{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 103 with $\\alpha_1 = 1$, $\\alpha_2 = 10$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 103\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020824.1</td>\n",
       "      <td>59372</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP020824.1        59372"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xl4VdW9//HPNwkJBBAFQjBMiUoC\nISFiAlTAEhQxXAW0OBDFVoVitVwHcODe2zrVoXqlKv1JC1hrVapMSlGxQP2ROlUlKCgEAlGZRAQE\nAmHMsO4f2aHHkIQTzc4J4f16Hh7O3nvtdb4HzkrO5+y19zbnnAAAAAAgLNQFAAAAAGgYCAcAAAAA\nJBEOAAAAAHgIBwAAAAAkEQ4AAAAAeAgHAAAAACQRDgAAAAB4CAcAAAAAJBEOAAAAAHgiQl1AbbVt\n29bFx8eHuowTxv79+9W8efNQlwE0eIwVIDiMldpZvnz5TudcTKjrAIJ1woWD+Ph45ebmhrqME0ZO\nTo4yMzNDXQbQ4DFWgOAwVmrHzDaGugagNphWBAAAAEAS4QAAAACAh3AAAAAAQBLhAAAAAICHcAAA\nAABAEuEAAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAAQBLhAAAAAICHcAAAAABAkhQR6gIaovhJb4S6\nhDrzXFbzUJcAAACAEwRHDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPBw\nKdPG7usV0n0jQl1F3bivMNQVAAAANGocOQAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAA\nIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAg\niXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgyedwYGZZZpZvZgVmNqmK\n7Z3NbKmZfWJmn5rZf/hZDwAAAIDq+RYOzCxc0tOShkpKlpRtZsmVmv1K0mznXC9JoyRN9aseAAAA\nADXz88hBH0kFzrkvnHNHJL0saUSlNk7SKd7jVpK2+lgPAAAAgBpE+Nh3B0mbA5a3SOpbqc19khab\n2X9Kai5psI/1AAAAAKiBn+HAqljnKi1nS3rOOTfZzM6V9IKZpTjnyr7Tkdk4SeMkKTY2Vjk5OX7U\ne9TE1BJf+69PRVFxykm6P9Rl1A2f/99xcisqKvL9ZwvQGDBWgMbNz3CwRVKngOWOOnba0BhJWZLk\nnPuXmTWV1FbS9sBGzrnpkqZLUkZGhsvMzPSp5HLXTXrD1/7r03NpXyoz/95Ql1E3sgtDXQEasZyc\nHPn9swVoDBgrQOPm5zkHyyR1NbMEM4tU+QnHCyq12STpAkkys+6Smkra4WNNAAAAAKrhWzhwzpVI\nGi9pkaQ1Kr8q0Woze8DMhnvNJkr6uZmtlPSSpOucc5WnHgEAAACoB35OK5JzbqGkhZXW3RPwOE9S\nfz9rAAAAABAc7pAMAAAAQBLhAAAAAICHcAAAAABAEuEAAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAA\nQBLhAAAAAICHcAAAAABAEuEAAAAAgIdwAAAAAECSFBHqAgAAaOziJ70R6hLqzHNZzUNdAgAfceQA\nAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP4QAA\nAACAJMIBAAAAAE9EqAsAAAAnkK9XSPeNCHUVdeO+wlBXADQ4HDkAAAAAIIlwAAAAAMBDOAAAAAAg\niXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJ\ncAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlw\nAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIMnncGBmWWaWb2YFZjapmjZX\nmlmema02s7/6WQ8AAACA6kX41bGZhUt6WtKFkrZIWmZmC5xzeQFtukr6L0n9nXO7zaydX/UAAAAA\nqJmfRw76SCpwzn3hnDsi6WVJIyq1+bmkp51zuyXJObfdx3oAAAAA1MDPcNBB0uaA5S3eukCJkhLN\n7D0z+8DMsnysBwAAAEANfJtWJMmqWOeqeP6ukjIldZT0jpmlOOf2fKcjs3GSxklSbGyscnJy6rzY\nQBNTS3ztvz4VRcUpJ+n+UJdRN3z+f8fJraioyPefLTh58XulgWLMA8fwMxxskdQpYLmjpK1VtPnA\nOVcs6Uszy1d5WFgW2Mg5N13SdEnKyMhwmZmZftUsSbpu0hu+9l+fnkv7Upn594a6jLqRXRjqCtCI\n5eTkyO+fLTh58XulgeL3CnAMP6cVLZPU1cwSzCxS0ihJCyq1mS9pkCSZWVuVTzP6wseaAAAAAFTD\nt3DgnCuRNF7SIklrJM12zq02swfMbLjXbJGkb80sT9JSSXc65771qyYAAAAA1fNzWpGccwslLay0\n7p6Ax07SBO8PAITO1yuk+ypfUO0EdR9TJQAA3w93SAYAAAAgyecjBwAat/hGdZJlqCsAACD0OHIA\nAAAAQBLhAAAAAICHcAAAAABAEuEAAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAAQBLhAAAAAICHcAAA\nAABAEuEAAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAAQBLhAAAAAICHcAAAAABAEuEAAAAAgIdwAAAA\nAEBSLcKBmQ0ws+u9xzFmluBfWQAAAADqW1DhwMzulXS3pP/yVjWR9KJfRQEAAACof8EeObhM0nBJ\n+yXJObdVUku/igIAAABQ/4INB0ecc06SkyQza+5fSQAAAABCIdhwMNvMpkk61cx+Lukfkmb4VxYA\nAACA+hYRTCPn3ONmdqGkvZKSJN3jnFvia2UAAAAA6tVxw4GZhUta5JwbLIlAAAAAADRSx51W5Jwr\nlXTAzFrVQz0AAAAAQiSoaUWSDkn6zMyWyLtikSQ5527xpSoAAAAA9S7YcPCG9wcAAABAIxXsCcl/\nMbNISYneqnznXLF/ZQEAAACob0GFAzPLlPQXSRskmaROZvYz59zb/pUGAAAAoD4FO61osqQhzrl8\nSTKzREkvSUr3qzAAAAAA9SvYm6A1qQgGkuScWyepiT8lAQAAAAiFYI8c5JrZnyS94C1fI2m5PyUB\nAAAACIVgw8FNkn4p6RaVn3PwtqSpfhUFAAAAoP4FGw4iJD3lnPuddPSuyVG+VQUAANBILV++vF1E\nRMQzklIU/BRvoC6USVpVUlIyNj09fXtVDYINB29JGiypyFtuJmmxpH4/uEQAAICTSERExDPt27fv\nHhMTszssLMyFuh6cPMrKymzHjh3J27Zte0bS8KraBJtWmzrnKoKBvMfRdVAjAADAySYlJiZmL8EA\n9S0sLMzFxMQUqvyoVdVtguxrv5mdU7FgZhmSDv7A+gAAAE5GYQQDhIr33qs2AwQ7rehWSXPMbKsk\nJylO0lU/vDwAAAAADUWw4SBBUi9JnSVdJulHKg8JAAAA+AHiJ71RpzeV3fDbixvM5eb79OmT9Pjj\nj2/+8Y9/fCCY9mVlZerXr1/iwoULC1q3bl3md30/1Ouvv95y8uTJsUuXLi3wo/+ZM2e2Wr16dbOH\nH354W+Vt0dHRvQ4cOPBJdfv269cv8W9/+9vnMTExpbV5zmCnFf3aObdX0qmSLpQ0XdIfavNEAAAA\nOHEUFxfX+3POnj27VY8ePQ6eCMGgPlxzzTWFVQWDYGRnZ3/7+OOPx9R2v2DDQUXiuFjSH51zf5MU\nWdsnAwAAQOjdeeedpyckJPTo169f12HDhiXcc889sVL5N/3jx4/v0Lt376QHH3wwduvWrREXXXTR\nmSkpKd1TUlK6L168uLkk7d27N+yKK66IT0lJ6d69e/fkF1988VRJKioqsksuueSMxMTE5IsvvviM\nQ4cOmSQ98cQTbceMGdOp4vknT57cduzYsR0r1zVz5szWl1122R5Jys/Pj0xISOhx1VVXdenatWuP\n4cOHJ8yfP7/lOeec061Lly4pS5cuja6plvz8/Mj09PSk5OTk7snJyd2XLFnSXCr/tr9Pnz5JWVlZ\nZyQkJPQYPnx4QllZzVlk5MiR8VdffXXn9PT0pPj4+JSXXnqpVeU2S5cuje7Vq1e37t27J/fq1avb\nypUroyQpNze3aWpqavdu3bolJyYmJn/22WdRwb62KVOmtPnpT3/aWZLWrl0befbZZ3dLSUnpfuut\nt8ZVPO/GjRubZGRkJHXr1i25a9euPf7+97+3kKRRo0bteeWVV9oE+ZY4KthpRV+Z2TSVX870UTOL\nEtflBQAAOOG8/fbb0a+99tppn332WV5xcbGdffbZyb169To67WfPnj3hy5Yty5ekYcOGJUyYMOGb\niy66qGj9+vWRF110Udcvvvhi9X//93+fPmjQoL1z5szZsHPnzvCMjIzuw4cP3/u73/0uplmzZmXr\n1q3L+/DDD5v1798/WZLGjBmzq0ePHsmHDx/eEhUV5V588cW206ZN21i5tuXLl7fo37//0fWbN29u\nOmvWrC/S09M39uzZs/vMmTPb5Obmrv3rX/966kMPPXT6oEGDPq+ulri4uJJ33nlnXXR0tPvss8+i\nsrOzz1i1atUaSVqzZk2zFStWfBEfH1+cnp7ebcmSJS0uuuiiosr1BNq8eXPURx99lJ+Xlxc1ePDg\npBEjRnwWuD0tLe3QRx99tLZJkyaaP39+y7vuuqvjokWLPv/9738fc/PNN39z00037Tp06JCVlJTo\nq6++ahLMawvs/+abb+48duzYHePHj//2kUceOXpE4Nlnn219wQUXFD766KPbSkpKtG/fvjBJiomJ\nKT1y5Iht27YtvH379kFPLQo2HFwpKUvS4865PWZ2uqQ7g30SAAAANAw5OTkthg4duqdFixZOkrvw\nwgv3BG7Pzs7eVfH4vffeO2X9+vXNKpaLiorCd+/eHZaTk3PKokWLTp0yZUp7STp8+LAVFBREvvvu\nuy1uueWW7ZLUt2/fg4mJiQck6ZRTTinr37//vlmzZrVKTU09VFxcbH369DnmypeFhYURp5122tGv\n8Tt06HC4ol1iYuLB888/f29YWJjOOeecAw8++GCc93qqrKVLly7FY8aM6ZKXl9csLCxMGzduPHoD\n39TU1P1nnnlmsST16NHjwOeff37cGTEjR47cFR4ertTU1MOdOnU6vGLFiqaB23ft2hV+1VVXJWzY\nsKGpmbni4mKTpHPPPXf/448/fvqWLVsiR40atTs1NfVwsK8t0Mcff9zizTff/FySbrzxxm9/85vf\ndJSkH/3oR/tvvPHG+OLi4rDLL798d79+/Y7+u7Zp06Zk06ZNke3btw/6KqNBhQPn3AFJrwQsfy3p\n62CfBAAAAA2DczVfU6Zly5ZHP5w755Sbm7vGCxLf6WPu3LkFaWlphyvvb2ZV9jtu3LidDz30UPvE\nxMRDo0eP3llVm/DwcFdaWqrw8HBJUmRk5NHnDQsLU9OmTZ3XTqWlpVZTLRMmTIhr165d8bx5874s\nKytTs2bNjp74HRUVdbTf8PBwlZSUVF10Da+r8vLdd9/dYeDAgfuWLFnyeX5+fuT555+fJEm/+MUv\ndp133nn7X3311VZDhw5NnDp16oakpKTDwby2yqq6BO7QoUOL3n777fx58+a1uu666xJuueWWb8aP\nH/+tVB6UoqOja3X+BlODAAAATiKZmZlFixYtanXgwAErLCwM+8c//nFqdW0HDBiw99FHH21Xsfz+\n++83k6RBgwbtnTx5cmzFXP333nuvmde+6MUXX2wtScuWLWu6bt26ozfNPf/88/d//fXXka+++mqb\nMWPG7FIVEhISDq1Zsyaqqm3Vqa6WwsLC8NNPP704PDxcU6dObVNaevyZNb/85S87PP/881X+e7zy\nyiunlZaWavXq1VGbN2+OSktLOxS4fe/eveEdO3Y8IknTpk1rW7E+Ly8vsnv37od/9atfbR8yZMie\nFStWNKvcdzDOOeecohkzZrSWpBkzZhw9l2DdunWRHTp0KJ44ceLO0aNH7/z444+jpfIrP+3YsaNJ\nUlLSMQGuJsFOKwIAAIAP6vvSowMHDjyQlZVVmJyc3KNDhw6He/bsub9Vq1ZVfnKePn365rFjx3ZO\nTExMLi0ttb59++7r16/fpt/+9rdbx40b17lbt27Jzjnr2LHj4aVLlxbccccd20eNGpWQmJiY3KNH\njwOpqan7A/u79NJLd3/66afR1V1ec8iQIYWLFy9umZKSEvQH2upque2227aPHDnyzPnz5582YMCA\nfc2aNTvuN+h5eXnNKk6Iruyss8463KdPn6Rvv/22yZNPPrkxOjr6O9/i33333dvGjh2bMGXKlPbn\nnXfe3or1L7zwQus5c+a0iYiIcDExMcWPPPLI1j179oQH+/oqTJ06ddOoUaPOmDp1auzw4cN3V6xf\ntGhRyylTprSPiIhw0dHRpTNnzvxSkt59993oXr167W/SpEmtnseOd2jphzCzLElPSQqX9Ixz7rfV\ntLtc0hxJvZ1zuTX1mZGR4XJza2zyg8VPesPX/uvTc2nrlZl/b6jLqBv3FYa6AlTCWGmgGCsNDmOl\ngaqHsWJmy51zGYHrVq5cuSEtLa3KaTX1pbCwMKxVq1Zl+/btCzv33HOT/vjHP24cMGBAUPci+CEG\nDRp01m233fbNiBEj9lW1fePGjU2ys7Pj33///fV+11KVAQMGdH333XePee6RI0fGX3LJJYXXX3/9\n7qr2a4iuv/76Tpdeeumeqv6tV65c2TYtLS2+qv18m1ZkZuGSnpY0VFKypGwzS66iXUtJt0j60K9a\nAAAA8G+jR4/u0q1bt+SePXt2HzZs2G6/g8HOnTvD4+PjU5o2bVpWXTCQpC5duhTfcMMNO3ft2hWS\nqe9VBYMTVUpKysGa/q2r4+e0oj6SCpxzX0iSmb0saYSkvErtfiPpMUl3+FgLAAAAPK+99tqX9fl8\nbdu2Ld2wYcOqYNqOHTu2wX07P2/evA2hrqG2Jk6c+L2OTvmZyjpI2hywvMVbd5SZ9ZLUyTn3uo91\nAAAAAAiCn0cOqroE09ETHMwsTNITkq47bkdm4ySNk6TY2Fjl5OTUTYXVmJha4mv/9akoKk45SfeH\nuoy64fP/O2qPsdJAMVYaHMZKA8VYAY7hZzjYIqlTwHJHSVsDlltKSpGU410ntr2kBWY2vPJJyc65\n6ZKmS+UnJGdmZvpYtnRdozpx7MvGc+JYNidZNjSMlQaKsdLgMFYaKMYKcAw/pxUtk9TVzBLMLFLS\nKEkLKjY65wqdc22dc/HOuXhJH0g6JhgAAAAAqB++HTlwzpWY2XhJi1R+KdNnnXOrzewBSbnOuQU1\n9wAAAHASuK9V+vEb1aa/wuPeN8HM0keMGLFr/vz5X0pScXGx2rVrl3b22WfvX7p0acGUKVPa5Obm\nNn/++ec3TZgwIa5FixalDzzwwDeBfYSHh6d37dr1YGlpqZ111lkHZ8+evSHw7spS+Y24+vXrl7hw\n4cKC1q1bl11xxRXxb731Vqs2bdqUrF+/fnVVtX3yySdNr7/++vi8vLzoSZMmfRX4vHPnzj3ljjvu\n6FxWVqbRo0fvfPjhh7dV3v/NN99sMXHixE7r1q2LnjFjxhcVlx/dunVrxFVXXZXwzjvvNJorEvnB\n18tEOecWOucSnXNnOuce8tbdU1UwcM5lctQAAADAf82aNSvLz89vVlRUZJL06quvnhIbG1tcmz6i\noqLK1q5dm7d+/frVTZo0cZMnT46p3Gb27NmtevTocbB169ZlknTDDTfsXLBgQY0fztu1a1fy1FNP\nbbrxxhu/E0ZKSkp0++23d164cOG6devWrZ43b17r5cuXN628/xlnnHHkz3/+84Zhw4Z9G7g+Li6u\nJDY2tnjx4sXNa/M6TzYhuYYsAAAAQuuCCy4onDNnzqmS9NJLL7UeOXLkru/b14ABA4oKCgqiKq+f\nOXNm68A7Dg8dOrQoJiamxjP0O3ToUDJw4MADTZo0+c6denNycpp36dLlcHJy8pGmTZu6n/zkJ7vm\nzp17auX9k5KSjvTt2/dgWNixH3MvvfTSPc8//3ybWr24kwzhAAAA4CR07bXX7po1a9ZpBw4csDVr\n1kSfe+65+79PP8XFxVq0aNEpqampBytvW758eYv+/fsft9/HHnss5rHHHjvmyEOgzZs3R3bo0OFI\nxXLHjh2PfPXVV5GSdNttt8XNnDmz1fGep3///vs/+uijFsdrdzLz82pFAAAAaKD69u17cMuWLVEz\nZsxoPXjw4Fpfuunw4cNh3bp1S/b62nfrrbcec9OtwsLCiNNOO63s2L2/66677tpxvDbOuWPWmZmT\npCeffHLrMRurEBcXV7J9+/bIYNqerAgHAAAAJ6msrKw99957b6fFixfnb9++vVafCyvOOaipTXh4\nuCstLVV4ePgPK1RS586djx4pkKQtW7ZExsXF1eo8iQMHDlhUVNRxw8rJjGlFAAAAJ6mbbrpp58SJ\nE7f26dPnmClBdSEhIeHQmjVrjjkX4fsYOHDg/g0bNjRdu3Zt5KFDh+yVV15pPXLkyD3H3/PfVq1a\n1TQxMdGX19pYcOQAAAAglIK49KhfzjzzzOJf//rX24/X7oknnjh92rRpsRXL33zzzafB9D9kyJDC\nxYsXt0xJSTksScOGDUv44IMPWu7evTsiNja256RJk7befvvtOyvON7jrrrt2bNq0KaJ3797J+/fv\nDzczN23atNg1a9asat26ddnkyZM3ZWVlJZaWlurqq6/emZGRcUgqP+egd+/e+6+55prCf/7zn9FX\nXnnlWXv37g1/6623Tn3ooYfiCgoKVkvSkiVLWmZlZXH3uxpYVfO3GrKMjAyXm+vvFU/jG9WdLNc3\nnjtZ3sdYbmgYKw0UY6XBYaw0UPUwVsxsuXMuI3DdypUrN6SlpR0zP78x2rhxY5Ps7Oz4999/v0Hc\nWyAjIyPpzTffLIiJiSkNdS2htHLlyrZpaWnxVW1jWhEAAAB80aVLl+Ibbrhh565du0L+mXPr1q0R\nt9566zcnezA4HqYVAQAAwDdjx47dHeoapPIrFV177bW1OkfhZBTyFAcAAHCSKSsrK7NQF4GTk/fe\nq/aKTYQDAACA+rVqx44drQgIqG9lZWW2Y8eOVpJWVdeGaUUAAAD1qKSkZOy2bdue2bZtW4r4ohb1\nq0zSqpKSkrHVNSAcAAAA1KP09PTtkoaHug6gKqRVAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyE\nAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQD\nAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMA\nAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAA\nAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSfA4HZpZlZvlmVmBm\nk6rYPsHM8szsUzN7y8y6+FkPAAAAgOr5Fg7MLFzS05KGSkqWlG1myZWafSIpwznXU9JcSY/5VQ8A\nAACAmvl55KCPpALn3BfOuSOSXpY0IrCBc26pc+6At/iBpI4+1gMAAACgBn6Ggw6SNgcsb/HWVWeM\npDd9rAcAAABADSJ87NuqWOeqbGg2WlKGpIHVbB8naZwkxcbGKicnp45KrNrE1BJf+69PRVFxykm6\nP9Rl1A2f/99Re4yVBoqx0uAwVhooxgpwDD/DwRZJnQKWO0raWrmRmQ2W9D+SBjrnDlfVkXNuuqTp\nkpSRkeEyMzPrvNhA1016w9f+69NzaV8qM//eUJdRN7ILQ10BKmGsNFCMlQaHsdJAMVaAY/g5rWiZ\npK5mlmBmkZJGSVoQ2MDMekmaJmm4c267j7UAAAAAOA7fwoFzrkTSeEmLJK2RNNs5t9rMHjCz4V6z\n/5XUQtIcM1thZguq6Q4AAACAz/ycViTn3EJJCyutuyfg8WA/nx8AAABA8LhDMgAAAABJhAMAAAAA\nHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAe\nwgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7C\nAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIB\nAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEA\nAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAA\nAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB5fw4GZZZlZvpkVmNmkKrZHmdks\nb/uHZhbvZz0AAAAAqudbODCzcElPSxoqKVlStpklV2o2RtJu59xZkp6Q9Khf9QAAAAComZ9HDvpI\nKnDOfeGcOyLpZUkjKrUZIekv3uO5ki4wM/OxJgAAAADV8DMcdJC0OWB5i7euyjbOuRJJhZLa+FgT\nAAAAgGpE+Nh3VUcA3PdoIzMbJ2mct1hkZvk/sLaTxiCpraSdoa6jTtzPQSX4h7ECBIexUmtd6uNJ\ngLriZzjYIqlTwHJHSVurabPFzCIktZK0q3JHzrnpkqb7VGejZma5zrmMUNcBNHSMFSA4jBWgcfNz\nWtEySV3NLMHMIiWNkrSgUpsFkn7mPb5c0v93zh1z5AAAAACA/3w7cuCcKzGz8ZIWSQqX9KxzbrWZ\nPSAp1zm3QNKfJL1gZgUqP2Iwyq96AAAAANTM+KK+cTOzcd60LAA1YKwAwWGsAI0b4QAAAACAJJ/v\nkAwAAADgxEE4AAAAACCJcHCsFBNHAAAEtklEQVTSM7N4M1vlPc40s9dDXRMQCmb2/nG2LzSzU+ur\nHqAxMLMcM8vwHm8ws7ahrglAzfy8zwF8ZGam8nNGykJdC9DQmFm4c660Nvs45/odZ/t//LCqgMaH\n30VA48ORgxOI9y3/GjObKuljSdea2b/M7GMzm2NmLbx2vc3sfTNbaWYfmVlLb993vLYfm1mNH4SA\nhsp7L681s7+Y2admNtfMor1vJe8xs3clXWFmZ5rZ381suffe7+btH2tmr3rjY2XFWDCzIu/v083s\nbTNbYWarzOw8b/3Rbz3NbIK3bZWZ3RZQ1xozm2Fmq81ssZk1C8k/ElCDyu9fM3vUzG4O2H6fmU30\nHt9pZsu8sXa/t67y76JOZvYHM8v13vv3h+aVAagLhIMTT5Kk5yVdKGmMpMHOuXMk5Uqa4N1wbpak\nW51zaZIGSzooabukC722V0maEorigTqSJGm6c66npL2SKj7YHHLODXDOvazyu6r/p3MuXdIdkqZ6\nbaZI+qc3Ps6RtLpS31dLWuScO1tSmqQVgRvNLF3S9ZL6SvqRpJ+bWS9vc1dJTzvnekjaI2lkXb1g\noC5U9f6V9LLKfy9UuFLSHDMbovL3dB9JZ0tKN7Mfe22SJD3vnOvlnNso6X+8uyb3lDTQzHrWywsC\nUOeYVnTi2eic+8DMLpGULOm98qO6ipT0L5X/wP7aObdMkpxzeyXJzJpL+n9mdrakUkmJoSgeqCOb\nnXPveY9flHSL93iWJHlH0fqp/ANOxT5R3t/nS/qpJHlTjwor9b1M0rNm1kTSfOfcikrbB0h61Tm3\n33uuVySdp/I7vn8Z0H65pPgf8BoBP1T3/m1nZnGSYiTtds5tMrNbJA2R9Im3bwuVh4VN8n4XBfR7\npZmNU/nnitNV/vvp0/p4QQDqFuHgxLPf+9skLXHOZQdu9L6tqermFbdL+kbl34SGSTrkZ5GAzyq/\nxyuWK8ZHmKQ93rf/tevYube9b0cvVvkd3P/XOfd8QBOrZldJOhzwuFQS04rQ0FT3/p0r6XJJ7VV+\nJKGi7SPOuWnf6cAsXv8eazKzBJUfnevtnNttZs9JalqnVQOoN0wrOnF9IKm/mZ0lSd6c60RJayXF\nmVlvb31LM4uQ1ErlRxTKJF0rKTxEdQN1obOZnes9zpb0buBG74jZl2Z2hVR+0qSZpXmb35J0k7c+\n3MxOCdzXzLpI2u6cmyHpTyqfehTobUmXemOuuaTLJL1Tdy8N8FV179+XJY1SeUCY67VdJOmGgPPZ\nOphZuyr6PEXlYaHQzGIlDfX5NQDwEeHgBOWc2yHpOkkvmdmnKg8L3ZxzR1Q+d/T3ZrZS0hKVf4Mz\nVdLPzOwDlU8p2l9lx8CJYY3K38+fSmot6Q9VtLlG0hhvHKyWNMJbf6ukQWb2mcqn/vSotF+mpBVm\n9onKzxl4KnCjc+5jSc9J+kjSh5Kecc59IuAEUN371zm3WlJLSV8557722i6W9FdJ//LGy1yvTeU+\nV6p86tFqSc9Keq9yGwAnDnOuqhkoANAweVMaXnfOpYS4FAAAGh2OHAAAAACQxJEDAAAAAB6OHAAA\nAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkCT9H7HrkPctvbaSAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f34e4c497f0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:10:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>37700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>50571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>3166</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0        37700\n",
       "1  plasmid_1        50571\n",
       "2  plasmid_2         3166"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.881796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.583352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.702179</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.881796\n",
       "1  precision  0.583352\n",
       "2   f1_score  0.702179"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP020824.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.851765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.046638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.898403</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP020824.1\n",
       "0  plasmid_0    0.000000\n",
       "1  plasmid_1    0.851765\n",
       "2  plasmid_2    0.046638\n",
       "3      Total    0.898403"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020824.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.874605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.874605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2\n",
       "0  CP020824.1        0.0        1.0   0.874605\n",
       "1       Total        0.0        1.0   0.874605"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.874605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.000000\n",
       "1  plasmid_1    1.000000\n",
       "2  plasmid_2    0.874605"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020824.1</td>\n",
       "      <td>0.881796</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP020824.1    0.881796"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020824.1</td>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.851765</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction    recall  precision\n",
       "0  CP020824.1  plasmid_1  0.851765        1.0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 10, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0110832</td>\n",
       "      <td>-0.00034013</td>\n",
       "      <td>0.00101681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[43, 26]</td>\n",
       "      <td>[40, 22]</td>\n",
       "      <td>[36, 46, 40]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.694261</td>\n",
       "      <td>-0.757027</td>\n",
       "      <td>-0.622523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.705344</td>\n",
       "      <td>-0.757368</td>\n",
       "      <td>-0.621507</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        plasmid_0   plasmid_1     plasmid_2\n",
       "GC_sum -0.0110832 -0.00034013    0.00101681\n",
       "chain    [43, 26]    [40, 22]  [36, 46, 40]\n",
       "gd_sum  -0.694261   -0.757027     -0.622523\n",
       "total   -0.705344   -0.757368     -0.621507"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0110238</td>\n",
       "      <td>-0.000104619</td>\n",
       "      <td>-0.000402845</td>\n",
       "      <td>0.000940107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[43, 39, 43, 26]</td>\n",
       "      <td>[52, 36, 52, 42, 47, 40, 22]</td>\n",
       "      <td>[46, 40, 4, 57]</td>\n",
       "      <td>[40]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.667969</td>\n",
       "      <td>-0.758001</td>\n",
       "      <td>-0.00221128</td>\n",
       "      <td>-0.685045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.678993</td>\n",
       "      <td>-0.758105</td>\n",
       "      <td>-0.00261412</td>\n",
       "      <td>-0.684105</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               plasmid_0                     plasmid_1        plasmid_2  \\\n",
       "GC_sum        -0.0110238                  -0.000104619     -0.000402845   \n",
       "chain   [43, 39, 43, 26]  [52, 36, 52, 42, 47, 40, 22]  [46, 40, 4, 57]   \n",
       "gd_sum         -0.667969                     -0.758001      -0.00221128   \n",
       "total          -0.678993                     -0.758105      -0.00261412   \n",
       "\n",
       "          plasmid_3  \n",
       "GC_sum  0.000940107  \n",
       "chain          [40]  \n",
       "gd_sum    -0.685045  \n",
       "total     -0.684105  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_0  from MILP completely contained in  plasmid_0  from greedy.\n",
      "\n",
      "plasmid_1  from MILP completely contained in  plasmid_1  from greedy.\n",
      "\n",
      "plasmid_3  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_3  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
