{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 23 with $\\alpha_1 = 1$, $\\alpha_2 = 10$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 23\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020836.1</td>\n",
       "      <td>37362</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP020836.1        37362"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt8FOW9x/HvLwkJBBAFQpBwSVQC\nhIQICaCAJShKqAJavIBiq0KxWuoFrHLOaa21XqpHqtIjLWCtRakKqBQVC9RD6q0qQUEg3KJyiYiA\nQDAgkMtz/siEs4ZN2NRMliSf9+vFy52ZZ5/57b72MfvdeWbGnHMCAAAAgIhwFwAAAADg5EA4AAAA\nACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAExXuAmqq\nbdu2LjExMdxl1BsHDx5U8+bNw10GcNJjrAChYazUzMqVK/c45+LCXQcQqnoXDhITE5WbmxvuMuqN\nnJwcZWVlhbsM4KTHWAFCw1ipGTPbGu4agJpgWhEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIB\nAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7fwoGZPWVmu8xsbRXbzcymm1m+mX1sZn38qgUAAADA\nifl55OBpSdnVbB8uqav3b6KkP/hYCwAAAIAT8C0cOOfelLS3miajJM1x5d6TdKqZne5XPQAAAACq\nFxXGfSdI2h6wXOCt+6JyQzObqPKjC4qPj1dOTo6vha35vNDX/utSUqtI398voCEoKipirAAhYKwA\nDVs4w4EFWeeCNXTOzZI0S5IyMzNdVlaWj2VJ1019zdf+69LT2c3l9/sFNAQ5OTmMFSAEjBWgYQvn\n1YoKJHUKWO4oaUeYagEAAAAavXCGg0WSfuhdtegcSYXOueOmFAEAAACoG75NKzKz5yRlSWprZgWS\nfiWpiSQ55/4oabGk70vKl3RI0vV+1QIAAADgxHwLB865sSfY7iT91K/9AwAAAKgZ7pAMAAAAQBLh\nAAAAAICHcAAAAABAEuEAAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAAQBLhAAAAAICHcAAAAABAEuEA\nAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAAQBLhAAAAAICHcAAAAABAEuEAAAAAgIdwAAAAAEAS4QAA\nAACAh3AAAAAAQBLhAAAAAICHcAAAAABAEuEAAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAAQBLhAAAA\nAICHcAAAAABAEuEAAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAAQBLhAAAAAICHcAAAAABAEuEAAAAA\ngIdwAAAAAEAS4QAAAACAJyrcBQAA0NAlTn0t3CXUmqezm4e7BAA+4sgBAAAAAEmEAwAAAAAewgEA\nAAAASYQDAAAAAB7CAQAAAABJPocDM8s2s41mlm9mU4Ns72xmy83sIzP72My+72c9AAAAAKrmWzgw\ns0hJT0gaLilF0lgzS6nU7BeS5jnneksaI2mGX/UAAAAAqJ6fRw76Scp3zn3qnDsq6XlJoyq1cZJO\n8R63krTDx3oAAAAAVMPPm6AlSNoesFwgqX+lNvdIWmpmP5PUXNJQH+sBAAAAUA0/w4EFWecqLY+V\n9LRzbpqZnSvpGTNLdc6Vfasjs4mSJkpSfHy8cnJy/Kj3mClpJb72X5eKiop8f7+AhoCxAj/xdwVA\nfeFnOCiQ1ClguaOOnzY0XlK2JDnn/mVmTSW1lbQrsJFzbpakWZKUmZnpsrKyfCq53HUN7Db3fr9f\nQEOQk5PDWIFv+LsCoL7w85yDFZK6mlmSmUWr/ITjRZXabJN0gSSZWQ9JTSXt9rEmAAAAAFXwLRw4\n50okTZK0RNJ6lV+VaJ2Z3WtmI71mUyT92MxWS3pO0nXOucpTjwAAAADUAT+nFck5t1jS4krr7g54\nnCdpoJ81AAAAAAgNd0gGAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgA\nAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAA\nAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAA\nACCJcAAAAADAQzgAAAAAIIlwAAAAAMATFe4CANRfiVNfC3cJtebp7ObhLgEAgLDjyAEAAAAASYQD\nAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMA\nAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHl/DgZllm9lGM8s3\ns6lVtLnSzPLMbJ2Z/dXPegAAAABULcqvjs0sUtITki6UVCBphZktcs7lBbTpKuk/JA10zu0zs3Z+\n1QMAAACgen4eOegnKd8596lz7qik5yWNqtTmx5KecM7tkyTn3C4f6wEAAABQDd+OHEhKkLQ9YLlA\nUv9KbZIlyczekRQp6R7n3N8rd2RmEyVNlKT4+Hjl5OT4Ue8xU9JKfO2/LhUVFfn+fqHxYqwAoWGs\nAKgv/AwHFmSdC7L/rpKyJHWU9JaZpTrn9n/rSc7NkjRLkjIzM11WVlatFxvouqmv+dp/XXo6u7n8\nfr/QeDFWgNAwVgDUF35OKyqQ1ClguaOkHUHa/M05V+yc+0zSRpWHBQAAAAB1zM9wsEJSVzNLMrNo\nSWMkLarUZqGkIZJkZm1VPs3oUx9rAgAAAFAF38KBc65E0iRJSyStlzTPObfOzO41s5FesyWSvjKz\nPEnLJf3cOfeVXzUBAAAAqJqf5xzIObdY0uJK6+4OeOwkTfb+AQAAAAgj7pAMAAAAQBLhAAAAAICH\ncAAAAABAUg3CgZkNMrPrvcdxZpbkX1kAAAAA6lpI4cDMfiXpLkn/4a1qIulZv4oCAAAAUPdCPXJw\nmaSRkg5KknNuh6SWfhUFAAAAoO6FGg6OepcddZJkZs39KwkAAABAOIQaDuaZ2UxJp5rZjyX9Q9Js\n/8oCAAAAUNdCugmac+4RM7tQ0gFJ3STd7Zxb5mtlAAAAAOrUCcOBmUVKWuKcGyqJQAAAAAA0UCec\nVuScK5V0yMxa1UE9AAAAAMIkpGlFkg5LWmNmy+RdsUiSnHO3+FIVAAAAgDoXajh4zfsHAAAAoIEK\n9YTkv5hZtKRkb9VG51yxf2UBAAAAqGshhQMzy5L0F0lbJJmkTmb2I+fcm/6VBgAAAKAuhTqtaJqk\ni5xzGyXJzJIlPScpw6/CAAAAANStUG+C1qQiGEiSc26TpCb+lAQAAAAgHEI9cpBrZn+S9Iy3fI2k\nlf6UBAAAACAcQg0HN0n6qaRbVH7OwZuSZvhVFAAAAIC6F2o4iJL0uHPud9KxuybH+FYVAABAA7Vy\n5cp2UVFRT0pKVehTvIHaUCZpbUlJyYSMjIxdwRqEGg7ekDRUUpG33EzSUkkDvnOJAAAAjUhUVNST\n7du37xEXF7cvIiLChbseNB5lZWW2e/fulJ07dz4paWSwNqGm1abOuYpgIO9xbC3UCAAA0NikxsXF\nHSAYoK5FRES4uLi4QpUftQreJsS+DppZn4oFM8uU9M13rA8AAKAxiiAYIFy8z16VGSDUaUW3Sppv\nZjskOUkdJF313csDAAAAcLIINRwkSeotqbOkyySdo/KQAAAAgO8gceprtXpT2S2/vfikudx8v379\nuj3yyCPbv/e97x0KpX1ZWZkGDBiQvHjx4vzWrVuX+V3fd/Xqq6+2nDZtWvzy5cvz/eh/7ty5rdat\nW9fsgQce2Fl5W2xsbO9Dhw59VNVzBwwYkPy3v/3tk7i4uNKa7DPUaUW/dM4dkHSqpAslzZL0h5rs\nCAAAAPVHcXFxne9z3rx5rXr27PlNfQgGdeGaa64pDBYMQjF27NivHnnkkbiaPi/UcFCROC6W9Efn\n3N8kRdd0ZwAAAAi/n//856cnJSX1HDBgQNcRI0Yk3X333fFS+S/9kyZNSujbt2+3++67L37Hjh1R\nw4YNOzM1NbVHampqj6VLlzaXpAMHDkRcccUViampqT169OiR8uyzz54qSUVFRXbJJZeckZycnHLx\nxRefcfjwYZOkRx99tO348eM7Vex/2rRpbSdMmNCxcl1z585tfdlll+2XpI0bN0YnJSX1vOqqq7p0\n7dq158iRI5MWLlzYsk+fPt27dOmSunz58tjqatm4cWN0RkZGt5SUlB4pKSk9li1b1lwq/7W/X79+\n3bKzs89ISkrqOXLkyKSysuqzyOjRoxOvvvrqzhkZGd0SExNTn3vuuVaV2yxfvjy2d+/e3Xv06JHS\nu3fv7qtXr46RpNzc3KZpaWk9unfvnpKcnJyyZs2amFBf2/Tp09v88Ic/7CxJGzZsiD777LO7p6am\n9rj11ls7VOx369atTTIzM7t17949pWvXrj3//ve/t5CkMWPG7H/ppZfahPiROCbUaUWfm9lMlV/O\n9CEzixHX5QUAAKh33nzzzdhXXnnltDVr1uQVFxfb2WefndK7d+9j0372798fuWLFio2SNGLEiKTJ\nkyd/OWzYsKLNmzdHDxs2rOunn3667j//8z9PHzJkyIH58+dv2bNnT2RmZmaPkSNHHvjd734X16xZ\ns7JNmzblvf/++80GDhyYIknjx4/f27Nnz5QjR44UxMTEuGeffbbtzJkzt1aubeXKlS0GDhx4bP32\n7dubvvDCC59mZGRs7dWrV4+5c+e2yc3N3fDXv/711Pvvv//0IUOGfFJVLR06dCh56623NsXGxro1\na9bEjB079oy1a9eul6T169c3W7Vq1aeJiYnFGRkZ3ZctW9Zi2LBhRZXrCbR9+/aYDz74YGNeXl7M\n0KFDu40aNWpN4Pb09PTDH3zwwYYmTZpo4cKFLe+8886OS5Ys+eT3v/993M033/zlTTfdtPfw4cNW\nUlKizz//vEkory2w/5tvvrnzhAkTdk+aNOmrBx988NgRgaeeeqr1BRdcUPjQQw/tLCkp0ddffx0h\nSXFxcaVHjx61nTt3RrZv3z7kqUWhhoMrJWVLesQ5t9/MTpf081B3AgAAgJNDTk5Oi+HDh+9v0aKF\nk+QuvPDC/YHbx44du7fi8TvvvHPK5s2bm1UsFxUVRe7bty8iJyfnlCVLlpw6ffr09pJ05MgRy8/P\nj3777bdb3HLLLbskqX///t8kJycfkqRTTjmlbODAgV+/8MILrdLS0g4XFxdbv379jrvyZWFhYdRp\np5127Gf8hISEIxXtkpOTvzn//PMPREREqE+fPofuu+++Dt7rCVpLly5disePH98lLy+vWUREhLZu\n3XrsBr5paWkHzzzzzGJJ6tmz56FPPvnkhDNiRo8evTcyMlJpaWlHOnXqdGTVqlVNA7fv3bs38qqr\nrkrasmVLUzNzxcXFJknnnnvuwUceeeT0goKC6DFjxuxLS0s7EuprC/Thhx+2eP311z+RpBtvvPGr\n3/zmNx0l6Zxzzjl44403JhYXF0dcfvnl+wYMGHDsfW3Tpk3Jtm3botu3bx/yVUZDCgfOuUOSXgpY\n/kLSF6HuBAAAACcH56q/pkzLli2PfTl3zik3N3e9FyS+1ceCBQvy09PTj1R+vpkF7XfixIl77r//\n/vbJycmHx40btydYm8jISFdaWqrIyEhJUnR09LH9RkREqGnTps5rp9LSUquulsmTJ3do165d8Ysv\nvvhZWVmZmjVrduzE75iYmGP9RkZGqqSkJHjR1byuyst33XVXwuDBg79etmzZJxs3bow+//zzu0nS\nT37yk73nnXfewZdffrnV8OHDk2fMmLGlW7duR0J5bZUFuwTu8OHDi958882NL774Yqvrrrsu6ZZb\nbvly0qRJX0nlQSk2NrZG528wNQgAAKARycrKKlqyZEmrQ4cOWWFhYcQ//vGPU6tqO2jQoAMPPfRQ\nu4rld999t5kkDRky5MC0adPiK+bqv/POO8289kXPPvtsa0lasWJF002bNh27ae75559/8Isvvoh+\n+eWX24wfP36vgkhKSjq8fv36mGDbqlJVLYWFhZGnn356cWRkpGbMmNGmtPTEM2t++tOfJsyZMyfo\n+/HSSy+dVlpaqnXr1sVs3749Jj09/XDg9gMHDkR27NjxqCTNnDmzbcX6vLy86B49ehz5xS9+seui\niy7av2rVqmaV+w5Fnz59imbPnt1akmbPnn3sXIJNmzZFJyQkFE+ZMmXPuHHj9nz44YexUvmVn3bv\n3t2kW7duxwW46oQ6rQgAAAA+qOtLjw4ePPhQdnZ2YUpKSs+EhIQjvXr1OtiqVaug35xnzZq1fcKE\nCZ2Tk5NTSktLrX///l8PGDBg229/+9sdEydO7Ny9e/cU55x17NjxyPLly/PvuOOOXWPGjElKTk5O\n6dmz56G0tLSDgf1deuml+z7++OPYqi6vedFFFxUuXbq0ZWpqashfaKuq5bbbbts1evToMxcuXHja\noEGDvm7WrNkJf0HPy8trVnFCdGVnnXXWkX79+nX76quvmjz22GNbY2Njv/Ur/l133bVzwoQJSdOn\nT29/3nnnHahY/8wzz7SeP39+m6ioKBcXF1f84IMP7ti/f39kqK+vwowZM7aNGTPmjBkzZsSPHDly\nX8X6JUuWtJw+fXr7qKgoFxsbWzp37tzPJOntt9+O7d2798EmTZrUaD92okNLJ5vMzEyXm5vr6z4S\np77ma/916ens5srKygp3GWigGCtAaBgrjZeZrXTOZQauW7169Zb09PSg02rqSmFhYUSrVq3Kvv76\n64hzzz232x//+MetgwYNCuleBN/FkCFDzrrtttu+HDVq1NfBtm/durXJ2LFjE999993NftcSzKBB\ng7q+/fbbx+179OjRiZdccknh9ddfvy/Y805G119/fadLL710f7D3evXq1W3T09MTgz2PaUUAAACN\nzLhx47p07949pVevXj1GjBixz+9gsGfPnsjExMTUpk2bllUVDCSpS5cuxTfccMOevXv3huU7arBg\nUF+lpqZ+U917XRWmFQEAADQyr7zyymd1ub+2bduWbtmyZW0obSdMmHDS/Tr/4osvbgl3DTU1ZcqU\nf+voFEcOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPBwQjIAAEA43dMq48SNatJf4Qnvm2BmGaNG\njdq7cOHCzySpuLhY7dq1Sz/77LMPLl++PH/69OltcnNzm8+ZM2fb5MmTO7Ro0aL03nvv/TKwj8jI\nyIyuXbt+U1paamedddY38+bN2xJ4d2Wp/EZcAwYMSF68eHF+69aty6644orEN954o1WbNm1KNm/e\nvC5YbR999FHT66+/PjEvLy926tSpnwfud8GCBafccccdncvKyjRu3Lg9DzzwwM7Kz3/99ddbTJky\npdOmTZtiZ8+e/WnF5Ud37NgRddVVVyW99dZbDeaKRH7gyAEAAEAj06xZs7KNGzc2KyoqMkl6+eWX\nT4mPjy+uSR8xMTFlGzZsyNu8efO6Jk2auGnTpsVVbjNv3rxWPXv2/KZ169ZlknTDDTfsWbRoUbVf\nztu1a1fy+OOPb7vxxhu/FUZKSkp0++23d168ePGmTZs2rXvxxRdbr1y5smnl559xxhlH//znP28Z\nMWLEV4HrO3ToUBIfH1+8dOnS5jV5nY0N4QAAAKARuuCCCwrnz59/qiQ999xzrUePHr333+1r0KBB\nRfn5+TGV18+dO7d14B2Hhw8fXhQXF1dSXV8JCQklgwcPPtSkSZNv3ak3JyeneZcuXY6kpKQcbdq0\nqfvBD36wd8GCBadWfn63bt2O9u/f/5uIiOO/5l566aX758yZ06ZGL66RIRwAAAA0Qtdee+3eF154\n4bRDhw7Z+vXrY88999yD/04/xcXFWrJkySlpaWnfVN62cuXKFgMHDjxhvw8//HDcww8/fNyRh0Db\nt2+PTkhIOFqx3LFjx6Off/55tCTddtttHebOndvqRPsZOHDgwQ8++KDFido1Zr6ec2Bm2ZIelxQp\n6Unn3G+raHe5pPmS+jrncv2sCQAAAFL//v2/KSgoiJk9e3broUOHFtb0+UeOHIno3r17itfX17fe\neutxN90qLCyMOu2008qOf/a33XnnnbtP1MY5d9w6M3OS9Nhjj+0IpeYOHTqU7Nq1KzqUto2Vb+HA\nzCIlPSHpQkkFklaY2SLnXF6ldi0l3SLpfb9qAQAAwPGys7P3/+pXv+q0dOnSjbt27arR98KKcw6q\naxMZGelKS0sVGRn53QqV1Llz52NHCiSpoKAgukOHDjU6T+LQoUMWExNzwrDSmPk5raifpHzn3KfO\nuaOSnpc0Kki730h6WNJhH2sBAABAJTfddNOeKVOm7OjXr99xU4JqQ1JS0uH169cfdy7Cv2Pw4MEH\nt2zZ0nTDhg3Rhw8ftpdeeqn16NGj95/4mf9v7dq1TZOTk315rQ2Fn9OKEiRtD1gukNQ/sIGZ9ZbU\nyTn3qpndUVVHZjZR0kRJio+PV05OTu1XG2BKWrXnydQrRUVFvr9faLwYK0BoGCuoVgiXHvXLmWee\nWfzLX/5y14naPfroo6fPnDkzvmL5yy+//DiU/i+66KLCpUuXtkxNTT0iSSNGjEh67733Wu7bty8q\nPj6+19SpU3fcfvvteyrON7jzzjt3b9u2Lapv374pBw8ejDQzN3PmzPj169evbd26ddm0adO2ZWdn\nJ5eWlurqq6/ek5mZeVgqP+egb9++B6+55prCf/7zn7FXXnnlWQcOHIh84403Tr3//vs75Ofnr5Ok\nZcuWtczOzq7xFKrGxILN36qVjs2ukDTMOTfBW75WUj/n3M+85QhJ/yvpOufcFjPLkXTHic45yMzM\ndLm5/p6WkDj1NV/7r0tPZzdXVlZWuMtAA8VYAULDWGm8zGylcy4zcN3q1au3pKenHzc/vyHaunVr\nk7Fjxya+++67J8W9BTIzM7u9/vrr+XFxcaXhriWcVq9e3TY9PT0x2DY/pxUVSOoUsNxRUuDJIi0l\npUrKMbMtks6RtMjMvjWAAAAAUD916dKl+IYbbtizd+/esF8hc8eOHVG33nrrl409GJyIn9OKVkjq\namZJkj6XNEbS1RUbnXOFktpWLId65AAAAAD1x4QJE/aFuwap/EpF1157bY3OUWiMfEtxzrkSSZMk\nLZG0XtI859w6M7vXzEb6tV8AAICTXFlZWZmFuwg0Tt5nr8orNvl6nwPn3GJJiyutu7uKtll+1gIA\nAHCSWLt79+6UuLi4woiICH9O/gSCKCsrs927d7eStLaqNr6GAwAAAHxbSUnJhJ07dz65c+fOVPl7\n/idQWZmktSUlJROqakA4AAAAqEMZGRm7JDHFGicl0ioAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAA\nHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAe\nwgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7C\nAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIB\nAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEA\nAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEk+hwMzyzazjWaWb2ZTg2yfbGZ5Zvaxmb1h\nZl38rAcAAABA1XwLB2YWKekJScMlpUgaa2YplZp9JCnTOddL0gJJD/tVDwAAAIDq+XnkoJ+kfOfc\np865o5KelzQqsIFzbrlz7pC3+J6kjj7WAwAAAKAaUT72nSBpe8BygaT+1bQfL+n1YBvMbKKkiZIU\nHx+vnJycWioxuClpJb72X5eKiop8f7/QeDFWgNAwVgDUF36GAwuyzgVtaDZOUqakwcG2O+dmSZol\nSZmZmS4rK6uWSgzuuqmv+dp/XXo6u7n8fr/QeDFWgNAwVgDUF36GgwJJnQKWO0raUbmRmQ2V9F+S\nBjvnjvhYDwAAAIBq+HnOwQpJXc0sycyiJY2RtCiwgZn1ljRT0kjn3C4fawEAAABwAr6FA+dciaRJ\nkpZIWi9pnnNunZnda2YjvWb/LamFpPlmtsrMFlXRHQAAAACf+TmtSM65xZIWV1p3d8DjoX7uHwAA\nAEDouEMyAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAA\nAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAA\nAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAA\nHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAe\nwgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAnqhwFwCffbFKumdUuKuoHfcUhrsCAACABo0jBwAA\nAAAkEQ4AAAAAeAgHAAAAACQRDgAAAAB4CAcAAAAAJPkcDsws28w2mlm+mU0Nsj3GzF7wtr9vZol+\n1gMAAACgar5dytTMIiU9IelCSQWSVpjZIudcXkCz8ZL2OefOMrMxkh6SdJVfNQEAgO+IS2QDDZqf\n9znoJynfOfepJJnZ85JGSQoMB6Mk3eM9XiDpf8zMnHPOx7oA4Hh84QEAwNdpRQmStgcsF3jrgrZx\nzpVIKpTUxseaAAAAAFTBzyMHFmRd5SMCobSRmU2UNNFbLDKzjd+xtkZjiNRW0p5w11Erfh3s4wLU\nDsYKEBrGSo11qYudALXFz3BQIKlTwHJHSTuqaFNgZlGSWknaW7kj59wsSbN8qrNBM7Nc51xmuOsA\nTnaMFSA0jBWgYfNzWtEKSV3NLMnMoiWNkbSoUptFkn7kPb5c0v9yvgEAAAAQHr4dOXDOlZjZJElL\nJEVKeso5t87M7pWU65xbJOlH+Xt/AAAE+ElEQVRPkp4xs3yVHzEY41c9AAAAAKpn/FDfsJnZRG9a\nFoBqMFaA0DBWgIaNcAAAAABAks93SAYAAABQfxAOAAAAAEgiHDR6ZpZoZmu9x1lm9mq4awLCwcze\nPcH2xWZ2al3VAzQEZpZjZpne4y1m1jbcNQGonp/3OYCPzMxUfs5IWbhrAU42ZhbpnCutyXOccwNO\nsP37360qoOHhbxHQ8HDkoB7xfuVfb2YzJH0o6Voz+5eZfWhm882shdeur5m9a2arzewDM2vpPfct\nr+2HZlbtFyHgZOV9ljeY2V/M7GMzW2Bmsd6vkneb2duSrjCzM83s72a20vvsd/eeH29mL3vjY3XF\nWDCzIu+/p5vZm2a2yszWmtl53vpjv3qa2WRv21ozuy2grvVmNtvM1pnZUjNrFpY3CahG5c+vmT1k\nZjcHbL/HzKZ4j39uZiu8sfZrb13lv0WdzOwPZpbrffZ/HZ5XBqA2EA7qn26S5ki6UNJ4SUOdc30k\n5Uqa7N1w7gVJtzrn0iUNlfSNpF2SLvTaXiVpejiKB2pJN0mznHO9JB2QVPHF5rBzbpBz7nmV31X9\nZ865DEl3SJrhtZku6Z/e+OgjaV2lvq+WtMQ5d7akdEmrAjeaWYak6yX1l3SOpB+bWW9vc1dJTzjn\nekraL2l0bb1goDYE+/xKel7lfxcqXClpvpldpPLPdD9JZ0vKMLPveW26SZrjnOvtnNsq6b+8uyb3\nkjTYzHrVyQsCUOuYVlT/bHXOvWdml0hKkfRO+VFdRUv6l8r/h/2Fc26FJDnnDkiSmTWX9D9mdrak\nUknJ4SgeqCXbnXPveI+flXSL9/gFSfKOog1Q+ReciufEeP89X9IPJcmbelRYqe8Vkp4ysyaSFjrn\nVlXaPkjSy865g96+XpJ0nsrv+P5ZQPuVkhK/w2sE/FDV57edmXWQFCdpn3Num5ndIukiSR95z22h\n8rCwTd7fooB+rzSziSr/XnG6yv8+fVwXLwhA7SIc1D8Hvf+apGXOubGBG71fa4LdvOJ2SV+q/JfQ\nCEmH/SwS8Fnlz3jFcsX4iJC03/v1v2YdO/em9+voxSq/g/t/O+fmBDSxKp4qSUcCHpdKYloRTjZV\nfX4XSLpcUnuVH0moaPugc27mtzowS9T/jzWZWZLKj871dc7tM7OnJTWt1aoB1BmmFdVf70kaaGZn\nSZI35zpZ0gZJHcysr7e+pZlFSWql8iMKZZKulRQZprqB2tDZzM71Ho+V9HbgRu+I2WdmdoVUftKk\nmaV7m9+QdJO3PtLMTgl8rpl1kbTLOTdb0p9UPvUo0JuSLvXGXHNJl0l6q/ZeGuCrqj6/z0sao/KA\nsMBru0TSDQHnsyWYWbsgfZ6i8rBQaGbxkob7/BoA+IhwUE8553ZLuk7Sc2b2scrDQnfn3FGVzx39\nvZmtlrRM5b/gzJD0IzN7T+VTig4G7RioH9ar/PP8saTWkv4QpM01ksZ742CdpFHe+lslDTGzNSqf\n+tOz0vOyJK0ys49Ufs7A44EbnXMfSnpa0geS3pf0pHPuIwH1QFWfX+fcOkktJX3unPvCa7tU0l8l\n/csbLwu8NpX7XK3yqUfrJD0l6Z3KbQDUH+ZcsBkoAHBy8qY0vOqcSw1zKQAANDgcOQAAAAAgiSMH\nAAAAADwcOQAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIEn6P3jEushu+16iAAAAAElF\nTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fd0e3149860>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:10:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>199780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>1780</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0       199780\n",
       "1  plasmid_1         1780"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.036668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.006623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.011220</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.036668\n",
       "1  precision  0.006623\n",
       "2   f1_score  0.011220"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP020836.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.036668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.033296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.069964</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP020836.1\n",
       "0  plasmid_0    0.036668\n",
       "1  plasmid_1    0.033296\n",
       "2      Total    0.069964"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020836.1</td>\n",
       "      <td>0.003499</td>\n",
       "      <td>0.357303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.003499</td>\n",
       "      <td>0.357303</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1\n",
       "0  CP020836.1   0.003499   0.357303\n",
       "1       Total   0.003499   0.357303"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.003499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.357303</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.003499\n",
       "1  plasmid_1    0.357303"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020836.1</td>\n",
       "      <td>0.036668</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP020836.1    0.036668"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No high mutual correspondence found.\n"
     ]
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 10, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>0.000276606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[158, 134, 68, 93, 153, 132, 116, 127, 146, 14...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.274998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.274721</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0\n",
       "GC_sum                                        0.000276606\n",
       "chain   [158, 134, 68, 93, 153, 132, 116, 127, 146, 14...\n",
       "gd_sum                                          -0.274998\n",
       "total                                           -0.274721"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_11</th>\n",
       "      <th>plasmid_12</th>\n",
       "      <th>plasmid_13</th>\n",
       "      <th>plasmid_14</th>\n",
       "      <th>plasmid_15</th>\n",
       "      <th>plasmid_16</th>\n",
       "      <th>plasmid_17</th>\n",
       "      <th>plasmid_18</th>\n",
       "      <th>plasmid_19</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_20</th>\n",
       "      <th>plasmid_21</th>\n",
       "      <th>plasmid_22</th>\n",
       "      <th>plasmid_23</th>\n",
       "      <th>plasmid_24</th>\n",
       "      <th>plasmid_25</th>\n",
       "      <th>plasmid_26</th>\n",
       "      <th>plasmid_27</th>\n",
       "      <th>plasmid_28</th>\n",
       "      <th>plasmid_29</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_30</th>\n",
       "      <th>plasmid_31</th>\n",
       "      <th>plasmid_32</th>\n",
       "      <th>plasmid_33</th>\n",
       "      <th>plasmid_34</th>\n",
       "      <th>plasmid_35</th>\n",
       "      <th>plasmid_36</th>\n",
       "      <th>plasmid_37</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00591572</td>\n",
       "      <td>0.00118194</td>\n",
       "      <td>0.00338981</td>\n",
       "      <td>5.51238e-05</td>\n",
       "      <td>0.000554073</td>\n",
       "      <td>0.000311993</td>\n",
       "      <td>0.00127204</td>\n",
       "      <td>-0.000966151</td>\n",
       "      <td>0.000250108</td>\n",
       "      <td>-0.00196978</td>\n",
       "      <td>-0.00354959</td>\n",
       "      <td>-0.00570872</td>\n",
       "      <td>-0.000221112</td>\n",
       "      <td>0.0020337</td>\n",
       "      <td>0.00125904</td>\n",
       "      <td>-0.00194935</td>\n",
       "      <td>0.00127404</td>\n",
       "      <td>-0.00446464</td>\n",
       "      <td>0.00143472</td>\n",
       "      <td>8.20636e-05</td>\n",
       "      <td>0.000141095</td>\n",
       "      <td>-0.00183224</td>\n",
       "      <td>-0.000265514</td>\n",
       "      <td>-0.00316684</td>\n",
       "      <td>0.00061996</td>\n",
       "      <td>-0.000334183</td>\n",
       "      <td>-0.00666212</td>\n",
       "      <td>0.00068075</td>\n",
       "      <td>-0.00151534</td>\n",
       "      <td>0.00514173</td>\n",
       "      <td>0.000486719</td>\n",
       "      <td>0.00154468</td>\n",
       "      <td>0.00298076</td>\n",
       "      <td>-0.000521492</td>\n",
       "      <td>-0.000761172</td>\n",
       "      <td>-1.20404e-05</td>\n",
       "      <td>-0.000746776</td>\n",
       "      <td>-0.0114199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[46, 120, 187, 73, 191, 187]</td>\n",
       "      <td>[23, 109, 99, 104, 114, 210, 120, 187, 188, 57...</td>\n",
       "      <td>[134, 68, 90, 153, 131, 116, 128, 146, 143, 148]</td>\n",
       "      <td>[106, 194, 212, 208, 210, 120, 187, 188, 15, 1...</td>\n",
       "      <td>[6, 178, 154, 135, 176, 49, 178]</td>\n",
       "      <td>[178, 154, 135, 176, 52]</td>\n",
       "      <td>[178, 154, 135]</td>\n",
       "      <td>[172, 211, 188, 187, 120, 210, 208, 212, 193, ...</td>\n",
       "      <td>[194, 212, 208, 210, 120, 187, 189, 44, 102, 86]</td>\n",
       "      <td>[192, 74, 102, 38, 192]</td>\n",
       "      <td>[195, 115, 102, 64, 192]</td>\n",
       "      <td>[195, 56, 102, 82, 189, 187, 120, 210, 208]</td>\n",
       "      <td>[151, 84, 158, 133, 68, 93, 153, 132, 116, 127...</td>\n",
       "      <td>[212, 208, 210, 120, 186, 190, 69]</td>\n",
       "      <td>[212, 208, 210, 120, 186, 191, 34]</td>\n",
       "      <td>[193, 212, 208, 210, 120, 186, 189, 58, 189, 1...</td>\n",
       "      <td>[212, 208, 210, 120, 186, 191, 42]</td>\n",
       "      <td>[208, 210, 120, 186, 190, 53, 188, 187, 120]</td>\n",
       "      <td>[212, 208, 210, 120, 186, 191, 71]</td>\n",
       "      <td>[120, 187, 188, 213, 192, 212, 208, 210, 120, ...</td>\n",
       "      <td>[192, 212, 208, 210, 120, 186, 190, 2]</td>\n",
       "      <td>[193, 212, 208, 210, 120, 186, 189, 59]</td>\n",
       "      <td>[192, 212, 208, 210, 120, 186, 190, 67]</td>\n",
       "      <td>[151, 126, 169, 111, 97, 130, 161, 85]</td>\n",
       "      <td>[192, 212, 208, 210, 120, 186, 191, 4]</td>\n",
       "      <td>[192, 212, 208, 210, 120, 186, 191, 66]</td>\n",
       "      <td>[208, 210, 120, 186, 190, 81]</td>\n",
       "      <td>[192, 212, 208, 210, 120, 187, 189, 17, 100, 8...</td>\n",
       "      <td>[193, 212, 208, 210, 120, 186, 189, 7]</td>\n",
       "      <td>[210, 120]</td>\n",
       "      <td>[41, 100, 14, 155]</td>\n",
       "      <td>[5, 100, 61, 96]</td>\n",
       "      <td>[12, 210, 120, 187, 189, 45]</td>\n",
       "      <td>[162, 181, 144, 137, 32, 194, 212, 208, 210, 1...</td>\n",
       "      <td>[107, 63, 101, 119, 159, 172, 107, 1, 102, 3, ...</td>\n",
       "      <td>[194, 40, 154, 135, 27, 194, 212, 208, 210, 12...</td>\n",
       "      <td>[101, 24, 161, 129, 97, 110]</td>\n",
       "      <td>[181, 162, 95, 101, 119]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.762375</td>\n",
       "      <td>-0.135748</td>\n",
       "      <td>-0.574526</td>\n",
       "      <td>-0.0435458</td>\n",
       "      <td>-0.0107323</td>\n",
       "      <td>-0.0197268</td>\n",
       "      <td>-0.701684</td>\n",
       "      <td>-0.0293437</td>\n",
       "      <td>-0.0406827</td>\n",
       "      <td>-0.0261927</td>\n",
       "      <td>-0.0784917</td>\n",
       "      <td>-0.0577454</td>\n",
       "      <td>-0.122538</td>\n",
       "      <td>-0.0494498</td>\n",
       "      <td>-0.010606</td>\n",
       "      <td>-0.099877</td>\n",
       "      <td>-0.0122251</td>\n",
       "      <td>-0.103556</td>\n",
       "      <td>-0.0566813</td>\n",
       "      <td>-0.021594</td>\n",
       "      <td>-0.00170143</td>\n",
       "      <td>-0.0220886</td>\n",
       "      <td>-0.03943</td>\n",
       "      <td>-0.167191</td>\n",
       "      <td>-0.00196339</td>\n",
       "      <td>-0.033229</td>\n",
       "      <td>-0.0890366</td>\n",
       "      <td>-0.00642522</td>\n",
       "      <td>-0.00297828</td>\n",
       "      <td>-0.656112</td>\n",
       "      <td>-0.00761477</td>\n",
       "      <td>-0.00547782</td>\n",
       "      <td>-0.120772</td>\n",
       "      <td>-0.154488</td>\n",
       "      <td>-0.00734948</td>\n",
       "      <td>-0.0157684</td>\n",
       "      <td>-0.0464605</td>\n",
       "      <td>-0.322296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.768291</td>\n",
       "      <td>-0.134566</td>\n",
       "      <td>-0.571137</td>\n",
       "      <td>-0.0434907</td>\n",
       "      <td>-0.0101782</td>\n",
       "      <td>-0.0194148</td>\n",
       "      <td>-0.700411</td>\n",
       "      <td>-0.0303099</td>\n",
       "      <td>-0.0404326</td>\n",
       "      <td>-0.0281625</td>\n",
       "      <td>-0.0820412</td>\n",
       "      <td>-0.0634542</td>\n",
       "      <td>-0.122759</td>\n",
       "      <td>-0.0474161</td>\n",
       "      <td>-0.00934698</td>\n",
       "      <td>-0.101826</td>\n",
       "      <td>-0.0109511</td>\n",
       "      <td>-0.108021</td>\n",
       "      <td>-0.0552466</td>\n",
       "      <td>-0.0215119</td>\n",
       "      <td>-0.00156034</td>\n",
       "      <td>-0.0239208</td>\n",
       "      <td>-0.0396956</td>\n",
       "      <td>-0.170358</td>\n",
       "      <td>-0.00134343</td>\n",
       "      <td>-0.0335632</td>\n",
       "      <td>-0.0956988</td>\n",
       "      <td>-0.00574447</td>\n",
       "      <td>-0.00449361</td>\n",
       "      <td>-0.650971</td>\n",
       "      <td>-0.00712805</td>\n",
       "      <td>-0.00393314</td>\n",
       "      <td>-0.117792</td>\n",
       "      <td>-0.155009</td>\n",
       "      <td>-0.00811065</td>\n",
       "      <td>-0.0157805</td>\n",
       "      <td>-0.0472073</td>\n",
       "      <td>-0.333716</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           plasmid_0  \\\n",
       "GC_sum                   -0.00591572   \n",
       "chain   [46, 120, 187, 73, 191, 187]   \n",
       "gd_sum                     -0.762375   \n",
       "total                      -0.768291   \n",
       "\n",
       "                                                plasmid_1  \\\n",
       "GC_sum                                         0.00118194   \n",
       "chain   [23, 109, 99, 104, 114, 210, 120, 187, 188, 57...   \n",
       "gd_sum                                          -0.135748   \n",
       "total                                           -0.134566   \n",
       "\n",
       "                                              plasmid_10  \\\n",
       "GC_sum                                        0.00338981   \n",
       "chain   [134, 68, 90, 153, 131, 116, 128, 146, 143, 148]   \n",
       "gd_sum                                         -0.574526   \n",
       "total                                          -0.571137   \n",
       "\n",
       "                                               plasmid_11  \\\n",
       "GC_sum                                        5.51238e-05   \n",
       "chain   [106, 194, 212, 208, 210, 120, 187, 188, 15, 1...   \n",
       "gd_sum                                         -0.0435458   \n",
       "total                                          -0.0434907   \n",
       "\n",
       "                              plasmid_12                plasmid_13  \\\n",
       "GC_sum                       0.000554073               0.000311993   \n",
       "chain   [6, 178, 154, 135, 176, 49, 178]  [178, 154, 135, 176, 52]   \n",
       "gd_sum                        -0.0107323                -0.0197268   \n",
       "total                         -0.0101782                -0.0194148   \n",
       "\n",
       "             plasmid_14                                         plasmid_15  \\\n",
       "GC_sum       0.00127204                                       -0.000966151   \n",
       "chain   [178, 154, 135]  [172, 211, 188, 187, 120, 210, 208, 212, 193, ...   \n",
       "gd_sum        -0.701684                                         -0.0293437   \n",
       "total         -0.700411                                         -0.0303099   \n",
       "\n",
       "                                              plasmid_16  \\\n",
       "GC_sum                                       0.000250108   \n",
       "chain   [194, 212, 208, 210, 120, 187, 189, 44, 102, 86]   \n",
       "gd_sum                                        -0.0406827   \n",
       "total                                         -0.0404326   \n",
       "\n",
       "                     plasmid_17                plasmid_18  \\\n",
       "GC_sum              -0.00196978               -0.00354959   \n",
       "chain   [192, 74, 102, 38, 192]  [195, 115, 102, 64, 192]   \n",
       "gd_sum               -0.0261927                -0.0784917   \n",
       "total                -0.0281625                -0.0820412   \n",
       "\n",
       "                                         plasmid_19  \\\n",
       "GC_sum                                  -0.00570872   \n",
       "chain   [195, 56, 102, 82, 189, 187, 120, 210, 208]   \n",
       "gd_sum                                   -0.0577454   \n",
       "total                                    -0.0634542   \n",
       "\n",
       "                                                plasmid_2  \\\n",
       "GC_sum                                       -0.000221112   \n",
       "chain   [151, 84, 158, 133, 68, 93, 153, 132, 116, 127...   \n",
       "gd_sum                                          -0.122538   \n",
       "total                                           -0.122759   \n",
       "\n",
       "                                plasmid_20  \\\n",
       "GC_sum                           0.0020337   \n",
       "chain   [212, 208, 210, 120, 186, 190, 69]   \n",
       "gd_sum                          -0.0494498   \n",
       "total                           -0.0474161   \n",
       "\n",
       "                                plasmid_21  \\\n",
       "GC_sum                          0.00125904   \n",
       "chain   [212, 208, 210, 120, 186, 191, 34]   \n",
       "gd_sum                           -0.010606   \n",
       "total                          -0.00934698   \n",
       "\n",
       "                                               plasmid_22  \\\n",
       "GC_sum                                        -0.00194935   \n",
       "chain   [193, 212, 208, 210, 120, 186, 189, 58, 189, 1...   \n",
       "gd_sum                                          -0.099877   \n",
       "total                                           -0.101826   \n",
       "\n",
       "                                plasmid_23  \\\n",
       "GC_sum                          0.00127404   \n",
       "chain   [212, 208, 210, 120, 186, 191, 42]   \n",
       "gd_sum                          -0.0122251   \n",
       "total                           -0.0109511   \n",
       "\n",
       "                                          plasmid_24  \\\n",
       "GC_sum                                   -0.00446464   \n",
       "chain   [208, 210, 120, 186, 190, 53, 188, 187, 120]   \n",
       "gd_sum                                     -0.103556   \n",
       "total                                      -0.108021   \n",
       "\n",
       "                                plasmid_25  \\\n",
       "GC_sum                          0.00143472   \n",
       "chain   [212, 208, 210, 120, 186, 191, 71]   \n",
       "gd_sum                          -0.0566813   \n",
       "total                           -0.0552466   \n",
       "\n",
       "                                               plasmid_26  \\\n",
       "GC_sum                                        8.20636e-05   \n",
       "chain   [120, 187, 188, 213, 192, 212, 208, 210, 120, ...   \n",
       "gd_sum                                          -0.021594   \n",
       "total                                          -0.0215119   \n",
       "\n",
       "                                    plasmid_27  \\\n",
       "GC_sum                             0.000141095   \n",
       "chain   [192, 212, 208, 210, 120, 186, 190, 2]   \n",
       "gd_sum                             -0.00170143   \n",
       "total                              -0.00156034   \n",
       "\n",
       "                                     plasmid_28  \\\n",
       "GC_sum                              -0.00183224   \n",
       "chain   [193, 212, 208, 210, 120, 186, 189, 59]   \n",
       "gd_sum                               -0.0220886   \n",
       "total                                -0.0239208   \n",
       "\n",
       "                                     plasmid_29  \\\n",
       "GC_sum                             -0.000265514   \n",
       "chain   [192, 212, 208, 210, 120, 186, 190, 67]   \n",
       "gd_sum                                 -0.03943   \n",
       "total                                -0.0396956   \n",
       "\n",
       "                                     plasmid_3  \\\n",
       "GC_sum                             -0.00316684   \n",
       "chain   [151, 126, 169, 111, 97, 130, 161, 85]   \n",
       "gd_sum                               -0.167191   \n",
       "total                                -0.170358   \n",
       "\n",
       "                                    plasmid_30  \\\n",
       "GC_sum                              0.00061996   \n",
       "chain   [192, 212, 208, 210, 120, 186, 191, 4]   \n",
       "gd_sum                             -0.00196339   \n",
       "total                              -0.00134343   \n",
       "\n",
       "                                     plasmid_31  \\\n",
       "GC_sum                             -0.000334183   \n",
       "chain   [192, 212, 208, 210, 120, 186, 191, 66]   \n",
       "gd_sum                                -0.033229   \n",
       "total                                -0.0335632   \n",
       "\n",
       "                           plasmid_32  \\\n",
       "GC_sum                    -0.00666212   \n",
       "chain   [208, 210, 120, 186, 190, 81]   \n",
       "gd_sum                     -0.0890366   \n",
       "total                      -0.0956988   \n",
       "\n",
       "                                               plasmid_33  \\\n",
       "GC_sum                                         0.00068075   \n",
       "chain   [192, 212, 208, 210, 120, 187, 189, 17, 100, 8...   \n",
       "gd_sum                                        -0.00642522   \n",
       "total                                         -0.00574447   \n",
       "\n",
       "                                    plasmid_34  plasmid_35  \\\n",
       "GC_sum                             -0.00151534  0.00514173   \n",
       "chain   [193, 212, 208, 210, 120, 186, 189, 7]  [210, 120]   \n",
       "gd_sum                             -0.00297828   -0.656112   \n",
       "total                              -0.00449361   -0.650971   \n",
       "\n",
       "                plasmid_36        plasmid_37                     plasmid_4  \\\n",
       "GC_sum         0.000486719        0.00154468                    0.00298076   \n",
       "chain   [41, 100, 14, 155]  [5, 100, 61, 96]  [12, 210, 120, 187, 189, 45]   \n",
       "gd_sum         -0.00761477       -0.00547782                     -0.120772   \n",
       "total          -0.00712805       -0.00393314                     -0.117792   \n",
       "\n",
       "                                                plasmid_5  \\\n",
       "GC_sum                                       -0.000521492   \n",
       "chain   [162, 181, 144, 137, 32, 194, 212, 208, 210, 1...   \n",
       "gd_sum                                          -0.154488   \n",
       "total                                           -0.155009   \n",
       "\n",
       "                                                plasmid_6  \\\n",
       "GC_sum                                       -0.000761172   \n",
       "chain   [107, 63, 101, 119, 159, 172, 107, 1, 102, 3, ...   \n",
       "gd_sum                                        -0.00734948   \n",
       "total                                         -0.00811065   \n",
       "\n",
       "                                                plasmid_7  \\\n",
       "GC_sum                                       -1.20404e-05   \n",
       "chain   [194, 40, 154, 135, 27, 194, 212, 208, 210, 12...   \n",
       "gd_sum                                         -0.0157684   \n",
       "total                                          -0.0157805   \n",
       "\n",
       "                           plasmid_8                 plasmid_9  \n",
       "GC_sum                  -0.000746776                -0.0114199  \n",
       "chain   [101, 24, 161, 129, 97, 110]  [181, 162, 95, 101, 119]  \n",
       "gd_sum                    -0.0464605                 -0.322296  \n",
       "total                     -0.0472073                 -0.333716  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_35  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
