{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 66 with $\\alpha_1 = 1$, $\\alpha_2 = 5$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 66\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016924.1</td>\n",
       "      <td>111539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016925.1</td>\n",
       "      <td>63291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016927.1</td>\n",
       "      <td>72093</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP016924.1       111539\n",
       "1  CP016925.1        63291\n",
       "2  CP016927.1        72093"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt0VeWd//HPNwkEAhgFQrhKUiWB\nXIiQCBpQgqLAVMEuqpLWdlQoHS2jFlplZlq11tbqQJ3Bn1Sw41IKRcALpUoF2iHFayUoyDUQkRgE\nBAQC4ZrL8/sjO8zxkIQTzc4hyfu1Fsuz937Os7/nrPN48jn72Xubc04AAAAAEBHuAgAAAACcHwgH\nAAAAACQRDgAAAAB4CAcAAAAAJBEOAAAAAHgIBwAAAAAkEQ4AAAAAeAgHAAAAACQRDgAAAAB4osJd\nQH117tzZJSQkhLuMJuPYsWNq165duMsAznuMFSA0jJX6Wbt27QHnXFy46wBC1eTCQUJCgvLz88Nd\nRpORl5ennJyccJcBnPcYK0BoGCv1Y2ZF4a4BqA+mFQEAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADg\nIRwAAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4PEtHJjZc2a2z8w21rLdzGymmRWa2UdmNtCvWgAA\nAACcm59HDp6XNKqO7aMl9fH+TZL0Ox9rAQAAAHAOvoUD59xqSQfraDJW0lxX5T1JF5pZN7/qAQAA\nAFC3qDDuu4ek4oDlXd66PcENzWySqo4uKD4+Xnl5eb4WtuGzEl/7b0yJsZG+v19ouRgrQGgYKwCa\ninCGA6thnaupoXNujqQ5kpSVleVycnJ8LEu6fdrrvvbfmJ4f1U5+v19ouRgrQGgYKwCainBerWiX\npF4Byz0l7Q5TLQAAAECLF85wsFTS972rFl0hqcQ5d9aUIgAAAACNw7dpRWa2QFKOpM5mtkvSQ5Ja\nSZJz7hlJyyT9k6RCSccl3eFXLQAAAADOzbdw4JzLPcd2J+lHfu0fAAAAQP1wh2QAAAAAkggHAAAA\nADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAA\nPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8\nhAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyE\nAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQD\nAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8voYD\nMxtlZgVmVmhm02rYfrGZrTKzD83sIzP7Jz/rAQAAAFA738KBmUVKelrSaEkpknLNLCWo2c8kLXLO\nDZA0XtIsv+oBAAAAUDc/jxwMklTonNvhnDst6UVJY4PaOEkXeI9jJe32sR4AAAAAdYjyse8ekooD\nlndJGhzU5mFJK8zsXyW1kzTCx3oAAAAA1MHPcGA1rHNBy7mSnnfOzTCzKyX9wczSnHOVX+rIbJKk\nSZIUHx+vvLw8P+o9Y2p6ua/9N6bS0lLf3y+0XIwVIDSMFQBNhZ/hYJekXgHLPXX2tKEJkkZJknPu\nXTNrI6mzpH2BjZxzcyTNkaSsrCyXk5PjU8lVbp/2uq/9N6bnR7WT3+8XWi7GChAaxgqApsLPcLBG\nUh8zS5T0mapOOP5OUJtPJV0r6Xkz6yepjaT9PtYEADXbs056OPi0qCbq4ZJwVwAAaKJ8OyHZOVcu\nabKk5ZK2qOqqRJvM7BEzG+M1myrpB2a2XtICSbc754KnHgEAAABoBH4eOZBzbpmkZUHrHgx4vFnS\nED9rAAAAABAa7pAMAAAAQBLhAAAAAICHcAAAAABAEuEAAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAA\nQBLhAAAAAICHcAAAAABAEuEAAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAAQBLhAAAAAICHcAAAAABA\nEuEAAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAAQBLhAAAAAICHcAAAAABAEuEAAAAAgIdwAAAAAEAS\n4QAAAACAh3AAAAAAQBLhAAAAAICHcAAAAABAEuEAAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAAQBLh\nAAAAAICHcAAAAABAEuEAAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAAQBLhAAAAAICHcAAAAABAEuEA\nAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAAQJLP4cDMRplZgZkVmtm0WtrcYmabzWyTmf3Rz3oAAAAA\n1C7Kr47NLFLS05Kuk7RL0hozW+qc2xzQpo+kf5M0xDl3yMy6+FUPAAAAgLr5eeRgkKRC59wO59xp\nSS9KGhvU5geSnnbOHZIk59w+H+sBAAAAUAffjhxI6iGpOGB5l6TBQW2SJMnM3pYUKelh59wbwR2Z\n2SRJkyQpPj5eeXl5ftR7xtT0cl/7b0ylpaW+v19ouZrVWInurrzkX4S7jIbBmD/vNKuxwvcK0Kz5\nGQ6shnWuhv33kZQjqaekN80szTl3+EtPcm6OpDmSlJWV5XJychq82EC3T3vd1/4b0/Oj2snv9wst\nV7MaKxmfKKfgoXCX0TByS8JdAYI0q7HC9wrQrPk5rWiXpF4Byz0l7a6hzZ+cc2XOuU8kFagqLAAA\nAABoZH6GgzWS+phZopm1ljRe0tKgNkskDZckM+usqmlGO3ysCQAAAEAtfAsHzrlySZMlLZe0RdIi\n59wmM3vEzMZ4zZZL+sLMNktaJemnzrkv/KoJAAAAQO38POdAzrllkpYFrXsw4LGTNMX7BwAAACCM\nuEMyAAAAAEmEAwAAAAAewgEAAAAASfUIB2Y21Mzu8B7HmVmif2UBAAAAaGwhhQMze0jSA5L+zVvV\nStI8v4oCAAAA0PhCPXLwLUljJB2TJOfcbkkd/CoKAAAAQOMLNRyc9i476iTJzNr5VxIAAACAcAg1\nHCwys9mSLjSzH0j6q6Rn/SsLAAAAQGML6SZozrnpZnadpCOSkiU96Jxb6WtlAAAAABrVOcOBmUVK\nWu6cGyGJQAAAAAA0U+ecVuScq5B03MxiG6EeAAAAAGES0rQiSSclbTCzlfKuWCRJzrl7fKkKAAAA\nQKMLNRy87v0DAAAA0EyFekLyC2bWWlKSt6rAOVfmX1kAAAAAGltI4cDMciS9IGmnJJPUy8z+2Tm3\n2r/SAAAAADSmUKcVzZB0vXOuQJLMLEnSAkmZfhUGAAAAoHGFehO0VtXBQJKcc9sktfKnJAAAAADh\nEOqRg3wz+x9Jf/CWvytprT8lAQAAAAiHUMPBXZJ+JOkeVZ1zsFrSLL+KAgAAAND4Qg0HUZL+2zn3\nW+nMXZOjfasKAACgmVq7dm2XqKio30tKU+hTvIGGUClpY3l5+cTMzMx9NTUINRz8TdIISaXecltJ\nKyRlf+0SAQAAWpCoqKjfd+3atV9cXNyhiIgIF+560HJUVlba/v37U/bu3ft7SWNqahNqWm3jnKsO\nBvIexzRAjQAAAC1NWlxc3BGCARpbRESEi4uLK1HVUaua24TY1zEzG1i9YGZZkk58zfoAAABaogiC\nAcLF++zVmgFCnVZ0r6TFZrZbkpPUXdKtX788AAAAAOeLUMNBoqQBki6W9C1JV6gqJAAAAOBrSJj2\neoPeVHbnb7553lxuftCgQcnTp08vvvrqq4+H0r6yslLZ2dlJy5YtK+zYsWOl3/V9Xa+99lqHGTNm\nxK9atarQj/7nz58fu2nTpra//vWv9wZvi4mJGXD8+PEPa3tudnZ20p/+9KeP4+LiKuqzz1CnFf3c\nOXdE0oWSrpM0R9Lv6rMjAAAANB1lZWWNvs9FixbFpqamnmgKwaAxfPe73y2pKRiEIjc394vp06fH\n1fd5oYaD6sTxTUnPOOf+JKl1fXcGAACA8PvpT3/aLTExMTU7O7vPjTfemPjggw/GS1W/9E+ePLnH\n5Zdfnvzoo4/G7969O2rkyJGXpKWl9UtLS+u3YsWKdpJ05MiRiJtvvjkhLS2tX79+/VLmzZt3oSSV\nlpbaDTfc8I2kpKSUb37zm984efKkSdKTTz7ZecKECb2q9z9jxozOEydO7Blc1/z58zt+61vfOixJ\nBQUFrRMTE1NvvfXW3n369EkdM2ZM4pIlSzoMHDiwb+/evdNWrVoVU1ctBQUFrTMzM5NTUlL6paSk\n9Fu5cmU7qerX/kGDBiWPGjXqG4mJialjxoxJrKysO4uMGzcu4Tvf+c7FmZmZyQkJCWkLFiyIDW6z\natWqmAEDBvTt169fyoABA/quX78+WpLy8/PbpKen9+vbt29KUlJSyoYNG6JDfW0zZ87s9P3vf/9i\nSdq6dWvryy67rG9aWlq/e++9t3v1fouKilplZWUl9+3bN6VPnz6pb7zxRntJGj9+/OFXXnmlU4gf\niTNCnVb0mZnNVtXlTB83s2hxXV4AAIAmZ/Xq1TF//vOfL9qwYcPmsrIyu+yyy1IGDBhwZtrP4cOH\nI9esWVMgSTfeeGPilClTPh85cmTp9u3bW48cObLPjh07Nv37v/97t+HDhx9ZvHjxzgMHDkRmZWX1\nGzNmzJHf/va3cW3btq3ctm3b5n/84x9thwwZkiJJEyZMOJiamppy6tSpXdHR0W7evHmdZ8+eXRRc\n29q1a9sPGTLkzPri4uI2Cxcu3JGZmVnUv3//fvPnz++Un5+/9Y9//OOFv/rVr7oNHz7849pq6d69\ne/mbb765LSYmxm3YsCE6Nzf3Gxs3btwiSVu2bGm7bt26HQkJCWWZmZl9V65c2X7kyJGlwfUEKi4u\njn7//fcLNm/eHD1ixIjksWPHbgjcnpGRcfL999/f2qpVKy1ZsqTD/fff33P58uUfP/XUU3F33333\n53fdddfBkydPWnl5uT777LNWoby2wP7vvvvuiydOnLh/8uTJXzz22GNnjgg899xzHa+99tqSxx9/\nfG95ebmOHj0aIUlxcXEVp0+ftr1790Z27do15KlFoYaDWySNkjTdOXfYzLpJ+mmoOwEAAMD5IS8v\nr/3o0aMPt2/f3kly11133eHA7bm5uQerH7/99tsXbN++vW31cmlpaeShQ4ci8vLyLli+fPmFM2fO\n7CpJp06dssLCwtZvvfVW+3vuuWefJA0ePPhEUlLScUm64IILKocMGXJ04cKFsenp6SfLysps0KBB\nZ135sqSkJOqiiy468zN+jx49TlW3S0pKOnHNNdcciYiI0MCBA48/+uij3b3XU2MtvXv3LpswYULv\nzZs3t42IiFBRUdGZG/imp6cfu+SSS8okKTU19fjHH398zhkx48aNOxgZGan09PRTvXr1OrVu3bo2\ngdsPHjwYeeuttybu3LmzjZm5srIyk6Qrr7zy2PTp07vt2rWr9fjx4w+lp6efCvW1Bfrggw/a/+Uv\nf/lYkn74wx9+8ctf/rKnJF1xxRXHfvjDHyaUlZVFfPvb3z6UnZ195n3t1KlT+aefftq6a9euIV9l\nNKRw4Jw7LumVgOU9kvaEuhMAAACcH5yr+5oyHTp0OPPHuXNO+fn5W7wg8aU+XnrppcKMjIxTwc83\nsxr7nTRp0oFf/epXXZOSkk7edtttB2pqExkZ6SoqKhQZGSlJat269Zn9RkREqE2bNs5rp4qKCqur\nlilTpnTv0qVL2csvv/xJZWWl2rZte+bE7+jo6DP9RkZGqry8vOai63hdwcsPPPBAj2HDhh1duXLl\nxwUFBa2vueaaZEn6l3/5l4NXXXXVsVdffTV29OjRSbNmzdqZnJx8KpTXFqymS+COHj26dPXq1QUv\nv/xy7O233554zz33fD558uQvpKqgFBMTU6/zN5gaBAAA0ILk5OSULl++PPb48eNWUlIS8de//vXC\n2toOHTr0yOOPP96levmdd95pK0nDhw8/MmPGjPjqufpvv/12W6996bx58zpK0po1a9ps27btzE1z\nr7nmmmN79uxp/eqrr3aaMGHCQdUgMTHx5JYtW6Jr2lab2mopKSmJ7NatW1lkZKRmzZrVqaLi3DNr\nfvSjH/WYO3duje/HK6+8clFFRYU2bdoUXVxcHJ2RkXEycPuRI0cie/bseVqSZs+e3bl6/ebNm1v3\n69fv1M9+9rN9119//eF169a1De47FAMHDix99tlnO0rSs88+e+Zcgm3btrXu0aNH2dSpUw/cdttt\nBz744IMYqerKT/v372+VnJx8VoCrS6jTigAAAOCDxr706LBhw46PGjWqJCUlJbVHjx6n+vfvfyw2\nNrbGv5znzJlTPHHixIuTkpJSKioqbPDgwUezs7M//c1vfrN70qRJF/ft2zfFOWc9e/Y8tWrVqsKf\n/OQn+8aPH5+YlJSUkpqaejw9Pf1YYH833XTToY8++iimtstrXn/99SUrVqzokJaWFvIftLXVct99\n9+0bN27cJUuWLLlo6NChR9u2bXvOX9A3b97ctvqE6GCXXnrpqUGDBiV/8cUXrf7rv/6rKCYm5ku/\n4j/wwAN7J06cmDhz5syuV1111ZHq9X/4wx86Ll68uFNUVJSLi4sre+yxx3YfPnw4MtTXV23WrFmf\njh8//huzZs2KHzNmzKHq9cuXL+8wc+bMrlFRUS4mJqZi/vz5n0jSW2+9FTNgwIBjrVq1qtd+7FyH\nls43WVlZLj8/39d9JEx73df+G9Pzo9opJycn3GWgmWpWYyVju3IKHgp3GQ3j4ZJwV4AgzWqs8L1S\nL2a21jmXFbhu/fr1OzMyMmqcVtNYSkpKImJjYyuPHj0aceWVVyY/88wzRUOHDg3pXgRfx/Dhwy+9\n7777Ph87duzRmrYXFRW1ys3NTXjnnXe2+11LTYYOHdrnrbfeOmvf48aNS7jhhhtK7rjjjkM1Pe98\ndMcdd/S66aabDtf0Xq9fv75zRkZGQk3PY1oRAABAC3Pbbbf17tu3b0r//v373XjjjYf8DgYHDhyI\nTEhISGvTpk1lbcFAknr37l125513Hjh48GBY/katKRg0VWlpaSfqeq9rw7QiAACAFubPf/7zJ425\nv86dO1fs3LlzYyhtJ06ceN79Ov/yyy/vDHcN9TV16tSvdHSKIwcAAAAAJBEOAAAAAHgIBwAAAAAk\nEQ4AAAAAeDghGQAAIJwejs08d6P69FdyzvsmmFnm2LFjDy5ZsuQTSSorK1OXLl0yLrvssmOrVq0q\nnDlzZqf8/Px2c+fO/XTKlCnd27dvX/HII498HthHZGRkZp8+fU5UVFTYpZdeemLRokU7A++uLFXd\niCs7Oztp2bJlhR07dqy8+eabE/72t7/FdurUqXz79u2baqrttdde65Cbm3tJjx49TkvSDTfccGj6\n9Ol7gtv9+te/jnvmmWfii4uLo3fv3r2+W7du5ZK0YMGC2Pfff7/dk08+uTv0Nw3VOHIAAADQwrRt\n27ayoKCgbWlpqUnSq6++ekF8fHxZffqIjo6u3Lp16+bt27dvatWqlZsxY0ZccJtFixbFpqamnujY\nsWOlJN15550Hli5des7LhWZlZZVu3bp189atWzfXFAwkadiwYaUrV67c1r1799OB62+99daSN954\n48KjR4/yd+5XwJsGAADQAl177bUlixcvvlCSFixY0HHcuHEHv2pfQ4cOLS0sLIwOXj9//vyOgXcc\nHj16dGlcXFz5V91PoCFDhpxITk4+Hbw+IiJC2dnZRxcuXBjbEPtpaQgHAAAALdD3vve9gwsXLrzo\n+PHjtmXLlpgrr7zy2Ffpp6ysTMuXL78gPT39RPC2tWvXth8yZMg5+33iiSfinnjiiTNHHj788MP2\nycnJKVdffXWf/Pz8NtXrhw0bdunOnTtbnau/rKysY2+++Wb7+rwOVPH1nAMzGyXpvyVFSvq9c+43\ntbT7tqTFki53zuX7WRMAAACkwYMHn9i1a1f0s88+23HEiBEl9X3+qVOnIvr27Zvi9XX03nvvPeum\nWyUlJVEXXXRR5dnP/rL7779/f/Xj7OzsY0VFRR/FxsZWLly4MHbcuHGXFhUVbZSkv//974Wh1Na1\na9fyvXv3tg791aCab+HAzCIlPS3pOkm7JK0xs6XOuc1B7TpIukfSP/yqBQAAAGcbNWrU4YceeqjX\nihUrCvbt21evvwurzzmoq01kZKSrqKhQZGRkyP1Wn58gVZ0/MGXKlIv37NkTVX3CcShOnDhhbdq0\nOWcowdn8nFY0SFKhc26Hc+60pBclja2h3S8lPSHppI+1AAAAIMhdd911YOrUqbsHDRp01pSghpCY\nmHhyy5YtZ52LUJdPP/00qrKy6u/6VatWxVRWVio+Pr5e5ykUFBS0SU1N9eU1NXd+TivqIak4YHmX\npMGBDcxsgKRezrnXzOwntXVkZpMkTZKk+Ph45eXlNXy1AaamN8h5MueF0tJS398vtFzNaqxEd1de\n8i/CXUbDYMyfd5rVWOF7peGFcOlRv1xyySVlP//5z/edq92TTz7Zbfbs2fHVy59//vlHofR//fXX\nl6xYsaJDWlraKUm68cYbE997770Ohw4dioqPj+8/bdq03T/+8Y8PVJ9vcP/99++fN2/eRc8991yX\nyMhI16ZNm8q5c+fuiIio+j172LBhl77wwgtFCQkJZY8++miXp556qusXX3zRKiMjI2X48OElCxcu\nLJKk1atXd3j88cc/+wpvSYtnzjl/Oja7WdJI59xEb/l7kgY55/7VW46Q9L+SbnfO7TSzPEk/Odc5\nB1lZWS4/39/TEhKmve5r/43p+VHtlJOTE+4y0Ew1q7GSsV05BQ+Fu4yG8XC9pw7DZ81qrPC9Ui9m\nttY5lxW4bv369TszMjLOmp/fHBUVFbXKzc1NeOedd855+dKGUlxcHHXLLbd84913393WWPtsatav\nX985IyMjoaZtfk4r2iWpV8ByT0mBN6PoIClNUp6Z7ZR0haSlZvalAQQAAICmqXfv3mV33nnngYMH\nDzbaFTJ37NjResaMGcXnboma+DmtaI2kPmaWKOkzSeMlfad6o3OuRFLn6uVQjxwAAACg6Zg4ceKh\nxtzfsGHDjjfm/pob31Kcc65c0mRJyyVtkbTIObfJzB4xszF+7RcAAOA8V1lZWWnhLgItk/fZq/VK\nTr7e58A5t0zSsqB1D9bSNsfPWgAAAM4TG/fv358SFxdXEhER4c/Jn0ANKisrbf/+/bGSNtbWxtdw\nAAAAgC8rLy+fuHfv3t/v3bs3Tf6e/wkEq5S0sby8fGJtDQgHAAAAjSgzM3OfJKZY47xEWgUAAAAg\niXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJ\ncAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlw\nAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAA\nAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAA\nAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIMnncGBm\no8yswMwKzWxaDdunmNlmM/vIzP5mZr39rAcAAABA7XwLB2YWKelpSaMlpUjKNbOUoGYfSspyzvWX\n9JKkJ/yqBwAAAEDd/DxyMEhSoXNuh3PutKQXJY0NbOCcW+WcO+4tviepp4/1AAAAAKhDlI9995BU\nHLC8S9LgOtpPkPSXmjaY2SRJkyQpPj5eeXl5DVRizaaml/vaf2MqLS31/f1Cy9Wsxkp0d+Ul/yLc\nZTQMxvx5p1mNFb5XgGbNz3BgNaxzNTY0u01SlqRhNW13zs2RNEeSsrKyXE5OTgOVWLPbp73ua/+N\n6flR7eT3+4WWq1mNlYxPlFPwULjLaBi5JeGuAEGa1VjhewVo1vwMB7sk9QpY7ilpd3AjMxsh6T8k\nDXPOnfKxHgAA8HXtWSc9PPbc7ZqChwnSQDA/zzlYI6mPmSWaWWtJ4yUtDWxgZgMkzZY0xjm3z8da\nAAAAAJyDb+HAOVcuabKk5ZK2SFrknNtkZo+Y2Riv2X9Kai9psZmtM7OltXQHAAAAwGd+TiuSc26Z\npGVB6x4MeDzCz/0DAAAACB13SAYAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAA\nAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAA\nwEM4AAAAACCJcAAAAADAQzgAAAAAIEmKCncB8NmeddLDY8NdRcN4uCTcFQAAADRrHDkAAAAAIIlw\nAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAA\nAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAA\nAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADA42s4MLNRZlZgZoVm\nNq2G7dFmttDb/g8zS/CzHgAAAAC18y0cmFmkpKcljZaUIinXzFKCmk2QdMg5d6mkJyU97lc9AAAA\nAOrm55GDQZIKnXM7nHOnJb0oaWxQm7GSXvAevyTpWjMzH2sCAAAAUAs/w0EPScUBy7u8dTW2cc6V\nSyqR1MnHmgAAAADUIsrHvms6AuC+QhuZ2SRJk7zFUjMr+Jq1tRjDpc6SDoS7jgbxCw4qwT+MFSA0\njJV6690YOwEaip/hYJekXgHLPSXtrqXNLjOLkhQr6WBwR865OZLm+FRns2Zm+c65rHDXAZzvGCtA\naBgrQPPm57SiNZL6mFmimbWWNF7S0qA2SyX9s/f425L+1zl31pEDAAAAAP7z7ciBc67czCZLWi4p\nUtJzzrlNZvaIpHzn3FJJ/yPCOS0mAAAE90lEQVTpD2ZWqKojBuP9qgcAAABA3Ywf6ps3M5vkTcsC\nUAfGChAaxgrQvBEOAAAAAEjy+Q7JAAAAAJoOwgEAAAAASYSDFs/MEsxso/c4x8xeC3dNQDiY2Tvn\n2L7MzC5srHqA5sDM8swsy3u808w6h7smAHXz8z4H8JGZmarOGakMdy3A+cbMIp1zFfV5jnMu+xzb\n/+nrVQU0P3wXAc0PRw6aEO9X/i1mNkvSB5K+Z2bvmtkHZrbYzNp77S43s3fMbL2ZvW9mHbznvum1\n/cDM6vxDCDhfeZ/lrWb2gpl9ZGYvmVmM96vkg2b2lqSbzewSM3vDzNZ6n/2+3vPjzexVb3ysrx4L\nZlbq/bebma02s3VmttHMrvLWn/nV08ymeNs2mtl9AXVtMbNnzWyTma0ws7ZheZOAOgR/fs3scTO7\nO2D7w2Y21Xv8UzNb4421X3jrgr+LepnZ78ws3/vs/yI8rwxAQyAcND3JkuZKuk7SBEkjnHMDJeVL\nmuLdcG6hpHudcxmSRkg6IWmfpOu8trdKmhmO4oEGkixpjnOuv6Qjkqr/sDnpnBvqnHtRVXdV/1fn\nXKakn0ia5bWZKenv3vgYKGlTUN/fkbTcOXeZpAxJ6wI3mlmmpDskDZZ0haQfmNkAb3MfSU8751Il\nHZY0rqFeMNAQavr8SnpRVd8L1W6RtNjMrlfVZ3qQpMskZZrZ1V6bZElznXMDnHNFkv7Du2tyf0nD\nzKx/o7wgAA2OaUVNT5Fz7j0zu0FSiqS3q47qqrWkd1X1P+w9zrk1kuScOyJJZtZO0v8zs8skVUhK\nCkfxQAMpds697T2eJ+ke7/FCSfKOomWr6g+c6udEe/+9RtL3JcmbelQS1PcaSc+ZWStJS5xz64K2\nD5X0qnPumLevVyRdpao7vn8S0H6tpISv8RoBP9T2+e1iZt0lxUk65Jz71MzukXS9pA+957ZXVVj4\nVN53UUC/t5jZJFX9XdFNVd9PHzXGCwLQsAgHTc8x778maaVzLjdwo/drTU03r/ixpM9V9UtohKST\nfhYJ+Cz4M169XD0+IiQd9n79r1/Hzq32fh39pqru4P6fzrm5AU2slqdK0qmAxxWSmFaE801tn9+X\nJH1bUldVHUmobvuYc272lzowS9D/jTWZWaKqjs5d7pw7ZGbPS2rToFUDaDRMK2q63pM0xMwulSRv\nznWSpK2SupvZ5d76DmYWJSlWVUcUKiV9T1JkmOoGGsLFZnal9zhX0luBG70jZp+Y2c1S1UmTZpbh\nbf6bpLu89ZFmdkHgc82st6R9zrlnJf2PqqYeBVot6SZvzLWT9C1JbzbcSwN8Vdvn90VJ41UVEF7y\n2i6XdGfA+Ww9zKxLDX1eoKqwUGJm8ZJG+/waAPiIcNBEOef2S7pd0gIz+0hVYaGvc+60quaOPmVm\n6yWtVNUvOLMk/bOZvaeqKUXHauwYaBq2qOrz/JGkjpJ+V0Ob70qa4I2DTZLGeuvvlTTczDaoaupP\natDzciStM7MPVXXOwH8HbnTOfSDpeUnvS/qHpN875z4U0ATU9vl1zm2S1EHSZ865PV7bFZL+KOld\nb7y85LUJ7nO9qqYebZL0nKS3g9sAaDrMuZpmoADA+cmb0vCacy4tzKUAANDscOQAAAAAgCSOHAAA\nAADwcOQAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAk6f8D0FeN4f3EzhIAAAAASUVO\nRK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7ff71e0a4908>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:5:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>11724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>1978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>9768</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0        11724\n",
       "1  plasmid_1         1978\n",
       "2  plasmid_2         9768"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.109581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.790371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.192476</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.109581\n",
       "1  precision  0.790371\n",
       "2   f1_score  0.192476"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP016924.1</th>\n",
       "      <th>CP016925.1</th>\n",
       "      <th>CP016927.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.170751</td>\n",
       "      <td>0.015314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.234520</td>\n",
       "      <td>0.040267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.405271</td>\n",
       "      <td>0.088899</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP016924.1  CP016925.1  CP016927.1\n",
       "0  plasmid_0         0.0    0.170751    0.015314\n",
       "1  plasmid_1         0.0    0.000000    0.033318\n",
       "2  plasmid_2         0.0    0.234520    0.040267\n",
       "3      Total         0.0    0.405271    0.088899"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016924.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016925.1</td>\n",
       "      <td>0.646622</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016927.1</td>\n",
       "      <td>0.087769</td>\n",
       "      <td>0.607179</td>\n",
       "      <td>0.185401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.734391</td>\n",
       "      <td>0.607179</td>\n",
       "      <td>1.185401</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2\n",
       "0  CP016924.1   0.000000   0.000000   0.000000\n",
       "1  CP016925.1   0.646622   0.000000   1.000000\n",
       "2  CP016927.1   0.087769   0.607179   0.185401\n",
       "3       Total   0.734391   0.607179   1.185401"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.646622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.607179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.646622\n",
       "1  plasmid_1    0.607179\n",
       "2  plasmid_2    1.000000"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016924.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016925.1</td>\n",
       "      <td>0.340917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016927.1</td>\n",
       "      <td>0.076027</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP016924.1    0.000000\n",
       "1  CP016925.1    0.340917\n",
       "2  CP016927.1    0.076027"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No high mutual correspondence found.\n"
     ]
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 5, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00821679</td>\n",
       "      <td>-0.0112825</td>\n",
       "      <td>-0.00576185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[90, 29, 49, 96, 31]</td>\n",
       "      <td>[48, 42]</td>\n",
       "      <td>[44, 32, 49, 90, 33]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.77057</td>\n",
       "      <td>-0.82912</td>\n",
       "      <td>-0.680027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.778787</td>\n",
       "      <td>-0.840403</td>\n",
       "      <td>-0.685788</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   plasmid_0  plasmid_1             plasmid_2\n",
       "GC_sum           -0.00821679 -0.0112825           -0.00576185\n",
       "chain   [90, 29, 49, 96, 31]   [48, 42]  [44, 32, 49, 90, 33]\n",
       "gd_sum              -0.77057   -0.82912             -0.680027\n",
       "total              -0.778787  -0.840403             -0.685788"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00898258</td>\n",
       "      <td>-0.00210305</td>\n",
       "      <td>-0.0138317</td>\n",
       "      <td>-0.0425805</td>\n",
       "      <td>-0.00875212</td>\n",
       "      <td>-0.00929555</td>\n",
       "      <td>-0.0400867</td>\n",
       "      <td>-0.00402962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[31, 96, 49, 32, 44, 50, 22, 42, 48, 81, 36, 8...</td>\n",
       "      <td>[90, 28, 49, 90, 53, 103, 44, 50, 35]</td>\n",
       "      <td>[16]</td>\n",
       "      <td>[40, 47]</td>\n",
       "      <td>[90, 29, 49, 90, 33, 49, 96, 26, 49]</td>\n",
       "      <td>[43, 49]</td>\n",
       "      <td>[39]</td>\n",
       "      <td>[44, 103]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.719564</td>\n",
       "      <td>-0.777015</td>\n",
       "      <td>-0.722274</td>\n",
       "      <td>-0.285966</td>\n",
       "      <td>-0.694023</td>\n",
       "      <td>-0.58847</td>\n",
       "      <td>-0.68657</td>\n",
       "      <td>-0.787083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.728547</td>\n",
       "      <td>-0.779118</td>\n",
       "      <td>-0.736105</td>\n",
       "      <td>-0.328546</td>\n",
       "      <td>-0.702775</td>\n",
       "      <td>-0.597765</td>\n",
       "      <td>-0.726657</td>\n",
       "      <td>-0.791113</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0  \\\n",
       "GC_sum                                        -0.00898258   \n",
       "chain   [31, 96, 49, 32, 44, 50, 22, 42, 48, 81, 36, 8...   \n",
       "gd_sum                                          -0.719564   \n",
       "total                                           -0.728547   \n",
       "\n",
       "                                    plasmid_1  plasmid_2  plasmid_3  \\\n",
       "GC_sum                            -0.00210305 -0.0138317 -0.0425805   \n",
       "chain   [90, 28, 49, 90, 53, 103, 44, 50, 35]       [16]   [40, 47]   \n",
       "gd_sum                              -0.777015  -0.722274  -0.285966   \n",
       "total                               -0.779118  -0.736105  -0.328546   \n",
       "\n",
       "                                   plasmid_4   plasmid_5  plasmid_6  \\\n",
       "GC_sum                           -0.00875212 -0.00929555 -0.0400867   \n",
       "chain   [90, 29, 49, 90, 33, 49, 96, 26, 49]    [43, 49]       [39]   \n",
       "gd_sum                             -0.694023    -0.58847   -0.68657   \n",
       "total                              -0.702775   -0.597765  -0.726657   \n",
       "\n",
       "         plasmid_7  \n",
       "GC_sum -0.00402962  \n",
       "chain    [44, 103]  \n",
       "gd_sum   -0.787083  \n",
       "total    -0.791113  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_1  from MILP completely contained in  plasmid_0  from greedy.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
