{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 11 with $\\alpha_1 = 1$, $\\alpha_2 = 5$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 11\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP011580.1</td>\n",
       "      <td>90452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP011579.1</td>\n",
       "      <td>33610</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP011580.1        90452\n",
       "1  CP011579.1        33610"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt0VeWd//HPNwl3MAqEcJWkSgIh\nEDARakAJigpTBbvwQlrtqFA6WkYttsrMtGqtrdWB2sEpFey41EK5WaWMMgLTHxSvlaAgcgkgEkFA\nQCAQrrl8f39kh4khgYNm54Twfq3l8uy9n/Ps78k6D+d8zn723ubuAgAAAICYaBcAAAAAoH4gHAAA\nAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAEIiLdgFn\nqm3btp6UlBTtMs4ahw4dUosWLaJdBlDvMVaAyDBWzsyKFSv2uHtCtOsAInXWhYOkpCTl5eVFu4yz\nxtKlS5WTkxPtMoB6j7ECRIaxcmbMrCDaNQBngmlFAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQI\nBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkSXHRLqA+Sprw\nWrRLqDXPD20R7RIAAABwluDIAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAA\nAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAA\nAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAAB\nwgEAAAAASSGHAzMbamb5ZrbJzCZUs/1CM1tiZh+Y2Ydm9g9h1gMAAACgZnFhdWxmsZJ+J+lqSdsk\nLTez+e6+tlKzn0qa4+6/N7M0SQskJYVVEwAA0ZA04bVol1Brnh/aItolAAhRmEcO+kna5O6b3f24\npFmSRlRp45LOCx7HS9oeYj0AAAAATiG0IweSOknaWml5m6T+Vdo8ImmRmf2zpBaShlTXkZmNlTRW\nkhITE7V06dLarvVL7u9VEmr/damoqCj0vxfQEDBWECY+VwCcLcIMB1bNOq+ynCvpeXefZGaXSfqj\nmaW7e9mXnuQ+TdI0ScrKyvKcnJww6j3h9gZ2+DfsvxfQECxdupSxgtDwuQLgbBHmtKJtkrpUWu6s\nk6cNjZY0R5Lc/R1JTSW1DbEmAAAAADUIMxwsl9TNzJLNrLGkUZLmV2nzqaSrJMnMeqg8HOwOsSYA\nAAAANQgtHLh7iaRxkhZKWqfyqxKtMbNHzWx40Ox+Sd83s1WSZkq63d2rTj0CAAAAUAfCPOdA7r5A\n5ZcnrbzuoUqP10oaEGYNAAAAACLDHZIBAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgi\nHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIc\nAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEiS4qJdAICzV9KE16JdQq15fmiLaJcAAEDUceQAAAAA\ngCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACA\nJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAk\nwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTC\nAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgKSQw4GZDTWzfDPbZGYTamhz\ns5mtNbM1ZvanMOsBAAAAULO4sDo2s1hJv5N0taRtkpab2Xx3X1upTTdJ/yJpgLvvM7N2YdUDAAAA\n4NTCPHLQT9Imd9/s7sclzZI0okqb70v6nbvvkyR33xViPQAAAABOIbQjB5I6SdpaaXmbpP5V2qRI\nkpm9JSlW0iPu/nrVjsxsrKSxkpSYmKilS5eGUe8J9/cqCbX/ulS0d5eWzvxttMuoHR36RLsCVNGg\nxkpRUej/tuDcxVgBcLYIMxxYNeu8mv13k5QjqbOkN8ws3d33f+lJ7tMkTZOkrKwsz8nJqfViK7t9\nwmuh9l+Xns/4RDn5D0e7jNqRWxjtClBFgxorQ1so7H9bcO5irAA4W4Q5rWibpC6VljtL2l5Nm7+4\ne7G7fyIpX+VhAQAAAEAdCzMcLJfUzcySzayxpFGS5ldpM0/SYEkys7Yqn2a0OcSaAAAAANQgtHDg\n7iWSxklaKGmdpDnuvsbMHjWz4UGzhZK+MLO1kpZI+om7fxFWTQAAAABqFuY5B3L3BZIWVFn3UKXH\nLml88B8AAACAKOIOyQAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAA\nAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAA\nACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACSdQTgws4FmdkfwOMHM\nksMrCwAAAEBdiygcmNnDkh6U9C/BqkaSpodVFAAAAIC6F+mRg29LGi7pkCS5+3ZJrcIqCgAAAEDd\nizQcHHd3l+SSZGYtwisJAAAAQDTERdhujplNlXS+mX1f0p2Sng2vLAAAUC/tWCk9MiLaVdSORwqj\nXQFQ70QUDtx9opldLemApFRJD7n74lArAwAAAFCnThsOzCxW0kJ3HyKJQAAAAAA0UKc958DdSyUd\nNrP4OqgHAAAAQJREes7BUUmrzWyxgisWSZK73xNKVQAAAADqXKTh4LXgPwAAAAANVKQnJL9gZo0l\npQSr8t29OLyyAAAAANS1iMKBmeVIekHSFkkmqYuZ/aO7LwuvNAAAAAB1KdJpRZMkXePu+ZJkZimS\nZkrKDKswAAAAAHUr0jskN6oIBpLk7hskNQqnJAAAAADREOmRgzwz+y9JfwyWvytpRTglAQAAAIiG\nSMPBXZJ+KOkelZ9zsEzSlLCKAgAAAFD3Ig0HcZL+w91/I524a3KT0KoCAABooFasWNEuLi7uD5LS\nFfkUb6A2lEn6qKSkZExmZuau6hpEGg7+KmmIpKJguZmkRZKyv3aJAAAA55C4uLg/tG/fvkdCQsK+\nmJgYj3Y9OHeUlZXZ7t2703bu3PkHScOraxNpWm3q7hXBQMHj5rVQIwAAwLkmPSEh4QDBAHUtJibG\nExISClV+1Kr6NhH2dcjMLqlYMLMsSUe+Zn0AAADnohiCAaIleO/VmAEinVZ0r6S5ZrZdkkvqKOmW\nr18eAAAAgPoi0nCQLKmvpAslfVvSN1UeEgAAAPA1JE14rVZvKrvl19+qN5eb79evX+rEiRO3XnHF\nFYcjaV9WVqbs7OyUBQsWbGrdunVZ2PV9Xa+++mqrSZMmJS5ZsmRTGP3PmDEjfs2aNc1+9atf7ay6\nrXnz5n0PHz78QU3Pzc7OTvnLX/7ycUJCQumZ7DPSaUU/c/cDks6XdLWkaZJ+fyY7AgAAwNmjuLi4\nzvc5Z86c+J49ex45G4JBXfjud79bWF0wiERubu4XEydOTDjT50UaDioSx7ckPePuf5HU+Ex3BgAA\ngOj7yU9+0iE5OblndnZ2t+uvvz75oYceSpTKf+kfN25cp0svvTT1scceS9y+fXvctddee1F6enqP\n9PT0HosWLWohSQcOHIi56aabktLT03v06NEjbfr06edLUlFRkV133XXfSElJSfvWt771jaNHj5ok\nPfXUU21Hjx7dpWL/kyZNajtmzJjOVeuaMWNG629/+9v7JSk/P79xcnJyz1tuuaVrt27deg4fPjx5\n3rx5rS655JLuXbt2TV+yZEnzU9WSn5/fODMzMzUtLa1HWlpaj8WLF7eQyn/t79evX+rQoUO/kZyc\n3HP48OHJZWWnziIjR45M+s53vnNhZmZmalJSUvrMmTPjq7ZZsmRJ8759+3bv0aNHWt++fbuvWrWq\niSTl5eU17dWrV4/u3bunpaSkpK1evbpJpK9t8uTJbb73ve9dKEnr169v3KdPn+7p6ek97r333o4V\n+y0oKGiUlZWV2r1797Ru3br1fP3111tK0qhRo/a//PLLbSJ8S5wQ6bSiz8xsqsovZ/qEmTUR1+UF\nAAA46yxbtqz5f//3f1+wevXqtcXFxdanT5+0vn37npj2s3///tjly5fnS9L111+fPH78+M+vvfba\noo0bNza+9tpru23evHnNv/7rv3YYPHjwgblz527Zs2dPbFZWVo/hw4cf+M1vfpPQrFmzsg0bNqz9\n+9//3mzAgAFpkjR69Oi9PXv2TDt27Ni2Jk2a+PTp09tOnTq1oGptK1asaDlgwIAT67du3dp09uzZ\nmzMzMwt69+7dY8aMGW3y8vLW/+lPfzr/l7/8ZYfBgwd/XFMtHTt2LHnjjTc2NG/e3FevXt0kNzf3\nGx999NE6SVq3bl2zlStXbk5KSirOzMzsvnjx4pbXXnttUdV6Ktu6dWuT9957L3/t2rVNhgwZkjpi\nxIjVlbdnZGQcfe+999Y3atRI8+bNa/XAAw90Xrhw4cdPP/10wt133/35XXfdtffo0aNWUlKizz77\nrFEkr61y/3ffffeFY8aM2T1u3LgvHn/88RNHBJ577rnWV111VeETTzyxs6SkRAcPHoyRpISEhNLj\nx4/bzp07Y9u3bx/x1KJIw8HNkoZKmuju+82sg6SfRLoTAAAA1A9Lly5tOWzYsP0tW7Z0SX711Vfv\nr7w9Nzd3b8Xjt95667yNGzc2q1guKiqK3bdvX8zSpUvPW7hw4fmTJ09uL0nHjh2zTZs2NX7zzTdb\n3nPPPbskqX///kdSUlIOS9J5551XNmDAgIOzZ8+O79Wr19Hi4mLr16/fSVe+LCwsjLvgggtO/Izf\nqVOnYxXtUlJSjlx55ZUHYmJidMkllxx+7LHHOgavp9paunbtWjx69Oiua9eubRYTE6OCgoITN/Dt\n1avXoYsuuqhYknr27Hn4448/Pu2MmJEjR+6NjY1Vr169jnXp0uXYypUrm1bevnfv3thbbrklecuW\nLU3NzIuLi02SLrvsskMTJ07ssG3btsajRo3a16tXr2ORvrbK3n///Zb/8z//87Ek/eAHP/jiF7/4\nRWdJ+uY3v3noBz/4QVJxcXHMjTfeuC87O/vE37VNmzYln376aeP27dtHfJXRiMKBux+W9HKl5R2S\ndkS6EwCo93aslB4ZEe0qascjhdGuAEA95n7qa8q0atXqxJdzd1deXt66IEh8qY+XXnppU0ZGxrGq\nzzezavsdO3bsnl/+8pftU1JSjt566617qmsTGxvrpaWlio2NlSQ1btz4xH5jYmLUtGlTD9qptLTU\nTlXL+PHjO7Zr1674z3/+8ydlZWVq1qzZiRO/mzRpcqLf2NhYlZSUVF/0KV5X1eUHH3yw06BBgw4u\nXrz44/z8/MZXXnllqiT90z/9097LL7/80CuvvBI/bNiwlClTpmxJTU09Fslrq6q6S+AOGzasaNmy\nZfl//vOf42+//fbke+655/Nx48Z9IZUHpebNm5/R+RtMDQIAADiH5OTkFC1cuDD+8OHDVlhYGPO/\n//u/59fUduDAgQeeeOKJdhXLb7/9djNJGjx48IFJkyYlVszVf+utt5oF7YumT5/eWpKWL1/edMOG\nDSdumnvllVce2rFjR+NXXnmlzejRo/eqGsnJyUfXrVvXpLptNamplsLCwtgOHToUx8bGasqUKW1K\nS08/s+aHP/xhpxdffLHav8fLL798QWlpqdasWdNk69atTTIyMo5W3n7gwIHYzp07H5ekqVOntq1Y\nv3bt2sY9evQ49tOf/nTXNddcs3/lypXNqvYdiUsuuaTo2WefbS1Jzz777IlzCTZs2NC4U6dOxfff\nf/+eW2+9dc/777/fXCq/8tPu3bsbpaamnhTgTiXSaUVfiZkNlfQfkmIl/cHdf11DuxslzZV0qbvn\nhVkTAABAfVLXlx4dNGjQ4aFDhxampaX17NSp07HevXsfio+Pr/ab87Rp07aOGTPmwpSUlLTS0lLr\n37//wezs7E9//etfbx87duyF3bt3T3N369y587ElS5Zs+vGPf7xr1KhRySkpKWk9e/Y83KtXr0OV\n+7vhhhv2ffjhh81rurzmNddcU7ho0aJW6enpEX+hramW++67b9fIkSMvmjdv3gUDBw482KxZs9P+\ngr527dpmFSdEV3XxxRcf69evX+oXX3zR6Le//W1B8+bNv/Qr/oMPPrhzzJgxyZMnT25/+eWXH6hY\n/8c//rH13Llz28TFxXlCQkLx448/vn3//v2xkb6+ClOmTPl01KhR35gyZUri8OHD91WsX7hwYavJ\nkye3j4uL8+bNm5fOmDHjE0l68803m/ft2/dQo0aNzmg/drpDS1+VmcVK2qDyS59uk7RcUq67r63S\nrpWk11R+9aNxpwsHWVlZnpcXbn5ImvBaqP3XpeczNion/+Fol1E7mCpR7zBW6inGSr3DWKmn6mCs\nmNkKd8+qvG7VqlVbMjIyqp1WU1cKCwtj4uPjyw4ePBhz2WWXpT7zzDMFAwcOjOheBF/H4MGDL77v\nvvs+HzFixMHqthcUFDTKzc1NevvttzeGXUt1Bg4c2O3NN988ad8jR45Muu666wrvuOOOfdU9rz66\n4447utxwww37q/tbr1q1qm1GRkZSdc8Lc1pRP0mb3H2zux+XNEtSdRN6fyHpSUlHq9kGAACAWnbr\nrbd27d69e1rv3r17XH/99fvCDgZ79uyJTUpKSm/atGlZTcFAkrp27Vp855137tm7d29Upr5XFwzO\nVunp6UdO9beuSZhHDm6UNNTdxwTLt0nq7+7jKrXpK+mn7j7SzJZK+nF1Rw7MbKyksZKUmJiYOWvW\nrFBqrrD6s4bzq1tys2NqeWx7tMuoHR36RLsCVMFYqacYK/UOY6WeqoOxMnjw4Hp55ADntlMdOQjz\nnIPqzrI+kUTMLEbSU5JuP11H7j5N5XdlVlZWlufk5NROhTW4vUEd/v2k4Rz+zW04H64NBWOlnmKs\n1DuMlXqKsQKcJMxDNtskdam03FlS5Z8aWklKl7TUzLZI+qak+Wb2pXQNAAAAoG6EGQ6WS+pmZslm\n1ljSKEnzKza6e6G7t3X3JHdPkvSupOFcrQgAAACIjtDCgbuXSBonaaGkdZLmuPsaM3vUzIaHtV8A\nAAAAX02o9zlw9wWSFlRZ91ANbXPCrAUAAKBeeiQ+8/SNzqS/wtPeN8HMMkeMGLF33rx5n0hScXGx\n2rVrl9GnT59DS5Ys2TR58uQ2eXl5LV588cVPx48f37Fly5aljz766OeV+4iNjc3s1q3bkdLSUrv4\n4ouPzJkzZ0vluytL5Tfiys7OTlmwYMGm1q1bl910001Jf/3rX+PbtGlTsnHjxjXV1fbqq6+2ys3N\nvahTp07HJem6667bN3HixB1V2/3qV79KeOaZZxK3bt3aZPv27as6dOhQIkkzZ86Mf++991o89dRT\nDeTM+brFHZIBAADOMc2aNSvLz89vVlRUZJL0yiuvnJeYmFh8Jn00adKkbP369Ws3bty4plGjRj5p\n0qSEqm3mzJkT37NnzyOtW7cuk6Q777xzz/z58097udCsrKyi9evXr12/fv3a6oKBJA0aNKho8eLF\nGzp27Hi88vpbbrml8PXXXz//4MGDfM/9CvijAQAAnIOuuuqqwrlz554vSTNnzmw9cuTIvV+1r4ED\nBxZt2rSpSdX1M2bMaF35jsPDhg0rSkhIKPmq+6lswIABR1JTU49XXR8TE6Ps7OyDs2fPjq+N/Zxr\nCAcAAADnoNtuu23v7NmzLzh8+LCtW7eu+WWXXXboq/RTXFyshQsXnterV68jVbetWLGi5YABA07b\n75NPPpnw5JNPnjjy8MEHH7RMTU1Nu+KKK7rl5eU1rVg/aNCgi7ds2dLodP1lZWUdeuONN1qeyetA\nuVDPOQAAAED91L9//yPbtm1r8uyzz7YeMmTIGd/04dixYzHdu3dPC/o6eO+99550Y7fCwsK4Cy64\noOzkZ3/ZAw88sLvicXZ29qGCgoIP4+Pjy2bPnh0/cuTIiwsKCj6SpL/97W+bIqmtffv2JTt37mwc\n+atBBcIBAADAOWro0KH7H3744S6LFi3K37Vr1xl9L6w45+BUbWJjY720tFSxsbER91txfoJUfv7A\n+PHjL9yxY0dcxQnHkThy5Ig1bdr0tKEEJ2NaEQAAwDnqrrvu2nP//fdv79ev30lTgmpDcnLy0XXr\n1p10LsKpfPrpp3FlZeXf65csWdK8rKxMiYmJZ3SeQn5+ftOePXuG8poaOo4cAAAARFMElx4Ny0UX\nXVT8s5/9bNfp2j311FMdpk6dmlix/Pnnn38YSf/XXHNN4aJFi1qlp6cfk6Trr78++d133221b9++\nuMTExN4TJkzY/qMf/WhPxfkGDzzwwO7p06df8Nxzz7WLjY31pk2blr344oubY2LKf88eNGjQxS+8\n8EJBUlJS8WOPPdbu6aefbv/FF180ysjISBs8eHDh7NmzCyRp2bJlrZ544onPvsKf5Jxn7h7tGs5I\nVlaW5+WFexPlpAmvhdp/XXo+Y6Ny8h+Odhm145Ezng6JkDFW6inGSr3DWKmn6mCsmNkKd8+qvG7V\nqlVbMjIyTpqf3xAVFBQ0ys3NTXr77bdPe/nS2rJ169a4m2+++RvvvPPOhrra59lm1apVbTMyMpKq\n28a0IgAAAISia9euxXfeeeeevXv31tl3zs2bNzeeNGnS1rraX0PDtCIAAACEZsyYMfvqcn+DBg06\nXJf7a2g4cgAAAFC3ysrKyizaReDcFLz3arySE+EAAACgbn20e/fueAIC6lpZWZnt3r07XtJHNbVh\nWhEAAEAdKikpGbNz584/7Ny5M138UIu6VSbpo5KSkjE1NSAcAAAA1KHMzMxdkoZHuw6gOqRVAAAA\nAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAA\nkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACS\nCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJII\nBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggH\nAAAAAAKEAwAAAACSQg4HZjbUzPLNbJOZTahm+3gzW2tmH5rZX82sa5j1AAAAAKhZaOHAzGIl/U7S\nMElpknLNLK1Ksw8kZbl7b0kvSXoyrHoAAAAAnFqYRw76Sdrk7pvd/bikWZJGVG7g7kvc/XCw+K6k\nziHWAwAAAOAUzN3D6djsRklD3X1MsHybpP7uPq6G9v8paae7P1bNtrGSxkpSYmJi5qxZs0KpucLq\nzwpD7b8uJTc7ppbHtke7jNrRoU+0K0AVjJV6irFS7zBW6qk6GCuDBw9e4e5Zoe8IqCVxIfZt1ayr\nNomY2a2SsiQNqm67u0+TNE2SsrKyPCcnp5ZKrN7tE14Ltf+69HzGJ8rJfzjaZdSO3Ibz4dpQMFbq\nKcZKvcNYqacYK8BJwgwH2yR1qbTcWdJJPzWY2RBJ/yZpkLsfC7EeAAAAAKcQ5jkHyyV1M7NkM2ss\naZSk+ZUbmFlfSVMlDXf3XSHWAgAAAOA0QgsH7l4iaZykhZLWSZrj7mvM7FEzGx40+3dJLSXNNbOV\nZja/hu4AAAAAhCzMaUVy9wWSFlRZ91Clx0PC3D8AAACAyHGHZAAAAACSCAcAAAAAAoQDAAAAAJII\nBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggH\nAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcA\nAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAA\nAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAA\nAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAA\nAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKhhgMzG2pm+Wa2ycwmVLO9iZnNDrb/3cySwqwH\nAAAAQM1CCwdmFivpd5KGSUqTlGtmaVWajZa0z90vlvSUpCfCqgcAAADAqYV55KCfpE3uvtndj0ua\nJWlElTYjJL0QPH5J0lVmZiHWBAAAAKAGYYaDTpK2VlreFqyrto27l0gqlNQmxJoAAAAA1CAuxL6r\nOwLgX6GNzGyspLHBYpGZ5X/N2s4Zg6W2kvZEu45a8XMOKiE8jBUgMoyVM9a1LnYC1JYww8E2SV0q\nLXeWtL2GNtvMLE5SvKS9VTty92mSpoVUZ4NmZnnunhXtOoD6jrECRIaxAjRsYU4rWi6pm5klm1lj\nSaMkza/SZr6kfwwe3yjp/7n7SUcOAAAAAIQvtCMH7l5iZuMkLZQUK+k5d19jZo9KynP3+ZL+S9If\nzWyTyo8YjAqrHgAAAACnZvxQ37CZ2dhgWhaAU2CsAJFhrAANG+EAAAAAgKSQ75AMAAAA4OxBOAAA\nAAAgiXBwzjOzJDP7KHicY2avRrsmIBrM7O3TbF9gZufXVT1AQ2BmS80sK3i8xczaRrsmAKcW5n0O\nECIzM5WfM1IW7VqA+sbMYt299Eye4+7Zp9n+D1+vKqDh4bMIaHg4cnAWCX7lX2dmUyS9L+k2M3vH\nzN43s7lm1jJod6mZvW1mq8yE0RC9AAAEMklEQVTsPTNrFTz3jaDt+2Z2yi9CQH0VvJfXm9kLZvah\nmb1kZs2DXyUfMrM3Jd1kZheZ2etmtiJ473cPnp9oZq8E42NVxVgws6Lg/x3MbJmZrTSzj8zs8mD9\niV89zWx8sO0jM7uvUl3rzOxZM1tjZovMrFlU/kjAKVR9/5rZE2Z2d6Xtj5jZ/cHjn5jZ8mCs/TxY\nV/WzqIuZ/d7M8oL3/s+j88oA1AbCwdknVdKLkq6WNFrSEHe/RFKepPHBDedmS7rX3TMkDZF0RNIu\nSVcHbW+RNDkaxQO1JFXSNHfvLemApIovNkfdfaC7z1L5XdX/2d0zJf1Y0pSgzWRJfwvGxyWS1lTp\n+zuSFrp7H0kZklZW3mhmmZLukNRf0jclfd/M+gabu0n6nbv3lLRf0sjaesFAbaju/Stplso/Fyrc\nLGmumV2j8vd0P0l9JGWa2RVBm1RJL7p7X3cvkPRvwV2Te0saZGa96+QFAah1TCs6+xS4+7tmdp2k\nNElvlR/VVWNJ76j8H+wd7r5cktz9gCSZWQtJ/2lmfSSVSkqJRvFALdnq7m8Fj6dLuid4PFuSgqNo\n2Sr/glPxnCbB/6+U9D1JCqYeFVbpe7mk58yskaR57r6yyvaBkl5x90PBvl6WdLnK7/j+SaX2KyQl\nfY3XCIShpvdvOzPrKClB0j53/9TM7pF0jaQPgue2VHlY+FTBZ1Glfm82s7Eq/17RQeWfTx/WxQsC\nULsIB2efQ8H/TdJid8+tvDH4taa6m1f8SNLnKv8lNEbS0TCLBEJW9T1esVwxPmIk7Q9+/T+zjt2X\nBb+Ofkvld3D/d3d/sVITq+GpknSs0uNSSUwrQn1T0/v3JUk3Smqv8iMJFW0fd/epX+rALEn/N9Zk\nZskqPzp3qbvvM7PnJTWt1aoB1BmmFZ293pU0wMwulqRgznWKpPWSOprZpcH6VmYWJyle5UcUyiTd\nJik2SnUDteFCM7sseJwr6c3KG4MjZp+Y2U1S+UmTZpYRbP6rpLuC9bFmdl7l55pZV0m73P1ZSf+l\n8qlHlS2TdEMw5lpI+rakN2rvpQGhqun9O0vSKJUHhJeCtgsl3VnpfLZOZtaumj7PU3lYKDSzREnD\nQn4NAEJEODhLuftuSbdLmmlmH6o8LHR39+Mqnzv6tJmtkrRY5b/gTJH0j2b2rsqnFB2qtmPg7LBO\n5e/nDyW1lvT7atp8V9LoYByskTQiWH+vpMFmtlrlU396VnlejqSVZvaBys8Z+I/KG939fUnPS3pP\n0t8l/cHdPxBwFqjp/evuayS1kvSZu+8I2i6S9CdJ7wTj5aWgTdU+V6l86tEaSc9JeqtqGwBnD3Ov\nbgYKANRPwZSGV909PcqlAADQ4HDkAAAAAIAkjhwAAAAACHDkAAAAAIAkwgEAAACAAOEAAAAAgCTC\nAQAAAIAA4QAAAACAJOn/A/TXgO6nOlqPAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f1758e28898>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:5:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>101525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>47991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>22407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>4651</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0       101525\n",
       "1  plasmid_1        47991\n",
       "2  plasmid_2        22407\n",
       "3  plasmid_3         4651"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.602038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.418176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.493539</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.602038\n",
       "1  precision  0.418176\n",
       "2   f1_score  0.493539"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP011579.1</th>\n",
       "      <th>CP011580.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.084885</td>\n",
       "      <td>0.256147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.917733</td>\n",
       "      <td>0.213307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.064743</td>\n",
       "      <td>0.091640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.084766</td>\n",
       "      <td>0.062409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Total</td>\n",
       "      <td>1.152127</td>\n",
       "      <td>0.623502</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP011579.1  CP011580.1\n",
       "0  plasmid_0    0.084885    0.256147\n",
       "1  plasmid_1    0.917733    0.213307\n",
       "2  plasmid_2    0.064743    0.091640\n",
       "3  plasmid_3    0.084766    0.062409\n",
       "4      Total    1.152127    0.623502"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP011579.1</td>\n",
       "      <td>0.027727</td>\n",
       "      <td>0.641933</td>\n",
       "      <td>0.095417</td>\n",
       "      <td>0.604386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP011580.1</td>\n",
       "      <td>0.191894</td>\n",
       "      <td>0.344419</td>\n",
       "      <td>0.201544</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.219621</td>\n",
       "      <td>0.986352</td>\n",
       "      <td>0.296961</td>\n",
       "      <td>1.604386</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2  plasmid_3\n",
       "0  CP011579.1   0.027727   0.641933   0.095417   0.604386\n",
       "1  CP011580.1   0.191894   0.344419   0.201544   1.000000\n",
       "2       Total   0.219621   0.986352   0.296961   1.604386"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.191894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.941635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.201544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.191894\n",
       "1  plasmid_1    0.941635\n",
       "2  plasmid_2    0.201544\n",
       "3  plasmid_3    1.000000"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP011579.1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP011580.1</td>\n",
       "      <td>0.454164</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP011579.1    1.000000\n",
       "1  CP011580.1    0.454164"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No high mutual correspondence found.\n"
     ]
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 5, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00420457</td>\n",
       "      <td>-0.0105304</td>\n",
       "      <td>-0.0128796</td>\n",
       "      <td>-0.00955603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[25, 39, 41, 35, 60, 54, 51, 64, 28, 80, 46, 7...</td>\n",
       "      <td>[37, 46, 79, 39, 25, 41, 30]</td>\n",
       "      <td>[41, 39, 42, 80, 46, 79, 36, 64, 51, 59, 55, 6...</td>\n",
       "      <td>[42, 39]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.714709</td>\n",
       "      <td>-0.705393</td>\n",
       "      <td>-0.463285</td>\n",
       "      <td>-0.632122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.718914</td>\n",
       "      <td>-0.715923</td>\n",
       "      <td>-0.476165</td>\n",
       "      <td>-0.641678</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0  \\\n",
       "GC_sum                                        -0.00420457   \n",
       "chain   [25, 39, 41, 35, 60, 54, 51, 64, 28, 80, 46, 7...   \n",
       "gd_sum                                          -0.714709   \n",
       "total                                           -0.718914   \n",
       "\n",
       "                           plasmid_1  \\\n",
       "GC_sum                    -0.0105304   \n",
       "chain   [37, 46, 79, 39, 25, 41, 30]   \n",
       "gd_sum                     -0.705393   \n",
       "total                      -0.715923   \n",
       "\n",
       "                                                plasmid_2   plasmid_3  \n",
       "GC_sum                                         -0.0128796 -0.00955603  \n",
       "chain   [41, 39, 42, 80, 46, 79, 36, 64, 51, 59, 55, 6...    [42, 39]  \n",
       "gd_sum                                          -0.463285   -0.632122  \n",
       "total                                           -0.476165   -0.641678  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_11</th>\n",
       "      <th>plasmid_12</th>\n",
       "      <th>plasmid_13</th>\n",
       "      <th>plasmid_14</th>\n",
       "      <th>plasmid_15</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00392364</td>\n",
       "      <td>-0.0164867</td>\n",
       "      <td>0.00127075</td>\n",
       "      <td>-0.000291464</td>\n",
       "      <td>-3.07637e-05</td>\n",
       "      <td>-0.00105619</td>\n",
       "      <td>-0.0171367</td>\n",
       "      <td>0.00104448</td>\n",
       "      <td>0.00172555</td>\n",
       "      <td>-0.0142262</td>\n",
       "      <td>0.00218943</td>\n",
       "      <td>-0.00955603</td>\n",
       "      <td>-0.0118359</td>\n",
       "      <td>-0.00363888</td>\n",
       "      <td>-0.00014304</td>\n",
       "      <td>-0.00518533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[26, 49, 10, 51, 54, 60, 35, 41, 39, 25, 41, 3...</td>\n",
       "      <td>[60, 65, 36, 64, 51, 54, 60, 35, 41, 39, 25, 41]</td>\n",
       "      <td>[43, 62, 6, 47, 11, 49]</td>\n",
       "      <td>[64, 32, 47, 5, 58, 70, 68, 72, 66]</td>\n",
       "      <td>[77, 23, 47, 9, 48, 2]</td>\n",
       "      <td>[33, 7, 47]</td>\n",
       "      <td>[64, 32, 47]</td>\n",
       "      <td>[53, 18, 48, 1]</td>\n",
       "      <td>[69, 27, 46, 31, 50, 44, 65, 55, 59, 51, 64, 2...</td>\n",
       "      <td>[33, 33, 40, 65, 55, 59, 51, 64, 28]</td>\n",
       "      <td>[43, 13, 46, 79, 37, 46]</td>\n",
       "      <td>[42, 39]</td>\n",
       "      <td>[46, 79]</td>\n",
       "      <td>[75, 81, 55, 65, 75, 16, 78, 70, 68, 82, 74, 66]</td>\n",
       "      <td>[55, 65, 4, 69, 78, 70, 68, 72, 66]</td>\n",
       "      <td>[61, 81, 55]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.189541</td>\n",
       "      <td>-0.66344</td>\n",
       "      <td>-0.0900824</td>\n",
       "      <td>-0.0188737</td>\n",
       "      <td>-0.00215583</td>\n",
       "      <td>-0.0046451</td>\n",
       "      <td>-0.095115</td>\n",
       "      <td>-0.00117337</td>\n",
       "      <td>-0.604879</td>\n",
       "      <td>-0.378054</td>\n",
       "      <td>-0.041496</td>\n",
       "      <td>-0.632122</td>\n",
       "      <td>-0.745809</td>\n",
       "      <td>-0.0088538</td>\n",
       "      <td>-0.0189617</td>\n",
       "      <td>-0.518072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.193465</td>\n",
       "      <td>-0.679926</td>\n",
       "      <td>-0.0888117</td>\n",
       "      <td>-0.0191652</td>\n",
       "      <td>-0.00218659</td>\n",
       "      <td>-0.00570128</td>\n",
       "      <td>-0.112252</td>\n",
       "      <td>-0.000128899</td>\n",
       "      <td>-0.603153</td>\n",
       "      <td>-0.392281</td>\n",
       "      <td>-0.0393066</td>\n",
       "      <td>-0.641678</td>\n",
       "      <td>-0.757645</td>\n",
       "      <td>-0.0124927</td>\n",
       "      <td>-0.0191048</td>\n",
       "      <td>-0.523258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0  \\\n",
       "GC_sum                                        -0.00392364   \n",
       "chain   [26, 49, 10, 51, 54, 60, 35, 41, 39, 25, 41, 3...   \n",
       "gd_sum                                          -0.189541   \n",
       "total                                           -0.193465   \n",
       "\n",
       "                                               plasmid_1  \\\n",
       "GC_sum                                        -0.0164867   \n",
       "chain   [60, 65, 36, 64, 51, 54, 60, 35, 41, 39, 25, 41]   \n",
       "gd_sum                                          -0.66344   \n",
       "total                                          -0.679926   \n",
       "\n",
       "                     plasmid_10                           plasmid_11  \\\n",
       "GC_sum               0.00127075                         -0.000291464   \n",
       "chain   [43, 62, 6, 47, 11, 49]  [64, 32, 47, 5, 58, 70, 68, 72, 66]   \n",
       "gd_sum               -0.0900824                           -0.0188737   \n",
       "total                -0.0888117                           -0.0191652   \n",
       "\n",
       "                    plasmid_12   plasmid_13    plasmid_14       plasmid_15  \\\n",
       "GC_sum            -3.07637e-05  -0.00105619    -0.0171367       0.00104448   \n",
       "chain   [77, 23, 47, 9, 48, 2]  [33, 7, 47]  [64, 32, 47]  [53, 18, 48, 1]   \n",
       "gd_sum             -0.00215583   -0.0046451     -0.095115      -0.00117337   \n",
       "total              -0.00218659  -0.00570128     -0.112252     -0.000128899   \n",
       "\n",
       "                                                plasmid_2  \\\n",
       "GC_sum                                         0.00172555   \n",
       "chain   [69, 27, 46, 31, 50, 44, 65, 55, 59, 51, 64, 2...   \n",
       "gd_sum                                          -0.604879   \n",
       "total                                           -0.603153   \n",
       "\n",
       "                                   plasmid_3                 plasmid_4  \\\n",
       "GC_sum                            -0.0142262                0.00218943   \n",
       "chain   [33, 33, 40, 65, 55, 59, 51, 64, 28]  [43, 13, 46, 79, 37, 46]   \n",
       "gd_sum                             -0.378054                 -0.041496   \n",
       "total                              -0.392281                -0.0393066   \n",
       "\n",
       "         plasmid_5  plasmid_6  \\\n",
       "GC_sum -0.00955603 -0.0118359   \n",
       "chain     [42, 39]   [46, 79]   \n",
       "gd_sum   -0.632122  -0.745809   \n",
       "total    -0.641678  -0.757645   \n",
       "\n",
       "                                               plasmid_7  \\\n",
       "GC_sum                                       -0.00363888   \n",
       "chain   [75, 81, 55, 65, 75, 16, 78, 70, 68, 82, 74, 66]   \n",
       "gd_sum                                        -0.0088538   \n",
       "total                                         -0.0124927   \n",
       "\n",
       "                                  plasmid_8     plasmid_9  \n",
       "GC_sum                          -0.00014304   -0.00518533  \n",
       "chain   [55, 65, 4, 69, 78, 70, 68, 72, 66]  [61, 81, 55]  \n",
       "gd_sum                           -0.0189617     -0.518072  \n",
       "total                            -0.0191048     -0.523258  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_3  from MILP completely contained in  plasmid_0  from greedy.\n",
      "\n",
      "plasmid_5  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n",
      "plasmid_3  from MILP completely contained in  plasmid_5  from greedy.\n",
      "\n",
      "plasmid_5  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_6  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_6  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_6  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
