{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 23 with $\\alpha_1 = 1$, $\\alpha_2 = 5$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 23\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020836.1</td>\n",
       "      <td>37362</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP020836.1        37362"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt0VOW9//HPNxNuAYwCIVwlqZJA\nLkRIBA0oQVGgVbCLqtDaHhVKj5aqhVY5Pa21Vmv1QD0Hj1SwdVkKRcALpUoF2kOK10pQkGsgIjEI\nCAgEwjWX5/fH7PAbwyQMNTtDwvu1FsvZez/z7O/MmsfMZ/az9zbnnAAAAAAgJtoFAAAAADg3EA4A\nAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPDERruA\ns9WhQweXlJQU7TIajSNHjqh169bRLgM45zFWgMgwVs7O6tWr9znnEqJdBxCpRhcOkpKSVFBQEO0y\nGo38/Hzl5eVFuwzgnMdYASLDWDk7ZlYc7RqAs8G0IgAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8\nhAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPL6FAzN7zsz2mNn6WrabmU03syIz+9DM+vlVCwAA\nAIAz8/PIwfOShtexfYSknt6/CZJ+62MtAAAAAM7At3DgnFspaX8dTUZJmu2C3pV0oZl19qseAAAA\nAHWLjeK+u0oqCVne4a3bVbOhmU1Q8OiCEhMTlZ+f72th6z4t9bX/hpQcH/D9/QKagrKyMsYKEAHG\nCtC0RTMcWJh1LlxD59wsSbMkKScnx+Xl5flYlnT7lNd87b8hPT+8tfx+v4CmID8/n7ECRICxAjRt\n0bxa0Q5J3UOWu0naGaVaAAAAgPNeNMPBYknf8a5adIWkUufcaVOKAAAAADQM36YVmdk8SXmSOpjZ\nDkk/l9RMkpxzz0haIumrkookHZV0h1+1AAAAADgz38KBc27sGbY7Sd/3a/8AAAAAzg53SAYAAAAg\niXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJ\ncAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlw\nAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAA\nAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAA\nAADAQzgAAAAAIIlwAAAAAMATG+0CAABo6pKmvBbtEurN88NbR7sEAD7iyAEAAAAASYQDAAAAAB7C\nAQAAAABJhAMAAAAAHsIBAAAAAEk+hwMzG25mhWZWZGZTwmy/2MxWmNkHZvahmX3Vz3oAAAAA1M63\ncGBmAUlPSxohKU3SWDNLq9Hsp5IWOOf6ShojaYZf9QAAAACom59HDvpLKnLObXPOnZT0gqRRNdo4\nSRd4j+Ml7fSxHgAAAAB18PMmaF0llYQs75A0oEabhyQtM7MfSGotaaiP9QAAAACog5/hwMKsczWW\nx0p63jk3zcyulPRHM8twzlV9oSOzCZImSFJiYqLy8/P9qPeUyZkVvvbfkMrKynx/v4CmgLECP/F3\nBUBj4Wc42CGpe8hyN50+bWicpOGS5Jx7x8xaSuogaU9oI+fcLEmzJCknJ8fl5eX5VHLQ7U3sNvd+\nv19AU5Cfn89YgW/4uwKgsfDznINVknqaWbKZNVfwhOPFNdp8IulaSTKz3pJaStrrY00AAAAAauFb\nOHDOVUiaKGmppE0KXpVog5k9bGYjvWaTJX3XzNZKmifpdudczalHAAAAABqAn9OK5JxbImlJjXUP\nhjzeKGmgnzUAAAAAiAx3SAYAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBD\nOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4\nAAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgA\nAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwBMb7QIANF5JU16Ldgn15vnhraNdAgAAUceRAwAAAACS\nCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJII\nBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8voYDMxtuZoVm\nVmRmU2ppc4uZbTSzDWb2Jz/rAQAAAFC7WL86NrOApKclXSdph6RVZrbYObcxpE1PSf8haaBz7oCZ\ndfSrHgAAAAB18/PIQX9JRc65bc65k5JekDSqRpvvSnraOXdAkpxze3ysBwAAAEAdfDtyIKmrpJKQ\n5R2SBtRokyJJZvaWpICkh5xzr9fsyMwmSJogSYmJicrPz/ej3lMmZ1b42n9DKisr8/39wvmLsQJE\nhrECoLHwMxxYmHUuzP57SsqT1E3SG2aW4Zw7+IUnOTdL0ixJysnJcXl5efVebKjbp7zma/8N6fnh\nreX3+4XzF2MFiAxjBUBj4ee0oh2Suocsd5O0M0ybPzvnyp1zH0sqVDAsAAAAAGhgfoaDVZJ6mlmy\nmTWXNEbS4hptFkkaIklm1kHBaUbbfKwJAAAAQC18CwfOuQpJEyUtlbRJ0gLn3AYze9jMRnrNlkr6\n3Mw2Sloh6cfOuc/9qgkAAABA7fw850DOuSWSltRY92DIYydpkvcPAAAAQBRxh2QAAAAAkggHAAAA\nADyEAwAAAACSziIcmNkgM7vDe5xgZsn+lQUAAACgoUUUDszs55IekPQf3qpmkub4VRQAAACAhhfp\nkYOvSxop6YgkOed2SmrrV1EAAAAAGl6k4eCkd9lRJ0lm1tq/kgAAAABEQ6ThYIGZzZR0oZl9V9Lf\nJD3rX1kAAAAAGlpEN0Fzzk01s+skHZKUKulB59xyXysDAAAA0KDOGA7MLCBpqXNuqCQCAQAAANBE\nnXFakXOuUtJRM4tvgHoAAAAARElE04okHZe0zsyWy7tikSQ55+7xpSoAAAAADS7ScPCa9w8AAABA\nExXpCcl/MLPmklK8VYXOuXL/ygIAAADQ0CIKB2aWJ+kPkrZLMkndzezfnHMr/SsNAAAAQEOKdFrR\nNEnXO+cKJcnMUiTNk5TtV2EAAAAAGlakN0FrVh0MJMk5t0VSM39KAgAAABANkR45KDCz30v6o7f8\nLUmr/SkJAAAAQDREGg7ukvR9SfcoeM7BSkkz/CoKAAAAQMOLNBzESvof59xvpFN3TW7hW1UAAABN\n1OrVqzvGxsb+TlKGIp/iDdSHKknrKyoqxmdnZ+8J1yDScPB3SUMllXnLrSQtk5T7pUsEAAA4j8TG\nxv6uU6dOvRMSEg7ExMS4aNeD80dVVZXt3bs3bffu3b+TNDJcm0jTakvnXHUwkPc4rh5qBAAAON9k\nJCQkHCIYoKHFxMS4hISEUgWPWoVvE2FfR8ysX/WCmeVIOvYl6wMAADgfxRAMEC3eZ6/WDBDptKJ7\nJS00s52SnKQukm798uUBAAAAOFdEGg6SJfWVdLGkr0u6QsGQAAAAgC8hacpr9XpT2e2//to5c7n5\n/v37p06dOrXk6quvPhpJ+6qqKuXm5qYsWbKkqF27dlV+1/dlvfrqq22nTZuWuGLFiiI/+p87d278\nhg0bWv3qV7/aXXNbXFxc36NHj35Q23Nzc3NT/vznP3+UkJBQeTb7jHRa0c+cc4ckXSjpOkmzJP32\nbHYEAACAxqO8vLzB97lgwYL49PT0Y40hGDSEb33rW6XhgkEkxo4d+/nUqVMTzvZ5kYaD6sTxNUnP\nOOf+LKn52e4MAAAA0ffjH/+4c3Jycnpubm7PG2+8MfnBBx9MlIK/9E+cOLHr5ZdfnvrII48k7ty5\nM3bYsGGXZGRk9M7IyOi9bNmy1pJ06NChmJtvvjkpIyOjd+/evdPmzJlzoSSVlZXZDTfc8JWUlJS0\nr33ta185fvy4SdKTTz7ZYdy4cd2r9z9t2rQO48eP71azrrlz57b7+te/flCSCgsLmycnJ6ffeuut\nPXr27Jk+cuTI5EWLFrXt169frx49emSsWLEirq5aCgsLm2dnZ6empaX1TktL6718+fLWUvDX/v79\n+6cOHz78K8nJyekjR45MrqqqO4uMHj066Zvf/ObF2dnZqUlJSRnz5s2Lr9lmxYoVcX379u3Vu3fv\ntL59+/Zau3ZtC0kqKChomZmZ2btXr15pKSkpaevWrWsR6WubPn16++985zsXS9LmzZubX3bZZb0y\nMjJ633vvvV2q91tcXNwsJycntVevXmk9e/ZMf/3119tI0pgxYw6+/PLL7SP8SJwS6bSiT81spoKX\nM33czFqI6/ICAAA0OitXroz7y1/+ctG6des2lpeX22WXXZbWt2/fU9N+Dh48GFi1alWhJN14443J\nkyZN+mzYsGFlW7dubT5s2LCe27Zt2/CTn/yk85AhQw4tXLhw+759+wI5OTm9R44ceeg3v/lNQqtW\nraq2bNmy8Z///GergQMHpknSuHHj9qenp6edOHFiR4sWLdycOXM6zJw5s7hmbatXr24zcODAU+tL\nSkpazp8/f1t2dnZxnz59es+dO7d9QUHB5j/96U8XPvroo52HDBnyUW21dOnSpeKNN97YEhcX59at\nW9di7NixX1m/fv0mSdq0aVOrNWvWbEtKSirPzs7utXz58jbDhg0rq1lPqJKSkhbvvfde4caNG1sM\nHTo0ddSoUetCt2dlZR1/7733Njdr1kyLFi1qe//993dbunTpR0899VTC3Xff/dldd921//jx41ZR\nUaFPP/20WSSvLbT/u+++++Lx48fvnThx4uePPfbYqSMCzz33XLtrr7229PHHH99dUVGhw4cPx0hS\nQkJC5cmTJ2337t2BTp06RTy1KNJwcIuk4ZKmOucOmllnST+OdCcAAAA4N+Tn57cZMWLEwTZt2jhJ\n7rrrrjsYun3s2LH7qx+/9dZbF2zdurVV9XJZWVngwIEDMfn5+RcsXbr0wunTp3eSpBMnTlhRUVHz\nN998s80999yzR5IGDBhwLCUl5agkXXDBBVUDBw48PH/+/PjMzMzj5eXl1r9//9OufFlaWhp70UUX\nnfoZv2vXrieq26WkpBy75pprDsXExKhfv35HH3nkkS7e6wlbS48ePcrHjRvXY+PGja1iYmJUXFx8\n6ga+mZmZRy655JJySUpPTz/60UcfnXFGzOjRo/cHAgFlZmae6N69+4k1a9a0DN2+f//+wK233pq8\nffv2lmbmysvLTZKuvPLKI1OnTu28Y8eO5mPGjDmQmZl5ItLXFur9999v89e//vUjSfre9773+S9/\n+ctuknTFFVcc+d73vpdUXl4e841vfONAbm7uqfe1ffv2FZ988knzTp06RXyV0YjCgXPuqKSXQ5Z3\nSdoV6U4AAABwbnCu7mvKtG3b9tSXc+ecCgoKNnlB4gt9vPjii0VZWVknaj7fzML2O2HChH2PPvpo\np5SUlOO33XbbvnBtAoGAq6ysVCAQkCQ1b9781H5jYmLUsmVL57VTZWWl1VXLpEmTunTs2LH8pZde\n+riqqkqtWrU6deJ3ixYtTvUbCARUUVERvug6XlfN5QceeKDr4MGDDy9fvvyjwsLC5tdcc02qJP37\nv//7/quuuurIK6+8Ej9ixIiUGTNmbE9NTT0RyWurKdwlcEeMGFG2cuXKwpdeein+9ttvT77nnns+\nmzhx4udSMCjFxcWd1fkbTA0CAAA4j+Tl5ZUtXbo0/ujRo1ZaWhrzt7/97cLa2g4aNOjQ448/3rF6\n+e23324lSUOGDDk0bdq0xOq5+m+99VYrr33ZnDlz2knSqlWrWm7ZsuXUTXOvueaaI7t27Wr+yiuv\ntB83btx+hZGcnHx806ZNLcJtq01ttZSWlgY6d+5cHggENGPGjPaVlWeeWfP973+/6+zZs8O+Hy+/\n/PJFlZWV2rBhQ4uSkpIWWVlZx0O3Hzp0KNCtW7eTkjRz5swO1es3btzYvHfv3id++tOf7rn++usP\nrlmzplXNviPRr1+/smeffbadJD377LOnziXYsmVL865du5ZPnjx532233bbv/fffj5OCV37au3dv\ns9TU1NMCXF0inVYEAAAAHzT0pUcHDx58dPjw4aVpaWnpXbt2PdGnT58j8fHxYb85z5o1q2T8+PEX\np6SkpFVWVtqAAQMO5+bmfvLrX/9654QJEy7u1atXmnPOunXrdmLFihVFP/rRj/aMGTMmOSUlJS09\nPf1oZmbmkdD+brrppgMffvhhXG2X17z++utLly1b1jYjIyPiL7S11XLfffftGT169CWLFi26aNCg\nQYdbtWp1xl/QN27c2Kr6hOiaLr300hP9+/dP/fzzz5v993//d3FcXNwXfsV/4IEHdo8fPz55+vTp\nna666qpD1ev/+Mc/tlu4cGH72NhYl5CQUP7YY4/tPHjwYCDS11dtxowZn4wZM+YrM2bMSBw5cuSB\n6vVLly5tO3369E6xsbEuLi6ucu7cuR9L0ptvvhnXt2/fI82aNTur/diZDi2da3JyclxBQYGv+0ia\n8pqv/Tek54e3Vl5eXrTLQBPFWAEiw1g5f5nZaudcTui6tWvXbs/Kygo7raahlJaWxsTHx1cdPnw4\n5sorr0x95plnigcNGhTRvQi+jCFDhlx63333fTZq1KjD4bYXFxc3Gzt2bNLbb7+91e9awhk0aFDP\nN99887R9jx49OumGG24oveOOOw6Ee9656I477uh+0003HQz3Xq9du7ZDVlZWUrjnMa0IAADgPHPb\nbbf16NWrV1qfPn1633jjjQf8Dgb79u0LJCUlZbRs2bKqtmAgST169Ci/88479+3fvz8q31HDBYPG\nKiMj41hd73VtmFYEAABwnvnLX/7ycUPur0OHDpXbt29fH0nb8ePHn3O/zr/00kvbo13D2Zo8efK/\ndHSKIwcAAAAAJBEOAAAAAHgIBwAAAAAkEQ4AAAAAeDghGQAAIJoeis8+c6Oz6a/0jPdNMLPsUaNG\n7V+0aNHHklReXq6OHTtmXXbZZUdWrFhRNH369PYFBQWtZ8+e/cmkSZO6tGnTpvLhhx/+LLSPQCCQ\n3bNnz2OVlZV26aWXHluwYMH20LsrS8EbceXm5qYsWbKkqF27dlU333xz0t///vf49u3bV2zdunVD\nuNpeffXVtmPHjr2ka9euJyXphhtuODB16tRdNdv96le/SnjmmWcSS0pKWuzcuXNt586dKyRp3rx5\n8e+9917rJ598cmfkbxqqceQAAADgPNOqVauqwsLCVmVlZSZJr7zyygWJiYnlZ9NHixYtqjZv3rxx\n69atG5o1a+amTZuWULPNggUL4tPT04+1a9euSpLuvPPOfYsXLz7j5UJzcnLKNm/evHHz5s0bwwUD\nSRo8eHDZ8uXLt3Tp0uVk6Ppbb7219PXXX7/w8OHDfM/9F/CmAQAAnIeuvfba0oULF14oSfPmzWs3\nevTo/f9qX4MGDSorKipqUXP93Llz24XecXjEiBFlCQkJFf/qfkINHDjwWGpq6sma62NiYpSbm3t4\n/vz58fWxn/MN4QAAAOA89O1vf3v//PnzLzp69Kht2rQp7sorrzzyr/RTXl6upUuXXpCZmXms5rbV\nq1e3GThw4Bn7feKJJxKeeOKJU0cePvjggzapqalpV199dc+CgoKW1esHDx586fbt25udqb+cnJwj\nb7zxRpuzeR0I8vWcAzMbLul/JAUk/c459+ta2n1D0kJJlzvnCvysCQAAANKAAQOO7dixo8Wzzz7b\nbujQoaVn+/wTJ07E9OrVK83r6/C999572k23SktLYy+66KKq05/9Rffff//e6se5ublHiouLP4yP\nj6+aP39+/OjRoy8tLi5eL0n/+Mc/iiKprVOnThW7d+9uHvmrQTXfwoGZBSQ9Lek6STskrTKzxc65\njTXatZV0j6R/+lULAAAATjd8+PCDP//5z7svW7ascM+ePWf1vbD6nIO62gQCAVdZWalAIBBxv9Xn\nJ0jB8wcmTZp08a5du2KrTziOxLFjx6xly5ZnDCU4nZ/TivpLKnLObXPOnZT0gqRRYdr9UtITko77\nWAsAAABquOuuu/ZNnjx5Z//+/U+bElQfkpOTj2/atOm0cxHq8sknn8RWVQW/169YsSKuqqpKiYmJ\nZ3WeQmFhYcv09HRfXlNT5+e0oq6SSkKWd0gaENrAzPpK6u6ce9XMflRbR2Y2QdIESUpMTFR+fn79\nVxticma9nCdzTigrK/P9/cL5i7ECRIaxgjpFcOlRv1xyySXlP/vZz/acqd2TTz7ZeebMmYnVy599\n9tmHkfR//fXXly5btqxtRkbGCUm68cYbk9999922Bw4ciE1MTOwzZcqUnT/84Q/3VZ9vcP/99++d\nM2fORc8991zHQCDgWrZsWTV79uxtMTHB37MHDx586R/+8IfipKSk8kceeaTjU0891enzzz9vlpWV\nlTZkyJDS+fPnF0vSypUr2z7++OOf/gtvyXnPnHP+dGx2s6Rhzrnx3vK3JfV3zv3AW46R9H+SbnfO\nbTezfEk/OtM5Bzk5Oa6gwN/TEpKmvOZr/w3p+eGtlZeXF+0y0EQxVoDIMFbOX2a22jmXE7pu7dq1\n27Oysk6bn98UFRcXNxs7dmzS22+/fcbLl9aXkpKS2FtuueUr77zzzpaG2mdjs3bt2g5ZWVlJ4bb5\nOa1oh6TuIcvdJIXejKKtpAxJ+Wa2XdIVkhab2RcGEAAAABqnHj16lN9555379u/f32BXyNy2bVvz\nadOmlZy5JcLxc1rRKkk9zSxZ0qeSxkj6ZvVG51yppA7Vy5EeOQAAAEDjMX78+AMNub/Bgwcfbcj9\nNTW+pTjnXIWkiZKWStokaYFzboOZPWxmI/3aLwAAwDmuqqqqyqJdBM5P3mev1is5+XqfA+fcEklL\naqx7sJa2eX7WAgAAcI5Yv3fv3rSEhITSmJgYf07+BMKoqqqyvXv3xktaX1sbX8MBAAAAvqiiomL8\n7t27f7d79+4M+Xv+J1BTlaT1FRUV42trQDgAAABoQNnZ2XskMcUa5yTSKgAAAABJhAMAAAAAHsIB\nAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEA\nAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAA\nAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAA\nAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAA\nSYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAAST6HAzMbbmaFZlZkZlPC\nbJ9kZhvN7EMz+7uZ9fCzHgAAAAC18y0cmFlA0tOSRkhKkzTWzNJqNPtAUo5zro+kFyU94Vc9AAAA\nAOrm55GD/pKKnHPbnHMnJb0gaVRoA+fcCufcUW/xXUndfKwHAAAAQB1ifey7q6SSkOUdkgbU0X6c\npL+G22BmEyRNkKTExETl5+fXU4nhTc6s8LX/hlRWVub7+4XzF2MFiAxjBUBj4Wc4sDDrXNiGZrdJ\nypE0ONx259wsSbMkKScnx+Xl5dVTieHdPuU1X/tvSM8Pby2/3y+cvxgrQGQYKwAaCz/DwQ5J3UOW\nu0naWbORmQ2V9J+SBjvnTvhYDwAAAIA6+HnOwSpJPc0s2cyaSxojaXFoAzPrK2mmpJHOuT0+1gIA\nAADgDHwLB865CkkTJS2VtEnSAufcBjN72MxGes3+S1IbSQvNbI2ZLa6lOwAAAAA+83NakZxzSyQt\nqbHuwZDHQ/3cPwAAAIDIcYdkAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAA\nPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8\nhAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyE\nAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQD\nAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPLHRLgA+27VGemhUtKuoHw+V\nRrsCAACAJo0jBwAAAAAkEQ4AAAAAeAgHAAAAACQRDgAAAAB4CAcAAAAAJPkcDsxsuJkVmlmRmU0J\ns72Fmc33tv/TzJL8rAcAAABA7Xy7lKmZBSQ9Lek6STskrTKzxc65jSHNxkk64Jy71MzGSHpc0q1+\n1QQAAL4kLpENNGl+3uegv6Qi59w2STKzFySNkhQaDkZJesh7/KKk/zUzc845H+sCgNPxhQcAAF+n\nFXWVVBKyvMNbF7aNc65CUqmk9j7WBAAAAKAWfh45sDDrah4RiKSNzGyCpAneYpmZFX7J2s4bQ6QO\nkvZFu4568YtwHxegfjBWgMgwVs5aj4bYCVBf/AwHOyR1D1nuJmlnLW12mFmspHhJ+2t25JybJWmW\nT3U2aWZW4JzLiXYdwLmOsQJEhrECNG1+TitaJamnmSWbWXNJYyQtrtFmsaR/8x5/Q9L/cb4BAAAA\nEB2+HTlwzlWY2URJSyUFJD3nnNtgZg9LKnDOLZb0e0l/NLMiBY8YjPGrHgAAAAB1M36ob9rMbII3\nLQtAHRgrQGQYK0DTRjgAAAAAIMnnOyQDAAAAaDwIBwAAAAAkEQ7Oe2aWZGbrvcd5ZvZqtGsCosHM\n3j7D9iVmdmFD1QM0BWaWb2ble+NqAAAEkElEQVQ53uPtZtYh2jUBqJuf9zmAj8zMFDxnpCratQDn\nGjMLOOcqz+Y5zrncM2z/6perCmh6+FsEND0cOWhEvF/5N5nZDEnvS/q2mb1jZu+b2UIza+O1u9zM\n3jaztWb2npm19Z77htf2fTOr84sQcK7yPsubzewPZvahmb1oZnHer5IPmtmbkm42s0vM7HUzW+19\n9nt5z080s1e88bG2eiyYWZn3385mttLM1pjZejO7ylt/6ldPM5vkbVtvZveF1LXJzJ41sw1mtszM\nWkXlTQLqUPPza2aPm9ndIdsfMrPJ3uMfm9kqb6z9wltX829RdzP7rZkVeJ/9X0TnlQGoD4SDxidV\n0mxJ10kaJ2moc66fpAJJk7wbzs2XdK9zLkvSUEnHJO2RdJ3X9lZJ06NRPFBPUiXNcs71kXRIUvUX\nm+POuUHOuRcUvKv6D5xz2ZJ+JGmG12a6pH9446OfpA01+v6mpKXOucskZUlaE7rRzLIl3SFpgKQr\nJH3XzPp6m3tKeto5ly7poKTR9fWCgfoQ7vMr6QUF/y5Uu0XSQjO7XsHPdH9Jl0nKNrOrvTapkmY7\n5/o654ol/ad31+Q+kgabWZ8GeUEA6h3TihqfYufcu2Z2g6Q0SW8Fj+qquaR3FPwf9i7n3CpJcs4d\nkiQzay3pf83sMkmVklKiUTxQT0qcc295j+dIusd7PF+SvKNouQp+wal+Tgvvv9dI+o4keVOPSmv0\nvUrSc2bWTNIi59yaGtsHSXrFOXfE29fLkq5S8I7vH4e0Xy0p6Uu8RsAPtX1+O5pZF0kJkg445z4x\ns3skXS/pA++5bRQMC5/I+1sU0u8tZjZBwe8VnRX8+/RhQ7wgAPWLcND4HPH+a5KWO+fGhm70fq0J\nd/OKH0r6TMFfQmMkHfezSMBnNT/j1cvV4yNG0kHv1/+z69i5ld6vo19T8A7u/+Wcmx3SxGp5qiSd\nCHlcKYlpRTjX1Pb5fVHSNyR1UvBIQnXbx5xzM7/QgVmS/v9Yk5klK3h07nLn3AEze15Sy3qtGkCD\nYVpR4/WupIFmdqkkeXOuUyRtltTFzC731rc1s1hJ8QoeUaiS9G1JgSjVDdSHi83sSu/xWElvhm70\njph9bGY3S8GTJs0sy9v8d0l3eesDZnZB6HPNrIekPc65ZyX9XsGpR6FWSrrJG3OtJX1d0hv199IA\nX9X2+X1B0hgFA8KLXtulku4MOZ+tq5l1DNPnBQqGhVIzS5Q0wufXAMBHhINGyjm3V9LtkuaZ2YcK\nhoVezrmTCs4dfcrM1kparuAvODMk/ZuZvavglKIjYTsGGodNCn6eP5TUTtJvw7T5lqRx3jjYIGmU\nt/5eSUPMbJ2CU3/SazwvT9IaM/tAwXMG/id0o3PufUnPS3pP0j8l/c4594GARqC2z69zboOktpI+\ndc7t8touk/QnSe944+VFr03NPtcqOPVog6TnJL1Vsw2AxsOcCzcDBQDOTd6UhledcxlRLgUAgCaH\nIwcAAAAAJHHkAAAAAICHIwcAAAAAJBEOAAAAAHgIBwAAAAAkEQ4AAAAAeAgHAAAAACRJ/w82cY0+\nD6xQigAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f137dc3e828>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:5:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>199780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>1780</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0       199780\n",
       "1  plasmid_1         1780"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.036668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.006623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.011220</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.036668\n",
       "1  precision  0.006623\n",
       "2   f1_score  0.011220"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP020836.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.036668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.033296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.069964</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP020836.1\n",
       "0  plasmid_0    0.036668\n",
       "1  plasmid_1    0.033296\n",
       "2      Total    0.069964"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020836.1</td>\n",
       "      <td>0.003499</td>\n",
       "      <td>0.357303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.003499</td>\n",
       "      <td>0.357303</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1\n",
       "0  CP020836.1   0.003499   0.357303\n",
       "1       Total   0.003499   0.357303"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.003499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.357303</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.003499\n",
       "1  plasmid_1    0.357303"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020836.1</td>\n",
       "      <td>0.036668</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP020836.1    0.036668"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No high mutual correspondence found.\n"
     ]
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 5, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>0.000507111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[158, 134, 68, 93, 153, 132, 116, 127, 146, 14...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.252081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.251574</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0\n",
       "GC_sum                                        0.000507111\n",
       "chain   [158, 134, 68, 93, 153, 132, 116, 127, 146, 14...\n",
       "gd_sum                                          -0.252081\n",
       "total                                           -0.251574"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_11</th>\n",
       "      <th>plasmid_12</th>\n",
       "      <th>plasmid_13</th>\n",
       "      <th>plasmid_14</th>\n",
       "      <th>plasmid_15</th>\n",
       "      <th>plasmid_16</th>\n",
       "      <th>plasmid_17</th>\n",
       "      <th>plasmid_18</th>\n",
       "      <th>plasmid_19</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_20</th>\n",
       "      <th>plasmid_21</th>\n",
       "      <th>plasmid_22</th>\n",
       "      <th>plasmid_23</th>\n",
       "      <th>plasmid_24</th>\n",
       "      <th>plasmid_25</th>\n",
       "      <th>plasmid_26</th>\n",
       "      <th>plasmid_27</th>\n",
       "      <th>plasmid_28</th>\n",
       "      <th>plasmid_29</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_30</th>\n",
       "      <th>plasmid_31</th>\n",
       "      <th>plasmid_32</th>\n",
       "      <th>plasmid_33</th>\n",
       "      <th>plasmid_34</th>\n",
       "      <th>plasmid_35</th>\n",
       "      <th>plasmid_36</th>\n",
       "      <th>plasmid_37</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0108455</td>\n",
       "      <td>0.00216688</td>\n",
       "      <td>0.00621464</td>\n",
       "      <td>0.00010106</td>\n",
       "      <td>0.0010158</td>\n",
       "      <td>0.000571987</td>\n",
       "      <td>0.00233207</td>\n",
       "      <td>-0.00177128</td>\n",
       "      <td>0.000458532</td>\n",
       "      <td>-0.00361126</td>\n",
       "      <td>-0.00650758</td>\n",
       "      <td>-0.010466</td>\n",
       "      <td>-0.000405371</td>\n",
       "      <td>0.00372844</td>\n",
       "      <td>0.00230825</td>\n",
       "      <td>-0.00357381</td>\n",
       "      <td>0.00233575</td>\n",
       "      <td>-0.00818517</td>\n",
       "      <td>0.00263032</td>\n",
       "      <td>0.00015045</td>\n",
       "      <td>0.000258675</td>\n",
       "      <td>-0.00335911</td>\n",
       "      <td>-0.000486776</td>\n",
       "      <td>-0.00580587</td>\n",
       "      <td>0.00113659</td>\n",
       "      <td>-0.000612669</td>\n",
       "      <td>-0.0122139</td>\n",
       "      <td>0.00124804</td>\n",
       "      <td>-0.00277812</td>\n",
       "      <td>0.00942651</td>\n",
       "      <td>0.000892318</td>\n",
       "      <td>0.00283191</td>\n",
       "      <td>0.00546473</td>\n",
       "      <td>-0.000956068</td>\n",
       "      <td>-0.00139548</td>\n",
       "      <td>-2.20741e-05</td>\n",
       "      <td>-0.00136909</td>\n",
       "      <td>-0.0209365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[46, 120, 187, 73, 191, 187]</td>\n",
       "      <td>[23, 109, 99, 104, 114, 210, 120, 187, 188, 57...</td>\n",
       "      <td>[134, 68, 90, 153, 131, 116, 128, 146, 143, 148]</td>\n",
       "      <td>[106, 194, 212, 208, 210, 120, 187, 188, 15, 1...</td>\n",
       "      <td>[6, 178, 154, 135, 176, 49, 178]</td>\n",
       "      <td>[178, 154, 135, 176, 52]</td>\n",
       "      <td>[178, 154, 135]</td>\n",
       "      <td>[172, 211, 188, 187, 120, 210, 208, 212, 193, ...</td>\n",
       "      <td>[194, 212, 208, 210, 120, 187, 189, 44, 102, 86]</td>\n",
       "      <td>[192, 74, 102, 38, 192]</td>\n",
       "      <td>[195, 115, 102, 64, 192]</td>\n",
       "      <td>[195, 56, 102, 82, 189, 187, 120, 210, 208]</td>\n",
       "      <td>[151, 84, 158, 133, 68, 93, 153, 132, 116, 127...</td>\n",
       "      <td>[212, 208, 210, 120, 186, 190, 69]</td>\n",
       "      <td>[212, 208, 210, 120, 186, 191, 34]</td>\n",
       "      <td>[193, 212, 208, 210, 120, 186, 189, 58, 189, 1...</td>\n",
       "      <td>[212, 208, 210, 120, 186, 191, 42]</td>\n",
       "      <td>[208, 210, 120, 186, 190, 53, 188, 187, 120]</td>\n",
       "      <td>[212, 208, 210, 120, 186, 191, 71]</td>\n",
       "      <td>[120, 187, 188, 213, 192, 212, 208, 210, 120, ...</td>\n",
       "      <td>[192, 212, 208, 210, 120, 186, 190, 2]</td>\n",
       "      <td>[193, 212, 208, 210, 120, 186, 189, 59]</td>\n",
       "      <td>[192, 212, 208, 210, 120, 186, 190, 67]</td>\n",
       "      <td>[151, 126, 169, 111, 97, 130, 161, 85]</td>\n",
       "      <td>[192, 212, 208, 210, 120, 186, 191, 4]</td>\n",
       "      <td>[192, 212, 208, 210, 120, 186, 191, 66]</td>\n",
       "      <td>[208, 210, 120, 186, 190, 81]</td>\n",
       "      <td>[192, 212, 208, 210, 120, 187, 189, 17, 100, 8...</td>\n",
       "      <td>[193, 212, 208, 210, 120, 186, 189, 7]</td>\n",
       "      <td>[210, 120]</td>\n",
       "      <td>[41, 100, 14, 155]</td>\n",
       "      <td>[5, 100, 61, 96]</td>\n",
       "      <td>[12, 210, 120, 187, 189, 45]</td>\n",
       "      <td>[162, 181, 144, 137, 32, 194, 212, 208, 210, 1...</td>\n",
       "      <td>[107, 63, 101, 119, 159, 172, 107, 1, 102, 3, ...</td>\n",
       "      <td>[194, 40, 154, 135, 27, 194, 212, 208, 210, 12...</td>\n",
       "      <td>[101, 24, 161, 129, 97, 110]</td>\n",
       "      <td>[181, 162, 95, 101, 119]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.698844</td>\n",
       "      <td>-0.124435</td>\n",
       "      <td>-0.526649</td>\n",
       "      <td>-0.039917</td>\n",
       "      <td>-0.00983792</td>\n",
       "      <td>-0.0180829</td>\n",
       "      <td>-0.64321</td>\n",
       "      <td>-0.0268984</td>\n",
       "      <td>-0.0372924</td>\n",
       "      <td>-0.02401</td>\n",
       "      <td>-0.0719507</td>\n",
       "      <td>-0.0529333</td>\n",
       "      <td>-0.112327</td>\n",
       "      <td>-0.045329</td>\n",
       "      <td>-0.00972219</td>\n",
       "      <td>-0.0915539</td>\n",
       "      <td>-0.0112063</td>\n",
       "      <td>-0.0949263</td>\n",
       "      <td>-0.0519579</td>\n",
       "      <td>-0.0197945</td>\n",
       "      <td>-0.00155965</td>\n",
       "      <td>-0.0202478</td>\n",
       "      <td>-0.0361442</td>\n",
       "      <td>-0.153258</td>\n",
       "      <td>-0.00179977</td>\n",
       "      <td>-0.0304599</td>\n",
       "      <td>-0.0816169</td>\n",
       "      <td>-0.00588979</td>\n",
       "      <td>-0.00273009</td>\n",
       "      <td>-0.601436</td>\n",
       "      <td>-0.00698021</td>\n",
       "      <td>-0.00502133</td>\n",
       "      <td>-0.110708</td>\n",
       "      <td>-0.141614</td>\n",
       "      <td>-0.00673702</td>\n",
       "      <td>-0.0144544</td>\n",
       "      <td>-0.0425888</td>\n",
       "      <td>-0.295438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.709689</td>\n",
       "      <td>-0.122268</td>\n",
       "      <td>-0.520434</td>\n",
       "      <td>-0.0398159</td>\n",
       "      <td>-0.00882212</td>\n",
       "      <td>-0.0175109</td>\n",
       "      <td>-0.640878</td>\n",
       "      <td>-0.0286697</td>\n",
       "      <td>-0.0368339</td>\n",
       "      <td>-0.0276212</td>\n",
       "      <td>-0.0784583</td>\n",
       "      <td>-0.0633993</td>\n",
       "      <td>-0.112732</td>\n",
       "      <td>-0.0416005</td>\n",
       "      <td>-0.00741394</td>\n",
       "      <td>-0.0951277</td>\n",
       "      <td>-0.00887059</td>\n",
       "      <td>-0.103111</td>\n",
       "      <td>-0.0493276</td>\n",
       "      <td>-0.0196441</td>\n",
       "      <td>-0.00130097</td>\n",
       "      <td>-0.0236069</td>\n",
       "      <td>-0.036631</td>\n",
       "      <td>-0.159064</td>\n",
       "      <td>-0.000663179</td>\n",
       "      <td>-0.0310726</td>\n",
       "      <td>-0.0938308</td>\n",
       "      <td>-0.00464174</td>\n",
       "      <td>-0.0055082</td>\n",
       "      <td>-0.59201</td>\n",
       "      <td>-0.00608789</td>\n",
       "      <td>-0.00218943</td>\n",
       "      <td>-0.105243</td>\n",
       "      <td>-0.14257</td>\n",
       "      <td>-0.00813251</td>\n",
       "      <td>-0.0144765</td>\n",
       "      <td>-0.0439579</td>\n",
       "      <td>-0.316375</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           plasmid_0  \\\n",
       "GC_sum                    -0.0108455   \n",
       "chain   [46, 120, 187, 73, 191, 187]   \n",
       "gd_sum                     -0.698844   \n",
       "total                      -0.709689   \n",
       "\n",
       "                                                plasmid_1  \\\n",
       "GC_sum                                         0.00216688   \n",
       "chain   [23, 109, 99, 104, 114, 210, 120, 187, 188, 57...   \n",
       "gd_sum                                          -0.124435   \n",
       "total                                           -0.122268   \n",
       "\n",
       "                                              plasmid_10  \\\n",
       "GC_sum                                        0.00621464   \n",
       "chain   [134, 68, 90, 153, 131, 116, 128, 146, 143, 148]   \n",
       "gd_sum                                         -0.526649   \n",
       "total                                          -0.520434   \n",
       "\n",
       "                                               plasmid_11  \\\n",
       "GC_sum                                         0.00010106   \n",
       "chain   [106, 194, 212, 208, 210, 120, 187, 188, 15, 1...   \n",
       "gd_sum                                          -0.039917   \n",
       "total                                          -0.0398159   \n",
       "\n",
       "                              plasmid_12                plasmid_13  \\\n",
       "GC_sum                         0.0010158               0.000571987   \n",
       "chain   [6, 178, 154, 135, 176, 49, 178]  [178, 154, 135, 176, 52]   \n",
       "gd_sum                       -0.00983792                -0.0180829   \n",
       "total                        -0.00882212                -0.0175109   \n",
       "\n",
       "             plasmid_14                                         plasmid_15  \\\n",
       "GC_sum       0.00233207                                        -0.00177128   \n",
       "chain   [178, 154, 135]  [172, 211, 188, 187, 120, 210, 208, 212, 193, ...   \n",
       "gd_sum         -0.64321                                         -0.0268984   \n",
       "total         -0.640878                                         -0.0286697   \n",
       "\n",
       "                                              plasmid_16  \\\n",
       "GC_sum                                       0.000458532   \n",
       "chain   [194, 212, 208, 210, 120, 187, 189, 44, 102, 86]   \n",
       "gd_sum                                        -0.0372924   \n",
       "total                                         -0.0368339   \n",
       "\n",
       "                     plasmid_17                plasmid_18  \\\n",
       "GC_sum              -0.00361126               -0.00650758   \n",
       "chain   [192, 74, 102, 38, 192]  [195, 115, 102, 64, 192]   \n",
       "gd_sum                 -0.02401                -0.0719507   \n",
       "total                -0.0276212                -0.0784583   \n",
       "\n",
       "                                         plasmid_19  \\\n",
       "GC_sum                                    -0.010466   \n",
       "chain   [195, 56, 102, 82, 189, 187, 120, 210, 208]   \n",
       "gd_sum                                   -0.0529333   \n",
       "total                                    -0.0633993   \n",
       "\n",
       "                                                plasmid_2  \\\n",
       "GC_sum                                       -0.000405371   \n",
       "chain   [151, 84, 158, 133, 68, 93, 153, 132, 116, 127...   \n",
       "gd_sum                                          -0.112327   \n",
       "total                                           -0.112732   \n",
       "\n",
       "                                plasmid_20  \\\n",
       "GC_sum                          0.00372844   \n",
       "chain   [212, 208, 210, 120, 186, 190, 69]   \n",
       "gd_sum                           -0.045329   \n",
       "total                           -0.0416005   \n",
       "\n",
       "                                plasmid_21  \\\n",
       "GC_sum                          0.00230825   \n",
       "chain   [212, 208, 210, 120, 186, 191, 34]   \n",
       "gd_sum                         -0.00972219   \n",
       "total                          -0.00741394   \n",
       "\n",
       "                                               plasmid_22  \\\n",
       "GC_sum                                        -0.00357381   \n",
       "chain   [193, 212, 208, 210, 120, 186, 189, 58, 189, 1...   \n",
       "gd_sum                                         -0.0915539   \n",
       "total                                          -0.0951277   \n",
       "\n",
       "                                plasmid_23  \\\n",
       "GC_sum                          0.00233575   \n",
       "chain   [212, 208, 210, 120, 186, 191, 42]   \n",
       "gd_sum                          -0.0112063   \n",
       "total                          -0.00887059   \n",
       "\n",
       "                                          plasmid_24  \\\n",
       "GC_sum                                   -0.00818517   \n",
       "chain   [208, 210, 120, 186, 190, 53, 188, 187, 120]   \n",
       "gd_sum                                    -0.0949263   \n",
       "total                                      -0.103111   \n",
       "\n",
       "                                plasmid_25  \\\n",
       "GC_sum                          0.00263032   \n",
       "chain   [212, 208, 210, 120, 186, 191, 71]   \n",
       "gd_sum                          -0.0519579   \n",
       "total                           -0.0493276   \n",
       "\n",
       "                                               plasmid_26  \\\n",
       "GC_sum                                         0.00015045   \n",
       "chain   [120, 187, 188, 213, 192, 212, 208, 210, 120, ...   \n",
       "gd_sum                                         -0.0197945   \n",
       "total                                          -0.0196441   \n",
       "\n",
       "                                    plasmid_27  \\\n",
       "GC_sum                             0.000258675   \n",
       "chain   [192, 212, 208, 210, 120, 186, 190, 2]   \n",
       "gd_sum                             -0.00155965   \n",
       "total                              -0.00130097   \n",
       "\n",
       "                                     plasmid_28  \\\n",
       "GC_sum                              -0.00335911   \n",
       "chain   [193, 212, 208, 210, 120, 186, 189, 59]   \n",
       "gd_sum                               -0.0202478   \n",
       "total                                -0.0236069   \n",
       "\n",
       "                                     plasmid_29  \\\n",
       "GC_sum                             -0.000486776   \n",
       "chain   [192, 212, 208, 210, 120, 186, 190, 67]   \n",
       "gd_sum                               -0.0361442   \n",
       "total                                 -0.036631   \n",
       "\n",
       "                                     plasmid_3  \\\n",
       "GC_sum                             -0.00580587   \n",
       "chain   [151, 126, 169, 111, 97, 130, 161, 85]   \n",
       "gd_sum                               -0.153258   \n",
       "total                                -0.159064   \n",
       "\n",
       "                                    plasmid_30  \\\n",
       "GC_sum                              0.00113659   \n",
       "chain   [192, 212, 208, 210, 120, 186, 191, 4]   \n",
       "gd_sum                             -0.00179977   \n",
       "total                             -0.000663179   \n",
       "\n",
       "                                     plasmid_31  \\\n",
       "GC_sum                             -0.000612669   \n",
       "chain   [192, 212, 208, 210, 120, 186, 191, 66]   \n",
       "gd_sum                               -0.0304599   \n",
       "total                                -0.0310726   \n",
       "\n",
       "                           plasmid_32  \\\n",
       "GC_sum                     -0.0122139   \n",
       "chain   [208, 210, 120, 186, 190, 81]   \n",
       "gd_sum                     -0.0816169   \n",
       "total                      -0.0938308   \n",
       "\n",
       "                                               plasmid_33  \\\n",
       "GC_sum                                         0.00124804   \n",
       "chain   [192, 212, 208, 210, 120, 187, 189, 17, 100, 8...   \n",
       "gd_sum                                        -0.00588979   \n",
       "total                                         -0.00464174   \n",
       "\n",
       "                                    plasmid_34  plasmid_35  \\\n",
       "GC_sum                             -0.00277812  0.00942651   \n",
       "chain   [193, 212, 208, 210, 120, 186, 189, 7]  [210, 120]   \n",
       "gd_sum                             -0.00273009   -0.601436   \n",
       "total                               -0.0055082    -0.59201   \n",
       "\n",
       "                plasmid_36        plasmid_37                     plasmid_4  \\\n",
       "GC_sum         0.000892318        0.00283191                    0.00546473   \n",
       "chain   [41, 100, 14, 155]  [5, 100, 61, 96]  [12, 210, 120, 187, 189, 45]   \n",
       "gd_sum         -0.00698021       -0.00502133                     -0.110708   \n",
       "total          -0.00608789       -0.00218943                     -0.105243   \n",
       "\n",
       "                                                plasmid_5  \\\n",
       "GC_sum                                       -0.000956068   \n",
       "chain   [162, 181, 144, 137, 32, 194, 212, 208, 210, 1...   \n",
       "gd_sum                                          -0.141614   \n",
       "total                                            -0.14257   \n",
       "\n",
       "                                                plasmid_6  \\\n",
       "GC_sum                                        -0.00139548   \n",
       "chain   [107, 63, 101, 119, 159, 172, 107, 1, 102, 3, ...   \n",
       "gd_sum                                        -0.00673702   \n",
       "total                                         -0.00813251   \n",
       "\n",
       "                                                plasmid_7  \\\n",
       "GC_sum                                       -2.20741e-05   \n",
       "chain   [194, 40, 154, 135, 27, 194, 212, 208, 210, 12...   \n",
       "gd_sum                                         -0.0144544   \n",
       "total                                          -0.0144765   \n",
       "\n",
       "                           plasmid_8                 plasmid_9  \n",
       "GC_sum                   -0.00136909                -0.0209365  \n",
       "chain   [101, 24, 161, 129, 97, 110]  [181, 162, 95, 101, 119]  \n",
       "gd_sum                    -0.0425888                 -0.295438  \n",
       "total                     -0.0439579                 -0.316375  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_35  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
