{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 109 with $\\alpha_1 = 1$, $\\alpha_2 = 10$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 109\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012925.1</td>\n",
       "      <td>2096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012927.1</td>\n",
       "      <td>3372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012928.1</td>\n",
       "      <td>62920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012929.1</td>\n",
       "      <td>96042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012926.1</td>\n",
       "      <td>37697</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP012925.1         2096\n",
       "1  CP012927.1         3372\n",
       "2  CP012928.1        62920\n",
       "3  CP012929.1        96042\n",
       "4  CP012926.1        37697"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt8FeWdx/HvLwm3AKJACFdJVAKE\nQIAEUMASFCVsBeziBRS7KhRXy3oBq+xuq12rtbpSW7rSAl3XUqkKqJRWLFCX1FutBAW5BqJyExEQ\nCAYEcvntH5mwMSThpDIJJJ/368XLMzPPeeZ3zus85nzPPDNj7i4AAAAAiKrtAgAAAACcGQgHAAAA\nACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAEYmq7gOpq\n3bq1JyQk1HYZZ43Dhw+radOmtV0GcMZjrACRYaxUz6pVq/a5e1xt1wFE6qwLBwkJCcrOzq7tMs4a\nWVlZysjIqO0ygDMeYwWIDGOlesxsW23XAFQH04oAAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAO\nAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAiEFg7M7Gkz22Nm6yrZbmY2w8xyzewDM+sbVi0AAAAA\nTi3MIwfPSMqsYvsISV2Cf5Mk/TLEWgAAAACcQmjhwN1fl7S/iiajJc31Eu9IOtfM2oVVDwAAAICq\nxdTivjtI2lFmeWew7tPyDc1skkqOLig+Pl5ZWVk1UV+dkJ+fz/uF0Kz9JK+2SzhtEltEM1aACPB3\nBajbajMcWAXrvKKG7j5b0mxJSk9P94yMjBDLqluysrLE+4Ww3Dztldou4bR5JrMpYwWIAH9XgLqt\nNsPBTkmdyix3lLSrlmoBACA0CXUsSAOou2rzUqaLJX07uGrRxZLy3P2kKUUAAAAAakZoRw7M7DlJ\nGZJam9lOSQ9KaiBJ7v4rSUsk/YOkXElHJN0SVi3VxS88AAAAqI9CCwfuPu4U213Sd8PaPwAAAIDq\n4Q7JAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAE\nCAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQI\nBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgH\nAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcA\nAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAA\nAAAkEQ4AAAAABEINB2aWaWY5ZpZrZtMq2H6+ma0ws/fN7AMz+4cw6wEAAABQudDCgZlFS3pK0ghJ\nyZLGmVlyuWbflzTf3ftIGitpZlj1AAAAAKhamEcO+kvKdfeP3P24pOcljS7XxiWdEzxuIWlXiPUA\nAAAAqEJMiH13kLSjzPJOSQPKtfmhpGVm9i+SmkoaFmI9AAAAAKoQZjiwCtZ5ueVxkp5x9+lmdomk\n35pZirsXf6Ujs0mSJklSfHy8srKywqj3hKk9C0Ptvybl5+eH/n6h/mKsAJFhrAA4W4QZDnZK6lRm\nuaNOnjY0QVKmJLn7X82ssaTWkvaUbeTusyXNlqT09HTPyMgIqeQSN097JdT+a9IzmU0V9vuF+oux\nAkSGsQLgbBHmOQcrJXUxs0Qza6iSE44Xl2uzXdLlkmRm3SU1lrQ3xJoAAAAAVCK0cODuhZImS1oq\naaNKrkq03sweMrNRQbOpkr5jZmskPSfpZncvP/UIAAAAQA0Ic1qR3H2JpCXl1j1Q5vEGSYPCrAEA\nAABAZLhDMgAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMA\nAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAA\nAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAA\nAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAA\nAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAAB\nwgEAAAAASYQDAAAAAAHCAQAAAABJIYcDM8s0sxwzyzWzaZW0uc7MNpjZejP7XZj1AAAAAKhcTFgd\nm1m0pKckXSFpp6SVZrbY3TeUadNF0r9KGuTuB8ysTVj1AAAAAKhamEcO+kvKdfeP3P24pOcljS7X\n5juSnnL3A5Lk7ntCrAcAAABAFUI7ciCpg6QdZZZ3ShpQrk2SJJnZW5KiJf3Q3f9UviMzmyRpkiTF\nx8crKysrjHpPmNqzMNT+a1J+fn7o7xfqL8YKEBnGCoCzRZjhwCpY5xXsv4ukDEkdJb1hZinufvAr\nT3KfLWm2JKWnp3tGRsZpL7asm6e9Emr/NemZzKYK+/1C/cVYASLDWAFwtghzWtFOSZ3KLHeUtKuC\nNr939wJ3/1hSjkrCAgAAAIAaFmY4WCmpi5klmllDSWMlLS7XZpGkoZJkZq1VMs3ooxBrAgAAAFCJ\n0MKBuxdKmixpqaSNkua7+3oze8jMRgXNlkr63Mw2SFoh6Xvu/nlYNQEAAACoXJjnHMjdl0haUm7d\nA2Ueu6QpwT8AAAAAtYg7JAMAAACQRDgAAAAAECAcAAAAAJBUjXBgZoPN7JbgcZyZJYZXFgAAAICa\nFlE4MLMHJd0v6V+DVQ0kPRtWUQAAAABqXqRHDr4laZSkw5Lk7rskNQ+rKAAAAAA1L9JwcDy47KhL\nkpk1Da8kAAAAALUh0nAw38xmSTrXzL4j6c+S5oRXFgAAAICaFtFN0Nz9CTO7QtIhSV0lPeDuy0Ot\nDAAAAECNOmU4MLNoSUvdfZgkAgEAAABQR51yWpG7F0k6YmYtaqAeAAAAALUkomlFko5KWmtmyxVc\nsUiS3P3OUKoCAAAAUOMiDQevBP8AAAAA1FGRnpD8GzNrKCkpWJXj7gXhlQUAAACgpkUUDswsQ9Jv\nJG2VZJI6mdk/ufvr4ZUGAAAAoCZFOq1ouqQr3T1HkswsSdJzktLCKgwAAABAzYr0JmgNSoOBJLn7\nZkkNwikJAAAAQG2I9MhBtpn9t6TfBss3SloVTkkAAAAAakOk4eB2Sd+VdKdKzjl4XdLMsIoCAAAA\nUPMiDQcxkn7u7j+VTtw1uVFoVQEAANRRq1atahMTE/NrSSmKfIo3cDoUS1pXWFg4MS0tbU9FDSIN\nB69JGiYpP1huImmZpIFfu0QAAIB6JCYm5tdt27btHhcXdyAqKsprux7UH8XFxbZ3797k3bt3/1rS\nqIraRJpWG7t7aTBQ8Dj2NNQIAABQ36TExcUdIhigpkVFRXlcXFyeSo5aVdwmwr4Om1nf0gUzS5f0\n5desDwAAoD6KIhigtgSfvUozQKTTiu6StMDMdklySe0lXf/1ywMAAABwpog0HCRK6iPpfEnfknSx\nSkICAAAAvoaEaa+c1pvKbv3JN8+Yy83379+/6xNPPLHjG9/4xpFI2hcXF2vgwIFJS5YsyW3ZsmVx\n2PV9XX/84x+bT58+PX7FihW5YfQ/b968FuvXr2/y4x//eHf5bbGxsX2OHDnyfmXPHThwYNLvf//7\nD+Pi4oqqs89IpxX9wN0PSTpX0hWSZkv6ZXV2BAAAgLNHQUFBje9z/vz5LXr06PHl2RAMasKNN96Y\nV1EwiMS4ceM+f+KJJ+Kq+7xIw0Fp4vimpF+5++8lNazuzgAAAFD7vve977VLTEzsMXDgwC4jR45M\nfOCBB+Klkl/6J0+e3KFfv35dH3744fhdu3bFDB8+/MKUlJTuKSkp3ZctW9ZUkg4dOhR17bXXJqSk\npHTv3r178rPPPnuuJOXn59tVV111QVJSUvI3v/nNC44ePWqS9OSTT7aeMGFCp9L9T58+vfXEiRM7\nlq9r3rx5Lb/1rW8dlKScnJyGiYmJPa6//vrOXbp06TFq1KjERYsWNe/bt2+3zp07p6xYsSK2qlpy\ncnIapqWldU1OTu6enJzcffny5U2lkl/7+/fv3zUzM/OCxMTEHqNGjUosLq46i4wZMybhhhtuOD8t\nLa1rQkJCynPPPdeifJsVK1bE9unTp1v37t2T+/Tp023NmjWNJCk7O7txz549u3fr1i05KSkpee3a\ntY0ifW0zZsxo9e1vf/t8Sdq0aVPD3r17d0tJSel+1113tS/d77Zt2xqkp6d37datW3KXLl16/OlP\nf2omSWPHjj340ksvtYrwI3FCpNOKPjGzWSq5nOljZtZIXJcXAADgrPP666/H/uEPfzhv7dq1GwoK\nCqx3797Jffr0OTHt5+DBg9ErV67MkaSRI0cmTpky5bPhw4fnb9mypeHw4cO7fPTRR+v/7d/+rd3Q\noUMPLViwYOu+ffui09PTu48aNerQT3/607gmTZoUb968ecPf/va3JoMGDUqWpAkTJuzv0aNH8rFj\nx3Y2atTIn3322dazZs3aVr62VatWNRs0aNCJ9Tt27Gj8wgsvfJSWlratV69e3efNm9cqOzt70+9+\n97tzH3nkkXZDhw79sLJa2rdvX/jGG29sjo2N9bVr1zYaN27cBevWrdsoSRs3bmyyevXqjxISEgrS\n0tK6LV++vNnw4cPzy9dT1o4dOxq9++67ORs2bGg0bNiwrqNHj15bdntqaurRd999d1ODBg20aNGi\n5vfdd1/HpUuXfviLX/wi7o477vjs9ttv33/06FErLCzUJ5980iCS11a2/zvuuOP8iRMn7p08efLn\njz766IkjAk8//XTLyy+/PO+xxx7bXVhYqC+++CJKkuLi4oqOHz9uu3fvjm7btm3EU4siDQfXScqU\n9IS7HzSzdpK+F+lOAAAAcGbIyspqNmLEiIPNmjVzSX7FFVccLLt93Lhx+0sfv/XWW+ds2bKlSely\nfn5+9IEDB6KysrLOWbp06bkzZsxoK0nHjh2z3Nzchm+++WazO++8c48kDRgw4MukpKQjknTOOecU\nDxo06IsXXnihRc+ePY8WFBRY//79T7ryZV5eXsx555134mf8Dh06HCttl5SU9OVll112KCoqSn37\n9j3y8MMPtw9eT4W1dO7cuWDChAmdN2zY0CQqKkrbtm07cQPfnj17Hr7wwgsLJKlHjx5HPvzww1PO\niBkzZsz+6Oho9ezZ81inTp2OrV69unHZ7fv374++/vrrE7du3drYzLygoMAk6ZJLLjn8xBNPtNu5\nc2fDsWPHHujZs+exSF9bWe+9916zV1999UNJuu222z7/0Y9+1FGSLr744sO33XZbQkFBQdQ111xz\nYODAgSfe11atWhVu3769Ydu2bSO+ymhE4cDdj0h6qczyp5I+jXQnAAAAODO4V31NmebNm5/4cu7u\nys7O3hgEia/0sXDhwtzU1NRj5Z9vZhX2O2nSpH2PPPJI26SkpKPjx4/fV1Gb6OhoLyoqUnR0tCSp\nYcOGJ/YbFRWlxo0be9BORUVFVlUtU6ZMad+mTZuCF1988ePi4mI1adLkxInfjRo1OtFvdHS0CgsL\nKy66itdVfvn+++/vMGTIkC+WL1/+YU5OTsPLLrusqyT98z//8/5LL7308Msvv9xixIgRSTNnztza\ntWvXY5G8tvIqugTuiBEj8l9//fWcF198scXNN9+ceOedd342efLkz6WSoBQbG1ut8zeYGgQAAFCP\nZGRk5C9durTFkSNHLC8vL+rPf/7zuZW1HTx48KHHHnusTeny22+/3USShg4demj69OnxpXP133rr\nrSZB+/xnn322pSStXLmy8ebNm0/cNPeyyy47/OmnnzZ8+eWXW02YMGG/KpCYmHh048aNjSraVpnK\nasnLy4tu165dQXR0tGbOnNmqqOjUM2u++93vdpg7d26F78dLL710XlFRkdavX99ox44djVJTU4+W\n3X7o0KHojh07HpekWbNmtS5dv2HDhobdu3c/9v3vf3/PlVdeeXD16tVNyvcdib59++bPmTOnpSTN\nmTPnxLkEmzdvbtihQ4eCqVOn7hs/fvy+9957L1YqufLT3r17G3Tt2vWkAFeVSKcVAQAAIAQ1fenR\nIUOGHMnMzMxLTk7u0aFDh2O9evU63KJFiwq/Oc+ePXvHxIkTz09KSkouKiqyAQMGfDFw4MDtP/nJ\nT3ZNmjTp/G7duiW7u3Xs2PHYihUrcu+99949Y8eOTUxKSkru0aPHkZ49ex4u29/VV1994IMPPoit\n7PKaV155Zd6yZcuap6SkRPyFtrJa7r777j1jxoy5cNGiRecNHjz4iyZNmpzyF/QNGzY0KT0huryL\nLrroWP/+/bt+/vnnDX72s59ti42N/cqv+Pfff//uiRMnJs6YMaPtpZdeeqh0/W9/+9uWCxYsaBUT\nE+NxcXEFjz766K6DBw9GR/r6Ss2cOXP72LFjL5g5c2b8qFGjDpSuX7p0afMZM2a0jYmJ8djY2KJ5\n8+Z9LElvvvlmbJ8+fQ43aNCgWvuxUx1aOtOkp6d7dnZ2qPtImPZKqP3XpGcymyojI6O2y0AdxVgB\nIsNYqb/MbJW7p5ddt2bNmq2pqakVTqupKXl5eVEtWrQo/uKLL6IuueSSrr/61a+2DR48OKJ7EXwd\nQ4cOvejuu+/+bPTo0V9UtH3btm0Nxo0bl/D2229vCbuWigwePLjLm2++edK+x4wZk3DVVVfl3XLL\nLQcqet6Z6JZbbul09dVXH6zovV6zZk3r1NTUhIqex7QiAACAemb8+PGdu3XrltyrV6/uI0eOPBB2\nMNi3b190QkJCSuPGjYsrCwaS1Llz54Jbb7113/79+2vlO2pFweBslZKS8mVV73VlmFYEAABQz/zh\nD3/4uCb317p166KtW7eui6TtxIkTz7hf51988cWttV1DdU2dOvXvOjrFkQMAAAAAkggHAAAAAAKE\nAwAAAACSCAcAAAAAApyQDAAAUJt+2CLt1I2q01/eKe+bYGZpo0eP3r9o0aKPJamgoEBt2rRJ7d27\n9+EVK1bkzpgxo1V2dnbTuXPnbp8yZUr7Zs2aFT300EOfle0jOjo6rUuXLl8WFRXZRRdd9OX8+fO3\nlr27slRyI66BAwcmLVmyJLdly5bF1157bcJrr73WolWrVoVbtmxZX1Ft77//fuNbbrklYcOGDbHT\npk37pOx+Fy5ceM699957fnFxscaPH7/vxz/+8e7yz3/11VebTZ06tdPmzZtj58yZ81Hp5Ud37doV\nc/311ye+8cYbdeaKRGHgyAEAAEA906RJk+KcnJwm+fn5Jkkvv/zyOfHx8QXV6aNRo0bFmzZt2rBl\ny5b1DRo08OnTp8eVbzN//vwWPXr0+LJly5bFknTrrbfuW7x4cZVfztu0aVP485//fPttt932lTBS\nWFioe+655/wlS5Zs3rx58/oXX3yx5apVqxqXf/4FF1xw/H/+53+2jhw58vOy69u3b18YHx9fsGzZ\nsqbVeZ31DeEAAACgHrr88svzFixYcK4kPffccy3HjBmz/+/ta/Dgwfm5ubmNyq+fN29ey7J3HB4x\nYkR+XFxcYVV9dejQoXDIkCFHGjRo8JU79WZlZTXt3LnzseTk5OONGzf2f/zHf9y/cOHCc8s/v2vX\nrscHDBjwZVTUyV9zr7766oNz585tVa0XV88QDgAAAOqhm266af8LL7xw3pEjR2zjxo2xl1xyyeG/\np5+CggItXbr0nJ49e35ZftuqVauaDRo06JT9Pv7443GPP/74SUceytqxY0fDDh06HC9d7tix4/FP\nPvmkoSTdfffd7efNm9fiVPsZNGjQ4XfffbfZqdrVZ6Gec2BmmZJ+Lila0q/d/SeVtLtG0gJJ/dw9\nO8yaAAAAIA0YMODLnTt3NpozZ07LYcOG5VX3+ceOHYvq1q1bctDXF3fddddJN93Ky8uLOe+884pP\nfvZX3XfffXtP1cbdT1pnZi5JP/vZz3ZFUnP79u0L9+zZ0zCStvVVaOHAzKIlPSXpCkk7Ja00s8Xu\nvqFcu+aS7pT0t7BqAQAAwMkyMzMPPvjgg52WLVuWs2fPnmp9Lyw956CqNtHR0V5UVKTo6OivV6ik\n888//8SRAknauXNnw/bt21frPIkjR45Yo0aNThlW6rMwpxX1l5Tr7h+5+3FJz0saXUG7H0l6XNLR\nEGsBAABAObfffvu+qVOn7urfv/9JU4JOh8TExKMbN2486VyEv8eQIUMOb926tfGmTZsaHj161F56\n6aWWY8aMOXjqZ/6/devWNU5KSgrltdYVYU4r6iBpR5nlnZIGlG1gZn0kdXL3P5rZvZV1ZGaTJE2S\npPj4eGVlZZ3+asuY2rPK82TOKvn5+aG/X6i/GCtAZBgrqFIElx4Ny4UXXljwgx/8YM+p2j355JPt\nZs2aFV+6/Nlnn30QSf9XXnll3rJly5qnpKQck6SRI0cmvvPOO80PHDgQEx8f32vatGm77rnnnn2l\n5xvcd999e7dv3x7Tr1+/5MOHD0ebmc+aNSt+48aN61q2bFk8ffr07ZmZmUlFRUW64YYb9qWnpx+V\nSs456Nev3+Ebb7wx7y9/+Uvsddddd9GhQ4eiX3vttXMfeeSR9rm5ueslafny5c0zMzOrPYWqPrGK\n5m+dlo7NrpU03N0nBss3Serv7v8SLEdJ+l9JN7v7VjPLknTvqc45SE9P9+zscE9LSJj2Sqj916Rn\nMpsqIyOjtstAHVWnxkrqFmXkPFjbZZweP+Tv3pmmTo0V/q5Ui5mtcvf0suvWrFmzNTU19aT5+XXR\ntm3bGowbNy7h7bffPiPuLZCent711VdfzY2Liyuq7Vpq05o1a1qnpqYmVLQtzGlFOyV1KrPcUVLZ\nk0WaS0qRlGVmWyVdLGmxmX1lAAEAAODs1Llz54Jbb7113/79+2v9Cpm7du2Kueuuuz6r78HgVMKc\nVrRSUhczS5T0iaSxkm4o3ejueZJaly5HeuQAAAAAZ4+JEyceqO0apJIrFd10003VOkehPgotHLh7\noZlNlrRUJZcyfdrd15vZQ5Ky3X1xWPsGAAAh+XS19MOKri9yFqq9KXjFxcXFFhUVFc7cbqAKxcXF\nJqnSKzaFep8Dd18iaUm5dQ9U0jYjzFoAAADOEOv27t2bHBcXl0dAQE0qLi62vXv3tpC0rrI2oYYD\nAAAAfFVhYeHE3bt3/3r37t0pCvf8T6C8YknrCgsLJ1bWgHBQ13H4FwCAM0paWtoeSaNquw6gIqRV\nAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMA\nAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAA\nAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAA\nAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAA\nkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACS\nQg4HZpZpZjlmlmtm0yrYPsXMNpjZB2b2mpl1DrMeAAAAAJULLRyYWbSkpySNkJQsaZyZJZdr9r6k\ndHfvJWmhpMfDqgcAAABA1cI8ctBfUq67f+TuxyU9L2l02QbuvsLdjwSL70jqGGI9AAAAAKpg7h5O\nx2bXSMp094nB8k2SBrj75Era/5ek3e7+cAXbJkmaJEnx8fFpzz//fCg1l1r7SV6o/dekxCbH1OzY\nrtou4/Ro17u2K0A5jJUzFGPljMNYOUPVwFgZOnToKndPD31HwGkSE2LfVsG6CpOImY2XlC5pSEXb\n3X22pNmSlJ6e7hkZGaepxIrdPO2VUPuvSc+kfqyMnAdru4zTY1zd+eNaVzBWzlCMlTMOY+UMxVgB\nThJmONgpqVOZ5Y6STvqpwcyGSfp3SUPc/ViI9QAAAACoQpjnHKyU1MXMEs2soaSxkhaXbWBmfSTN\nkjTK3feEWAsAAACAUwgtHLh7oaTJkpZK2ihpvruvN7OHzGxU0Ow/JTWTtMDMVpvZ4kq6AwAAABCy\nMKcVyd2XSFpSbt0DZR4PC3P/AAAAACLHHZIBAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAA\nAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAA\nSCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABI\nIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgi\nHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIc\nAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIhBoOzCzTzHLMLNfMplWwvZGZvRBs/5uZ\nJYRZDwAAAIDKhRYOzCxa0lOSRkhKljTOzJLLNZsg6YC7XyTpSUmPhVUPAAAAgKqFeeSgv6Rcd//I\n3Y9Lel7S6HJtRkv6TfB4oaTLzcxCrAkAAABAJcIMBx0k7SizvDNYV2Ebdy+UlCepVYg1AQAAAKhE\nTIh9V3QEwP+ONjKzSZImBYv5ZpbzNWurN4ZKrSXtq+06Tov/4KASwsNYASLDWKm2zjWxE+B0CTMc\n7JTUqcxyR0m7Kmmz08xiJLWQtL98R+4+W9LskOqs08ws293Ta7sO4EzHWAEiw1gB6rYwpxWtlNTF\nzBLNrKGksZIWl2uzWNI/BY/DfVxcAAAFHUlEQVSvkfS/7n7SkQMAAAAA4QvtyIG7F5rZZElLJUVL\netrd15vZQ5Ky3X2xpP+W9Fszy1XJEYOxYdUDAAAAoGrGD/V1m5lNCqZlAagCYwWIDGMFqNsIBwAA\nAAAkhXyHZAAAAABnD8IBAAAAAEmEg3rPzBLMbF3wOMPM/ljbNQG1wczePsX2JWZ2bk3VA9QFZpZl\nZunB461m1rq2awJQtTDvc4AQmZmp5JyR4tquBTjTmFm0uxdV5znuPvAU2//h61UF1D38LQLqHo4c\nnEWCX/k3mtlMSe9JusnM/mpm75nZAjNrFrTrZ2Zvm9kaM3vXzJoHz30jaPuemVX5RQg4UwWf5U1m\n9hsz+8DMFppZbPCr5ANm9qaka83sQjP7k5mtCj773YLnx5vZy8H4WFM6FswsP/hvOzN73cxWm9k6\nM7s0WH/iV08zmxJsW2dmd5epa6OZzTGz9Wa2zMya1MqbBFSh/OfXzB4zszvKbP+hmU0NHn/PzFYG\nY+0/gnXl/xZ1MrNfmll28Nn/j9p5ZQBOB8LB2aerpLmSrpA0QdIwd+8rKVvSlOCGcy9IusvdUyUN\nk/SlpD2SrgjaXi9pRm0UD5wmXSXNdvdekg5JKv1ic9TdB7v78yq5q/q/uHuapHslzQzazJD0l2B8\n9JW0vlzfN0ha6u69JaVKWl12o5mlSbpF0gBJF0v6jpn1CTZ3kfSUu/eQdFDSmNP1goHToaLPr6Tn\nVfJ3odR1khaY2ZUq+Uz3l9RbUpqZfSNo01XSXHfv4+7bJP17cNfkXpKGmFmvGnlBAE47phWdfba5\n+ztmdpWkZElvlRzVVUNJf1XJ/7A/dfeVkuTuhyTJzJpK+i8z6y2pSFJSbRQPnCY73P2t4PGzku4M\nHr8gScFRtIEq+YJT+pxGwX8vk/RtSQqmHuWV63ulpKfNrIGkRe6+utz2wZJedvfDwb5eknSpSu74\n/nGZ9qskJXyN1wiEobLPbxszay8pTtIBd99uZndKulLS+8Fzm6kkLGxX8LeoTL/XmdkklXyvaKeS\nv08f1MQLAnB6EQ7OPoeD/5qk5e4+ruzG4Neaim5ecY+kz1TyS2iUpKNhFgmErPxnvHS5dHxESToY\n/PpfvY7dXw9+Hf2mSu7g/p/uPrdME6vkqZJ0rMzjIklMK8KZprLP70JJ10hqq5IjCaVtH3X3WV/p\nwCxB/z/WZGaJKjk618/dD5jZM5Ian9aqAdQYphWdvd6RNMjMLpKkYM51kqRNktqbWb9gfXMzi5HU\nQiVHFIol3SQpupbqBk6H883skuDxOElvlt0YHDH72MyulUpOmjSz1GDza5JuD9ZHm9k5ZZ9rZp0l\n7XH3OZL+WyVTj8p6XdLVwZhrKulbkt44fS8NCFVln9/nJY1VSUBYGLRdKunWMuezdTCzNhX0eY5K\nwkKemcVLGhHyawAQIsLBWcrd90q6WdJzZvaBSsJCN3c/rpK5o78wszWSlqvkF5yZkv7JzN5RyZSi\nwxV2DJwdNqrk8/yBpJaSfllBmxslTQjGwXpJo4P1d0kaamZrVTL1p0e552VIWm1m76vknIGfl93o\n7u9JekbSu5L+JunX7v6+gLNAZZ9fd18vqbmkT9z906DtMkm/k/TXYLwsDNqU73ONSqYerZf0tKS3\nyrcBcPYw94pmoADAmSmY0vBHd0+p5VIAAKhzOHIAAAAAQBJHDgAAAAAEOHIAAAAAQBLhAAAAAECA\ncAAAAABAEuEAAAAAQIBwAAAAAECS9H8P59En5EnRlAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f026c3316a0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:10:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>59036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>88349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>7460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>97919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>83275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>3287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>2894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>plasmid_7</td>\n",
       "      <td>2652</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0        59036\n",
       "1  plasmid_1        88349\n",
       "2  plasmid_2         7460\n",
       "3  plasmid_3        97919\n",
       "4  plasmid_4        83275\n",
       "5  plasmid_5         3287\n",
       "6  plasmid_6         2894\n",
       "7  plasmid_7         2652"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.375719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.398847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.386938</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.375719\n",
       "1  precision  0.398847\n",
       "2   f1_score  0.386938"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP012925.1</th>\n",
       "      <th>CP012926.1</th>\n",
       "      <th>CP012927.1</th>\n",
       "      <th>CP012928.1</th>\n",
       "      <th>CP012929.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.919310</td>\n",
       "      <td>0.086816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002750</td>\n",
       "      <td>0.008725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.088620</td>\n",
       "      <td>0.081745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.946774</td>\n",
       "      <td>0.104132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.027257</td>\n",
       "      <td>0.040992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.072044</td>\n",
       "      <td>0.003957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.028655</td>\n",
       "      <td>0.020324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>plasmid_7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.039701</td>\n",
       "      <td>0.015847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.125111</td>\n",
       "      <td>0.362539</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP012925.1  CP012926.1  CP012927.1  CP012928.1  CP012929.1\n",
       "0  plasmid_0         0.0         0.0         0.0    0.919310    0.086816\n",
       "1  plasmid_1         0.0         0.0         0.0    0.002750    0.008725\n",
       "2  plasmid_2         0.0         0.0         0.0    0.088620    0.081745\n",
       "3  plasmid_3         0.0         0.0         0.0    0.946774    0.104132\n",
       "4  plasmid_4         0.0         0.0         0.0    0.027257    0.040992\n",
       "5  plasmid_5         0.0         0.0         0.0    0.072044    0.003957\n",
       "6  plasmid_6         0.0         0.0         0.0    0.028655    0.020324\n",
       "7  plasmid_7         0.0         0.0         0.0    0.039701    0.015847\n",
       "8      Total         0.0         0.0         0.0    2.125111    0.362539"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012925.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012926.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012927.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012928.1</td>\n",
       "      <td>0.959008</td>\n",
       "      <td>0.001981</td>\n",
       "      <td>0.690483</td>\n",
       "      <td>0.603203</td>\n",
       "      <td>0.020630</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.953697</td>\n",
       "      <td>0.907240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012929.1</td>\n",
       "      <td>0.131953</td>\n",
       "      <td>0.007504</td>\n",
       "      <td>0.967024</td>\n",
       "      <td>0.100726</td>\n",
       "      <td>0.044899</td>\n",
       "      <td>0.060846</td>\n",
       "      <td>0.942294</td>\n",
       "      <td>0.574284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Total</td>\n",
       "      <td>1.090961</td>\n",
       "      <td>0.009485</td>\n",
       "      <td>1.657507</td>\n",
       "      <td>0.703929</td>\n",
       "      <td>0.065530</td>\n",
       "      <td>1.060846</td>\n",
       "      <td>1.895992</td>\n",
       "      <td>1.481523</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2  plasmid_3  plasmid_4  \\\n",
       "0  CP012925.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "1  CP012926.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "2  CP012927.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "3  CP012928.1   0.959008   0.001981   0.690483   0.603203   0.020630   \n",
       "4  CP012929.1   0.131953   0.007504   0.967024   0.100726   0.044899   \n",
       "5       Total   1.090961   0.009485   1.657507   0.703929   0.065530   \n",
       "\n",
       "   plasmid_5  plasmid_6  plasmid_7  \n",
       "0   0.000000   0.000000   0.000000  \n",
       "1   0.000000   0.000000   0.000000  \n",
       "2   0.000000   0.000000   0.000000  \n",
       "3   1.000000   0.953697   0.907240  \n",
       "4   0.060846   0.942294   0.574284  \n",
       "5   1.060846   1.895992   1.481523  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.959059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.007504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.967024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.620166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.044899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>plasmid_7</td>\n",
       "      <td>0.908371</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.959059\n",
       "1  plasmid_1    0.007504\n",
       "2  plasmid_2    0.967024\n",
       "3  plasmid_3    0.620166\n",
       "4  plasmid_4    0.044899\n",
       "5  plasmid_5    1.000000\n",
       "6  plasmid_6    1.000000\n",
       "7  plasmid_7    0.908371"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012925.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012926.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012927.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012928.1</td>\n",
       "      <td>0.995598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012929.1</td>\n",
       "      <td>0.138481</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP012925.1    0.000000\n",
       "1  CP012926.1    0.000000\n",
       "2  CP012927.1    0.000000\n",
       "3  CP012928.1    0.995598\n",
       "4  CP012929.1    0.138481"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012928.1</td>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.91931</td>\n",
       "      <td>0.959008</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction   recall  precision\n",
       "0  CP012928.1  plasmid_0  0.91931   0.959008"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 10, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.000688194</td>\n",
       "      <td>-0.00252895</td>\n",
       "      <td>0.00441765</td>\n",
       "      <td>-0.000795468</td>\n",
       "      <td>-0.00117698</td>\n",
       "      <td>0.00528784</td>\n",
       "      <td>0.00419741</td>\n",
       "      <td>0.00457638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[27, 46, 26, 36, 13, 48, 30]</td>\n",
       "      <td>[37, 14, 49, 15, 39]</td>\n",
       "      <td>[48, 31, 52, 25, 46, 24]</td>\n",
       "      <td>[46, 24, 49, 84, 48, 31, 52, 23, 36, 13, 85, 4...</td>\n",
       "      <td>[66, 32, 62, 78, 42, 86, 88, 40, 72, 56, 41, 6...</td>\n",
       "      <td>[52, 23]</td>\n",
       "      <td>[49, 84, 48, 30, 31]</td>\n",
       "      <td>[46, 26]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.824366</td>\n",
       "      <td>-0.804928</td>\n",
       "      <td>-0.800756</td>\n",
       "      <td>-0.509084</td>\n",
       "      <td>-0.460907</td>\n",
       "      <td>-0.691706</td>\n",
       "      <td>-0.784696</td>\n",
       "      <td>-0.815508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.825054</td>\n",
       "      <td>-0.807457</td>\n",
       "      <td>-0.796338</td>\n",
       "      <td>-0.509879</td>\n",
       "      <td>-0.462084</td>\n",
       "      <td>-0.686418</td>\n",
       "      <td>-0.780498</td>\n",
       "      <td>-0.810932</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           plasmid_0             plasmid_1  \\\n",
       "GC_sum                  -0.000688194           -0.00252895   \n",
       "chain   [27, 46, 26, 36, 13, 48, 30]  [37, 14, 49, 15, 39]   \n",
       "gd_sum                     -0.824366             -0.804928   \n",
       "total                      -0.825054             -0.807457   \n",
       "\n",
       "                       plasmid_2  \\\n",
       "GC_sum                0.00441765   \n",
       "chain   [48, 31, 52, 25, 46, 24]   \n",
       "gd_sum                 -0.800756   \n",
       "total                  -0.796338   \n",
       "\n",
       "                                                plasmid_3  \\\n",
       "GC_sum                                       -0.000795468   \n",
       "chain   [46, 24, 49, 84, 48, 31, 52, 23, 36, 13, 85, 4...   \n",
       "gd_sum                                          -0.509084   \n",
       "total                                           -0.509879   \n",
       "\n",
       "                                                plasmid_4   plasmid_5  \\\n",
       "GC_sum                                        -0.00117698  0.00528784   \n",
       "chain   [66, 32, 62, 78, 42, 86, 88, 40, 72, 56, 41, 6...    [52, 23]   \n",
       "gd_sum                                          -0.460907   -0.691706   \n",
       "total                                           -0.462084   -0.686418   \n",
       "\n",
       "                   plasmid_6   plasmid_7  \n",
       "GC_sum            0.00419741  0.00457638  \n",
       "chain   [49, 84, 48, 30, 31]    [46, 26]  \n",
       "gd_sum             -0.784696   -0.815508  \n",
       "total              -0.780498   -0.810932  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_11</th>\n",
       "      <th>plasmid_12</th>\n",
       "      <th>plasmid_13</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.000760813</td>\n",
       "      <td>0.00506309</td>\n",
       "      <td>-0.00717595</td>\n",
       "      <td>1.50511e-05</td>\n",
       "      <td>-0.000298372</td>\n",
       "      <td>0.00020777</td>\n",
       "      <td>0.00578519</td>\n",
       "      <td>-0.00985356</td>\n",
       "      <td>-0.00134298</td>\n",
       "      <td>0.00528784</td>\n",
       "      <td>0.00324073</td>\n",
       "      <td>0.00671698</td>\n",
       "      <td>-0.00346476</td>\n",
       "      <td>-0.00556522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[15, 39, 15, 49, 24, 46, 27, 36, 26, 46, 25]</td>\n",
       "      <td>[36, 26, 46, 25, 52, 31, 48, 84, 49, 24, 46, 27]</td>\n",
       "      <td>[39]</td>\n",
       "      <td>[19, 75, 35, 7, 43]</td>\n",
       "      <td>[1, 75, 35, 6, 76]</td>\n",
       "      <td>[4, 75, 35, 3, 58]</td>\n",
       "      <td>[36, 26, 46, 27]</td>\n",
       "      <td>[16]</td>\n",
       "      <td>[13, 48, 30]</td>\n",
       "      <td>[52, 23]</td>\n",
       "      <td>[28]</td>\n",
       "      <td>[48]</td>\n",
       "      <td>[14, 37, 14]</td>\n",
       "      <td>[37]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.828581</td>\n",
       "      <td>-0.821305</td>\n",
       "      <td>-0.835989</td>\n",
       "      <td>-0.00256197</td>\n",
       "      <td>-0.000601923</td>\n",
       "      <td>-0.000464338</td>\n",
       "      <td>-0.848407</td>\n",
       "      <td>-0.747057</td>\n",
       "      <td>-0.821934</td>\n",
       "      <td>-0.691706</td>\n",
       "      <td>-0.429389</td>\n",
       "      <td>-0.909091</td>\n",
       "      <td>-0.782242</td>\n",
       "      <td>-0.909091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.829341</td>\n",
       "      <td>-0.816242</td>\n",
       "      <td>-0.843165</td>\n",
       "      <td>-0.00254692</td>\n",
       "      <td>-0.000900295</td>\n",
       "      <td>-0.000256568</td>\n",
       "      <td>-0.842621</td>\n",
       "      <td>-0.756911</td>\n",
       "      <td>-0.823277</td>\n",
       "      <td>-0.686418</td>\n",
       "      <td>-0.426149</td>\n",
       "      <td>-0.902374</td>\n",
       "      <td>-0.785707</td>\n",
       "      <td>-0.914656</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           plasmid_0  \\\n",
       "GC_sum                                  -0.000760813   \n",
       "chain   [15, 39, 15, 49, 24, 46, 27, 36, 26, 46, 25]   \n",
       "gd_sum                                     -0.828581   \n",
       "total                                      -0.829341   \n",
       "\n",
       "                                               plasmid_1  plasmid_10  \\\n",
       "GC_sum                                        0.00506309 -0.00717595   \n",
       "chain   [36, 26, 46, 25, 52, 31, 48, 84, 49, 24, 46, 27]        [39]   \n",
       "gd_sum                                         -0.821305   -0.835989   \n",
       "total                                          -0.816242   -0.843165   \n",
       "\n",
       "                 plasmid_11          plasmid_12          plasmid_13  \\\n",
       "GC_sum          1.50511e-05        -0.000298372          0.00020777   \n",
       "chain   [19, 75, 35, 7, 43]  [1, 75, 35, 6, 76]  [4, 75, 35, 3, 58]   \n",
       "gd_sum          -0.00256197        -0.000601923        -0.000464338   \n",
       "total           -0.00254692        -0.000900295        -0.000256568   \n",
       "\n",
       "               plasmid_2   plasmid_3     plasmid_4   plasmid_5   plasmid_6  \\\n",
       "GC_sum        0.00578519 -0.00985356   -0.00134298  0.00528784  0.00324073   \n",
       "chain   [36, 26, 46, 27]        [16]  [13, 48, 30]    [52, 23]        [28]   \n",
       "gd_sum         -0.848407   -0.747057     -0.821934   -0.691706   -0.429389   \n",
       "total          -0.842621   -0.756911     -0.823277   -0.686418   -0.426149   \n",
       "\n",
       "         plasmid_7     plasmid_8   plasmid_9  \n",
       "GC_sum  0.00671698   -0.00346476 -0.00556522  \n",
       "chain         [48]  [14, 37, 14]        [37]  \n",
       "gd_sum   -0.909091     -0.782242   -0.909091  \n",
       "total    -0.902374     -0.785707   -0.914656  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_7  from MILP completely contained in  plasmid_0  from greedy.\n",
      "\n",
      "plasmid_2  from MILP completely contained in  plasmid_1  from greedy.\n",
      "\n",
      "plasmid_7  from MILP completely contained in  plasmid_1  from greedy.\n",
      "\n",
      "plasmid_2  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_7  from MILP completely contained in  plasmid_2  from greedy.\n",
      "\n",
      "plasmid_4  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_5  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_5  from MILP completely contained in  plasmid_5  from greedy.\n",
      "\n",
      "plasmid_5  from greedy completely contained in  plasmid_5  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_6  from MILP.\n",
      "\n",
      "plasmid_8  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_9  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_10  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_10  from greedy completely contained in  plasmid_4  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
