{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 117 with $\\alpha_1 = 1$, $\\alpha_2 = 10$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 117\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012934.1</td>\n",
       "      <td>3372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012932.1</td>\n",
       "      <td>2096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012933.1</td>\n",
       "      <td>3319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012935.1</td>\n",
       "      <td>3905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012936.1</td>\n",
       "      <td>98999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CP012931.1</td>\n",
       "      <td>236176</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP012934.1         3372\n",
       "1  CP012932.1         2096\n",
       "2  CP012933.1         3319\n",
       "3  CP012935.1         3905\n",
       "4  CP012936.1        98999\n",
       "5  CP012931.1       236176"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt8FeWdx/HvLwkEAogCIVwlUQkQ\nAgESQAFLUNTQCtjFCyh2VSiulvWGVXa3VWu1VldqS1daoOtaK1UBldKKBeqSeqsVUJBrICo3EQGB\nYEAgl9/+kQkbQxIONZMT4PN+vXx5ZuY5z/zO8TzmfM88M2PuLgAAAACIiXYBAAAAAOoHwgEAAAAA\nSYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAIG4aBdwolq1\nauXJycnRLuOkceDAATVp0iTaZQD1HmMFiAxj5cQsX758t7snRrsOIFInXThITk7WsmXLol3GSSM3\nN1fZ2dnRLgOo9xgrQGQYKyfGzDZHuwbgRDCtCAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAA\nAACAJMIBAAAAgADhAAAAAIAkwgEAAACAQGjhwMyeMrOdZra6mu1mZlPNLN/MPjCzPmHVAgAAAOD4\nwjxy8LSknBq2D5PUOfhngqRfhVgLAAAAgOMILRy4++uS9tTQZKSkZ7zMO5LONLO2YdUDAAAAoGZx\nUdx3e0lbKyxvC9Z9WrmhmU1Q2dEFJSUlKTc3ty7qOyUUFhbyfiE0qz4piHYJtSaleSxjBYgAf1eA\nU1s0w4FVsc6raujuMyTNkKSsrCzPzs4OsaxTS25urni/EJYbJr8S7RJqzdM5TRgrCE3yKTVWmjJW\ngFNYNMPBNkkdKyx3kLQ9SrV8xan1P/Em0S4BAAAAJ4loXsp0vqTvBFctOl9SgbsfM6UIAAAAQN0I\n7ciBmT0nKVtSKzPbJul+SQ0kyd1/LWmBpG9Kypd0UNKNYdUCAAAA4PhCCwfuPuY4213S98LaPwAA\nAIATwx2SAQAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAA\nAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAA\nAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAA\nCBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAI\nEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQ\nDgAAAABIIhwAAAAACIQaDswsx8zyzCzfzCZXsf1sM1tiZu+b2Qdm9s0w6wEAAABQvdDCgZnFSnpS\n0jBJaZLGmFlapWY/kDTb3XtLGi1pWlj1AAAAAKhZmEcO+knKd/eP3P2IpOcljazUxiWdETxuLml7\niPUAAAAAqEFciH23l7S1wvI2Sf0rtXlA0iIz+1dJTSQNDbEeAAAAADUIMxxYFeu80vIYSU+7+xQz\nu0DS78ws3d1Lv9KR2QRJEyQpKSlJubm5YdR71KQexaH2X5cKCwtDf79w+mKsAJFhrAA4WYQZDrZJ\n6lhhuYOOnTY0TlKOJLn738yskaRWknZWbOTuMyTNkKSsrCzPzs4OqeQyN0x+JdT+69LTOU0U9vuF\n0xdjBYgMYwXAySLMcw6WSupsZilm1lBlJxzPr9Rmi6SLJcnMuklqJGlXiDUBAAAAqEZo4cDdiyVN\nlLRQ0jqVXZVojZk9aGYjgmaTJH3XzFZKek7SDe5eeeoRAAAAgDoQ5rQiufsCSQsqrbuvwuO1kgaG\nWQMAAACAyHCHZAAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACS\nCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJII\nBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggH\nAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcA\nAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAA\nAAAChAMAAAAAkggHAAAAAAKEAwAAAACSQg4HZpZjZnlmlm9mk6tpc7WZrTWzNWb2+zDrAQAAAFC9\nuLA6NrNYSU9KukTSNklLzWy+u6+t0KazpH+TNNDd95pZ67DqAQAAAFCzMI8c9JOU7+4fufsRSc9L\nGlmpzXclPenueyXJ3XeGWA8AAACAGoR25EBSe0lbKyxvk9S/UptUSTKztyTFSnrA3f9cuSMzmyBp\ngiQlJSUpNzc3jHqPmtSjONT+61JhYWHo7xdOX4wVIDKMFQAnizDDgVWxzqvYf2dJ2ZI6SHrDzNLd\nfd9XnuQ+Q9IMScrKyvLs7OxaL7aiGya/Emr/denpnCYK+/3C6YuxAkSGsQLgZBHmtKJtkjpWWO4g\naXsVbf7g7kXu/rGkPJWFBQAAAAB1LMxwsFRSZzNLMbOGkkZLml+pzTxJQyTJzFqpbJrRRyHWBAAA\nAKAaoYUDdy+WNFHSQknrJM129zVm9qCZjQiaLZT0uZmtlbRE0vfd/fOwagIAAABQvTDPOZC7L5C0\noNK6+yo8dkl3Bf8AAAAAiCLukAwAAABAEuEAAAAAQIBwAAAAAEDSCYQDMxtkZjcGjxPNLCW8sgAA\nAADUtYjCgZndL+leSf8WrGog6dmwigIAAABQ9yK9WtG3JfWW9J4kuft2M2sWWlWoPZ+ukB4YGe0q\nascDBdGuAAAA4JQW6bSiI8FlR12SzKxJeCUBAAAAiIZIw8FsM5su6Uwz+66kv0iaGV5ZAAAAAOpa\nRNOK3P1xM7tE0n5JXSTd5+6LQ60MAAAAQJ06bjgws1hJC919qCQCAQAAAHCKOu60IncvkXTQzJrX\nQT0AAAAAoiTSqxUdkrTKzBZLOlC+0t1vC6UqAAAAAHUu0nDwSvAPAAAAgFNUpCck/9bMGkpKDVbl\nuXtReGUBAAAAqGsRhQMzy5b0W0mbJJmkjmb2z+7+enilAQAAAKhLkU4rmiLpUnfPkyQzS5X0nKTM\nsAoDAAD10KcrpAdGRruK2vFAQbQrAOqdSG+C1qA8GEiSu2+Q1CCckgAAAABEQ6RHDpaZ2X9L+l2w\nfJ2k5eGUBAAAACAaIg0Ht0j6nqTbVHbOweuSpoVVFAAAAIC6F2k4iJP0C3f/mXT0rsnxoVUFAABw\nilq+fHnruLi430hKV+RTvIHaUCppdXFx8fjMzMydVTWINBy8JmmopMJgubGkRZIGfO0SAQAATiNx\ncXG/adOmTbfExMS9MTExHu16cPooLS21Xbt2pe3YseM3kkZU1SbStNrI3cuDgYLHCbVQIwAAwOkm\nPTExcT/BAHUtJibGExMTC1R21KrqNhH2dcDM+pQvmFmWpC+/Zn0AAACnoxiCAaIl+OxVmwEinVZ0\nu6Q5ZrZdkktqJ+mar18eANQTXLsdAICIw0GKpN6Szpb0bUnnqywkAAAA4GtInvxKrd5UdtNPv1Vv\nLjffr1+/Lo8//vjWb3zjGwcjaV9aWqoBAwakLliwIL9FixalYdf3df3pT39qNmXKlKQlS5bkh9H/\nrFmzmq9Zs6bxT37ykx2VtyUkJPQ+ePDg+9U9d8CAAal/+MMfPkxMTCw5kX1GOq3oh+6+X9KZki6R\nNEPSr05kRwAAADh5FBUV1fk+Z8+e3bx79+5fngzBoC5cd911BVUFg0iMGTPm88cffzzxRJ8XaTgo\nTxzfkvRrd/+DpIYnujMAAABE3/e///22KSkp3QcMGNB5+PDhKffdd1+SVPZL/8SJE9v37du3y0MP\nPZS0ffv2uMsuu+zc9PT0bunp6d0WLVrURJL2798fc9VVVyWnp6d369atW9qzzz57piQVFhba5Zdf\nfk5qamrat771rXMOHTpkkvTEE0+0GjduXMfy/U+ZMqXV+PHjO1Sua9asWS2+/e1v75OkvLy8hikp\nKd2vueaaTp07d+4+YsSIlHnz5jXr06dP106dOqUvWbIkoaZa8vLyGmZmZnZJS0vrlpaW1m3x4sVN\npLJf+/v169clJyfnnJSUlO4jRoxIKS2tOYuMGjUq+dprrz07MzOzS3Jycvpzzz3XvHKbJUuWJPTu\n3btrt27d0nr37t115cqV8ZK0bNmyRj169OjWtWvXtNTU1LRVq1bFR/rapk6d2vI73/nO2ZK0fv36\nhr169eqanp7e7fbbb29Xvt/Nmzc3yMrK6tK1a9e0zp07d//zn//cVJJGjx6976WXXmoZ4UfiqEin\nFX1iZtNVdjnTR80sXlyXFwAA4KTz+uuvJ/zxj388a9WqVWuLioqsV69eab179z467Wffvn2xS5cu\nzZOk4cOHp9x1112fXXbZZYUbN25seNlll3X+6KOP1vz7v/972yFDhuyfM2fOpt27d8dmZWV1GzFi\nxP6f/exniY0bNy7dsGHD2r///e+NBw4cmCZJ48aN29O9e/e0w4cPb4uPj/dnn3221fTp0zdXrm35\n8uVNBw4ceHT91q1bG73wwgsfZWZmbu7Zs2e3WbNmtVy2bNn63//+92c+/PDDbYcMGfJhdbW0a9eu\n+I033tiQkJDgq1atih8zZsw5q1evXidJ69ata7xixYqPkpOTizIzM7suXry46WWXXVZYuZ6Ktm7d\nGv/uu+/mrV27Nn7o0KFdRo4cuari9oyMjEPvvvvu+gYNGmjevHnN7rnnng4LFy788Je//GXirbfe\n+tktt9yy59ChQ1ZcXKxPPvmkQSSvrWL/t95669njx4/fNXHixM8feeSRo0cEnnrqqRYXX3xxwaOP\nPrqjuLhYX3zxRYwkJSYmlhw5csR27NgR26ZNm4inFkUaDq6WlCPpcXffZ2ZtJX0/0p0AAACgfsjN\nzW06bNiwfU2bNnVJfskll+yruH3MmDF7yh+/9dZbZ2zcuLFx+XJhYWHs3r17Y3Jzc89YuHDhmVOn\nTm0jSYcPH7b8/PyGb775ZtPbbrttpyT179//y9TU1IOSdMYZZ5QOHDjwixdeeKF5jx49DhUVFVm/\nfv2OufJlQUFB3FlnnXX0Z/z27dsfLm+Xmpr65UUXXbQ/JiZGffr0OfjQQw+1C15PlbV06tSpaNy4\ncZ3Wrl3bOCYmRps3bz56A98ePXocOPfcc4skqXv37gc//PDD486IGTVq1J7Y2Fj16NHjcMeOHQ+v\nWLGiUcXte/bsib3mmmtSNm3a1MjMvKioyCTpggsuOPD444+33bZtW8PRo0fv7dGjx+FIX1tF7733\nXtNXX331Q0m6+eabP//xj3/cQZLOP//8AzfffHNyUVFRzJVXXrl3wIABR9/Xli1bFm/ZsqVhmzZt\nIr7KaEThwN0PSnqpwvKnkj6NdCcAAACoH9xrvqZMs2bNjn45d3ctW7ZsXRAkvtLH3Llz8zMyMg5X\nfr6ZVdnvhAkTdj/88MNtUlNTD40dO3Z3VW1iY2O9pKREsbGxkqSGDRse3W9MTIwaNWrkQTuVlJRY\nTbXcdddd7Vq3bl304osvflxaWqrGjRsfPfE7Pj7+aL+xsbEqLi6uuugaXlfl5Xvvvbf94MGDv1i8\nePGHeXl5DS+66KIukvQv//Ivey688MIDL7/8cvNhw4alTps2bVOXLl0OR/LaKqvqErjDhg0rfP31\n1/NefPHF5jfccEPKbbfd9tnEiRM/l8qCUkJCwgmdv8HUIAAAgNNIdnZ24cKFC5sfPHjQCgoKYv7y\nl7+cWV3bQYMG7X/00Udbly+//fbbjSVpyJAh+6dMmZJUPlf/rbfeahy0L3z22WdbSNLSpUsbbdiw\n4ehNcy+66KIDn376acOXX3655bhx4/aoCikpKYfWrVsXX9W26lRXS0FBQWzbtm2LYmNjNW3atJYl\nJcefWfO9732v/TPPPFPl+/HSSy+dVVJSojVr1sRv3bo1PiMj41DF7fv374/t0KHDEUmaPn16q/L1\na9eubditW7fDP/jBD3Zeeuml+1asWNG4ct+R6NOnT+HMmTNbSNLMmTOPnkuwYcOGhu3bty+aNGnS\n7rFjx+5+7733EqSyKz/t2rWrQZcuXY4JcDWJdFoRAAAAQlDXlx4dPHjwwZycnIK0tLTu7du3P9yz\nZ88DzZs3r/Kb84wZM7aOHz/+7NTU1LSSkhLr37//FwMGDNjy05/+dPuECRPO7tq1a5q7W4cOHQ4v\nWbIk/+677945evTolNTU1LTu3bsf7NGjx4GK/V1xxRV7P/jgg4TqLq956aWXFixatKhZenp6xF9o\nq6vljjvu2Dlq1Khz582bd9agQYO+aNy48XF/QV+7dm3j8hOiKzvvvPMO9+vXr8vnn3/e4Oc///nm\nhISEr/yKf++99+4YP358ytSpU9tceOGF+8vX/+53v2sxZ86clnFxcZ6YmFj0yCOPbN+3b19spK+v\n3LRp07aMHj36nGnTpiWNGDFib/n6hQsXNps6dWqbuLg4T0hIKJk1a9bHkvTmm28m9O7d+0CDBg1O\naD92vENL9U1WVpYvW7Ys1H0kT34l1P7r0tMZG5Wdd3+0y6gd3Nip3mGs1FOMlXqHsVJP1cFYMbPl\n7p5Vcd3KlSs3ZWRkVDmtpq4UFBTENG/evPSLL76IueCCC7r8+te/3jxo0KCI7kXwdQwZMuS8O+64\n47ORI0d+UdX2zZs3NxgzZkzy22+/vTHsWqoyaNCgzm+++eYx+x41alTy5ZdfXnDjjTfurep59dGN\nN97Y8YorrthX1Xu9cuXKVhkZGclVPY9pRQAAAKeZsWPHduratWtaz549uw0fPnxv2MFg9+7dscnJ\nyemNGjUqrS4YSFKnTp2Kbrrppt179uyJynfUqoLBySo9Pf3Lmt7r6jCtCAAA4DTzxz/+8eO63F+r\nVq1KNm3atDqStuPHj693v86/+OKLm6Jdw4maNGnSP3R0iiMHAAAAACQRDgAAAAAECAcAAAAAJBEO\nAAAAAAQ4IRkAACCaHmieefxGJ9JfwXHvm2BmmSNHjtwzb968jyWpqKhIrVu3zujVq9eBJUuW5E+d\nOrXlsmXLmjzzzDNb7rrrrnZNmzYtefDBBz+r2EdsbGxm586dvywpKbHzzjvvy9mzZ2+qeHdlqexG\nXAMGDEhdsGBBfosWLUqvuuqq5Ndee615y5Ytizdu3Limqtref//9RjfeeGPy2rVrEyZPnvxJxf3O\nnTv3jLvvvvvs0tJSjR07dvdPfvKTHZWf/+qrrzadNGlSxw0bNiTMnDnzo/LLj27fvj3ummuuSXnj\njTdOmSsShYEjBwAAAKeZxo0bl+bl5TUuLCw0SXr55ZfPSEpKKjqRPuLj40vXr1+/duPGjWsaNGjg\nU6ZMSazcZvbs2c27d+/+ZYsWLUol6aabbto9f/78Gr+ct27duvgXv/jFlptvvvkrYaS4uFh33nnn\n2QsWLNiwYcOGNS+++GKL5cuXN6r8/HPOOefI//zP/2waPnz45xXXt2vXrjgpKalo0aJFTU7kdZ5u\nCAcAAACnoYsvvrhgzpw5Z0rSc88912LUqFF7/tG+Bg0aVJifnx9fef2sWbNaVLzj8LBhwwoTExOL\na+qrffv2xYMHDz7YoEGDr9ypNzc3t0mnTp0Op6WlHWnUqJH/0z/90565c+eeWfn5Xbp0OdK/f/8v\nY2KO/Zp7xRVX7HvmmWdantCLO80QDgAAAE5D119//Z4XXnjhrIMHD9q6desSLrjgggP/SD9FRUVa\nuHDhGT169Piy8rbly5c3HThw4HH7feyxxxIfe+yxY448VLR169aG7du3P1K+3KFDhyOffPJJQ0m6\n44472s2aNav58fYzcODAA++++27T47U7nYV6zoGZ5Uj6haRYSb9x959W0+5KSXMk9XX3ZWHWBAAA\nAKl///5fbtu2LX7mzJkthg4dWnCizz98+HBM165d04K+vrj99tuPuelWQUFB3FlnnVV67LO/6p57\n7tl1vDbufsw6M3NJ+vnPf749kprbtWtXvHPnzoaRtD1dhRYOzCxW0pOSLpG0TdJSM5vv7msrtWsm\n6TZJfw+rFgAAABwrJydn3/33399x0aJFeTt37jyh74Xl5xzU1CY2NtZLSkoUGxv79QqVdPbZZx89\nUiBJ27Zta9iuXbsTOk/i4MGDFh8ff9ywcjoLc1pRP0n57v6Rux+R9LykkVW0+7GkxyQdCrEWAAAA\nVHLLLbfsnjRp0vZ+/fodMyWoNqSkpBxat27dMeci/CMGDx58YNOmTY3Wr1/f8NChQ/bSSy+1GDVq\n1L7jP/P/rV69ulFqamoor/VUEea0ovaStlZY3iapf8UGZtZbUkd3/5OZ3V1dR2Y2QdIESUpKSlJu\nbm7tV1vBpB41nidzUimMb6fcLj+Kdhm1I+T/7jhxjJV6irFS7zBW6qn6MlYiuPRoWM4999yiH/7w\nhzuP1+6JJ55oO3369KTy5c8+++yDSPq/9NJLCxYtWtQsPT39sCQNHz485Z133mm2d+/euKSkpJ6T\nJ0/efuedd+4uP9/gnnvu2bVly5a4vn37ph04cCDWzHz69OlJ69atW92iRYvSKVOmbMnJyUktKSnR\ntddeuzsrK+uQVHbOQd++fQ9cd911BX/9618Trr766vP2798f+9prr5358MMPt8vPz18jSYsXL26W\nk5NzwlOoTidW1fytWunY7CpJl7n7+GD5ekn93P1fg+UYSf8r6QZ332RmuZLuPt45B1lZWb5sWbin\nJSRPfiXU/uvS0xkblZ13f7TLqB0PMJbrG8ZKPcVYqXcYK/VUHYwVM1vu7lkV161cuXJTRkbGMfPz\nT0WbN29uMGbMmOS33367XtxbICsrq8urr76an5iYWBLtWqJp5cqVrTIyMpKr2hbmtKJtkjpWWO4g\nqeLJIs0kpUvKNbNNks6XNN/MvjKAAAAAcHLq1KlT0U033bR7z549Ub9C5vbt2+Nuv/32z073YHA8\nYU4rWiqps5mlSPpE0mhJ15ZvdPcCSa3KlyM9cgAAAICTx/jx4/dGuwap7EpF119//Qmdo3A6Ci3F\nuXuxpImSFkpaJ2m2u68xswfNbERY+wUAAKjnSktLSy3aReD0FHz2qr1iU6j3OXD3BZIWVFp3XzVt\ns8OsBQAAoJ5YvWvXrrTExMSCmJiYcE7+BKpQWlpqu3btai5pdXVtQg0HAAAA+Kri4uLxO3bs+M2O\nHTvSFe75n0BlpZJWFxcXj6+uAeEAAACgDmVmZu6UxBRr1EukVQAAAACSCAcAAAAAAoQDAAAAAJII\nBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggH\nAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcA\nAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAA\nAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAA\nAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkkIOB2aWY2Z5ZpZvZpOr2H6Xma01\nsw/M7DUz6xRmPQAAAACqF1o4MLNYSU9KGiYpTdIYM0ur1Ox9SVnu3lPSXEmPhVUPAAAAgJqFeeSg\nn6R8d//I3Y9Iel7SyIoN3H2Jux8MFt+R1CHEegAAAADUwNw9nI7NrpSU4+7jg+XrJfV394nVtP8v\nSTvc/aEqtk2QNEGSkpKSMp9//vlQai636pOCUPuvSymND6vp4e3RLqN2tO0V7QpQCWOlnmKs1DuM\nlXqqDsbKkCFDlrt7Vug7AmpJXIh9WxXrqkwiZjZWUpakwVVtd/cZkmZIUlZWlmdnZ9dSiVW7YfIr\nofZfl57O+FjZefdHu4zaMebU+eN6qmCs1FOMlXqHsVJPMVaAY4QZDrZJ6lhhuYOkY35qMLOhkv5D\n0mB3PxxiPQAAAABqEOY5B0sldTazFDNrKGm0pPkVG5hZb0nTJY1w950h1gIAAADgOEILB+5eLGmi\npIWS1kma7e5rzOxBMxsRNPtPSU0lzTGzFWY2v5ruAAAAAIQszGlFcvcFkhZUWndfhcdDw9w/AAAA\ngMhxh2QAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAA\nAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAA\nAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAC\nhAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKE\nAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQD\nAAAAAJIIBwAAAAACoYYDM8sxszwzyzezyVVsjzezF4Ltfzez5DDrAQAAAFC90MKBmcVKelLSMElp\nksaYWVqlZuMk7XX38yQ9IenRsOoBAAAAULMwjxz0k5Tv7h+5+xFJz0saWanNSEm/DR7PlXSxmVmI\nNQEAAACoRpjhoL2krRWWtwXrqmzj7sWSCiS1DLEmAAAAANWIC7Hvqo4A+D/QRmY2QdKEYLHQzPK+\nZm2njSFSK0m7o11HrfgRB5UQHsYKEBnGygnrVBc7AWpLmOFgm6SOFZY7SNpeTZttZhYnqbmkPZU7\ncvcZkmaEVOcpzcyWuXtWtOssmpVqAAAFP0lEQVQA6jvGChAZxgpwagtzWtFSSZ3NLMXMGkoaLWl+\npTbzJf1z8PhKSf/r7sccOQAAAAAQvtCOHLh7sZlNlLRQUqykp9x9jZk9KGmZu8+X9N+Sfmdm+So7\nYjA6rHoAAAAA1Mz4of7UZmYTgmlZAGrAWAEiw1gBTm2EAwAAAACSQr5DMgAAAICTB+EAAAAAgCTC\nwWnPzJLNbHXwONvM/hTtmoBoMLO3j7N9gZmdWVf1AKcCM8s1s6zg8SYzaxXtmgDULMz7HCBEZmYq\nO2ekNNq1APWNmcW6e8mJPMfdBxxn+ze/XlXAqYe/RcCphyMHJ5HgV/51ZjZN0nuSrjezv5nZe2Y2\nx8yaBu36mtnbZrbSzN41s2bBc98I2r5nZjV+EQLqq+CzvN7MfmtmH5jZXDNLCH6VvM/M3pR0lZmd\na2Z/NrPlwWe/a/D8JDN7ORgfK8vHgpkVBv9ua2avm9kKM1ttZhcG64/+6mlmdwXbVpvZHRXqWmdm\nM81sjZktMrPGUXmTgBpU/vya2aNmdmuF7Q+Y2aTg8ffNbGkw1n4UrKv8t6ijmf3KzJYFn/0fReeV\nAagNhIOTTxdJz0i6RNI4SUPdvY+kZZLuCm4494Kk2909Q9JQSV9K2inpkqDtNZKmRqN4oJZ0kTTD\n3XtK2i+p/IvNIXcf5O7Pq+yu6v/q7pmS7pY0LWgzVdJfg/HRR9KaSn1fK2mhu/eSlCFpRcWNZpYp\n6UZJ/SWdL+m7ZtY72NxZ0pPu3l3SPkmjausFA7Whqs+vpOdV9neh3NWS5pjZpSr7TPeT1EtSppl9\nI2jTRdIz7t7b3TdL+o/grsk9JQ02s5518oIA1DqmFZ18Nrv7O2Z2uaQ0SW+VHdVVQ0l/U9n/sD91\n96WS5O77JcnMmkj6LzPrJalEUmo0igdqyVZ3fyt4/Kyk24LHL0hScBRtgMq+4JQ/Jz7490WSviNJ\nwdSjgkp9L5X0lJk1kDTP3VdU2j5I0svufiDY10uSLlTZHd8/rtB+uaTkr/EagTBU9/ltbWbtJCVK\n2uvuW8zsNkmXSno/eG5TlYWFLQr+FlXo92ozm6Cy7xVtVfb36YO6eEEAahfh4ORzIPi3SVrs7mMq\nbgx+ranq5hV3SvpMZb+Exkg6FGaRQMgqf8bLl8vHR4ykfcGv/yfWsfvrwa+j31LZHdz/092fqdDE\nqnmqJB2u8LhEEtOKUN9U9/mdK+lKSW1UdiShvO0j7j79Kx2YJev/x5rMLEVlR+f6uvteM3taUqNa\nrRpAnWFa0cnrHUkDzew8SQrmXKdKWi+pnZn1DdY3M7M4Sc1VdkShVNL1kmKjVDdQG842swuCx2Mk\nvVlxY3DE7GMzu0oqO2nSzDKCza9JuiVYH2tmZ1R8rpl1krTT3WdK+m+VTT2q6HVJVwRjromkb0t6\no/ZeGhCq6j6/z0sarbKAMDdou1DSTRXOZ2tvZq2r6PMMlYWFAjNLkjQs5NcAIESEg5OUu++SdIOk\n58zsA5WFha7ufkRlc0d/aWYrJS1W2S840yT9s5m9o7IpRQeq7Bg4OaxT2ef5A0ktJP2qijbXSRoX\njIM1kkYG62+XNMTMVqls6k/3Ss/LlrTCzN5X2TkDv6i40d3fk/S0pHcl/V3Sb9z9fQEngeo+v+6+\nRlIzSZ+4+6dB20WSfi/pb8F4mRu0qdznSpVNPVoj6SlJb1VuA+DkYe5VzUABgPopmNLwJ3dPj3Ip\nAACccjhyAAAAAEASRw4AAAAABDhyAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAkvR/\nTIXZ7zgi4zUAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fcaf3cb1c88>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:10:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>194757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>92126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>70243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>168157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>5752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>13397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>64580</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0       194757\n",
       "1  plasmid_1        92126\n",
       "2  plasmid_2        70243\n",
       "3  plasmid_3       168157\n",
       "4  plasmid_4         5752\n",
       "5  plasmid_5        13397\n",
       "6  plasmid_6        64580"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.559688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.488846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.521874</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.559688\n",
       "1  precision  0.488846\n",
       "2   f1_score  0.521874"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP012931.1</th>\n",
       "      <th>CP012932.1</th>\n",
       "      <th>CP012933.1</th>\n",
       "      <th>CP012934.1</th>\n",
       "      <th>CP012935.1</th>\n",
       "      <th>CP012936.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.245152</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.010416</td>\n",
       "      <td>0.040076</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.920595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.012033</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.252341</td>\n",
       "      <td>0.040076</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.530844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.022890</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>0.001617</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.085758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>0.121608</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.666058</td>\n",
       "      <td>0.080153</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.537197</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP012931.1  CP012932.1  CP012933.1  CP012934.1  CP012935.1  \\\n",
       "0  plasmid_0    0.245152    0.000000         0.0         0.0         0.0   \n",
       "1  plasmid_1    0.010416    0.040076         0.0         0.0         0.0   \n",
       "2  plasmid_2    0.012033    0.000000         0.0         0.0         0.0   \n",
       "3  plasmid_3    0.252341    0.040076         0.0         0.0         0.0   \n",
       "4  plasmid_4    0.022890    0.000000         0.0         0.0         0.0   \n",
       "5  plasmid_5    0.001617    0.000000         0.0         0.0         0.0   \n",
       "6  plasmid_6    0.121608    0.000000         0.0         0.0         0.0   \n",
       "7      Total    0.666058    0.080153         0.0         0.0         0.0   \n",
       "\n",
       "   CP012936.1  \n",
       "0    0.000000  \n",
       "1    0.920595  \n",
       "2    0.000000  \n",
       "3    0.530844  \n",
       "4    0.000000  \n",
       "5    0.085758  \n",
       "6    0.000000  \n",
       "7    1.537197  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012931.1</td>\n",
       "      <td>0.280611</td>\n",
       "      <td>0.008901</td>\n",
       "      <td>0.014393</td>\n",
       "      <td>0.345392</td>\n",
       "      <td>0.880042</td>\n",
       "      <td>0.014257</td>\n",
       "      <td>0.427872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012932.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000912</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000500</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012933.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012934.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012935.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CP012936.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.980016</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.311227</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.570725</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.280611</td>\n",
       "      <td>0.989829</td>\n",
       "      <td>0.014393</td>\n",
       "      <td>0.657118</td>\n",
       "      <td>0.880042</td>\n",
       "      <td>0.584982</td>\n",
       "      <td>0.427872</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2  plasmid_3  plasmid_4  \\\n",
       "0  CP012931.1   0.280611   0.008901   0.014393   0.345392   0.880042   \n",
       "1  CP012932.1   0.000000   0.000912   0.000000   0.000500   0.000000   \n",
       "2  CP012933.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "3  CP012934.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "4  CP012935.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "5  CP012936.1   0.000000   0.980016   0.000000   0.311227   0.000000   \n",
       "6       Total   0.280611   0.989829   0.014393   0.657118   0.880042   \n",
       "\n",
       "   plasmid_5  plasmid_6  \n",
       "0   0.014257   0.427872  \n",
       "1   0.000000   0.000000  \n",
       "2   0.000000   0.000000  \n",
       "3   0.000000   0.000000  \n",
       "4   0.000000   0.000000  \n",
       "5   0.570725   0.000000  \n",
       "6   0.584982   0.427872  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.280611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.988917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.014393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.656619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.880042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>0.584982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>0.427872</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.280611\n",
       "1  plasmid_1    0.988917\n",
       "2  plasmid_2    0.014393\n",
       "3  plasmid_3    0.656619\n",
       "4  plasmid_4    0.880042\n",
       "5  plasmid_5    0.584982\n",
       "6  plasmid_6    0.427872"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012931.1</td>\n",
       "      <td>0.407222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012932.1</td>\n",
       "      <td>0.040076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012933.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012934.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012935.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CP012936.1</td>\n",
       "      <td>0.994323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP012931.1    0.407222\n",
       "1  CP012932.1    0.040076\n",
       "2  CP012933.1    0.000000\n",
       "3  CP012934.1    0.000000\n",
       "4  CP012935.1    0.000000\n",
       "5  CP012936.1    0.994323"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012936.1</td>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.920595</td>\n",
       "      <td>0.980016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction    recall  precision\n",
       "0  CP012936.1  plasmid_1  0.920595   0.980016"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 10, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00615695</td>\n",
       "      <td>-0.00230566</td>\n",
       "      <td>-0.000499113</td>\n",
       "      <td>-0.00196103</td>\n",
       "      <td>0.00805273</td>\n",
       "      <td>0.000914784</td>\n",
       "      <td>-0.00028907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[43, 11, 42, 35, 15]</td>\n",
       "      <td>[49, 17, 62, 16, 47, 43]</td>\n",
       "      <td>[48, 92, 108, 104, 34, 106, 100, 27, 40, 19, 7...</td>\n",
       "      <td>[49, 17, 62, 23, 71, 19, 40, 27, 100, 106, 34,...</td>\n",
       "      <td>[26, 53, 44]</td>\n",
       "      <td>[62, 23, 58, 109, 35, 28]</td>\n",
       "      <td>[71, 19, 40, 27, 100, 106, 34, 86, 72, 107, 48...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.742865</td>\n",
       "      <td>-0.808754</td>\n",
       "      <td>-0.383676</td>\n",
       "      <td>-0.612209</td>\n",
       "      <td>-0.840024</td>\n",
       "      <td>-0.801672</td>\n",
       "      <td>-0.355148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.749022</td>\n",
       "      <td>-0.81106</td>\n",
       "      <td>-0.384175</td>\n",
       "      <td>-0.61417</td>\n",
       "      <td>-0.831971</td>\n",
       "      <td>-0.800757</td>\n",
       "      <td>-0.355437</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   plasmid_0                 plasmid_1  \\\n",
       "GC_sum           -0.00615695               -0.00230566   \n",
       "chain   [43, 11, 42, 35, 15]  [49, 17, 62, 16, 47, 43]   \n",
       "gd_sum             -0.742865                 -0.808754   \n",
       "total              -0.749022                  -0.81106   \n",
       "\n",
       "                                                plasmid_2  \\\n",
       "GC_sum                                       -0.000499113   \n",
       "chain   [48, 92, 108, 104, 34, 106, 100, 27, 40, 19, 7...   \n",
       "gd_sum                                          -0.383676   \n",
       "total                                           -0.384175   \n",
       "\n",
       "                                                plasmid_3     plasmid_4  \\\n",
       "GC_sum                                        -0.00196103    0.00805273   \n",
       "chain   [49, 17, 62, 23, 71, 19, 40, 27, 100, 106, 34,...  [26, 53, 44]   \n",
       "gd_sum                                          -0.612209     -0.840024   \n",
       "total                                            -0.61417     -0.831971   \n",
       "\n",
       "                        plasmid_5  \\\n",
       "GC_sum                0.000914784   \n",
       "chain   [62, 23, 58, 109, 35, 28]   \n",
       "gd_sum                  -0.801672   \n",
       "total                   -0.800757   \n",
       "\n",
       "                                                plasmid_6  \n",
       "GC_sum                                        -0.00028907  \n",
       "chain   [71, 19, 40, 27, 100, 106, 34, 86, 72, 107, 48...  \n",
       "gd_sum                                          -0.355148  \n",
       "total                                           -0.355437  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_11</th>\n",
       "      <th>plasmid_12</th>\n",
       "      <th>plasmid_13</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0062533</td>\n",
       "      <td>-0.00259313</td>\n",
       "      <td>-5.94052e-05</td>\n",
       "      <td>-0.000119691</td>\n",
       "      <td>0.000544323</td>\n",
       "      <td>0.000550089</td>\n",
       "      <td>-0.000836153</td>\n",
       "      <td>-0.00267233</td>\n",
       "      <td>-0.00770681</td>\n",
       "      <td>-0.000701078</td>\n",
       "      <td>0.00336402</td>\n",
       "      <td>-0.00539855</td>\n",
       "      <td>-0.00456856</td>\n",
       "      <td>-0.00700929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[53, 44, 43, 11, 42, 58, 24, 58]</td>\n",
       "      <td>[35, 42, 15, 35, 28, 43, 20, 43, 54, 53, 26, 53]</td>\n",
       "      <td>[21, 91, 45, 6, 91]</td>\n",
       "      <td>[1, 45, 8, 93]</td>\n",
       "      <td>[45, 4]</td>\n",
       "      <td>[43]</td>\n",
       "      <td>[17, 49, 17, 62, 23]</td>\n",
       "      <td>[62, 23, 62, 16, 47, 16]</td>\n",
       "      <td>[32]</td>\n",
       "      <td>[29]</td>\n",
       "      <td>[33]</td>\n",
       "      <td>[47]</td>\n",
       "      <td>[58]</td>\n",
       "      <td>[49]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.751537</td>\n",
       "      <td>-0.762542</td>\n",
       "      <td>-0.00174809</td>\n",
       "      <td>-0.000431608</td>\n",
       "      <td>-0.00126228</td>\n",
       "      <td>-0.909091</td>\n",
       "      <td>-0.827408</td>\n",
       "      <td>-0.795688</td>\n",
       "      <td>-0.561505</td>\n",
       "      <td>-0.702596</td>\n",
       "      <td>-0.429389</td>\n",
       "      <td>-0.909091</td>\n",
       "      <td>-0.909091</td>\n",
       "      <td>-0.835989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.75779</td>\n",
       "      <td>-0.765135</td>\n",
       "      <td>-0.00180749</td>\n",
       "      <td>-0.000551299</td>\n",
       "      <td>-0.000717953</td>\n",
       "      <td>-0.908541</td>\n",
       "      <td>-0.828244</td>\n",
       "      <td>-0.79836</td>\n",
       "      <td>-0.569212</td>\n",
       "      <td>-0.703297</td>\n",
       "      <td>-0.426025</td>\n",
       "      <td>-0.914489</td>\n",
       "      <td>-0.913659</td>\n",
       "      <td>-0.842998</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               plasmid_0  \\\n",
       "GC_sum                        -0.0062533   \n",
       "chain   [53, 44, 43, 11, 42, 58, 24, 58]   \n",
       "gd_sum                         -0.751537   \n",
       "total                           -0.75779   \n",
       "\n",
       "                                               plasmid_1           plasmid_10  \\\n",
       "GC_sum                                       -0.00259313         -5.94052e-05   \n",
       "chain   [35, 42, 15, 35, 28, 43, 20, 43, 54, 53, 26, 53]  [21, 91, 45, 6, 91]   \n",
       "gd_sum                                         -0.762542          -0.00174809   \n",
       "total                                          -0.765135          -0.00180749   \n",
       "\n",
       "            plasmid_11   plasmid_12   plasmid_13             plasmid_2  \\\n",
       "GC_sum    -0.000119691  0.000544323  0.000550089          -0.000836153   \n",
       "chain   [1, 45, 8, 93]      [45, 4]         [43]  [17, 49, 17, 62, 23]   \n",
       "gd_sum    -0.000431608  -0.00126228    -0.909091             -0.827408   \n",
       "total     -0.000551299 -0.000717953    -0.908541             -0.828244   \n",
       "\n",
       "                       plasmid_3   plasmid_4    plasmid_5   plasmid_6  \\\n",
       "GC_sum               -0.00267233 -0.00770681 -0.000701078  0.00336402   \n",
       "chain   [62, 23, 62, 16, 47, 16]        [32]         [29]        [33]   \n",
       "gd_sum                 -0.795688   -0.561505    -0.702596   -0.429389   \n",
       "total                   -0.79836   -0.569212    -0.703297   -0.426025   \n",
       "\n",
       "         plasmid_7   plasmid_8   plasmid_9  \n",
       "GC_sum -0.00539855 -0.00456856 -0.00700929  \n",
       "chain         [47]        [58]        [49]  \n",
       "gd_sum   -0.909091   -0.909091   -0.835989  \n",
       "total    -0.914489   -0.913659   -0.842998  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_2  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_8  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n",
      "plasmid_8  from greedy completely contained in  plasmid_5  from MILP.\n",
      "\n",
      "plasmid_8  from greedy completely contained in  plasmid_6  from MILP.\n",
      "\n",
      "plasmid_9  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_9  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_13  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_13  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_13  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
