{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 39 with $\\alpha_1 = 1$, $\\alpha_2 = 1$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 39\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP018949.1</td>\n",
       "      <td>16559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP018951.1</td>\n",
       "      <td>43380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP018952.1</td>\n",
       "      <td>132345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP018950.1</td>\n",
       "      <td>35225</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP018949.1        16559\n",
       "1  CP018951.1        43380\n",
       "2  CP018952.1       132345\n",
       "3  CP018950.1        35225"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xl4VfWdx/HPNythEQVCWCVUWWSL\nSAwKWIIiQqeCfbBKWpzKUhitYy20ykwXLbW1OKIjjlTQ+iCVKqCVUqUF2iHFXQKCAjGACAQR2QNh\nzfKdP3LCxJCEi+YkAd6v5+HxnnN+93e+9/b+mvu553fOMXcXAAAAAETVdgEAAAAA6gbCAQAAAABJ\nhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAgZjaLuBMNWvW\nzJOTk2u7jLPG4cOH1aBBg9ouA6jzGCtAZBgrZ2blypV73D2xtusAInXWhYPk5GRlZWXVdhlnjczM\nTKWnp9d2GUCdx1gBIsNYOTNmtrW2awDOBNOKAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAA\nAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIhBYOzOxZM9tlZmsr2W5mNs3MNpnZB2Z2RVi1AAAAADi9\nMI8czJI0uIrtQyR1CP6Nk/S7EGsBAAAAcBqhhQN3Xy5pXxVNhkma7SXekXShmbUMqx4AAAAAVYup\nxX23lpRbZnl7sO6z8g3NbJxKji4oKSlJmZmZNVHfOSE/P5/3C6H58NO82i6h2rRvHM1YASLA3xXg\n3Fab4cAqWOcVNXT3mZJmSlJqaqqnp6eHWNa5JTMzU7xfCMvtk16r7RKqzazBDRgrQAT4uwKc22rz\nakXbJbUts9xG0o5aqgUAAAA479VmOFgo6V+DqxZdJSnP3U+ZUgQAAACgZoQ2rcjMXpCULqmZmW2X\ndL+kWEly96ckLZL0DUmbJB2RNCqsWgAAAACcXmjhwN0zTrPdJf0grP0DAFBXJJ9j5+cAOHdxh2QA\nAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAArV5E7Q6ixPHAAAAcD7iyAEAAAAASYQDAAAAAAHCAQAA\nAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAA\nAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAA\nSYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJ\nhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmE\nAwAAAAABwgEAAAAASSGHAzMbbGY5ZrbJzCZVsP1iM1tmZu+b2Qdm9o0w6wEAAABQudDCgZlFS3pS\n0hBJXSRlmFmXcs1+Jmmeu/eUNELS9LDqAQAAAFC1MI8cpEna5O6b3f2EpBclDSvXxiVdEDxuLGlH\niPUAAAAAqEJMiH23lpRbZnm7pN7l2jwgaYmZ/bukBpIGhlgPAAAAgCqEGQ6sgnVebjlD0ix3n2pm\nV0v6g5l1c/fiL3RkNk7SOElKSkpSZmZmGPWeNLF7Yaj916T8/PzQ3y+cvxgrQGQYKwDOFmGGg+2S\n2pZZbqNTpw2NkTRYktz9bTOrJ6mZpF1lG7n7TEkzJSk1NdXT09NDKrnE7ZNeC7X/mjRrcAOF/X7h\n/MVYASLDWAFwtgjznIMVkjqYWXszi1PJCccLy7XZJuk6STKzyyTVk7Q7xJoAAAAAVCK0cODuhZLu\nkrRYUrZKrkq0zswmm9nQoNlESd83szWSXpB0u7uXn3oEAAAAoAaEOa1I7r5I0qJy635R5vF6SX3D\nrAEAAABAZLhDMgAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJ\nhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmE\nAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQD\nAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMA\nAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAA\nAAABwgEAAAAASYQDAAAAAAHCAQAAAABJIYcDMxtsZjlmtsnMJlXS5hYzW29m68zsj2HWAwAAAKBy\nMWF1bGbRkp6UdL2k7ZJWmNlCd19fpk0HSf8hqa+77zez5mHVAwAAAKBqYR45SJO0yd03u/sJSS9K\nGlauzfclPenu+yXJ3XeFWA8AAACAKoR25EBSa0m5ZZa3S+pdrk1HSTKzNyVFS3rA3f9WviMzGydp\nnCQlJSUpMzMzjHpPmti9MNT+a1J+fn7o7xfOX4wVIDKMFQBnizDDgVWwzivYfwdJ6ZLaSHrdzLq5\n+4EvPMl9pqSZkpSamurp6enVXmxZt096LdT+a9KswQ0U9vuF8xdjBYgMYwXA2SLMaUXbJbUts9xG\n0o4K2vzZ3Qvc/RNJOSoJCwAAAABqWJjhYIWkDmbW3sziJI2QtLBcmwWSBkiSmTVTyTSjzSHWBAAA\nAKASoYUDdy+UdJekxZKyJc1z93VmNtnMhgbNFkvaa2brJS2T9BN33xtWTQAAAAAqF+Y5B3L3RZIW\nlVv3izKPXdKE4B8AAACAWsQdkgEAAABIIhwAAAAACBAOAAAAAEg6g3BgZv3MbFTwONHM2odXFgAA\nAICaFlE4MLP7Jd0n6T+CVbGSng+rKAAAAAA1L9IjB9+SNFTSYUly9x2SGoVVFAAAAICaF2k4OBFc\ndtQlycwahFcSAAAAgNoQaTiYZ2YzJF1oZt+X9HdJT4dXFgAAAICaFtFN0Nz9ETO7XtJBSZ0k/cLd\nl4ZaGQAAAIAaddpwYGbRkha7+0BJBAIAAADgHHXaaUXuXiTpiJk1roF6AAAAANSSiKYVSTom6UMz\nW6rgikWS5O53h1IVAAAAgBoXaTh4LfgHAAAA4BwV6QnJz5lZnKSOwaocdy8IrywAAAAANS2icGBm\n6ZKek7RFkklqa2bfc/fl4ZUGAAAAoCZFOq1oqqRB7p4jSWbWUdILknqFVRgAAACAmhXpTdBiS4OB\nJLn7Bkmx4ZQEAAAAoDZEeuQgy8x+L+kPwfJ3Ja0MpyQAAAAAtSHScHCHpB9Iulsl5xwslzQ9rKIA\nAAAA1LxIw0GMpMfd/VHp5F2T40OrCgAA4By1cuXK5jExMc9I6qbIp3gD1aFY0trCwsKxvXr12lVR\ng0jDwT8kDZSUHywnSFoiqc9XLhEAAOA8EhMT80yLFi0uS0xM3B8VFeW1XQ/OH8XFxbZ79+4uO3fu\nfEbS0IraRJpW67l7aTBQ8Lh+NdQIAABwvumWmJh4kGCAmhYVFeWJiYl5KjlqVaFIjxwcNrMr3H2V\nJJlZqqSj1VAjANQNn62WHhhW21VUjwfyarsCAFWLIhigtgSfvUoPEEQaDn4oab6Z7ZDkklpJuvWr\nlwcAAACgrog0HLSX1FPSxZK+JekqlYQEAAAAfAXJk16r1pvKbvntv9SZy82npaV1euSRR3K//vWv\nH4mkfXFxsfr06dNx0aJFm5o0aVIcdn1f1auvvtpo6tSpScuWLdsURv9z5sxpvG7duoTf/OY3O8tv\nq1+/fs8jR468X9lz+/Tp0/HPf/7zx4mJiUVnss9Izzn4ubsflHShpOslzZT0uzPZEQAAAM4eBQUF\nNb7PefPmNe7atevRsyEY1ITvfve7eRUFg0hkZGTsfeSRRxLP9HmRHjkoTRz/Iukpd/+zmT1wpjsD\nAABnOc7POSf85Cc/afnSSy81admy5YmmTZsW9uzZ88jkyZM/T0tL65SWlpb/7rvvNvzGN75xYPz4\n8XtHjRrV7tNPP42TpEcffXTboEGDDh88eDBqzJgxF2dnZycUFRXZT3/60x0jR448kJ+fbyNGjGi/\nYcOGeh06dDh27Ngxk6THHnus2dq1axN+//vf50rS1KlTm2VnZ9d75plntpeta86cOU3Gjx+/R5Jy\ncnLiBg8e3CEtLS1/1apVDS+77LIjo0eP3jN58uTWe/fujZk1a9bmAQMGHKmslpycnLjvfOc77Y8e\nPRolSY8//vi266+//vCrr77aaPLkya2aNGlSkJOTk9C9e/cjCxYs+CQqqvLfzIcPH54cHx9fnJOT\nk7B3797Yhx56KDcjI+MLH6Bly5bVnzBhwsXHjh2LqlevXvGsWbM+SUlJOZ6VlVVv1KhR7QsKCqy4\nuFgvv/zyx3FxcR7Ja5s2bVrTrKysBrNnz9720UcfxY0YMeJrhYWFdt11153c99atW2OHDx/+tfz8\n/OiioiJ74okntg4ePDh/xIgRB/r06dN5ypQpZxQuIg0Hn5rZDJVcznSKmcWL6/ICAACcdZYvX17/\nL3/5y0Uffvjh+oKCArv88su79OzZ8+S0nwMHDkSvWLEiR5JuvPHG9hMmTPj8hhtuyN+4cWPcDTfc\n0GHz5s3r/vM//7PlgAEDDs6fP3/Lnj17olNTUy8bOnTowUcffTQxISGheMOGDevffffdhL59+3aR\npDFjxuzr2rVrl+PHj2+Pj4/3559/vtmMGTO2lq9t5cqVDfv27XtyfW5ubr25c+du7tWr19YePXpc\nNmfOnKZZWVkf/fGPf7zw17/+dcsBAwZ8XFktrVq1Knz99dc31K9f3z/88MP4jIyMr61duzZbkrKz\nsxNWr169OTk5uaBXr16dly5d2vCGG27IL19PWbm5ufHvvfdezvr16+MHDhzYadiwYR+W3Z6SknLs\nvffe+yg2NlYLFixodO+997ZZvHjxx0888UTinXfe+fkdd9yx79ixY1ZYWKhPP/00NpLXVrb/O++8\n8+KxY8fuvuuuu/Y+9NBDJ48IPPvss02uu+66vClTpuwsLCzUoUOHoiQpMTGx6MSJE7Zz587oFi1a\nRDy1KNJwcIukwZIecfcDZtZS0k8i3QkAAADqhszMzIZDhgw50LBhQ5fk119//YGy2zMyMvaVPn7z\nzTcv2LhxY0Lpcn5+fvT+/fujMjMzL1i8ePGF06ZNayFJx48ft02bNsW98cYbDe++++5dktS7d++j\nHTt2PCJJF1xwQXHfvn0PzZ07t3H37t2PFRQUWFpa2ilXvszLy4u56KKLTk4pat269fHSdh07djx6\n7bXXHoyKitIVV1xx5MEHH2wVvJ4Ka2nXrl3BmDFj2q1fvz4hKipKW7duPXkD3+7dux++5JJLCiSp\na9euRz7++OO4071vw4cP3xcdHa3u3bsfb9u27fHVq1fXK7t937590bfeemv7LVu21DMzLygoMEm6\n+uqrDz/yyCMtt2/fHjdixIj93bt3Px7paytr1apVDf/6179+LEnjx4/f+6tf/aqNJF111VWHx48f\nn1xQUBB188037+/Tp8/J97Vp06aF27Zti2vRokXEVxmNKBy4+xFJfyqz/JmkzyLdCWoRh38BAEAZ\n7lVfU6ZRo0Ynv5y7u7KysrKDIPGFPl566aVNKSkpx8s/38wq7HfcuHF7fv3rX7fo2LHjsZEjR+6p\nqE10dLQXFRUpOjpakhQXF3dyv1FRUapXr54H7VRUVGRV1TJhwoRWzZs3L3j55Zc/KS4uVkJCwskT\nv+Pj40/2Gx0drcLCwoqLruJ1lV++7777Wvfv3//Q0qVLP87JyYm79tprO0nSv/3bv+275pprDr/y\nyiuNhwwZ0nH69OlbOnXqdDyS11ZeRZfAHTJkSP7y5ctzXn755ca33357+7vvvvvzu+66a69UEpTq\n169/RudvMDUIAADgPJKenp6/ePHixkeOHLG8vLyov//97xdW1rZfv34Hp0yZ0rx0+a233kqQpAED\nBhycOnVqUnFxyffON998MyFon//88883kaQVK1bU27Bhw8mb5l577bWHP/vss7hXXnml6ZgxY/ap\nAu3btz+WnZ0dX9G2ylRWS15eXnTLli0LoqOjNX369KZFRaefWfODH/yg9ezZsyt8P/70pz9dVFRU\npHXr1sXn5ubGp6SkHCu7/eDBg9Ft2rQ5IUkzZsxoVrp+/fr1cZdddtnxn/3sZ7sGDRp0YPXq1Qnl\n+47EFVdckf/00083kaSnn366aen6DRs2xLVu3bpg4sSJe0aOHLln1apV9aWSKz/t3r07tlOnTqcE\nuKpEOq0IAAAAIajpS4/279//yODBg/O6dOnStXXr1sd79OhxuHHjxhV+c545c2bu2LFjL+7YsWOX\noqIi692796E+ffps++1vf7tj3LhxF3fu3LmLu1ubNm2OL1u2bNOPf/zjXSNGjGjfsWPHLl27dj3S\nvXv3w2X7u+mmm/Z/8MEH9Su7vOagQYPylixZ0qhbt24Rf6GtrJZ77rln1/Dhwy9ZsGDBRf369TuU\nkJBw2l/Q169fn/Ctb33rQEXbLr300uNpaWmd9u7dG/vf//3fW+vXr/+FX/Hvu+++nWPHjm0/bdq0\nFtdcc83B0vV/+MMfmsyfP79pTEyMJyYmFjz00EM7Dhw4EB3p6ys1ffr0bSNGjPja9OnTk4YOHbq/\ndP3ixYsbTZs2rUVMTIzXr1+/aM6cOZ9I0htvvFG/Z8+eh2NjY89oP3a6Q0t1TWpqqmdlZYW6j+RJ\nr4Xaf02albJR6Tn313YZ1YNpRXUOY6WOYqzUOYyVOqoGxoqZrXT31LLr1qxZsyUlJaXCaTU1JS8v\nL6px48bFhw4dirr66qs7PfXUU1v79esX0b0IvooBAwZces8993w+bNiwQxVt37p1a2xGRkbyW2+9\ntTHsWirSr1+/Dm+88cYp+x4+fHjyN7/5zbxRo0btr+h5ddGoUaPa3nTTTQcqeq/XrFnTLCUlJbmi\n5zGtCAAA4DwzcuTIdp07d+7So0ePy2688cb9YQeDPXv2RCcnJ3erV69ecWXBQJLatWtXMHr06D37\n9u2rle+oFQWDs1W3bt2OVvVeV4ZpRQAAAOeZv/zlL5/U5P6aNWtWtGXLlrWRtB07dmyd+3X+5Zdf\n3lLbNZypiRMnfqmjUxw5AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACDACckAAAC16YHGvU7f6Ez6\nyzvtfRPMrNewYcP2LViw4BNJKigoUPPmzVMuv/zyw8uWLds0bdq0pllZWQ1mz569bcKECa0aNmxY\nNHny5M/L9hEdHd2rQ4cOR4uKiuzSSy89Om/evC1l764sldyIq0+fPh0XLVq0qUmTJsXf/va3k//x\nj380btq0aeHGjRvXVVTb+++/X2/UqFHJ69evrz9p0qRPy++31G9+85vEp556Kik3Nzd+x44da1q2\nbFkoSS+88ELj9957r8Fjjz22I9K3DP+PIwcAAADnmYSEhOKcnJyE/Px8k6RXXnnlgqSkpIIz6SM+\nPr74o48+Wr9x48Z1sbGxPnXq1MTybebNm9e4a9euR5s0aVIsSaNHj96zcOHCKi8X2rx588LHH398\n2/jx4ysMBaX69++fv3Tp0g2tWrU6UXb9rbfemve3v/3twkOHDvE990vgTQMAADgPXXfddXnz58+/\nUJJeeOGFJsOHD9/3Zfvq169f/qZNm+LLr58zZ06TsnccHjJkSH5iYmJhVX21bt26sH///kdiY2Or\nvFNv3759j3bq1OlE+fVRUVHq06fPoblz5zY+k9eAEoQDAACA89Btt922b+7cuRcdOXLEsrOz6199\n9dWHv0w/BQUFWrx48QXdu3c/Wn7bypUrG/bt2/e0/T788MOJDz/88ClHHsrr37//pVu2bIk9XbvU\n1NTDr7/+esPTtcOpQj3nwMwGS3pcUrSkZ9z9t5W0u1nSfElXuntWmDUBAABA6t2799Ht27fHP/30\n000GDhyYd6bPP378eFTnzp27BH0d+uEPf3jKTbfy8vJiLrroouJTn/1F99577+5I9vnPf/5zUyTt\nWrRoUbhz5864SNrii0ILB2YWLelJSddL2i5phZktdPf15do1knS3pHfDqgUAAACnGjx48IH777+/\n7ZIlS3J27dp1Rt8LS885qKpNdHS0FxUVKTo6+qsVeoaOHj1q9erVO20owanCnFaUJmmTu2929xOS\nXpQ0rIJ2v5L0sKRjIdYCAACAcu644449EydO3JGWlnbKlKDq0L59+2PZ2dmnnIsQtpycnHpdu3YN\n5TWd68KcVtRaUm6Z5e2SepdtYGY9JbV191fN7MeVdWRm4ySNk6SkpCRlZmZWf7VlTOxe5XkyZ5X8\n+FbK7PTL2i6jeoT8vzvOHGOljmKs1DmMlTqqroyVCC49GpZLLrmk4Oc///mu07V77LHHWs6YMSOp\ndPnzzz//IJL+Bw0alLdkyZJG3bp1Oy5JN954Y/t33nmn0f79+2OSkpJ6TJo0acePfvSjPaXnG9x7\n7727t23bFnPllVd2OXz4cLSZ+YwZM5Kys7PXNmnSpLh///6XPvfcc1uTk5MLHnzwweZPPPFEi717\n98ampKR0GTBgQN7cuXO3StLy5csbTZky5dMv966c38y9yhPBv3zHZt+WdIO7jw2Wb5OU5u7/HixH\nSfpfSbe7+xYzy5T049Odc5CamupZWeGelpA86bVQ+69Js1I2Kj3n/touo3o8cMbTIREyxkodxVip\ncxgrdVQNjBUzW+nuqWXXrVmzZktKSsop8/PPRVu3bo3NyMhIfuutt6q8fGl1ys3Njbnlllu+9vbb\nb2+oqX2ebdasWdMsJSUluaJtYU4r2i6pbZnlNpLK3oyikaRukjLNbIukqyQtNLMvDCAAAACcndq1\na1cwevToPfv27auxK2Ru3rw5burUqbmnb4mKhDmtaIWkDmbWXtKnkkZI+k7pRnfPk9SsdDnSIwcA\nAAA4e4wdO3Z/Te6vf//+R2pyf+ea0FKcuxdKukvSYknZkua5+zozm2xmQ8PaLwAAQB1XXFxcbLVd\nBM5PwWev0is5hXqfA3dfJGlRuXW/qKRtepi1AAAA1BFrd+/e3SUxMTEvKioqnJM/gQoUFxfb7t27\nG0taW1mbUMMBAAAAvqiwsHDszp07n9m5c2c3hXv+J1BesaS1hYWFYytrQDgAAACoQb169doliSnW\nqJNIqwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAA\nBAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAE\nCAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQI\nBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgH\nAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcA\nAAAAJIUcDsxssJnlmNkmM5tUwfYJZrbezD4ws3+YWbsw6wEAAABQudDCgZlFS3pS0hBJXSRlmFmX\ncs3el5Tq7j0kvSTp4bDqAQAAAFC1MI8cpEna5O6b3f2EpBclDSvbwN2XufuRYPEdSW1CrAcAAABA\nFczdw+nY7GZJg919bLB8m6Te7n5XJe3/R9JOd3+wgm3jJI2TpKSkpF4vvvhiKDWX+vDTvFD7r0nt\nE46r4fEdtV1G9Wh5eW1XgHIYK3UUY6XOYazUUTUwVgYMGLDS3VND3xFQTWJC7NsqWFdhEjGzkZJS\nJfWvaLu7z5Q0U5JSU1M9PT29mkqs2O2TXgu1/5o0K+UTpefcX9tlVI+Mc+eP67mCsVJHMVbqHMZK\nHcVYAU4RZjjYLqltmeU2kk75qcHMBkr6qaT+7n48xHoAAAAAVCHMcw5WSOpgZu3NLE7SCEkLyzYw\ns56SZkga6u67QqwFAAAAwGmEFg7cvVDSXZIWS8qWNM/d15nZZDMbGjT7L0kNJc03s9VmtrCS7gAA\nAACELMxpRXL3RZIWlVv3izKPB4a5fwAAAACR4w7JAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQI\nBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgH\nAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcA\nAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAA\nAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAA\nACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABEINB2Y22MxyzGyTmU2qYHu8mc0N\ntr9rZslh1gMAAACgcqGFAzOLlvSkpCGSukjKMLMu5ZqNkbTf3S+V9JikKWHVAwAAAKBqYR45SJO0\nyd03u/sJSS9KGlauzTBJzwWPX5J0nZlZiDUBAAAAqESY4aC1pNwyy9uDdRW2cfdCSXmSmoZYEwAA\nAIBKxITYd0VHAPxLtJGZjZM0LljMN7Ocr1jbeWOA1EzSntquo1r8koNKCA9jBYgMY+WMtauJnQDV\nJcxwsF1S2zLLbSTtqKTNdjOLkdRY0r7yHbn7TEkzQ6rznGZmWe6eWtt1AHUdYwWIDGMFOLeFOa1o\nhaQOZtbezOIkjZC0sFybhZK+Fzy+WdL/uvspRw4AAAAAhC+0IwfuXmhmd0laLCla0rPuvs7MJkvK\ncveFkn4v6Q9mtkklRwxGhFUPAAAAgKoZP9Sf28xsXDAtC0AVGCtAZBgrwLmNcAAAAABAUsh3SAYA\nAABw9iAcAAAAAJBEODjvmVlt1HvzAAAErklEQVSyma0NHqeb2au1XRNQG8zsrdNsX2RmF9ZUPcC5\nwMwyzSw1eLzFzJrVdk0AqhbmfQ4QIjMzlZwzUlzbtQB1jZlFu3vRmTzH3fucZvs3vlpVwLmHv0XA\nuYcjB2eR4Ff+bDObLmmVpNvM7G0zW2Vm882sYdDuSjN7y8zWmNl7ZtYoeO7rQdtVZlblFyGgrgo+\nyx+Z2XNm9oGZvWRm9YNfJX9hZm9I+raZXWJmfzOzlcFnv3Pw/CQzeyUYH2tKx4KZ5Qf/bWlmy81s\ntZmtNbNrgvUnf/U0swnBtrVmdk+ZurLN7GkzW2dmS8wsoVbeJKAK5T+/ZjbFzO4ss/0BM5sYPP6J\nma0Ixtovg3Xl/xa1NbPfmVlW8Nn/Ze28MgDVgXBw9ukkabak6yWNkTTQ3a+QlCVpQnDDubmSfuju\nKZIGSjoqaZek64O2t0qaVhvFA9Wkk6SZ7t5D0kFJpV9sjrl7P3d/USV3Vf93d+8l6ceSpgdtpkn6\nZzA+rpC0rlzf35G02N0vl5QiaXXZjWbWS9IoSb0lXSXp+2bWM9jcQdKT7t5V0gFJw6vrBQPVoaLP\nr6QXVfJ3odQtkuab2SCVfKbTJF0uqZeZfT1o00nSbHfv6e5bJf00uGtyD0n9zaxHjbwgANWOaUVn\nn63u/o6ZfVNSF0lvlhzVVZykt1Xyf9ifufsKSXL3g5JkZg0k/Y+ZXS6pSFLH2igeqCa57v5m8Ph5\nSXcHj+dKUnAUrY9KvuCUPic++O+1kv5VkoKpR3nl+l4h6Vkzi5W0wN1Xl9veT9Ir7n442NefJF2j\nkju+f1Km/UpJyV/hNQJhqOzz29zMWklKlLTf3beZ2d2SBkl6P3huQ5WEhW0K/haV6fcWMxunku8V\nLVXy9+mDmnhBAKoX4eDsczj4r0la6u4ZZTcGv9ZUdPOKH0n6XCW/hEZJOhZmkUDIyn/GS5dLx0eU\npAPBr/9n1rH78uDX0X9RyR3c/8vdZ5dpYpU8VZKOl3lcJIlpRahrKvv8viTpZkktVHIkobTtQ+4+\n4wsdmCXr/8eazKy9So7OXenu+81slqR61Vo1gBrDtKKz1zuS+prZpZIUzLnuKOkjSa3M7MpgfSMz\ni5HUWCVHFIol3SYpupbqBqrDxWZ2dfA4Q9IbZTcGR8w+MbNvSyUnTZpZSrD5H5LuCNZHm9kFZZ9r\nZu0k7XL3pyX9XiVTj8paLummYMw1kPQtSa9X30sDQlXZ5/dFSSNUEhBeCtouljS6zPlsrc2seQV9\nXqCSsJBnZkmShoT8GgCEiHBwlnL33ZJul/SCmX2gkrDQ2d1PqGTu6BNmtkbSUpX8gjNd0vfM7B2V\nTCk6XGHHwNkhWyWf5w8kNZH0uwrafFfSmGAcrJM0LFj/Q0kDzOxDlUz96VrueemSVpvZ+yo5Z+Dx\nshvdfZWkWZLek/SupGfc/X0BZ4HKPr/uvk5SI0mfuvtnQdslkv4o6e1gvLwUtCnf5xqVTD1aJ+lZ\nSW+WbwPg7GHuFc1AAYC6KZjS8Kq7d6vlUgAAOOdw5AAAAACAJI4cAAAAAAhw5AAAAACAJMIBAAAA\ngADhAAAAAIAkwgEAAACAAOEAAAAAgCTp/wCbverJcnhJSQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f2638b01940>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:1:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>158604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>43380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>25501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>6549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>2708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>1588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>1659</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0       158604\n",
       "1  plasmid_1        43380\n",
       "2  plasmid_2        25501\n",
       "3  plasmid_3         6549\n",
       "4  plasmid_4         2708\n",
       "5  plasmid_5         1588\n",
       "6  plasmid_6         1659"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.468518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.492198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.480066</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.468518\n",
       "1  precision  0.492198\n",
       "2   f1_score  0.480066"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP018949.1</th>\n",
       "      <th>CP018950.1</th>\n",
       "      <th>CP018951.1</th>\n",
       "      <th>CP018952.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.463282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.022484</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.007896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.058907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.050648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018926</td>\n",
       "      <td>0.021588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018903</td>\n",
       "      <td>0.019782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.022484</td>\n",
       "      <td>1.037828</td>\n",
       "      <td>0.635045</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP018949.1  CP018950.1  CP018951.1  CP018952.1\n",
       "0  plasmid_0         0.0    0.000000    0.000000    0.463282\n",
       "1  plasmid_1         0.0    0.022484    1.000000    0.007896\n",
       "2  plasmid_2         0.0    0.000000    0.000000    0.058907\n",
       "3  plasmid_3         0.0    0.000000    0.000000    0.050648\n",
       "4  plasmid_4         0.0    0.000000    0.018926    0.021588\n",
       "5  plasmid_5         0.0    0.000000    0.018903    0.019782\n",
       "6  plasmid_6         0.0    0.000000    0.000000    0.012943\n",
       "7      Total         0.0    0.022484    1.037828    0.635045"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP018949.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP018950.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018257</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP018951.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.303176</td>\n",
       "      <td>0.516373</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP018952.1</td>\n",
       "      <td>0.370413</td>\n",
       "      <td>0.019825</td>\n",
       "      <td>0.210384</td>\n",
       "      <td>0.945182</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.782746</td>\n",
       "      <td>0.29355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.370413</td>\n",
       "      <td>1.038082</td>\n",
       "      <td>0.210384</td>\n",
       "      <td>0.945182</td>\n",
       "      <td>1.303176</td>\n",
       "      <td>1.299118</td>\n",
       "      <td>0.29355</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2  plasmid_3  plasmid_4  \\\n",
       "0  CP018949.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "1  CP018950.1   0.000000   0.018257   0.000000   0.000000   0.000000   \n",
       "2  CP018951.1   0.000000   1.000000   0.000000   0.000000   0.303176   \n",
       "3  CP018952.1   0.370413   0.019825   0.210384   0.945182   1.000000   \n",
       "4       Total   0.370413   1.038082   0.210384   0.945182   1.303176   \n",
       "\n",
       "   plasmid_5  plasmid_6  \n",
       "0   0.000000    0.00000  \n",
       "1   0.000000    0.00000  \n",
       "2   0.516373    0.00000  \n",
       "3   0.782746    0.29355  \n",
       "4   1.299118    0.29355  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.370413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.210384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.945182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>0.782746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>0.293550</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.370413\n",
       "1  plasmid_1    1.000000\n",
       "2  plasmid_2    0.210384\n",
       "3  plasmid_3    0.945182\n",
       "4  plasmid_4    1.000000\n",
       "5  plasmid_5    0.782746\n",
       "6  plasmid_6    0.293550"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP018949.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP018950.1</td>\n",
       "      <td>0.022484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP018951.1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP018952.1</td>\n",
       "      <td>0.471646</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP018949.1    0.000000\n",
       "1  CP018950.1    0.022484\n",
       "2  CP018951.1    1.000000\n",
       "3  CP018952.1    0.471646"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP018951.1</td>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction  recall  precision\n",
       "0  CP018951.1  plasmid_1     1.0        1.0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 1, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.000817127</td>\n",
       "      <td>-0.019466</td>\n",
       "      <td>0.0136153</td>\n",
       "      <td>-0.00443081</td>\n",
       "      <td>-0.0128611</td>\n",
       "      <td>0.00802071</td>\n",
       "      <td>0.00676867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[113, 83, 92, 89, 30, 56, 46, 53, 76, 36, 125,...</td>\n",
       "      <td>[99, 22, 71]</td>\n",
       "      <td>[56, 46, 53, 76, 36, 125, 115, 82, 73]</td>\n",
       "      <td>[60, 49, 101, 81]</td>\n",
       "      <td>[60, 71, 99]</td>\n",
       "      <td>[89, 92, 83, 71, 99]</td>\n",
       "      <td>[89, 92, 83, 73, 82]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.391434</td>\n",
       "      <td>-0.436007</td>\n",
       "      <td>-0.209992</td>\n",
       "      <td>-0.429302</td>\n",
       "      <td>-0.472674</td>\n",
       "      <td>-0.432935</td>\n",
       "      <td>-0.25859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.392251</td>\n",
       "      <td>-0.455473</td>\n",
       "      <td>-0.196377</td>\n",
       "      <td>-0.433733</td>\n",
       "      <td>-0.485535</td>\n",
       "      <td>-0.424914</td>\n",
       "      <td>-0.251821</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0     plasmid_1  \\\n",
       "GC_sum                                       -0.000817127     -0.019466   \n",
       "chain   [113, 83, 92, 89, 30, 56, 46, 53, 76, 36, 125,...  [99, 22, 71]   \n",
       "gd_sum                                          -0.391434     -0.436007   \n",
       "total                                           -0.392251     -0.455473   \n",
       "\n",
       "                                     plasmid_2          plasmid_3  \\\n",
       "GC_sum                               0.0136153        -0.00443081   \n",
       "chain   [56, 46, 53, 76, 36, 125, 115, 82, 73]  [60, 49, 101, 81]   \n",
       "gd_sum                               -0.209992          -0.429302   \n",
       "total                                -0.196377          -0.433733   \n",
       "\n",
       "           plasmid_4             plasmid_5             plasmid_6  \n",
       "GC_sum    -0.0128611            0.00802071            0.00676867  \n",
       "chain   [60, 71, 99]  [89, 92, 83, 71, 99]  [89, 92, 83, 73, 82]  \n",
       "gd_sum     -0.472674             -0.432935              -0.25859  \n",
       "total      -0.485535             -0.424914             -0.251821  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_11</th>\n",
       "      <th>plasmid_12</th>\n",
       "      <th>plasmid_13</th>\n",
       "      <th>plasmid_14</th>\n",
       "      <th>plasmid_15</th>\n",
       "      <th>plasmid_16</th>\n",
       "      <th>plasmid_17</th>\n",
       "      <th>plasmid_18</th>\n",
       "      <th>plasmid_19</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_20</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00413489</td>\n",
       "      <td>-0.045737</td>\n",
       "      <td>0.00838095</td>\n",
       "      <td>0.000116057</td>\n",
       "      <td>0.0113787</td>\n",
       "      <td>-0.00358467</td>\n",
       "      <td>0.00743855</td>\n",
       "      <td>-0.000892041</td>\n",
       "      <td>0.00275658</td>\n",
       "      <td>0.0234336</td>\n",
       "      <td>0.0201804</td>\n",
       "      <td>1.10986e-05</td>\n",
       "      <td>0.0283625</td>\n",
       "      <td>-0.054933</td>\n",
       "      <td>-0.0051549</td>\n",
       "      <td>-0.00121919</td>\n",
       "      <td>0.0276208</td>\n",
       "      <td>0.029781</td>\n",
       "      <td>0.0123917</td>\n",
       "      <td>0.00107272</td>\n",
       "      <td>-0.00185382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[99, 87, 81, 101, 49, 60, 43, 104, 111, 65, 74...</td>\n",
       "      <td>[79, 24, 79, 62]</td>\n",
       "      <td>[65, 104]</td>\n",
       "      <td>[27, 56, 2, 72, 41]</td>\n",
       "      <td>[56]</td>\n",
       "      <td>[108, 29, 83, 92, 89, 35, 98]</td>\n",
       "      <td>[83, 92, 89]</td>\n",
       "      <td>[52, 77, 61, 55]</td>\n",
       "      <td>[58, 57, 55, 10]</td>\n",
       "      <td>[55, 34, 103]</td>\n",
       "      <td>[79]</td>\n",
       "      <td>[130, 13, 73, 82, 115]</td>\n",
       "      <td>[31]</td>\n",
       "      <td>[42, 73]</td>\n",
       "      <td>[67, 18, 83, 92, 89, 32, 102, 81, 90, 40, 101,...</td>\n",
       "      <td>[72, 19, 56, 15, 77, 51, 56, 46, 53, 76, 36, 1...</td>\n",
       "      <td>[81, 90]</td>\n",
       "      <td>[90]</td>\n",
       "      <td>[71]</td>\n",
       "      <td>[88, 125, 115, 82, 73, 45, 60, 4, 97]</td>\n",
       "      <td>[108, 14, 55, 33, 65, 104, 1, 68, 11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.441833</td>\n",
       "      <td>-0.43895</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.00699786</td>\n",
       "      <td>-0.485387</td>\n",
       "      <td>-0.00823149</td>\n",
       "      <td>-0.361328</td>\n",
       "      <td>-0.1319</td>\n",
       "      <td>-0.00529888</td>\n",
       "      <td>-0.0704741</td>\n",
       "      <td>-0.430952</td>\n",
       "      <td>-0.00105601</td>\n",
       "      <td>-0.440214</td>\n",
       "      <td>-0.0143141</td>\n",
       "      <td>-0.0445676</td>\n",
       "      <td>-0.0171148</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.00298832</td>\n",
       "      <td>-0.0174707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.445968</td>\n",
       "      <td>-0.484687</td>\n",
       "      <td>-0.491619</td>\n",
       "      <td>-0.0068818</td>\n",
       "      <td>-0.474008</td>\n",
       "      <td>-0.0118162</td>\n",
       "      <td>-0.35389</td>\n",
       "      <td>-0.132792</td>\n",
       "      <td>-0.0025423</td>\n",
       "      <td>-0.0470405</td>\n",
       "      <td>-0.410772</td>\n",
       "      <td>-0.00104491</td>\n",
       "      <td>-0.411851</td>\n",
       "      <td>-0.0692471</td>\n",
       "      <td>-0.0497225</td>\n",
       "      <td>-0.018334</td>\n",
       "      <td>-0.472379</td>\n",
       "      <td>-0.470219</td>\n",
       "      <td>-0.487608</td>\n",
       "      <td>-0.0019156</td>\n",
       "      <td>-0.0193245</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0         plasmid_1  \\\n",
       "GC_sum                                        -0.00413489         -0.045737   \n",
       "chain   [99, 87, 81, 101, 49, 60, 43, 104, 111, 65, 74...  [79, 24, 79, 62]   \n",
       "gd_sum                                          -0.441833          -0.43895   \n",
       "total                                           -0.445968         -0.484687   \n",
       "\n",
       "        plasmid_10           plasmid_11 plasmid_12  \\\n",
       "GC_sum  0.00838095          0.000116057  0.0113787   \n",
       "chain    [65, 104]  [27, 56, 2, 72, 41]       [56]   \n",
       "gd_sum        -0.5          -0.00699786  -0.485387   \n",
       "total    -0.491619           -0.0068818  -0.474008   \n",
       "\n",
       "                           plasmid_13    plasmid_14        plasmid_15  \\\n",
       "GC_sum                    -0.00358467    0.00743855      -0.000892041   \n",
       "chain   [108, 29, 83, 92, 89, 35, 98]  [83, 92, 89]  [52, 77, 61, 55]   \n",
       "gd_sum                    -0.00823149     -0.361328           -0.1319   \n",
       "total                      -0.0118162      -0.35389         -0.132792   \n",
       "\n",
       "              plasmid_16     plasmid_17 plasmid_18              plasmid_19  \\\n",
       "GC_sum        0.00275658      0.0234336  0.0201804             1.10986e-05   \n",
       "chain   [58, 57, 55, 10]  [55, 34, 103]       [79]  [130, 13, 73, 82, 115]   \n",
       "gd_sum       -0.00529888     -0.0704741  -0.430952             -0.00105601   \n",
       "total         -0.0025423     -0.0470405  -0.410772             -0.00104491   \n",
       "\n",
       "        plasmid_2 plasmid_20  \\\n",
       "GC_sum  0.0283625  -0.054933   \n",
       "chain        [31]   [42, 73]   \n",
       "gd_sum  -0.440214 -0.0143141   \n",
       "total   -0.411851 -0.0692471   \n",
       "\n",
       "                                                plasmid_3  \\\n",
       "GC_sum                                         -0.0051549   \n",
       "chain   [67, 18, 83, 92, 89, 32, 102, 81, 90, 40, 101,...   \n",
       "gd_sum                                         -0.0445676   \n",
       "total                                          -0.0497225   \n",
       "\n",
       "                                                plasmid_4  plasmid_5  \\\n",
       "GC_sum                                        -0.00121919  0.0276208   \n",
       "chain   [72, 19, 56, 15, 77, 51, 56, 46, 53, 76, 36, 1...   [81, 90]   \n",
       "gd_sum                                         -0.0171148       -0.5   \n",
       "total                                           -0.018334  -0.472379   \n",
       "\n",
       "       plasmid_6  plasmid_7                              plasmid_8  \\\n",
       "GC_sum  0.029781  0.0123917                             0.00107272   \n",
       "chain       [90]       [71]  [88, 125, 115, 82, 73, 45, 60, 4, 97]   \n",
       "gd_sum      -0.5       -0.5                            -0.00298832   \n",
       "total  -0.470219  -0.487608                             -0.0019156   \n",
       "\n",
       "                                    plasmid_9  \n",
       "GC_sum                            -0.00185382  \n",
       "chain   [108, 14, 55, 33, 65, 104, 1, 68, 11]  \n",
       "gd_sum                             -0.0174707  \n",
       "total                              -0.0193245  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_1  from MILP completely contained in  plasmid_0  from greedy.\n",
      "\n",
      "plasmid_3  from MILP completely contained in  plasmid_0  from greedy.\n",
      "\n",
      "plasmid_4  from MILP completely contained in  plasmid_0  from greedy.\n",
      "\n",
      "plasmid_5  from MILP completely contained in  plasmid_0  from greedy.\n",
      "\n",
      "plasmid_5  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_6  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_4  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_5  from MILP.\n",
      "\n",
      "plasmid_12  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_12  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n",
      "plasmid_14  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_14  from greedy completely contained in  plasmid_5  from MILP.\n",
      "\n",
      "plasmid_14  from greedy completely contained in  plasmid_6  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
