{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 125 with $\\alpha_1 = 1$, $\\alpha_2 = 5$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 125\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016582.1</td>\n",
       "      <td>2094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016584.1</td>\n",
       "      <td>6647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016583.1</td>\n",
       "      <td>37697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP016585.1</td>\n",
       "      <td>99473</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP016582.1         2094\n",
       "1  CP016584.1         6647\n",
       "2  CP016583.1        37697\n",
       "3  CP016585.1        99473"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt0VeWd//HPNwlXwSgQwlWSKgmE\nQIBE0IASFAWmCnbhhbTaUaF0tIxabJWZadVaW6sDtYMjFey4lEK5eaFUGYF2oHitBAWRSwCRCAIC\nAuEOuXx/f2SHXwwJHDQ7J8D7tRbLs/d+zrO/5/Q8zfns/eyzzd0FAAAAADHRLgAAAABA3UA4AAAA\nACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgEBftAk5X\nixYtPCkpKdplnDEOHjyo8847L9plAHUeYwWIDGPl9CxbtmyXuydEuw4gUmdcOEhKSlJeXl60yzhj\nLF68WDk5OdEuA6jzGCtAZBgrp8fMCqJdA3A6mFYEAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBw\nAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAEAgtHBgZs+b2Q4z+7ia7WZmE8xsg5l9ZGY9w6oFAAAA\nwKmFeebgBUmDTrJ9sKSOwb9Rkn4fYi0AAAAATiG0cODuSyTtPkmToZKmeJn3JF1gZq3DqgcAAADA\nycVFcd9tJW2usLwlWLetckMzG6WyswtKTEzU4sWLQy1s5eeFofZfm5LjY0N/v3DuYqwAkWGsADhT\nRDMcWBXrvKqG7j5Z0mRJysrK8pycnBDLkm4f+3qo/demFwadp7DfL5y7GCtAZBgrAM4U0QwHWyS1\nr7DcTtLWKNUC4Fy3bbn0yNBoV1EzHjl7jlIDAGpXNH/KdK6k7we/WnSZpEJ3P2FKEQAAAIDaEdqZ\nAzObLilHUgsz2yLpYUn1JMndn5U0T9I/Sdog6ZCkO8KqBQAAAMCphRYO3D33FNtd0o/C2j8AAACA\n08MdkgEAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAA\nCBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAI\nEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQ\nDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAO\nAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4A\nAAAASCIcAAAAAAiEGg7MbJCZ5ZvZBjMbW8X2i8xskZl9aGYfmdk/hVkPAAAAgOqFFg7MLFbSM5IG\nS0qTlGtmaZWa/UzSLHfvIWm4pIlh1QMAAADg5MI8c9BL0gZ33+juxyTNkDS0UhuXdH7wOF7S1hDr\nAQAAAHAScSH23VbS5grLWyT1rtTmEUkLzOxfJZ0naUCI9QAAgG9q23LpkcrH+s5QjxRGuwKgzgkz\nHFgV67zScq6kF9x9vJldLumPZpbu7qVf6chslKRRkpSYmKjFixeHUe9x93ctDrX/2nTgwIHQ3y+c\nu86qsdKgjRan/iLaZdQMxnydw1ipoxgrwAnCDAdbJLWvsNxOJ04bGiFpkCS5+7tm1lBSC0k7KjZy\n98mSJktSVlaW5+TkhFRymdvHvh5q/7XphUHnKez3C+eus2qsZHyqnPyHo11GzcjlaGhdw1ipoxgr\nwAnCvOZgqaSOZpZsZvVVdsHx3EptPpN0tSSZWWdJDSXtDLEmAAAAANUILRy4e7Gk0ZLmS1qjsl8l\nWmVmj5rZkKDZ/ZJ+YGYrJE2XdLu7V556BAAAAKAWhDmtSO4+T9K8SuseqvB4taQ+YdYAAAAAIDLc\nIRkAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADh\nAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEA\nAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAA\nAACAJMIBAAAAgADhAAAAAIAkKS7aBSBk25ZLjwyNdhU145HCaFcAAABwVuPMAQAAAABJhAMAAAAA\nAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAAB\nwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEkhhwMzG2Rm+Wa2wczGVtPmZjNbbWar\nzOxPYdYDAAAAoHpxYXVsZrGSnpF0jaQtkpaa2Vx3X12hTUdJ/yapj7vvMbOWYdUDAAAA4OTCPHPQ\nS9IGd9/o7sckzZA0tFKbH0h6xt33SJK77wixHgAAAAAnYe4eTsdmN0oa5O4jg+XbJPV299EV2syR\ntE5SH0mxkh5x9zeq6GuUpFGSlJiYmDljxoxQai638vPCUPuvTcmNjqrJ0a3RLqNmtO4e7QpQCWOl\njmKs1DmMlTqqFsZK//79l7l7Vug7AmpIaNOKJFkV6yonkThJHSXlSGon6U0zS3f3vV95kvtkSZMl\nKSsry3Nycmq82IpuH/t6qP3XphcyPlVO/sPRLqNm5J49f1zPFoyVOoqxUucwVuooxgpwgjCnFW2R\n1L7CcjtJlQ81bJH0Z3cvcvdPJeWrLCwAAAAAqGVhhoOlkjqaWbKZ1Zc0XNLcSm3mSOovSWbWQlKK\npI0h1gQAAACgGqGFA3cvljRa0nxJayTNcvdVZvaomQ0Jms2X9KWZrZa0SNJP3f3LsGoCAAAAUL0w\nrzmQu8+TNK/SuocqPHZJY4J/AAAAAKKIOyQDAAAAkEQ4AAAAABAgHAAAAACQdBrhwMz6mtkdweME\nM0sOrywAAAAAtS2icGBmD0t6UNK/BavqSZoaVlEAAAAAal+kZw6+I2mIpIOS5O5bJTUNqygAAAAA\ntS/ScHAs+NlRlyQzOy+8kgAAAABEQ6ThYJaZTZJ0gZn9QNJfJT0XXlkAAAAAaltEN0Fz93Fmdo2k\nfZJSJT3k7gtDrQwAAABArTplODCzWEnz3X2AJAIBAAAAcJY65bQidy+RdMjM4muhHgAAAABREtG0\nIklHJK00s4UKfrFIktz9nlCqAgAAAFDrIg0Hrwf/AAAAAJylIr0g+UUzqy8pJViV7+5F4ZUFAAAA\noLZFFA7MLEfSi5I2STJJ7c3sn919SXilAQAAAKhNkU4rGi/pWnfPlyQzS5E0XVJmWIUBAAAAqF2R\n3gStXnkwkCR3XyepXjglAQAAAIiGSM8c5JnZ/0j6Y7D8PUnLwikJAAAAQDREGg7ukvQjSfeo7JqD\nJZImhlUUAAAAgNoXaTiIk/Rf7v5b6fhdkxuEVhUAAMBZatmyZS3j4uL+ICldkU/xBmpCqaSPi4uL\nR2ZmZu6oqkGk4eBvkgZIOhAsN5K0QFL2Ny4RAADgHBIXF/eHVq1adU5ISNgTExPj0a4H547S0lLb\nuXNn2vbt2/8gaUhVbSJNqw3dvTwYKHjcuAZqBAAAONekJyQk7CMYoLbFxMR4QkJCocrOWlXdJsK+\nDppZz/IFM8uSdPgb1gcAAHAuiiEYIFqCz161GSDSaUX3SpptZlsluaQ2km755uUBAAAAqCsiDQfJ\nknpIukjSdyRdprKQAAAAgG8gaezrNXpT2U2/+Xad+bn5Xr16pY4bN27zlVdeeSiS9qWlpcrOzk6Z\nN2/ehmbNmpWGXd839dprrzUdP3584qJFizaE0f+0adPiV61a1ejXv/719srbGjdu3OPQoUMfVvfc\n7OzslD//+c+fJCQklJzOPiOdVvRzd98n6QJJ10iaLOn3p7MjAAAAnDmKiopqfZ+zZs2K79Kly+Ez\nIRjUhu9973uFVQWDSOTm5n45bty4hNN9XqThoDxxfFvSs+7+Z0n1T3dnAAAAiL6f/vSnrZOTk7tk\nZ2d3vP7665MfeuihRKnsSP/o0aPbXnrppamPPfZY4tatW+MGDhx4cXp6euf09PTOCxYsOE+S9u3b\nF3PTTTclpaend+7cuXPa1KlTL5CkAwcO2HXXXfetlJSUtG9/+9vfOnLkiEnSU0891WLEiBHty/c/\nfvz4FiNHjmxXua5p06Y1+853vrNXkvLz8+snJyd3ueWWWzp07Nixy5AhQ5LnzJnTtGfPnp06dOiQ\nvmjRosYnqyU/P79+ZmZmalpaWue0tLTOCxcuPE8qO9rfq1ev1EGDBn0rOTm5y5AhQ5JLS0+eRYYN\nG5b03e9+96LMzMzUpKSk9OnTp8dXbrNo0aLGPXr06NS5c+e0Hj16dFqxYkUDScrLy2vYtWvXzp06\ndUpLSUlJW7lyZYNIX9uECROaf//7379IktauXVu/e/fundLT0zvfe++9bcr3W1BQUC8rKyu1U6dO\naR07duzyxhtvNJGk4cOH733llVeaR/iROC7SaUWfm9kklf2c6RNm1kD8Li8AAMAZZ8mSJY3/8pe/\nXLhy5crVRUVF1r1797QePXocn/azd+/e2KVLl+ZL0vXXX588ZsyYLwYOHHhg/fr19QcOHNhx48aN\nq/793/+9df/+/ffNnj17065du2KzsrI6DxkyZN9vf/vbhEaNGpWuW7du9T/+8Y9Gffr0SZOkESNG\n7O7SpUva0aNHtzRo0MCnTp3aYtKkSQWVa1u2bFmTPn36HF+/efPmhjNnztyYmZlZ0K1bt87Tpk1r\nnpeXt/ZPf/rTBb/61a9a9+/f/5PqamnTpk3xm2++ua5x48a+cuXKBrm5ud/6+OOP10jSmjVrGi1f\nvnxjUlJSUWZmZqeFCxc2GThw4IHK9VS0efPmBu+//37+6tWrGwwYMCB16NChKytuz8jIOPL++++v\nrVevnubMmdP0gQceaDd//vxPnn766YS77777i7vuumv3kSNHrLi4WJ9//nm9SF5bxf7vvvvui0aO\nHLlz9OjRXz7++OPHzwg8//zzza6++urCJ554YntxcbH2798fI0kJCQklx44ds+3bt8e2atUq4qlF\nkYaDmyUNkjTO3feaWWtJP410JwAAAKgbFi9e3GTw4MF7mzRp4pL8mmuu2Vtxe25u7u7yx2+//fb5\n69evb1S+fODAgdg9e/bELF68+Pz58+dfMGHChFaSdPToUduwYUP9t956q8k999yzQ5J69+59OCUl\n5ZAknX/++aV9+vTZP3PmzPiuXbseKSoqsl69ep3wy5eFhYVxF1544fHD+G3btj1a3i4lJeXwVVdd\ntS8mJkY9e/Y89Nhjj7UJXk+VtXTo0KFoxIgRHVavXt0oJiZGBQUFx2/g27Vr14MXX3xxkSR16dLl\n0CeffHLKGTHDhg3bHRsbq65dux5t37790eXLlzesuH337t2xt9xyS/KmTZsampkXFRWZJF1++eUH\nx40b13rLli31hw8fvqdr165HI31tFX3wwQdN/vd///cTSfrhD3/45S9/+ct2knTZZZcd/OEPf5hU\nVFQUc+ONN+7Jzs4+/r42b968+LPPPqvfqlWriH9lNKJw4O6HJL1SYXmbpG2R7gQAAAB1g/vJf1Om\nadOmx7+cu7vy8vLWBEHiK3289NJLGzIyMo5Wfr6ZVdnvqFGjdv3qV79qlZKScuTWW2/dVVWb2NhY\nLykpUWxsrCSpfv36x/cbExOjhg0betBOJSUldrJaxowZ06Zly5ZFL7/88qelpaVq1KjR8Qu/GzRo\ncLzf2NhYFRcXV130SV5X5eUHH3ywbb9+/fYvXLjwk/z8/PpXXXVVqiT9y7/8y+4rrrji4Kuvvho/\nePDglIkTJ25KTU09Gslrq6yqn8AdPHjwgSVLluS//PLL8bfffnvyPffc88Xo0aO/lMqCUuPGjU/r\n+g2mBgEAAJxDcnJyDsyfPz/+0KFDVlhYGPPXv/71gura9u3bd98TTzzRsnz5nXfeaSRJ/fv33zd+\n/PjE8rn6b7/9dqOg/YGpU6c2k6SlS5c2XLdu3fGb5l511VUHt23bVv/VV19tPmLEiN2qQnJy8pE1\na9Y0qGpbdaqrpbCwMLZ169ZFsbGxmjhxYvOSklPPrPnRj37UdsqUKVW+H6+88sqFJSUlWrVqVYPN\nmzc3yMjIOFJx+759+2LbtWt3TJImTZrUonz96tWr63fu3Pnoz372sx3XXnvt3uXLlzeq3Hckevbs\neeC5555rJknPPffc8WsJ1q1bV79t27ZF999//65bb7111wcffNBYKvvlp507d9ZLTU09IcCdTKTT\nigAAABCC2v7p0X79+h0aNGhQYVpaWpe2bdse7dat28H4+PgqvzlPnjx588iRIy9KSUlJKykpsd69\ne+/Pzs7+7De/+c3WUaNGXdSpU6c0d7d27dodXbRo0Yaf/OQnO4YPH56ckpKS1qVLl0Ndu3Y9WLG/\nG264Yc9HH33UuLqf17z22msLFyxY0DQ9PT3iL7TV1XLfffftGDZs2MVz5sy5sG/fvvsbNWp0yiPo\nq1evblR+QXRll1xyydFevXqlfvnll/V+97vfFTRu3PgrR/EffPDB7SNHjkyeMGFCqyuuuGJf+fo/\n/vGPzWbPnt08Li7OExISih5//PGte/fujY309ZWbOHHiZ8OHD//WxIkTE4cMGbKnfP38+fObTpgw\noVVcXJw3bty4ZNq0aZ9K0ltvvdW4R48eB+vVq3da+7FTnVqqa7KysjwvLy/UfSSNfT3U/mvTCxnr\nlZP/cLTLqBmPFEa7AlTCWKmjGCt1DmOljqqFsWJmy9w9q+K6FStWbMrIyKhyWk1tKSwsjImPjy/d\nv39/zOWXX5767LPPFvTt2zeiexF8E/3797/kvvvu+2Lo0KH7q9peUFBQLzc3N+mdd95ZH3YtVenb\nt2/Ht95664R9Dxs2LOm6664rvOOOO/ZU9by66I477mh/ww037K3qvV6xYkWLjIyMpKqex7QiAACA\nc8ytt97aoVOnTmndunXrfP311+8JOxjs2rUrNikpKb1hw4al1QUDSerQoUPRnXfeuWv37t1R+Y5a\nVTA4U6Wnpx8+2XtdHaYVAQAAnGP+8pe/fFqb+2vRokXJpk2bPo6k7ciRI+vc0fmXX355U7RrOF33\n33//1zo7xZkDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAJckAwAABBNj8RnnrrR6fRXeMr7JphZ\n5tChQ3fPmTPnU0kqKipSy5YtM7p3735w0aJFGyZMmNA8Ly/vvClTpnw2ZsyYNk2aNCl59NFHv6jY\nR2xsbGbHjh0Pl5SU2CWXXHJ41qxZmyreXVkquxFXdnZ2yrx58zY0a9as9Kabbkr629/+Ft+8efPi\n9evXr6qqttdee61pbm7uxW3btj0mSdddd92ecePGbavc7te//nXCs88+m7h58+YGW7duXdG6deti\nSZo+fXr8+++/f95TTz21NfI3DeU4cwAAAHCOadSoUWl+fn6jAwcOmCS9+uqr5ycmJhadTh8NGjQo\nXbt27er169evqlevno8fPz6hcptZs2bFd+nS5XCzZs1KJenOO+/cNXfu3FP+XGhWVtaBtWvXrl67\ndu3qqoKBJPXr1+/AwoUL17Vp0+ZYxfW33HJL4RtvvHHB/v37+Z77NfCmAQAAnIOuvvrqwtmzZ18g\nSdOnT282bNiw3V+3r759+x7YsGFDg8rrp02b1qziHYcHDx58ICEhofjr7qeiPn36HE5NTT1WeX1M\nTIyys7P3z5w5M74m9nOuIRwAAACcg2677bbdM2fOvPDQoUO2Zs2axpdffvnBr9NPUVGR5s+ff37X\nrl0PV962bNmyJn369Dllv08++WTCk08+efzMw4cfftgkNTU17corr+yYl5fXsHx9v379Ltm0aVO9\nU/WXlZV18M0332xyOq8DZUK95sDMBkn6L0mxkv7g7r+ppt2NkmZLutTd88KsCQAAAFLv3r0Pb9my\npcFzzz3XbMCAAYWn+/yjR4/GdOrUKS3oa/+99957wk23CgsL4y688MLSE5/9VQ888MDO8sfZ2dkH\nCwoKPoqPjy+dOXNm/LBhwy4pKCj4WJL+/ve/b4iktlatWhVv3769fuSvBuVCCwdmFivpGUnXSNoi\naamZzXX31ZXaNZV0j6R/hFULAAAATjRo0KC9Dz/8cPsFCxbk79ix47S+F5Zfc3CyNrGxsV5SUqLY\n2NiI+y2/PkEqu35gzJgxF23bti2u/ILjSBw+fNgaNmx4ylCCE4U5raiXpA3uvtHdj0maIWloFe1+\nKelJSUdCrAUAAACV3HXXXbvuv//+rb169TphSlBNSE5OPrJmzZoTrkU4mc8++yyutLTse/2iRYsa\nl5aWKjEx8bSuU8jPz2/YpUuXUF7T2S7MaUVtJW2usLxFUu+KDcysh6T27v6amf2kuo7MbJSkUZKU\nmJioxYsX13y1FdzftUauk6kTDjRoo8Wpv4h2GTUj5P/dcfoYK3UUY6XOYazUUXVlrETw06Nhufji\ni4t+/vOf7zhVu6eeeqr1pEmTEsuXv/jii48i6f/aa68tXLBgQdP09PSjknT99dcnv/fee0337NkT\nl5iY2G3s2LFbf/zjH+8qv97ggQce2Dl16tQLn3/++ZaxsbHesGHD0ilTpmyMiSk7nt2vX79LXnzx\nxYKkpKSixx57rOXTTz/d6ssvv6yXkZGR1r9//8KZM2cWSNKSJUuaPvHEE59/jbfknGfuHk7HZjdJ\nGujuI4Pl2yT1cvd/DZZjJP2fpNvdfZOZLZb0k1Ndc5CVleV5eeFelpA09vVQ+69NL2SsV07+w9Eu\no2Y8ctrTIREyxkodxVipcxgrdVQtjBUzW+buWRXXrVixYlNGRsYJ8/PPRgUFBfVyc3OT3nnnnVP+\nfGlN2bx5c9zNN9/8rXfffXddbe3zTLNixYoWGRkZSVVtC3Na0RZJ7Ssst5NU8WYUTSWlS1psZpsk\nXSZprpl9ZQABAADgzNShQ4eiO++8c9fu3btr7RcyN27cWH/8+PGbT90SVQlzWtFSSR3NLFnS55KG\nS/pu+UZ3L5TUonw50jMHAAAAOHOMHDlyT23ur1+/fodqc39nm9BSnLsXSxotab6kNZJmufsqM3vU\nzIaEtV8AAIA6rrS0tNSiXQTOTcFnr9pfcgr1PgfuPk/SvErrHqqmbU6YtQAAANQRH+/cuTMtISGh\nMCYmJpyLP4EqlJaW2s6dO+MlfVxdm1DDAQAAAL6quLh45Pbt2/+wffv2dIV7/SdQWamkj4uLi0dW\n14BwAAAAUIsyMzN3SGKKNeok0ioAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAA\nAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAA\nAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAA\nAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAAB\nwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHC\nAQAAAABJhAMAAAAAAcIBAAAAAEkhhwMzG2Rm+Wa2wczGVrF9jJmtNrOPzOxvZtYhzHoAAAAAVC+0\ncGBmsZKekTRYUpqkXDNLq9TsQ0lZ7t5N0kuSngyrHgAAAAAnF+aZg16SNrj7Rnc/JmmGpKEVG7j7\nInc/FCy+J6ldiPUAAAAAOAlz93A6NrtR0iB3Hxks3yapt7uPrqb9f0va7u6PVbFtlKRRkpSYmJg5\nY8aMUGout/LzwlD7r03JjY6qydGt0S6jZrTuHu0KUAljpY5irNQ5jJU6qhbGSv/+/Ze5e1boOwJq\nSFyIfVsV66pMImZ2q6QsSf2q2u7ukyVNlqSsrCzPycmpoRKrdvvY10Ptvza9kPGpcvIfjnYZNSP3\n7PnjerZgrNRRjJU6h7FSRzFWgBOEGQ62SGpfYbmdpBMONZjZAEn/Iamfux8NsR4AAAAAJxHmNQdL\nJXU0s2Qzqy9puKS5FRuYWQ9JkyQNcfcdIdYCAAAA4BRCCwfuXixptKT5ktZImuXuq8zsUTMbEjT7\nT0lNJM02s+VmNrea7gAAAACELMxpRXL3eZLmVVr3UIXHA8LcPwAAAIDIcYdkAAAAAJIIBwAAAAAC\nhAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKE\nAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQD\nAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMA\nAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAA\nAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAqGGAzMb\nZGb5ZrbBzMZWsb2Bmc0Mtv/DzJLCrAcAAABA9UILB2YWK+kZSYMlpUnKNbO0Ss1GSNrj7pdIekrS\nE2HVAwAAAODkwjxz0EvSBnff6O7HJM2QNLRSm6GSXgwevyTpajOzEGsCAAAAUI0ww0FbSZsrLG8J\n1lXZxt2LJRVKah5iTQAAAACqERdi31WdAfCv0UZmNkrSqGDxgJnlf8Pazhn9pRaSdkW7jhrxC04q\nITyMFSAyjJXT1qE2dgLUlDDDwRZJ7Ssst5O0tZo2W8wsTlK8pN2VO3L3yZImh1TnWc3M8tw9K9p1\nAHUdYwWIDGMFOLuFOa1oqaRoxXl+AAAFLUlEQVSOZpZsZvUlDZc0t1KbuZL+OXh8o6T/c/cTzhwA\nAAAACF9oZw7cvdjMRkuaLylW0vPuvsrMHpWU5+5zJf2PpD+a2QaVnTEYHlY9AAAAAE7OOFB/djOz\nUcG0LAAnwVgBIsNYAc5uhAMAAAAAkkK+QzIAAACAMwfhAAAAAIAkwsE5z8ySzOzj4HGOmb0W7ZqA\naDCzd06xfZ6ZXVBb9QBnAzNbbGZZweNNZtYi2jUBOLkw73OAEJmZqeyakdJo1wLUNWYW6+4lp/Mc\nd88+xfZ/+mZVAWcf/hYBZx/OHJxBgqP8a8xsoqQPJN1mZu+a2QdmNtvMmgTtLjWzd8xshZm9b2ZN\ng+e+GbT9wMxO+kUIqKuCz/JaM3vRzD4ys5fMrHFwVPIhM3tL0k1mdrGZvWFmy4LPfqfg+Ylm9mow\nPlaUjwUzOxD8t7WZLTGz5Wb2sZldEaw/ftTTzMYE2z42s/sq1LXGzJ4zs1VmtsDMGkXlTQJOovLn\n18yeMLO7K2x/xMzuDx7/1MyWBmPtF8G6yn+L2pvZ780sL/js/yI6rwxATSAcnHlSJU2RdI2kEZIG\nuHtPSXmSxgQ3nJsp6V53z5A0QNJhSTskXRO0vUXShGgUD9SQVEmT3b2bpH2Syr/YHHH3vu4+Q2V3\nVf9Xd8+U9BNJE4M2EyT9PRgfPSWtqtT3dyXNd/fukjIkLa+40cwyJd0hqbekyyT9wMx6BJs7SnrG\n3btI2itpWE29YKAmVPX5lTRDZX8Xyt0sabaZXauyz3QvSd0lZZrZlUGbVElT3L2HuxdI+o/grsnd\nJPUzs2618oIA1DimFZ15Ctz9PTO7TlKapLfLzuqqvqR3VfZ/2Nvcfakkufs+STKz8yT9t5l1l1Qi\nKSUaxQM1ZLO7vx08nirpnuDxTEkKzqJlq+wLTvlzGgT/vUrS9yUpmHpUWKnvpZKeN7N6kua4+/JK\n2/tKetXdDwb7ekXSFSq74/unFdovk5T0DV4jEIbqPr8tzayNpARJe9z9MzO7R9K1kj4MnttEZWHh\nMwV/iyr0e7OZjVLZ94rWKvv79FFtvCAANYtwcOY5GPzXJC1099yKG4OjNVXdvOLHkr5Q2ZHQGElH\nwiwSCFnlz3j5cvn4iJG0Nzj6f3oduy8Jjo5+W2V3cP9Pd59SoYlV81RJOlrhcYkkphWhrqnu8/uS\npBsltVLZmYTyto+7+6SvdGCWpP8/1mRmySo7O3epu+8xsxckNazRqgHUGqYVnbnek9THzC6RpGDO\ndYqktZLamNmlwfqmZhYnKV5lZxRKJd0mKTZKdQM14SIzuzx4nCvprYobgzNmn5rZTVLZRZNmlhFs\n/puku4L1sWZ2fsXnmlkHSTvc/TlJ/6OyqUcVLZF0QzDmzpP0HUlv1txLA0JV3ed3hqThKgsILwVt\n50u6s8L1bG3NrGUVfZ6vsrBQaGaJkgaH/BoAhIhwcIZy952Sbpc03cw+UllY6OTux1Q2d/RpM1sh\naaHKjuBMlPTPZvaeyqYUHayyY+DMsEZln+ePJDWT9Psq2nxP0ohgHKySNDRYf6+k/ma2UmVTf7pU\nel6OpOVm9qHKrhn4r4ob3f0DSS9Iel/SPyT9wd0/FHAGqO7z6+6rJDWV9Lm7bwvaLpD0J0nvBuPl\npaBN5T5XqGzq0SpJz0t6u3IbAGcOc69qBgoA1E3BlIbX3D09yqUAAHDW4cwBAAAAAEmcOQAAAAAQ\n4MwBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJ0v8DFkrDNytc8kYAAAAASUVORK5C\nYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f29f9c3b8d0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:5:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>99471</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0        99471"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.676515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.992018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.804437</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.676515\n",
       "1  precision  0.992018\n",
       "2   f1_score  0.804437"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP016582.1</th>\n",
       "      <th>CP016583.1</th>\n",
       "      <th>CP016584.1</th>\n",
       "      <th>CP016585.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.99234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.99234</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP016582.1  CP016583.1  CP016584.1  CP016585.1\n",
       "0  plasmid_0         0.0         0.0         0.0     0.99234\n",
       "1      Total         0.0         0.0         0.0     0.99234"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016582.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016583.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016584.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP016585.1</td>\n",
       "      <td>0.992018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.992018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0\n",
       "0  CP016582.1   0.000000\n",
       "1  CP016583.1   0.000000\n",
       "2  CP016584.1   0.000000\n",
       "3  CP016585.1   0.992018\n",
       "4       Total   0.992018"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.992018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.992018"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016582.1</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016583.1</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016584.1</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP016585.1</td>\n",
       "      <td>0.99234</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP016582.1     0.00000\n",
       "1  CP016583.1     0.00000\n",
       "2  CP016584.1     0.00000\n",
       "3  CP016585.1     0.99234"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016585.1</td>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.99234</td>\n",
       "      <td>0.992018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction   recall  precision\n",
       "0  CP016585.1  plasmid_0  0.99234   0.992018"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 5, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.000902813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[31, 13, 28]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.744874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.745777</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           plasmid_0\n",
       "GC_sum  -0.000902813\n",
       "chain   [31, 13, 28]\n",
       "gd_sum     -0.744874\n",
       "total      -0.745777"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0181409</td>\n",
       "      <td>-0.00580319</td>\n",
       "      <td>0.00548066</td>\n",
       "      <td>-0.00095548</td>\n",
       "      <td>-0.000442171</td>\n",
       "      <td>-0.00074848</td>\n",
       "      <td>0.000958468</td>\n",
       "      <td>-0.000603517</td>\n",
       "      <td>-0.00992073</td>\n",
       "      <td>-0.00934865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[17]</td>\n",
       "      <td>[21]</td>\n",
       "      <td>[24]</td>\n",
       "      <td>[31, 81, 13, 31, 13, 28, 31, 80, 28]</td>\n",
       "      <td>[75, 6, 32, 69, 20]</td>\n",
       "      <td>[69, 5, 32, 69, 9]</td>\n",
       "      <td>[43, 11, 32, 69]</td>\n",
       "      <td>[70, 8, 32, 1]</td>\n",
       "      <td>[32]</td>\n",
       "      <td>[28]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.684803</td>\n",
       "      <td>-0.70859</td>\n",
       "      <td>-0.393983</td>\n",
       "      <td>-0.743462</td>\n",
       "      <td>-0.00127934</td>\n",
       "      <td>-0.000763168</td>\n",
       "      <td>-0.00237317</td>\n",
       "      <td>-0.000305021</td>\n",
       "      <td>-0.572605</td>\n",
       "      <td>-0.400569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.702943</td>\n",
       "      <td>-0.714394</td>\n",
       "      <td>-0.388502</td>\n",
       "      <td>-0.744418</td>\n",
       "      <td>-0.00172151</td>\n",
       "      <td>-0.00151165</td>\n",
       "      <td>-0.0014147</td>\n",
       "      <td>-0.000908538</td>\n",
       "      <td>-0.582526</td>\n",
       "      <td>-0.409918</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        plasmid_0   plasmid_1   plasmid_2  \\\n",
       "GC_sum -0.0181409 -0.00580319  0.00548066   \n",
       "chain        [17]        [21]        [24]   \n",
       "gd_sum  -0.684803    -0.70859   -0.393983   \n",
       "total   -0.702943   -0.714394   -0.388502   \n",
       "\n",
       "                                   plasmid_3            plasmid_4  \\\n",
       "GC_sum                           -0.00095548         -0.000442171   \n",
       "chain   [31, 81, 13, 31, 13, 28, 31, 80, 28]  [75, 6, 32, 69, 20]   \n",
       "gd_sum                             -0.743462          -0.00127934   \n",
       "total                              -0.744418          -0.00172151   \n",
       "\n",
       "                 plasmid_5         plasmid_6       plasmid_7   plasmid_8  \\\n",
       "GC_sum         -0.00074848       0.000958468    -0.000603517 -0.00992073   \n",
       "chain   [69, 5, 32, 69, 9]  [43, 11, 32, 69]  [70, 8, 32, 1]        [32]   \n",
       "gd_sum        -0.000763168       -0.00237317    -0.000305021   -0.572605   \n",
       "total          -0.00151165        -0.0014147    -0.000908538   -0.582526   \n",
       "\n",
       "         plasmid_9  \n",
       "GC_sum -0.00934865  \n",
       "chain         [28]  \n",
       "gd_sum   -0.400569  \n",
       "total    -0.409918  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_0  from MILP completely contained in  plasmid_3  from greedy.\n",
      "\n",
      "plasmid_9  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
