{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 23 with $\\alpha_1 = 1$, $\\alpha_2 = 1$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 23\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020836.1</td>\n",
       "      <td>37362</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP020836.1        37362"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xl4VPW9x/HPNwtLAFEghFUSlQBZ\nCJAICliCokAV0AcXaLUVoVgtVQsu3Htba61L9UqteKUC1seNIuBCUalAvaS4S1AQCAYiAkFE9kBY\ns/zuHznhjsMkDMrJkPB+PQ+Pc875ze98Z575mfnM+Z1zzDknAAAAAIiKdAEAAAAATg2EAwAAAACS\nCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPDGRLuBEtWjR\nwiUmJka6jFpj//79atSoUaTLAE55jBUgPIyVE7Ns2bIdzrn4SNcBhKvWhYPExETl5uZGuoxaIycn\nR9nZ2ZEuAzjlMVaA8DBWToyZbYx0DcCJYFoRAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAA\nAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAe38KBmT1rZtvMbFUV283MJptZgZl9bmY9/KoFAAAAwPH5\neeTgOUmDqtk+WFJH799YSX/1sRYAAAAAx+FbOHDOLZG0q5omwyS94Cp8JOlMM2vtVz0AAAAAqhcT\nwX23lVQYsLzZW/dNcEMzG6uKowtKSEhQTk6Or4Wt/LrI1/5rUlLTaN/fL6AuKC4uZqwAYWCsAHVb\nJMOBhVjnQjV0zk2TNE2SsrKyXHZ2to9lSTdOfMvX/mvSc4Maye/3C6gLcnJyGCtAGBgrQN0WyasV\nbZbUPmC5naQtEaoFAAAAOO1FMhzMk/Qz76pFF0gqcs4dM6UIAAAAQM3wbVqRmc2UlC2phZltlvR7\nSbGS5Jx7WtJ8ST+WVCDpgKRRftUCAAAA4Ph8CwfOuZHH2e4k/cqv/QMAAAA4MdwhGQAAAIAkwgEA\nAAAAD+EAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAAD+EAAAAAgCTCAQAA\nAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAA\nAA/hAAAAAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAA\nD+EAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP\n4QAAAACAJMIBAAAAAE9MpAsAAKCuS5z4VqRLOGmeG9Qo0iUA8BFHDgAAAABIIhwAAAAA8BAOAAAA\nAEgiHAAAAADwEA4AAAAASPI5HJjZIDPLN7MCM5sYYvvZZrbYzD4zs8/N7Md+1gMAAACgar6FAzOL\nlvSUpMGSUiSNNLOUoGa/lTTbOddd0ghJU/yqBwAAAED1/Dxy0FNSgXNuvXPuiKSXJQ0LauMkneE9\nbippi4/1AAAAAKiGnzdBayupMGB5s6ReQW3uk7TQzH4tqZGkAT7WAwAAAKAafoYDC7HOBS2PlPSc\nc26SmV0o6UUzS3POlX+nI7OxksZKUkJCgnJycvyo96gJ6aW+9l+TiouLfX+/gLqAsQI/8XcFQG3h\nZzjYLKl9wHI7HTttaLSkQZLknPvQzBpIaiFpW2Aj59w0SdMkKSsry2VnZ/tUcoUb69ht7v1+v4C6\nICcnh7EC3/B3BUBt4ec5B0sldTSzJDOrp4oTjucFtdkk6RJJMrMukhpI2u5jTQAAAACq4Fs4cM6V\nShonaYGkNaq4KtFqM7vfzIZ6zSZI+oWZrZA0U9KNzrngqUcAAAAAaoCf04rknJsvaX7QunsDHudJ\n6uNnDQAAAADCwx2SAQAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAA\nAEgiHAAAAADwEA4AAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAA\nSCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPAQDgAAAABI\nIhwAAAAA8BAOAAAAAEgiHAAAAADwxES6AAC1V+LEtyJdwknz3KBGkS4BAICI48gBAAAAAEmEAwAA\nAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAA\nAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB5fw4GZDTKzfDMrMLOJ\nVbS51szyzGy1mf3dz3oAAAAAVC3Gr47NLFrSU5IulbRZ0lIzm+ecywto01HSf0jq45zbbWYt/aoH\nAAAAQPX8PHLQU1KBc269c+6IpJclDQtq8wtJTznndkuSc26bj/UAAAAAqIZvRw4ktZVUGLC8WVKv\noDbJkmRm70uKlnSfc+7t4I7MbKyksZKUkJCgnJwcP+o9akJ6qa/916Ti4mLf3y+cvhgrQHgYKwBq\nCz/DgYVY50Lsv6OkbEntJL1rZmnOuT3feZJz0yRNk6SsrCyXnZ190osNdOPEt3ztvyY9N6iR/H6/\ncPpirADhYawAqC38nFa0WVL7gOV2kraEaPMP51yJc+4rSfmqCAsAAAAAapif4WCppI5mlmRm9SSN\nkDQvqM1cSf0lycxaqGKa0XofawIAAABQBd/CgXOuVNI4SQskrZE02zm32szuN7OhXrMFknaaWZ6k\nxZLucs7t9KsmAAAAAFXz85wDOefmS5oftO7egMdO0njvHwAAAIAI4g7JAAAAACQRDgAAAAB4CAcA\nAAAAJJ1AODCzvmY2ynscb2ZJ/pUFAAAAoKaFFQ7M7PeS7pH0H96qWEkv+VUUAAAAgJoX7pGDqyQN\nlbRfkpxzWyQ18asoAAAAADUv3HBwxLvsqJMkM2vkX0kAAAAAIiHccDDbzKZKOtPMfiHpX5Km+1cW\nAAAAgJoW1k3QnHOPmdmlkvZK6iTpXufcIl8rAwAAAFCjjhsOzCxa0gLn3ABJBAIAAACgjjrutCLn\nXJmkA2bWtAbqAQAAABAhYU0rknRI0kozWyTvikWS5Jy7zZeqAAAAANS4cMPBW94/AAAAAHVUuCck\nP29m9SQle6vynXMl/pUFAAAAoKaFFQ7MLFvS85I2SDJJ7c3s5865Jf6VBgAAAKAmhTutaJKky5xz\n+ZJkZsmSZkrK9KswAAAAADUr3JugxVYGA0lyzq2VFOtPSQAAAAAiIdwjB7lm9jdJL3rLP5W0zJ+S\nAAAAAERCuOHgFkm/knSbKs45WCJpil9FAQAAAKh54YaDGElPOOf+LB29a3J936oCAACoo5YtW9Yy\nJibmGUlpCn+KN3AylEtaVVpaOiYzM3NbqAbhhoN3JA2QVOwtN5S0UFLvH1wiAADAaSQmJuaZVq1a\ndYmPj98dFRXlIl0PTh/l5eW2ffv2lK1btz4jaWioNuGm1QbOucpgIO9x3EmoEQAA4HSTFh8fv5dg\ngJoWFRXl4uPji1Rx1Cp0mzD72m9mPSoXzCxL0sEfWB8AAMDpKIpggEjxPntVZoBwpxXdLmmOmW2R\n5CS1kXTdDy8PAAAAwKki3HCQJKm7pLMlXSXpAlWEBAAAAPwAiRPfOqk3ld3wp8tPmcvN9+zZs9Nj\njz1W+KMf/ehAOO3Ly8vVu3fv5Pnz5xc0a9as3O/6fqg333yzyaRJkxIWL15c4Ef/M2bMaLp69eqG\nDz300NbgbXFxcd0PHDjwWVXP7d27d/I//vGPL+Pj48tOZJ/hTiv6nXNur6QzJV0qaZqkv57IjgAA\nAFB7lJSU1Pg+Z8+e3TQ1NfVgbQgGNeGnP/1pUahgEI6RI0fufOyxx+JP9HnhhoPKxHG5pKedc/+Q\nVO9EdwYAAIDIu+uuu1onJSWl9u7du+OQIUOS7r333gSp4pf+cePGtT3//PM7PfDAAwlbtmyJGThw\n4LlpaWld0tLSuixcuLCRJO3duzfqmmuuSUxLS+vSpUuXlJdeeulMSSouLrYrrrjinOTk5JTLL7/8\nnEOHDpkkPf744y1Gjx7dvnL/kyZNajFmzJh2wXXNmDGj2VVXXbVHkvLz8+slJSWlXnfddR06duyY\nOnTo0KS5c+c26dGjR+cOHTqkLV68OK66WvLz8+tlZmZ2SklJ6ZKSktJl0aJFjaSKX/t79uzZadCg\nQeckJSWlDh06NKm8vPosMnz48MSf/OQnZ2dmZnZKTExMmzlzZtPgNosXL47r3r175y5duqR07969\n84oVK+pLUm5uboP09PQunTt3TklOTk5ZuXJl/XBf2+TJk5v/7Gc/O1uSvvjii3rdunXrnJaW1uX2\n229vU7nfjRs3xmZlZXXq3LlzSseOHVPffvvtxpI0YsSIPa+99lrzMD8SR4U7rehrM5uqisuZPmJm\n9cV1eQEAAGqdJUuWxL3xxhtnrVy5Mq+kpMS6deuW0r1796PTfvbs2RO9dOnSfEkaMmRI0vjx478d\nOHBg8bp16+oNHDiw4/r161f/53/+Z+v+/fvvnTNnzoYdO3ZEZ2VldRk6dOjeP//5z/ENGzYsX7t2\nbd7HH3/csE+fPimSNHr06F2pqakphw8f3ly/fn330ksvtZg6derG4NqWLVvWuE+fPkfXFxYWNpg1\na9b6zMzMjV27du0yY8aM5rm5uV/8/e9/P/PBBx9s3b9//y+rqqVNmzal77777tq4uDi3cuXK+iNH\njjxn1apVayRpzZo1DZcvX74+MTGxJDMzs/OiRYsaDxw4sDi4nkCFhYX1P/nkk/y8vLz6AwYM6DRs\n2LCVgdszMjIOffLJJ1/ExsZq7ty5Te6+++52CxYs+PLJJ5+Mv/XWW7+95ZZbdh06dMhKS0v19ddf\nx4bz2gL7v/XWW88eM2bM9nHjxu18+OGHjx4RePbZZ5tdcsklRY888sjW0tJS7du3L0qS4uPjy44c\nOWJbt26NbtWqVdhTi8INB9dKGiTpMefcHjNrLemucHcCAACAU0NOTk7jwYMH72ncuLGT5C699NI9\ngdtHjhy5q/Lx+++/f8a6desaVi4XFxdH7969OyonJ+eMBQsWnDl58uRWknT48GErKCio99577zW+\n7bbbtklSr169DiYnJx+QpDPOOKO8T58++2bNmtU0PT39UElJifXs2fOYK18WFRXFnHXWWUd/xm/b\ntu3hynbJyckHL7744r1RUVHq0aPHgQceeKCN93pC1tKhQ4eS0aNHd8jLy2sYFRWljRs3Hr2Bb3p6\n+v5zzz23RJJSU1MPfPnll8edETN8+PBd0dHRSk9PP9y+ffvDy5cvbxC4fdeuXdHXXXdd0oYNGxqY\nmSspKTFJuvDCC/c/9thjrTdv3lxvxIgRu9PT0w+H+9oCffrpp43/+c9/filJN998884//vGP7STp\nggsu2H/zzTcnlpSURF199dW7e/fuffR9bd68eemmTZvqtWrVKuyrjIYVDpxzByS9FrD8jaRvwt0J\nAAAATg3OVX9NmSZNmhz9cu6cU25u7hovSHynj1deeaUgIyPjcPDzzSxkv2PHjt3x4IMPtkpOTj50\n/fXX7wjVJjo62pWVlSk6OlqSVK9evaP7jYqKUoMGDZzXTmVlZVZdLePHj2/TsmXLkldfffWr8vJy\nNWzY8OiJ3/Xr1z/ab3R0tEpLS0MXXc3rCl6+55572vbr12/fokWLvszPz6938cUXd5KkX/7yl7su\nuuii/a+//nrTwYMHJ0+ZMmVDp06dDofz2oKFugTu4MGDi5csWZL/6quvNr3xxhuTbrvttm/HjRu3\nU6oISnFxcSd0/gZTgwAAAE4j2dnZxQsWLGh64MABKyoqivrXv/51ZlVt+/btu/eRRx5pWbn8wQcf\nNJSk/v377500aVJC5Vz9999/v6HXvvill15qJklLly5tsHbt2qM3zb344ov3f/PNN/Vef/315qNH\nj96lEJKSkg6tWbOmfqhtVamqlqKioujWrVuXREdHa8qUKc3Lyo4/s+ZXv/pV2xdeeCHk+/Haa6+d\nVVZWptWrV9cvLCysn5GRcShw+969e6PbtWt3RJKmTp3aonJ9Xl5evS5duhz+7W9/u+2yyy7bs3z5\n8obBfYejR48exdOnT28mSdOnTz96LsHatWvrtW3btmTChAk7rr/++h2ffvppnFRx5aft27fHdurU\n6ZgAV51wpxUBAADABzV96dF+/fodGDRoUFFKSkpq27ZtD3ft2nV/06ZNQ35znjZtWuGYMWPOTk5O\nTikrK7NevXrt692796Y//elPW8aOHXt2586dU5xz1q5du8OLFy8uuPPOO7eNGDEiKTk5OSU1NfVA\nenr6/sD+rrzyyt2ff/55XFWX17zsssuKFi5c2CQtLS3sL7RV1XLHHXdsGz58+Llz5849q2/fvvsa\nNmx43F/Q8/LyGlaeEB3svPPOO9yzZ89OO3fujP3LX/6yMS4u7ju/4t9zzz1bx4wZkzR58uRWF110\n0d7K9S+++GKzOXPmNI+JiXHx8fElDz/88JY9e/ZEh/v6Kk2ZMmXTiBEjzpkyZUrC0KFDd1euX7Bg\nQZPJkye3iomJcXFxcWUzZsz4SpLee++9uO7du++PjY09of3Y8Q4tnWqysrJcbm6ur/tInPiWr/3X\npOcGNVJ2dnaky0AdxVgBwsNYOX2Z2TLnXFbguhUrVmzIyMgIOa2mphQVFUU1bdq0fN++fVEXXnhh\np6effnpj3759w7oXwQ/Rv3//8+64445vhw0bti/U9o0bN8aOHDky8YMPPljndy2h9O3bt+N77713\nzL6HDx+eeMUVVxSNGjVqd6jnnYpGjRrV/sorr9wT6r1esWJFi4yMjMRQz2NaEQAAwGnm+uuv79C5\nc+eUrl27dhkyZMhuv4PBjh07ohMTE9MaNGhQXlUwkKQOHTqU3HTTTTt27doVke+ooYJBbZWWlnaw\nuve6KkwrAgAAOM288cYbX9Xk/lq0aFG2YcOGVeG0HTNmzCn36/yrr766IdI1nKgJEyZ8r6NTHDkA\nAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwMMJyQAAAJF0X9PM4zc6kf6KjnvfBDPLHDZs2K65c+d+\nJUklJSVq2bJlRrdu3fYvXry4YPLkyc1zc3MbvfDCC5vGjx/fpnHjxmX333//t4F9REdHZ3bs2PFg\nWVmZnXfeeQdnz569IfDuylLFjbh69+6dPH/+/IJmzZqVX3PNNYnvvPNO0+bNm5euW7dudajaPvvs\nswajRo1KzMvLi5s4ceLXwfut9NBDD8U//fTTCYWFhfW3bNmyonXr1qWSNHPmzKaffPJJo8cff3xL\nuG8Z/h9HDgAAAE4zDRs2LM/Pz29YXFxskvT666+fkZCQUHIifdSvX7/8iy++yFu3bt3q2NhYN2nS\npPjgNrNnz26ampp6sFmzZuWSdNNNN+2YN29etZcLbdmyZekTTzyx6eabbw4ZCir169eveNGiRWvb\ntGlzJHD9ddddV/T222+fuW/fPr7nfg+8aQAAAKehSy65pGjOnDlnStLMmTObDR8+fNf37atv377F\nBQUF9YPXz5gxo1ngHYcHDx5cHB8fX1pdX23bti3t16/fgdjY2Grv1NunT5+DnTp1OhK8PioqSr17\n9943a9aspifyGlCBcAAAAHAauuGGG3bNmjXrrAMHDtiaNWviLrzwwv3fp5+SkhItWLDgjPT09IPB\n25YtW9a4T58+x+330UcfjX/00UePOfIQrF+/fudt2LAh9njtsrKy9r/77ruNj9cOx/L1nAMzGyTp\nCUnRkp5xzv2pinZXS5oj6XznXK6fNQEAAEDq1avXwc2bN9efPn16swEDBhSd6PMPHz4c1blz5xSv\nr3233377MTfdKioqijnrrLPKj332d919993bw9nnv//974Jw2rVq1ap069at9cJpi+/yLRyYWbSk\npyRdKmmzpKVmNs85lxfUromk2yR97FctAAAAONagQYP2/P73v2+/cOHC/G3btp3Q98LKcw6qaxMd\nHe3KysoUHR39wwo9QQcPHrQGDRocN5TgWH5OK+opqcA5t945d0TSy5KGhWj3R0mPSjrkYy0AAAAI\ncsstt+yYMGHClp49ex4zJehkSEpKOrRmzZpjzkXwW35+foPU1FRfXlNd5+e0oraSCgOWN0vqFdjA\nzLpLau+ce9PM7qyqIzMbK2msJCUkJCgnJ+fkVxtgQnq158nUKsXFxb6/Xzh9MVaA8DBWUK0wLj3q\nl3PPPbfkd7/73bbjtXv88cdbT506NaFy+dtvv/08nP4vu+yyooULFzZJS0s7LElDhgxJ+uijj5rs\n3r07JiEhoevEiRO3/OY3v9lReb7B3XffvX3Tpk0x559/fsr+/fujzcxNnTo1Yc2aNauaNWtW3q9f\nv/Oef/75jYmJiSUPPPBAyyeffLLVzp07YzMyMlL69+9fNGvWrI2StGTJkiaPPPLI19/vXTm9mXPV\nngj+/Ts2u0bSQOfcGG/5Bkk9nXO/9pajJP2vpBudcxvMLEfSncc75yArK8vl5vp7WkLixLd87b8m\nPTeokbKzsyNdBuooxgoQHsbK6cvMljnnsgLXrVixYkNGRsYx8/Proo0bN8aOHDky8YMPPqj28qUn\nU2FhYcy11157zocffri2pvZZ26xYsaJFRkZGYqhtfk4r2iypfcByO0mBN6NoIilNUo6ZbZB0gaR5\nZvadAQQAAIDaqUOHDiU33XTTjl27dtXYFTLXr19fb9KkSYXHb4lQ/JxWtFRSRzNLkvS1pBGSflK5\n0TlXJKlF5XK4Rw4AAABQe4wZM2Z3Te6vX79+B2pyf3WNbynOOVcqaZykBZLWSJrtnFttZveb2VC/\n9gsAAHCKKy8vL7dIF4HTk/fZq/JKTr7e58A5N1/S/KB191bRNtvPWgAAAE4Rq7Zv354SHx9fFBUV\n5c/Jn0AI5eXltn379qaSVlXVxtdwAAAAgO8qLS0ds3Xr1me2bt2aJn/P/wSClUtaVVpaOqaqBoQD\nAACAGpSZmblNElOscUoirQIAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAh\nHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEc\nAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwA\nAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAA\nAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEcAAAA\nAJBEOAAAAADgIRwAAAAAkORzODCzQWaWb2YFZjYxxPbxZpZnZp+b2Ttm1sHPegAAAABUzbdwYGbR\nkp6SNFhSiqSRZpYS1OwzSVnOua6SXpH0qF/1AAAAAKien0cOekoqcM6td84dkfSypGGBDZxzi51z\nB7zFjyS187EeAAAAANWI8bHvtpIKA5Y3S+pVTfvRkv4ZaoOZjZU0VpISEhKUk5NzkkoMbUJ6qa/9\n16Ti4mLf3y+cvhgrQHgYKwBqCz/DgYVY50I2NLteUpakfqG2O+emSZomSVlZWS47O/sklRjajRPf\n8rX/mvTcoEby+/3C6YuxAoSHsQKgtvAzHGyW1D5guZ2kLcGNzGyApP+S1M85d9jHegAAAABUw89z\nDpZK6mhmSWZWT9IISfMCG5hZd0lTJQ11zm3zsRYAAAAAx+FbOHDOlUoaJ2mBpDWSZjvnVpvZ/WY2\n1Gv235IaS5pjZsvNbF4V3QEAAADwmZ/TiuScmy9pftC6ewMeD/Bz/wAAAADCxx2SAQAAAEgiHAAA\nAADwEA4AAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAA\nAPAQDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA\n8BAOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADw\nEA4AAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPAQ\nDgAAAABIIhwAAAAA8MREugD47Jvl0n3DIl3FyXFfUaQrAAAAqNM4cgAAAABAEuEAAAAAgIdwAAAA\nAEAS4QAAAACAh3AAAAAAQJLP4cDMBplZvpkVmNnEENvrm9ksb/vHZpboZz0AAAAAquZbODCzaElP\nSRosKUXSSDNLCWo2WtJu59x5kh6X9Ihf9QAAAAConp/3OegpqcA5t16SzOxlScMk5QW0GSbpPu/x\nK5L+x8zMOed8rAsAAHxf3D8HqNP8DAdtJRUGLG+W1KuqNs65UjMrktRc0g4f6wKAY/GFBwAAX8OB\nhVgXfEQgnDYys7GSxnqLxWaW/wNrO230l1qoroStP4T6uAAnB2MFCA9j5YR1qImdACeLn+Fgs6T2\nAcvtJG2pos1mM4uR1FTSruCOnHPTJE3zqc46zcxynXNZka4DONUxVoDwMFaAus3PqxUtldTRzJLM\nrJ6kEZLmBbWZJ+nn3uOrJf0v5xsAAAAAkeHbkQPvHIJxkhZIipb0rHNutZndLynXOTdP0t8kvWhm\nBao4YjDCr3oAAAAAVM/4ob5uM7Ox3rQsANVgrADhYawAdRvhAAAAAIAkn++QDAAAAKD2IBwAAAAA\nkEQ4OO2ZWaKZrfIeZ5vZm5GuCYgEM/vgONvnm9mZNVUPUBeYWY6ZZXmPN5hZi0jXBKB6ft7nAD4y\nM1PFOSPlka4FONWYWbRzruxEnuOc632c7T/+YVUBdQ9/i4C6hyMHtYj3K/8aM5si6VNJN5jZh2b2\nqZnNMbPGXrvzzewDM1thZp+8FOVbAAAEN0lEQVSYWRPvue96bT81s2q/CAGnKu+z/IWZPW9mn5vZ\nK2YW5/0qea+ZvSfpGjM718zeNrNl3me/s/f8BDN73RsfKyrHgpkVe/9tbWZLzGy5ma0ys4u89Ud/\n9TSz8d62VWZ2R0Bda8xsupmtNrOFZtYwIm8SUI3gz6+ZPWJmtwZsv8/MJniP7zKzpd5Y+4O3Lvhv\nUXsz+6uZ5Xqf/T9E5pUBOBkIB7VPJ0kvSLpU0mhJA5xzPSTlShrv3XBulqTbnXMZkgZIOihpm6RL\nvbbXSZocieKBk6STpGnOua6S9kqq/GJzyDnX1zn3siruqv5r51ympDslTfHaTJb0b2989JC0Oqjv\nn0ha4JzrJilD0vLAjWaWKWmUpF6SLpD0CzPr7m3uKOkp51yqpD2Shp+sFwycDKE+v5JeVsXfhUrX\nSppjZpep4jPdU1I3SZlm9iOvTSdJLzjnujvnNkr6L++uyV0l9TOzrjXyggCcdEwrqn02Ouc+MrMr\nJKVIer/iqK7qSfpQFf/D/sY5t1SSnHN7JcnMGkn6HzPrJqlMUnIkigdOkkLn3Pve45ck3eY9niVJ\n3lG03qr4glP5nPrefy+W9DNJ8qYeFQX1vVTSs2YWK2muc2550Pa+kl53zu339vWapItUccf3rwLa\nL5OU+ANeI+CHqj6/Lc2sjaR4Sbudc5vM7DZJl0n6zHtuY1WEhU3y/hYF9HutmY1VxfeK1qr4+/R5\nTbwgACcX4aD22e/91yQtcs6NDNzo/VoT6uYVv5H0rSp+CY2SdMjPIgGfBX/GK5crx0eUpD3er/8n\n1rFzS7xfRy9XxR3c/9s590JAE6viqZJ0OOBxmSSmFeFUU9Xn9xVJV0tqpYojCZVtH3bOTf1OB2aJ\n+v+xJjNLUsXRufOdc7vN7DlJDU5q1QBqDNOKaq+PJPUxs/MkyZtznSzpC0ltzOx8b30TM4uR1FQV\nRxTKJd0gKTpCdQMnw9lmdqH3eKSk9wI3ekfMvjKza6SKkybNLMPb/I6kW7z10WZ2RuBzzayDpG3O\nuemS/qaKqUeBlki60htzjSRdJendk/fSAF9V9fl9WdIIVQSEV7y2CyTdFHA+W1szaxmizzNUERaK\nzCxB0mCfXwMAHxEOainn3HZJN0qaaWafqyIsdHbOHVHF3NEnzWyFpEWq+AVniqSfm9lHqphStD9k\nx0DtsEYVn+fPJTWT9NcQbX4qabQ3DlZLGuatv11SfzNbqYqpP6lBz8uWtNzMPlPFOQNPBG50zn0q\n6TlJn0j6WNIzzrnPBNQCVX1+nXOrJTWR9LVz7huv7UJJf5f0oTdeXvHaBPe5QhVTj1ZLelbS+8Ft\nANQe5lyoGSgAcGrypjS86ZxLi3ApAADUORw5AAAAACCJIwcAAAAAPBw5AAAAACCJcAAAAADAQzgA\nAAAAIIlwAAAAAMBDOAAAAAAgSfo/expmtIydA90AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f8c60d9fa58>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:1:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>199780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>3463</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0       199780\n",
       "1  plasmid_1         3463"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.036668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.003439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.006289</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.036668\n",
       "1  precision  0.003439\n",
       "2   f1_score  0.006289"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP020836.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.036668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.036668</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP020836.1\n",
       "0  plasmid_0    0.036668\n",
       "1  plasmid_1    0.000000\n",
       "2      Total    0.036668"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020836.1</td>\n",
       "      <td>0.003499</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.003499</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1\n",
       "0  CP020836.1   0.003499        0.0\n",
       "1       Total   0.003499        0.0"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.003499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.003499\n",
       "1  plasmid_1    0.000000"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020836.1</td>\n",
       "      <td>0.036668</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP020836.1    0.036668"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No high mutual correspondence found.\n"
     ]
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 1, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>0.00152133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[158, 134, 68, 93, 153, 132, 116, 128, 146, 14...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.151249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.149728</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0\n",
       "GC_sum                                         0.00152133\n",
       "chain   [158, 134, 68, 93, 153, 132, 116, 128, 146, 14...\n",
       "gd_sum                                          -0.151249\n",
       "total                                           -0.149728"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_11</th>\n",
       "      <th>plasmid_12</th>\n",
       "      <th>plasmid_13</th>\n",
       "      <th>plasmid_14</th>\n",
       "      <th>plasmid_15</th>\n",
       "      <th>plasmid_16</th>\n",
       "      <th>plasmid_17</th>\n",
       "      <th>plasmid_18</th>\n",
       "      <th>plasmid_19</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_20</th>\n",
       "      <th>plasmid_21</th>\n",
       "      <th>plasmid_22</th>\n",
       "      <th>plasmid_23</th>\n",
       "      <th>plasmid_24</th>\n",
       "      <th>plasmid_25</th>\n",
       "      <th>plasmid_26</th>\n",
       "      <th>plasmid_27</th>\n",
       "      <th>plasmid_28</th>\n",
       "      <th>plasmid_29</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_30</th>\n",
       "      <th>plasmid_31</th>\n",
       "      <th>plasmid_32</th>\n",
       "      <th>plasmid_33</th>\n",
       "      <th>plasmid_34</th>\n",
       "      <th>plasmid_35</th>\n",
       "      <th>plasmid_36</th>\n",
       "      <th>plasmid_37</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0325364</td>\n",
       "      <td>0.00650064</td>\n",
       "      <td>0.0186439</td>\n",
       "      <td>0.000303181</td>\n",
       "      <td>0.0030474</td>\n",
       "      <td>0.00171596</td>\n",
       "      <td>0.00699621</td>\n",
       "      <td>-0.00531383</td>\n",
       "      <td>0.0013756</td>\n",
       "      <td>-0.0108338</td>\n",
       "      <td>-0.0195227</td>\n",
       "      <td>-0.031398</td>\n",
       "      <td>-0.00121611</td>\n",
       "      <td>0.0111853</td>\n",
       "      <td>0.00692474</td>\n",
       "      <td>-0.0107214</td>\n",
       "      <td>0.00700724</td>\n",
       "      <td>-0.0245555</td>\n",
       "      <td>0.00789096</td>\n",
       "      <td>0.00045135</td>\n",
       "      <td>0.000776025</td>\n",
       "      <td>-0.0100773</td>\n",
       "      <td>-0.00146033</td>\n",
       "      <td>-0.0174176</td>\n",
       "      <td>0.00340978</td>\n",
       "      <td>-0.00183801</td>\n",
       "      <td>-0.0366416</td>\n",
       "      <td>0.00374412</td>\n",
       "      <td>-0.00833435</td>\n",
       "      <td>0.0282795</td>\n",
       "      <td>0.00267696</td>\n",
       "      <td>0.00849572</td>\n",
       "      <td>0.0163942</td>\n",
       "      <td>-0.0028682</td>\n",
       "      <td>-0.00418644</td>\n",
       "      <td>-6.62224e-05</td>\n",
       "      <td>-0.00410727</td>\n",
       "      <td>-0.0628096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[46, 120, 187, 73, 191, 187]</td>\n",
       "      <td>[23, 109, 99, 104, 114, 210, 120, 187, 188, 57...</td>\n",
       "      <td>[134, 68, 90, 153, 131, 116, 128, 146, 143, 148]</td>\n",
       "      <td>[106, 194, 212, 208, 210, 120, 187, 188, 15, 1...</td>\n",
       "      <td>[6, 178, 154, 135, 176, 49, 178]</td>\n",
       "      <td>[178, 154, 135, 176, 52]</td>\n",
       "      <td>[178, 154, 135]</td>\n",
       "      <td>[172, 211, 188, 187, 120, 210, 208, 212, 193, ...</td>\n",
       "      <td>[194, 212, 208, 210, 120, 187, 189, 44, 102, 86]</td>\n",
       "      <td>[192, 74, 102, 38, 192]</td>\n",
       "      <td>[195, 115, 102, 64, 192]</td>\n",
       "      <td>[195, 56, 102, 82, 189, 187, 120, 210, 208]</td>\n",
       "      <td>[151, 84, 158, 133, 68, 93, 153, 132, 116, 127...</td>\n",
       "      <td>[212, 208, 210, 120, 186, 190, 69]</td>\n",
       "      <td>[212, 208, 210, 120, 186, 191, 34]</td>\n",
       "      <td>[193, 212, 208, 210, 120, 186, 189, 58, 189, 1...</td>\n",
       "      <td>[212, 208, 210, 120, 186, 191, 42]</td>\n",
       "      <td>[208, 210, 120, 186, 190, 53, 188, 187, 120]</td>\n",
       "      <td>[212, 208, 210, 120, 186, 191, 71]</td>\n",
       "      <td>[120, 187, 188, 213, 192, 212, 208, 210, 120, ...</td>\n",
       "      <td>[192, 212, 208, 210, 120, 186, 190, 2]</td>\n",
       "      <td>[193, 212, 208, 210, 120, 186, 189, 59]</td>\n",
       "      <td>[192, 212, 208, 210, 120, 186, 190, 67]</td>\n",
       "      <td>[151, 126, 169, 111, 97, 130, 161, 85]</td>\n",
       "      <td>[192, 212, 208, 210, 120, 186, 191, 4]</td>\n",
       "      <td>[192, 212, 208, 210, 120, 186, 191, 66]</td>\n",
       "      <td>[208, 210, 120, 186, 190, 81]</td>\n",
       "      <td>[192, 212, 208, 210, 120, 187, 189, 17, 100, 8...</td>\n",
       "      <td>[193, 212, 208, 210, 120, 186, 189, 7]</td>\n",
       "      <td>[210, 120]</td>\n",
       "      <td>[41, 100, 14, 155]</td>\n",
       "      <td>[5, 100, 61, 96]</td>\n",
       "      <td>[12, 210, 120, 187, 189, 45]</td>\n",
       "      <td>[162, 181, 144, 137, 32, 194, 212, 208, 210, 1...</td>\n",
       "      <td>[107, 63, 101, 119, 159, 172, 107, 1, 102, 3, ...</td>\n",
       "      <td>[194, 40, 154, 135, 27, 194, 212, 208, 210, 12...</td>\n",
       "      <td>[101, 24, 161, 129, 97, 110]</td>\n",
       "      <td>[181, 162, 95, 101, 119]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.419306</td>\n",
       "      <td>-0.0746612</td>\n",
       "      <td>-0.315989</td>\n",
       "      <td>-0.0239502</td>\n",
       "      <td>-0.00590275</td>\n",
       "      <td>-0.0108497</td>\n",
       "      <td>-0.385926</td>\n",
       "      <td>-0.016139</td>\n",
       "      <td>-0.0223755</td>\n",
       "      <td>-0.014406</td>\n",
       "      <td>-0.0431704</td>\n",
       "      <td>-0.03176</td>\n",
       "      <td>-0.067396</td>\n",
       "      <td>-0.0271974</td>\n",
       "      <td>-0.00583331</td>\n",
       "      <td>-0.0549323</td>\n",
       "      <td>-0.0067238</td>\n",
       "      <td>-0.0569558</td>\n",
       "      <td>-0.0311747</td>\n",
       "      <td>-0.0118767</td>\n",
       "      <td>-0.000935788</td>\n",
       "      <td>-0.0121487</td>\n",
       "      <td>-0.0216865</td>\n",
       "      <td>-0.0919549</td>\n",
       "      <td>-0.00107986</td>\n",
       "      <td>-0.0182759</td>\n",
       "      <td>-0.0489702</td>\n",
       "      <td>-0.00353387</td>\n",
       "      <td>-0.00163805</td>\n",
       "      <td>-0.360862</td>\n",
       "      <td>-0.00418812</td>\n",
       "      <td>-0.0030128</td>\n",
       "      <td>-0.0664248</td>\n",
       "      <td>-0.0849682</td>\n",
       "      <td>-0.00404221</td>\n",
       "      <td>-0.00867263</td>\n",
       "      <td>-0.0255533</td>\n",
       "      <td>-0.177263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.451843</td>\n",
       "      <td>-0.0681605</td>\n",
       "      <td>-0.297346</td>\n",
       "      <td>-0.023647</td>\n",
       "      <td>-0.00285535</td>\n",
       "      <td>-0.00913377</td>\n",
       "      <td>-0.37893</td>\n",
       "      <td>-0.0214529</td>\n",
       "      <td>-0.0209999</td>\n",
       "      <td>-0.0252398</td>\n",
       "      <td>-0.0626932</td>\n",
       "      <td>-0.063158</td>\n",
       "      <td>-0.0686121</td>\n",
       "      <td>-0.0160121</td>\n",
       "      <td>0.00109142</td>\n",
       "      <td>-0.0656538</td>\n",
       "      <td>0.000283434</td>\n",
       "      <td>-0.0815113</td>\n",
       "      <td>-0.0232838</td>\n",
       "      <td>-0.0114254</td>\n",
       "      <td>-0.000159763</td>\n",
       "      <td>-0.022226</td>\n",
       "      <td>-0.0231469</td>\n",
       "      <td>-0.109372</td>\n",
       "      <td>0.00232992</td>\n",
       "      <td>-0.0201139</td>\n",
       "      <td>-0.0856118</td>\n",
       "      <td>0.000210252</td>\n",
       "      <td>-0.0099724</td>\n",
       "      <td>-0.332582</td>\n",
       "      <td>-0.00151117</td>\n",
       "      <td>0.00548292</td>\n",
       "      <td>-0.0500306</td>\n",
       "      <td>-0.0878364</td>\n",
       "      <td>-0.00822866</td>\n",
       "      <td>-0.00873885</td>\n",
       "      <td>-0.0296605</td>\n",
       "      <td>-0.240073</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           plasmid_0  \\\n",
       "GC_sum                    -0.0325364   \n",
       "chain   [46, 120, 187, 73, 191, 187]   \n",
       "gd_sum                     -0.419306   \n",
       "total                      -0.451843   \n",
       "\n",
       "                                                plasmid_1  \\\n",
       "GC_sum                                         0.00650064   \n",
       "chain   [23, 109, 99, 104, 114, 210, 120, 187, 188, 57...   \n",
       "gd_sum                                         -0.0746612   \n",
       "total                                          -0.0681605   \n",
       "\n",
       "                                              plasmid_10  \\\n",
       "GC_sum                                         0.0186439   \n",
       "chain   [134, 68, 90, 153, 131, 116, 128, 146, 143, 148]   \n",
       "gd_sum                                         -0.315989   \n",
       "total                                          -0.297346   \n",
       "\n",
       "                                               plasmid_11  \\\n",
       "GC_sum                                        0.000303181   \n",
       "chain   [106, 194, 212, 208, 210, 120, 187, 188, 15, 1...   \n",
       "gd_sum                                         -0.0239502   \n",
       "total                                           -0.023647   \n",
       "\n",
       "                              plasmid_12                plasmid_13  \\\n",
       "GC_sum                         0.0030474                0.00171596   \n",
       "chain   [6, 178, 154, 135, 176, 49, 178]  [178, 154, 135, 176, 52]   \n",
       "gd_sum                       -0.00590275                -0.0108497   \n",
       "total                        -0.00285535               -0.00913377   \n",
       "\n",
       "             plasmid_14                                         plasmid_15  \\\n",
       "GC_sum       0.00699621                                        -0.00531383   \n",
       "chain   [178, 154, 135]  [172, 211, 188, 187, 120, 210, 208, 212, 193, ...   \n",
       "gd_sum        -0.385926                                          -0.016139   \n",
       "total          -0.37893                                         -0.0214529   \n",
       "\n",
       "                                              plasmid_16  \\\n",
       "GC_sum                                         0.0013756   \n",
       "chain   [194, 212, 208, 210, 120, 187, 189, 44, 102, 86]   \n",
       "gd_sum                                        -0.0223755   \n",
       "total                                         -0.0209999   \n",
       "\n",
       "                     plasmid_17                plasmid_18  \\\n",
       "GC_sum               -0.0108338                -0.0195227   \n",
       "chain   [192, 74, 102, 38, 192]  [195, 115, 102, 64, 192]   \n",
       "gd_sum                -0.014406                -0.0431704   \n",
       "total                -0.0252398                -0.0626932   \n",
       "\n",
       "                                         plasmid_19  \\\n",
       "GC_sum                                    -0.031398   \n",
       "chain   [195, 56, 102, 82, 189, 187, 120, 210, 208]   \n",
       "gd_sum                                     -0.03176   \n",
       "total                                     -0.063158   \n",
       "\n",
       "                                                plasmid_2  \\\n",
       "GC_sum                                        -0.00121611   \n",
       "chain   [151, 84, 158, 133, 68, 93, 153, 132, 116, 127...   \n",
       "gd_sum                                          -0.067396   \n",
       "total                                          -0.0686121   \n",
       "\n",
       "                                plasmid_20  \\\n",
       "GC_sum                           0.0111853   \n",
       "chain   [212, 208, 210, 120, 186, 190, 69]   \n",
       "gd_sum                          -0.0271974   \n",
       "total                           -0.0160121   \n",
       "\n",
       "                                plasmid_21  \\\n",
       "GC_sum                          0.00692474   \n",
       "chain   [212, 208, 210, 120, 186, 191, 34]   \n",
       "gd_sum                         -0.00583331   \n",
       "total                           0.00109142   \n",
       "\n",
       "                                               plasmid_22  \\\n",
       "GC_sum                                         -0.0107214   \n",
       "chain   [193, 212, 208, 210, 120, 186, 189, 58, 189, 1...   \n",
       "gd_sum                                         -0.0549323   \n",
       "total                                          -0.0656538   \n",
       "\n",
       "                                plasmid_23  \\\n",
       "GC_sum                          0.00700724   \n",
       "chain   [212, 208, 210, 120, 186, 191, 42]   \n",
       "gd_sum                          -0.0067238   \n",
       "total                          0.000283434   \n",
       "\n",
       "                                          plasmid_24  \\\n",
       "GC_sum                                    -0.0245555   \n",
       "chain   [208, 210, 120, 186, 190, 53, 188, 187, 120]   \n",
       "gd_sum                                    -0.0569558   \n",
       "total                                     -0.0815113   \n",
       "\n",
       "                                plasmid_25  \\\n",
       "GC_sum                          0.00789096   \n",
       "chain   [212, 208, 210, 120, 186, 191, 71]   \n",
       "gd_sum                          -0.0311747   \n",
       "total                           -0.0232838   \n",
       "\n",
       "                                               plasmid_26  \\\n",
       "GC_sum                                         0.00045135   \n",
       "chain   [120, 187, 188, 213, 192, 212, 208, 210, 120, ...   \n",
       "gd_sum                                         -0.0118767   \n",
       "total                                          -0.0114254   \n",
       "\n",
       "                                    plasmid_27  \\\n",
       "GC_sum                             0.000776025   \n",
       "chain   [192, 212, 208, 210, 120, 186, 190, 2]   \n",
       "gd_sum                            -0.000935788   \n",
       "total                             -0.000159763   \n",
       "\n",
       "                                     plasmid_28  \\\n",
       "GC_sum                               -0.0100773   \n",
       "chain   [193, 212, 208, 210, 120, 186, 189, 59]   \n",
       "gd_sum                               -0.0121487   \n",
       "total                                 -0.022226   \n",
       "\n",
       "                                     plasmid_29  \\\n",
       "GC_sum                              -0.00146033   \n",
       "chain   [192, 212, 208, 210, 120, 186, 190, 67]   \n",
       "gd_sum                               -0.0216865   \n",
       "total                                -0.0231469   \n",
       "\n",
       "                                     plasmid_3  \\\n",
       "GC_sum                              -0.0174176   \n",
       "chain   [151, 126, 169, 111, 97, 130, 161, 85]   \n",
       "gd_sum                              -0.0919549   \n",
       "total                                -0.109372   \n",
       "\n",
       "                                    plasmid_30  \\\n",
       "GC_sum                              0.00340978   \n",
       "chain   [192, 212, 208, 210, 120, 186, 191, 4]   \n",
       "gd_sum                             -0.00107986   \n",
       "total                               0.00232992   \n",
       "\n",
       "                                     plasmid_31  \\\n",
       "GC_sum                              -0.00183801   \n",
       "chain   [192, 212, 208, 210, 120, 186, 191, 66]   \n",
       "gd_sum                               -0.0182759   \n",
       "total                                -0.0201139   \n",
       "\n",
       "                           plasmid_32  \\\n",
       "GC_sum                     -0.0366416   \n",
       "chain   [208, 210, 120, 186, 190, 81]   \n",
       "gd_sum                     -0.0489702   \n",
       "total                      -0.0856118   \n",
       "\n",
       "                                               plasmid_33  \\\n",
       "GC_sum                                         0.00374412   \n",
       "chain   [192, 212, 208, 210, 120, 187, 189, 17, 100, 8...   \n",
       "gd_sum                                        -0.00353387   \n",
       "total                                         0.000210252   \n",
       "\n",
       "                                    plasmid_34  plasmid_35  \\\n",
       "GC_sum                             -0.00833435   0.0282795   \n",
       "chain   [193, 212, 208, 210, 120, 186, 189, 7]  [210, 120]   \n",
       "gd_sum                             -0.00163805   -0.360862   \n",
       "total                               -0.0099724   -0.332582   \n",
       "\n",
       "                plasmid_36        plasmid_37                     plasmid_4  \\\n",
       "GC_sum          0.00267696        0.00849572                     0.0163942   \n",
       "chain   [41, 100, 14, 155]  [5, 100, 61, 96]  [12, 210, 120, 187, 189, 45]   \n",
       "gd_sum         -0.00418812        -0.0030128                    -0.0664248   \n",
       "total          -0.00151117        0.00548292                    -0.0500306   \n",
       "\n",
       "                                                plasmid_5  \\\n",
       "GC_sum                                         -0.0028682   \n",
       "chain   [162, 181, 144, 137, 32, 194, 212, 208, 210, 1...   \n",
       "gd_sum                                         -0.0849682   \n",
       "total                                          -0.0878364   \n",
       "\n",
       "                                                plasmid_6  \\\n",
       "GC_sum                                        -0.00418644   \n",
       "chain   [107, 63, 101, 119, 159, 172, 107, 1, 102, 3, ...   \n",
       "gd_sum                                        -0.00404221   \n",
       "total                                         -0.00822866   \n",
       "\n",
       "                                                plasmid_7  \\\n",
       "GC_sum                                       -6.62224e-05   \n",
       "chain   [194, 40, 154, 135, 27, 194, 212, 208, 210, 12...   \n",
       "gd_sum                                        -0.00867263   \n",
       "total                                         -0.00873885   \n",
       "\n",
       "                           plasmid_8                 plasmid_9  \n",
       "GC_sum                   -0.00410727                -0.0628096  \n",
       "chain   [101, 24, 161, 129, 97, 110]  [181, 162, 95, 101, 119]  \n",
       "gd_sum                    -0.0255533                 -0.177263  \n",
       "total                     -0.0296605                 -0.240073  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_35  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
