{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 103 with $\\alpha_1 = 1$, $\\alpha_2 = 1$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 103\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020824.1</td>\n",
       "      <td>59372</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP020824.1        59372"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xl4VdW9//HPNwmEBDAKhDBKopJA\nBgIkYgVsgiLDVUAfHEhrWxWK1XLVQlVub+tUh+KVWvFKBayPE0XAgaJSgfojxVmCgkAwgMgkInMg\njBnW74/scI8hCQfJzgnh/XoeHs7ee521vyc5K8nn7LX3NuecAAAAACAs1AUAAAAAqB8IBwAAAAAk\nEQ4AAAAAeAgHAAAAACQRDgAAAAB4CAcAAAAAJBEOAAAAAHgIBwAAAAAkEQ4AAAAAeCJCXcDJatWq\nlYuPjw91GaeNAwcOqGnTpqEuA6j3GCtAcBgrJ2fp0qU7nXOxoa4DCNZpFw7i4+OVl5cX6jJOG7m5\nucrOzg51GUC9x1gBgsNYOTlmtjHUNQAng2lFAAAAACQRDgAAAAB4CAcAAAAAJBEOAAAAAHgIBwAA\nAAAkEQ4AAAAAeAgHAAAAACQRDgAAAAB4CAcAAAAAJBEOAAAAAHgIBwAAAAAkSRGhLqA+ih//dqhL\nqDXPD2oa6hIAAABwmuDIAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7C\nAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIB\nAAAAAElSRKgLgM++XSbdPyzUVdSO+wtDXQEAAECDxpEDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAA\nADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADy+hgMzG2RmBWa2zszG\nV7H9XDNbZGafm9kXZvYfftYDAAAAoHq+hQMzC5f0tKTBkpIl5ZhZcqVmv5c0yznXQ9IISZP9qgcA\nAABAzfw8ctBL0jrn3Hrn3FFJr0gaVqmNk3SW9zhG0lYf6wEAAABQgwgf+24vaXPA8hZJF1Vqc7+k\nBWb2n5KaSurvYz0AAAAAauBnOLAq1rlKyzmSnnfOTTSziyW9ZGapzrmy73VkNlrSaEmKi4tTbm6u\nH/UeMy6txNf+61JRZDvlJj0Q6jJqh8/fd5zZioqKfP/ZAjQEjBWgYfMzHGyR1DFguYOOnzY0UtIg\nSXLOfWRmTSS1krQ9sJFzbqqkqZKUmZnpsrOzfSq53I3j3/a1/7r0fPrXyi64L9Rl1I6cwlBXgAYs\nNzdXfv9sARoCxgrQsPl5zsESSZ3NLMHMGqv8hOO5ldpsknSZJJlZV0lNJO3wsSYAAAAA1fAtHDjn\nSiSNkTRf0mqVX5VolZk9aGZDvWbjJP3SzJZLmiHpRudc5alHAAAAAOqAn9OK5JybJ2lepXX3BjzO\nl9THzxoAAAAABIc7JAMAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAA\nAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkCRFhLoAAAAauvjxb4e6hFrz/KCmoS4BgI84\ncgAAAABAEuEAAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAAQBLhAAAAAICHcAAAAABAEuEAAAAAgIdw\nAAAAAEAS4QAAAACAh3AAAAAAQBLhAAAAAICHcAAAAABAEuEAAAAAgIdwAAAAAECSFBHqAgAAwGnk\n22XS/cNCXUXtuL8w1BUA9Q5HDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAA\nAPAQDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA\n8BAOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA8PgaDsxskJkVmNk6Mxtf\nTZvrzCzfzFaZ2d/9rAcAAABA9SL86tjMwiU9LelySVskLTGzuc65/IA2nSX9l6Q+zrk9Ztbar3oA\nAAAA1MzPIwe9JK1zzq13zh2V9IqkYZXa/FLS0865PZLknNvuYz0AAAAAauBnOGgvaXPA8hZvXaBE\nSYlm9oGZfWxmg3ysBwAAAEANfJtWJMmqWOeq2H9nSdmSOkh6z8xSnXN7v9eR2WhJoyUpLi5Oubm5\ntV5soHFpJb72X5eKItspN+mBUJdRO3z+vuPMVlRU5PvPFpy5+L1STzHmgeP4GQ62SOoYsNxB0tYq\n2nzsnCuW9LWZFag8LCwJbOScmyppqiRlZma67Oxsv2qWJN04/m1f+69Lz6d/reyC+0JdRu3IKQx1\nBWjAcnNz5ffPFpy5+L1ST/F7BTiOn9OKlkjqbGYJZtZY0ghJcyu1mSOpnySZWSuVTzNa72NNAAAA\nAKrhWzhwzpVIGiNpvqTVkmY551aZ2YNmNtRrNl/SLjPLl7RI0l3OuV1+1QQAAACgen5OK5Jzbp6k\neZXW3Rvw2Eka6/0DAAAAEELcIRkAAACAJJ+PHABo2OIb0kmWg5qGugQAAEKOcAAAkvTtMun+yvdp\nPE3dzxVYAAA/DNOKAAAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAA\nAEgiHAAAAADwEA4AAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAA\nSCIcAAAAAPAEHQ7MrK+Z3eQ9jjWzBP/KAgAAAFDXggoHZnafpHsk/Ze3qpGkl/0qCgAAAEDdC/bI\nwdWShko6IEnOua2SmvtVFAAAAIC6F2w4OOqcc5KcJJlZU/9KAgAAABAKwYaDWWY2RdLZZvZLSf+S\nNM2/sgAAAADUtYhgGjnnHjezyyXtk5Qk6V7n3EJfKwMAAABQp04YDswsXNJ851x/SQQCAAAAoIE6\n4bQi51yppINmFlMH9QAAAAAIkaCmFUk6LGmFmS2Ud8UiSXLO3e5LVQAAAADqXLDh4G3vHwAAAIAG\nKtgTkl8ws8aSEr1VBc65Yv/KAgAAAFDXggoHZpYt6QVJGySZpI5m9gvn3GL/SgMAAABQl4KdVjRR\n0gDnXIEkmVmipBmSMvwqDAAAAEDdCvYmaI0qgoEkOefWSGrkT0kAAAAAQiHYIwd5ZvY3SS95yz+V\ntNSfkgAAAACEQrDh4FZJv5Z0u8rPOVgsabJfRQEAAACoe8GGgwhJTzrn/iwdu2typG9VAQAANFBL\nly5tHRER8aykVAU/xRuoDWWSVpaUlIzKyMjYXlWDYMPBu5L6SyrylqMkLZDU+5RLBAAAOINEREQ8\n26ZNm66xsbF7wsLCXKjrwZmjrKzMduzYkbxt27ZnJQ2tqk2wabWJc64iGMh7HF0LNQIAAJxpUmNj\nY/cRDFDXwsLCXGxsbKHKj1pV3SbIvg6YWc+KBTPLlHToFOsDAAA4E4URDBAq3nuv2gwQ7LSiOyTN\nNrOtkpykdpKuP/XyAAAAANQXwYaDBEk9JJ0r6WpJP1J5SAAAAMApiB//dq3eVHbDn66oN5eb79Wr\nV9Ljjz+++cc//vHBYNqXlZWpd+/eifPmzVvXokWLMr/rO1VvvfVW84kTJ8YtWrRonR/9T58+PWbV\nqlVRjzzyyLbK26Kjo3scPHjw8+qe27t378R//OMfX8XGxpaezD6DnVb0B+fcPklnS7pc0lRJfz2Z\nHQEAAOD0UVxcXOf7nDVrVkxKSsqh0yEY1IWf/vSnhVUFg2Dk5OTsevzxx2NP9nnBhoOKxHGFpGec\nc/+Q1PhkdwYAAIDQu+uuu9omJCSk9O7du/OQIUMS7r333jip/JP+MWPGtL/wwguTHnroobitW7dG\nDBw48PzU1NSuqampXRcsWNBUkvbt2xd27bXXxqempnbt2rVr8ssvv3y2JBUVFdmVV155XmJiYvIV\nV1xx3uHDh02SnnjiiVYjR47sWLH/iRMntho1alSHynVNnz69xdVXX71XkgoKChonJCSkXH/99Z06\nd+6cMnTo0IQ5c+Y079mzZ5dOnTqlLlq0KLqmWgoKChpnZGQkJScnd01OTu66cOHCplL5p/29evVK\nGjRo0HkJCQkpQ4cOTSgrqzmLDB8+PP4nP/nJuRkZGUnx8fGpM2bMiKncZtGiRdE9evTo0rVr1+Qe\nPXp0Wb58eaQk5eXlNUlLS+vapUuX5MTExOQVK1ZEBvvaJk2a1PLnP//5uZL05ZdfNu7evXuX1NTU\nrnfccUe7iv1u3LixUWZmZlKXLl2SO3funPLOO+80k6QRI0bsff3111sG+ZY4JthpRd+Y2RSVX850\ngplFiuvyAgAAnHYWL14c/eabb56zYsWK/OLiYuvevXtyjx49jk372bt3b/iSJUsKJGnIkCEJY8eO\n/W7gwIFFa9eubTxw4MDO69evX/W73/2ubb9+/fbNnj17w86dO8MzMzO7Dh06dN+f//zn2KioqLI1\na9bkf/LJJ1F9+vRJlqSRI0fuTklJST5y5MiWyMhI9/LLL7eaMmXKxsq1LV26tFmfPn2Ord+8eXOT\nmTNnrs/IyNjYrVu3rtOnT2+Zl5f35d///vezH3744bb9+vX7qrpa2rVrV/Lee++tiY6OditWrIjM\nyck5b+XKlaslafXq1VHLli1bHx8fX5yRkdFl4cKFzQYOHFhUuZ5Amzdvjvz0008L8vPzI/v37580\nbNiwFYHb09PTD3/66adfNmrUSHPmzGl+9913d5g/f/5XTz31VOxtt9323a233rr78OHDVlJSom++\n+aZRMK8tsP/bbrvt3FGjRu0YM2bMrkcfffTYEYHnnnuuxWWXXVY4YcKEbSUlJdq/f3+YJMXGxpYe\nPXrUtm3bFt6mTZugpxYFGw6ukzRI0uPOub1m1lbSXcHuBAAAAPVDbm5us8GDB+9t1qyZk+Quv/zy\nvYHbc3Jydlc8/uCDD85au3ZtVMVyUVFR+J49e8Jyc3PPmj9//tmTJk1qI0lHjhyxdevWNX7//feb\n3X777dsl6aKLLjqUmJh4UJLOOuussj59+uyfOXNmTFpa2uHi4mLr1avXcVe+LCwsjDjnnHOOfYzf\nvn37IxXtEhMTD1166aX7wsLC1LNnz4MPPfRQO+/1VFlLp06dikeOHNkpPz8/KiwsTBs3bjx2A9+0\ntLQD559/frEkpaSkHPzqq69OOCNm+PDhu8PDw5WWlnakY8eOR5YtW9YkcPvu3bvDr7/++oQNGzY0\nMTNXXFxsknTxxRcfePzxx9tu2bKl8YgRI/akpaUdCfa1Bfrss8+a/fOf//xKkm655ZZdf/zjHztI\n0o9+9KMDt9xyS3xxcXHYNddcs6d3797Hvq4tW7Ys2bRpU+M2bdoEfZXRoMKBc+6gpNcDlr+V9G2w\nOwEAAED94FzN15Rp3rz5sT/OnXPKy8tb7QWJ7/Xx6quvrktPTz9S+flmVmW/o0eP3vnwww+3SUxM\nPHzDDTfsrKpNeHi4Ky0tVXh4uCSpcePGx/YbFhamJk2aOK+dSktLraZaxo4d265169bFr7322tdl\nZWWKioo6duJ3ZGTksX7Dw8NVUlJSddE1vK7Ky/fcc0/7rKys/QsXLvyqoKCg8aWXXpokSb/61a92\nX3LJJQfeeOONmMGDBydOnjx5Q1JS0pFgXltlVV0Cd/DgwUWLFy8ueO2112JuvPHGhNtvv/27MWPG\n7JLKg1J0dPRJnb/B1CAAAIAzSHZ2dtH8+fNjDh48aIWFhWH/+te/zq6ubd++ffdNmDChdcXyhx9+\nGCVJ/fr12zdx4sS4irn6H3zwQZTXvujll19uIUlLlixpsmbNmmM3zb300ksPfPvtt43feOONliNH\njtytKiQkJBxevXp1ZFXbqlNdLYWFheFt27YtDg8P1+TJk1uWlp54Zs2vf/3r9i+++GKVX4/XX3/9\nnNLSUq1atSpy8+bNkenp6YcDt+/bty+8Q4cORyVpypQprSrW5+fnN+7ateuR3//+99sHDBiwd9my\nZVGV+w5Gz549i6ZNm9ZCkqZNm3bsXII1a9Y0bt++ffG4ceN23nDDDTs/++yzaKn8yk87duxolJSU\ndFyAq0mw04oAAADgg7q+9GhWVtbBQYMGFSYnJ6e0b9/+SLdu3Q7ExMRU+Zfz1KlTN48aNercxMTE\n5NLSUrvooov29+7de9Of/vSnraNHjz63S5cuyc4569Chw5FFixat++1vf7t9xIgRCYmJickpKSkH\n09LSDgT2d9VVV+354osvoqu7vOaAAQMKFyxY0Dw1NTXoP2irq+XOO+/cPnz48PPnzJlzTt++ffdH\nRUWd8BP0/Pz8qIoToiu74IILjvTq1Stp165djf7yl79sjI6O/t6n+Pfcc8+2UaNGJUyaNKnNJZdc\nsq9i/UsvvdRi9uzZLSMiIlxsbGzxo48+unXv3r3hwb6+CpMnT940YsSI8yZPnhw3dOjQPRXr58+f\n33zSpEltIiIiXHR0dOn06dO/lqT3338/ukePHgcaNWp0UvuxEx1aOhVmNkjSk5LCJT3rnPtTNe2u\nkTRb0oXOubya+szMzHR5eTU2OWXx49/2tf+69Hz6WmUX3BfqMmrH/YWhrgCVMFbqKcZKvcNYqafq\nYKyY2VLnXGbguuXLl29IT0+vclpNXSksLAyLiYkp279/f9jFF1+c9Mwzz2zs27dvUPciOBX9+vW7\n4M477/xu2LBh+6vavnHjxkY5OTnxH3744Vq/a6lK3759O7///vvH7Xv48OHxV155ZeFNN920p6rn\n1Uc33XRTx6uuumpvVV/r5cuXt0pPT4+v6nm+TSsys3BJT0saLClZUo6ZJVfRrrmk2yV94lctAAAA\n+D833HBDpy5duiR369at65AhQ/b4HQx27twZHh8fn9qkSZOy6oKBJHXq1Kn45ptv3rl79+6QTH2v\nKhicrlJTUw/V9LWujp/TinpJWuecWy9JZvaKpGGS8iu1+6OkxyT91sdaAAAA4HnzzTe/rsv9tWrV\nqnTDhg0rg2k7atSoevfp/GuvvbYh1DWcrHHjxv2go1N+prL2kjYHLG/x1h1jZj0kdXTOveVjHQAA\nAACC4OeRg6ouwXTsBAczC5P0hKQbT9iR2WhJoyUpLi5Oubm5tVNhNcallfjaf10qimyn3KQHQl1G\n7fD5+46Tx1ippxgr9Q5jpZ5irADH8TMcbJHUMWC5g6StAcvNJaVKyvWuE9tG0lwzG1r5pGTn3FRJ\nU6XyE5Kzs7N9LFu6sUGdOPZ1wzlxLIeTLOsbxko9xVipdxgr9RRjBTiOn9OKlkjqbGYJZtZY0ghJ\ncys2OucKnXOtnHPxzrl4SR9LOi4YAAAAAKgbvh05cM6VmNkYSfNVfinT55xzq8zsQUl5zrm5NfcA\nAABwBrg/JuPEjU6mv8IT3jfBzDKGDRu2e86cOV9LUnFxsVq3bp3evXv3A4sWLVo3adKklnl5eU1f\nfPHFTWPHjm3XrFmz0gcffPC7wD7Cw8MzOnfufKi0tNQuuOCCQ7NmzdoQeHdlqfxGXL17906cN2/e\nuhYtWpRde+218e+++25My5YtS9auXbuqqto+//zzJjfddFN8fn5+9Pjx47+pvN8KjzzySOwzzzwT\nt3nz5sitW7cub9u2bYkkzZgxI+bTTz9t+sQTT2yt6nmoma+XiXLOzXPOJTrnznfOPeytu7eqYOCc\ny+aoAQAAgP+ioqLKCgoKooqKikyS3njjjbPi4uKKT6aPyMjIsi+//DJ/7dq1qxo1auQmTpwYW7nN\nrFmzYlJSUg61aNGiTJJuvvnmnXPnzq3xcqGtW7cuefLJJzfdcsstVYaCCllZWUULFy5c065du6OB\n66+//vrCd9555+z9+/eH5HKopzu+aAAAAGegyy67rHD27NlnS9KMGTNaDB8+fPcP7atv375F69at\ni6y8fvr06S0C7zg8ePDgotjY2BrP0G/fvn1JVlbWwUaNGtV4p94+ffocSkpKOlp5fVhYmHr37r1/\n5syZMSfzGlCOcAAAAHAG+tnPfrZ75syZ5xw8eNBWr14dffHFFx/4If0UFxdr/vz5Z6WlpR2qvG3p\n0qXN+vTpc8J+H3vssdjHHnvsuCMPlWVlZV2wYcOGRidql5mZeeC9995rdqJ2OJ6fVysCAABAPXXR\nRRcd2rJlS+S0adNa9O/f/6Qv3XTkyJGwLl26JHt97b/jjjuOu+lWYWFhxDnnnFN2/LO/7+67794R\nzD7//e9/rwumXZs2bUq2bdvWOJi2+D7CAQAAwBlq0KBBe++7776OCxYsKNi+fftJ/V1Ycc5BTW3C\nw8NdaWmpwsPDT63Qk3To0CFr0qTJCUMJjse0IgAAgDPUrbfeunPcuHFbe/XqddyUoNqQkJBwePXq\n1cedi+C3goKCJikpKb68poaOIwcAAAChFMSlR/1y/vnnF//hD3/YfqJ2TzzxRNspU6bEVSx/9913\nXwTT/4ABAwoXLFjQPDU19YgkDRkyJOHjjz9uvmfPnoi4uLhu48eP3/qb3/xmZ8X5BnffffeOTZs2\nRVx44YXJBw4cCDczN2XKlLjVq1evbNGiRVlWVtYFL7zwwsb4+Pjihx56qPVTTz3VZteuXY3S09OT\n+/XrVzhz5syNkrR48eLmEyZM+OaHfVXObOZcjSeC1zuZmZkuL8/fK57GN6g7Wa5tOHeyvJ87WdY3\njJV6irFS7zBW6qk6GCtmttQ5lxm4bvny5RvS09OPm5/fEG3cuLFRTk5O/Icffljj5Utr0+bNmyOu\nu+668z766KM1dbXP083y5ctbpaenx1e1jWlFAAAA8EWnTp2Kb7755p27d++us785169f33jixImb\n62p/DQ3TigAAAOCbUaNG7anL/WVlZR2sy/01NBw5AAAAqFtlZWVlFuoicGby3nvVXsmJcAAAAFC3\nVu7YsSOGgIC6VlZWZjt27IiRtLK6NkwrAgAAqEMlJSWjtm3b9uy2bdtSxQe1qFtlklaWlJSMqq4B\n4QAAAKAOZWRkbJc0NNR1AFUhrQIAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAA\nAOAhHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA\n4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADg\nIRwAAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAh\nHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkORzODCzQWZWYGbrzGx8FdvHmlm+mX1h\nZu+aWSc/6wEAAABQPd/CgZmFS3pa0mBJyZJyzCy5UrPPJWU657pJelXSY37VAwAAAKBmfh456CVp\nnXNuvXPuqKRXJA0LbOCcW+ScO+gtfiypg4/1AAAAAKiBn+GgvaTNActbvHXVGSnpnz7WAwAAAKAG\nET72bVWsc1U2NLtBUqakrGq2j5Y0WpLi4uKUm5tbSyVWbVxaia/916WiyHbKTXog1GXUDp+/7zh5\njJV6irFS7zBW6inGCnAcP8PBFkkdA5Y7SNpauZGZ9Zf035KynHNHqurIOTdV0lRJyszMdNnZ2bVe\nbKAbx7/ta/916fn0r5VdcF+oy6gdOYWhrgCVMFbqKcZKvcNYqacYK8Bx/JxWtERSZzNLMLPGkkZI\nmhvYwMx6SJoiaahzbruPtQAAAAA4Ad/CgXOuRNIYSfMlrZY0yzm3ysweNLOhXrP/kdRM0mwzW2Zm\nc6vpDgAAAIDP/JxWJOfcPEnzKq27N+Bxfz/3DwAAACB43CEZAAAAgCTCAQAAAAAP4QAAAACAJMIB\nAAAAAA/hAAAAAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEA\nAAAAD+EAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAAD+EAAAAAgCTCAQAA\nAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAA\nAA/hAAAAAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAA\nD+EAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP\n4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAAj6/hwMwGmVmBma0zs/FVbI80s5ne9k/MLN7PegAA\nAABUz7dwYGbhkp6WNFhSsqQcM0uu1GykpD3OuQskPSFpgl/1AAAAAKiZn0cOekla55xb75w7KukV\nScMqtRkm6QXv8auSLjMz87EmAAAAANXwMxy0l7Q5YHmLt67KNs65EkmFklr6WBMAAACAakT42HdV\nRwDcD2gjMxstabS3WGRmBadY2xmjn9RK0s5Q11ErHuCgEvzDWAGCw1g5aZ3qYidAbfEzHGyR1DFg\nuYOkrdW02WJmEZJiJO2u3JFzbqqkqT7V2aCZWZ5zLjPUdQD1HWMFCA5jBWjY/JxWtERSZzNLMLPG\nkkZImlupzVxJv/AeXyPp/znnjjtyAAAAAMB/vh05cM6VmNkYSfMlhUt6zjm3yswelJTnnJsr6W+S\nXjKzdSo/YjDCr3oAAAAA1Mz4oL5hM7PR3rQsADVgrADBYawADRvhAAAAAIAkn++QDAAAAOD0QTgA\nAAAAIIlwcMYzs3gzW+k9zjazt0JdExAKZvbhCbbPM7Oz66oeoCEws1wzy/QebzCzVqGuCUDN/LzP\nAXxkZqbyc0bKQl0LUN+YWbhzrvRknuOc632C7f9xalUBDQ+/i4CGhyMHpxHvU/7VZjZZ0meSfmZm\nH5nZZ2Y228yaee0uNLMPzWy5mX1qZs29577ntf3MzGr8Qwior7z38pdm9oKZfWFmr5pZtPep5L1m\n9r6ka83sfDN7x8yWeu/9Lt4LxoWFAAAEAklEQVTz48zsDW98LK8YC2ZW5P3f1swWm9kyM1tpZpd4\n64996mlmY71tK83szoC6VpvZNDNbZWYLzCwqJF8koAaV379mNsHMbgvYfr+ZjfMe32VmS7yx9oC3\nrvLvoo5m9lczy/Pe+w+E5pUBqA2Eg9NPkqQXJV0uaaSk/s65npLyJI31bjg3U9Idzrl0Sf0lHZK0\nXdLlXtvrJU0KRfFALUmSNNU5103SPkkVf9gcds71dc69ovK7qv+ncy5D0m8lTfbaTJL0b2989JS0\nqlLfP5E03znXXVK6pGWBG80sQ9JNki6S9CNJvzSzHt7mzpKeds6lSNoraXhtvWCgNlT1/pX0isp/\nL1S4TtJsMxug8vd0L0ndJWWY2Y+9NkmSXnTO9XDObZT0395dk7tJyjKzbnXyggDUOqYVnX42Ouc+\nNrMrJSVL+qD8qK4aS/pI5T+wv3XOLZEk59w+STKzppL+18y6SyqVlBiK4oFastk594H3+GVJt3uP\nZ0qSdxStt8r/wKl4TqT3/6WSfi5J3tSjwkp9L5H0nJk1kjTHObes0va+kt5wzh3w9vW6pEtUfsf3\nrwPaL5UUfwqvEfBDde/f1mbWTlKspD3OuU1mdrukAZI+957bTOVhYZO830UB/V5nZqNV/ndFW5X/\nfvqiLl4QgNpFODj9HPD+N0kLnXM5gRu9T2uqunnFbyR9p/JPQsMkHfazSMBnld/jFcsV4yNM0l7v\n0/+T69i5xd6no1eo/A7u/+OcezGgiVXzVEk6EvC4VBLTilDfVPf+fVXSNZLaqPxIQkXbR51zU77X\ngVm8/m+sycwSVH507kLn3B4ze15Sk1qtGkCdYVrR6etjSX3M7AJJ8uZcJ0r6UlI7M7vQW9/czCIk\nxaj8iEKZpJ9JCg9R3UBtONfMLvYe50h6P3Cjd8TsazO7Vio/adLM0r3N70q61VsfbmZnBT7XzDpJ\n2u6cmybpbyqfehRosaSrvDHXVNLVkt6rvZcG+Kq69+8rkkaoPCC86rWdL+nmgPPZ2ptZ6yr6PEvl\nYaHQzOIkDfb5NQDwEeHgNOWc2yHpRkkzzOwLlYeFLs65oyqfO/qUmS2XtFDln+BMlvQLM/tY5VOK\nDlTZMXB6WK3y9/MXklpI+msVbX4qaaQ3DlZJGuatv0NSPzNbofKpPymVnpctaZmZfa7ycwaeDNzo\nnPtM0vOSPpX0iaRnnXOfCzgNVPf+dc6tktRc0jfOuW+9tgsk/V3SR954edVrU7nP5SqferRK0nOS\nPqjcBsDpw5yragYKANRP3pSGt5xzqSEuBQCABocjBwAAAAAkceQAAAAAgIcjBwAAAAAkEQ4AAAAA\neAgHAAAAACQRDgAAAAB4CAcAAAAAJEn/HxL+QaQAq313AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f2cc9b74b70>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:1:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>37700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>50571</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0        37700\n",
       "1  plasmid_1        50571"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.851765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.572906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.685044</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.851765\n",
       "1  precision  0.572906\n",
       "2   f1_score  0.685044"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP020824.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.851765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.851765</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP020824.1\n",
       "0  plasmid_0    0.000000\n",
       "1  plasmid_1    0.851765\n",
       "2      Total    0.851765"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020824.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1\n",
       "0  CP020824.1        0.0        1.0\n",
       "1       Total        0.0        1.0"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0         0.0\n",
       "1  plasmid_1         1.0"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020824.1</td>\n",
       "      <td>0.851765</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP020824.1    0.851765"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020824.1</td>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.851765</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction    recall  precision\n",
       "0  CP020824.1  plasmid_1  0.851765        1.0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 1, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0609575</td>\n",
       "      <td>-0.00187071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[43, 26]</td>\n",
       "      <td>[40, 22]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.381844</td>\n",
       "      <td>-0.416365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.442801</td>\n",
       "      <td>-0.418236</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        plasmid_0   plasmid_1\n",
       "GC_sum -0.0609575 -0.00187071\n",
       "chain    [43, 26]    [40, 22]\n",
       "gd_sum  -0.381844   -0.416365\n",
       "total   -0.442801   -0.418236"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0606308</td>\n",
       "      <td>-0.000575404</td>\n",
       "      <td>-0.00221565</td>\n",
       "      <td>0.00517059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[43, 39, 43, 26]</td>\n",
       "      <td>[52, 36, 52, 42, 47, 40, 22]</td>\n",
       "      <td>[46, 40, 4, 57]</td>\n",
       "      <td>[40]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.367383</td>\n",
       "      <td>-0.4169</td>\n",
       "      <td>-0.0012162</td>\n",
       "      <td>-0.376775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.428014</td>\n",
       "      <td>-0.417476</td>\n",
       "      <td>-0.00343185</td>\n",
       "      <td>-0.371604</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               plasmid_0                     plasmid_1        plasmid_2  \\\n",
       "GC_sum        -0.0606308                  -0.000575404      -0.00221565   \n",
       "chain   [43, 39, 43, 26]  [52, 36, 52, 42, 47, 40, 22]  [46, 40, 4, 57]   \n",
       "gd_sum         -0.367383                       -0.4169       -0.0012162   \n",
       "total          -0.428014                     -0.417476      -0.00343185   \n",
       "\n",
       "         plasmid_3  \n",
       "GC_sum  0.00517059  \n",
       "chain         [40]  \n",
       "gd_sum   -0.376775  \n",
       "total    -0.371604  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_0  from MILP completely contained in  plasmid_0  from greedy.\n",
      "\n",
      "plasmid_1  from MILP completely contained in  plasmid_1  from greedy.\n",
      "\n",
      "plasmid_3  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
