{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 109 with $\\alpha_1 = 1$, $\\alpha_2 = 1$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 109\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012925.1</td>\n",
       "      <td>2096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012927.1</td>\n",
       "      <td>3372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012928.1</td>\n",
       "      <td>62920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012929.1</td>\n",
       "      <td>96042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012926.1</td>\n",
       "      <td>37697</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP012925.1         2096\n",
       "1  CP012927.1         3372\n",
       "2  CP012928.1        62920\n",
       "3  CP012929.1        96042\n",
       "4  CP012926.1        37697"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt0VeWZx/HfkwuQAKJACFdJqgRI\nAgESQQFLUBSYKtiFF2i1VaE4WkYttMrMtNZaW6sjtcWRCnZcaqUIeKG0UoF2SPFaCQpyMxCRm4iA\nQCBcc3nmj+wwMSThoNkJhO9nLZZn7/2edz/nrPOa8zv73XubuwsAAAAAouq7AAAAAACnB8IBAAAA\nAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAACBmPou4FS1\nbt3ak5KS6ruMM8bBgwfVtGnT+i4DOO0xVoDIMFZOzfLly3e7e0J91wFE6owLB0lJScrNza3vMs4Y\nOTk5ys7Oru8ygNMeYwWIDGPl1JjZ5vquATgVTCsCAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4\nAAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACAQWjgws6fNbKeZra5mu5nZVDPLN7MPzKxPWLUAAAAA\nOLkwjxw8I2lYDduHS+oS/Bsv6Xch1gIAAADgJEILB+6+VNKeGpqMlPScl3lH0rlm1i6segAAAADU\nLKYe991B0tYKy9uCdZ9Wbmhm41V2dEGJiYnKycmpi/oahMLCQt4vhGbVJwX1XUKtSW4RzVgBIsDf\nFaBhq89wYFWs86oauvsMSTMkKSsry7Ozs0Msq2HJyckR7xfCcvPkV+u7hFrzzLCmjBUgAvxdARq2\n+gwH2yR1qrDcUdL2eqoFAIDQJDWwIA2g4arPS5nOl/Sd4KpFF0sqcPcTphQBAAAAqBuhHTkws1mS\nsiW1NrNtkn4qKVaS3P1JSQsk/YukfEmHJN0SVi2nil94AAAAcDYKLRy4+5iTbHdJ3w9r/wAAAABO\nDXdIBgAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAg\nQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBA\nOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4\nAAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgA\nAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAA\nAAAgiXAAAAAAIBBqODCzYWaWZ2b5Zja5iu3nm9kSM3vfzD4ws38Jsx4AAAAA1QstHJhZtKQnJA2X\nlCppjJmlVmr2Y0lz3L23pNGSpoVVDwAAAICahXnkoK+kfHff6O7HJL0gaWSlNi7pnOBxC0nbQ6wH\nAAAAQA1iQuy7g6StFZa3SepXqc39khaZ2b9JaippSIj1AAAAAKhBmOHAqljnlZbHSHrG3aeY2SWS\n/mBm6e5e+oWOzMZLGi9JiYmJysnJCaPe4yb1KA61/7pUWFgY+vuFsxdjBYgMYwXAmSLMcLBNUqcK\nyx114rShsZKGSZK7v21mTSS1lrSzYiN3nyFphiRlZWV5dnZ2SCWXuXnyq6H2X5eeGdZUYb9fOHsx\nVoDIMFYAnCnCPOdgmaQuZpZsZo1UdsLx/Epttki6XJLMrLukJpJ2hVgTAAAAgGqEFg7cvVjSBEkL\nJa1T2VWJ1pjZA2Y2Img2SdL3zGylpFmSbnb3ylOPAAAAANSBMKcVyd0XSFpQad19FR6vlTQgzBoA\nAAAARIY7JAMAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgA\nAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAA\nAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAA\nABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAA\nECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQ\nIBwAAAAAkEQ4AAAAABAgHAAAAACQFHI4MLNhZpZnZvlmNrmaNteb2VozW2NmfwyzHgAAAADViwmr\nYzOLlvSEpCskbZO0zMzmu/vaCm26SPp3SQPcfa+ZtQmrHgAAAAA1C/PIQV9J+e6+0d2PSXpB0shK\nbb4n6Ql33ytJ7r4zxHoAAAAA1CC0IweSOkjaWmF5m6R+ldqkSJKZvSkpWtL97v5a5Y7MbLyk8ZKU\nmJionJycMOo9blKP4lD7r0uFhYWhv184ezFWgMgwVgCcKcIMB1bFOq9i/10kZUvqKOl1M0t3931f\neJL7DEkzJCkrK8uzs7NrvdiKbp78aqj916VnhjVV2O8Xzl6MFSAyjBUAZ4owpxVtk9SpwnJHSdur\naPMndy9y948l5aksLAAAAACoY2GGg2WSuphZspk1kjRa0vxKbeZJGixJZtZaZdOMNoZYEwAAAIBq\nhBYO3L1Y0gRJCyWtkzTH3deY2QNmNiJotlDS52a2VtISST9y98/DqgkAAABA9cI850DuvkDSgkrr\n7qvw2CVNDP4BAAAAqEfcIRkAAACAJMIBAAAAgADhAAAAAICkUwgHZjbQzG4JHieYWXJ4ZQEAAACo\naxGFAzP7qaR7Jf17sCpW0vNhFQUAAACg7kV65OCbkkZIOihJ7r5dUvOwigIAAABQ9yINB8eCy466\nJJlZ0/BKAgAAAFAfIg0Hc8xsuqRzzex7kv4m6anwygIAAABQ1yK6CZq7P2pmV0jaL6mrpPvcfXGo\nlQEAAACoUycNB2YWLWmhuw+RRCAAAAAAGqiTTity9xJJh8ysRR3UAwAAAKCeRDStSNIRSavMbLGC\nKxZJkrvfGUpVAAAAAOpcpOHg1eAfAAAAgAYq0hOSnzWzRpJSglV57l4UXlkAAAAA6lpE4cDMsiU9\nK2mTJJPUycy+6+5LwysNAAAAQF2KdFrRFElXunueJJlZiqRZkjLDKgwAAABA3Yr0Jmix5cFAktx9\nvaTYcEoCAAAAUB8iPXKQa2b/I+kPwfK3JS0PpyQAAAAA9SHScHC7pO9LulNl5xwslTQtrKIAAAAA\n1L1Iw0GMpN+6+6+l43dNbhxaVQAAAA3U8uXL28TExPxeUroin+IN1IZSSauLi4vHZWZm7qyqQaTh\n4O+ShkgqDJbjJC2S1P8rlwgAAHAWiYmJ+X3btm27JyQk7I2KivL6rgdnj9LSUtu1a1fqjh07fi9p\nRFVtIk2rTdy9PBgoeBxfCzUCAACcbdITEhL2EwxQ16KiojwhIaFAZUetqm4TYV8HzaxP+YKZZUk6\n/BXrAwAAOBtFEQxQX4LPXrUZINJpRXdJmmtm2yW5pPaSbvjq5QEAAAA4XUQaDpIl9ZZ0vqRvSrpY\nZSEBAAAAX0HS5Fdr9aaym371jdPmcvN9+/bt+uijj279+te/fiiS9qWlperfv3/KggUL8lu2bFka\ndn1f1V/+8pfmU6ZMSVyyZEl+GP3PnDmzxZo1a+J++ctf7qi8LT4+vvehQ4fer+65/fv3T/nTn/70\nUUJCQsmp7DPSaUU/cff9ks6VdIWkGZJ+dyo7AgAAwJmjqKiozvc5Z86cFmlpaYfPhGBQF7797W8X\nVBUMIjFmzJjPH3300YRTfV6k4aA8cXxD0pPu/idJjU51ZwAAAKh/P/rRj9olJyen9e/fv8vVV1+d\nfN999yVKZb/0T5gwocNFF13U9cEHH0zcvn17zNChQy9IT0/vnp6e3n3RokVNJWn//v1R1113XVJ6\nenr37t27pz7//PPnSlJhYaFdddVVX0tJSUn9xje+8bUjR46YJD322GOtx44d26l8/1OmTGk9bty4\njpXrmjlzZstvfvOb+yQpLy+vUXJyctoNN9zQuUuXLmkjRoxInjdvXvM+ffp069y5c/qSJUvia6ol\nLy+vUWZmZtfU1NTuqamp3RcvXtxUKvu1v2/fvl2HDRv2teTk5LQRI0Ykl5bWnEVGjRqV9K1vfev8\nzMzMrklJSemzZs1qUbnNkiVL4nv37t2te/fuqb179+62cuXKxpKUm5vbpEePHt27deuWmpKSkrpq\n1arGkb62qVOntvrOd75zviR9+OGHjXr16tUtPT29+1133dW+fL+bN2+OzcrK6tqtW7fULl26pL32\n2mvNJGn06NH7Xn755VYRfiSOi3Ra0SdmNl1llzN92Mwai+vyAgAAnHGWLl0a/+c///m8VatWrS0q\nKrJevXql9u7d+/i0n3379kUvW7YsT5Kuvvrq5IkTJ342dOjQwg0bNjQaOnRol40bN675j//4j3aD\nBw/eP3fu3E27d++OzsrK6j5ixIj9v/71rxPi4uJK169fv/af//xn3IABA1IlaezYsXvS0tJSjx49\nuq1x48b+/PPPt54+ffrmyrUtX7682YABA46v37p1a5PZs2dvzMzM3NyzZ8/uM2fObJWbm/vhH//4\nx3N/8YtftBs8ePBH1dXSvn374tdff319fHy8r1q1qvGYMWO+tnr16nWStG7durgVK1ZsTEpKKsrM\nzOy2ePHiZkOHDi2sXE9FW7dubfzuu+/mrV27tvGQIUO6jhw5clXF7RkZGUfefffdD2NjYzVv3rzm\n99xzT8eFCxd+9Pjjjyfccccdn91+++17jhw5YsXFxfrkk09iI3ltFfu/4447zh83btyuCRMmfP7Q\nQw8dPyLw9NNPt7z88ssLHn744R3FxcU6cOBAlCQlJCSUHDt2zHbs2BHdtm3biKcWRRoOrpc0TNKj\n7r7PzNpJ+lGkOwEAAMDpIScnp9nw4cP3NWvWzCX5FVdcsa/i9jFjxuwpf/zmm2+es2HDhrjy5cLC\nwui9e/dG5eTknLNw4cJzp06d2laSjh49avn5+Y3eeOONZnfeeedOSerXr9/hlJSUQ5J0zjnnlA4Y\nMODA7NmzW/To0eNIUVGR9e3b94QrXxYUFMScd955x3/G79Chw9HydikpKYcvu+yy/VFRUerTp8+h\nBx98sH3weqqspXPnzkVjx47tvHbt2rioqCht3rz5+A18e/TocfCCCy4okqS0tLRDH3300UlnxIwa\nNWpPdHS0evTocbRTp05HV6xY0aTi9j179kTfcMMNyZs2bWpiZl5UVGSSdMkllxx89NFH223btq3R\n6NGj9/bo0eNopK+tovfee6/ZX//6148k6bbbbvv85z//eUdJuvjiiw/edtttSUVFRVHXXnvt3v79\n+x9/X1u1alW8ZcuWRm3bto34KqMRhQN3PyTp5QrLn0r6NNKdAAAA4PTgXvM1ZZo3b378y7m7Kzc3\nd10QJL7Qx4svvpifkZFxtPLzzazKfsePH7/7F7/4RduUlJQjN9544+6q2kRHR3tJSYmio6MlSY0a\nNTq+36ioKDVp0sSDdiopKbGaapk4cWL7Nm3aFL300ksfl5aWKi4u7viJ340bNz7eb3R0tIqLi6su\nuobXVXn53nvv7TBo0KADixcv/igvL6/RZZdd1lWS/vVf/3XPpZdeevCVV15pMXz48JRp06Zt6tq1\n69FIXltlVV0Cd/jw4YVLly7Ne+mll1rcfPPNyXfeeednEyZM+FwqC0rx8fGndP4GU4MAAADOItnZ\n2YULFy5scejQISsoKIj629/+dm51bQcOHLj/4YcfblO+/NZbb8VJ0uDBg/dPmTIlsXyu/ptvvhkX\ntC98/vnnW0rSsmXLmqxfv/74TXMvu+yyg59++mmjV155pdXYsWP3qArJyclH1q1b17iqbdWprpaC\ngoLodu3aFUVHR2vatGmtSkpOPrPm+9//fofnnnuuyvfj5ZdfPq+kpERr1qxpvHXr1sYZGRlHKm7f\nv39/dMeOHY9J0vTp01uXr1+7dm2j7t27H/3xj3+888orr9y3YsWKuMp9R6JPnz6FTz31VEtJeuqp\np46fS7B+/fpGHTp0KJo0adLuG2+8cfd7770XL5Vd+WnXrl2xXbt2PSHA1STSaUUAAAAIQV1fenTQ\noEGHhg0bVpCamprWoUOHoz179jzYokWLKr85z5gxY+u4cePOT0lJSS0pKbF+/fod6N+//5Zf/epX\n28ePH39+t27dUt3dOnbseHTJkiX5P/zhD3eOHj06OSUlJTUtLe1Qjx49Dlbs75prrtn7wQcfxFd3\nec0rr7yyYNGiRc3T09Mj/kJbXS133333zlGjRl0wb9688wYOHHggLi7upL+gr127Nq78hOjKLrzw\nwqN9+/bt+vnnn8f+5je/2RwfH/+FX/HvvffeHePGjUueOnVq20svvXR/+fo//OEPLefOndsqJibG\nExISih566KHt+/bti4709ZWbNm3altGjR39t2rRpiSNGjNhbvn7hwoXNp06d2jYmJsbj4+NLZs6c\n+bEkvfHGG/G9e/c+GBsbe0r7sZMdWjrdZGVleW5ubqj7SJr8aqj916VnhjVVdnZ2fZeBBoqxAkSG\nsXL2MrPl7p5Vcd3KlSs3ZWRkVDmtpq4UFBREtWjRovTAgQNRl1xySdcnn3xy88CBAyO6F8FXMXjw\n4Avvvvvuz0aOHHmgqu2bN2+OHTNmTNJbb721IexaqjJw4MAub7zxxgn7HjVqVNJVV11VcMstt+yt\n6nmno1tuuaXTNddcs6+q93rlypWtMzIykqp6HtOKAAAAzjI33nhj527duqX27Nmz+9VXX7037GCw\ne/fu6KSkpPQmTZqUVhcMJKlz585Ft9566+49e/bUy3fUqoLBmSo9Pf1wTe91dZhWBAAAcJb585//\n/HFd7q9169YlmzZtWh1J23Hjxp12v86/9NJLm+q7hlM1adKkL3V0iiMHAAAAACQRDgAAAAAECAcA\nAAAAJBEOAAAAAAQ4IRkAAKA+3d8i8+SNTqW/gpPeN8HMMkeOHLln3rx5H0tSUVGR2rRpk9GrV6+D\nS5YsyZ86dWqr3Nzcps8999yWiRMntm/WrFnJAw888FnFPqKjozO7dOlyuKSkxC688MLDc+bM2VTx\n7spS2Y24+vfvn7JgwYL8li1bll533XVJf//731u0atWqeMOGDWuqqu39999vcssttyStXbs2fvLk\nyZ9U3m+5X/7ylwlPPvlk4tatWxtv3759Zbt27YoladasWS3efffdpo899tj2SN8y/D+OHAAAAJxl\n4uLiSvPy8uIKCwtNkl555ZVzEhMTi06lj8aNG5d++OGHazds2LAmNjbWp0yZklC5zZw5c1qkpaUd\nbtmyZakk3Xrrrbvnz59f4+VC27RpU/zb3/52y2233VZlKCg3aNCgwsWLF69v3779sYrrb7jhhoLX\nXnvt3AMHDvA990vgTQMAADgLXX755QVz5849V5JmzZrVctSoUXu+bF8DBw4szM/Pb1x5/cyZM1tW\nvOPw8OHDCxMSEopr6qtDhw7FgwYNOhQbG1vjnXoHDBhwuGvXrscqr4+KilL//v0PzJ49u8WpvAaU\nIRwAAACchW666aY9s2fPPu/QoUO2bt26+EsuueTgl+mnqKhICxcuPKdHjx6HK29bvnx5swEDBpy0\n30ceeSThkUceOeHIQ2WDBg26cNOmTbEna5eVlXXw9ddfb3aydjhRqOccmNkwSb+VFC3p9+7+q2ra\nXStprqSL3D03zJoAAAAg9evX7/C2bdsaP/XUUy2HDBlScKrPP3r0aFS3bt1Sg74O3HXXXSfcdKug\noCDmvPPOKz3x2V90zz337Ipkn//4xz/yI2nXtm3b4h07djSKpC2+KLRwYGbRkp6QdIWkbZKWmdl8\nd19bqV1zSXdK+mdYtQAAAOBEw4YN2/fTn/6006JFi/J27tx5St8Ly885qKlNdHS0l5SUKDo6+qsV\neooOHz5sTZo0OWkowYnCnFbUV1K+u29092OSXpA0sop2P5f0iKQjIdYCAACASm6//fbdkyZN2t63\nb98TpgTVhuTk5CPr1q074VyEsOXl5TVJS0sL5TU1dGFOK+ogaWuF5W2S+lVsYGa9JXVy97+Y2Q+r\n68jMxksaL0mJiYnKycmp/WormNSjxvNkziiFhYWhv184ezFWgMgwVlCjCC49GpYLLrig6Cc/+cnO\nk7V77LHH2k2fPj2xfPmzzz77IJL+r7zyyoJFixY1T09PPypJV199dfI777zTfO/evTGJiYk9J0+e\nvP0HP/jB7vLzDe65555dW7ZsibnoootSDx48GG1mPn369MR169atbtmyZemgQYMufPbZZzcnJSUV\nPfjgg20ef/zxtp9//nlsRkZG6uDBgwtmz569WZKWLl3a/OGHH/7ky70rZzdzr/FE8C/fsdl1koa6\n+7hg+SZJfd3934LlKEn/K+lmd99kZjmSfniycw6ysrI8Nzfc0xKSJr8aav916ZlhTZWdnV3fZaCB\nYqwAkWGsnL3MbLm7Z1Vct3Llyk0ZGRknzM9viDZv3hw7ZsyYpLfeeqvGy5fWpq1bt8Zcf/31X3v7\n7bfX19U+zzQrV65snZGRkVTVtjCnFW2T1KnCckdJFW9G0VxSuqQcM9sk6WJJ883sCwMIAAAAZ6bO\nnTsX3Xrrrbv37NlTZ1fI3LhxY6MpU6ZsPXlLVCXMaUXLJHUxs2RJn0gaLelb5RvdvUBS6/LlSI8c\nAEAoPl0h3V/VaVFnoPtP+aIjABCacePG7a3L/Q0aNOhQXe6voQktxbl7saQJkhZKWidpjruvMbMH\nzGxEWPsFAAA4zZWWlpZafReBs1Pw2av2Sk6h3ufA3RdIWlBp3X3VtM0OsxYAAFALOMpWG1bv2rUr\nNSEhoSAqKiqckz+BKpSWltquXbtaSFpdXZtQwwEAAAC+qLi4eNyOHTt+v2PHjnSFe/4nUFmppNXF\nxcXjqmtAOGjo+IUHAIDTSmZm5k5JTLHGaYm0CgAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAA\nAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAA\nAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAA\nQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABA\nEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS\n4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAUsjhwMyGmVmemeWb2eQqtk80s7Vm9oGZ/d3MOodZ\nDwAAAIDqhRYOzCxa0hOShktKlTTGzFIrNXtfUpa795T0oqRHwqoHAAAAQM3CPHLQV1K+u29092OS\nXpA0smIDd1/i7oeCxXckdQyxHgAAAAA1MHcPp2OzayUNc/dxwfJNkvq5+4Rq2v+3pB3u/mAV28ZL\nGi9JiYmJmS+88EIoNZdb9UlBqP3XpeS4o2p2dHt9l1E72vWq7wpQCWPlNMVYOe0wVk5TdTBWBg8e\nvNzds0LfEVBLYkLs26pYV2USMbMbJWVJGlTVdnefIWmGJGVlZXl2dnYtlVi1mye/Gmr/demZjI+V\nnffT+i6jdoxpOH9cGwrGymmKsXLaYaycphgrwAnCDAfbJHWqsNxR0gk/NZjZEEn/KWmQux8NsR4A\nAAAANQjznINlkrqYWbKZNZI0WtL8ig3MrLek6ZJGuPvOEGsBAAAAcBKhhQN3L5Y0QdJCSeskzXH3\nNWb2gJmNCJr9l6Rmkuaa2Qozm19NdwAAAABCFua0Irn7AkkLKq27r8LjIWHuHwAAAEDkuEMyAAAA\nAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAA\nSYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJ\nhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmE\nAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQD\nAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMA\nAAAAgVDDgZkNM7M8M8s3s8lVbG9sZrOD7f80s6Qw6wEAAABQvdDCgZlFS3pC0nBJqZLGmFlqpWZj\nJe119wslPSbp4bDqAQAAAFCzMI8c9JWU7+4b3f2YpBckjazUZqSkZ4PHL0q63MwsxJoAAAAAVCPM\ncNBB0tYKy9uCdVW2cfdiSQWSWoVYEwAAAIBqxITYd1VHAPxLtJGZjZc0PlgsNLO8r1jbWWOw1FrS\n7vquo1b8jINKCA9jBYgMY+WUda6LnQC1JcxwsE1SpwrLHSVtr6bNNjOLkdRC0p7KHbn7DEkzQqqz\nQTOzXHfPqu86gNMdYwWIDGMFaNjCnFa0TFIXM0s2s0aSRkuaX6nNfEnfDR5fK+l/3f2EIwcAAAAA\nwhfakQN3LzazCZIWSoqW9LS7rzGzByTluvt8Sf8j6Q9mlq+yIwajw6oHAAAAQM2MH+obNjMbH0zL\nAlADxgoQGcYK0LARDgAAAABICvkOyQAAAADOHIQDAAAAAJIIB2c9M0sys9XB42wz+0t91wTUBzN7\n6yTbF5jZuXVVD9AQmFmOmWUFjzeZWev6rglAzcK8zwFCZGamsnNGSuu7FuB0Y2bR7l5yKs9x9/4n\n2f4vX60qoOHhbxHQ8HDk4AyYsVCfAAAEU0lEQVQS/Mq/zsymSXpP0k1m9raZvWdmc82sWdDuIjN7\ny8xWmtm7ZtY8eO7rQdv3zKzGL0LA6Sr4LH9oZs+a2Qdm9qKZxQe/St5nZm9Ius7MLjCz18xsefDZ\n7xY8P9HMXgnGx8rysWBmhcF/25nZUjNbYWarzezSYP3xXz3NbGKwbbWZ3V2hrnVm9pSZrTGzRWYW\nVy9vElCDyp9fM3vYzO6osP1+M5sUPP6RmS0LxtrPgnWV/xZ1MrPfmVlu8Nn/Wf28MgC1gXBw5ukq\n6TlJV0gaK2mIu/eRlCtpYnDDudmS7nL3DElDJB2WtFPSFUHbGyRNrY/igVrSVdIMd+8pab+k8i82\nR9x9oLu/oLK7qv+bu2dK+qGkaUGbqZL+EYyPPpLWVOr7W5IWunsvSRmSVlTcaGaZkm6R1E/SxZK+\nZ2a9g81dJD3h7mmS9kkaVVsvGKgNVX1+Jb2gsr8L5a6XNNfMrlTZZ7qvpF6SMs3s60GbrpKec/fe\n7r5Z0n8Gd03uKWmQmfWskxcEoNYxrejMs9nd3zGzqySlSnqz7KiuGkl6W2X/w/7U3ZdJkrvvlyQz\nayrpv82sl6QSSSn1UTxQS7a6+5vB4+cl3Rk8ni1JwVG0/ir7glP+nMbBfy+T9B1JCqYeFVTqe5mk\np80sVtI8d19RaftASa+4+8FgXy9LulRld3z/uEL75ZKSvsJrBMJQ3ee3jZm1l5Qgaa+7bzGzOyVd\nKen94LnNVBYWtij4W1Sh3+vNbLzKvle0U9nfpw/q4gUBqF2EgzPPweC/Jmmxu4+puDH4taaqm1f8\nQNJnKvslNErSkTCLBEJW+TNevlw+PqIk7Qt+/T+1jt2XBr+OfkNld3D/L3d/rkITq+apknS0wuMS\nSUwrwummus/vi5KuldRWZUcSyts+5O7Tv9CBWZL+f6zJzJJVdnTuInffa2bPSGpSq1UDqDNMKzpz\nvSNpgJldKEnBnOsUSR9Kam9mFwXrm5tZjKQWKjuiUCrpJknR9VQ3UBvON7NLgsdjJL1RcWNwxOxj\nM7tOKjtp0swygs1/l3R7sD7azM6p+Fwz6yxpp7s/Jel/VDb1qKKlkq4JxlxTSd+U9HrtvTQgVNV9\nfl+QNFplAeHFoO1CSbdWOJ+tg5m1qaLPc1QWFgrMLFHS8JBfA4AQEQ7OUO6+S9LNkmaZ2QcqCwvd\n3P2YyuaOPm5mKyUtVtkvONMkfdfM3lHZlKKDVXYMnBnWqezz/IGklpJ+V0Wbb0saG4yDNZJGBuvv\nkjTYzFapbOpPWqXnZUtaYWbvq+ycgd9W3Oju70l6RtK7kv4p6ffu/r6AM0B1n193XyOpuaRP3P3T\noO0iSX+U9HYwXl4M2lTuc6XKph6tkfS0pDcrtwFw5jD3qmagAMDpKZjS8Bd3T6/nUgAAaHA4cgAA\nAABAEkcOAAAAAAQ4cgAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQJL0f8eSflEKPVfm\nAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f010fd88390>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:1:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>59036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>88349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>7460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>97955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>83174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>3287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0        59036\n",
       "1  plasmid_1        88349\n",
       "2  plasmid_2         7460\n",
       "3  plasmid_3        97955\n",
       "4  plasmid_4        83174\n",
       "5  plasmid_5         3287"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.375719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.389918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.382687</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.375719\n",
       "1  precision  0.389918\n",
       "2   f1_score  0.382687"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP012925.1</th>\n",
       "      <th>CP012926.1</th>\n",
       "      <th>CP012927.1</th>\n",
       "      <th>CP012928.1</th>\n",
       "      <th>CP012929.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.919310</td>\n",
       "      <td>0.086816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002750</td>\n",
       "      <td>0.008725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.088620</td>\n",
       "      <td>0.081745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.948188</td>\n",
       "      <td>0.102830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.027257</td>\n",
       "      <td>0.040992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.072044</td>\n",
       "      <td>0.003957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.058169</td>\n",
       "      <td>0.325066</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP012925.1  CP012926.1  CP012927.1  CP012928.1  CP012929.1\n",
       "0  plasmid_0         0.0         0.0         0.0    0.919310    0.086816\n",
       "1  plasmid_1         0.0         0.0         0.0    0.002750    0.008725\n",
       "2  plasmid_2         0.0         0.0         0.0    0.088620    0.081745\n",
       "3  plasmid_3         0.0         0.0         0.0    0.948188    0.102830\n",
       "4  plasmid_4         0.0         0.0         0.0    0.027257    0.040992\n",
       "5  plasmid_5         0.0         0.0         0.0    0.072044    0.003957\n",
       "6      Total         0.0         0.0         0.0    2.058169    0.325066"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012925.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012926.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012927.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012928.1</td>\n",
       "      <td>0.959008</td>\n",
       "      <td>0.001981</td>\n",
       "      <td>0.690483</td>\n",
       "      <td>0.604625</td>\n",
       "      <td>0.020655</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012929.1</td>\n",
       "      <td>0.131953</td>\n",
       "      <td>0.007504</td>\n",
       "      <td>0.967024</td>\n",
       "      <td>0.099444</td>\n",
       "      <td>0.044954</td>\n",
       "      <td>0.060846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Total</td>\n",
       "      <td>1.090961</td>\n",
       "      <td>0.009485</td>\n",
       "      <td>1.657507</td>\n",
       "      <td>0.704068</td>\n",
       "      <td>0.065609</td>\n",
       "      <td>1.060846</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2  plasmid_3  plasmid_4  \\\n",
       "0  CP012925.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "1  CP012926.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "2  CP012927.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "3  CP012928.1   0.959008   0.001981   0.690483   0.604625   0.020655   \n",
       "4  CP012929.1   0.131953   0.007504   0.967024   0.099444   0.044954   \n",
       "5       Total   1.090961   0.009485   1.657507   0.704068   0.065609   \n",
       "\n",
       "   plasmid_5  \n",
       "0   0.000000  \n",
       "1   0.000000  \n",
       "2   0.000000  \n",
       "3   1.000000  \n",
       "4   0.060846  \n",
       "5   1.060846  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.959059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.007504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.967024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.620305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.044954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.959059\n",
       "1  plasmid_1    0.007504\n",
       "2  plasmid_2    0.967024\n",
       "3  plasmid_3    0.620305\n",
       "4  plasmid_4    0.044954\n",
       "5  plasmid_5    1.000000"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012925.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012926.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012927.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012928.1</td>\n",
       "      <td>0.995598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012929.1</td>\n",
       "      <td>0.138481</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP012925.1    0.000000\n",
       "1  CP012926.1    0.000000\n",
       "2  CP012927.1    0.000000\n",
       "3  CP012928.1    0.995598\n",
       "4  CP012929.1    0.138481"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012928.1</td>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.91931</td>\n",
       "      <td>0.959008</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction   recall  precision\n",
       "0  CP012928.1  plasmid_0  0.91931   0.959008"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 1, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00378507</td>\n",
       "      <td>-0.0139092</td>\n",
       "      <td>0.0242971</td>\n",
       "      <td>-0.00441789</td>\n",
       "      <td>-0.0064734</td>\n",
       "      <td>0.0290831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[27, 46, 26, 36, 13, 48, 30]</td>\n",
       "      <td>[37, 14, 49, 15, 39]</td>\n",
       "      <td>[48, 31, 52, 25, 46, 24]</td>\n",
       "      <td>[66, 32, 63, 78, 42, 86, 88, 40, 72, 56, 41, 6...</td>\n",
       "      <td>[66, 32, 62, 78, 42, 86, 88, 40, 72, 56, 41, 6...</td>\n",
       "      <td>[52, 23]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.453401</td>\n",
       "      <td>-0.44271</td>\n",
       "      <td>-0.440416</td>\n",
       "      <td>-0.279863</td>\n",
       "      <td>-0.253499</td>\n",
       "      <td>-0.380438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.457186</td>\n",
       "      <td>-0.456619</td>\n",
       "      <td>-0.416118</td>\n",
       "      <td>-0.284281</td>\n",
       "      <td>-0.259972</td>\n",
       "      <td>-0.351355</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           plasmid_0             plasmid_1  \\\n",
       "GC_sum                   -0.00378507            -0.0139092   \n",
       "chain   [27, 46, 26, 36, 13, 48, 30]  [37, 14, 49, 15, 39]   \n",
       "gd_sum                     -0.453401              -0.44271   \n",
       "total                      -0.457186             -0.456619   \n",
       "\n",
       "                       plasmid_2  \\\n",
       "GC_sum                 0.0242971   \n",
       "chain   [48, 31, 52, 25, 46, 24]   \n",
       "gd_sum                 -0.440416   \n",
       "total                  -0.416118   \n",
       "\n",
       "                                                plasmid_3  \\\n",
       "GC_sum                                        -0.00441789   \n",
       "chain   [66, 32, 63, 78, 42, 86, 88, 40, 72, 56, 41, 6...   \n",
       "gd_sum                                          -0.279863   \n",
       "total                                           -0.284281   \n",
       "\n",
       "                                                plasmid_4  plasmid_5  \n",
       "GC_sum                                         -0.0064734  0.0290831  \n",
       "chain   [66, 32, 62, 78, 42, 86, 88, 40, 72, 56, 41, 6...   [52, 23]  \n",
       "gd_sum                                          -0.253499  -0.380438  \n",
       "total                                           -0.259972  -0.351355  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_11</th>\n",
       "      <th>plasmid_12</th>\n",
       "      <th>plasmid_13</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00418447</td>\n",
       "      <td>0.027847</td>\n",
       "      <td>-0.0394678</td>\n",
       "      <td>8.2781e-05</td>\n",
       "      <td>-0.00164105</td>\n",
       "      <td>0.00114274</td>\n",
       "      <td>0.0318185</td>\n",
       "      <td>-0.0541946</td>\n",
       "      <td>-0.0073864</td>\n",
       "      <td>0.0290831</td>\n",
       "      <td>0.017824</td>\n",
       "      <td>0.0369434</td>\n",
       "      <td>-0.0190562</td>\n",
       "      <td>-0.0306087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[15, 39, 15, 49, 24, 46, 27, 36, 26, 46, 25]</td>\n",
       "      <td>[36, 26, 46, 25, 52, 31, 48, 84, 49, 24, 46, 27]</td>\n",
       "      <td>[39]</td>\n",
       "      <td>[19, 75, 35, 7, 43]</td>\n",
       "      <td>[1, 75, 35, 6, 76]</td>\n",
       "      <td>[4, 75, 35, 3, 58]</td>\n",
       "      <td>[36, 26, 46, 27]</td>\n",
       "      <td>[16]</td>\n",
       "      <td>[13, 48, 30]</td>\n",
       "      <td>[52, 23]</td>\n",
       "      <td>[28]</td>\n",
       "      <td>[48]</td>\n",
       "      <td>[14, 37, 14]</td>\n",
       "      <td>[37]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.455719</td>\n",
       "      <td>-0.451718</td>\n",
       "      <td>-0.459794</td>\n",
       "      <td>-0.00140908</td>\n",
       "      <td>-0.000331057</td>\n",
       "      <td>-0.000255386</td>\n",
       "      <td>-0.466624</td>\n",
       "      <td>-0.410882</td>\n",
       "      <td>-0.452064</td>\n",
       "      <td>-0.380438</td>\n",
       "      <td>-0.236164</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.430233</td>\n",
       "      <td>-0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.459904</td>\n",
       "      <td>-0.423871</td>\n",
       "      <td>-0.499262</td>\n",
       "      <td>-0.0013263</td>\n",
       "      <td>-0.0019721</td>\n",
       "      <td>0.000887351</td>\n",
       "      <td>-0.434805</td>\n",
       "      <td>-0.465076</td>\n",
       "      <td>-0.45945</td>\n",
       "      <td>-0.351355</td>\n",
       "      <td>-0.21834</td>\n",
       "      <td>-0.463057</td>\n",
       "      <td>-0.449289</td>\n",
       "      <td>-0.530609</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           plasmid_0  \\\n",
       "GC_sum                                   -0.00418447   \n",
       "chain   [15, 39, 15, 49, 24, 46, 27, 36, 26, 46, 25]   \n",
       "gd_sum                                     -0.455719   \n",
       "total                                      -0.459904   \n",
       "\n",
       "                                               plasmid_1 plasmid_10  \\\n",
       "GC_sum                                          0.027847 -0.0394678   \n",
       "chain   [36, 26, 46, 25, 52, 31, 48, 84, 49, 24, 46, 27]       [39]   \n",
       "gd_sum                                         -0.451718  -0.459794   \n",
       "total                                          -0.423871  -0.499262   \n",
       "\n",
       "                 plasmid_11          plasmid_12          plasmid_13  \\\n",
       "GC_sum           8.2781e-05         -0.00164105          0.00114274   \n",
       "chain   [19, 75, 35, 7, 43]  [1, 75, 35, 6, 76]  [4, 75, 35, 3, 58]   \n",
       "gd_sum          -0.00140908        -0.000331057        -0.000255386   \n",
       "total            -0.0013263          -0.0019721         0.000887351   \n",
       "\n",
       "               plasmid_2  plasmid_3     plasmid_4  plasmid_5 plasmid_6  \\\n",
       "GC_sum         0.0318185 -0.0541946    -0.0073864  0.0290831  0.017824   \n",
       "chain   [36, 26, 46, 27]       [16]  [13, 48, 30]   [52, 23]      [28]   \n",
       "gd_sum         -0.466624  -0.410882     -0.452064  -0.380438 -0.236164   \n",
       "total          -0.434805  -0.465076      -0.45945  -0.351355  -0.21834   \n",
       "\n",
       "        plasmid_7     plasmid_8  plasmid_9  \n",
       "GC_sum  0.0369434    -0.0190562 -0.0306087  \n",
       "chain        [48]  [14, 37, 14]       [37]  \n",
       "gd_sum       -0.5     -0.430233       -0.5  \n",
       "total   -0.463057     -0.449289  -0.530609  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_2  from MILP completely contained in  plasmid_1  from greedy.\n",
      "\n",
      "plasmid_2  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_4  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_4  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_5  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_5  from MILP completely contained in  plasmid_5  from greedy.\n",
      "\n",
      "plasmid_5  from greedy completely contained in  plasmid_5  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_8  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_9  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_10  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_10  from greedy completely contained in  plasmid_4  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
