{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 117 with $\\alpha_1 = 1$, $\\alpha_2 = 1$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 117\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012934.1</td>\n",
       "      <td>3372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012932.1</td>\n",
       "      <td>2096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012933.1</td>\n",
       "      <td>3319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012935.1</td>\n",
       "      <td>3905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012936.1</td>\n",
       "      <td>98999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CP012931.1</td>\n",
       "      <td>236176</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP012934.1         3372\n",
       "1  CP012932.1         2096\n",
       "2  CP012933.1         3319\n",
       "3  CP012935.1         3905\n",
       "4  CP012936.1        98999\n",
       "5  CP012931.1       236176"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt0VeWZx/HfkwQCAUSBEK6SVAmQ\nBAIkggKWoCgwVbALL9Bqq0JxtIxaaJWZadVaW6sjdYojFey41EoR8EJpZQTaIcVrJSjIzUBEbiIC\nAoFwzeWZP7LDxJCEg2bnAt/PWi7P3vs9737O8bzm/M5+997m7gIAAACAqLouAAAAAED9QDgAAAAA\nIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACAQU9cFnK42\nbdp4YmJiXZfRYBw6dEjNmjWr6zKAeo+xAkSGsXJ6VqxYscfd4+u6DiBSDS4cJCYmKicnp67LaDCy\ns7OVlZVV12UA9R5jBYgMY+X0mNmWuq4BOB1MKwIAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgA\nAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIBBaODCzZ8xsl5mtqWK7mdk0M8szsw/NrG9YtQAAAAA4\ntTCPHDwraXg120dI6hr8M0HS70KsBQAAAMAphBYO3H2ZpL3VNBkl6Xkv9a6kc82sfVj1AAAAAKhe\nTB3uu6OkbeWWtwfrPqvY0MwmqPToghISEpSdnV0b9Z0RCgoKeL8QmtWf5td1CTUmqWU0YwWIAH9X\ngDNbXYYDq2SdV9bQ3WdKmilJmZmZnpWVFWJZZ5bs7GzxfiEsN095ra5LqDHPDm/GWEFoEs+osdKc\nsQKcweoyHGyX1LnccidJO+qoli85s/4n3qyuSwAAAEADUZeXMl0g6XvBVYsulpTv7idNKQIAAABQ\nO0I7cmBmsyVlSWpjZtsl3S+pkSS5+1OSFkr6J0l5kg5LuiWsWgAAAACcWmjhwN3HnmK7S/phWPsH\nAAAAcHq4QzIAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQD\nAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMA\nAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAA\nAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAA\nAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAA\nAcIBAAAAAEmEAwAAAACBUMOBmQ03s1wzyzOzKZVsP9/MlprZB2b2oZn9U5j1AAAAAKhaaOHAzKIl\nPSlphKQUSWPNLKVCs59KmuvufSSNkTQ9rHoAAAAAVC/MIwf9JOW5+yZ3Py7pRUmjKrRxSecEj1tK\n2hFiPQAAAACqERNi3x0lbSu3vF1S/wptHpC02Mz+RVIzSUNDrAcAAABANcIMB1bJOq+wPFbSs+4+\n1cwukfQHM0tz95IvdWQ2QdIESUpISFB2dnYY9Z4wuWdRqP3XpoKCgtDfL5y9GCtAZBgrABqKMMPB\ndkmdyy130snThsZJGi5J7v6OmTWR1EbSrvKN3H2mpJmSlJmZ6VlZWSGVXOrmKa+F2n9tenZ4M4X9\nfuHsxVgBIsNYAdBQhHnOwXJJXc0sycwaq/SE4wUV2myVdLkkmVkPSU0k7Q6xJgAAAABVCC0cuHuR\npImSFklar9KrEq01swfNbGTQbLKkH5jZKkmzJd3s7hWnHgEAAACoBWFOK5K7L5S0sMK6+8o9Xidp\nYJg1AAAAAIgMd0gGAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAA\nIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAg\niXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJ\ncAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlw\nAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAA\nAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAICnkcGBmw80s18zyzGxKFW2uN7N1ZrbWzP4YZj0AAAAA\nqhYTVsdmFi3pSUlXSNouabmZLXD3deXadJX0r5IGuvs+M2sbVj0AAAAAqhfmkYN+kvLcfZO7H5f0\noqRRFdr8QNKT7r5Pktx9V4j1AAAAAKhGaEcOJHWUtK3c8nZJ/Su0SZYkM3tLUrSkB9z99YodmdkE\nSRMkKSEhQdnZ2WHUe8LknkWh9l+bCgoKQn+/cPZirACRYawAaCjCDAdWyTqvZP9dJWVJ6iTpDTNL\nc/f9X3qS+0xJMyUpMzPTs7KyarzY8m6e8lqo/demZ4c3U9jvF85ejBUgMowVAA1FmNOKtkvqXG65\nk6QdlbT5k7sXuvsnknJVGhYAAAAA1LIww8FySV3NLMnMGksaI2lBhTbzJQ2RJDNro9JpRptCrAkA\nAABAFUILB+5eJGmipEWS1kua6+5rzexBMxsZNFsk6QszWydpqaSfuPsXYdUEAAAAoGphnnMgd18o\naWGFdfeVe+ySJgX/AAAAAKhD3CEZAAAAgCTCAQAAAIAA4QAAAACApNMIB2Y2yMxuCR7Hm1lSeGUB\nAAAAqG0RhQMzu1/SvZL+NVjVSNILYRUFAAAAoPZFerWib0vqI+l9SXL3HWbWIrSqUHM+Wyk9MKqu\nq6gZD+TXdQUAAABntEinFR0PLjvqkmRmzcIrCQAAAEBdiPTIwVwzmyHpXDP7gaRbJT0dXlkAAKBe\n4og0cEaLKBy4+2NmdoWkA5K6SbrP3ZeEWhkAAACAWnXKcGBm0ZIWuftQSQQCAAAA4Ax1ynDg7sVm\ndtjMWro7x98AnJmYKgEAQMTnHByVtNrMlkg6VLbS3e8MpSoAAAAAtS7ScPBa8A8AAACAM1SkJyQ/\nZ2aNJSUHq3LdvTC8sgAAAADUtojCgZllSXpO0mZJJqmzmX3f3ZeFVxoAAACA2hTptKKpkq5091xJ\nMrNkSbMlZYRVGAAAAIDaFekdkhuVBQNJcvcNkhqFUxIAAACAuhDpkYMcM/tvSX8Ilr8raUU4JQEA\nAACoC5GGg9sl/VDSnSo952CZpOlhFQUAAACg9kUaDmIk/dbdfyOduGtybGhVAQAAnKFWrFjRNiYm\n5veS0hT5FG+gJpRIWlNUVDQ+IyNjV2UNIg0Hf5M0VFJBsNxU0mJJA752iQAAAGeRmJiY37dr165H\nfHz8vqioKK/renD2KCkpsd27d6fs3Lnz95JGVtYm0rTaxN3LgoGCx3E1UCMAAMDZJi0+Pv4AwQC1\nLSoqyuPj4/NVetSq8jYR9nXIzPqWLZhZpqQjX7M+AACAs1EUwQB1JfjsVZkBIp1WdJekeWa2Q5JL\n6iDphq9fHgAAAID6ItJwkCSpj6TzJX1b0sUqDQkAAAD4GhKnvFajN5Xd/Otv1ZvLzffr16/bY489\ntu2b3/zm4Ujal5SUaMCAAckLFy7Ma9WqVUnY9X1df/nLX1pMnTo1YenSpXlh9D9r1qyWa9eubfqr\nX/1qZ8VtcXFxfQ4fPvxBVc8dMGBA8p/+9KeP4+Pji09nn5FOK/qZux+QdK6kKyTNlPS709kRAAAA\nGo7CwsJa3+fcuXNbpqamHmkIwaA2fPe7382vLBhEYuzYsV889thj8af7vEjDQVni+Jakp9z9T5Ia\nn+7OAAAAUPd+8pOftE9KSkodMGBA16uvvjrpvvvuS5BKf+mfOHFix4suuqjbQw89lLBjx46YYcOG\nXZCWltYjLS2tx+LFi5tJ0oEDB6Kuu+66xLS0tB49evRIeeGFF86VpIKCArvqqqu+kZycnPKtb33r\nG0ePHjVJevzxx9uMGzeuc9n+p06d2mb8+PGdKtY1a9asVt/+9rf3S1Jubm7jpKSk1BtuuKFL165d\nU0eOHJk0f/78Fn379u3epUuXtKVLl8ZVV0tubm7jjIyMbikpKT1SUlJ6LFmypJlU+mt/v379ug0f\nPvwbSUlJqSNHjkwqKak+i4wePTrxO9/5zvkZGRndEhMT02bPnt2yYpulS5fG9enTp3uPHj1S+vTp\n033VqlWxkpSTk9OkZ8+ePbp3756SnJycsnr16thIX9u0adNaf+973ztfkj766KPGvXv37p6Wltbj\nrrvu6lC23y1btjTKzMzs1r1795SuXbumvv76680lacyYMftfeeWV1hF+JE6IdFrRp2Y2Q6WXM33E\nzGLFdXkBAAAanGXLlsX9+c9/Pm/16tXrCgsLrXfv3il9+vQ5Me1n//790cuXL8+VpKuvvjpp0qRJ\nnw8bNqxg48aNjYcNG9Z106ZNa//t3/6t/ZAhQw7Mmzdv8549e6IzMzN7jBw58sBvfvOb+KZNm5Zs\n2LBh3T/+8Y+mAwcOTJGkcePG7U1NTU05duzY9tjYWH/hhRfazJgxY0vF2lasWNF84MCBJ9Zv27at\nyZw5czZlZGRs6dWrV49Zs2a1zsnJ+eiPf/zjub/85S/bDxky5OOqaunQoUPRG2+8sSEuLs5Xr14d\nO3bs2G+sWbNmvSStX7++6cqVKzclJiYWZmRkdF+yZEnzYcOGFVSsp7xt27bFvvfee7nr1q2LHTp0\naLdRo0atLr89PT396HvvvfdRo0aNNH/+/Bb33HNPp0WLFn38xBNPxN9xxx2f33777XuPHj1qRUVF\n+vTTTxtF8trK93/HHXecP378+N0TJ0784uGHHz5xROCZZ55pdfnll+c/8sgjO4uKinTw4MEoSYqP\njy8+fvy47dy5M7pdu3YRTy2KNBxcL2m4pMfcfb+ZtZf0k0h3AgAAgPohOzu7+YgRI/Y3b97cJfkV\nV1yxv/z2sWPH7i17/NZbb52zcePGpmXLBQUF0fv27YvKzs4+Z9GiRedOmzatnSQdO3bM8vLyGr/5\n5pvN77zzzl2S1L9//yPJycmHJemcc84pGThw4ME5c+a07Nmz59HCwkLr16/fSVe+zM/PjznvvPNO\n/IzfsWPHY2XtkpOTj1x22WUHoqKi1Ldv38MPPfRQh+D1VFpLly5dCseNG9dl3bp1TaOiorRly5YT\nN/Dt2bPnoQsuuKBQklJTUw9//PHHp5wRM3r06L3R0dHq2bPnsc6dOx9buXJlk/Lb9+7dG33DDTck\nbd68uYmZeWFhoUnSJZdccuixxx5rv3379sZjxozZ17Nnz2ORvrby3n///eb/8z//87Ek3XbbbV/8\n4he/6CRJF1988aHbbrstsbCwMOraa6/dN2DAgBPva+vWrYu2bt3auF27dhFfZTSicODuhyW9Um75\nM0mfRboTAAAA1A/u1V9TpkWLFie+nLu7cnJy1gdB4kt9vPTSS3np6enHKj7fzCrtd8KECXt++ctf\ntktOTj5644037qmsTXR0tBcXFys6OlqS1Lhx4xP7jYqKUpMmTTxop+LiYquulkmTJnVo27Zt4csv\nv/xJSUmJmjZteuLE79jY2BP9RkdHq6ioqPKiq3ldFZfvvffejoMHDz64ZMmSj3Nzcxtfdtll3STp\nn//5n/deeumlh1599dWWI0aMSJ4+ffrmbt26HYvktVVU2SVwR4wYUbBs2bLcl19+ueXNN9+cdOed\nd34+ceLEL6TSoBQXF3da528wNQgAAOAskpWVVbBo0aKWhw8ftvz8/Ki//vWv51bVdtCgQQceeeSR\ntmXLb7/9dlNJGjJkyIGpU6cmlM3Vf+utt5oG7QteeOGFVpK0fPnyJhs2bDhx09zLLrvs0Geffdb4\n1VdfbT1u3Li9qkRSUtLR9evXx1a2rSpV1ZKfnx/dvn37wujoaE2fPr11cfGpZ9b88Ic/7Pj8889X\n+n688sor5xUXF2vt2rWx27Zti01PTz9afvuBAweiO3XqdFySZsyY0aZs/bp16xr36NHj2E9/+tNd\nV1555f6VK1c2rdh3JPr27Vvw9NNPt5Kkp59++sS5BBs2bGjcsWPHwsmTJ++58cYb97z//vtxUumV\nn3bv3t2oW7duJwW46kQ6rQgAAAAhqO1Ljw4ePPjw8OHD81NSUlI7dux4rFevXodatmxZ6TfnmTNn\nbhs/fvz5ycnJKcXFxda/f/+DAwYM2PrrX/96x4QJE87v3r17irtbp06dji1dujTvxz/+8a4xY8Yk\nJScnp6Smph7u2bPnofL9XXPNNfs+/PDDuKour3nllVfmL168uEVaWlrEX2irquXuu+/eNXr06Avm\nz59/3qBBgw42bdr0lL+gr1u3rmnZCdEVXXjhhcf69evX7Ysvvmj0n//5n1vi4uK+9Cv+vffeu3P8\n+PFJ06ZNa3fppZceKFv/hz/8odW8efNax8TEeHx8fOHDDz+8Y//+/dGRvr4y06dP3zpmzJhvTJ8+\nPWHkyJH7ytYvWrSoxbRp09rFxMR4XFxc8axZsz6RpDfffDOuT58+hxo1anRa+7FTHVqqbzIzMz0n\nJyfUfSROeS3U/mvTs+kblZV7f12XUTMeyK/rClABY6WeYqzUO4yVeqoWxoqZrXD3zPLrVq1atTk9\nPb3SaTW1JT8/P6ply5YlBw8ejLrkkku6PfXUU1sGDRoU0b0Ivo4hQ4ZcePfdd38+atSog5Vt37Jl\nS6OxY8cmvv322xvDrqUygwYN6vrmm2+etO/Ro0cnXnXVVfm33HLLvsqeVx/dcsstna+55pr9lb3X\nq1atapOenp5Y2fOYVgQAAHCWufHGG7t07949pVevXj2uvvrqfWEHgz179kQnJiamNWnSpKSqYCBJ\nXbp0Kbz11lv37N27t06+o1YWDBqqtLS0I9W911VhWhEAAMBZ5s9//vMntbm/Nm3aFG/evHlNJG3H\njx9f736df/nllzfXdQ2na/LkyV/p6BRHDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIcEIyAABA\nXXqgZcapG51Of/mnvG+CmWWMGjVq7/z58z+RpMLCQrVt2za9d+/eh5YuXZo3bdq01jk5Oc2ef/75\nrZMmTerQvHnz4gcffPDz8n1ER0dndO3a9UhxcbFdeOGFR+bOnbu5/N2VpdIbcQ0YMCB54cKFea1a\ntSq57rrrEv/2t7+1bN26ddHGjRvXVlbbBx980OSWW25JXLduXdyUKVM+rbjfMr/61a/in3rqqYRt\n27bF7tixY1X79u2LJGn27Nkt33vvvWaPP/74jkjfMvw/jhwAAACcZZo2bVqSm5vbtKCgwCTp1Vdf\nPSchIaHwdPqIjY0t+eijj9Zt3LhxbaNGjXzq1KnxFdvMnTu3ZWpq6pFWrVqVSNKtt966Z8GCBdVe\nLrRt27ZFv/3tb7fedtttlYaCMoMHDy5YsmTJhg4dOhwvv/6GG27If/311889ePAg33O/At40AACA\ns9Dll1+eP2/evHMlafbs2a1Gjx6996v2NWjQoIK8vLzYiutnzZrVqvwdh0eMGFEQHx9fVF1fHTt2\nLBo8ePDhRo0aVXun3oEDBx7p1q3b8Yrro6KiNGDAgINz5sxpeTqvAaUIBwAAAGehm266ae+cOXPO\nO3z4sK1fvz7ukksuOfRV+iksLNSiRYvO6dmz55GK21asWNF84MCBp+z30UcfjX/00UdPOvJQ0eDB\ngy/cvHlzo1O1y8zMPPTGG280P1U7nCzUcw7MbLik30qKlvR7d/91Fe2ulTRP0kXunhNmTQAAAJD6\n9+9/ZPv27bFPP/10q6FDh+af7vOPHTsW1b1795Sgr4N33XXXSTfdys/PjznvvPNKTn72l91zzz27\nI9nn3//+97xI2rVr165o586djSNpiy8LLRyYWbSkJyVdIWm7pOVmtsDd11Vo10LSnZL+EVYtAAAA\nONnw4cP333///Z0XL16cu2vXrtP6Xlh2zkF1baKjo724uFjR0dFfr9DTdOTIEWvSpMkpQwlOFua0\non6S8tx9k7sfl/SipFGVtPuFpEclHQ2xFgAAAFRw++2375k8efKOfv36nTQlqCYkJSUdXb9+/Unn\nIoQtNze3SWpqaiiv6UwX5rSijpK2lVveLql/+QZm1kdSZ3f/i5n9uKqOzGyCpAmSlJCQoOzs7Jqv\ntpzJPas9T6ZBKYjtoOxuP6/rMmpGyP/dcfoYK/UUY6XeYazUU/VlrERw6dGwXHDBBYU/+9nPdp2q\n3eOPP95+xowZCWXLn3/++YeR9H/llVfmL168uEVaWtoxSbr66quT3n333Rb79u2LSUhI6DVlypQd\nP/rRj/aUnW9wzz337N66dWvMRRddlHLo0KFoM/MZM2YkrF+/fk2rVq1KBg8efOFzzz23JTExsfCh\nhx5q+8QTT7T74osvGqWnp6cMGTIkf86cOVskadmyZS0eeeSRT7/au3J2M/dqTwT/6h2bXSdpmLuP\nD5ZvktTP3f8lWI6S9L+Sbnb3zWaWLenHpzrnIDMz03Nywj0tIXHKa6H2X5ueTd+orNz767qMmvHA\naU+HRMgYK/UUY6XeYazUU7UwVsxshbtnll+3atWqzenp6SfNzz8TbdmypdHYsWMT33777WovX1qT\ntm3bFnP99dd/45133tlQW/tsaFatWtUmPT09sbJtYU4r2i6pc7nlTpLK34yihaQ0SdlmtlnSxZIW\nmNmXBhAAAAAapi5duhTeeuute/bu3VtrV8jctGlT46lTp247dUtUJsxpRcsldTWzJEmfShoj6Ttl\nG909X1KbsuVIjxwAAACg4Rg/fvy+2tzf4MGDD9fm/s40oaU4dy+SNFHSIknrJc1197Vm9qCZjQxr\nvwAAAPVcSUlJidV1ETg7BZ+9Kq/kFOp9Dtx9oaSFFdbdV0XbrDBrAQAAqCfW7N69OyU+Pj4/Kioq\nnJM/gUqUlJTY7t27W0paU1WbUMMBAAAAvqyoqGj8zp07f79z5840hXv+J1BRiaQ1RUVF46tqQDgA\nAACoRRkZGbskMcUa9RJpFQAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADh\nAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEA\nAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAA\nAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAA\nAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAA\ngCTCAQAAAIAA4QAAAACApJDDgZkNN7NcM8szsymVbJ9kZuvM7EMz+5uZdQmzHgAAAABVCy0cmFm0\npCcljZCUImmsmaVUaPaBpEx37yXpJUmPhlUPAAAAgOqFeeSgn6Q8d9/k7sclvShpVPkG7r7U3Q8H\ni+9K6hRiPQAAAACqYe4eTsdm10oa7u7jg+WbJPV394lVtP8vSTvd/aFKtk2QNEGSEhISMl588cVQ\nai6z+tP8UPuvTUlNj6n5sR11XUbNaN+7ritABYyVeoqxUu8wVuqpWhgrQ4YMWeHumaHvCKghMSH2\nbZWsqzSJmNmNkjIlDa5su7vPlDRTkjIzMz0rK6uGSqzczVNeC7X/2vRs+ifKyr2/rsuoGWPPnD+u\nZwrGSj3FWKl3GCv1FGMFOEmY4WC7pM7lljtJOumnBjMbKunfJQ1292Mh1gMAAACgGmGec7BcUlcz\nSzKzxpLGSFpQvoGZ9ZE0Q9JId98VYi0AAAAATiG0cODuRZImSlokab2kue6+1sweNLORQbP/kNRc\n0jwzW2lmC6roDgAAAEDIwpxWJHdfKGlhhXX3lXs8NMz9AwAAAIgcd0gGAAAAIIlwAAAAACBAOAAA\nAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAA\nACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAA\nIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAg\niXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJ\ncAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgEGo4MLPhZpZr\nZnlmNqWS7bFmNifY/g8zSwyzHgAAAABVCy0cmFm0pCcljZCUImmsmaVUaDZO0j53v1DS45IeCase\nAAAAANUL88hBP0l57r7J3Y9LelHSqAptRkl6Lnj8kqTLzcxCrAkAAABAFcIMBx0lbSu3vD1YV2kb\ndy+SlC+pdYg1AQAAAKhCTIh9V3YEwL9CG5nZBEkTgsUCM8v9mrWdNYZIbSTtqes6asTPOaiE8DBW\ngMgwVk5bl9rYCVBTwgwH2yV1LrfcSdKOKtpsN7MYSS0l7a3YkbvPlDQzpDrPaGaW4+6ZdV0HUN8x\nVoDIMFaAM1uY04qWS+pqZklm1ljSGEkLKrRZIOn7weNrJf2vu5905AAAAABA+EI7cuDuRWY2UdIi\nSdGSnnH3tWb2oKQcd18g6b8l/cHM8lR6xGBMWPUAAAAAqJ7xQ/2ZzcwmBNOyAFSDsQJEhrECnNkI\nBwAAAAAkhXyHZAAAAAANB+EAAAAAgCTCwVnPzBLNbE3wOMvM/lLXNQF1wczePsX2hWZ2bm3VA5wJ\nzCzbzDKDx5vNrE1d1wSgemF4WsrjAAAEf0lEQVTe5wAhMjNT6TkjJXVdC1DfmFm0uxefznPcfcAp\ntv/T16sKOPPwtwg483DkoAEJfuVfb2bTJb0v6SYze8fM3jezeWbWPGh3kZm9bWarzOw9M2sRPPeN\noO37ZlbtFyGgvgo+yx+Z2XNm9qGZvWRmccGvkveZ2ZuSrjOzC8zsdTNbEXz2uwfPTzCzV4Pxsaps\nLJhZQfDv9ma2zMxWmtkaM7s0WH/iV08zmxRsW2Nmd5era72ZPW1ma81ssZk1rZM3CahGxc+vmT1i\nZneU2/6AmU0OHv/EzJYHY+3nwbqKf4s6m9nvzCwn+Oz/vG5eGYCaQDhoeLpJel7SFZLGSRrq7n0l\n5UiaFNxwbo6ku9w9XdJQSUck7ZJ0RdD2BknT6qJ4oIZ0kzTT3XtJOiCp7IvNUXcf5O4vqvSu6v/i\n7hmSfixpetBmmqS/B+Ojr6S1Ffr+jqRF7t5bUrqkleU3mlmGpFsk9Zd0saQfmFmfYHNXSU+6e6qk\n/ZJG19QLBmpCZZ9fSS+q9O9CmeslzTOzK1X6me4nqbekDDP7ZtCmm6Tn3b2Pu2+R9O/BXZN7SRps\nZr1q5QUBqHFMK2p4trj7u2Z2laQUSW+VHtVVY0nvqPR/2J+5+3JJcvcDkmRmzST9l5n1llQsKbku\nigdqyDZ3fyt4/IKkO4PHcyQpOIo2QKVfcMqeExv8+zJJ35OkYOpRfoW+l0t6xswaSZrv7isrbB8k\n6VV3PxTs6xVJl6r0ju+flGu/QlLi13iNQBiq+vy2NbMOkuIl7XP3rWZ2p6QrJX0QPLe5SsPCVgV/\ni8r1e72ZTVDp94r2Kv379GFtvCAANYtw0PAcCv5tkpa4+9jyG4Nfayq7ecWPJH2u0l9CoyQdDbNI\nIGQVP+Nly2XjI0rS/uDX/9Pr2H1Z8Ovot1R6B/f/cPfnyzWxKp4qScfKPS6WxLQi1DdVfX5fknSt\npHYqPZJQ1vZhd5/xpQ7MEvX/Y01mlqTSo3MXufs+M3tWUpMarRpArWFaUcP1rqSBZnahJAVzrpMl\nfSSpg5ldFKxvYWYxklqq9IhCiaSbJEXXUd1ATTjfzC4JHo+V9Gb5jcERs0/M7Dqp9KRJM0sPNv9N\n0u3B+mgzO6f8c82si6Rd7v60pP9W6dSj8pZJuiYYc80kfVvSGzX30oBQVfX5fVHSGJUGhJeCtosk\n3VrufLaOZta2kj7PUWlYyDezBEkjQn4NAEJEOGig3H23pJslzTazD1UaFrq7+3GVzh19wsxWSVqi\n0l9wpkv6vpm9q9IpRYcq7RhoGNar9PP8oaRWkn5XSZvvShoXjIO1kkYF6++SNMTMVqt06k9qhedl\nSVppZh+o9JyB35bf6O7vS3pW0nuS/iHp9+7+gYAGoKrPr7uvldRC0qfu/lnQdrGkP0p6JxgvLwVt\nKva5SqVTj9ZKekbSWxXbAGg4zL2yGSgAUD8FUxr+4u5pdVwKAABnHI4cAAAAAJDEkQMAAAAAAY4c\nAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQJP0fZxuOEaA6Mf4AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f14fa6bbc88>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:1:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>194757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>92126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>70104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>140574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>13397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>5752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>64580</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0       194757\n",
       "1  plasmid_1        92126\n",
       "2  plasmid_2        70104\n",
       "3  plasmid_3       140574\n",
       "4  plasmid_4        13397\n",
       "5  plasmid_5         5752\n",
       "6  plasmid_6        64580"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.559786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.541724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.550607</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.559786\n",
       "1  precision  0.541724\n",
       "2   f1_score  0.550607"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP012931.1</th>\n",
       "      <th>CP012932.1</th>\n",
       "      <th>CP012933.1</th>\n",
       "      <th>CP012934.1</th>\n",
       "      <th>CP012935.1</th>\n",
       "      <th>CP012936.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.245152</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.010416</td>\n",
       "      <td>0.040076</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.920595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.114347</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.252341</td>\n",
       "      <td>0.040076</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.458197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.029110</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.085112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>0.023034</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>0.093447</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.767847</td>\n",
       "      <td>0.080153</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.463904</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP012931.1  CP012932.1  CP012933.1  CP012934.1  CP012935.1  \\\n",
       "0  plasmid_0    0.245152    0.000000         0.0         0.0         0.0   \n",
       "1  plasmid_1    0.010416    0.040076         0.0         0.0         0.0   \n",
       "2  plasmid_2    0.114347    0.000000         0.0         0.0         0.0   \n",
       "3  plasmid_3    0.252341    0.040076         0.0         0.0         0.0   \n",
       "4  plasmid_4    0.029110    0.000000         0.0         0.0         0.0   \n",
       "5  plasmid_5    0.023034    0.000000         0.0         0.0         0.0   \n",
       "6  plasmid_6    0.093447    0.000000         0.0         0.0         0.0   \n",
       "7      Total    0.767847    0.080153         0.0         0.0         0.0   \n",
       "\n",
       "   CP012936.1  \n",
       "0    0.000000  \n",
       "1    0.920595  \n",
       "2    0.000000  \n",
       "3    0.458197  \n",
       "4    0.085112  \n",
       "5    0.000000  \n",
       "6    0.000000  \n",
       "7    1.463904  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012931.1</td>\n",
       "      <td>0.280611</td>\n",
       "      <td>0.008901</td>\n",
       "      <td>0.381975</td>\n",
       "      <td>0.413163</td>\n",
       "      <td>0.415093</td>\n",
       "      <td>0.940195</td>\n",
       "      <td>0.324868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012932.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000912</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000598</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012933.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012934.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012935.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CP012936.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.980016</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.317904</td>\n",
       "      <td>0.570799</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.280611</td>\n",
       "      <td>0.989829</td>\n",
       "      <td>0.381975</td>\n",
       "      <td>0.731664</td>\n",
       "      <td>0.985892</td>\n",
       "      <td>0.940195</td>\n",
       "      <td>0.324868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2  plasmid_3  plasmid_4  \\\n",
       "0  CP012931.1   0.280611   0.008901   0.381975   0.413163   0.415093   \n",
       "1  CP012932.1   0.000000   0.000912   0.000000   0.000598   0.000000   \n",
       "2  CP012933.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "3  CP012934.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "4  CP012935.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "5  CP012936.1   0.000000   0.980016   0.000000   0.317904   0.570799   \n",
       "6       Total   0.280611   0.989829   0.381975   0.731664   0.985892   \n",
       "\n",
       "   plasmid_5  plasmid_6  \n",
       "0   0.940195   0.324868  \n",
       "1   0.000000   0.000000  \n",
       "2   0.000000   0.000000  \n",
       "3   0.000000   0.000000  \n",
       "4   0.000000   0.000000  \n",
       "5   0.000000   0.000000  \n",
       "6   0.940195   0.324868  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.280611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.988917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.381975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.731067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.985892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>0.940195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>0.324868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.280611\n",
       "1  plasmid_1    0.988917\n",
       "2  plasmid_2    0.381975\n",
       "3  plasmid_3    0.731067\n",
       "4  plasmid_4    0.985892\n",
       "5  plasmid_5    0.940195\n",
       "6  plasmid_6    0.324868"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012931.1</td>\n",
       "      <td>0.407366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012932.1</td>\n",
       "      <td>0.040076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012933.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012934.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012935.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CP012936.1</td>\n",
       "      <td>0.994323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP012931.1    0.407366\n",
       "1  CP012932.1    0.040076\n",
       "2  CP012933.1    0.000000\n",
       "3  CP012934.1    0.000000\n",
       "4  CP012935.1    0.000000\n",
       "5  CP012936.1    0.994323"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012936.1</td>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.920595</td>\n",
       "      <td>0.980016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction    recall  precision\n",
       "0  CP012936.1  plasmid_1  0.920595   0.980016"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 1, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0338632</td>\n",
       "      <td>-0.0126811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[43, 11, 42, 35, 15]</td>\n",
       "      <td>[49, 17, 62, 16, 47, 43]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.408576</td>\n",
       "      <td>-0.444815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.442439</td>\n",
       "      <td>-0.457496</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   plasmid_0                 plasmid_1\n",
       "GC_sum            -0.0338632                -0.0126811\n",
       "chain   [43, 11, 42, 35, 15]  [49, 17, 62, 16, 47, 43]\n",
       "gd_sum             -0.408576                 -0.444815\n",
       "total              -0.442439                 -0.457496"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_11</th>\n",
       "      <th>plasmid_12</th>\n",
       "      <th>plasmid_13</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0343931</td>\n",
       "      <td>-0.0142622</td>\n",
       "      <td>-0.000326729</td>\n",
       "      <td>-0.000658298</td>\n",
       "      <td>0.00299377</td>\n",
       "      <td>0.00302549</td>\n",
       "      <td>-0.00459884</td>\n",
       "      <td>-0.0146978</td>\n",
       "      <td>-0.0423875</td>\n",
       "      <td>-0.00385593</td>\n",
       "      <td>0.0185021</td>\n",
       "      <td>-0.029692</td>\n",
       "      <td>-0.0251271</td>\n",
       "      <td>-0.0385511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[53, 44, 43, 11, 42, 58, 24, 58]</td>\n",
       "      <td>[35, 42, 15, 35, 28, 43, 20, 43, 54, 53, 26, 53]</td>\n",
       "      <td>[21, 91, 45, 6, 91]</td>\n",
       "      <td>[1, 45, 8, 93]</td>\n",
       "      <td>[45, 4]</td>\n",
       "      <td>[43]</td>\n",
       "      <td>[17, 49, 17, 62, 23]</td>\n",
       "      <td>[62, 23, 62, 16, 47, 16]</td>\n",
       "      <td>[32]</td>\n",
       "      <td>[29]</td>\n",
       "      <td>[33]</td>\n",
       "      <td>[47]</td>\n",
       "      <td>[58]</td>\n",
       "      <td>[49]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.413345</td>\n",
       "      <td>-0.419398</td>\n",
       "      <td>-0.000961448</td>\n",
       "      <td>-0.000237385</td>\n",
       "      <td>-0.000694252</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.455074</td>\n",
       "      <td>-0.437628</td>\n",
       "      <td>-0.308828</td>\n",
       "      <td>-0.386428</td>\n",
       "      <td>-0.236164</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.459794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.447739</td>\n",
       "      <td>-0.43366</td>\n",
       "      <td>-0.00128818</td>\n",
       "      <td>-0.000895682</td>\n",
       "      <td>0.00229952</td>\n",
       "      <td>-0.496975</td>\n",
       "      <td>-0.459673</td>\n",
       "      <td>-0.452326</td>\n",
       "      <td>-0.351215</td>\n",
       "      <td>-0.390284</td>\n",
       "      <td>-0.217662</td>\n",
       "      <td>-0.529692</td>\n",
       "      <td>-0.525127</td>\n",
       "      <td>-0.498345</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               plasmid_0  \\\n",
       "GC_sum                        -0.0343931   \n",
       "chain   [53, 44, 43, 11, 42, 58, 24, 58]   \n",
       "gd_sum                         -0.413345   \n",
       "total                          -0.447739   \n",
       "\n",
       "                                               plasmid_1           plasmid_10  \\\n",
       "GC_sum                                        -0.0142622         -0.000326729   \n",
       "chain   [35, 42, 15, 35, 28, 43, 20, 43, 54, 53, 26, 53]  [21, 91, 45, 6, 91]   \n",
       "gd_sum                                         -0.419398         -0.000961448   \n",
       "total                                           -0.43366          -0.00128818   \n",
       "\n",
       "            plasmid_11   plasmid_12  plasmid_13             plasmid_2  \\\n",
       "GC_sum    -0.000658298   0.00299377  0.00302549           -0.00459884   \n",
       "chain   [1, 45, 8, 93]      [45, 4]        [43]  [17, 49, 17, 62, 23]   \n",
       "gd_sum    -0.000237385 -0.000694252        -0.5             -0.455074   \n",
       "total     -0.000895682   0.00229952   -0.496975             -0.459673   \n",
       "\n",
       "                       plasmid_3  plasmid_4   plasmid_5  plasmid_6 plasmid_7  \\\n",
       "GC_sum                -0.0146978 -0.0423875 -0.00385593  0.0185021 -0.029692   \n",
       "chain   [62, 23, 62, 16, 47, 16]       [32]        [29]       [33]      [47]   \n",
       "gd_sum                 -0.437628  -0.308828   -0.386428  -0.236164      -0.5   \n",
       "total                  -0.452326  -0.351215   -0.390284  -0.217662 -0.529692   \n",
       "\n",
       "        plasmid_8  plasmid_9  \n",
       "GC_sum -0.0251271 -0.0385511  \n",
       "chain        [58]       [49]  \n",
       "gd_sum       -0.5  -0.459794  \n",
       "total   -0.525127  -0.498345  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_7  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_9  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_13  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_13  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
