{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 125 with $\\alpha_1 = 1$, $\\alpha_2 = 10$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 125\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016582.1</td>\n",
       "      <td>2094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016584.1</td>\n",
       "      <td>6647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016583.1</td>\n",
       "      <td>37697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP016585.1</td>\n",
       "      <td>99473</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP016582.1         2094\n",
       "1  CP016584.1         6647\n",
       "2  CP016583.1        37697\n",
       "3  CP016585.1        99473"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt4VeWZ/vH7SQIhAUSBEI6SqAQI\ngYCJoIAleAxTATt4AA8dFYqjZTxhlZlfq9ZqrY5US0daoONYKxUBldKKBeqQeqoVUJCTgaicRAQE\ngoCEHJ7fH1lhYkjCpmZlh/D9XBcXe6317nc9O+yX7Huvd61l7i4AAAAAiIl2AQAAAAAaBsIBAAAA\nAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAACBuGgXcLza\ntm3rKSkp0S7jhHHgwAE1b9482mUADR5jBYgMY+X4LF++fJe7J0W7DiBSJ1w4SElJ0bJly6Jdxgkj\nLy9POTk50S4DaPAYK0BkGCvHx8w2RbsG4HgwrQgAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEA\nAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgEBo4cDMnjazHWa2uobtZmZTzKzAzD4ws7PDqgUAAADA\nsYV55OAZSbm1bB8mqVvwZ7ykX4VYCwAAAIBjCC0cuPvrknbX0mSkpGe93DuSTjWzDmHVAwAAAKB2\ncVHcdydJWyotbw3WfVa1oZmNV/nRBSUnJysvLy/UwlZ9Whhq//UptVVs6D8vnLwa1VhJKFLe809G\nu4y60aFvtCtAFY1qrPB7BWjUohkOrJp1Xl1Dd58uabokZWdne05OTohlSTdMeiXU/uvTM7nNFfbP\nCyevRjVWMj9RTv790S6jboxpPB9EG4tGNVb4vQI0atG8WtFWSV0qLXeWtC1KtQAAAAAnvWiGg/mS\nvhtctehcSYXuftSUIgAAAAD1I7RpRWb2vKQcSW3NbKuk+yU1kSR3/7WkBZL+SVKBpIOSbgyrFgAA\nAADHFlo4cPcxx9jukr4f1v4BAAAAHB/ukAwAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAA\nQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABA\nEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS\n4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLh\nAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEA\nAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAEAg1HBgZrlmlm9mBWY2qZrtp5vZEjN738w+\nMLN/CrMeAAAAADWLC6tjM4uV9JSkiyVtlbTUzOa7+9pKzX4oaba7/8rM0iUtkJQSVk0AAOAb+myF\n9MDIaFdRNx4ojHYFQIMT5pGD/pIK3P1jdz8saZakqv+buKRTgsetJG0LsR4AAAAAtQjtyIGkTpK2\nVFreKmlAlTYPSFpkZv8mqbmki0KsBwAAAEAtwgwHVs06r7I8RtIz7j7ZzM6T9Dszy3D3sq91ZDZe\n0nhJSk5OVl5eXhj1HjGxd0mo/den/fv3h/7zwsmrUY2V+I7K6/7jaJdRNxjzDQ5jpYFirABHCTMc\nbJXUpdJyZx09bWispFxJcve/mVkzSW0l7ajcyN2nS5ouSdnZ2Z6TkxNSyeVumPRKqP3Xp2dymyvs\nnxdOXo1qrGR+opz8+6NdRt0Ywzzqhoax0kAxVoCjhHnOwVJJ3cws1cyaShotaX6VNpslXShJZtZT\nUjNJO0OsCQAAAEANQgsH7l4iaYKkhZLWqfyqRGvM7EEzGxE0myjpe2a2UtLzkm5w96pTjwAAAADU\ngzCnFcndF6j88qSV191X6fFaSYPCrAEAAABAZLhDMgAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAAB\nwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHC\nAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIB\nAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJUly0C0DIPlshPTAy2lXUjQcKo10BAABA\no8aRAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAA\nAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAC\nhAMAAAAAkkIOB2aWa2b5ZlZgZpNqaHOVma01szVm9vsw6wEAAABQs7iwOjazWElPSbpY0lZJS81s\nvruvrdSmm6R/lzTI3feYWbuw6gEAAABQuzCPHPSXVODuH7v7YUmzJI2s0uZ7kp5y9z2S5O47QqwH\nAAAAQC3M3cPp2OwKSbnuPi5Yvl7SAHefUKnNPEnrJQ2SFCvpAXf/czV9jZc0XpKSk5OzZs2aFUrN\nFVZ9Whhq//UpNaFILYq2RbuMutGhb7QrQBWMlQaKsdLgMFYaqHoYK0OHDl3u7tmh7wioI6FNK5Jk\n1ayrmkTiJHWTlCOps6Q3zCzD3fd+7Unu0yVNl6Ts7GzPycmp82Iru2HSK6H2X5+eyfxEOfn3R7uM\nujGm8fxybSwYKw0UY6XBYaw0UIwV4ChhTivaKqlLpeXOkqp+1bBV0h/cvdjdP5GUr/KwAAAAAKCe\nhRkOlkrqZmapZtZU0mhJ86u0mSdpqCSZWVtJaZI+DrEmAAAAADUILRy4e4mkCZIWSlonaba7rzGz\nB81sRNBsoaQvzGytpCWSfuDuX4RVEwAAAICahXnOgdx9gaQFVdbdV+mxS7or+AMAAAAgirhDMgAA\nAABJhAMAAAAAAcIBAAAAAEnHEQ7MbLCZ3Rg8TjKz1PDKAgAAAFDfIgoHZna/pHsl/Xuwqomk58Iq\nCgAAAED9i/TIwXckjZB0QJLcfZuklmEVBQAAAKD+RRoODgeXHXVJMrPm4ZUEAAAAIBoiDQezzWya\npFPN7HuS/iJpRnhlAQAAAKhvEd0Ezd0fN7OLJe2T1F3Sfe6+ONTKAAAAANSrY4YDM4uVtNDdL5JE\nIAAAAAAaqWNOK3L3UkkHzaxVPdQDAAAAIEoimlYk6ZCkVWa2WMEViyTJ3W8LpSoAAAAA9S7ScPBK\n8AcAAABAIxXpCcm/NbOmktKCVfnuXhxeWQAAAADqW0ThwMxyJP1W0kZJJqmLmf2Lu78eXmkAAAAA\n6lOk04omS7rE3fMlyczSJD0vKSuswgAAAADUr0hvgtakIhhIkruvl9QknJIAAAAAREOkRw6Wmdl/\nS/pdsHytpOXhlAQAAAAgGiINB7dI+r6k21R+zsHrkqaGVRQAAACA+hdpOIiT9At3/7l05K7J8aFV\nBQAA0EgtX768XVxc3G8kZSjyKd5AXSiTtLqkpGRcVlbWjuoaRBoOXpN0kaT9wXKCpEWSBn7jEgEA\nAE4icXFxv2nfvn3PpKSkPTExMR7tenDyKCsrs507d6Zv3779N5JGVNcm0rTazN0rgoGCx4l1UCMA\nAMDJJiMpKWkfwQD1LSYmxpOSkgpVftSq+jYR9nXAzM6uWDCzbElffcP6AAAATkYxBANES/DeqzED\nRDqt6HZJc8xsmySX1FHS1d+8PAAAAAANRaThIFVSP0mnS/qOpHNVHhIAAADwDaRMeqVObyq78Wff\nbjCXm+/fv3/3xx9/fMu3vvWtg5G0Lysr08CBA9MWLFhQ0Lp167Kw6/um/vSnP7WcPHly8pIlSwrC\n6H/mzJmt1qxZk/DTn/50e9VtiYmJ/Q4ePPh+Tc8dOHBg2h/+8IePkpKSSo9nn5FOK/qRu++TdKqk\niyVNl/Sr49kRAAAAThzFxcX1vs/Zs2e36tWr11cnQjCoD9dee21hdcEgEmPGjPni8ccfTzre50Ua\nDioSx7cl/drd/yCp6fHuDAAAANH3gx/8oENqamqvgQMHdhs+fHjqfffdlyyVf9M/YcKETuecc073\nhx56KHnbtm1xl1566ZkZGRk9MzIyei5atKi5JO3bty/myiuvTMnIyOjZs2fP9Oeee+5USdq/f79d\ndtllZ6SlpaV/+9vfPuPQoUMmSU888UTbsWPHdqnY/+TJk9uOGzeuc9W6Zs6c2fo73/nOXknKz89v\nmpqa2uvqq6/u2q1bt14jRoxInTdvXsuzzz67R9euXTOWLFmSWFst+fn5TbOysrqnp6f3TE9P77l4\n8eLmUvm3/f379++em5t7Rmpqaq8RI0aklpXVnkVGjRqVcs0115yelZXVPSUlJeP5559vVbXNkiVL\nEvv169ejZ8+e6f369euxcuXKeElatmxZs969e/fs0aNHelpaWvqqVaviI31tU6ZMafPd7373dEn6\n8MMPm/bt27dHRkZGz9tvv71jxX43bdrUJDs7u3uPHj3Su3Xr1uvPf/5zC0kaPXr03pdeeqlNhG+J\nIyKdVvSpmU1T+eVMHzWzeHFdXgAAgBPO66+/nvjHP/7xtFWrVq0tLi62vn37pvfr1+/ItJ+9e/fG\nLl26NF+Shg8fnnrXXXd9fumll+7fsGFD00svvbTbxx9/vOY//uM/OgwdOnTfnDlzNu7atSs2Ozu7\n54gRI/b9/Oc/T0pISChbv3792r///e8JgwYNSpeksWPH7u7Vq1d6UVHR1vj4eH/uuefaTps2bVPV\n2pYvX95i0KBBR9Zv2bKl2QsvvPBxVlbWpj59+vScOXNmm2XLln34+9///tSHH364w9ChQz+qqZaO\nHTuWvPHGG+sTExN91apV8WPGjDlj9erV6yRp3bp1CStWrPg4JSWlOCsrq8fixYtbXHrppfur1lPZ\nli1b4t999938tWvXxl900UXdR44cuary9szMzEPvvvvuh02aNNG8efNa3nPPPZ0XLlz40S9/+cuk\nW2+99fNbbrll96FDh6ykpESffvppk0heW+X+b7311tPHjRu3c8KECV888sgjR44IPP30060vvPDC\nwkcffXR7SUmJvvzyyxhJSkpKKj18+LBt3749tn379hFPLYo0HFwlKVfS4+6+18w6SPpBpDsBAABA\nw5CXl9di2LBhe1u0aOGS/OKLL95befuYMWN2Vzx+6623TtmwYUNCxfL+/ftj9+zZE5OXl3fKwoUL\nT50yZUp7SSoqKrKCgoKmb775ZovbbrtthyQNGDDgq7S0tIOSdMopp5QNGjToyxdeeKFV7969DxUX\nF1v//v2PuvJlYWFh3GmnnXbka/xOnToVVbRLS0v76oILLtgXExOjs88+++BDDz3UMXg91dbStWvX\n4rFjx3Zdu3ZtQkxMjDZt2nTkBr69e/c+cOaZZxZLUq9evQ5+9NFHx5wRM2rUqN2xsbHq3bt3UZcu\nXYpWrFjRrPL23bt3x1599dWpGzdubGZmXlxcbJJ03nnnHXj88cc7bN26teno0aP39O7duyjS11bZ\ne++91+LVV1/9SJJuvvnmL37yk590lqRzzz33wM0335xSXFwcc8UVV+wZOHDgkZ9rmzZtSjZv3ty0\nffv2EV9lNKJw4O4HJb1UafkzSZ9FuhMAAAA0DO61X1OmZcuWRz6cu7uWLVu2LggSX+tj7ty5BZmZ\nmUVVn29m1fY7fvz4XQ8//HD7tLS0Q9ddd92u6trExsZ6aWmpYmNjJUlNmzY9st+YmBg1a9bMg3Yq\nLS212mq56667OrZr1674xRdf/KSsrEwJCQlHTvyOj48/0m9sbKxKSkqqL7qW11V1+d577+00ZMiQ\nLxcvXvxRfn5+0wsuuKC7JP3rv/7r7vPPP//Ayy+/3GrYsGFpU6dO3di9e/eiSF5bVdVdAnfYsGH7\nX3/99fwXX3yx1Q033JB62223fT5hwoQvpPKglJiYeFznbzA1CAAA4CSSk5Ozf+HCha0OHjxohYWF\nMX/5y19Orant4MGD9z366KPtKpbffvvtBEkaOnTovsmTJydXzNV/6623EoL2+5977rnWkrR06dJm\n69evP3LT3AsuuODAZ5991vTll19uM3bs2N2qRmpq6qF169bFV7etJjXVUlhYGNuhQ4fi2NhYTZ06\ntU1p6bFn1nz/+9/v9Oyzz1b783jppZdOKy0t1Zo1a+K3bNkSn5mZeajy9n379sV27tz5sCRNmzat\nbcX6tWvXNu3Zs2fRD3/4wx2XXHLJ3hUrViRU7TsSZ5999v4ZM2a0lqQZM2YcOZdg/fr1TTt16lQ8\nceLEXdddd92u9957L1Eqv/LTzp07m3Tv3v2oAFebSKcVAQAAIAT1fenRIUOGHMzNzS1MT0/v1alT\np6I+ffocaNWqVbWfnKdPn75l3Lhxp6elpaWXlpbagAEDvhw4cODmn/3sZ9vGjx9/eo8ePdLd3Tp3\n7ly0ZMmSgrvvvnvH6NGjU9PS0tJ79ep1sHfv3gcq93f55Zfv+eCDDxJrurzmJZdcUrho0aKWGRkZ\nEX+gramWO+64Y8eoUaPOnDdv3mmDBw/+MiEh4ZjfoK9duzah4oToqs4666yi/v37d//iiy+aPPnk\nk5sSExO/9i3+vffeu33cuHGpU6ZMaX/++efvq1j/u9/9rvWcOXPaxMXFeVJSUvEjjzyybe/evbGR\nvr4KU6dO3Tx69Ogzpk6dmjxixIg9FesXLlzYcsqUKe3j4uI8MTGxdObMmZ9I0ptvvpnYr1+/A02a\nNDmu/dixDi01NNnZ2b5s2bJQ95Ey6ZVQ+69Pz2RuUE7+/dEuo248UBjtClAFY6WBYqw0OIyVBqoe\nxoqZLXf37MrrVq5cuTEzM7PaaTX1pbCwMKZVq1ZlX375Zcx5553X/de//vWmwYMHR3Qvgm9i6NCh\nZ91xxx2fjxw58svqtm/atKnJmDFjUt5+++0NYddSncGDB3d78803j9r3qFGjUi677LLCG2+8cU91\nz2uIbrzxxi6XX3753up+1itXrmybmZmZUt3zmFYEAABwkrnuuuu69ujRI71Pnz49hw8fvifsYLBr\n167YlJSUjGbNmpXVFAwkqWvXrsU33XTTrt27d0flM2p1weBElZGR8VVtP+uaMK0IAADgJPPHP/7x\nk/rcX9u2bUs3bty4OpK248aNa3Dfzr/44osbo13D8Zo4ceI/dHSKIwcAAAAAJBEOAAAAAAQIBwAA\nAAAkEQ4AAAAABDghGQAAIJoeaJV17EbH01/hMe+bYGZZI0eO3D1v3rxPJKm4uFjt2rXL7Nu374El\nS5YUTJkypc2yZcuaP/vss5vvuuuuji1atCh98MEHP6/cR2xsbFa3bt2+Ki0ttbPOOuur2bNnb6x8\nd2Wp/EZcAwcOTFuwYEFB69aty6688sqU1157rVWbNm1KNmzYsKa62t5///1mN954Y8ratWsTJ02a\n9Gnl/c6dO/eUu++++/SysjJdd911u376059ur/r8V199tcXEiRO7rF+/PnHGjBkfV1x+dNu2bXFX\nX3116htvvNForkgUBo4cAAAAnGQSEhLK8vPzE/bv32+S9PLLL5+SnJxcfDx9xMfHl3344YdrN2zY\nsKZJkyY+efLkpKptZs+e3apXr15ftW7dukySbrrppl3z58+v9cN5u3btSn7xi19svvnmm78WRkpK\nSnTnnXeevmDBgvXr169f8+KLL7Zevnx5s6rPP+OMMw7/z//8z8bhw4d/UXl9x44dS5KTk4sXLVrU\n/Hhe58mGcAAAAHASuvDCCwvnzJlzqiQ9//zzrUeNGrX7H+1r8ODB+wsKCuKrrp85c2brynccHjZs\n2P6kpKSS2vrq1KlTyZAhQw42adLka3fqzcvLa961a9ei9PT0w82aNfN//ud/3j137txTqz6/e/fu\nhwcMGPBVTMzRH3Mvv/zyvc8++2yb43pxJxnCAQAAwEno+uuv3/3CCy+cdvDgQVu3bl3ieeedd+Af\n6ae4uFgLFy48pXfv3l9V3bZ8+fIWgwYNOma/jz32WNJjjz121JGHyrZs2dK0U6dOhyuWO3fufPjT\nTz9tKkl33HFHx5kzZ7Y61n4GDRp04N13321xrHYns1DPOTCzXEm/kBQr6Tfu/rMa2l0haY6kc9x9\nWZg1AQAAQBowYMBXW7dujZ8xY0briy66qPB4n19UVBTTo0eP9KCvL2+//fajbrpVWFgYd9ppp5Ud\n/eyvu+eee3Yeq427H7XOzFySnnzyyW2R1NyxY8eSHTt2NI2k7ckqtHBgZrGSnpJ0saStkpaa2Xx3\nX1ulXUtJt0n6e1i1AAAA4Gi5ubl777///i6LFi3K37Fjx3F9Lqw456C2NrGxsV5aWqrY2NhvVqik\n008//ciRAknaunVr044dOx7XeRIHDx60+Pj4Y4aVk1mY04r6Sypw94/d/bCkWZJGVtPuJ5Iek3Qo\nxFoAAABQxS233LJr4sSJ2/r373/UlKC6kJqaemjdunVHnYvwjxgyZMiBjRs3Nvvwww+bHjp0yF56\n6aXWo0aN2nvsZ/6f1atXN0tLSwvltTYWYU4r6iRpS6XlrZIGVG5gZv0kdXH3P5nZ3TV1ZGbjJY2X\npOTkZOXl5dV9tZVM7F3reTInlP3xHZXX/cfRLqNuhPzvjuPHWGmgGCsNDmOlgWooYyWCS4+G5cwz\nzyz+0Y9+tONY7Z544okO06ZNS65Y/vzzzz+IpP9LLrmkcNGiRS0zMjKKJGn48OGp77zzTss9e/bE\nJScn95k0adK2O++8c1fF+Qb33HPPzs2bN8edc8456QcOHIg1M582bVryunXrVrdu3bps8uTJm3Nz\nc9NKS0t1zTXX7MrOzj4klZ9zcM455xy49tprC//6178mXnXVVWft27cv9rXXXjv14Ycf7lhQULBG\nkhYvXtwyNzf3uKdQnUysuvlbddKx2ZWSLnX3ccHy9ZL6u/u/Bcsxkv5X0g3uvtHM8iTdfaxzDrKz\ns33ZsnBPS0iZ9Eqo/denZzI3KCf//miXUTceYCw3NIyVBoqx0uAwVhqoehgrZrbc3bMrr1u5cuXG\nzMzMo+bnN0abNm1qMmbMmJS33367QdxbIDs7u/urr75akJSUVBrtWqJp5cqVbTMzM1Oq2xbmtKKt\nkrpUWu4sqfLJIi0lZUjKM7ONks6VNN/MvjaAAAAAcGLq2rVr8U033bRr9+7dUb9C5rZt2+Juv/32\nz0/2YHAsYU4rWiqpm5mlSvpU0mhJ11RsdPdCSW0rliM9cgAAAIATx7hx4/ZEuwap/EpF119//XGd\no3AyCi3FuXuJpAmSFkpaJ2m2u68xswfNbERY+wUAAGjgysrKyizaReDkFLz3arxiU6j3OXD3BZIW\nVFl3Xw1tc8KsBQAAoIFYvXPnzvSkpKTCmJiYcE7+BKpRVlZmO3fubCVpdU1tQg0HAAAA+LqSkpJx\n27dv/8327dszFO75n0BVZZJWl5SUjKupAeEAAACgHmVlZe2QxBRrNEikVQAAAACSCAcAAAAAAoQD\nAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMA\nAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAA\nAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAA\nAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAA\nkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkkIOB2aWa2b5ZlZgZpOq\n2X6Xma01sw/M7DUz6xpmPQAAAABqFlo4MLNYSU9JGiYpXdIYM0uv0ux9Sdnu3kfSXEmPhVUPAAAA\ngNqFeeSgv6QCd//Y3Q9LmiVpZOUG7r7E3Q8Gi+9I6hxiPQAAAABqYe4eTsdmV0jKdfdxwfL1kga4\n+4Qa2v+XpO3u/lA128ZLGi9JycnJWbNmzQql5gqrPi0Mtf/6lJpQpBZF26JdRt3o0DfaFaAKxkoD\nxVhpcBgrDVQ9jJWhQ4cud/fs0HcE1JG4EPu2atZVm0TM7DpJ2ZKGVLfd3adLmi5J2dnZnpOTU0cl\nVu+GSa+E2n99eibzE+Xk3x/tMurGmMbzy7WxYKw0UIyVBoex0kAxVoCjhBkOtkrqUmm5s6Sjvmow\ns4sk/T9JQ9y9KMR6AAAAANQizHMOlkrqZmapZtZU0mhJ8ys3MLN+kqZJGuHuO0KsBQAAAMAxhBYO\n3L1E0gRJCyWtkzTb3deY2YNmNiJo9p+SWkiaY2YrzGx+Dd0BAAAACFmY04rk7gskLaiy7r5Kjy8K\nc/8AAAAAIscdkgEAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABI\nIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgi\nHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIc\nAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwA\nAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAA\nAAAIEA4AAAAASCIcAAAAAAiEGg7MLNfM8s2swMwmVbM93sxeCLb/3cxSwqwHAAAAQM1CCwdmFivp\nKUnDJKVLGmNm6VWajZW0x93PkvSEpEfDqgcAAABA7cI8ctBfUoG7f+zuhyXNkjSySpuRkn4bPJ4r\n6UIzsxBrAgAAAFCDMMNBJ0ksIq/wAAAFtklEQVRbKi1vDdZV28bdSyQVSmoTYk0AAAAAahAXYt/V\nHQHwf6CNzGy8pPHB4n4zy/+GtZ00hkptJe2Kdh114sccVEJ4GCtAZBgrx61rfewEqCthhoOtkrpU\nWu4saVsNbbaaWZykVpJ2V+3I3adLmh5SnY2amS1z9+xo1wE0dIwVIDKMFaBxC3Na0VJJ3cws1cya\nShotaX6VNvMl/Uvw+ApJ/+vuRx05AAAAABC+0I4cuHuJmU2QtFBSrKSn3X2NmT0oaZm7z5f035J+\nZ2YFKj9iMDqsegAAAADUzviivnEzs/HBtCwAtWCsAJFhrACNG+EAAAAAgKSQ75AMAAAA4MRBOAAA\nAAAgiXBw0jOzFDNbHTzOMbM/RbsmIBrM7O1jbF9gZqfWVz1AY2BmeWaWHTzeaGZto10TgNqFeZ8D\nhMjMTOXnjJRFuxagoTGzWHcvPZ7nuPvAY2z/p29WFdD48LsIaHw4cnACCb7lX2dmUyW9J+l6M/ub\nmb1nZnPMrEXQ7hwze9vMVprZu2bWMnjuG0Hb98ys1g9CQEMVvJc/NLPfmtkHZjbXzBKDbyXvM7M3\nJV1pZmea2Z/NbHnw3u8RPD/ZzF4OxsfKirFgZvuDvzuY2etmtsLMVpvZ+cH6I996mtldwbbVZnZH\npbrWmdkMM1tjZovMLCEqPySgFlXfv2b2qJndWmn7A2Y2MXj8AzNbGoy1Hwfrqv4u6mJmvzKzZcF7\n/8fReWUA6gLh4MTTXdKzki6WNFbSRe5+tqRlku4Kbjj3gqTb3T1T0kWSvpK0Q9LFQdurJU2JRvFA\nHekuabq795G0T1LFB5tD7j7Y3Wep/K7q/+buWZLuljQ1aDNF0l+D8XG2pDVV+r5G0kJ37yspU9KK\nyhvNLEvSjZIGSDpX0vfMrF+wuZukp9y9l6S9kkbV1QsG6kJ1719Js1T+e6HCVZLmmNklKn9P95fU\nV1KWmX0raNNd0rPu3s/dN0n6f8Fdk/tIGmJmferlBQGoc0wrOvFscvd3zOwySemS3io/qqumkv6m\n8v+wP3P3pZLk7vskycyaS/ovM+srqVRSWjSKB+rIFnd/K3j8nKTbgscvSFJwFG2gyj/gVDwnPvj7\nAknflaRg6lFhlb6XSnrazJpImufuK6psHyzpZXc/EOzrJUnnq/yO759Uar9cUso3eI1AGGp6/7Yz\ns46SkiTtcffNZnabpEskvR88t4XKw8JmBb+LKvV7lZmNV/nnig4q//30QX28IAB1i3Bw4jkQ/G2S\nFrv7mMobg29rqrt5xZ2SPlf5N6Exkg6FWSQQsqrv8YrlivERI2lv8O3/8XXs/nrw7ei3VX4H9/90\n92crNbEanipJRZUel0piWhEamprev3MlXSGpvcqPJFS0fcTdp32tA7MU/d9Yk5mlqvzo3DnuvsfM\nnpHUrE6rBlBvmFZ04npH0iAzO0uSgjnXaZI+lNTRzM4J1rc0szhJrVR+RKFM0vWSYqNUN1AXTjez\n84LHYyS9WXljcMTsEzO7Uio/adLMMoPNr0m6JVgfa2anVH6umXWVtMPdZ0j6b5VPParsdUmXB2Ou\nuaTvSHqj7l4aEKqa3r+zJI1WeUCYG7RdKOmmSuezdTKzdtX0eYrKw0KhmSVLGhbyawAQIsLBCcrd\nd0q6QdLzZvaBysNCD3c/rPK5o780s5WSFqv8G5ypkv7FzN5R+ZSiA9V2DJwY1qn8/fyBpNaSflVN\nm2sljQ3GwRpJI4P1t0saamarVD71p1eV5+VIWmFm76v8nIFfVN7o7u9JekbSu5L+Luk37v6+gBNA\nTe9fd18jqaWkT939s6DtIkm/l/S3YLzMDdpU7XOlyqcerZH0tKS3qrYBcOIw9+pmoABAwxRMafiT\nu2dEuRQAABodjhwAAAAAkMSRAwAAAAABjhwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAA\nAJAk/X/po/aVEehNgAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fac57dfd9b0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:10:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>99471</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0        99471"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.681724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.810744</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.681724\n",
       "1  precision  1.000000\n",
       "2   f1_score  0.810744"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP016582.1</th>\n",
       "      <th>CP016583.1</th>\n",
       "      <th>CP016584.1</th>\n",
       "      <th>CP016585.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.99998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.99998</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP016582.1  CP016583.1  CP016584.1  CP016585.1\n",
       "0  plasmid_0         0.0         0.0         0.0     0.99998\n",
       "1      Total         0.0         0.0         0.0     0.99998"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016582.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016583.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016584.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP016585.1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Total</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0\n",
       "0  CP016582.1        0.0\n",
       "1  CP016583.1        0.0\n",
       "2  CP016584.1        0.0\n",
       "3  CP016585.1        1.0\n",
       "4       Total        1.0"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0         1.0"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016582.1</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016583.1</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016584.1</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP016585.1</td>\n",
       "      <td>0.99998</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP016582.1     0.00000\n",
       "1  CP016583.1     0.00000\n",
       "2  CP016584.1     0.00000\n",
       "3  CP016585.1     0.99998"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016585.1</td>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.99998</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction   recall  precision\n",
       "0  CP016585.1  plasmid_0  0.99998        1.0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 10, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.000492444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[13, 28, 31]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.81259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.813082</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           plasmid_0\n",
       "GC_sum  -0.000492444\n",
       "chain   [13, 28, 31]\n",
       "gd_sum      -0.81259\n",
       "total      -0.813082"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00989503</td>\n",
       "      <td>-0.00316538</td>\n",
       "      <td>0.00298945</td>\n",
       "      <td>-0.000521171</td>\n",
       "      <td>-0.000241184</td>\n",
       "      <td>-0.000408262</td>\n",
       "      <td>0.000522801</td>\n",
       "      <td>-0.000329191</td>\n",
       "      <td>-0.00541131</td>\n",
       "      <td>-0.00509926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[17]</td>\n",
       "      <td>[21]</td>\n",
       "      <td>[24]</td>\n",
       "      <td>[31, 81, 13, 31, 13, 28, 31, 80, 28]</td>\n",
       "      <td>[75, 6, 32, 69, 20]</td>\n",
       "      <td>[69, 5, 32, 69, 9]</td>\n",
       "      <td>[43, 11, 32, 69]</td>\n",
       "      <td>[70, 8, 32, 1]</td>\n",
       "      <td>[32]</td>\n",
       "      <td>[28]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.747057</td>\n",
       "      <td>-0.773008</td>\n",
       "      <td>-0.429799</td>\n",
       "      <td>-0.81105</td>\n",
       "      <td>-0.00139564</td>\n",
       "      <td>-0.000832547</td>\n",
       "      <td>-0.00258891</td>\n",
       "      <td>-0.00033275</td>\n",
       "      <td>-0.62466</td>\n",
       "      <td>-0.436984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.756952</td>\n",
       "      <td>-0.776173</td>\n",
       "      <td>-0.42681</td>\n",
       "      <td>-0.811571</td>\n",
       "      <td>-0.00163682</td>\n",
       "      <td>-0.00124081</td>\n",
       "      <td>-0.00206611</td>\n",
       "      <td>-0.000661941</td>\n",
       "      <td>-0.630071</td>\n",
       "      <td>-0.442084</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         plasmid_0   plasmid_1   plasmid_2  \\\n",
       "GC_sum -0.00989503 -0.00316538  0.00298945   \n",
       "chain         [17]        [21]        [24]   \n",
       "gd_sum   -0.747057   -0.773008   -0.429799   \n",
       "total    -0.756952   -0.776173    -0.42681   \n",
       "\n",
       "                                   plasmid_3            plasmid_4  \\\n",
       "GC_sum                          -0.000521171         -0.000241184   \n",
       "chain   [31, 81, 13, 31, 13, 28, 31, 80, 28]  [75, 6, 32, 69, 20]   \n",
       "gd_sum                              -0.81105          -0.00139564   \n",
       "total                              -0.811571          -0.00163682   \n",
       "\n",
       "                 plasmid_5         plasmid_6       plasmid_7   plasmid_8  \\\n",
       "GC_sum        -0.000408262       0.000522801    -0.000329191 -0.00541131   \n",
       "chain   [69, 5, 32, 69, 9]  [43, 11, 32, 69]  [70, 8, 32, 1]        [32]   \n",
       "gd_sum        -0.000832547       -0.00258891     -0.00033275    -0.62466   \n",
       "total          -0.00124081       -0.00206611    -0.000661941   -0.630071   \n",
       "\n",
       "         plasmid_9  \n",
       "GC_sum -0.00509926  \n",
       "chain         [28]  \n",
       "gd_sum   -0.436984  \n",
       "total    -0.442084  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_0  from MILP completely contained in  plasmid_3  from greedy.\n",
      "\n",
      "plasmid_9  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
