{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 117 with $\\alpha_1 = 1$, $\\alpha_2 = 5$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 117\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012934.1</td>\n",
       "      <td>3372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012932.1</td>\n",
       "      <td>2096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012933.1</td>\n",
       "      <td>3319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012935.1</td>\n",
       "      <td>3905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012936.1</td>\n",
       "      <td>98999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CP012931.1</td>\n",
       "      <td>236176</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP012934.1         3372\n",
       "1  CP012932.1         2096\n",
       "2  CP012933.1         3319\n",
       "3  CP012935.1         3905\n",
       "4  CP012936.1        98999\n",
       "5  CP012931.1       236176"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt0VeWd//HPNwl3MAqEcJWkSgIh\nECARNKAERYGpgl14gVY7KpSOllELrTIzrVpra3WgdnCkgh2XUigCXihVRqAdUrxWgoJcA4hEEBAQ\nCIRrLt/fH9nhF0MCB81OArxfa7k8e+/nPPt7jucx53P2s/c2dxcAAAAARNV2AQAAAADqBsIBAAAA\nAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAACBmNou4Ey1\nbNnSExISaruMs8ahQ4fUpEmT2i4DqPMYK0BkGCtnZvny5XvcPa626wAiddaFg4SEBOXk5NR2GWeN\n7OxsZWVl1XYZQJ3HWAEiw1g5M2aWV9s1AGeCaUUAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgH\nAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAARCCwdm9ryZ7TKz1VVsNzObbGabzOxjM+sVVi0AAAAA\nTi/MIwcvSBp8iu1DJHUK/hkj6fch1gIAAADgNEILB+6+VNLeUzQZJmm6l3pf0oVm1iasegAAAACc\nWkwt7rudpK3llrcF63ZUbGhmY1R6dEHx8fHKzs6uifrOCQUFBbxfCM2qz/Nru4RqkxgbzVgBIsDf\nFeDcVpvhwCpZ55U1dPdpkqZJUkZGhmdlZYVY1rklOztbvF8Iyx0T3qjtEqrNC4ObMFYQmoRzaqw0\nZawA57DaDAfbJHUot9xe0vZaquUrzq3/iTep7RIAAABwlqjNS5nOl/T94KpFl0vKd/eTphQBAAAA\nqBmhHTkws1mSsiS1NLNtkh6WVE+S3P1ZSQsk/ZOkTZIOS7ozrFoAAAAAnF5o4cDdR55mu0v6UVj7\nBwAAAHBmuEMyAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmE\nAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQD\nAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMA\nAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAA\nAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAA\nAAHCAQAAAABJhAMAAAAAgVDDgZkNNrNcM9tkZhMq2X6xmS0xs4/M7GMz+6cw6wEAAABQtdDCgZlF\nS3pG0hBJKZJGmllKhWY/kzTH3XtKGiFpSlj1AAAAADi1MI8c9Ja0yd03u/txSS9JGlahjUu6IHgc\nK2l7iPUAAAAAOIWYEPtuJ2lrueVtkvpUaPOIpEVm9q+SmkgaGGI9AAAAAE4hzHBglazzCssjJb3g\n7pPM7ApJfzSzVHcv+UpHZmMkjZGk+Ph4ZWdnh1HvCeO7FYXaf00qKCgI/f3C+YuxAkSGsQLgbBFm\nONgmqUO55fY6edrQKEmDJcnd3zOzhpJaStpVvpG7T5M0TZIyMjI8KysrpJJL3THhjVD7r0kvDG6i\nsN8vnL8YK0BkGCsAzhZhnnOwTFInM0s0s/oqPeF4foU2n0m6RpLMrIukhpJ2h1gTAAAAgCqEFg7c\nvUjSWEkLJa1T6VWJ1pjZo2Y2NGg2XtIPzGylpFmS7nD3ilOPAAAAANSAMKcVyd0XSFpQYd1D5R6v\nldQ3zBoAAAAARIY7JAMAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAA\nAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAA\nAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAA\nkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQ\nRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBE\nOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQFHI4MLPBZpZrZpvMbEIVbW4xs7VmtsbM/hRmPQAA\nAACqFhNWx2YWLekZSddK2iZpmZnNd/e15dp0kvRvkvq6+z4zaxVWPQAAAABOLcwjB70lbXL3ze5+\nXNJLkoZVaPMDSc+4+z5JcvddIdYDAAAA4BRCO3IgqZ2kreWWt0nqU6FNkiSZ2TuSoiU94u5vVuzI\nzMZIGiNJ8fHxys7ODqPeE8Z3Kwq1/5pUUFAQ+vuF8xdjBYgMYwXA2SLMcGCVrPNK9t9JUpak9pLe\nMrNUd9//lSe5T5M0TZIyMjI8Kyur2ost744Jb4Taf016YXAThf1+4fzFWAEiw1gBcLYIc1rRNkkd\nyi23l7S9kjZ/dvdCd/9UUq5KwwIAAACAGhZmOFgmqZOZJZpZfUkjJM2v0GaepAGSZGYtVTrNaHOI\nNQEAAACoQmjhwN2LJI2VtFDSOklz3H2NmT1qZkODZgslfWlmayUtkfRTd/8yrJoAAAAAVC3Mcw7k\n7gskLaiw7qFyj13SuOAfAAAAALWIOyQDAAAAkEQ4AAAAABAgHAAAAACQdAbhwMz6mdmdweM4M0sM\nrywAAAAANS2icGBmD0t6UNK/BavqSZoRVlEAAAAAal6kRw6+I2mopEOS5O7bJTULqygAAAAANS/S\ncHA8uOyoS5KZNQmvJAAAAAC1IdJwMMfMpkq60Mx+IOmvkp4LrywAAAAANS2im6C5+0Qzu1bSAUnJ\nkh5y98WhVgYAAACgRp02HJhZtKSF7j5QEoEAAAAAOEedNhy4e7GZHTazWHfPr4miUI12rJAeGVbb\nVVSPR/j4AQAAhCmiaUWSjkpaZWaLFVyxSJLc/d5QqgIAAABQ4yINB28E/wAAAAA4R0V6QvKLZlZf\nUlKwKtfdC8MrCwAA1ElMVwXOaRGFAzPLkvSipC2STFIHM/tnd18aXmkAAAAAalKk04omSbrO3XMl\nycySJM2SlB5WYQAAAABqVqThoF5ZMJAkd99gZvVCqgkAah5TJQAAiDgc5JjZ/0j6Y7D8PUnLwykJ\nAAAAQG2INBzcLelHku5V6TkHSyVNCasoAAAAADUv0nAQI+m/3P230om7JjcIrSoAAIBz1PLly1vF\nxMT8QVKqpKjargfnlRJJq4uKikanp6fvqqxBpOHgb5IGSioIlhtJWiQp8xuXCAAAcB6JiYn5Q+vW\nrbvExcXti4qK8tquB+ePkpIS2717d8rOnTv/IGloZW0iTasN3b0sGCh43LgaagQAADjfpMbFxR0g\nGKCmRUVFeVxcXL5Kj1pV3ibCvg6ZWa+yBTPLkHTkG9YHAABwPooiGKC2BJ+9KjNApNOK7pM018y2\nS3JJbSXd+s3LAwAAAFBXRBoOEiX1lHSxpO9IulylIQEAAADfQMKEN6r1prJbfvPtOnO5+d69eydP\nnDhx61VXXXU4kvYlJSXKzMxMWrBgwabmzZuXhF3fN/X66683mzRpUvySJUs2hdH/zJkzY9esWdPo\n17/+9c6K2xo3btzz8OHDH1X13MzMzKQ///nPn8TFxRWfyT4jnVb0c3c/IOlCSddKmibp92eyIwAA\nAJw9CgsLa3yfc+bMie3ateuRsyEY1ITvfe97+ZUFg0iMHDnyy4kTJ8ad6fMiDQdliePbkp519z9L\nqn+mOwMAAEDt++lPf9omMTGxa2ZmZqcbbrgh8aGHHoqXSn/pHzt2bLvLLrss+bHHHovfvn17zKBB\ngy5JTU3tkpqa2mXRokVNJOnAgQNRN998c0JqamqXLl26pMyYMeNCSSooKLDrr7/+W0lJSSnf/va3\nv3X06FGTpKeeeqrlqFGjOpTtf9KkSS1Hjx7dvmJdM2fObP6d73xnvyTl5ubWT0xM7Hrrrbd27NSp\nU9ehQ4cmzps3r1mvXr06d+zYMXXJkiWNT1VLbm5u/fT09OSUlJQuKSkpXRYvXtxEKv21v3fv3smD\nBw/+VmJiYtehQ4cmlpScOosMHz484bvf/e7F6enpyQkJCamzZs2KrdhmyZIljXv27Nm5S5cuKT17\n9uy8cuXKBpKUk5PTsFu3bl06d+6ckpSUlLJq1aoGkb62yZMnt/j+979/sSStX7++fo8ePTqnpqZ2\nue+++9qW7TcvL69eRkZGcufOnVM6derU9c0332wqSSNGjNj/6quvtojwI3FCpNOKPjezqSq9nOkT\nZtZAXJcXAADgrLN06dLGf/nLXy5atWrV2sLCQuvRo0dKz549T0z72b9/f/SyZctyJemGG25IHDdu\n3BeDBg0q2LhxY/1BgwZ12rx585p///d/bzNgwIADc+fO3bJnz57ojIyMLkOHDj3w29/+Nq5Ro0Yl\nGzZsWPuPf/yjUd++fVMkadSoUXu7du2acuzYsW0NGjTwGTNmtJw6dWpexdqWL1/etG/fvifWb926\nteHs2bM3p6en53Xv3r3LzJkzW+Tk5Kz/05/+dOGvfvWrNgMGDPikqlratm1b9NZbb21o3Lixr1q1\nqsHIkSO/tXr16nWStG7dukYrVqzYnJCQUJient558eLFTQcNGlRQsZ7ytm7d2uCDDz7IXbt2bYOB\nAwcmDxs2bFX57WlpaUc/+OCD9fXq1dO8efOaPfDAA+0XLlz4ydNPPx13zz33fHH33XfvPXr0qBUV\nFenzzz+vF8lrK9//Pffcc/Ho0aN3jx079svHH3/8xBGB559/vvk111yT/8QTT+wsKirSwYMHoyQp\nLi6u+Pjx47Zz587o1q1bRzy1KNJwcIukwZImuvt+M2sj6aeR7gQAAAB1Q3Z2dtMhQ4bsb9q0qUvy\na6+9dn/57SNHjtxb9vidd965YOPGjY3KlgsKCqL37dsXlZ2dfcHChQsvnDx5cmtJOnbsmG3atKn+\n22+/3fTee+/dJUl9+vQ5kpSUdFiSLrjggpK+ffsenD17dmy3bt2OFhYWWu/evU+68mV+fn7MRRdd\ndOJn/Hbt2h0ra5eUlHTk6quvPhAVFaVevXodfuyxx9oGr6fSWjp27Fg4atSojmvXrm0UFRWlvLy8\nEzfw7dat26FLLrmkUJK6du16+JNPPjntjJjhw4fvjY6OVrdu3Y516NDh2IoVKxqW3753797oW2+9\nNXHLli0NzcwLCwtNkq644opDEydObLNt27b6I0aM2NetW7djkb628j788MOm//u///uJJP3whz/8\n8pe//GV7Sbr88ssP/fCHP0woLCyMuummm/ZlZmaeeF9btGhR9Nlnn9Vv3bp1xFcZjSgcuPthSa+W\nW94haUekOwEAAEDd4H7qa8o0a9bsxJdzd1dOTs66IEh8pY+XX355U1pa2rGKzzezSvsdM2bMnl/9\n6letk5KSjt522217KmsTHR3txcXFio6OliTVr1//xH6joqLUsGFDD9qpuLjYTlXLuHHj2rZq1arw\nlVde+bSkpESNGjU6ceJ3gwYNTvQbHR2toqKiyos+xeuquPzggw+269+//8HFixd/kpubW//qq69O\nlqR/+Zd/2XvllVceeu2112KHDBmSNGXKlC3JycnHInltFVV2CdwhQ4YULF26NPeVV16JveOOOxLv\nvffeL8aOHfulVBqUGjdufEbnbzA1CAAA4DySlZVVsHDhwtjDhw9bfn5+1F//+tcLq2rbr1+/A088\n8USrsuV33323kSQNGDDgwKRJk+LL5uq/8847jYL2BTNmzGguScuWLWu4YcOGEzfNvfrqqw/t2LGj\n/muvvdZi1KhRe1WJxMTEo+vWrWtQ2baqVFVLfn5+dJs2bQqjo6M1ZcqUFsXFp59Z86Mf/ajd9OnT\nK30/Xn311YuKi4u1Zs2aBlu3bm2QlpZ2tPz2AwcORLdv3/64JE2dOrVl2fq1a9fW79Kly7Gf/exn\nu6677rr9K1asaFSx70j06tWr4LnnnmsuSc8999yJcwk2bNhQv127doXjx4/fc9ttt+358MMPG0ul\nV37avXt3veTk5JMC3KlEOq0IAAAAIajpS4/279//8ODBg/NTUlK6tmvX7lj37t0PxcbGVvrNedq0\naVtHjx59cVJSUkpxcbH16dPnYGZm5me/+c1vto8ZM+bizp07p7i7tW/f/tiSJUs2/eQnP9k1YsSI\nxKSkpJSuXbse7tat26Hy/d144437Pv7448ZVXV7zuuuuy1+0aFGz1NTUiL/QVlXL/fffv2v48OGX\nzJs376J+/fodbNSo0Wl/QV+7dm2jshOiK7r00kuP9e7dO/nLL7+s97vf/S6vcePGX/kV/8EHH9w5\nevToxMmTJ7e+8sorD5St/+Mf/9h87ty5LWJiYjwuLq7w8ccf375///7oSF9fmSlTpnw2YsSIb02Z\nMiV+6NCh+8rWL1y4sNnkyZNbx8TEeOPGjYtnzpz5qSS9/fbbjXv27HmoXr16Z7QfO92hpbomIyPD\nc3JyQt1HwoQ3Qu2/Jr2QtlFZuQ/XdhnV45H82q4AFTBW6ijGSp3DWKmjamCsmNlyd88ov27lypVb\n0tLSKp1WU1Py8/OjYmNjSw4ePBh1xRVXJD/77LN5/fr1i+heBN/EgAEDLr3//vu/GDZs2MHKtufl\n5dUbOXJkwrvvvrsx7Foq069fv05vv/32SfsePnx4wvXXX59/55137qvseXXRnXfe2eHGG2/cX9l7\nvXLlypZpaWkJlT2PaUUAAADnmdtuu61j586dU7p3797lhhtu2Bd2MNizZ090QkJCasOGDUuqCgaS\n1LFjx8K77rprz969e2vlO2plweBslZqaeuRU73VVmFYEAABwnvnLX/7yaU3ur2XLlsVbtmxZHUnb\n0aNH17lf51955ZUttV3DmRo/fvzXOjrFkQMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAApyQDAAA\nUJseiU0/faMz6S//tPdNMLP0YcOG7Z03b96nklRYWKhWrVql9ejR49CSJUs2TZ48uUVOTk6T6dOn\nfzZu3Li2TZs2LX700Ue/KN9HdHR0eqdOnY4UFxfbpZdeemTOnDlbyt9dWSq9EVdmZmbSggULNjVv\n3rzk5ptvTvjb3/4W26JFi6KNGzeuqay2119/vdnIkSMvadeu3XFJuv766/dNnDhxR8V2v/71r+Oe\nffbZ+K1btzbYvn37yjZt2hRJ0qxZs2I/+OCDJk899dT2yN80lOHIAQAAwHmmUaNGJbm5uY0KCgpM\nkl577bUL4uPjC8+kjwYNGpSsX79+7caNG9fUq1fPJ02aFFexzZw5c2K7du16pHnz5iWSdNddd+2Z\nP3/+aS8XmpGRUbB+/fq169evX1tZMJCk/v37FyxevHhD27Ztj5dff+utt+a/+eabFx48eJDvuV8D\nbxoAAMB56JprrsmfO3fuhZI0a9as5sOHD9/7dfvq169fwaZNmxpUXD9z5szm5e84PGTIkIK4uLii\nr7uf8vr27XskOTn5eMX1UVFRyszMPDh79uzY6tjP+YZwAAAAcB66/fbb986ePfuiw4cP27p16xpf\nccUVh75OP4WFhVq4cOEF3bp1O1Jx2/Lly5v27dv3tP0++eSTcU8++eSJIw8fffRR0+Tk5JSrrrqq\nU05OTsOy9f379790y5Yt9U7XX0ZGxqG33nqr6Zm8DpQK9ZwDMxss6b8kRUv6g7v/pop2N0maK+ky\nd88JsyYAAABIffr0ObJt27YGzz33XPOBAwfmn+nzjx07FtW5c+eUoK+D991330k33crPz4+56KKL\nSk5+9lc98MADu8seZ2ZmHsrLy/s4Nja2ZPbs2bHDhw+/NC8vb7Uk/f3vf98USW2tW7cu2rlzZ/3I\nXw3KhBYOzCxa0jOSrpW0TdIyM5vv7msrtGsm6V5J/wirFgAAAJxs8ODB+x9++OEOixYtyt21a9cZ\nfS8sO+fgVG2io6O9uLhY0dHREfdbdn6CVHr+wLhx4y7esWNHTNkJx5E4cuSINWzY8LShBCcLc1pR\nb0mb3H2zux+X9JKkYZW0+6WkJyUdDbEWAAAAVHD33XfvGT9+/PbevXufNCWoOiQmJh5dt27dSeci\nnMpnn30WU1JS+r1+yZIljUtKShQfH39G5ynk5uY27Nq1ayiv6VwX5rSidpK2llveJqlP+QZm1lNS\nB3d/3cx+UlVHZjZG0hhJio+PV3Z2dvVXW874btVynkydUNCgrbKTf1HbZVSPkP+748wxVuooxkqd\nw1ipo+rKWIng0qNhueSSSwp//vOf7zpdu6eeeqrN1KlT48uWv/jii48j6f+6667LX7RoUbPU1NRj\nknTDDTckvv/++8327dsXEx8f333ChAnbf/zjH+8pO9/ggQce2D1jxoyLnn/++VbR0dHesGHDkunT\np2+Oiir9Pbt///6Xvvjii3kJCQmFjz32WKunn3669ZdfflkvLS0tZcCAAfmzZ8/Ok6SlS5c2e+KJ\nJz7/Gm/Jec/cPZyOzW6WNMjdRwfLt0vq7e7/GixHSfo/SXe4+xYzy5b0k9Odc5CRkeE5OeGelpAw\n4Y1Q+69JL6RtVFbuw7VdRvV45IynQyJkjJU6irFS5zBW6qgaGCtmttzdM8qvW7ly5Za0tLST5uef\ni/Ly8uqNHDky4d133z3t5Uury9atW2NuueWWb7333nsbamqfZ5uVK1e2TEtLS6hsW5jTirZJ6lBu\nub2k8jejaCYpVVK2mW2RdLmk+Wb2lQEEAACAs1PHjh0L77rrrj179+6tsStkbt68uf6kSZO2nr4l\nKhPmtKJlkjqZWaKkzyWNkPTdso3uni+pZdlypEcOAAAAcPYYPXr0vprcX//+/Q/X5P7ONaGlOHcv\nkjRW0kJJ6yTNcfc1ZvaomQ0Na78AAAB1XElJSYnVdhE4PwWfvSqv5BTqfQ7cfYGkBRXWPVRF26ww\nawEAAKgjVu/evTslLi4uPyoqKpyTP4FKlJSU2O7du2Mlra6qTajhAAAAAF9VVFQ0eufOnX/YuXNn\nqsI9/xOoqETS6qKiotFVNSAcAAAA1KD09PRdkphijTqJtAoAAABAEuEAAAAAQIBwAAAAAEAS4QAA\nAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAA\nAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAA\nQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABA\ngHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECA\ncAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQFLI4cDMBptZrpltMrMJlWwfZ2Zrzexj\nM/ubmXUMsx4AAAAAVQstHJhZtKRnJA2RlCJppJmlVGj2kaQMd+8u6WVJT4ZVDwAAAIBTC/PIQW9J\nm9x9s7sfl/SSpGHlG7j7Enc/HCy+L6l9iPUAAAAAOAVz93A6NrtJ0mB3Hx0s3y6pj7uPraL9f0va\n6e6PVbJtjKQxkhQfH5/+0ksvhVJzmVWf54faf01KbHRMTY9tr+0yqkebHrVdASpgrNRRjJU6h7FS\nR9XAWBkwYMByd88IfUdANYkJsW+rZF2lScTMbpOUIal/ZdvdfZqkaZKUkZHhWVlZ1VRi5e6Y8Eao\n/dekF9I+VVbuw7VdRvUYee78cT1XMFbqKMZKncNYqaMYK8BJwgwH2yR1KLfcXtJJPzWY2UBJ/yGp\nv7sfC7EeAAAAAKcQ5jkHyyR1MrNEM6svaYSk+eUbmFlPSVMlDXX3XSHWAgAAAOA0QgsH7l4kaayk\nhZLWSZrj7mvM7FEzGxo0+09JTSXNNbMVZja/iu4AAAAAhCzMaUVy9wWSFlRY91C5xwPD3D8AAACA\nyHGHZAAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAA\nAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAC\nhAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKE\nAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQD\nAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMA\nAAAAkggHAAAAAAKhhgMzG2xmuWa2ycwmVLK9gZnNDrb/w8wSwqwHAAAAQNVCCwdmFi3pGUlDJKVI\nGmlmKRWajZK0z90vlfSUpCfCqgcAAADAqYV55KC3pE3uvtndj0t6SdKwCm2GSXoxePyypGvMzEKs\nCQAAAEAVwgwH7SRtLbe8LVhXaRt3L5KUL6lFiDUBAAAAqEJMiH1XdgTAv0YbmdkYSWOCxQIzy/2G\ntZ03BkgtJe2p7TqqxS84qITwMFaAyDBWzljHmtgJUF3CDAfbJHUot9xe0vYq2mwzsxhJsZL2VuzI\n3adJmhZSnec0M8tx94zargOo6xgrQGQYK8C5LcxpRcskdTKzRDOrL2mEpPkV2syX9M/B45sk/Z+7\nn3TkAAAAAED4Qjty4O5FZjZW0kJJ0ZKed/c1ZvaopBx3ny/pfyT90cw2qfSIwYiw6gEAAABwasYP\n9ec2MxsTTMsCcAqMFSAyjBXTRItMAAAEzUlEQVTg3EY4AAAAACAp5DskAwAAADh7EA4AAAAASCIc\nnPfMLMHMVgePs8zs9dquCagNZvbuabYvMLMLa6oe4FxgZtlmlhE83mJmLWu7JgCnFuZ9DhAiMzOV\nnjNSUtu1AHWNmUW7e/GZPMfdM0+z/Z++WVXAuYe/RcC5hyMHZ5HgV/51ZjZF0oeSbjez98zsQzOb\na2ZNg3aXmdm7ZrbSzD4ws2bBc98K2n5oZqf8IgTUVcFneb2ZvWhmH5vZy2bWOPhV8iEze1vSzWZ2\niZm9aWbLg89+5+D58Wb2WjA+VpaNBTMrCP7dxsyWmtkKM1ttZlcG60/86mlm44Jtq83s/nJ1rTOz\n58xsjZktMrNGtfImAadQ8fNrZk+Y2T3ltj9iZuODxz81s2XBWPtFsK7i36IOZvZ7M8sJPvu/qJ1X\nBqA6EA7OPsmSpku6VtIoSQPdvZekHEnjghvOzZZ0n7unSRoo6YikXZKuDdreKmlybRQPVJNkSdPc\nvbukA5LKvtgcdfd+7v6SSu+q/q/uni7pJ5KmBG0mS/p7MD56SVpToe/vSlro7j0kpUlaUX6jmaVL\nulNSH0mXS/qBmfUMNneS9Iy7d5W0X9Lw6nrBQHWo7PMr6SWV/l0oc4ukuWZ2nUo/070l9ZCUbmZX\nBW2SJU13957unifpP4K7JneX1N/MutfICwJQ7ZhWdPbJc/f3zex6SSmS3ik9qqv6kt5T6f+wd7j7\nMkly9wOSZGZNJP23mfWQVCwpqTaKB6rJVnd/J3g8Q9K9wePZkhQcRctU6Recsuc0CP59taTvS1Iw\n9Si/Qt/LJD1vZvUkzXP3FRW295P0mrsfCvb1qqQrVXrH90/LtV8uKeEbvEYgDFV9fluZWVtJcZL2\nuftnZnavpOskfRQ8t6lKw8JnCv4Wlev3FjMbo9LvFW1U+vfp45p4QQCqF+Hg7HMo+LdJWuzuI8tv\nDH6tqezmFT+W9IVKfwmNknQ0zCKBkFX8jJctl42PKEn7g1//z6xj96XBr6PfVukd3P/T3aeXa2JV\nPFWSjpV7XCyJaUWoa6r6/L4s6SZJrVV6JKGs7ePuPvUrHZgl6P+PNZlZokqPzl3m7vvM7AVJDau1\nagA1hmlFZ6/3JfU1s0slKZhznSRpvaS2ZnZZsL6ZmcVIilXpEYUSSbdLiq6luoHqcLGZXRE8Hinp\n7fIbgyNmn5rZzVLpSZNmlhZs/puku4P10WZ2QfnnmllHSbvc/TlJ/6PSqUflLZV0YzDmmkj6jqS3\nqu+lAaGq6vP7kqQRKg0ILwdtF0q6q9z5bO3MrFUlfV6g0rCQb2bxkoaE/BoAhIhwcJZy992S7pA0\ny8w+VmlY6Ozux1U6d/RpM1spabFKf8GZIumfzex9lU4pOlRpx8DZYZ1KP88fS2ou6feVtPmepFHB\nOFgjaViw/j5JA8xslUqn/nSt8LwsSSvM7COVnjPwX+U3uvuHkl6Q9IGkf0j6g7t/JOAsUNXn193X\nSGom6XN33xG0XSTpT5LeC8bLy0Gbin2uVOnUozWSnpf0TsU2AM4e5l7ZDBQAqJuCKQ2vu3tqLZcC\nAMA5hyMHAAAAACRx5AAAAABAgCMHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkSf8P\n83ywnCat580AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f210d35fcf8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:5:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>194757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>92126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>70124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>168157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>5752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>13397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>64582</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0       194757\n",
       "1  plasmid_1        92126\n",
       "2  plasmid_2        70124\n",
       "3  plasmid_3       168157\n",
       "4  plasmid_4         5752\n",
       "5  plasmid_5        13397\n",
       "6  plasmid_6        64582"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.540569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.518229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.529163</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.540569\n",
       "1  precision  0.518229\n",
       "2   f1_score  0.529163"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP012931.1</th>\n",
       "      <th>CP012932.1</th>\n",
       "      <th>CP012933.1</th>\n",
       "      <th>CP012934.1</th>\n",
       "      <th>CP012935.1</th>\n",
       "      <th>CP012936.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.245152</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.010416</td>\n",
       "      <td>0.040076</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.920595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.086186</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.252341</td>\n",
       "      <td>0.040076</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.530844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.022890</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>0.029110</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.085112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>0.093447</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.739542</td>\n",
       "      <td>0.080153</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.536551</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP012931.1  CP012932.1  CP012933.1  CP012934.1  CP012935.1  \\\n",
       "0  plasmid_0    0.245152    0.000000         0.0         0.0         0.0   \n",
       "1  plasmid_1    0.010416    0.040076         0.0         0.0         0.0   \n",
       "2  plasmid_2    0.086186    0.000000         0.0         0.0         0.0   \n",
       "3  plasmid_3    0.252341    0.040076         0.0         0.0         0.0   \n",
       "4  plasmid_4    0.022890    0.000000         0.0         0.0         0.0   \n",
       "5  plasmid_5    0.029110    0.000000         0.0         0.0         0.0   \n",
       "6  plasmid_6    0.093447    0.000000         0.0         0.0         0.0   \n",
       "7      Total    0.739542    0.080153         0.0         0.0         0.0   \n",
       "\n",
       "   CP012936.1  \n",
       "0    0.000000  \n",
       "1    0.920595  \n",
       "2    0.000000  \n",
       "3    0.530844  \n",
       "4    0.000000  \n",
       "5    0.085112  \n",
       "6    0.000000  \n",
       "7    1.536551  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012931.1</td>\n",
       "      <td>0.280611</td>\n",
       "      <td>0.008901</td>\n",
       "      <td>0.287006</td>\n",
       "      <td>0.345392</td>\n",
       "      <td>0.880042</td>\n",
       "      <td>0.415093</td>\n",
       "      <td>0.324858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012932.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000912</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000500</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012933.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012934.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012935.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CP012936.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.980016</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.311227</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.570799</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.280611</td>\n",
       "      <td>0.989829</td>\n",
       "      <td>0.287006</td>\n",
       "      <td>0.657118</td>\n",
       "      <td>0.880042</td>\n",
       "      <td>0.985892</td>\n",
       "      <td>0.324858</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2  plasmid_3  plasmid_4  \\\n",
       "0  CP012931.1   0.280611   0.008901   0.287006   0.345392   0.880042   \n",
       "1  CP012932.1   0.000000   0.000912   0.000000   0.000500   0.000000   \n",
       "2  CP012933.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "3  CP012934.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "4  CP012935.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "5  CP012936.1   0.000000   0.980016   0.000000   0.311227   0.000000   \n",
       "6       Total   0.280611   0.989829   0.287006   0.657118   0.880042   \n",
       "\n",
       "   plasmid_5  plasmid_6  \n",
       "0   0.415093   0.324858  \n",
       "1   0.000000   0.000000  \n",
       "2   0.000000   0.000000  \n",
       "3   0.000000   0.000000  \n",
       "4   0.000000   0.000000  \n",
       "5   0.570799   0.000000  \n",
       "6   0.985892   0.324858  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.280611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.988917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.287006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.656619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.880042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>0.985892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>0.324858</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.280611\n",
       "1  plasmid_1    0.988917\n",
       "2  plasmid_2    0.287006\n",
       "3  plasmid_3    0.656619\n",
       "4  plasmid_4    0.880042\n",
       "5  plasmid_5    0.985892\n",
       "6  plasmid_6    0.324858"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012931.1</td>\n",
       "      <td>0.379061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012932.1</td>\n",
       "      <td>0.040076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012933.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012934.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012935.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CP012936.1</td>\n",
       "      <td>0.994323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP012931.1    0.379061\n",
       "1  CP012932.1    0.040076\n",
       "2  CP012933.1    0.000000\n",
       "3  CP012934.1    0.000000\n",
       "4  CP012935.1    0.000000\n",
       "5  CP012936.1    0.994323"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012936.1</td>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.920595</td>\n",
       "      <td>0.980016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction    recall  precision\n",
       "0  CP012936.1  plasmid_1  0.920595   0.980016"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 5, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0112877</td>\n",
       "      <td>-0.00422704</td>\n",
       "      <td>-0.000947772</td>\n",
       "      <td>-0.00359523</td>\n",
       "      <td>0.0147633</td>\n",
       "      <td>0.0016771</td>\n",
       "      <td>-0.000529961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[43, 11, 42, 35, 15]</td>\n",
       "      <td>[49, 17, 62, 16, 47, 43]</td>\n",
       "      <td>[71, 19, 40, 27, 100, 106, 34, 86, 72, 107, 48...</td>\n",
       "      <td>[49, 17, 62, 23, 71, 19, 40, 27, 100, 106, 34,...</td>\n",
       "      <td>[26, 53, 44]</td>\n",
       "      <td>[28, 35, 109, 58, 23, 62]</td>\n",
       "      <td>[71, 19, 40, 27, 100, 106, 34, 86, 72, 107, 48...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.68096</td>\n",
       "      <td>-0.741358</td>\n",
       "      <td>-0.351989</td>\n",
       "      <td>-0.561198</td>\n",
       "      <td>-0.770022</td>\n",
       "      <td>-0.734866</td>\n",
       "      <td>-0.325552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.692247</td>\n",
       "      <td>-0.745585</td>\n",
       "      <td>-0.352937</td>\n",
       "      <td>-0.564794</td>\n",
       "      <td>-0.755259</td>\n",
       "      <td>-0.733189</td>\n",
       "      <td>-0.326082</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   plasmid_0                 plasmid_1  \\\n",
       "GC_sum            -0.0112877               -0.00422704   \n",
       "chain   [43, 11, 42, 35, 15]  [49, 17, 62, 16, 47, 43]   \n",
       "gd_sum              -0.68096                 -0.741358   \n",
       "total              -0.692247                 -0.745585   \n",
       "\n",
       "                                                plasmid_2  \\\n",
       "GC_sum                                       -0.000947772   \n",
       "chain   [71, 19, 40, 27, 100, 106, 34, 86, 72, 107, 48...   \n",
       "gd_sum                                          -0.351989   \n",
       "total                                           -0.352937   \n",
       "\n",
       "                                                plasmid_3     plasmid_4  \\\n",
       "GC_sum                                        -0.00359523     0.0147633   \n",
       "chain   [49, 17, 62, 23, 71, 19, 40, 27, 100, 106, 34,...  [26, 53, 44]   \n",
       "gd_sum                                          -0.561198     -0.770022   \n",
       "total                                           -0.564794     -0.755259   \n",
       "\n",
       "                        plasmid_5  \\\n",
       "GC_sum                  0.0016771   \n",
       "chain   [28, 35, 109, 58, 23, 62]   \n",
       "gd_sum                  -0.734866   \n",
       "total                   -0.733189   \n",
       "\n",
       "                                                plasmid_6  \n",
       "GC_sum                                       -0.000529961  \n",
       "chain   [71, 19, 40, 27, 100, 106, 34, 86, 72, 107, 48...  \n",
       "gd_sum                                          -0.325552  \n",
       "total                                           -0.326082  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_11</th>\n",
       "      <th>plasmid_12</th>\n",
       "      <th>plasmid_13</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0114644</td>\n",
       "      <td>-0.00475408</td>\n",
       "      <td>-0.00010891</td>\n",
       "      <td>-0.000219433</td>\n",
       "      <td>0.000997925</td>\n",
       "      <td>0.0010085</td>\n",
       "      <td>-0.00153295</td>\n",
       "      <td>-0.00489927</td>\n",
       "      <td>-0.0141292</td>\n",
       "      <td>-0.00128531</td>\n",
       "      <td>0.00616738</td>\n",
       "      <td>-0.00989734</td>\n",
       "      <td>-0.00837569</td>\n",
       "      <td>-0.0128504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[53, 44, 43, 11, 42, 58, 24, 58]</td>\n",
       "      <td>[35, 42, 15, 35, 28, 43, 20, 43, 54, 53, 26, 53]</td>\n",
       "      <td>[21, 91, 45, 6, 91]</td>\n",
       "      <td>[1, 45, 8, 93]</td>\n",
       "      <td>[45, 4]</td>\n",
       "      <td>[43]</td>\n",
       "      <td>[17, 49, 17, 62, 23]</td>\n",
       "      <td>[62, 23, 62, 16, 47, 16]</td>\n",
       "      <td>[32]</td>\n",
       "      <td>[29]</td>\n",
       "      <td>[33]</td>\n",
       "      <td>[47]</td>\n",
       "      <td>[58]</td>\n",
       "      <td>[49]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.688909</td>\n",
       "      <td>-0.698997</td>\n",
       "      <td>-0.00160241</td>\n",
       "      <td>-0.000395641</td>\n",
       "      <td>-0.00115709</td>\n",
       "      <td>-0.833333</td>\n",
       "      <td>-0.758457</td>\n",
       "      <td>-0.729381</td>\n",
       "      <td>-0.514713</td>\n",
       "      <td>-0.644046</td>\n",
       "      <td>-0.393607</td>\n",
       "      <td>-0.833333</td>\n",
       "      <td>-0.833333</td>\n",
       "      <td>-0.766323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.700373</td>\n",
       "      <td>-0.703751</td>\n",
       "      <td>-0.00171132</td>\n",
       "      <td>-0.000615073</td>\n",
       "      <td>-0.000159161</td>\n",
       "      <td>-0.832325</td>\n",
       "      <td>-0.75999</td>\n",
       "      <td>-0.73428</td>\n",
       "      <td>-0.528842</td>\n",
       "      <td>-0.645331</td>\n",
       "      <td>-0.387439</td>\n",
       "      <td>-0.843231</td>\n",
       "      <td>-0.841709</td>\n",
       "      <td>-0.779173</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               plasmid_0  \\\n",
       "GC_sum                        -0.0114644   \n",
       "chain   [53, 44, 43, 11, 42, 58, 24, 58]   \n",
       "gd_sum                         -0.688909   \n",
       "total                          -0.700373   \n",
       "\n",
       "                                               plasmid_1           plasmid_10  \\\n",
       "GC_sum                                       -0.00475408          -0.00010891   \n",
       "chain   [35, 42, 15, 35, 28, 43, 20, 43, 54, 53, 26, 53]  [21, 91, 45, 6, 91]   \n",
       "gd_sum                                         -0.698997          -0.00160241   \n",
       "total                                          -0.703751          -0.00171132   \n",
       "\n",
       "            plasmid_11   plasmid_12 plasmid_13             plasmid_2  \\\n",
       "GC_sum    -0.000219433  0.000997925  0.0010085           -0.00153295   \n",
       "chain   [1, 45, 8, 93]      [45, 4]       [43]  [17, 49, 17, 62, 23]   \n",
       "gd_sum    -0.000395641  -0.00115709  -0.833333             -0.758457   \n",
       "total     -0.000615073 -0.000159161  -0.832325              -0.75999   \n",
       "\n",
       "                       plasmid_3  plasmid_4   plasmid_5   plasmid_6  \\\n",
       "GC_sum               -0.00489927 -0.0141292 -0.00128531  0.00616738   \n",
       "chain   [62, 23, 62, 16, 47, 16]       [32]        [29]        [33]   \n",
       "gd_sum                 -0.729381  -0.514713   -0.644046   -0.393607   \n",
       "total                   -0.73428  -0.528842   -0.645331   -0.387439   \n",
       "\n",
       "         plasmid_7   plasmid_8  plasmid_9  \n",
       "GC_sum -0.00989734 -0.00837569 -0.0128504  \n",
       "chain         [47]        [58]       [49]  \n",
       "gd_sum   -0.833333   -0.833333  -0.766323  \n",
       "total    -0.843231   -0.841709  -0.779173  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_2  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_8  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n",
      "plasmid_8  from greedy completely contained in  plasmid_5  from MILP.\n",
      "\n",
      "plasmid_8  from greedy completely contained in  plasmid_6  from MILP.\n",
      "\n",
      "plasmid_9  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_9  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_13  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_13  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_13  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
