{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 56 with $\\alpha_1 = 1$, $\\alpha_2 = 5$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 56\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP017931.1</td>\n",
       "      <td>111395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP017933.1</td>\n",
       "      <td>11638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP017930.1</td>\n",
       "      <td>114139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP017929.1</td>\n",
       "      <td>113105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP017932.1</td>\n",
       "      <td>76186</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP017931.1       111395\n",
       "1  CP017933.1        11638\n",
       "2  CP017930.1       114139\n",
       "3  CP017929.1       113105\n",
       "4  CP017932.1        76186"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt0VeWd//HPNycQEsAoEMJVkioJ\nJIEAiaABJSgKTAXswgtp1VGhdLSMWmiVmflVrdpaHagzWKlgx4UWioAXSpURaEuK10pQkGsgIgEE\n5B4I11ye3x/ZYWJIwolm5yTh/VqL5dl7P+fZ33PWeTz5nP3svc05JwAAAAAIC3UBAAAAABoGwgEA\nAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAJ7wUBdQ\nW+3atXNxcXGhLqPROH78uFq2bBnqMoAGj7ECBIexUjurV68+4JyLCXUdQLAaXTiIi4tTTk5OqMto\nNLKzs5WZmRnqMoAGj7ECBIexUjtmlh/qGoDaYFoRAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7C\nAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAe38KBmb1kZvvMbH01283MpptZnpl9Zmb9/KoFAAAA\nwPn5eeRgtqThNWwfIam792+CpN/5WAsAAACA8/AtHDjnVko6VEOT0ZJecWU+knSxmXX0qx4AAAAA\nNQsP4b47S9pZYXmXt25P5YZmNkFlRxcUGxur7Ozs+qivSSgsLOT9gm/WfVkQ6hLqTHx0gLECBIHv\nFaBpC2U4sCrWuaoaOudmSZolSenp6S4zM9PHspqW7Oxs8X7BL3dNeTvUJdSZ2cNbMlaAIPC9AjRt\nobxa0S5JXSssd5G0O0S1AAAAABe8UIaDxZLu9K5adKWkAufcOVOKAAAAANQP36YVmdk8SZmS2pnZ\nLkmPSmomSc65FyQtkfRPkvIknZB0t1+1AAAAADg/38KBcy7rPNudpB/7tX8AAAAAtcMdkgEAAABI\nIhwAAAAA8BAOAAAAAEgK7X0OAAC4IMQ1sXuCAGi6OHIAAAAAQBLhAAAAAICHcAAAAABAEuEAAAAA\ngIdwAAAAAEAS4QAAAACAh3AAAAAAQBLhAAAAAICHm6BVgZvVAAAA4ELEkQMAAAAAkggHAAAAADyE\nAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQD\nAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMA\nAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAA\nAACSCAcAAAAAPIQDAAAAAJJ8DgdmNtzMcs0sz8ymVLH9UjNbYWafmtlnZvZPftYDAAAAoHq+hQMz\nC0h6XtIISUmSsswsqVKz/ydpgXOur6Sxkmb4VQ8AAACAmvl55KC/pDzn3Dbn3BlJr0oaXamNk3SR\n9zha0m4f6wEAAABQg3Af++4saWeF5V2SBlRq85ikZWb2r5JaShrqYz0AAAAAauBnOLAq1rlKy1mS\nZjvnppnZVZL+YGYpzrnSr3VkNkHSBEmKjY1Vdna2H/WeNblXsa/916fCwkLf3y9cuBgrQHAYKwAa\nCz/DwS5JXSssd9G504bGSRouSc65D82shaR2kvZVbOScmyVpliSlp6e7zMxMn0ouc9eUt33tvz7N\nHt5Sfr9fuHAxVoDgMFYANBZ+nnOwSlJ3M4s3s+YqO+F4caU2OyRdJ0lm1lNSC0n7fawJAAAAQDV8\nCwfOuWJJEyUtlbRJZVcl2mBmj5vZKK/ZZEk/NLO1kuZJuss5V3nqEQAAAIB64Oe0IjnnlkhaUmnd\nIxUeb5Q00M8aAAAAAASHOyQDAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADg\nIRwAAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAh\nHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEc\nAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwA\nAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAA\nAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkORzODCz4WaWa2Z5Zjalmja3mtlGM9tgZn/0\nsx4AAAAA1Qv3q2MzC0h6XtL1knZJWmVmi51zGyu06S7p3yQNdM4dNrP2ftUDAAAAoGZ+HjnoLynP\nObfNOXdG0quSRldq80NJzzvnDkuSc26fj/UAAAAAqIFvRw4kdZa0s8LyLkkDKrVJkCQze19SQNJj\nzrl3KndkZhMkTZCk2NhYZWdn+1HvWZN7Ffvaf30qLCz0/f3ChYuxAgSHsQKgsfAzHFgV61wV++8u\nKVNSF0nvmlmKc+7I157k3CxJsyQpPT3dZWZm1nmxFd015W1f+69Ps4e3lN/vFy5cjBUgOIwVAI2F\nn9OKdknqWmG5i6TdVbT5k3OuyDn3haRclYUFAAAAAPXMz3CwSlJ3M4s3s+aSxkpaXKnNIklDJMnM\n2qlsmtE2H2sCAAAAUA3fwoFzrljSRElLJW2StMA5t8HMHjezUV6zpZIOmtlGSSsk/cw5d9CvmgAA\nAABUz89zDuScWyJpSaV1j1R47CRN8v4BAAAACCHukAwAAABAEuEAAAAAgIdwAAAAAEBSLcKBmQ0y\ns7u9xzFmFu9fWQAAAADqW1DhwMwelfSwpH/zVjWTNMevogAAAADUv2CPHHxP0ihJxyXJObdbUmu/\nigIAAABQ/4INB2e8y446STKzlv6VBAAAACAUgg0HC8xspqSLzeyHkv4i6UX/ygIAAABQ34K6CZpz\nbqqZXS/pqKRESY8455b7WhkAAACAenXecGBmAUlLnXNDJREIAAAAgCbqvNOKnHMlkk6YWXQ91AMA\nAAAgRIKaViTplKR1ZrZc3hWLJMk5d78vVQEAAACod8GGg7e9fwAAAACaqGBPSH7ZzJpLSvBW5Trn\nivwrCwAAAEB9CyocmFmmpJclbZdkkrqa2T8751b6VxoAAACA+hTstKJpkm5wzuVKkpklSJonKc2v\nwgAAAADUr2BvgtasPBhIknNui6Rm/pQEAAAAIBSCPXKQY2b/I+kP3vIPJK32pyQAAAAAoRBsOLhX\n0o8l3a+ycw5WSprhV1EAAAAA6l+w4SBc0n87534jnb1rcoRvVQEAADRRq1evbh8eHv57SSkKfoo3\nUBdKJa0vLi4en5aWtq+qBsGGg79KGiqp0FuOlLRMUsa3LhEAAOACEh4e/vsOHTr0jImJORwWFuZC\nXQ8uHKWlpbZ///6kvXv3/l7SqKraBJtWWzjnyoOBvMdRdVAjAADAhSYlJibmKMEA9S0sLMzFxMQU\nqOyoVdVtguzruJn1K18ws3RJJ79lfQAAABeiMIIBQsX77FWbAYKdVvSApIVmtluSk9RJ0m3fvjwA\nAAAADUWw4SBeUl9Jl0r6nqQrVRYSAAAA8C3ETXm7Tm8qu/3X320wl5vv379/4tSpU3dec801J4Jp\nX1paqoyMjIQlS5bktWnTptTv+r6tt956q/W0adNiV6xYkedH/3Pnzo3esGFD5K9+9au9lbdFRUX1\nPXHixKfVPTcjIyPhT3/60+cxMTEltdlnsNOKfu6cOyrpYknXS5ol6Xe12REAAAAaj6Kionrf54IF\nC6KTk5NPNoZgUB9+8IMfFFQVDIKRlZV1cOrUqTG1fV6w4aA8cXxX0gvOuT9Jal7bnQEAACD0fvaz\nn3WMj49PzsjI6D5y5Mj4Rx55JFYq+6V/4sSJna+44orEJ598Mnb37t3hw4YNuywlJaVnSkpKz2XL\nlrWUpKNHj4bdcsstcSkpKT179uyZNGfOnIslqbCw0G688cbvJCQkJH33u9/9zqlTp0ySnn322Xbj\nxo3rWr7/adOmtRs/fnyXynXNnTu3zfe+970jkpSbm9s8Pj4++bbbbuvWvXv35FGjRsUvWrSodb9+\n/Xp069YtZcWKFVE11ZKbm9s8LS0tMSkpqWdSUlLP5cuXt5TKfu3v379/4vDhw78THx+fPGrUqPjS\n0pqzyJgxY+K+//3vX5qWlpYYFxeXMm/evOjKbVasWBHVt2/fHj179kzq27dvj7Vr10ZIUk5OTote\nvXr17NGjR1JCQkLSunXrIoJ9bdOnT2975513XipJmzdvbt6nT58eKSkpPR944IFO5fvNz89vlp6e\nntijR4+k7t27J7/zzjutJGns2LFH3njjjbZBfiTOCnZa0ZdmNlNllzN92swixHV5AQAAGp2VK1dG\n/fnPf75k3bp1G4uKiqxPnz5Jffv2PTvt58iRI4FVq1blStLIkSPjJ02a9NWwYcMKt27d2nzYsGHd\nt23btuHf//3fOw4ZMuTowoULtx84cCCQnp7ec9SoUUd/85vfxERGRpZu2bJl4z/+8Y/IgQMHJknS\nuHHjDiUnJyedPn16V0REhJszZ067mTNn5leubfXq1a0GDhx4dv3OnTtbzJ8/f1taWlp+7969e86d\nO7dtTk7O5j/+8Y8X//KXv+w4ZMiQz6urpVOnTsXvvvvulqioKLdu3bqIrKys76xfv36TJG3atCly\nzZo12+Li4orS0tJ6LF++vNWwYcMKK9dT0c6dOyM+/vjj3I0bN0YMHTo0cfTo0esqbk9NTT318ccf\nb27WrJkWLVrU+qGHHuqydOnSz5977rmY++6776t777330KlTp6y4uFhffvlls2BeW8X+77vvvkvH\njx+/f+LEiQefeuqps0cEXnrppTbXXXddwdNPP723uLhYx44dC5OkmJiYkjNnztjevXsDHTp0CHpq\nUbDh4FZJwyVNdc4dMbOOkn4W7E4AAADQMGRnZ7caMWLEkVatWjlJ7vrrrz9ScXtWVtah8sfvv//+\nRVu3bo0sXy4sLAwcPnw4LDs7+6KlS5dePH369A6SdPr0acvLy2v+3nvvtbr//vv3SdKAAQNOJiQk\nnJCkiy66qHTgwIHH5s+fH92rV69TRUVF1r9//3OufFlQUBB+ySWXnP0Zv3PnzqfL2yUkJJy89tpr\nj4aFhalfv34nnnzyyU7e66mylm7duhWNGzeu28aNGyPDwsKUn59/9ga+vXr1On7ZZZcVSVJycvKJ\nzz///LwzYsaMGXMoEAioV69ep7t27Xp6zZo1LSpuP3ToUOC2226L3759ewszc0VFRSZJV1111fGp\nU6d23LVrV/OxY8ce7tWr1+lgX1tFn3zySav//d///VySfvSjHx184oknukjSlVdeefxHP/pRXFFR\nUdjNN998OCMj4+z72rZt2+IdO3Y079ChQ9BXGQ0qHDjnTkh6o8LyHkl7gt0JAAAAGgbnar6mTOvW\nrc/+ce6cU05OziYvSHytj9deey0vNTX1dOXnm1mV/U6YMOHAL3/5yw4JCQmnbr/99gNVtQkEAq6k\npESBQECS1Lx587P7DQsLU4sWLZzXTiUlJVZTLZMmTerUvn37otdff/2L0tJSRUZGnj3xOyIi4my/\ngUBAxcXFVRddw+uqvPzwww93Hjx48LHly5d/npub2/zaa69NlKR/+Zd/OXT11Vcff/PNN6NHjBiR\nMGPGjO2JiYmng3ltlVV1CdwRI0YUrly5Mvf111+Pvuuuu+Lvv//+ryZOnHhQKgtKUVFRtTp/I9gj\nBwDQtO1ZIz02OtRV1I3HCkJdAYAGLDMzs/Dee+/tduLEiT1FRUX2l7/85eI777xzf1VtBw0adPTp\np59u/8QTT3wlSR988EFkRkbGySFDhhydNm1a7OzZs3eEhYXp/fffjxw4cODJQYMGFc6ZM6fNyJEj\nj61atarFli1bzt4099prrz0+ceLE5hs2bGi5bt26DVXtLz4+/tSmTZsiUlJSzgkd1amuloKCgkCX\nLl3OBAIB/fa3v21bUnL+mTU//vGPOw8YMOD4nXfeeaTytjfeeOOSiRMnHty8eXPEzp07I1JTU0/9\n7W9/a1W+/ejRo4EuXbqckaSZM2e2K1+/cePG5j179jydnJy8b9u2bRFr1qyJTExMDPr1levXr1/h\niy++2Oa+++479OKLL549l2DLli3N4+Pjz0yePPnA8ePHwz755JMoSQdLS0u1f//+ZrXdF+EAAAAg\nhOr70qODBw8+MXz48IKkpKTkzp07n+7du/fx6OjoKv9ynjVr1s7x48dfmpCQkFRSUmIDBgw4lpGR\nsePXv/717gkTJlzao0ePJOecdenS5fSKFSvyfvrTn+4bO3ZsfEJCQlJycvKJXr16Ha/Y30033XT4\ns88+i6ru8po33HBDwbJly1rXJhxUV8uDDz64b8yYMZctWrTokkGDBh2LjIw87y/oGzdujCw/Ibqy\nyy+//HT//v0TDx482Oy//uu/8qOior72K/7DDz+8d/z48fHTp0/vcPXVVx8tX/+HP/yhzcKFC9uG\nh4e7mJiYoqeeemr3kSNHAsG+vnIzZszYMXbs2O/MmDEjdtSoUYfL1y9durT19OnTO4SHh7uoqKiS\nuXPnfiFJ7733XlTfvn2PN2vWrFb7sfMdWmpo0tPTXU5Ojq/7iJvytq/916fZw1sqMzMz1GWgiWpS\nYyV1qzJzHw11GXWDIwcNTpMaK3yv1IqZrXbOpVdct3bt2u2pqalVTqupLwUFBWHR0dGlx44dC7vq\nqqsSX3jhhfxBgwYFdS+Cb2PIkCGXP/jgg1+NHj36WFXb8/Pzm2VlZcV98MEHW/2upSqDBg3q/t57\n752z7zFjxsTdeOONBXfffffhqp7XEN19991db7rppiNVvddr165tl5qaGlfV87jiEAAAwAXm9ttv\n79ajR4+k3r179xw5cuRhv4PBgQMHAnFxcSktWrQorS4YSFK3bt2K7rnnngOHDh0Kyd+oVQWDxiol\nJeVkTe91dZhWBAAAcIH585///EV97q9du3Yl27dvXx9M2/Hjxze4X+dff/317aGuobYmT578jY5O\nceQAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA8nJAMAAITSY9Fp529Um/4KznvfBDNLGz169KFF\nixZ9IUlFRUVq3759ap8+fY6vWLEib/r06W1zcnJavvLKKzsmTZrUqVWrViWPP/74VxX7CAQCad27\ndz9ZUlJil19++ckFCxZsr3h3ZUkqLS1VRkZGwpIlS/LatGlTesstt8T99a9/jW7btm3x1q1bq7wR\n2ltvvdU6Kyvrss6dO5+RpBtvvPHw1KlT91Ru96tf/SrmhRdeiN25c2fE7t2713bs2LFYkubNmxf9\n8ccft3z22Wd3B/+moRxHDgAAAC4wkZGRpbm5uZGFhYUmSW+++eZFsbGxRbXpIyIionTz5s0bt27d\nuqFZs2Zu2rRpMZXbLFiwIDo5OflkmzZtSiXpnnvuObB48eLzXi40PT29cPPmzRs3b968sapgIEmD\nBw8uXL58+ZZOnTqdqbj+tttuK3jnnXcuPnbsGH/nfgO8aQAAABeg6667rmDhwoUXS9K8efPajBkz\n5tA37WvQoEGFeXl5EZXXz507t03FOw6PGDGiMCYmpvib7qeigQMHnkxMTDxTeX1YWJgyMjKOzZ8/\nP7ou9nOhIRwAAABcgO64445D8+fPv+TEiRO2adOmqKuuuur4N+mnqKhIS5cuvahXr14nK29bvXp1\nq4EDB56332eeeSbmmWeeOXvk4dNPP22VmJiYdM0113TPyclpUb5+8ODBl2/fvr3Z+fpLT08//u67\n77aqzetAGV/POTCz4ZL+W1JA0u+dc7+upt3NkhZKusI5l+NnTQAAAJAGDBhwcteuXREvvvhim6FD\nhxbU9vmnT58O69GjR5LX17EHHnjgnJtuFRQUhF9yySWl5z776x566KH95Y8zMjKO5+fnfxYdHV06\nf/786DFjxlyen5+/XpL+/ve/5wVTW4cOHYr37t3bPPhXg3K+hQMzC0h6XtL1knZJWmVmi51zGyu1\nay3pfkn/8KsWAAAAnGv48OFHHn300a7Lli3L3bdvX63+Liw/56CmNoFAwJWUlCgQCATdb/n5CVLZ\n+QOTJk26dM+ePeHlJxwH4+TJk9aiRYvzhhKcy89pRf0l5Tnntjnnzkh6VdLoKto9IekZSad8rAUA\nAACV3HvvvQcmT568u3///udMCaoL8fHxpzZt2nTOuQg12bFjR3hpadnf9StWrIgqLS1VbGxsrc5T\nyM3NbZGcnOzLa2rq/JxW1FnSzgrLuyQNqNjAzPpK6uqce8vMflpdR2Y2QdIESYqNjVV2dnbdV1vB\n5F51cp5Mg1BYWOj7+4ULV5MaKxGdlJ34i1CXUTcY8w1OkxorfK/UvSAuPeqXyy67rOjnP//5vvO1\ne/bZZzvOnDkztnz5q6+++iyY/m+44YaCZcuWtU5JSTktSSNHjoz/6KOPWh8+fDg8Nja295QpU3b/\n5Cc/OVB+vsFDDz20f86cOZe89NJL7QOBgGvRokXpK6+8si0srOz37MGDB1/+8ssv58fFxRU9+eST\n7Z977rkOBw8ebJaampo0ZMiQgvnz5+dL0sqVK1s//fTTX36Dt+SCZ845fzo2u0XSMOfceG/5Dkn9\nnXP/6i2HSfqbpLucc9vNLFvST893zkF6errLyfH3tIS4KW/72n99mj28pTIzM0NdBpqoJjVWUrcq\nM/fRUJdRNx6r9dRh+KxJjRW+V2rFzFY759Irrlu7du321NTUc+bnN0X5+fnNsrKy4j744IPzXr60\nruzcuTP81ltv/c6HH364pb722disXbu2XWpqalxV2/ycVrRLUtcKy10kVbwZRWtJKZKyzWy7pCsl\nLTazrw0gAAAANE7dunUruueeew4cOnSo3q6QuW3btubTpk3bef6WqIqf04pWSepuZvGSvpQ0VtL3\nyzc65woktStfDvbIAQAAABqP8ePHH67P/Q0ePPhEfe6vqfEtxTnniiVNlLRU0iZJC5xzG8zscTMb\n5dd+AQAAGrjS0tJSC3URuDB5n71qr+Tk630OnHNLJC2ptO6Ratpm+lkLAABAA7F+//79STExMQVh\nYWH+nPwJVKG0tNT2798fLWl9dW18DQcAAAD4uuLi4vF79+79/d69e1Pk7/mfQGWlktYXFxePr64B\n4QAAAKAepaWl7ZPEFGs0SKRVAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAA\nPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8\n4aEuAAAANCJ71kiPjQ51FXXjsYJQVwA0OBw5AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAA\nAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAA\nACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACBJCg91AfDZnjXSY6ND\nXUXdeKwg1BUAAAA0aRw5AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwONr\nODCz4WaWa2Z5Zjaliu2TzGyjmX1mZn81s25+1gMAAACger6FAzMLSHpe0ghJSZKyzCypUrNPJaU7\n53pLek3SM37VAwAAAKBmfh456C8pzzm3zTl3RtKrkr52Ny7n3Arn3Alv8SNJXXysBwAAAEANzDnn\nT8dmN0sa7pwb7y3fIWmAc25iNe1/K2mvc+7JKrZNkDRBkmJjY9NeffVVX2out+7LpnMn3vjI02p1\neneoy6gbHfuEugJUwlhpoBgrDQ5jpYGqh7EyZMiQ1c65dN93BNSRcB/7tirWVZlEzOx2SemSBle1\n3Tk3S9IsSUpPT3eZmZl1VGLV7prytq/916fZqV8oM/fRUJdRN7KazpdrU8FYaaAYKw0OY6WBYqwA\n5/AzHOyS1LXCchdJ5/zUYGZDJf2HpMHOudM+1gMAAACgBn6ec7BKUnczizez5pLGSlpcsYGZ9ZU0\nU9Io59w+H2sBAAAAcB6+hQPnXLGkiZKWStokaYFzboOZPW5mo7xm/ymplaSFZrbGzBZX0x0AAAAA\nn/k5rUjOuSWSllRa90iFx0P93D8AAACA4HGHZAAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMA\nAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAA\nAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAA\nAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAA\nkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACS\nCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADy+hgMzG25muWaWZ2ZTqtgeYWbzve3/\nMLM4P+sBAAAAUD3fwoGZBSQ9L2mEpCRJWWaWVKnZOEmHnXOXS3pW0tN+1QMAAACgZn4eOegvKc85\nt805d0bSq5JGV2ozWtLL3uPXJF1nZuZjTQAAAACq4Wc46CxpZ4XlXd66Kts454olFUhq62NNAAAA\nAKoR7mPfVR0BcN+gjcxsgqQJ3mKhmeV+y9ouGEOkdpIOhLqOOvELDirBP4wVIDiMlVrrVh87AeqK\nn+Fgl6SuFZa7SNpdTZtdZhYuKVrSocodOedmSZrlU51NmpnlOOfSQ10H0NAxVoDgMFaAps3PaUWr\nJHU3s3gzay5prKTFldoslvRhyliAAAAFJklEQVTP3uObJf3NOXfOkQMAAAAA/vPtyIFzrtjMJkpa\nKikg6SXn3AYze1xSjnNusaT/kfQHM8tT2RGDsX7VAwAAAKBmxg/1TZuZTfCmZQGoAWMFCA5jBWja\nCAcAAAAAJPl8h2QAAAAAjQfhAAAAAIAkwsEFz8zizGy99zjTzN4KdU1AKJjZB+fZvsTMLq6veoCm\nwMyyzSzde7zdzNqFuiYANfPzPgfwkZmZys4ZKQ11LUBDY2YB51xJbZ7jnMs4z/Z/+nZVAU0P30VA\n08ORg0bE+5V/k5nNkPSJpDvM7EMz+8TMFppZK6/dFWb2gZmtNbOPzay199x3vbafmFmNfwgBDZX3\nWd5sZi+b2Wdm9pqZRXm/Sj5iZu9JusXMLjOzd8xstffZ7+E9P9bM3vTGx9rysWBmhd5/O5rZSjNb\nY2brzexqb/3ZXz3NbJK3bb2ZPVihrk1m9qKZbTCzZWYWGZI3CahB5c+vmT1tZvdV2P6YmU32Hv/M\nzFZ5Y+0X3rrK30Vdzex3ZpbjffZ/EZpXBqAuEA4an0RJr0i6XtI4SUOdc/0k5Uia5N1wbr6kB5xz\nqZKGSjopaZ+k6722t0maHorigTqSKGmWc663pKOSyv+wOeWcG+Sce1Vld1X/V+dcmqSfSprhtZku\n6e/e+OgnaUOlvr8vaalzro+kVElrKm40szRJd0saIOlKST80s77e5u6SnnfOJUs6ImlMXb1goC5U\n9fmV9KrKvhfK3SppoZndoLLPdH9JfSSlmdk1XptESa845/o65/Il/Yd31+TekgabWe96eUEA6hzT\nihqffOfcR2Z2o6QkSe+XHdVVc0kfqux/2Hucc6skyTl3VJLMrKWk35pZH0klkhJCUTxQR3Y65973\nHs+RdL/3eL4keUfRMlT2B075cyK8/14r6U5J8qYeFVTqe5Wkl8ysmaRFzrk1lbYPkvSmc+64t683\nJF2tsju+f1Gh/WpJcd/iNQJ+qO7z297MOkmKkXTYObfDzO6XdIOkT73ntlJZWNgh77uoQr+3mtkE\nlf1d0VFl30+f1ccLAlC3CAeNz3HvvyZpuXMuq+JG79eaqm5e8RNJX6nsl9AwSaf8LBLwWeXPePly\n+fgIk3TE+/W/dh07t9L7dfS7KruD+386516p0MSqeaokna7wuEQS04rQ0FT3+X1N0s2SOqjsSEJ5\n26ecczO/1oFZnP5vrMnM4lV2dO4K59xhM5stqUWdVg2g3jCtqPH6SNJAM7tckrw51wmSNkvqZGZX\neOtbm1m4pGiVHVEolXSHpECI6gbqwqVmdpX3OEvSexU3ekfMvjCzW6SykybNLNXb/FdJ93rrA2Z2\nUcXnmlk3Sfuccy9K+h+VTT2qaKWkm7wx11LS9yS9W3cvDfBVdZ/fVyWNVVlAeM1ru1TSPRXOZ+ts\nZu2r6PMilYWFAjOLlTTC59cAwEeEg0bKObdf0l2S5pnZZyoLCz2cc2dUNnf0OTNbK2m5yn7BmSHp\nn83sI5VNKTpeZcdA47BJZZ/nzyS1kfS7Ktr8QNI4bxxskDTaW/+ApCFmtk5lU3+SKz0vU9IaM/tU\nZecM/HfFjc65TyTNlvSxpH9I+r1z7lMBjUB1n1/n3AZJrSV96Zzb47VdJumPkj70xstrXpvKfa5V\n2dSjDZJekvR+5TYAGg9zrqoZKADQMHlTGt5yzqWEuBQAAJocjhwAAAAAkMSRAwAAAAAejhwAAAAA\nkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEcAAAAAJAk/X+Q8rp5LGuW5AAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7ff2059b1be0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:5:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>179980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>11638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>11638</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0       179980\n",
       "1  plasmid_1        11638\n",
       "2  plasmid_2        11638"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.239406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.463750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.315790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.239406\n",
       "1  precision  0.463750\n",
       "2   f1_score  0.315790"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP017929.1</th>\n",
       "      <th>CP017930.1</th>\n",
       "      <th>CP017931.1</th>\n",
       "      <th>CP017932.1</th>\n",
       "      <th>CP017933.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.031015</td>\n",
       "      <td>0.059296</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.931588</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.026400</td>\n",
       "      <td>0.054530</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.026400</td>\n",
       "      <td>0.054530</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.083816</td>\n",
       "      <td>0.168356</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.931588</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP017929.1  CP017930.1  CP017931.1  CP017932.1  CP017933.1\n",
       "0  plasmid_0    0.031015    0.059296         0.0    0.931588         0.0\n",
       "1  plasmid_1    0.026400    0.054530         0.0    0.000000         1.0\n",
       "2  plasmid_2    0.026400    0.054530         0.0    0.000000         1.0\n",
       "3      Total    0.083816    0.168356         0.0    0.931588         2.0"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP017929.1</td>\n",
       "      <td>0.019491</td>\n",
       "      <td>0.256058</td>\n",
       "      <td>0.256058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP017930.1</td>\n",
       "      <td>0.019986</td>\n",
       "      <td>0.534800</td>\n",
       "      <td>0.534800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP017931.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP017932.1</td>\n",
       "      <td>0.394344</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP017933.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.433820</td>\n",
       "      <td>1.790858</td>\n",
       "      <td>1.790858</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2\n",
       "0  CP017929.1   0.019491   0.256058   0.256058\n",
       "1  CP017930.1   0.019986   0.534800   0.534800\n",
       "2  CP017931.1   0.000000   0.000000   0.000000\n",
       "3  CP017932.1   0.394344   0.000000   0.000000\n",
       "4  CP017933.1   0.000000   1.000000   1.000000\n",
       "5       Total   0.433820   1.790858   1.790858"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.394399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.394399\n",
       "1  plasmid_1    1.000000\n",
       "2  plasmid_2    1.000000"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP017929.1</td>\n",
       "      <td>0.057416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP017930.1</td>\n",
       "      <td>0.113826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP017931.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP017932.1</td>\n",
       "      <td>0.931588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP017933.1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP017929.1    0.057416\n",
       "1  CP017930.1    0.113826\n",
       "2  CP017931.1    0.000000\n",
       "3  CP017932.1    0.931588\n",
       "4  CP017933.1    1.000000"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP017933.1</td>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP017933.1</td>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction  recall  precision\n",
       "0  CP017933.1  plasmid_1     1.0        1.0\n",
       "1  CP017933.1  plasmid_2     1.0        1.0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 5, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00241854</td>\n",
       "      <td>0.00559941</td>\n",
       "      <td>-0.0116921</td>\n",
       "      <td>-0.00887644</td>\n",
       "      <td>-0.00425202</td>\n",
       "      <td>-0.00311174</td>\n",
       "      <td>0.00777197</td>\n",
       "      <td>-0.0112146</td>\n",
       "      <td>-0.00439763</td>\n",
       "      <td>0.00548618</td>\n",
       "      <td>0.000933598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[19, 211, 107, 129, 91, 62, 86, 71, 106, 65, 8...</td>\n",
       "      <td>[53, 49]</td>\n",
       "      <td>[110, 194, 155, 216, 143, 150, 145, 151, 187, ...</td>\n",
       "      <td>[101, 86, 59, 91, 156, 60, 106, 67, 116, 107, ...</td>\n",
       "      <td>[53, 49, 63, 72, 48, 68, 47]</td>\n",
       "      <td>[212, 177, 38, 97, 56, 180, 152, 215, 157, 88,...</td>\n",
       "      <td>[107, 116, 54, 65]</td>\n",
       "      <td>[55, 195, 155, 216, 143, 150, 145, 151, 187, 1...</td>\n",
       "      <td>[212, 177, 38, 97, 56, 180, 152, 226, 220, 157...</td>\n",
       "      <td>[210, 72]</td>\n",
       "      <td>[109, 72]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.717612</td>\n",
       "      <td>-0.624749</td>\n",
       "      <td>-0.471059</td>\n",
       "      <td>-0.685021</td>\n",
       "      <td>-0.637577</td>\n",
       "      <td>-0.490536</td>\n",
       "      <td>-0.801158</td>\n",
       "      <td>-0.620555</td>\n",
       "      <td>-0.568184</td>\n",
       "      <td>-0.750735</td>\n",
       "      <td>-0.548187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.720031</td>\n",
       "      <td>-0.61915</td>\n",
       "      <td>-0.482751</td>\n",
       "      <td>-0.693897</td>\n",
       "      <td>-0.641829</td>\n",
       "      <td>-0.493647</td>\n",
       "      <td>-0.793386</td>\n",
       "      <td>-0.631769</td>\n",
       "      <td>-0.572582</td>\n",
       "      <td>-0.745249</td>\n",
       "      <td>-0.547254</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0   plasmid_1  \\\n",
       "GC_sum                                        -0.00241854  0.00559941   \n",
       "chain   [19, 211, 107, 129, 91, 62, 86, 71, 106, 65, 8...    [53, 49]   \n",
       "gd_sum                                          -0.717612   -0.624749   \n",
       "total                                           -0.720031    -0.61915   \n",
       "\n",
       "                                               plasmid_10  \\\n",
       "GC_sum                                         -0.0116921   \n",
       "chain   [110, 194, 155, 216, 143, 150, 145, 151, 187, ...   \n",
       "gd_sum                                          -0.471059   \n",
       "total                                           -0.482751   \n",
       "\n",
       "                                                plasmid_2  \\\n",
       "GC_sum                                        -0.00887644   \n",
       "chain   [101, 86, 59, 91, 156, 60, 106, 67, 116, 107, ...   \n",
       "gd_sum                                          -0.685021   \n",
       "total                                           -0.693897   \n",
       "\n",
       "                           plasmid_3  \\\n",
       "GC_sum                   -0.00425202   \n",
       "chain   [53, 49, 63, 72, 48, 68, 47]   \n",
       "gd_sum                     -0.637577   \n",
       "total                      -0.641829   \n",
       "\n",
       "                                                plasmid_4           plasmid_5  \\\n",
       "GC_sum                                        -0.00311174          0.00777197   \n",
       "chain   [212, 177, 38, 97, 56, 180, 152, 215, 157, 88,...  [107, 116, 54, 65]   \n",
       "gd_sum                                          -0.490536           -0.801158   \n",
       "total                                           -0.493647           -0.793386   \n",
       "\n",
       "                                                plasmid_6  \\\n",
       "GC_sum                                         -0.0112146   \n",
       "chain   [55, 195, 155, 216, 143, 150, 145, 151, 187, 1...   \n",
       "gd_sum                                          -0.620555   \n",
       "total                                           -0.631769   \n",
       "\n",
       "                                                plasmid_7   plasmid_8  \\\n",
       "GC_sum                                        -0.00439763  0.00548618   \n",
       "chain   [212, 177, 38, 97, 56, 180, 152, 226, 220, 157...   [210, 72]   \n",
       "gd_sum                                          -0.568184   -0.750735   \n",
       "total                                           -0.572582   -0.745249   \n",
       "\n",
       "          plasmid_9  \n",
       "GC_sum  0.000933598  \n",
       "chain     [109, 72]  \n",
       "gd_sum    -0.548187  \n",
       "total     -0.547254  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_11</th>\n",
       "      <th>plasmid_12</th>\n",
       "      <th>plasmid_13</th>\n",
       "      <th>plasmid_14</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00347185</td>\n",
       "      <td>0.00216905</td>\n",
       "      <td>0.00559941</td>\n",
       "      <td>-0.00421322</td>\n",
       "      <td>-0.0149064</td>\n",
       "      <td>-0.00340827</td>\n",
       "      <td>-0.00112779</td>\n",
       "      <td>-0.00138778</td>\n",
       "      <td>-0.00298008</td>\n",
       "      <td>-0.0139741</td>\n",
       "      <td>-0.0123872</td>\n",
       "      <td>-0.00976398</td>\n",
       "      <td>-0.0103123</td>\n",
       "      <td>-0.00507725</td>\n",
       "      <td>-0.00740681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[28, 155, 216, 143, 130, 151, 186, 167, 104, 1...</td>\n",
       "      <td>[42, 165, 140, 125, 98, 49, 63, 72, 8, 107, 11...</td>\n",
       "      <td>[49, 53]</td>\n",
       "      <td>[53]</td>\n",
       "      <td>[107, 129]</td>\n",
       "      <td>[31, 109, 72, 30, 64]</td>\n",
       "      <td>[96, 40, 72, 13, 148]</td>\n",
       "      <td>[168, 66, 106, 65, 19, 211, 107, 116, 54, 65, ...</td>\n",
       "      <td>[68, 48, 72, 63, 49, 98, 125, 174, 125]</td>\n",
       "      <td>[62]</td>\n",
       "      <td>[168, 187, 151, 145, 150, 143, 216, 155, 194, 41]</td>\n",
       "      <td>[16]</td>\n",
       "      <td>[205, 143, 216, 155, 195, 55, 194, 155]</td>\n",
       "      <td>[60]</td>\n",
       "      <td>[67, 116, 107]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.705353</td>\n",
       "      <td>-0.0947092</td>\n",
       "      <td>-0.624749</td>\n",
       "      <td>-0.458687</td>\n",
       "      <td>-0.585823</td>\n",
       "      <td>-0.0148062</td>\n",
       "      <td>-0.00737403</td>\n",
       "      <td>-0.72755</td>\n",
       "      <td>-0.679217</td>\n",
       "      <td>-0.7026</td>\n",
       "      <td>-0.677174</td>\n",
       "      <td>-0.724038</td>\n",
       "      <td>-0.593107</td>\n",
       "      <td>-0.690018</td>\n",
       "      <td>-0.642381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.708825</td>\n",
       "      <td>-0.0925401</td>\n",
       "      <td>-0.61915</td>\n",
       "      <td>-0.462901</td>\n",
       "      <td>-0.60073</td>\n",
       "      <td>-0.0182144</td>\n",
       "      <td>-0.00850182</td>\n",
       "      <td>-0.728938</td>\n",
       "      <td>-0.682197</td>\n",
       "      <td>-0.716574</td>\n",
       "      <td>-0.689561</td>\n",
       "      <td>-0.733802</td>\n",
       "      <td>-0.603419</td>\n",
       "      <td>-0.695095</td>\n",
       "      <td>-0.649787</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0  \\\n",
       "GC_sum                                        -0.00347185   \n",
       "chain   [28, 155, 216, 143, 130, 151, 186, 167, 104, 1...   \n",
       "gd_sum                                          -0.705353   \n",
       "total                                           -0.708825   \n",
       "\n",
       "                                                plasmid_1  plasmid_10  \\\n",
       "GC_sum                                         0.00216905  0.00559941   \n",
       "chain   [42, 165, 140, 125, 98, 49, 63, 72, 8, 107, 11...    [49, 53]   \n",
       "gd_sum                                         -0.0947092   -0.624749   \n",
       "total                                          -0.0925401    -0.61915   \n",
       "\n",
       "        plasmid_11  plasmid_12             plasmid_13             plasmid_14  \\\n",
       "GC_sum -0.00421322  -0.0149064            -0.00340827            -0.00112779   \n",
       "chain         [53]  [107, 129]  [31, 109, 72, 30, 64]  [96, 40, 72, 13, 148]   \n",
       "gd_sum   -0.458687   -0.585823             -0.0148062            -0.00737403   \n",
       "total    -0.462901    -0.60073             -0.0182144            -0.00850182   \n",
       "\n",
       "                                                plasmid_2  \\\n",
       "GC_sum                                        -0.00138778   \n",
       "chain   [168, 66, 106, 65, 19, 211, 107, 116, 54, 65, ...   \n",
       "gd_sum                                           -0.72755   \n",
       "total                                           -0.728938   \n",
       "\n",
       "                                      plasmid_3  plasmid_4  \\\n",
       "GC_sum                              -0.00298008 -0.0139741   \n",
       "chain   [68, 48, 72, 63, 49, 98, 125, 174, 125]       [62]   \n",
       "gd_sum                                -0.679217    -0.7026   \n",
       "total                                 -0.682197  -0.716574   \n",
       "\n",
       "                                                plasmid_5   plasmid_6  \\\n",
       "GC_sum                                         -0.0123872 -0.00976398   \n",
       "chain   [168, 187, 151, 145, 150, 143, 216, 155, 194, 41]        [16]   \n",
       "gd_sum                                          -0.677174   -0.724038   \n",
       "total                                           -0.689561   -0.733802   \n",
       "\n",
       "                                      plasmid_7   plasmid_8       plasmid_9  \n",
       "GC_sum                               -0.0103123 -0.00507725     -0.00740681  \n",
       "chain   [205, 143, 216, 155, 195, 55, 194, 155]        [60]  [67, 116, 107]  \n",
       "gd_sum                                -0.593107   -0.690018       -0.642381  \n",
       "total                                 -0.603419   -0.695095       -0.649787  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_5  from MILP completely contained in  plasmid_1  from greedy.\n",
      "\n",
      "plasmid_5  from MILP completely contained in  plasmid_2  from greedy.\n",
      "\n",
      "plasmid_4  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_4  from greedy completely contained in  plasmid_7  from MILP.\n",
      "\n",
      "plasmid_5  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n",
      "plasmid_8  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n",
      "plasmid_9  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n",
      "plasmid_1  from MILP completely contained in  plasmid_10  from greedy.\n",
      "\n",
      "plasmid_10  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_10  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_11  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_11  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_12  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_12  from greedy completely contained in  plasmid_7  from MILP.\n",
      "\n",
      "plasmid_9  from MILP completely contained in  plasmid_13  from greedy.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
