{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 125 with $\\alpha_1 = 1$, $\\alpha_2 = 1$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 125\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016582.1</td>\n",
       "      <td>2094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016584.1</td>\n",
       "      <td>6647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016583.1</td>\n",
       "      <td>37697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP016585.1</td>\n",
       "      <td>99473</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP016582.1         2094\n",
       "1  CP016584.1         6647\n",
       "2  CP016583.1        37697\n",
       "3  CP016585.1        99473"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt0VeWZx/HfkwRCuIgCIVwlqRIg\nCQRIBAUsQVFgqmAXXqDVjgrF0TJqoVVmplVrba2O1BZHKthxqZVy80JpZQTageK1EhTkZgCRm4iA\nQLjHXJ75IztMDEk4aHZOCN/PWi7P3vs9737O8bzm/M5+997m7gIAAACAmGgXAAAAAKBuIBwAAAAA\nkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABCIi3YBp6tV\nq1aenJwc7TLOGEeOHFGTJk2iXQZQ5zFWgMgwVk7PihUr9rp7YrTrACJ1xoWD5ORk5ebmRruMM8bS\npUuVk5MT7TKAOo+xAkSGsXJ6zGxrtGsATgfTigAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4A\nAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACIQWDszsGTPbbWZrqthuZjbFzDaZ2Qdm1jusWgAAAACc\nWphHDp6VNLSa7cMkdQ7+GSfpdyHWAgAAAOAUQgsH7r5M0r5qmoyQ9LyXekfSuWbWNqx6AAAAAFQv\nLor7bi9pe7nlHcG6Tys2NLNxKj26oKSkJC1dujTUwlZ/kh9q/7UppXls6O8Xzl71aqwkFGjpzN9E\nu4ya0bZntCtABfVqrPB3BajXohkOrJJ1XllDd58uabokZWdne05OTohlSTdPejXU/mvTs0ObKOz3\nC2evejVWMj9WTt790S6jZoyuP19E64t6NVb4uwLUa9G8WtEOSR3LLXeQtDNKtQAAAABnvWiGg/mS\nvhdctehiSfnuftKUIgAAAAC1I7RpRWY2U1KOpFZmtkPS/ZIaSJK7PyVpgaR/krRJ0lFJt4RVCwAA\nAIBTCy0cuPvoU2x3ST8Ia/8AAAAATg93SAYAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAA\nIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAg\niXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJ\ncAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlw\nAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAA\nAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACAQajgws6Fmlmdmm8xsUiXbzzezJWb2vpl9\nYGb/FGY9AAAAAKoWF1bHZhYr6UlJV0jaIWm5mc1393Xlmv1E0hx3/52ZpUlaICk5rJoAAMDX9OlK\n6YER0a6iZjyQH+0KgDonzCMHfSRtcvfN7v6FpFmSKv7fxCWdEzxuLmlniPUAAAAAqEZoRw4ktZe0\nvdzyDkl9K7R5QNIiM/tXSU0kDQ6xHgAAAADVCDMcWCXrvMLyaEnPuvtkM7tE0h/MLMPdS77Ukdk4\nSeMkKSkpSUuXLg2j3hMmdi8Ktf/adPjw4dDfL5y96tVYiW+npV1+Fu0yagZjvs5hrNRRjBXgJGGG\ngx2SOpZb7qCTpw2NkTRUktz9bTNrJKmVpN3lG7n7dEnTJSk7O9tzcnJCKrnUzZNeDbX/2vTs0CYK\n+/3C2atejZXMj5WTd3+0y6gZo5lHXdcwVuooxgpwkjDPOVguqbOZpZhZQ0mjJM2v0GabpMslycy6\nSWokaU+INQEAAACoQmjhwN2LJI2XtFDSepVelWitmT1oZsODZhMlfd/MVkmaKelmd6849QgAAABA\nLQhzWpHcfYFKL09aft195R6vk9Q/zBoAAAAARIY7JAMAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQ\nIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAg\nHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAc\nAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQJMVFuwCE7NOV0gMjol1FzXggP9oVAAAA\n1GscOQAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAA\nIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAg\nQDgAAAAAICnkcGBmQ80sz8w2mdmkKtpcb2brzGytmf0xzHoAAAAAVC0urI7NLFbSk5KukLRD0nIz\nm+/u68q16Szp3yT1d/f9ZtY6rHoAAAAAVC/MIwd9JG1y983u/oWkWZJGVGjzfUlPuvt+SXL33SHW\nAwAAAKAa5u7hdGx2raSh7j42WL5JUl93H1+uzTxJGyT1lxQr6QF3f62SvsZJGidJSUlJWbNmzQql\n5jKrP8kPtf/alJJQoKYFO6NdRs1o2zPaFaACxkodxVipcxgrdVQtjJVBgwatcPfs0HcE1JDQphVJ\nskrWVUwicZI6S8qR1EHS62aW4e4HvvQk9+mSpktSdna25+Tk1Hix5d086dVQ+69Nz2Z+rJy8+6Nd\nRs0YXX/+uNYXjJU6irFS5zBW6ijGCnCSMKcV7ZDUsdxyB0kVf2rYIelP7l7o7h9LylNpWAAAAABQ\ny8IMB8sldTazFDNrKGmUpPkV2syTNEiSzKyVpFRJm0OsCQAAAEAVQgsH7l4kabykhZLWS5rj7mvN\n7EEzGx40WyjpczNbJ2mJpB+7++dh1QQAAACgamGecyB3XyBpQYV195V77JImBP8AAAAAiCLukAwA\nAABAEuEAAAAAQIBwAAAAAEDSaYQDMxtgZrcEjxPNLCW8sgAAAADUtojCgZndL+leSf8WrGog6YWw\nigIAAABQ+yI9cvBtScMlHZEkd98pqVlYRQEAAACofZGGgy+Cy466JJlZk/BKAgAAABANkYaDOWY2\nTdK5ZvZ9SX+V9HR4ZQEAAACobRHdBM3dHzOzKyQdlNRF0n3uvjjUygAAAADUqlOGAzOLlbTQ3QdL\nIhAAAAAA9dQppxW5e7Gko2bWvBbqAQAAABAlEU0rknRc0mozW6zgikWS5O53hlIVAAAAgFoXaTh4\nNfgHAAAAQD0V6QnJz5lZQ0mpwao8dy8MrywAAAAAtS2icGBmOZKek7RFkknqaGb/7O7LwisNAAAA\nQG2KdFrRZElXunueJJlZqqSZkrLCKgwAAABA7Yr0JmgNyoKBJLn7BkkNwikJAAAAQDREeuQg18z+\nW9IfguXvSloRTkkAAAAAoiHScHC7pB9IulOl5xwskzQ1rKIAAAAA1L5Iw0GcpN+6+6+lE3dNjg+t\nKgAAgHpqxYoVrePi4n4vKUORT/EGakKJpDVFRUVjs7KydlfWINJw8DdJgyUdDpYTJC2S1O9rlwgA\nAHAWiYuL+32bNm26JSYm7o+JifFo14OzR0lJie3Zsydt165dv5c0vLI2kabVRu5eFgwUPG5cAzUC\nAACcbTISExMPEgxQ22JiYjwxMTFfpUetKm8TYV9HzKx32YKZZUs69jXrAwAAOBvFEAwQLcFnr8oM\nEOm0orskzTWznZJcUjtJN3z98gAAAADUFZGGgxRJvSSdL+nbki5WaUgAAADA15A86dUavansll99\nq85cbr5Pnz5dHnvsse3f/OY3j0bSvqSkRP369UtdsGDBphYtWpSEXd/X9Ze//KXZ5MmTk5YsWbIp\njP5nzJjRfO3atQm//OUvd1Xc1rhx415Hjx59v6rn9uvXL/VPf/rTR4mJicWns89IpxX91N0PSjpX\n0hWSpkv63ensCAAAAGeOwsLCWt/nnDlzmqenpx87E4JBbfjud7+bX1kwiMTo0aM/f+yxxxJP93mR\nhoOyxPEtSU+5+58kNTzdnQEAACD6fvzjH7dNSUlJ79evX+err7465b777kuSSn/pHz9+fPuLLrqo\ny0MPPZS0c+fOuCFDhlyQkZHRLSMjo9uiRYuaSNLBgwdjrrvuuuSMjIxu3bp1S3vhhRfOlaTDhw/b\nVVdd9Y3U1NS0b33rW984fvy4SdLjjz/easyYMR3L9j958uRWY8eO7VCxrhkzZrT49re/fUCS8vLy\nGqakpKTfcMMNnTp37pw+fPjwlHnz5jXr3bt3106dOmUsWbKkcXW15OXlNczKyuqSlpbWLS0trdvi\nxYubSKW/9vfp06fL0KFDv5GSkpI+fPjwlJKS6rPIyJEjk7/zne+cn5WV1SU5OTlj5syZzSu2WbJk\nSeNevXp17datW1qvXr26rlq1Kl6ScnNzG3Xv3r1b165d01JTU9NWr14dH+lrmzJlSsvvfe9750vS\nhx9+2LBnz55dMzIyut11113tyva7devWBtnZ2V26du2a1rlz5/TXXnutqSSNGjXqwMsvv9wywo/E\nCZFOK/rEzKap9HKmj5hZvLguLwAAwBln2bJljf/85z+ft3r16nWFhYXWs2fPtF69ep2Y9nPgwIHY\n5cuX50nS1VdfnTJhwoTPhgwZcnjjxo0NhwwZ0nnz5s1r//3f/73toEGDDs6dO3fL3r17Y7Ozs7sN\nHz784K9//evEhISEkg0bNqz7xz/+kdC/f/80SRozZsy+9PT0tIKCgh3x8fH+wgsvtJo2bdrWirWt\nWLGiaf/+/U+s3759e6PZs2dvzsrK2tqjR49uM2bMaJmbm/vhH//4x3N/8YtftB00aNBHVdXSrl27\notdff31D48aNffXq1fGjR4/+xpo1a9ZL0vr16xNWrly5OTk5uTArK6vr4sWLmw4ZMuRwxXrK2759\ne/y7776bt27duvjBgwd3GTFixOry2zMzM4+/++67HzZo0EDz5s1rds8993RYuHDhR0888UTiHXfc\n8dntt9++7/jx41ZUVKRPPvmkQSSvrXz/d9xxx/ljx47dM378+M8ffvjhE0cEnnnmmRaXX355/iOP\nPLKrqKhIhw4dipGkxMTE4i+++MJ27doV26ZNm4inFkUaDq6XNFTSY+5+wMzaSvpxpDsBAABA3bB0\n6dKmw4YNO9C0aVOX5FdcccWB8ttHjx69r+zxm2++ec7GjRsTypYPHz4cu3///pilS5ees3DhwnOn\nTJnSRpIKCgps06ZNDd94442md955525J6tu377HU1NSjknTOOeeU9O/f/9Ds2bObd+/e/XhhYaH1\n6dPnpCtf5ufnx5133nknfsZv3759QVm71NTUY5dddtnBmJgY9e7d++hDDz3ULng9ldbSqVOnwjFj\nxnRat25dQkxMjLZu3XriBr7du3c/csEFFxRKUnp6+tGPPvrolDNiRo4cuS82Nlbdu3cv6NixY8HK\nlSsbld++b9++2BtuuCFly5YtjczMCwsLTZIuueSSI4899ljbHTt2NBw1atT+7t27F0T62sp77733\nmv7P//zPR5J02223ff7zn/+8gyRdfPHFR2677bbkwsLCmGuvvXZ/v379TryvLVu2LNq2bVvDNm3a\nRHyV0YjCgbsflfRyueVPJX0a6U4AAABQN7hXf02ZZs2anfhy7u7Kzc1dHwSJL/Xx4osvbsrMzCyo\n+Hwzq7TfcePG7f3FL37RJjU19fiNN964t7I2sbGxXlxcrNjYWElSw4YNT+w3JiZGjRo18qCdiouL\nrbpaJkyY0K5169aFL7300sclJSVKSEg4ceJ3fHz8iX5jY2NVVFRUedHVvK6Ky/fee2/7gQMHHlq8\nePFHeXl5DS+77LIukvQv//Iv+y699NIjr7zySvNhw4alTp06dUuXLl0KInltFVV2Cdxhw4YdXrZs\nWd5LL73U/Oabb0658847Pxs/fvznUmlQaty48Wmdv8HUIAAAgLNITk7O4YULFzY/evSo5efnx/z1\nr389t6q2AwYMOPjII4+0Llt+6623EiRp0KBBBydPnpxUNlf/zTffTAjaH37hhRdaSNLy5csbbdiw\n4cRNcy+77LIjn376acNXXnml5ZgxY/apEikpKcfXr18fX9m2qlRVS35+fmzbtm0LY2NjNXXq1JbF\nxaeeWfODH/yg/fPPP1/p+/Hyyy+fV1xcrLVr18Zv3749PjMz83j57QcPHozt0KHDF5I0bdq0VmXr\n161b17Bbt24FP/nJT3ZfeeWVB1auXJlQse9I9O7d+/DTTz/dQpKefvrpE+cSbNiwoWH79u0LJ06c\nuPfGG2/c+9577zWWSq/8tGfPngZdunQ5KcBVJ9JpRQAAAAhBbV96dODAgUeHDh2an5aWlt6+ffuC\nHj16HGnevHml35ynT5++fezYseenpqamFRcXW9++fQ/169dv269+9aud48aNO79r165p7m4dOnQo\nWLJkyaYf/ehHu0eNGpWSmpqalp6efrR79+5Hyvd3zTXX7P/ggw8aV3V5zSuvvDJ/0aJFzTIyMiL+\nQltVLXfffffukSNHXjBv3rzzBgwYcCghIeGUv6CvW7cuoeyE6IouvPDCgj59+nT5/PPPG/zmN7/Z\n2rhx4y/9in/vvffuGjt2bMqUKVPaXHrppQfL1v/hD39oMXfu3JZxcXGemJhY+PDDD+88cOBAbKSv\nr8zUqVO3jRo16htTp05NGj58+P6y9QsXLmw2ZcqUNnFxcd64cePiGTNmfCxJb7zxRuNevXodadCg\nwWntx051aKmuyc7O9tzc3FD3kTzp1VD7r03PZm5UTt790S6jZjyQH+0KUAFjpY5irNQ5jJU6qhbG\nipmtcPfs8utWrVq1JTMzs9JpNbUlPz8/pnnz5iWHDh2KueSSS7o89dRTWwcMGBDRvQi+jkGDBl14\n9913fzZixIhDlW3funVrg9GjRye/9dZbG8OupTIDBgzo/MYbb5y075EjRyZfddVV+bfccsv+yp5X\nF91yyy0dr7nmmgOVvderVq1qlZmZmVzZ85hWBAAAcJa58cYbO3Xt2jWtR48e3a6++ur9YQeDvXv3\nxiYnJ2c0atSopKpgIEmdOnUqvPXWW/fu27cvKt9RKwsGZ6qMjIxj1b3XVWFaEQAAwFnmz3/+88e1\nub9WrVoVb9myZU0kbceOHVvnfp1/6aWXtkS7htM1ceLEr3R0iiMHAAAAACQRDgAAAAAECAcAAAAA\nJBEOAAAAAAQ4IRkAACCaHmiedepGp9Nf/invm2BmWSNGjNg3b968jyWpsLBQrVu3zuzZs+eRJUuW\nbJoyZUrL3NzcJs8///y2CRMmtGvatGnxgw8++Fn5PmJjY7M6d+58rLi42C688MJjc+bM2VL+7spS\n6Y24+vXrl7pgwYJNLVq0KLnuuuuS//a3vzVv2bJl0caNG9dWVtv777/f6JZbbklet25d40mTJn1S\ncb9lfvnLXyY+9dRTSdu3b4/fuXPnqrZt2xZJ0syZM5u/++67TR5//PGdkb5l+H8cOQAAADjLJCQk\nlOTl5SUcPnzYJOmVV145JykpqfB0+oiPjy/58MMP123cuHFtgwYNfPLkyYkV28yZM6d5enr6sRYt\nWpRI0q233rp3/vz51V4utHXr1kW//e1vt912222VhoIyAwcOPLx48eIN7dq1+6L8+htuuCH/tdde\nO/fQoUN8z/0KeNMAAADOQpdffnn+3Llzz5WkmTNnthg5cuS+r9rXgAEDDm/atCm+4voZM2a0KH/H\n4WHDhh1OTEwsqq6v9u3bFw0cOPBogwYNqr1Tb//+/Y916dLli4rrY2Ji1K9fv0OzZ89ufjqvAaUI\nBwAAAGehm266ad/s2bPPO3r0qK1fv77xJZdccuSr9FNYWKiFCxee071792MVt61YsaJp//79T9nv\no48+mvjoo4+edOShooEDB164ZcuWBqdql52dfeT1119veqp2OFmo5xyY2VBJv5UUK+n37v6rKtpd\nK2mupIvcPTfMmgAAACD17dv32I4dO+KffvrpFoMHD84/3ecXFBTEdO3aNS3o69Bdd9110k238vPz\n484777ySk5/9Zffcc8+eSPb597//fVMk7dq0aVO0a9euhpG0xZeFFg7MLFbSk5KukLRD0nIzm+/u\n6yq0aybpTkn/CKsWAAAAnGzo0KEH7r///o6LFi3K271792l9Lyw756C6NrGxsV5cXKzY2NivV+hp\nOnbsmDVq1OiUoQQnC3NaUR9Jm9x9s7t/IWmWpBGVtPu5pEclHQ+xFgAAAFRw++237504ceLOPn36\nnDQlqCakpKQcX79+/UnnIoQtLy+vUXp6eiivqb4Lc1pRe0nbyy3vkNS3fAMz6yWpo7v/xcx+VFVH\nZjZO0jhJSkpK0tKlS2u+2nImdq/2PJkzyuH4dlra5WfRLqNmhPzfHaePsVJHMVbqHMZKHVVXxkoE\nlx4NywUXXFD405/+dPep2j3++ONtp02bllS2/Nlnn30QSf9XXnll/qJFi5plZGQUSNLVV1+d8s47\n7zTbv39/XFJSUo9Jkybt/OEPf7i37HyDe+65Z8+2bdviLrroorQjR47EmplPmzYtaf369WtatGhR\nMnDgwAufe+65rcnJyYUPPfRQ6yeeeKLN559/3iAzMzNt0KBB+bNnz94qScuWLWv2yCOPfPLV3pWz\nm7lXeyL4V+/Y7DpJQ9x9bLB8k6Q+7v6vwXKMpP+VdLO7bzGzpZJ+dKpzDrKzsz03N9zTEpInvRpq\n/7Xp2cyNysm7P9pl1IwHTns6JELGWKmjGCt1DmOljqqFsWJmK9w9u/y6VatWbcnMzDxpfn59tHXr\n1gajR49Ofuutt6q9fGlN2r59e9z111//jbfffntDbe3zTLNq1apWmZmZyZVtC3Na0Q5JHcstd5BU\n/mYUzSRlSFpqZlskXSxpvpl9aQABAADgzNSpU6fCW2+9de++fftq7QqZmzdvbjh58uTtp26JyoQ5\nrWi5pM5mliLpE0mjJH2nbKO750tqVbYc6ZEDAAAAnDnGjh27vzb3N3DgwKO1ub/6JrQU5+5FksZL\nWihpvaQ57r7WzB40s+Fh7RcAAKCOKykpKbFoF4GzU/DZq/JKTqHe58DdF0haUGHdfVW0zQmzFgAA\ngDpizZ49e9ISExPzY2Jiwjn5E6hESUmJ7dmzp7mkNVW1CTUcAAAA4MuKiorG7tq16/e7du3KULjn\nfwIVlUhaU1RUNLaqBoQDAACAWpSVlbVbElOsUSeRVgEAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAI\nEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQ\nDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAO\nAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4A\nAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAA\nAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASAo5HJjZUDPLM7NNZjapku0TzGydmX1gZn8z\ns05h1gMAAACgaqGFAzOLlfSkpGGS0iSNNrO0Cs3el5Tt7j0kvSjp0bDqAQAAAFC9MI8c9JG0yd03\nu/sXkmZJGlG+gbsvcfejweI7kjqEWA8AAACAapi7h9Ox2bWShrr72GD5Jkl93X18Fe3/S9Iud3+o\nkm3jJI2TpKSkpKxZs2aFUnOZ1Z/kh9p/bUpJKFDTgp3RLqNmtO0Z7QpQAWOljmKs1DmMlTqqFsbK\noEGDVrh7dug7AmpIXIh9WyXrKk0iZnajpGxJAyvb7u7TJU2XpOzsbM/JyamhEit386RXQ+2/Nj2b\n+bFy8u6Pdhk1Y3T9+eNaXzBW6ijGSp3DWKmjGCvAScIMBzskdSy33EHSST81mNlgSf8haaC7F4RY\nDwAAAIBqhHnOwXJJnc0sxcwaSholaX75BmbWS9I0ScPdfXeItQAAAAA4hdDCgbsXSRovaaGk9ZLm\nuPtaM3vQzIYHzf5TUlNJc81spZnNr6I7AAAAACELc1qR3H2BpAUV1t1X7vHgMPcPAAAAIHLcIRkA\nAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAA\nAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAA\ngCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACA\nJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAk\nwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTC\nAQAAAIBAqOHAzIaaWZ6ZbTKzSZVsjzez2cH2f5hZcpj1AAAAAKhaaOHAzGIlPSlpmKQ0SaPNLK1C\nszGS9rv7hZIel/RIWPUAAAAAqF6YRw76SNrk7pvd/QtJsySNqNBmhKTngscvSrrczCzEmgAAAABU\nIcxw0F7S9nLLO4J1lbZx9yJJ+ZJahlgTAAAAgCrEhdh3ZUcA/Cu0kZmNkzQuWDxsZnlfs7azxiCp\nlaS90a6jRvyMg0oID2MFiAxj5bR1qo2dADUlzHCwQ1LHcssdJO2sos0OM4uT1FzSvooduft0SdND\nqrNeM7Ncd8+Odh1AXcdYASLDWAHqtzCnFS2X1NnMUsysoaRRkuZXaDNf0j8Hj6+V9L/uftKRAwAA\nAADhC+3IgbsXmdl4SQslxUoZlVSLAAAE/0lEQVR6xt3XmtmDknLdfb6k/5b0BzPbpNIjBqPCqgcA\nAABA9Ywf6us3MxsXTMsCUA3GChAZxgpQvxEOAAAAAEgK+Q7JAAAAAM4chAMAAAAAkggHZz0zSzaz\nNcHjHDP7S7RrAqLBzN46xfYFZnZubdUD1AdmttTMsoPHW8ysVbRrAlC9MO9zgBCZman0nJGSaNcC\n1DVmFuvuxafzHHfvd4rt//T1qgLqH/4WAfUPRw7OIMGv/OvNbKqk9yTdZGZvm9l7ZjbXzJoG7S4y\ns7fMbJWZvWtmzYLnvh60fc/Mqv0iBNRVwWf5QzN7zsw+MLMXzaxx8KvkfWb2hqTrzOwCM3vNzFYE\nn/2uwfOTzOyVYHysKhsLZnY4+HdbM1tmZivNbI2ZXRqsP/Grp5lNCLatMbO7y9W13syeNrO1ZrbI\nzBKi8iYB1aj4+TWzR8zsjnLbHzCzicHjH5vZ8mCs/SxYV/FvUUcz+52Z5Qaf/Z9F55UBqAmEgzNP\nF0nPS7pC0hhJg929t6RcSROCG87NlnSXu2dKGizpmKTdkq4I2t4gaUo0igdqSBdJ0929h6SDksq+\n2Bx39wHuPkuld1X/V3fPkvQjSVODNlMk/T0YH70lra3Q93ckLXT3npIyJa0sv9HMsiTdIqmvpIsl\nfd/MegWbO0t60t3TJR2QNLKmXjBQEyr7/EqapdK/C2WulzTXzK5U6We6j6SekrLM7JtBmy6Snnf3\nXu6+VdJ/BHdN7iFpoJn1qJUXBKDGMa3ozLPV3d8xs6skpUl6s/SorhpKelul/8P+1N2XS5K7H5Qk\nM2si6b/MrKekYkmp0SgeqCHb3f3N4PELku4MHs+WpOAoWj+VfsEpe0588O/LJH1PkoKpR/kV+l4u\n6RkzayBpnruvrLB9gKRX3P1IsK+XJV2q0ju+f1yu/QpJyV/jNQJhqOrz29rM2klKlLTf3beZ2Z2S\nrpT0fvDcpioNC9sU/C0q1+/1ZjZOpd8r2qr079MHtfGCANQswsGZ50jwb5O02N1Hl98Y/FpT2c0r\nfijpM5X+Ehoj6XiYRQIhq/gZL1suGx8xkg4Ev/6fXsfuy4JfR7+l0ju4/6e7P1+uiVXxVEkqKPe4\nWBLTilDXVPX5fVHStZLaqPRIQlnbh9192pc6MEvW/481mVmKSo/OXeTu+83sWUmNarRqALWGaUVn\nrnck9TezCyUpmHOdKulDSe3M7KJgfTMzi5PUXKVHFEok3SQpNkp1AzXhfDO7JHg8WtIb5TcGR8w+\nNrPrpNKTJs0sM9j8N0m3B+tjzeyc8s81s06Sdrv705L+W6VTj8pbJumaYMw1kfRtSa/X3EsDQlXV\n53eWpFEqDQgvBm0XSrq13Pls7c2sdSV9nqPSsJBvZkmShoX8GgCEiHBwhnL3PZJuljTTzD5QaVjo\n6u5fqHTu6BNmtkrSYpX+gjNV0j+b2TsqnVJ0pNKOgTPDepV+nj+Q1ELS7ypp811JY4JxsFbSiGD9\nXZIGmdlqlU79Sa/wvBxJK83sfZWeM/Db8hvd/T1Jz0p6V9I/JP3e3d8XcAao6vPr7mslNZP0ibt/\nGrRdJOmPkt4OxsuLQZuKfa5S6dSjtZKekfRmxTYAzhzmXtkMFACom4IpDX9x94wolwIAQL3DkQMA\nAAAAkjhyAAAAACDAkQMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJKk/wMpRaccEeod\nsAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f103b35c978>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:1:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>99471</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0        99471"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.681724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.810744</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.681724\n",
       "1  precision  1.000000\n",
       "2   f1_score  0.810744"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP016582.1</th>\n",
       "      <th>CP016583.1</th>\n",
       "      <th>CP016584.1</th>\n",
       "      <th>CP016585.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.99998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.99998</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP016582.1  CP016583.1  CP016584.1  CP016585.1\n",
       "0  plasmid_0         0.0         0.0         0.0     0.99998\n",
       "1      Total         0.0         0.0         0.0     0.99998"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016582.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016583.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016584.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP016585.1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Total</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0\n",
       "0  CP016582.1        0.0\n",
       "1  CP016583.1        0.0\n",
       "2  CP016584.1        0.0\n",
       "3  CP016585.1        1.0\n",
       "4       Total        1.0"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0         1.0"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016582.1</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016583.1</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016584.1</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP016585.1</td>\n",
       "      <td>0.99998</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP016582.1     0.00000\n",
       "1  CP016583.1     0.00000\n",
       "2  CP016584.1     0.00000\n",
       "3  CP016585.1     0.99998"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016585.1</td>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.99998</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction   recall  precision\n",
       "0  CP016585.1  plasmid_0  0.99998        1.0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 1, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00270844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[13, 28, 31]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.446924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.449633</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           plasmid_0\n",
       "GC_sum   -0.00270844\n",
       "chain   [13, 28, 31]\n",
       "gd_sum     -0.446924\n",
       "total      -0.449633"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0544227</td>\n",
       "      <td>-0.0174096</td>\n",
       "      <td>0.016442</td>\n",
       "      <td>-0.00286644</td>\n",
       "      <td>-0.00132651</td>\n",
       "      <td>-0.00224544</td>\n",
       "      <td>0.0028754</td>\n",
       "      <td>-0.00181055</td>\n",
       "      <td>-0.0297622</td>\n",
       "      <td>-0.0280459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[17]</td>\n",
       "      <td>[21]</td>\n",
       "      <td>[24]</td>\n",
       "      <td>[31, 81, 13, 31, 13, 28, 31, 80, 28]</td>\n",
       "      <td>[75, 6, 32, 69, 20]</td>\n",
       "      <td>[69, 5, 32, 69, 9]</td>\n",
       "      <td>[43, 11, 32, 69]</td>\n",
       "      <td>[70, 8, 32, 1]</td>\n",
       "      <td>[32]</td>\n",
       "      <td>[28]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.410882</td>\n",
       "      <td>-0.425154</td>\n",
       "      <td>-0.23639</td>\n",
       "      <td>-0.446077</td>\n",
       "      <td>-0.000767602</td>\n",
       "      <td>-0.000457901</td>\n",
       "      <td>-0.0014239</td>\n",
       "      <td>-0.000183013</td>\n",
       "      <td>-0.343563</td>\n",
       "      <td>-0.240341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.465304</td>\n",
       "      <td>-0.442564</td>\n",
       "      <td>-0.219948</td>\n",
       "      <td>-0.448944</td>\n",
       "      <td>-0.00209412</td>\n",
       "      <td>-0.00270334</td>\n",
       "      <td>0.0014515</td>\n",
       "      <td>-0.00199356</td>\n",
       "      <td>-0.373325</td>\n",
       "      <td>-0.268387</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        plasmid_0  plasmid_1 plasmid_2                             plasmid_3  \\\n",
       "GC_sum -0.0544227 -0.0174096  0.016442                           -0.00286644   \n",
       "chain        [17]       [21]      [24]  [31, 81, 13, 31, 13, 28, 31, 80, 28]   \n",
       "gd_sum  -0.410882  -0.425154  -0.23639                             -0.446077   \n",
       "total   -0.465304  -0.442564 -0.219948                             -0.448944   \n",
       "\n",
       "                  plasmid_4           plasmid_5         plasmid_6  \\\n",
       "GC_sum          -0.00132651         -0.00224544         0.0028754   \n",
       "chain   [75, 6, 32, 69, 20]  [69, 5, 32, 69, 9]  [43, 11, 32, 69]   \n",
       "gd_sum         -0.000767602        -0.000457901        -0.0014239   \n",
       "total           -0.00209412         -0.00270334         0.0014515   \n",
       "\n",
       "             plasmid_7  plasmid_8  plasmid_9  \n",
       "GC_sum     -0.00181055 -0.0297622 -0.0280459  \n",
       "chain   [70, 8, 32, 1]       [32]       [28]  \n",
       "gd_sum    -0.000183013  -0.343563  -0.240341  \n",
       "total      -0.00199356  -0.373325  -0.268387  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_0  from MILP completely contained in  plasmid_3  from greedy.\n",
      "\n",
      "plasmid_9  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
