{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 66 with $\\alpha_1 = 1$, $\\alpha_2 = 1$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 66\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016924.1</td>\n",
       "      <td>111539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016925.1</td>\n",
       "      <td>63291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016927.1</td>\n",
       "      <td>72093</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP016924.1       111539\n",
       "1  CP016925.1        63291\n",
       "2  CP016927.1        72093"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt0VeWZx/Hfk4Q7iAIhXCWpEiAJ\nBEgEBSxBUaBVsAsv0GpHheJoGbXQKjPTqrVaqyO14kgFrUutFAEvlFZGoB0oXitBQW4GELmJCAgE\nwjWXZ/7IDhNDEg6anZOE72ctlmfv/Z53P+es85rzO/vde5u7CwAAAABiol0AAAAAgJqBcAAAAABA\nEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQCAu2gWcrlat\nWnliYmK0y6g1Dh06pCZNmkS7DKDGY6wAkWGsnJ7ly5fvcff4aNcBRKrWhYPExERlZ2dHu4xaY8mS\nJcrKyop2GUCNx1gBIsNYOT1mtiXaNQCng2lFAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAA\nAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAEQgsHZvasme0ys9UVbDczm2JmG83sIzPrHVYtAAAAAE4t\nzCMHz0kaWsn2YZI6B//GSfp9iLUAAAAAOIXQwoG7L5W0t5ImIyS94MXek3S2mbUNqx4AAAAAlYuL\n4r7bS9pWanl7sO7zsg3NbJyKjy4oISFBS5YsCbWwVZ/lhtp/dUpqHhv6+4UzF2MFiAxjBUBtEc1w\nYOWs8/Iauvt0SdMlKTMz07OyskIsS7px0uuh9l+dnhvaRGG/XzhzMVaAyDBWANQW0bxa0XZJHUst\nd5C0I0q1AAAAAGe8aIaDeZJ+GFy16EJJue5+0pQiAAAAANUjtGlFZjZTUpakVma2XdK9kupJkrs/\nJWm+pO9I2ijpsKSbwqoFAAAAwKmFFg7cffQptrukH4e1fwAAAACnhzskAwAAAJBEOAAAAAAQIBwA\nAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAA\nAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAA\nAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAA\nkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQ\nRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAEAg1HJjZUDPL\nMbONZjapnO3nmtliM/vQzD4ys++EWQ8AAACAioUWDswsVtKTkoZJSpE02sxSyjT7uaTZ7t5L0ihJ\nU8OqBwAAAEDlwjxy0EfSRnff5O7HJb0kaUSZNi7prOBxc0k7QqwHAAAAQCXiQuy7vaRtpZa3S+pb\nps19khaa2b9JaiJpcIj1AAAAAKhEmOHAylnnZZZHS3rO3Seb2UWS/mhmae5e9JWOzMZJGidJCQkJ\nWrJkSRj1njCxe0Go/VenvLy80N8vnLkYK0BkGCsAaosww8F2SR1LLXfQydOGxkgaKknu/q6ZNZTU\nStKu0o3cfbqk6ZKUmZnpWVlZIZVc7MZJr4faf3V6bmgThf1+4czFWAEiw1gBUFuEec7BMkmdzSzJ\nzOqr+ITjeWXabJV0qSSZWTdJDSXtDrEmAAAAABUILRy4e4Gk8ZIWSFqn4qsSrTGz+81seNBsoqQf\nmdlKSTMl3ejuZaceAQAAAKgGYU4rkrvPlzS/zLp7Sj1eK6l/mDUAAAAAiAx3SAYAAAAgiXAAAAAA\nIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAg\nQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBA\nOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4\nAAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgA\nAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAA\nAAAgKeRwYGZDzSzHzDaa2aQK2lxrZmvNbI2Z/SnMegAAAABULC6sjs0sVtKTki6TtF3SMjOb5+5r\nS7XpLOnfJfV3931m1jqsegAAAABULswjB30kbXT3Te5+XNJLkkaUafMjSU+6+z5JcvddIdYDAAAA\noBKhHTmQ1F7StlLL2yX1LdMmWZLM7G1JsZLuc/c3ynZkZuMkjZOkhIQELVmyJIx6T5jYvSDU/qtT\nXl5e6O8XzlyMFSAyjBUAtUWY4cDKWefl7L+zpCxJHSS9aWZp7r7/K09yny5puiRlZmZ6VlZWlRdb\n2o2TXg+1/+r03NAmCvv9wpmLsQJEhrECoLYIc1rRdkkdSy13kLSjnDZ/dvd8d/9UUo6KwwIAAACA\nahZmOFgmqbOZJZlZfUmjJM0r02aupEGSZGatVDzNaFOINQEAAACoQGjhwN0LJI2XtEDSOkmz3X2N\nmd1vZsODZgskfWlmayUtlvQzd/8yrJoAAAAAVCzMcw7k7vMlzS+z7p5Sj13ShOAfAAAAgCjiDskA\nAAAAJBEOAAAAAAQIBwAAAAAknUY4MLMBZnZT8DjezJLCKwsAAABAdYsoHJjZvZLulvTvwap6kl4M\nqygAAAAA1S/SIwffkzRc0iFJcvcdkpqFVRQAAACA6hdpODgeXHbUJcnMmoRXEgAAAIBoiDQczDaz\naZLONrMfSfqbpKfDKwsAAABAdYvoJmju/qiZXSbpgKQuku5x90WhVgYAAACgWp0yHJhZrKQF7j5Y\nEoEAAAAAqKNOOa3I3QslHTaz5tVQDwAAAIAoiWhakaSjklaZ2SIFVyySJHe/PZSqAAAAAFS7SMPB\n68E/AAAAAHVUpCckP29m9SUlB6ty3D0/vLIAAAAAVLeIwoGZZUl6XtJmSSapo5n9i7svDa80AAAA\nANUp0mlFkyVd7u45kmRmyZJmSsoIqzAAAAAA1SvSm6DVKwkGkuTu6yXVC6ckAAAAANEQ6ZGDbDP7\ng6Q/Bss/kLQ8nJIAAAAAREOk4eBWST+WdLuKzzlYKmlqWEUBAAAAqH6RhoM4SY+7+2+lE3dNbhBa\nVQAAAHXU8uXLW8fFxT0jKU2RT/EGqkKRpNUFBQVjMzIydpXXINJw8HdJgyXlBcuNJC2U1O8blwgA\nNcHnK6T7RkS7iqpxX260KwBQibi4uGfatGnTLT4+fl9MTIxHux6cOYqKimz37t0pO3fufEbS8PLa\nRJpWG7p7STBQ8LhxFdQIAABwpkmLj48/QDBAdYuJifH4+PhcFR+1Kr9NhH0dMrPeJQtmlinpyDes\nDwAA4EwUQzBAtASfvQozQKTTiu6QNMfMdkhySe0kXffNywMAAABQU0QaDpIk9ZJ0rqTvSbpQxSEB\nAAAA30DipNer9Kaym3/z3Rpzufk+ffp0efTRR7d9+9vfPhxJ+6KiIvXr1y95/vz5G1u0aFEUdn3f\n1F//+tdmkydPTli8ePHGMPqfMWNG8zVr1jT69a9/vbPstsaNG/c6fPjwhxU9t1+/fsl//vOfP4mP\njy88nX1GOq3oF+5+QNLZki6TNF3S709nRwAAAKg98vPzq32fs2fPbp6amnqkNgSD6vCDH/wgt7xg\nEInRo0d/+eijj8af7vMiDQclieO7kp5y9z9Lqn+6OwMAAED0/exnP2ublJSU2q9fv85XXnll0j33\n3JMgFf/SP378+PYXXHBBlwceeCBhx44dcUOGDDkvLS2tW1paWreFCxc2kaQDBw7EXHPNNYlpaWnd\nunXrlvLiiy+eLUl5eXl2xRVXfCs5OTnlu9/97reOHj1qkvTYY4+1GjNmTMeS/U+ePLnV2LFjO5St\na8aMGS2+973v7ZeknJyc+klJSanXXXddp86dO6cOHz48ae7cuc169+7dtVOnTmmLFy9uXFktOTk5\n9TMyMrqkpKR0S0lJ6bZo0aImUvGv/X369OkydOjQbyUlJaUOHz48qaio8iwycuTIxO9///vnZmRk\ndElMTEybOXNm87JtFi9e3LhXr15du3XrltKrV6+uK1eubCBJ2dnZDbt3796ta9euKcnJySmrVq1q\nEOlrmzJlSssf/vCH50rSxx9/XL9nz55d09LSut1xxx3tSva7ZcuWepmZmV26du2a0rlz59Q33nij\nqSSNGjVq/6uvvtoywo/ECZFOK/rMzKap+HKmD5tZA3FdXgAAgFpn6dKljf/yl7+cs2rVqrX5+fnW\ns2fPlF69ep2Y9rN///7YZcuW5UjSlVdemTRhwoQvhgwZkrdhw4b6Q4YM6bxp06Y1//Ef/9F20KBB\nB+bMmbN5z549sZmZmd2GDx9+4Le//W18o0aNitavX7/2n//8Z6P+/funSNKYMWP2pqamphw7dmx7\ngwYN/MUXX2w1bdq0LWVrW758edP+/fufWL9t27aGs2bN2pSRkbGlR48e3WbMmNEyOzv74z/96U9n\nP/jgg20HDRr0SUW1tGvXruDNN99c37hxY1+1alWD0aNHf2v16tXrJGndunWNVqxYsSkxMTE/IyOj\n66JFi5oOGTIkr2w9pW3btq3B+++/n7N27doGgwcP7jJixIhVpbenp6cfff/99z+uV6+e5s6d2+yu\nu+7qsGDBgk+eeOKJ+Ntuu+2LW2+9de/Ro0etoKBAn332Wb1IXlvp/m+77bZzx44du3v8+PFfPvTQ\nQyeOCDz77LMtLr300tyHH354Z0FBgQ4ePBgjSfHx8YXHjx+3nTt3xrZp0ybiqUWRhoNrJQ2V9Ki7\n7zeztpJ+FulOAAAAUDMsWbKk6bBhw/Y3bdrUJflll122v/T20aNH7y15/Pbbb5+1YcOGRiXLeXl5\nsfv27YtZsmTJWQsWLDh7ypQpbSTp2LFjtnHjxvpvvfVW09tvv32XJPXt2/dIcnLyYUk666yzivr3\n739w1qxZzbt37340Pz/f+vTpc9KVL3Nzc+POOeecEz/jt2/f/lhJu+Tk5COXXHLJgZiYGPXu3fvw\nAw880C54PeXW0qlTp/wxY8Z0Wrt2baOYmBht2bLlxA18u3fvfui8887Ll6TU1NTDn3zyySlnxIwc\nOXJvbGysunfvfqxjx47HVqxY0bD09r1798Zed911SZs3b25oZp6fn2+SdNFFFx169NFH227fvr3+\nqFGj9nXv3v1YpK+ttA8++KDp//zP/3wiSbfccsuXv/rVrzpI0oUXXnjolltuSczPz4+5+uqr9/Xr\n1+/E+9qyZcuCrVu31m/Tpk3EVxmNKBy4+2FJr5Za/lzS55HuBAAAADWDe+XXlGnWrNmJL+furuzs\n7HVBkPhKHy+//PLG9PT0Y2Wfb2bl9jtu3Lg9Dz74YJvk5OSj119//Z7y2sTGxnphYaFiY2MlSfXr\n1z+x35iYGDVs2NCDdiosLLTKapkwYUK71q1b57/yyiufFhUVqVGjRidO/G7QoMGJfmNjY1VQUFB+\n0ZW8rrLLd999d/uBAwceXLRo0Sc5OTn1L7nkki6S9K//+q97L7744kOvvfZa82HDhiVPnTp1c5cu\nXY5F8trKKu8SuMOGDctbunRpziuvvNL8xhtvTLr99tu/GD9+/JdScVBq3LjxaZ2/wdQgAACAM0hW\nVlbeggULmh8+fNhyc3Nj/va3v51dUdsBAwYcePjhh1uXLL/zzjuNJGnQoEEHJk+enFAyV//tt99u\nFLTPe/HFF1tI0rJlyxquX7/+xE1zL7nkkkOff/55/ddee63lmDFj9qocSUlJR9etW9egvG0VqaiW\n3Nzc2LZt2+bHxsZq6tSpLQsLTz2z5sc//nH7F154odz349VXXz2nsLBQa9asabBt27YG6enpR0tv\nP3DgQGyHDh2OS9K0adNalaxfu3Zt/W7duh37+c9/vuvyyy/fv2LFikZl+45E7969855++ukWkvT0\n00+fOJdg/fr19du3b58/ceLEPddff/2eDz74oLFUfOWn3bt31+vSpctJAa4ykU4rAgAAQAiq+9Kj\nAwcOPDx06NDclJSU1Pbt2x/r0aPHoebNm5f7zXn69Onbxo4de25ycnJKYWGh9e3b92C/fv22/uY3\nv9kxbty4c7t27Zri7tahQ4djixcv3vjTn/5016hRo5KSk5NTUlNTD3fv3v1Q6f6uuuqqfR999FHj\nii6vefnll+cuXLiwWVpaWsRfaCuq5c4779w1cuTI8+bOnXvOgAEDDjZq1OiUv6CvXbu2UckJ0WWd\nf/75x/r06dPlyy+/rPe73/1uS+PGjb/yK/7dd9+9c+zYsUlTpkxpc/HFFx8oWf/HP/6xxZw5c1rG\nxcV5fHx8/kMPPbRj//79sZG+vhJTp07dOmrUqG9NnTo1Yfjw4ftK1i9YsKDZlClT2sTFxXnjxo0L\nZ8yY8akkvfXWW4179ep1qF69eqe1HzvVoaWaJjMz07Ozs0PdR+Kk10Ptvzo9N7SJsrKyol0G6qg6\nNVbSNygr595ol1E17suNdgUoo06NFf6unBYzW+7umaXXrVy5cnN6enq502qqS25ubkzz5s2LDh48\nGHPRRRd1eeqpp7YMGDAgonsRfBODBg06/8477/xixIgRB8vbvmXLlnqjR49OfOeddzaEXUt5BgwY\n0Pmtt946ad8jR45MvOKKK3JvuummfeU9rya66aabOl511VX7y3uvV65c2So9PT2xvOcxrQgAAOAM\nc/3113fq2rVrSo8ePbpdeeWV+8IOBnv27IlNTExMa9iwYVFFwUCSOnXqlH/zzTfv2bt3b1S+o5YX\nDGqrtLS0I5W91xVhWhEAAMAZ5i9/+cun1bm/Vq1aFW7evHl1JG3Hjh1b436df+WVVzZHu4bTNXHi\nxK91dIojBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAEOCEZAAAgmu5rnnHqRqfTX+4p75tgZhkj\nRozYO3fu3E8lKT8/X61bt07v2bPnocWLF2+cMmVKy+zs7CYvvPDC1gkTJrRr2rRp4f333/9F6T5i\nY2MzOnfufKSwsNDOP//8I7Nnz95c+u7KUvGNuPr165c8f/78jS1atCi65pprEv/+9783b9myZcGG\nDRvWlFfbhx9+2PCmm25KXLt2beNJkyZ9Vna/JX7961/HP/XUUwnbtm1rsGPHjpVt27YtkKSZM2c2\nf//995s89thjOyJ9y/D/OHIAAABwhmnUqFFRTk5Oo7y8PJOk11577ayEhIT80+mjQYMGRR9//PHa\nDRs2rKlXr55Pnjw5vmyb2bNnN09NTT3SokWLIkm6+eab98ybN6/Sy4W2bt264PHHH996yy23lBsK\nSgwcODBv0aJF69u1a3e89Prrrrsu94033jj74MGDfM/9GnjTAAAAzkCXXnpp7pw5c86WpJkzZ7YY\nOXLk3q/b14ABA/I2btzYoOz6GTNmtCh9x+Fhw4blxcfHF1TWV/v27QsGDhx4uF69epXeqbd///5H\nunTpcrzs+piYGPXr1+/grFmzmp/Oa0AxwgEAAMAZ6IYbbtg7a9ascw4fPmzr1q1rfNFFFx36Ov3k\n5+drwYIFZ3Xv3v1I2W3Lly9v2r9//1P2+8gjj8Q/8sgjJx15KGvgwIHnb968ud6p2mVmZh568803\nm56qHU4W6jkHZjZU0uOSYiU94+6/qaDd1ZLmSLrA3bPDrAkAAABS3759j2zfvr3B008/3WLw4MG5\np/v8Y8eOxXTt2jUl6OvgHXfccdJNt3Jzc+POOeecopOf/VV33XXX7kj2+Y9//GNjJO3atGlTsHPn\nzvqRtMVXhRYOzCxW0pOSLpO0XdIyM5vn7mvLtGsm6XZJ/wyrFgAAAJxs6NCh+++9996OCxcuzNm1\na9dpfS8sOeegsjaxsbFeWFio2NjYb1boaTpy5Ig1bNjwlKEEJwtzWlEfSRvdfZO7H5f0kqQR5bT7\nlaRHJB0NsRYAAACUceutt+6ZOHHijj59+pw0JagqJCUlHV23bt1J5yKELScnp2Fqamoor6muC3Na\nUXtJ20otb5fUt3QDM+slqaO7/9XMflpRR2Y2TtI4SUpISNCSJUuqvtpSJnav9DyZWiUvLy/09wtn\nrjo1Vhq005Iuv4x2GVWDMV/j1Kmxwt+VqhfBpUfDct555+X/4he/2HWqdo899ljbadOmJZQsf/HF\nFx9F0v/ll1+eu3DhwmZpaWnHJOnKK69Meu+995rt27cvLiEhocekSZN2/OQnP9lTcr7BXXfdtXvr\n1q1xF1xwQcqhQ4dizcynTZuWsG7dutUtWrQoGjhw4PnPP//8lsTExPwHHnig9RNPPNHmyy+/rJee\nnp4yaNCg3FmzZm2RpKVLlzZ7+OGHP/t678qZzdwrPRH863dsdo2kIe4+Nli+QVIfd/+3YDlG0v9K\nutHdN5vZEkk/PdU5B5mZmZ6dHe5pCYmTXg+1/+r03NAmysrKinYZqKPq1FhJ36CsnHujXUbVuO+0\npw4jZHVqrPB35bSY2XJ3zyy9buXKlZvT09NPmp9fF23ZsqXe6NGjE995551KL19albZt2xZ37bXX\nfuvdd99dX137rG1WrlzZKj09PbG8bWFOK9ouqWOp5Q6SSt+MopmkNElLzGyzpAslzTOzrwwgAAAA\n1E6dOnXKv/nmm/fs3bu32q6QuWnTpvqTJ0/eduqWKE+Y04qWSepsZkmSPpM0StL3Sza6e66kViXL\nkR45AAAAQO0xduzYfdW5v4EDBx6uzv3VNaGlOHcvkDRe0gJJ6yTNdvc1Zna/mQ0Pa78AAAA1XFFR\nUZFFuwicmYLPXoVXcgr1PgfuPl/S/DLr7qmgbVaYtQAAANQQq3fv3p0SHx+fGxMTE87Jn0A5ioqK\nbPfu3c0lra6oTajhAAAAAF9VUFAwdufOnc/s3LkzTeGe/wmUVSRpdUFBwdiKGhAOAAAAqlFGRsYu\nSUyxRo1EWgUAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAA\nAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAA\nAAAgQDgAAAAAIIlwAAAAACAQF+0CAABALfL5Cum+EdGuomrclxvtCoAahyMHAAAAACQRDgAAAAAE\nCAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQI\nBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgH\nAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAATiwuzczIZKelxSrKRn3P03ZbZPkDRWUoGk3ZJudvct\nYdZ0xvl8hXTfiGhXUTXuy412BQAAAHVaaEcOzCxW0pOShklKkTTazFLKNPtQUqa795D0sqRHwqoH\nAAAAQOXCnFbUR9JGd9/k7sclvSTpKz9hu/tidz8cLL4nqUOI9QAAAACohLl7OB2bXS1pqLuPDZZv\nkNTX3cdX0P6/Je109wfK2TZO0jhJSkhIyHjppZdCqbnEqs/qzvSVpEbH1PTYjmiXUTXa9ox2BSiD\nsVJDMVZqHMZKDVUNY2XQoEHL3T0z9B0BVSTMcw6snHXlJhEzu15SpqSB5W139+mSpktSZmamZ2Vl\nVVGJ5btx0uuh9l+dnkv/VFk590a7jKoxuu78ca0rGCs1FGOlxmGs1FCMFeAkYYaD7ZI6llruIOmk\nnxrMbLCk/5Q00N2PhVgPAAAAgEqEec7BMkmdzSzJzOpLGiVpXukGZtZL0jRJw919V4i1AAAAADiF\n0MKBuxdIGi9pgaR1kma7+xozu9/MhgfN/ktSU0lzzGyFmc2roDsAAAAAIQv1PgfuPl/S/DLr7in1\neHCY+wcAAAAQOe6QDAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAA\nAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAA\nQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABA\nEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS\n4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLh\nAAAAAECAcAAAAABAEuEAAAAAQCDUcGBmQ80sx8w2mtmkcrY3MLNZwfZ/mllimPUAAAAAqFho4cDM\nYiU9KWmYpBRJo80spUyzMZL2ufv5kh6T9HBY9QAAAACoXJhHDvpI2ujum9z9uKSXJI0o02aEpOeD\nxy9LutTMLMSaAAAAAFQgzHDQXtK2Usvbg3XltnH3Akm5klqGWBMAAACACsSF2Hd5RwD8a7SRmY2T\nNC5YzDOznG9Y2xljkNRK0p5o11ElfslBJYSHsQJEhrFy2jpVx06AqhJmONguqWOp5Q6SdlTQZruZ\nxUlqLmlv2Y7cfbqk6SHVWaeZWba7Z0a7DqCmY6wAkWGsAHVbmNOKlknqbGZJZlZf0ihJ88q0mSfp\nX4LHV0v6X3c/6cgBAAAAgPCFduTA3QvMbLykBZJiJT3r7mvM7H5J2e4+T9IfJP3RzDaq+IjBqLDq\nAQAAAFA544f6us3MxgXTsgBUgrECRIaxAtRthAMAAAAAkkK+QzIAAACA2oNwAAAAAEAS4eCMZ2aJ\nZrY6eJxlZn+Ndk1ANJjZO6d0XbasAAAEmUlEQVTYPt/Mzq6ueoC6wMyWmFlm8HizmbWKdk0AKhfm\nfQ4QIjMzFZ8zUhTtWoCaxsxi3b3wdJ7j7v1Osf0736wqoO7hbxFQ93DkoBYJfuVfZ2ZTJX0g6QYz\ne9fMPjCzOWbWNGh3gZm9Y2Yrzex9M2sWPPfNoO0HZlbpFyGgpgo+yx+b2fNm9pGZvWxmjYNfJe8x\ns7ckXWNm55nZG2a2PPjsdw2en2BmrwXjY2XJWDCzvOC/bc1sqZmtMLPVZnZxsP7Er55mNiHYttrM\n7ixV1zoze9rM1pjZQjNrFJU3CahE2c+vmT1sZreV2n6fmU0MHv/MzJYFY+2Xwbqyf4s6mtnvzSw7\n+Oz/MjqvDEBVIBzUPl0kvSDpMkljJA12996SsiVNCG44N0vSHe6eLmmwpCOSdkm6LGh7naQp0Sge\nqCJdJE139x6SDkgq+WJz1N0HuPtLKr6r+r+5e4akn0qaGrSZIukfwfjoLWlNmb6/L2mBu/eUlC5p\nRemNZpYh6SZJfSVdKOlHZtYr2NxZ0pPunippv6SRVfWCgapQ3udX0ksq/rtQ4lpJc8zschV/pvtI\n6ikpw8y+HbTpIukFd+/l7lsk/Wdw1+QekgaaWY9qeUEAqhzTimqfLe7+npldISlF0tvFR3VVX9K7\nKv4f9ufuvkyS3P2AJJlZE0n/bWY9JRVKSo5G8UAV2ebubwePX5R0e/B4liQFR9H6qfgLTslzGgT/\nvUTSDyUpmHqUW6bvZZKeNbN6kua6+4oy2wdIes3dDwX7elXSxSq+4/unpdovl5T4DV4jEIaKPr+t\nzaydpHhJ+9x9q5ndLulySR8Gz22q4rCwVcHfolL9Xmtm41T8vaKtiv8+fVQdLwhA1SIc1D6Hgv+a\npEXuPrr0xuDXmvJuXvETSV+o+JfQGElHwywSCFnZz3jJcsn4iJG0P/j1//Q6dl8a/Dr6XRXfwf2/\n3P2FUk2sgqdK0rFSjwslMa0INU1Fn9+XJV0tqY2KjySUtH3I3ad9pQOzRP3/WJOZJan46NwF7r7P\nzJ6T1LBKqwZQbZhWVHu9J6m/mZ0vScGc62RJH0tqZ2YXBOubmVmcpOYqPqJQJOkGSbFRqhuoCuea\n2UXB49GS3iq9MThi9qmZXSMVnzRpZunB5r9LujVYH2tmZ5V+rpl1krTL3Z+W9AcVTz0qbamkq4Ix\n10TS9yS9WXUvDQhVRZ/flySNUnFAeDlou0DSzaXOZ2tvZq3L6fMsFYeFXDNLkDQs5NcAIESEg1rK\n3XdLulHSTDP7SMVhoau7H1fx3NEnzGylpEUq/gVnqqR/MbP3VDyl6FC5HQO1wzoVf54/ktRC0u/L\nafMDSWOCcbBG0ohg/R2SBpnZKhVP/Ukt87wsSSvM7EMVnzPweOmN7v6BpOckvS/pn5KecfcPBdQC\nFX1+3X2NpGaSPnP3z4O2CyX9SdK7wXh5OWhTts+VKp56tEbSs5LeLtsGQO1h7uXNQAGAmimY0vBX\nd0+LcikAANQ5HDkAAAAAIIkjBwAAAAACHDkAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAA\nACBJ+j+WoIWuny8pYwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f75773c9828>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:1:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>66705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>11428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>1978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>2535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>5860</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0        66705\n",
       "1  plasmid_1        11428\n",
       "2  plasmid_2         1978\n",
       "3  plasmid_3         2535\n",
       "4  plasmid_4         5860"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.213075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.502734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.299297</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.213075\n",
       "1  precision  0.502734\n",
       "2   f1_score  0.299297"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP016924.1</th>\n",
       "      <th>CP016925.1</th>\n",
       "      <th>CP016927.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.091403</td>\n",
       "      <td>0.357025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.232055</td>\n",
       "      <td>0.015411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.027081</td>\n",
       "      <td>0.037438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.171494</td>\n",
       "      <td>0.039824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.522033</td>\n",
       "      <td>0.483015</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP016924.1  CP016925.1  CP016927.1\n",
       "0  plasmid_0         0.0    0.091403    0.357025\n",
       "1  plasmid_1         0.0    0.232055    0.015411\n",
       "2  plasmid_2         0.0    0.000000    0.033318\n",
       "3  plasmid_3         0.0    0.027081    0.037438\n",
       "4  plasmid_4         0.0    0.171494    0.039824\n",
       "5      Total         0.0    0.522033    0.483015"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016924.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016925.1</td>\n",
       "      <td>0.039427</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.676134</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016927.1</td>\n",
       "      <td>0.364171</td>\n",
       "      <td>0.087679</td>\n",
       "      <td>0.607179</td>\n",
       "      <td>0.676134</td>\n",
       "      <td>0.303413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.403598</td>\n",
       "      <td>1.087679</td>\n",
       "      <td>0.607179</td>\n",
       "      <td>1.352268</td>\n",
       "      <td>1.303413</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2  plasmid_3  plasmid_4\n",
       "0  CP016924.1   0.000000   0.000000   0.000000   0.000000   0.000000\n",
       "1  CP016925.1   0.039427   1.000000   0.000000   0.676134   1.000000\n",
       "2  CP016927.1   0.364171   0.087679   0.607179   0.676134   0.303413\n",
       "3       Total   0.403598   1.087679   0.607179   1.352268   1.303413"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.364171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.607179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.676134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.364171\n",
       "1  plasmid_1    1.000000\n",
       "2  plasmid_2    0.607179\n",
       "3  plasmid_3    0.676134\n",
       "4  plasmid_4    1.000000"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016924.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016925.1</td>\n",
       "      <td>0.379264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016927.1</td>\n",
       "      <td>0.396835</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP016924.1    0.000000\n",
       "1  CP016925.1    0.379264\n",
       "2  CP016927.1    0.396835"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No high mutual correspondence found.\n"
     ]
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 1, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0303546</td>\n",
       "      <td>-0.0273721</td>\n",
       "      <td>-0.0338475</td>\n",
       "      <td>-0.0399687</td>\n",
       "      <td>-0.00856283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[36, 81, 42, 27, 39, 43, 49, 90, 53, 103, 44, ...</td>\n",
       "      <td>[33, 49, 90, 29]</td>\n",
       "      <td>[48, 42]</td>\n",
       "      <td>[50, 44, 103, 47, 106]</td>\n",
       "      <td>[44, 32, 49, 96]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.423447</td>\n",
       "      <td>-0.409783</td>\n",
       "      <td>-0.497472</td>\n",
       "      <td>-0.331755</td>\n",
       "      <td>-0.458959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.453801</td>\n",
       "      <td>-0.437155</td>\n",
       "      <td>-0.53132</td>\n",
       "      <td>-0.371724</td>\n",
       "      <td>-0.467522</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0         plasmid_1  \\\n",
       "GC_sum                                         -0.0303546        -0.0273721   \n",
       "chain   [36, 81, 42, 27, 39, 43, 49, 90, 53, 103, 44, ...  [33, 49, 90, 29]   \n",
       "gd_sum                                          -0.423447         -0.409783   \n",
       "total                                           -0.453801         -0.437155   \n",
       "\n",
       "        plasmid_2               plasmid_3         plasmid_4  \n",
       "GC_sum -0.0338475              -0.0399687       -0.00856283  \n",
       "chain    [48, 42]  [50, 44, 103, 47, 106]  [44, 32, 49, 96]  \n",
       "gd_sum  -0.497472               -0.331755         -0.458959  \n",
       "total    -0.53132               -0.371724         -0.467522  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0269478</td>\n",
       "      <td>-0.00630914</td>\n",
       "      <td>-0.0414951</td>\n",
       "      <td>-0.127742</td>\n",
       "      <td>-0.0262564</td>\n",
       "      <td>-0.0278866</td>\n",
       "      <td>-0.12026</td>\n",
       "      <td>-0.0120889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[31, 96, 49, 32, 44, 50, 22, 42, 48, 81, 36, 8...</td>\n",
       "      <td>[90, 28, 49, 90, 53, 103, 44, 50, 35]</td>\n",
       "      <td>[16]</td>\n",
       "      <td>[40, 47]</td>\n",
       "      <td>[90, 29, 49, 90, 33, 49, 96, 26, 49]</td>\n",
       "      <td>[43, 49]</td>\n",
       "      <td>[39]</td>\n",
       "      <td>[44, 103]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.431739</td>\n",
       "      <td>-0.466209</td>\n",
       "      <td>-0.433364</td>\n",
       "      <td>-0.171579</td>\n",
       "      <td>-0.416414</td>\n",
       "      <td>-0.353082</td>\n",
       "      <td>-0.411942</td>\n",
       "      <td>-0.47225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.458686</td>\n",
       "      <td>-0.472518</td>\n",
       "      <td>-0.474859</td>\n",
       "      <td>-0.299321</td>\n",
       "      <td>-0.44267</td>\n",
       "      <td>-0.380968</td>\n",
       "      <td>-0.532202</td>\n",
       "      <td>-0.484339</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0  \\\n",
       "GC_sum                                         -0.0269478   \n",
       "chain   [31, 96, 49, 32, 44, 50, 22, 42, 48, 81, 36, 8...   \n",
       "gd_sum                                          -0.431739   \n",
       "total                                           -0.458686   \n",
       "\n",
       "                                    plasmid_1  plasmid_2 plasmid_3  \\\n",
       "GC_sum                            -0.00630914 -0.0414951 -0.127742   \n",
       "chain   [90, 28, 49, 90, 53, 103, 44, 50, 35]       [16]  [40, 47]   \n",
       "gd_sum                              -0.466209  -0.433364 -0.171579   \n",
       "total                               -0.472518  -0.474859 -0.299321   \n",
       "\n",
       "                                   plasmid_4  plasmid_5 plasmid_6  plasmid_7  \n",
       "GC_sum                            -0.0262564 -0.0278866  -0.12026 -0.0120889  \n",
       "chain   [90, 29, 49, 90, 33, 49, 96, 26, 49]   [43, 49]      [39]  [44, 103]  \n",
       "gd_sum                             -0.416414  -0.353082 -0.411942   -0.47225  \n",
       "total                               -0.44267  -0.380968 -0.532202  -0.484339  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_2  from MILP completely contained in  plasmid_0  from greedy.\n",
      "\n",
      "plasmid_4  from MILP completely contained in  plasmid_0  from greedy.\n",
      "\n",
      "plasmid_1  from MILP completely contained in  plasmid_4  from greedy.\n",
      "\n",
      "plasmid_5  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_6  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
