{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 66 with $\\alpha_1 = 1$, $\\alpha_2 = 10$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 66\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016924.1</td>\n",
       "      <td>111539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016925.1</td>\n",
       "      <td>63291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016927.1</td>\n",
       "      <td>72093</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP016924.1       111539\n",
       "1  CP016925.1        63291\n",
       "2  CP016927.1        72093"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt8FOW9x/HvLwkJCSAKhHAnUUkg\nBAIkgAKWoCihFbDFCyi2KhSrpV7AKuecVq31Uj1SW3qkBVprqVQFVEorFqiH1FutBAWBQCAql4gI\nCAQDEnJ5zh+ZcNaQhE3NZHP5vF8vXu7MPPvMb/e1j9nvzjMz5pwTAAAAAISFugAAAAAADQPhAAAA\nAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAAT0SoC6it\nDh06uPj4+FCX0WgcO3ZMrVq1CnUZQIPHWAGCw1ipnfXr1x90zsWGug4gWI0uHMTHxys7OzvUZTQa\nWVlZysjICHUZQIPHWAGCw1ipHTPbFeoagNpgWhEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIB\nAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7fwoGZPWVm+81sczXbzczmmlmemb1vZoP8qgUAAADA\nmfl55OBpSZk1bB8rqZf3b7qkX/tYCwAAAIAz8C0cOOdek3SohiYTJC1y5d6WdLaZdfarHgAAAAA1\niwjhvrtK2hOwnO+t+6RyQzObrvKjC4qLi1NWVpavhW36uMDX/utTQttw398vNF+MFSA4jBUAjUUo\nw4FVsc5V1dA5t0DSAklKT093GRkZPpYl3TD7ZV/7r09PZ7aS3+8Xmi/GChAcxgqAxiKUVyvKl9Q9\nYLmbpL0hqgUAAABo9kIZDlZI+rZ31aILJBU4506bUgQAAACgfvg2rcjMnpWUIamDmeVLuk9SC0ly\nzv1G0kpJX5eUJ+m4pBv9qgUAAADAmfkWDpxzk8+w3Un6vl/7BwAAAFA73CEZAAAAgCTCAQAAAAAP\n4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/h\nAAAAAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAAD+EA\nAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP4QAA\nAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAA\nAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAkwgEAAAAAj6/hwMwy\nzSzXzPLMbHYV23uY2Voze8/M3jezr/tZDwAAAIDq+RYOzCxc0pOSxkpKljTZzJIrNfuRpCXOuYGS\nJkma51c9AAAAAGrm55GDIZLynHMfOudOSnpO0oRKbZyks7zHbSXt9bEeAAAAADWI8LHvrpL2BCzn\nSxpaqc39klab2Q8ktZI02sd6AAAAANTAz3BgVaxzlZYnS3raOTfHzC6U9EczS3HOlX2pI7PpkqZL\nUlxcnLKysvyo95RZ/Up87b8+FRYW+v5+oflirADBYawAaCz8DAf5kroHLHfT6dOGpkrKlCTn3D/N\nrKWkDpL2BzZyzi2QtECS0tPTXUZGhk8ll7th9su+9l+fns5sJb/fLzRfjBUgOIwVAI2Fn+ccrJPU\ny8wSzCxS5Sccr6jUZrekSyTJzPpIainpgI81AQAAAKiGb+HAOVciaYakVZK2qvyqRFvM7AEzG+81\nmyXpu2a2UdKzkm5wzlWeegQAAACgHvg5rUjOuZWSVlZad2/A4xxJw/2sAQAAAEBwuEMyAAAAAEmE\nAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQD\nAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMA\nAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAA\nAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAA\nAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAewgEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAA\nHsIBAAAAAEk+hwMzyzSzXDPLM7PZ1bS52sxyzGyLmf3Jz3oAAAAAVC/Cr47NLFzSk5IulZQvaZ2Z\nrXDO5QS06SXpPyQNd84dNrOOftUDAAAAoGZ+HjkYIinPOfehc+6kpOckTajU5ruSnnTOHZYk59x+\nH+sBAAAAUAPfjhxI6ippT8ByvqShldokSpKZvSkpXNL9zrm/Ve7IzKZLmi5JcXFxysrK8qPeU2b1\nK/G1//pUWFjo+/uF5ouxAgSHsQKgsfAzHFgV61wV++8lKUNSN0mvm1mKc+7Il57k3AJJCyQpPT3d\nZWRk1HmxgW6Y/bKv/denpzNbye/3C80XYwUIDmMFQGPh57SifEndA5a7SdpbRZs/O+eKnXMfScpV\neVgAAAAAUM/8DAfrJPUyswQzi5Q0SdKKSm2WSxolSWbWQeXTjD70sSYAAAAA1fAtHDjnSiTNkLRK\n0lZJS5xzW8zsATMb7zVbJekzM8uRtFbSD51zn/lVEwAAAIDq+XnOgZxzKyWtrLTu3oDHTtJM7x8A\nAACAEOIOyQAAAAAkEQ4AAAAAeAgHAAAAACTVIhyY2Qgzu9F7HGtmCf6VBQAAAKC+BRUOzOw+SfdI\n+g9vVQtJz/hVFAAAAID6F+yRg29KGi/pmCQ55/ZKauNXUQAAAADqX7Dh4KR32VEnSWbWyr+SAAAA\nAIRCsOFgiZnNl3S2mX1X0t8lLfSvLAAAAAD1LaiboDnnHjezSyUdlZQk6V7n3BpfKwMAAABQr84Y\nDswsXNIq59xoSQQCAAAAoIk647Qi51yppONm1rYe6gEAAAAQIkFNK5J0QtImM1sj74pFkuScu82X\nqgAAAADUu2DDwcvePwAAAABNVLAnJP/BzCIlJXqrcp1zxf6VBQAAAKC+BRUOzCxD0h8k7ZRkkrqb\n2Xecc6/5VxoAAACA+hTstKI5ki5zzuVKkpklSnpWUppfhQEAAACoX8HeBK1FRTCQJOfcdkkt/CkJ\nAAAAQCgEe+Qg28x+J+mP3vJ1ktb7UxIAAACAUAg2HNwi6fuSblP5OQevSZrnV1EAAAAA6l+w4SBC\n0i+dcz+XTt01Ocq3qgAAAJqo9evXd4yIiPitpBQFP8UbqAtlkjaXlJRMS0tL219Vg2DDwauSRksq\n9JajJa2WNOwrlwgAANCMRERE/LZTp059YmNjD4eFhblQ14Pmo6yszA4cOJC8b9++30oaX1WbYNNq\nS+dcRTCQ9zimDmoEAABoblJiY2OPEgxQ38LCwlxsbGyByo9aVd0myL6OmdmgigUzS5f0xVesDwAA\noDkKIxggVLzPXrUZINhpRbdLWmpmeyU5SV0kXfPVywMAAADQUAQbDhIkDZTUQ9I3JV2g8pAAAACA\nryB+9st1elPZnT/7RoO53PyQIUOSHn/88T1f+9rXjgfTvqysTMOGDUtcuXJlXrt27cr8ru+r+utf\n/9pmzpw5cWvXrs3zo//Fixe33bJlS/TDDz+8r/K2mJiYgcePH3+vuucOGzYs8c9//vMHsbGxpbXZ\nZ7DTin7snDsq6WxJl0paIOnXtdkRAAAAGo/i4uJ63+eSJUva9u3b94vGEAzqw3XXXVdQVTAIxuTJ\nkz97/PHHY2v7vGDDQUXi+Iak3zjn/iwpsrY7AwAAQOj98Ic/7JyQkNB32LBhvcaNG5dw7733xknl\nv/TPmDGj6+DBg5MefPDBuL1790aMGTPmvJSUlD4pKSl9Vq9e3UqSjh49GnbVVVfFp6Sk9OnTp0/y\nM888c7YkFRYW2uWXX35uYmJi8je+8Y1zT5w4YZL0xBNPdJg6dWr3iv3PmTOnw7Rp07pVrmvx4sXt\nvvnNbx6RpNzc3MiEhIS+11xzTc9evXr1HT9+fMLy5cvbDBo0qHfPnj1T1q5dG1NTLbm5uZFpaWlJ\nycnJfZKTk/usWbOmlVT+a/+QIUOSMjMzz01ISOg7fvz4hLKymrPIxIkT46+99toeaWlpSfHx8SnP\nPvts28pt1q5dGzNw4MDeffr0SR44cGDvjRs3RklSdnZ2y379+vXp3bt3cmJiYvKmTZuign1tc+fO\nbf/tb3+7hyRt27YtcsCAAb1TUlL63H777V0q9rtr164W6enpSb17907u1atX37/97W+tJWnSpElH\nXnzxxfZBfiROCXZa0cdmNl/llzN91MyixHV5ATQln2yQ7p8Q6irqxv0Foa4AQAP22muvxfzlL385\nZ9OmTTnFxcU2YMCA5IEDB56a9nPkyJHwdevW5UrSuHHjEmbOnPnpmDFjCnfs2BE5ZsyYXh9++OGW\n//zP/+w8atSoo0uXLt158ODB8PT09D7jx48/+vOf/zw2Ojq6bPv27Tn/+te/oocPH54sSVOnTj3U\nt2/f5KKiovyoqCj3zDPPdJg/f/6uyrWtX7++9fDhw0+t37NnT8vnn3/+w7S0tF39+/fvs3jx4vbZ\n2dnb/vSnP5390EMPdR41atQH1dXSpUuXktdff317TEyM27RpU9TkyZPP3bx581ZJ2rp1a/SGDRs+\njI+PL05LS+u9Zs2a1mPGjCmsXE+gPXv2RL3zzju5OTk5UaNHj06aMGHCpsDtqampJ955551tLVq0\n0PLly9vcfffd3VatWvXBr371q9hbb73101tuueXQiRMnrKSkRB9//HGLYF5bYP+33nprj2nTph2Y\nMWPGZ4888sipIwJPPfVUu0suuaTg0Ucf3VdSUqLPP/88TJJiY2NLT548afv27Qvv1KlT0FOLgg0H\nV0vKlPS4c+6ImXWW9MNgdwIAAICGISsrq/XYsWOPtG7d2klyl1566ZHA7ZMnTz5U8fjNN988a8eO\nHdEVy4WFheGHDx8Oy8rKOmvVqlVnz507t5MkFRUVWV5eXuQbb7zR+rbbbtsvSUOHDv0iMTHxuCSd\nddZZZcOHD//8+eefb9uvX78TxcXFNmTIkNOufFlQUBBxzjnnnPoZv2vXrkUV7RITE7+4+OKLj4aF\nhWnQoEHHH3zwwS7e66mylp49exZPnTq1Z05OTnRYWJh27dp16ga+/fr1O3beeecVS1Lfvn2Pf/DB\nB2ecETNx4sRD4eHh6tevX1H37t2LNmzY0DJw+6FDh8KvueaahJ07d7Y0M1dcXGySdOGFFx57/PHH\nO+fn50dOmjTpcL9+/YqCfW2B3n333davvPLKB5J08803f/bTn/60myRdcMEFx26++eb44uLisCuv\nvPLwsGHDTr2v7du3L9m9e3dkp06dgr7KaFDhwDl3XNKLAcufSPok2J0AAACgYXCu5mvKtGnT5tSX\nc+ecsrOzt3pB4kt9LFu2LC81NbWo8vPNrMp+p0+ffvChhx7qlJiYeGLKlCkHq2oTHh7uSktLFR4e\nLkmKjIw8td+wsDC1bNnSee1UWlpqNdUyc+bMLh07dix+4YUXPiorK1N0dPSpE7+joqJO9RseHq6S\nkpKqi67hdVVevueee7qOHDny8zVr1nyQm5sbefHFFydJ0ve+971DF1100bGXXnqp7dixYxPnzZu3\nMykpqSiY11ZZVZfAHTt2bOFrr72W+8ILL7S94YYbEm677bZPZ8yY8ZlUHpRiYmJqdf4GU4MAAACa\nkYyMjMJVq1a1PX78uBUUFIT9/e9/P7u6tiNGjDj66KOPdqxYfuutt6IladSoUUfnzJkTVzFX/803\n34z22hc+88wz7SRp3bp1Lbdv337qprkXX3zxsU8++STypZdeaj916tRDqkJCQsKJrVu3RlW1rTrV\n1VJQUBDeuXPn4vDwcM2bN699aemZZ9Z8//vf77po0aIq348XX3zxnNLSUm3ZsiVqz549UampqScC\ntx89ejS8W7duJyVp/vz5HSrW5+TkRPbp06foRz/60f7LLrvsyIYNG6Ir9x2MQYMGFS5cuLCdJC1c\nuPDUuQTbt2+P7Nq1a/GsWbMOTpky5eC7774bI5Vf+enAgQMtkpKSTgtwNQl2WhEAAAB8UN+XHh05\ncuTxzMzMguTk5L5du3Yt6t+//7G2bdtW+c15wYIFe6ZNm9YjMTExubS01IYOHfr5sGHDdv/sZz/b\nO3369B69e/dOds5Zt27ditauXZt311137Z80aVJCYmJict++fY/369fvWGB/V1xxxeH3338/prrL\na1522WUFq1evbpOSkhL0F9rqarnjjjv2T5w48bzly5efM2LEiM+jo6PP+At6Tk5OdMUJ0ZWdf/75\nRUOGDEn67LPPWvziF7/YFRMT86Vf8e+5555906ZNS5g7d26niy666GjF+j/+8Y/tli5d2j4iIsLF\nxsYWP/LII3uPHDkSHuzrqzBv3rzdkyZNOnfevHlx48ePP1yxftWqVW3mzp3bKSIiwsXExJQuXrz4\nI0l64403YgYOHHisRYsWtdqPnenQUkOTnp7usrOzfd1H/OyXfe2/Pj2d2UoZGRmhLgNNVJMaK6k7\nlJF7X6jLqBuckNzgNKmxwt+VWjGz9c659MB1Gzdu3JmamlrltJr6UlBQENa2bduyzz//POzCCy9M\n+s1vfrNrxIgRQd2L4KsYNWrU+XfcccenEyZM+Lyq7bt27WoxefLk+LfeemuH37VUZcSIEb3eeOON\n0/Y9ceLE+Msvv7zgxhtvPFzV8xqiG2+8sfsVV1xxpKr3euPGjR1SU1Pjq3oe04oAAACamSlTpvTs\n3bt3cv/+/fuMGzfusN/B4ODBg+Hx8fEpLVu2LKsuGEhSz549i2+66aaDhw4dCsl31KqCQWOVkpLy\nRU3vdXWYVgQAANDM/OUvf/moPvfXoUOH0p07d24Opu20adMa3K/zL7zwws5Q11Bbs2bN+reOTnHk\nAAAAAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAPJyQDAACE0v1t087cqDb9FZzxvglmljZhwoRDy5cv\n/0iSiouL1bFjx9QBAwYcW7t2bd7cuXPbZ2dnt1q0aNHumTNndmndunXpAw888GlgH+Hh4Wm9evX6\norS01M4///wvlixZsjPw7spS+Y24hg0blrhy5cq8du3alV111VXxr776atv27duX7NixY0tVtb33\n3nstb7zxxvicnJyY2bNnfxy432XLlp1111139SgrK9OUKVMOPvzww/sqP/+VV15pPWvWrO7bt2+P\nWbhw4YcVlx/du3dvxDXXXJPw+uuvN5krEvmBIwcAAADNTHR0dFlubm50YWGhSdJLL710VlxcXHFt\n+oiKiirbtm1bzo4dO7a0aNHCzZkzJ7ZymyVLlrTt27fvF+3atSuTpJtuuungihUravxy3rFjx5Jf\n/vKXu2+++eYvhZGSkhLdeeedPVauXLl9+/btW1544YV269evb1n5+eeee+7J3//+9zvHjRv3WeD6\nLl26lMTFxRWvXr26VW1eZ3NDOAAAAGiGLrnkkoKlS5eeLUnPPvtsu4kTJx76d/saMWJEYV5eXlTl\n9YsXL24XeMfhsWPHFsbGxpbU1FfXrl1LRo4cebxFixZfulNvVlZWq549exYlJyefbNmypfvWt751\naNmyZWdXfn5SUtLJoUOHfhEWdvrX3CuuuOLIokWL2tfqxTUzhAMAAIBm6Prrrz/0/PPPn3P8+HHb\nunVrzIUXXnjs3+mnuLhYq1atOqtfv35fVN62fv361sOHDz9jv4899ljsY489dtqRh0B79uyJ7Nq1\n68mK5W7dup38+OOPIyXpjjvu6LJ48eK2Z9rP8OHDj73zzjutz9SuOfP1nAMzy5T0S0nhkn7rnPtZ\nNe2ulLRU0mDnXLafNQEAAEAaOnToF/n5+VELFy5sN3r06ILaPr+oqCisd+/eyV5fn99+++2n3XSr\noKAg4pxzzik7/dlfdvfddx84Uxvn3GnrzMxJ0i9+8Yu9wdTcpUuXkv3790cG07a58i0cmFm4pCcl\nXSopX9I6M1vhnMup1K6NpNsk/cuvWgAAAHC6zMzMI/fdd1/31atX5+7fv79W3wsrzjmoqU14eLgr\nLS1VeHj4VytUUo8ePU4dKZCk/Pz8yC5dutTqPInjx49bVFTUGcNKc+bntKIhkvKccx86505Kek7S\nhCra/VTSY5JO+FgLAAAAKrnlllsOzpo1a++QIUNOmxJUFxISEk5s3br1tHMR/h0jR448tnPnzpbb\ntm2LPHHihL344ovtJk6ceOTMz/x/mzdvbpmYmOjLa20q/JxW1FXSnoDlfElDAxuY2UBJ3Z1zfzWz\nu6rryMymS5ouSXFxccrKyqr7agPM6lfjeTKNSmFhoe/vF5qvJjVWorooK+knoS6jbjDmG5wmNVb4\nu1L3grj0qF/OO++84h//+Mf7z9TuiSee6Dx//vy4iuVPP/30/WD6v+yyywpWr17dJiUlpUiSxo0b\nl/D222+3OXz4cERcXFz/2bNn773zzjsPVpxvcPfddx/YvXt3xODBg5OPHTsWbmZu/vz5cVu3bt3c\nrl27sjlz5uzOzMxMLC0t1bXXXnswPT39hFR+zsHgwYOPXXfddQX/+Mc/Yq6++urzjx49Gv7qq6+e\n/dBDD3XJy8vbIklr1qxpk5mZWespVM2JVTV/q046NrtK0hjn3DRv+XpJQ5xzP/CWwyT9r6QbnHM7\nzSxL0l1nOucgPT3dZWf7e1pC/OyXfe2/Pj2d2UoZGRmhLgNNVJMaK6k7lJF7X6jLqBv383evoWlS\nY4W/K7ViZuudc+mB6zZu3LgzNTX1tPn5TdGuXbtaTJ48Of6tt95qEPcWSE9PT3rllVfyYmNjS0Nd\nSyht3LixQ2pqanxV2/ycVpQvqXvAcjdJgSeLtJGUIinLzHZKukDSCjP70gACAABA49SzZ8/im266\n6eChQ4dCfoXMvXv3Rtx+++2fNvdgcCZ+TitaJ6mXmSVI+ljSJEnXVmx0zhVI6lCxHOyRAwAAADQe\n06ZNOxzqGqTyKxVdf/31tTpHoTnyLcU550okzZC0StJWSUucc1vM7AEzG+/XfgEAABq4srKyMgt1\nEWievM9etVds8vU+B865lZJWVlp3bzVtM/ysBQAAoIHYfODAgeTY2NiCsLAwf07+BKpQVlZmBw4c\naCtpc3VtfA0HAAAA+LKSkpJp+/bt++2+fftS5O/5n0BlZZI2l5SUTKuuAeEAAACgHqWlpe2XxBRr\nNEikVQAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAA\nPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8\nhAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJKkiFAXAAAAGpFPNkj3Twh1FXXj/oJQ\nVwA0OBw5AAAAACCJcAAAAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAA\nAADAQzgAAAAAIIlwAAAAAMBDOAAAAAAgiXAAAAAAwEM4AAAAACCJcAAAAADAQzgAAAAAIIlwAAAA\nAMBDOAAAAAAgiXAAAAAAwEM4AAAAACDJ53BgZplmlmtmeWY2u4rtM80sx8zeN7NXzaynn/UAAAAA\nqF6EXx2bWbikJyVdKilf0jozW+Gcywlo9p6kdOfccTO7RdJjkq7xq6Zm6ZMN0v0TQl1F3bi/INQV\nAAAANGl+HjkYIinPOfehc+6kpOckfelbqnNurXPuuLf4tqRuPtYDAAAAoAbmnPOnY7MrJWU656Z5\ny9dLGuqcm1FN+/+RtM8592AV26ZLmi5JcXFxac8995wvNVfY9HHT+YU6IbpIrYv2hrqMutF5QKgr\nQCWMlQaKsdLgMFYaqHoYK6NGjVrvnEv3fUdAHfFtWpEkq2JdlUnEzKZISpc0sqrtzrkFkhZIUnp6\nusvIyKijEqt2w+yXfe2/Pj2d+pEycu8LdRl1Y3LT+ePaVDBWGijGSoPDWGmgGCvAafwMB/mSugcs\nd5N02k8NZjZa0n9JGumcK/KxHgAAAAA18POcg3WSeplZgplFSpokaUVgAzMbKGm+pPHOuf0+1gIA\nAADgDHwLB865EkkzJK2StFXSEufcFjN7wMzGe83+W1JrSUvNbIOZraimOwAAAAA+83NakZxzKyWt\nrLTu3oDHo/3cPwAAAIDgcYdkAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAA\nPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8\nhAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyE\nAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQD\nAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMA\nAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPL6GAzPLNLNcM8szs9lVbI8ys+e97f8ys3g/6wEAAABQ\nPd/CgZmFS3pS0lhJyZImm1lypWZTJR12zp0v6QlJj/pVDwAAAICa+XnkYIikPOfch865k5KekzSh\nUpsJkv7gPV4m6RIzMx9rAgAAAFANP8NBV0l7ApbzvXVVtnHOlUgqkNTex5oAAAAAVCPCx76rOgLg\n/o02MrPpkqZ7i4VmlvsVa2s2RkkdJB0MdR114iccVIJ/GCtAcBgrtdazPnYC1BU/w0G+pO4By90k\n7a2mTb6ZRUhqK+lQ5Y6ccwuR1+YiAAAFVUlEQVQkLfCpzibNzLKdc+mhrgNo6BgrQHAYK0DT5ue0\nonWSeplZgplFSpokaUWlNiskfcd7fKWk/3XOnXbkAAAAAID/fDty4JwrMbMZklZJCpf0lHNui5k9\nICnbObdC0u8k/dHM8lR+xGCSX/UAAAAAqJnxQ33TZmbTvWlZAGrAWAGCw1gBmjbCAQAAAABJPt8h\nGQAAAEDjQTgAAAAAIIlw0OyZWbyZbfYeZ5jZX0NdExAKZvbWGbavNLOz66seoCkwsywzS/ce7zSz\nDqGuCUDN/LzPAXxkZqbyc0bKQl0L0NCYWbhzrrQ2z3HODTvD9q9/taqApoe/RUDTw5GDRsT7lX+r\nmc2T9K6k683sn2b2rpktNbPWXrvBZvaWmW00s3fMrI333Ne9tu+aWY1fhICGyvssbzOzP5jZ+2a2\nzMxivF8l7zWzNyRdZWbnmdnfzGy999nv7T0/zsxe8sbHxoqxYGaF3n87m9lrZrbBzDab2UXe+lO/\neprZTG/bZjO7I6CurWa20My2mNlqM4sOyZsE1KDy59fMHjWzWwO2329ms7zHPzSzdd5Y+4m3rvLf\nou5m9mszy/Y++z8JzSsDUBcIB41PkqRFki6VNFXSaOfcIEnZkmZ6N5x7XtLtzrlUSaMlfSFpv6RL\nvbbXSJobiuKBOpIkaYFzrr+ko5IqvticcM6NcM49p/K7qv/AOZcm6S5J87w2cyX9wxsfgyRtqdT3\ntZJWOecGSEqVtCFwo5mlSbpR0lBJF0j6rpkN9Db3kvSkc66vpCOSJtbVCwbqQlWfX0nPqfzvQoWr\nJS01s8tU/pkeImmApDQz+5rXJknSIufcQOfcLkn/5d01ub+kkWbWv15eEIA6x7SixmeXc+5tM7tc\nUrKkN8uP6ipS0j9V/j/sT5xz6yTJOXdUksyslaT/MbMBkkolJYaieKCO7HHOvek9fkbSbd7j5yXJ\nO4o2TOVfcCqeE+X992JJ35Ykb+pRQaW+10l6ysxaSFrunNtQafsISS855455+3pR0kUqv+P7RwHt\n10uK/wqvEfBDdZ/fjmbWRVKspMPOud1mdpukyyS95z23tcrDwm55f4sC+r3azKar/HtFZ5X/fXq/\nPl4QgLpFOGh8jnn/NUlrnHOTAzd6v9ZUdfOKOyV9qvJfQsMknfCzSMBnlT/jFcsV4yNM0hHv1//a\ndezca96vo99Q+R3c/9s5tyigiVXzVEkqCnhcKolpRWhoqvv8LpN0paROKj+SUNH2Eefc/C91YBav\n/x9rMrMElR+dG+ycO2xmT0tqWadVA6g3TCtqvN6WNNzMzpckb851oqRtkrqY2WBvfRszi5DUVuVH\nFMokXS8pPER1A3Whh5ld6D2eLOmNwI3eEbOPzOwqqfykSTNL9Ta/KukWb324mZ0V+Fwz6ylpv3Nu\noaTfqXzqUaDXJF3hjblWkr4p6fW6e2mAr6r7/D4naZLKA8Iyr+0qSTcFnM/W1cw6VtHnWSoPCwVm\nFidprM+vAYCPCAeNlHPugKQbJD1rZu+rPCz0ds6dVPnc0V+Z2UZJa1T+C848Sd8xs7dVPqXoWJUd\nA43DVpV/nt+X1E7Sr6toc52kqd442CJpgrf+dkmjzGyTyqf+9K30vAxJG8zsPZWfM/DLwI3OuXcl\nPS3pHUn/kvRb59x7AhqB6j6/zrktktpI+tg594nXdrWkP0n6pzdelnltKve5UeVTj7ZIekrSm5Xb\nAGg8zLmqZqAAQMPkTWn4q3MuJcSlAADQ5HDkAAAAAIAkjhwAAAAA8HDkAAAAAIAkwgEAAAAAD+EA\nAAAAgCTCAQAAAAAP4QAAAACAJOn/APnY1Q/piun+AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fe90c088828>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:10:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>66705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>9768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>11724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>1978</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0        66705\n",
       "1  plasmid_1         9768\n",
       "2  plasmid_2        11724\n",
       "3  plasmid_3         1978"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.207174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.475098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.288530</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.207174\n",
       "1  precision  0.475098\n",
       "2   f1_score  0.288530"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP016924.1</th>\n",
       "      <th>CP016925.1</th>\n",
       "      <th>CP016927.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.091403</td>\n",
       "      <td>0.357025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.234520</td>\n",
       "      <td>0.040267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.170751</td>\n",
       "      <td>0.015314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.496674</td>\n",
       "      <td>0.445924</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP016924.1  CP016925.1  CP016927.1\n",
       "0  plasmid_0         0.0    0.091403    0.357025\n",
       "1  plasmid_1         0.0    0.234520    0.040267\n",
       "2  plasmid_2         0.0    0.170751    0.015314\n",
       "3  plasmid_3         0.0    0.000000    0.033318\n",
       "4      Total         0.0    0.496674    0.445924"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016924.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016925.1</td>\n",
       "      <td>0.039427</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.646622</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016927.1</td>\n",
       "      <td>0.364171</td>\n",
       "      <td>0.185401</td>\n",
       "      <td>0.087769</td>\n",
       "      <td>0.607179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.403598</td>\n",
       "      <td>1.185401</td>\n",
       "      <td>0.734391</td>\n",
       "      <td>0.607179</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2  plasmid_3\n",
       "0  CP016924.1   0.000000   0.000000   0.000000   0.000000\n",
       "1  CP016925.1   0.039427   1.000000   0.646622   0.000000\n",
       "2  CP016927.1   0.364171   0.185401   0.087769   0.607179\n",
       "3       Total   0.403598   1.185401   0.734391   0.607179"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.364171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.646622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.607179</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.364171\n",
       "1  plasmid_1    1.000000\n",
       "2  plasmid_2    0.646622\n",
       "3  plasmid_3    0.607179"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP016924.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP016925.1</td>\n",
       "      <td>0.367762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP016927.1</td>\n",
       "      <td>0.386723</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP016924.1    0.000000\n",
       "1  CP016925.1    0.367762\n",
       "2  CP016927.1    0.386723"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No high mutual correspondence found.\n"
     ]
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 10, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00551902</td>\n",
       "      <td>-0.00314283</td>\n",
       "      <td>-0.00448188</td>\n",
       "      <td>-0.0061541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[36, 81, 42, 22, 50, 44, 103, 53, 90, 49, 43, ...</td>\n",
       "      <td>[44, 32, 49, 90, 33]</td>\n",
       "      <td>[90, 29, 49, 96, 31]</td>\n",
       "      <td>[48, 42]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.769903</td>\n",
       "      <td>-0.741847</td>\n",
       "      <td>-0.840622</td>\n",
       "      <td>-0.904495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.775422</td>\n",
       "      <td>-0.74499</td>\n",
       "      <td>-0.845104</td>\n",
       "      <td>-0.910649</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0  \\\n",
       "GC_sum                                        -0.00551902   \n",
       "chain   [36, 81, 42, 22, 50, 44, 103, 53, 90, 49, 43, ...   \n",
       "gd_sum                                          -0.769903   \n",
       "total                                           -0.775422   \n",
       "\n",
       "                   plasmid_1             plasmid_2  plasmid_3  \n",
       "GC_sum           -0.00314283           -0.00448188 -0.0061541  \n",
       "chain   [44, 32, 49, 90, 33]  [90, 29, 49, 96, 31]   [48, 42]  \n",
       "gd_sum             -0.741847             -0.840622  -0.904495  \n",
       "total               -0.74499             -0.845104  -0.910649  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00489959</td>\n",
       "      <td>-0.00114712</td>\n",
       "      <td>-0.00754457</td>\n",
       "      <td>-0.0232257</td>\n",
       "      <td>-0.00477389</td>\n",
       "      <td>-0.0050703</td>\n",
       "      <td>-0.0218655</td>\n",
       "      <td>-0.00219798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[31, 96, 49, 32, 44, 50, 22, 42, 48, 81, 36, 8...</td>\n",
       "      <td>[90, 28, 49, 90, 53, 103, 44, 50, 35]</td>\n",
       "      <td>[16]</td>\n",
       "      <td>[40, 47]</td>\n",
       "      <td>[90, 29, 49, 90, 33, 49, 96, 26, 49]</td>\n",
       "      <td>[43, 49]</td>\n",
       "      <td>[39]</td>\n",
       "      <td>[44, 103]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.784979</td>\n",
       "      <td>-0.847652</td>\n",
       "      <td>-0.787935</td>\n",
       "      <td>-0.311963</td>\n",
       "      <td>-0.757116</td>\n",
       "      <td>-0.641967</td>\n",
       "      <td>-0.748986</td>\n",
       "      <td>-0.858636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.789879</td>\n",
       "      <td>-0.8488</td>\n",
       "      <td>-0.795479</td>\n",
       "      <td>-0.335188</td>\n",
       "      <td>-0.76189</td>\n",
       "      <td>-0.647037</td>\n",
       "      <td>-0.770851</td>\n",
       "      <td>-0.860834</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0  \\\n",
       "GC_sum                                        -0.00489959   \n",
       "chain   [31, 96, 49, 32, 44, 50, 22, 42, 48, 81, 36, 8...   \n",
       "gd_sum                                          -0.784979   \n",
       "total                                           -0.789879   \n",
       "\n",
       "                                    plasmid_1   plasmid_2  plasmid_3  \\\n",
       "GC_sum                            -0.00114712 -0.00754457 -0.0232257   \n",
       "chain   [90, 28, 49, 90, 53, 103, 44, 50, 35]        [16]   [40, 47]   \n",
       "gd_sum                              -0.847652   -0.787935  -0.311963   \n",
       "total                                 -0.8488   -0.795479  -0.335188   \n",
       "\n",
       "                                   plasmid_4  plasmid_5  plasmid_6   plasmid_7  \n",
       "GC_sum                           -0.00477389 -0.0050703 -0.0218655 -0.00219798  \n",
       "chain   [90, 29, 49, 90, 33, 49, 96, 26, 49]   [43, 49]       [39]   [44, 103]  \n",
       "gd_sum                             -0.757116  -0.641967  -0.748986   -0.858636  \n",
       "total                               -0.76189  -0.647037  -0.770851   -0.860834  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_3  from MILP completely contained in  plasmid_0  from greedy.\n",
      "\n",
      "plasmid_5  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_6  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
