{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 56 with $\\alpha_1 = 1$, $\\alpha_2 = 1$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 56\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP017931.1</td>\n",
       "      <td>111395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP017933.1</td>\n",
       "      <td>11638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP017930.1</td>\n",
       "      <td>114139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP017929.1</td>\n",
       "      <td>113105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP017932.1</td>\n",
       "      <td>76186</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP017931.1       111395\n",
       "1  CP017933.1        11638\n",
       "2  CP017930.1       114139\n",
       "3  CP017929.1       113105\n",
       "4  CP017932.1        76186"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt0VeWZx/Hfk4Q7iAIhXCWpEiAJ\nBEgEBSxBUWCqYBdeoNWOCsVqqVpolZlprbW2VkdqiyMV7LjUShHwQmllBNohxWslKMjNACI3EQGB\nQLjm8swf2WFiSMJBs3OS8P2sxfLsvd/z7uecdV5zfme/e29zdwEAAABATLQLAAAAAFA7EA4AAAAA\nSCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAjERbuAM9Wm\nTRtPTEyMdhl1xuHDh9WsWbNolwHUeowVIDKMlTOzYsWKve4eH+06gEjVuXCQmJionJycaJdRZ2Rn\nZysrKyvaZQC1HmMFiAxj5cyY2dZo1wCcCaYVAQAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAA\nAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQCC0cmNnTZrbbzNZUst3MbJqZbTKzD8ysb1i1AAAAADi9\nMI8cPCNpeBXbR0jqGvybIOn3IdYCAAAA4DRCCwfuvkzSviqajJL0nJd4R9K5ZtY+rHoAAAAAVC0u\nivvuKGl7meUdwbpPyzc0swkqObqghIQEZWdn10R99UJ+fj7vF0Kz+pO8aJdQbZJaxjJWgAjwdwWo\n36IZDqyCdV5RQ3efKWmmJGVmZnpWVlaIZdUv2dnZ4v1CWG6e8mq0S6g2zwxvxlgBIsDfFaB+i+bV\ninZI6lxmuZOknVGqBQAAADjrRTMcLJD0neCqRRdLynP3U6YUAQAAAKgZoU0rMrPZkrIktTGzHZJ+\nJqmBJLn7k5IWSvoXSZskHZF0S1i1AAAAADi90MKBu489zXaX9P2w9g8AAADgzHCHZAAAAACSCAcA\nAAAAAoQDAAAAAJKie58DAADOCon17J4gAOovjhwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAc\nAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABDgJmgV4GY1AAAAOBtx5AAAAACAJMIBAAAAgADhAAAA\nAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAA\ngCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACA\nJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAk\nwgEAAACAAOEAAAAAgKSQw4GZDTezXDPbZGZTKth+vpktNbP3zewDM/uXMOsBAAAAULnQwoGZxUp6\nQtIISSmSxppZSrlmP5E01937SBojaXpY9QAAAACoWphHDvpJ2uTum939hKQXJI0q18YlnRM8bilp\nZ4j1AAAAAKhCXIh9d5S0vczyDkn9y7W5X9JiM/uBpGaShoZYDwAAAIAqhBkOrIJ1Xm55rKRn3H2q\nmV0i6Y9mlubuxV/oyGyCpAmSlJCQoOzs7DDqPWlyz8JQ+69J+fn5ob9fOHsxVoDIMFYA1BVhhoMd\nkjqXWe6kU6cNjZM0XJLc/W0zayypjaTdZRu5+0xJMyUpMzPTs7KyQiq5xM1TXg21/5r0zPBmCvv9\nwtmLsQJEhrECoK4I85yD5ZK6mlmSmTVUyQnHC8q12Sbpckkysx6SGkvaE2JNAAAAACoRWjhw90JJ\nEyUtkrReJVclWmtmD5jZyKDZZEnfNbNVkmZLutndy089AgAAAFADwpxWJHdfKGlhuXX3lXm8TtLA\nMGsAAAAAEBnukAwAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABA\nEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS\n4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLh\nAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEA\nAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAA\nAABAgHAAAAAAQBLhAAAAAECAcAAAAABAUsjhwMyGm1mumW0ysymVtLnezNaZ2Voz+1OY9QAAAACo\nXFxYHZtZrKQnJF0haYek5Wa2wN3XlWnTVdK/SRro7vvNrG1Y9QAAAACoWphHDvpJ2uTum939hKQX\nJI0q1+a7kp5w9/2S5O67Q6wHAAAAQBVCO3IgqaOk7WWWd0jqX65NsiSZ2ZuSYiXd7+6vle/IzCZI\nmiBJCQkJys7ODqPekyb3LAy1/5qUn58f+vuFsxdjBYgMYwVAXRFmOLAK1nkF++8qKUtSJ0mvm1ma\nux/4wpPcZ0qaKUmZmZmelZVV7cWWdfOUV0PtvyY9M7yZwn6/cPZirACRYawAqCvCnFa0Q1LnMsud\nJO2soM2f3b3A3T+WlKuSsAAAAACghoUZDpZL6mpmSWbWUNIYSQvKtZkvaYgkmVkblUwz2hxiTQAA\nAAAqEVo4cPdCSRMlLZK0XtJcd19rZg+Y2cig2SJJn5vZOklLJf3Y3T8PqyYAAAAAlQvznAO5+0JJ\nC8utu6/MY5c0KfgHAAAAIIq4QzIAAAAASYQDAAAAAAHCAQAAAABJZxAOzGyQmd0SPI43s6TwygIA\nAABQ0yIKB2b2M0n3Svq3YFUDSc+HVRQAAACAmhfpkYNvShop6bAkuftOSS3CKgoAAABAzYs0HJwI\nLjvqkmRmzcIrCQAAAEA0RBoO5prZDEnnmtl3Jf1N0lPhlQUAAACgpkV0EzR3f9TMrpB0UFI3Sfe5\n+5JQKwMAAABQo04bDswsVtIidx8qiUAAAAAA1FOnnVbk7kWSjphZyxqoBwAAAECURDStSNIxSavN\nbImCKxZJkrvfGUpVAAAAAGpcpOHg1eAfAAAAgHoq0hOSnzWzhpKSg1W57l4QXlkAAAAAalpE4cDM\nsiQ9K2mLJJPU2cz+1d2XhVcaAAAAgJoU6bSiqZKudPdcSTKzZEmzJWWEVRgAAACAmhXpTdAalAYD\nSXL3DZIahFMSAAAAgGiI9MhBjpn9t6Q/BsvflrQinJIAAAAAREOk4eB2Sd+XdKdKzjlYJml6WEUB\nAAAAqHmRhoM4Sb9z999IJ++a3Ci0qgAAAOqpFStWtI2Li/uDpDRFPsUbqA7FktYUFhaOz8jI2F1R\ng0jDwd8lDZWUHyw3kbRY0oCvXCIAAMBZJC4u7g/t2rXrER8fvz8mJsajXQ/OHsXFxbZnz56UXbt2\n/UHSyIraRJpWG7t7aTBQ8LhpNdQIAABwtkmLj48/SDBATYuJifH4+Pg8lRy1qrhNhH0dNrO+pQtm\nlinp6FesDwAA4GwUQzBAtASfvUozQKTTiu6SNM/MdkpySR0k3fDVywMAAABQW0QaDpIk9ZF0vqRv\nSrpYJSEBAAAAX0HilFer9aayW379jVpzufl+/fp1e/TRR7d//etfPxJJ++LiYg0YMCB54cKFm1q1\nalUcdn1f1V//+tcWU6dOTVi6dOmmMPqfNWtWy7Vr1zb51a9+tav8tqZNm/Y5cuTI+5U9d8CAAcl/\n/vOfP4qPjy86k31GOq3op+5+UNK5kq6QNFPS789kRwAAAKg7CgoKanyfc+fObZmamnq0LgSDmvDt\nb387r6JgEImxY8d+/uijj8af6fMiDQelieMbkp509z9LanimOwMAAED0/fjHP26flJSUOmDAgK5X\nX3110n333ZcglfzSP3HixI4XXXRRtwcffDBh586dccOGDbsgLS2tR1paWo/Fixc3k6SDBw/GXHfd\ndYlpaWk9evTokfL888+fK0n5+fl21VVXfS05OTnlG9/4xteOHTtmkvTYY4+1GTduXOfS/U+dOrXN\n+PHjO5Wva9asWa2++c1vHpCk3NzchklJSak33HBDl65du6aOHDkyaf78+S369u3bvUuXLmlLly5t\nWlUtubm5DTMyMrqlpKT0SElJ6bFkyZJmUsmv/f369es2fPjwryUlJaWOHDkyqbi46iwyevToxG99\n61vnZ2RkdEtMTEybPXt2y/Jtli5d2rRPnz7de/TokdKnT5/uq1ataiRJOTk5jXv27Nmje/fuKcnJ\nySmrV69uFOlrmzZtWuvvfOc750vShx9+2LB3797d09LSetx1110dSve7devWBpmZmd26d++e0rVr\n19TXXnutuSSNGTPmwMsvv9w6wo/ESZFOK/rEzGao5HKmD5tZI3FdXgAAgDpn2bJlTf/yl7+ct3r1\n6nUFBQXWu3fvlD59+pyc9nPgwIHY5cuX50rS1VdfnTRp0qTPhg0blr9x48aGw4YN67p58+a1//7v\n/95+yJAhB+fNm7dl7969sZmZmT1Gjhx58De/+U18kyZNijds2LDun//8Z5OBAwemSNK4ceP2paam\nphw/fnxHo0aN/Pnnn28zY8aMreVrW7FiRfOBAweeXL99+/bGc+bM2ZyRkbG1V69ePWbNmtU6Jyfn\nwz/96U/n/vKXv2w/ZMiQjyqrpUOHDoWvv/76hqZNm/rq1asbjR079mtr1qxZL0nr169vsnLlys2J\niYkFGRkZ3ZcsWdJ82LBh+eXrKWv79u2N3n333dx169Y1Gjp0aLdRo0atLrs9PT392LvvvvthgwYN\nNH/+/Bb33HNPp0WLFn30+OOPx99xxx2f3X777fuOHTtmhYWF+uSTTxpE8trK9n/HHXecP378+D0T\nJ078/KGHHjp5RODpp59udfnll+c9/PDDuwoLC3Xo0KEYSYqPjy86ceKE7dq1K7Zdu3YRTy2KNBxc\nL2m4pEfd/YCZtZf040h3AgAAgNohOzu7+YgRIw40b97cJfkVV1xxoOz2sWPH7it9/Oabb56zcePG\nJqXL+fn5sfv374/Jzs4+Z9GiRedOmzatnSQdP37cNm3a1PCNN95ofuedd+6WpP79+x9NTk4+Iknn\nnHNO8cCBAw/NmTOnZc+ePY8VFBRYv379TrnyZV5eXtx555138mf8jh07Hi9tl5ycfPSyyy47GBMT\no759+x558MEHOwSvp8JaunTpUjBu3Lgu69ataxITE6OtW7eevIFvz549D19wwQUFkpSamnrko48+\nOu2MmNGjR++LjY1Vz549j3fu3Pn4ypUrG5fdvm/fvtgbbrghacuWLY3NzAsKCkySLrnkksOPPvpo\n+x07djQcM2bM/p49ex6P9LWV9d577zX/n//5n48k6bbbbvv8F7/4RSdJuvjiiw/fdtttiQUFBTHX\nXnvt/gEDBpx8X1u3bl24bdu2hu3atYv4KqMRhQN3PyLp5TLLn0r6NNKdAAAAoHZwr/qaMi1atDj5\n5dzdlZOTsz4IEl/o48UXX9yUnp5+vPzzzazCfidMmLD3l7/8Zbvk5ORjN954496K2sTGxnpRUZFi\nY2MlSQ0bNjy535iYGDVu3NiDdioqKrKqapk0aVKHtm3bFrz00ksfFxcXq0mTJidP/G7UqNHJfmNj\nY1VYWFhx0VW8rvLL9957b8fBgwcfWrJkyUe5ubkNL7vssm6S9L3vfW/fpZdeeviVV15pOWLEiOTp\n06dv6dat2/FIXlt5FV0Cd8SIEfnLli3Lfemll1refPPNSXfeeednEydO/FwqCUpNmzY9o/M3mBoE\nAABwFsnKyspftGhRyyNHjlheXl7M3/72t3Mrazto0KCDDz/8cNvS5bfeequJJA0ZMuTg1KlTE0rn\n6r/55ptNgvb5zz//fCtJWr58eeMNGzacvGnuZZdddvjTTz9t+Morr7QeN27cPlUgKSnp2Pr16xtV\ntK0yldWSl5cX2759+4LY2FhNnz69dVHR6WfWfP/73+/43HPPVfh+vPzyy+cVFRVp7dq1jbZv394o\nPT39WNntBw8ejO3UqdMJSZoxY0ab0vXr1q1r2KNHj+M/+clPdl955ZUHVq5c2aR835Ho27dv/lNP\nPdVKkp566qmT5xJs2LChYceOHQsmT56898Ybb9z73nvvNZVKrvy0Z8+eBt26dTslwFUl0mlFAAAA\nCEFNX3p08ODBR4YPH56XkpKS2rFjx+O9evU63LJlywq/Oc+cOXP7+PHjz09OTk4pKiqy/v37Hxow\nYMC2X//61zsnTJhwfvfu3VPc3Tp16nR86dKlm370ox/tHjNmTFJycnJKamrqkZ49ex4u298111yz\n/4MPPmha2eU1r7zyyrzFixe3SEtLi/gLbWW13H333btHjx59wfz5888bNGjQoSZNmpz2F/R169Y1\nKT0hurwLL7zweL9+/bp9/vnnDX77299ubdq06Rd+xb/33nt3jR8/PmnatGntLr300oOl6//4xz+2\nmjdvXuu4uDiPj48veOihh3YeOHAgNtLXV2r69OnbxowZ87Xp06cnjBw5cn/p+kWLFrWYNm1au7i4\nOG/atGnRrFmzPpakN954o2mfPn0ON2jQ4Iz2Y6c7tFTbZGZmek5OTqj7SJzyaqj916RnhjdTVlZW\ntMtAPcVYASLDWDl7mdkKd88su27VqlVb0tPTK5xWU1Py8vJiWrZsWXzo0KGYSy65pNuTTz65ddCg\nQRHdi+CrGDJkyIV33333Z6NGjTpU0fatW7c2GDt2bOJbb721MexaKjJo0KCub7zxxin7Hj16dOJV\nV12Vd8stt+yv6Hm10S233NL5mmuuOVDRe71q1ao26enpiRU9j2lFAAAAZ5kbb7yxS/fu3VN69erV\n4+qrr94fdjDYu3dvbGJiYlrjxo2LKwsGktSlS5eCW2+9de++ffui8h21omBQV6WlpR2t6r2uDNOK\nAAAAzjJ/+ctfPq7J/bVp06Zoy5YtayJpO378+Fr36/xLL720Jdo1nKnJkyd/qaNTHDkAAAAAIIlw\nAAAAACBAOAAAAAAgiXAAAAAAIMAJyQAAANF0f8uM0zc6k/7yTnvfBDPLGDVq1L758+d/LEkFBQVq\n27Zteu/evQ8vXbp007Rp01rn5OQ0e+6557ZNmjSpQ/PmzYseeOCBz8r2ERsbm9G1a9ejRUVFduGF\nFx6dO3fulrJ3V5ZKbsQ1YMCA5IULF25q1apV8XXXXZf497//vWXr1q0LN27cuLai2t5///3Gt9xy\nS+K6deuaTpky5ZPy+y31q1/9Kv7JJ59M2L59e6OdO3euat++faEkzZ49u+W7777b7LHHHtsZ6VuG\n/8eRAwAAgLNMkyZNinNzc5vk5+ebJL3yyivnJCQkFJxJH40aNSr+8MMP123cuHFtgwYNfOrUqfHl\n28ydO7dlamrq0VatWhVL0q233rp3wYIFVV4utG3btoW/+93vtt12220VhoJSgwcPzl+yZMmGDh06\nnCi7/oYbbsh77bXXzj106BDfc78E3jQAAICz0OWXX543b968cyVp9uzZrUaPHr3vy/Y1aNCg/E2b\nNjUqv37WrFmtyt5xeMSIEfnx8fGFVfXVsWPHwsGDBx9p0KBBlXfqHThw4NFu3bqdKL8+JiZGAwYM\nODRnzpyWZ/IaUIJwAAAAcBa66aab9s2ZM+e8I0eO2Pr165tecsklh79MPwUFBVq0aNE5PXv2PFp+\n24oVK5oPHDjwtP0+8sgj8Y888sgpRx7KGzx48IVbtmxpcLp2mZmZh19//fXmp2uHU4V6zoGZDZf0\nO0mxkv7g7r+upN21kuZJusjdc8KsCQAAAFL//v2P7tixo9FTTz3VaujQoXln+vzjx4/HdO/ePSXo\n69Bdd911yk238vLy4s4777ziU5/9Rffcc8+eSPb5j3/8Y1Mk7dq1a1e4a9euhpG0xReFFg7MLFbS\nE5KukLRD0nIzW+Du68q1ayHpTkn/DKsWAAAAnGr48OEHfvazn3VevHhx7u7du8/oe2HpOQdVtYmN\njfWioiLFxsZ+tULP0NGjR61x48anDSU4VZjTivpJ2uTum939hKQXJI2qoN0vJD0i6ViItQAAAKCc\n22+/fe/kyZN39uvX75QpQdUhKSnp2Pr16085FyFsubm5jVNTU0N5TfVdmNOKOkraXmZ5h6T+ZRuY\nWR9Jnd39r2b2o8o6MrMJkiZIUkJCgrKzs6u/2jIm96zyPJk6JT8/P/T3C2cvxgoQGcYKqhTBpUfD\ncsEFFxT89Kc/3X26do899lj7GTNmJJQuf/bZZx9E0v+VV16Zt3jx4hZpaWnHJenqq69Oeuedd1rs\n378/LiEhodeUKVN2/vCHP9xber7BPffcs2fbtm1xF110Ucrhw4djzcxnzJiRsH79+jWtWrUqHjx4\n8IXPPvvs1sTExIIHH3yw7eOPP97u888/b5Cenp4yZMiQvDlz5myVpGXLlrV4+OGHP/ly78rZzdyr\nPBH8y3dsdp2kYe4+Pli+SVI/d/9BsBwj6X8l3ezuW8wsW9KPTnfOQWZmpufkhHtaQuKUV0PtvyY9\nM7yZsrKyol0G6inGChAZxsrZy8xWuHtm2XWrVq3akp6efsr8/Ppo69atDcaOHZv41ltvVXn50uq0\nffv2uOuvv/5rb7/99oaa2mdds2rVqjbp6emJFW0Lc1rRDkmdyyx3klT2ZhQtJKVJyjazLZIulrTA\nzL4wgAAAAFA3denSpeDWW2/du2/fvhq7QubmzZsbTp06dfvpW6IiYU4rWi6pq5klSfpE0hhJ3yrd\n6O55ktqULkd65AAAAAB1x/jx4/fX5P4GDx58pCb3V9+EluLcvVDSREmLJK2XNNfd15rZA2Y2Mqz9\nAgAA1HLFxcXFFu0icHYKPnuVXskp1PscuPtCSQvLrbuvkrZZYdYCAABQS6zZs2dPSnx8fF5MTEw4\nJ38CFSguLrY9e/a0lLSmsjahhgMAAAB8UWFh4fhdu3b9YdeuXWkK9/xPoLxiSWsKCwvHV9aAcAAA\nAFCDMjIydktiijVqJdIqAAAAAEkcOaj/Pl0p3V/RjanroPvzol0BAABAvcaRAwAAAACSCAcAAAAA\nAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAC\nhAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKE\nAwAAAACSCAcAAAAAAoQDAAAAAJKkuGgXAAAA6pBPV0r3j4p2FdXj/rxoVwDUOhw5AAAAACCJcAAA\nAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAA\nACBAOAAAAAAgiXAAAAAAIEA4AAAAACBJiot2AQBQK3y6Urp/VLSrqB7350W7AgBAHcWRAwAAAACS\nCAcAAAAAAoQDAAAAAJJCDgdmNtzMcs1sk5lNqWD7JDNbZ2YfmNnfzaxLmPUAAAAAqFxo4cDMYiU9\nIWmEpBRJY80spVyz9yVlunsvSS9KeiSsegAAAABULcwjB/0kbXL3ze5+QtILkr5wKRB3X+ruR4LF\ndyR1CrEeAAAAAFUwdw+nY7NrJQ139/HB8k2S+rv7xEra/5ekXe7+YAXbJkiaIEkJCQkZL7zwQig1\nl1r9Sf25DGBSk+NqfnxntMuoHu17R7sClMNYqaUYK7UOY6WWqoGxMmTIkBXunhn6joBqEuZ9DqyC\ndRUmETO7UVKmpMEVbXf3mZJmSlJmZqZnZWVVU4kVu3nKq6H2X5OeSf9YWbk/i3YZ1WNs/fnjWl8w\nVmopxkqtw1ippRgrwCnCDAc7JHUus9xJ0ik/NZjZUEn/IWmwux8PsR4AAAAAVQjznIPlkrqaWZKZ\nNZQ0RtKCsg3MrI+kGZJGuvvuEGsBAAAAcBqhhQN3L5Q0UdIiSeslzXX3tWb2gJmNDJr9p6TmkuaZ\n2UozW1BJdwAAAABCFua0Irn7QkkLy627r8zjoWHuHwAAAEDkuEMyAAAAAEmEAwAAAAABwgEAAAAA\nSYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJ\nhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmE\nAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQD\nAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMA\nAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAgVDDgZkNN7NcM9tk\nZlMq2N7IzOYE2/9pZolh1gMAAACgcqGFAzOLlfSEpBGSUiSNNbOUcs3GSdrv7hdKekzSw2HVAwAA\nAKBqYR456Cdpk7tvdvcTkl6QNKpcm1GSng0evyjpcjOzEGsCAAAAUIkww0FHSdvLLO8I1lXYxt0L\nJeVJah1iTQAAAAAqERdi3xUdAfAv0UZmNkHShGAx38xyv2JtZ40hUhtJe6NdR7X4OQeVEB7GChAZ\nxsoZ61ITOwGqS5jhYIekzmWWO0naWUmbHWYWJ6mlpH3lO3L3mZJmhlRnvWZmOe6eGe06gNqOsQJE\nhrEC1G9hTitaLqmrmSWZWUNJYyQtKNdmgaR/DR5fK+l/3f2UIwcAAAAAwhfakQN3LzSziZIWSYqV\n9LS7rzWzByTluPsCSf8t6Y9mtkklRwzGhFUPAAAAgKoZP9TXb2Y2IZiWBaAKjBUgMowVoH4jHAAA\nAACQFPIdkgEAAADUHYQDAAAAAJIIB2c9M0s0szXB4ywz+2u0awKiwczeOs32hWZ2bk3VA9QHZpZt\nZpnB4y1m1ibaNQGoWpj3OUDF8erWAAAEe0lEQVSIzMxUcs5IcbRrAWobM4t196IzeY67DzjN9n/5\nalUB9Q9/i4D6hyMHdUjwK/96M5su6T1JN5nZ22b2npnNM7PmQbuLzOwtM1tlZu+aWYvgua8Hbd8z\nsyq/CAG1VfBZ/tDMnjWzD8zsRTNrGvwqeZ+ZvSHpOjO7wMxeM7MVwWe/e/D8BDN7JRgfq0rHgpnl\nB/9tb2bLzGylma0xs0uD9Sd/9TSzScG2NWZ2d5m61pvZU2a21swWm1mTqLxJQBXKf37N7GEzu6PM\n9vvNbHLw+MdmtjwYaz8P1pX/W9TZzH5vZjnBZ//n0XllAKoD4aDu6SbpOUlXSBonaai795WUI2lS\ncMO5OZLucvd0SUMlHZW0W9IVQdsbJE2LRvFANekmaaa795J0UFLpF5tj7j7I3V9QyV3Vf+DuGZJ+\nJGl60GaapH8E46OvpLXl+v6WpEXu3ltSuqSVZTeaWYakWyT1l3SxpO+aWZ9gc1dJT7h7qqQDkkZX\n1wsGqkNFn19JL6jk70Kp6yXNM7MrVfKZ7iept6QMM/t60KabpOfcvY+7b5X0H8Fdk3tJGmxmvWrk\nBQGodkwrqnu2uvs7ZnaVpBRJb5Yc1VVDSW+r5H/Yn7r7ckly94OSZGbNJP2XmfWWVCQpORrFA9Vk\nu7u/GTx+XtKdweM5khQcRRugki84pc9pFPz3MknfkaRg6lFeub6XS3razBpImu/uK8ttHyTpFXc/\nHOzrZUmXquSO7x+Xab9CUuJXeI1AGCr7/LY1sw6S4iXtd/dtZnanpCslvR88t7lKwsI2BX+LyvR7\nvZlNUMn3ivYq+fv0QU28IADVi3BQ9xwO/muSlrj72LIbg19rKrp5xQ8lfaaSX0JjJB0Ls0ggZOU/\n46XLpeMjRtKB4Nf/M+vYfVnw6+g3VHIH9/909+fKNLFKnipJx8s8LpLEtCLUNpV9fl+UdK2kdio5\nklDa9iF3n/GFDswS9f9jTWaWpJKjcxe5+34ze0ZS42qtGkCNYVpR3fWOpIFmdqEkBXOukyV9KKmD\nmV0UrG9hZnGSWqrkiEKxpJskxUapbqA6nG9mlwSPx0p6o+zG4IjZx2Z2nVRy0qSZpQeb/y7p9mB9\nrJmdU/a5ZtZF0m53f0rSf6tk6lFZyyRdE4y5ZpK+Ken16ntpQKgq+/y+IGmMSgLCi0HbRZJuLXM+\nW0cza1tBn+eoJCzkmVmCpBEhvwYAISIc1FHuvkfSzZJmm9kHKgkL3d39hErmjj5uZqskLVHJLzjT\nJf2rmb2jkilFhyvsGKgb1qvk8/yBpFaSfl9Bm29LGheMg7WSRgXr75I0xMxWq2TqT2q552VJWmlm\n76vknIHfld3o7u9JekbSu5L+KekP7v6+gDqgss+vu6+V1ELSJ+7+adB2saQ/SXo7GC8vBm3K97lK\nJVOP1kp6WtKb5dsAqDvMvaIZKABQOwVTGv7q7mlRLgUAgHqHIwcAAAAAJHHkAAAAAECAIwcAAAAA\nJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACRJ/wdXbH5ZJMleSwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f429b72d9b0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:1:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>179980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>11638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>37921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>32234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>199840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>22286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>9298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>plasmid_7</td>\n",
       "      <td>170832</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0       179980\n",
       "1  plasmid_1        11638\n",
       "2  plasmid_2        37921\n",
       "3  plasmid_3        32234\n",
       "4  plasmid_4       199840\n",
       "5  plasmid_5        22286\n",
       "6  plasmid_6         9298\n",
       "7  plasmid_7       170832"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.365968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.228919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.281657</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.365968\n",
       "1  precision  0.228919\n",
       "2   f1_score  0.281657"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP017929.1</th>\n",
       "      <th>CP017930.1</th>\n",
       "      <th>CP017931.1</th>\n",
       "      <th>CP017932.1</th>\n",
       "      <th>CP017933.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.031015</td>\n",
       "      <td>0.059296</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.931588</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.026400</td>\n",
       "      <td>0.054530</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.033721</td>\n",
       "      <td>0.293563</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.024020</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.026400</td>\n",
       "      <td>0.118846</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076398</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.020082</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.058078</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.013218</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>0.082207</td>\n",
       "      <td>0.037849</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.045875</td>\n",
       "      <td>0.256058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>plasmid_7</td>\n",
       "      <td>0.035816</td>\n",
       "      <td>0.099975</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.066797</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.235560</td>\n",
       "      <td>0.798535</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.101580</td>\n",
       "      <td>2.256058</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP017929.1  CP017930.1  CP017931.1  CP017932.1  CP017933.1\n",
       "0  plasmid_0    0.031015    0.059296         0.0    0.931588    0.000000\n",
       "1  plasmid_1    0.026400    0.054530         0.0    0.000000    1.000000\n",
       "2  plasmid_2    0.033721    0.293563         0.0    0.024020    0.000000\n",
       "3  plasmid_3    0.026400    0.118846         0.0    0.000000    1.000000\n",
       "4  plasmid_4    0.000000    0.076398         0.0    0.020082    0.000000\n",
       "5  plasmid_5    0.000000    0.058078         0.0    0.013218    0.000000\n",
       "6  plasmid_6    0.082207    0.037849         0.0    0.045875    0.256058\n",
       "7  plasmid_7    0.035816    0.099975         0.0    0.066797    0.000000\n",
       "8      Total    0.235560    0.798535         0.0    1.101580    2.256058"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP017929.1</td>\n",
       "      <td>0.019491</td>\n",
       "      <td>0.256058</td>\n",
       "      <td>0.100498</td>\n",
       "      <td>0.092449</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.023713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP017930.1</td>\n",
       "      <td>0.019986</td>\n",
       "      <td>0.534800</td>\n",
       "      <td>0.771578</td>\n",
       "      <td>0.354284</td>\n",
       "      <td>0.019706</td>\n",
       "      <td>0.050570</td>\n",
       "      <td>0.464724</td>\n",
       "      <td>0.036504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP017931.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP017932.1</td>\n",
       "      <td>0.394344</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.045648</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007656</td>\n",
       "      <td>0.045185</td>\n",
       "      <td>0.375887</td>\n",
       "      <td>0.029790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP017933.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.361296</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.321144</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.433820</td>\n",
       "      <td>1.790858</td>\n",
       "      <td>0.917724</td>\n",
       "      <td>0.808029</td>\n",
       "      <td>0.027362</td>\n",
       "      <td>0.095755</td>\n",
       "      <td>2.161755</td>\n",
       "      <td>0.090007</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2  plasmid_3  plasmid_4  \\\n",
       "0  CP017929.1   0.019491   0.256058   0.100498   0.092449   0.000000   \n",
       "1  CP017930.1   0.019986   0.534800   0.771578   0.354284   0.019706   \n",
       "2  CP017931.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "3  CP017932.1   0.394344   0.000000   0.045648   0.000000   0.007656   \n",
       "4  CP017933.1   0.000000   1.000000   0.000000   0.361296   0.000000   \n",
       "5       Total   0.433820   1.790858   0.917724   0.808029   0.027362   \n",
       "\n",
       "   plasmid_5  plasmid_6  plasmid_7  \n",
       "0   0.000000   1.000000   0.023713  \n",
       "1   0.050570   0.464724   0.036504  \n",
       "2   0.000000   0.000000   0.000000  \n",
       "3   0.045185   0.375887   0.029790  \n",
       "4   0.000000   0.321144   0.000000  \n",
       "5   0.095755   2.161755   0.090007  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.394399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.771578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.522244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.019706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>0.050570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>plasmid_7</td>\n",
       "      <td>0.052279</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.394399\n",
       "1  plasmid_1    1.000000\n",
       "2  plasmid_2    0.771578\n",
       "3  plasmid_3    0.522244\n",
       "4  plasmid_4    0.019706\n",
       "5  plasmid_5    0.050570\n",
       "6  plasmid_6    1.000000\n",
       "7  plasmid_7    0.052279"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP017929.1</td>\n",
       "      <td>0.116043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP017930.1</td>\n",
       "      <td>0.521338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP017931.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP017932.1</td>\n",
       "      <td>0.942483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP017933.1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP017929.1    0.116043\n",
       "1  CP017930.1    0.521338\n",
       "2  CP017931.1    0.000000\n",
       "3  CP017932.1    0.942483\n",
       "4  CP017933.1    1.000000"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP017933.1</td>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction  recall  precision\n",
       "0  CP017933.1  plasmid_1     1.0        1.0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 1, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00725563</td>\n",
       "      <td>0.0167982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[19, 211, 107, 129, 91, 62, 86, 71, 106, 65, 8...</td>\n",
       "      <td>[53, 49]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.430567</td>\n",
       "      <td>-0.37485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.437823</td>\n",
       "      <td>-0.358051</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0  plasmid_1\n",
       "GC_sum                                        -0.00725563  0.0167982\n",
       "chain   [19, 211, 107, 129, 91, 62, 86, 71, 106, 65, 8...   [53, 49]\n",
       "gd_sum                                          -0.430567   -0.37485\n",
       "total                                           -0.437823  -0.358051"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_11</th>\n",
       "      <th>plasmid_12</th>\n",
       "      <th>plasmid_13</th>\n",
       "      <th>plasmid_14</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0104156</td>\n",
       "      <td>0.00650715</td>\n",
       "      <td>0.0167982</td>\n",
       "      <td>-0.0126396</td>\n",
       "      <td>-0.0447193</td>\n",
       "      <td>-0.0102248</td>\n",
       "      <td>-0.00338337</td>\n",
       "      <td>-0.00416335</td>\n",
       "      <td>-0.00894023</td>\n",
       "      <td>-0.0419224</td>\n",
       "      <td>-0.0371617</td>\n",
       "      <td>-0.029292</td>\n",
       "      <td>-0.0309369</td>\n",
       "      <td>-0.0152317</td>\n",
       "      <td>-0.0222204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[28, 155, 216, 143, 130, 151, 186, 167, 104, 1...</td>\n",
       "      <td>[42, 165, 140, 125, 98, 49, 63, 72, 8, 107, 11...</td>\n",
       "      <td>[49, 53]</td>\n",
       "      <td>[53]</td>\n",
       "      <td>[107, 129]</td>\n",
       "      <td>[31, 109, 72, 30, 64]</td>\n",
       "      <td>[96, 40, 72, 13, 148]</td>\n",
       "      <td>[168, 66, 106, 65, 19, 211, 107, 116, 54, 65, ...</td>\n",
       "      <td>[68, 48, 72, 63, 49, 98, 125, 174, 125]</td>\n",
       "      <td>[62]</td>\n",
       "      <td>[168, 187, 151, 145, 150, 143, 216, 155, 194, 41]</td>\n",
       "      <td>[16]</td>\n",
       "      <td>[205, 143, 216, 155, 195, 55, 194, 155]</td>\n",
       "      <td>[60]</td>\n",
       "      <td>[67, 116, 107]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.423212</td>\n",
       "      <td>-0.0568255</td>\n",
       "      <td>-0.37485</td>\n",
       "      <td>-0.275212</td>\n",
       "      <td>-0.351494</td>\n",
       "      <td>-0.0088837</td>\n",
       "      <td>-0.00442442</td>\n",
       "      <td>-0.43653</td>\n",
       "      <td>-0.40753</td>\n",
       "      <td>-0.42156</td>\n",
       "      <td>-0.406304</td>\n",
       "      <td>-0.434423</td>\n",
       "      <td>-0.355864</td>\n",
       "      <td>-0.414011</td>\n",
       "      <td>-0.385428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.433627</td>\n",
       "      <td>-0.0503184</td>\n",
       "      <td>-0.358051</td>\n",
       "      <td>-0.287852</td>\n",
       "      <td>-0.396213</td>\n",
       "      <td>-0.0191085</td>\n",
       "      <td>-0.00780779</td>\n",
       "      <td>-0.440693</td>\n",
       "      <td>-0.41647</td>\n",
       "      <td>-0.463482</td>\n",
       "      <td>-0.443466</td>\n",
       "      <td>-0.463715</td>\n",
       "      <td>-0.386801</td>\n",
       "      <td>-0.429243</td>\n",
       "      <td>-0.407649</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0  \\\n",
       "GC_sum                                         -0.0104156   \n",
       "chain   [28, 155, 216, 143, 130, 151, 186, 167, 104, 1...   \n",
       "gd_sum                                          -0.423212   \n",
       "total                                           -0.433627   \n",
       "\n",
       "                                                plasmid_1 plasmid_10  \\\n",
       "GC_sum                                         0.00650715  0.0167982   \n",
       "chain   [42, 165, 140, 125, 98, 49, 63, 72, 8, 107, 11...   [49, 53]   \n",
       "gd_sum                                         -0.0568255   -0.37485   \n",
       "total                                          -0.0503184  -0.358051   \n",
       "\n",
       "       plasmid_11  plasmid_12             plasmid_13             plasmid_14  \\\n",
       "GC_sum -0.0126396  -0.0447193             -0.0102248            -0.00338337   \n",
       "chain        [53]  [107, 129]  [31, 109, 72, 30, 64]  [96, 40, 72, 13, 148]   \n",
       "gd_sum  -0.275212   -0.351494             -0.0088837            -0.00442442   \n",
       "total   -0.287852   -0.396213             -0.0191085            -0.00780779   \n",
       "\n",
       "                                                plasmid_2  \\\n",
       "GC_sum                                        -0.00416335   \n",
       "chain   [168, 66, 106, 65, 19, 211, 107, 116, 54, 65, ...   \n",
       "gd_sum                                           -0.43653   \n",
       "total                                           -0.440693   \n",
       "\n",
       "                                      plasmid_3  plasmid_4  \\\n",
       "GC_sum                              -0.00894023 -0.0419224   \n",
       "chain   [68, 48, 72, 63, 49, 98, 125, 174, 125]       [62]   \n",
       "gd_sum                                 -0.40753   -0.42156   \n",
       "total                                  -0.41647  -0.463482   \n",
       "\n",
       "                                                plasmid_5 plasmid_6  \\\n",
       "GC_sum                                         -0.0371617 -0.029292   \n",
       "chain   [168, 187, 151, 145, 150, 143, 216, 155, 194, 41]      [16]   \n",
       "gd_sum                                          -0.406304 -0.434423   \n",
       "total                                           -0.443466 -0.463715   \n",
       "\n",
       "                                      plasmid_7  plasmid_8       plasmid_9  \n",
       "GC_sum                               -0.0309369 -0.0152317      -0.0222204  \n",
       "chain   [205, 143, 216, 155, 195, 55, 194, 155]       [60]  [67, 116, 107]  \n",
       "gd_sum                                -0.355864  -0.414011       -0.385428  \n",
       "total                                 -0.386801  -0.429243       -0.407649  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_4  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_1  from MILP completely contained in  plasmid_10  from greedy.\n",
      "\n",
      "plasmid_10  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_11  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_12  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
