{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 109 with $\\alpha_1 = 1$, $\\alpha_2 = 5$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 109\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012925.1</td>\n",
       "      <td>2096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012927.1</td>\n",
       "      <td>3372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012928.1</td>\n",
       "      <td>62920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012929.1</td>\n",
       "      <td>96042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012926.1</td>\n",
       "      <td>37697</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP012925.1         2096\n",
       "1  CP012927.1         3372\n",
       "2  CP012928.1        62920\n",
       "3  CP012929.1        96042\n",
       "4  CP012926.1        37697"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt0VeWd//HPNwl3MAqEcJWkSgJJ\nIEAiaEAJigJTBbvwAq22KpRWy6iFVpmZ1nasrdWBOoMjFey4lEIR8EKpMgLtkOK1EhTkGkAkgoiA\nQCBcc/n+/sgOvxiScNDshMv7tRbLs/d+zrO/56zzmPM5+9l7m7sLAAAAAKLquwAAAAAAZwbCAQAA\nAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAgZj6LuB0\ntW7d2hMSEuq7jLPGoUOH1KxZs/ouAzjjMVaAyDBWTs+KFSv2uHtcfdcBROqsCwcJCQnKzc2t7zLO\nGjk5OcrOzq7vMoAzHmMFiAxj5fSYWX591wCcDqYVAQAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAg\nHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQCC0cmNmzZrbLzNZUs93MbIqZbTazD82sd1i1AAAA\nADi1MI8cPCdpSA3bh0rqEvwbK+n3IdYCAAAA4BRCCwfuvkzS3hqaDJc0w8u8K+lCM2sXVj0AAAAA\nahZTj/vuIGlbheXtwbrPKjc0s7EqO7qg+Ph45eTk1EV954TCwkLeL4Rm9acF9V1CrUmMjWasABHg\n7wpwbqvPcGBVrPOqGrr7dEnTJSkzM9Ozs7NDLOvckpOTI94vhOWOia/Vdwm15rkhzRgrQAT4uwKc\n2+ozHGyX1KnCckdJO+qpFgAAQpNwjgVpAOeu+ryU6QJJ3w2uWnS5pAJ3P2lKEQAAAIC6EdqRAzOb\nLSlbUmsz2y7pF5IaSJK7Py1poaR/krRZ0mFJd4ZVy+niFx4AAACcj0ILB+4+6hTbXdKPwto/AAAA\ngNPDHZIBAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAA\nAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAA\nCBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAI\nEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQ\nDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAO\nAAAAAEgiHAAAAAAIhBoOzGyImeWZ2WYzm1jF9ovNbKmZfWBmH5rZP4VZDwAAAIDqhRYOzCxa0lOS\nhkpKkTTKzFIqNfuZpLnu3kvSSElTw6oHAAAAQM3CPHLQR9Jmd9/i7sclvSBpeKU2LumC4HGspB0h\n1gMAAACgBjEh9t1B0rYKy9sl9a3U5peSFpvZP0tqJmlQiPUAAAAAqEGY4cCqWOeVlkdJes7dJ5vZ\nFZL+aGZp7l76pY7MxkoaK0nx8fHKyckJo94TJnQvDrX/ulRYWBj6+4XzF2MFiAxjBcDZIsxwsF1S\npwrLHXXytKHRkoZIkru/Y2aNJbWWtKtiI3efLmm6JGVmZnp2dnZIJZe5Y+JrofZfl54b0kxhv184\nfzFWgMgwVgCcLcI852C5pC5mlmhmDVV2wvGCSm0+kXSNJJlZN0mNJe0OsSYAAAAA1QgtHLh7saRx\nkhZJWq+yqxKtNbOHzWxY0GyCpO+b2SpJsyXd4e6Vpx4BAAAAqANhTiuSuy+UtLDSuocqPF4nqV+Y\nNQAAAACIDHdIBgAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJ\ncAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlw\nAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAA\nAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAA\nAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAA\nACBAOAAAAAAgiXAAAAAAIEA4AAAAACAp5HBgZkPMLM/MNpvZxGra3GJm68xsrZn9Kcx6AAAAAFQv\nJqyOzSxa0lOSrpW0XdJyM1vg7usqtOki6V8k9XP3fWbWJqx6AAAAANQszCMHfSRtdvct7n5c0guS\nhldq831JT7n7Pkly910h1gMAAACgBqEdOZDUQdK2CsvbJfWt1CZJkszsLUnRkn7p7q9X7sjMxkoa\nK0nx8fHKyckJo94TJnQvDrX/ulRYWBj6+4XzF2MFiAxjBcDZIsxwYFWs8yr230VStqSOkt4wszR3\n3/+lJ7lPlzRdkjIzMz07O7vWi63ojomvhdp/XXpuSDOF/X7h/MVYASLDWAFwtghzWtF2SZ0qLHeU\ntKOKNn929yJ3/1hSnsrCAgAAAIA6FmY4WC6pi5klmllDSSMlLajUZr6kgZJkZq1VNs1oS4g1AQAA\nAKhGaOHA3YsljZO0SNJ6SXPdfa2ZPWxmw4JmiyR9YWbrJC2V9FN3/yKsmgAAAABUL8xzDuTuCyUt\nrLTuoQqPXdL44B8AAACAesQdkgEAAABIIhwAAAAACBAOAAAAAEg6jXBgZv3N7M7gcZyZJYZXFgAA\nAIC6FlE4MLNfSHpQ0r8EqxpImhlWUQAAAADqXqRHDr4laZikQ5Lk7jsktQirKAAAAAB1L9JwcDy4\n7KhLkpk1C68kAAAAAPUh0nAw18ymSbrQzL4v6a+SngmvLAAAAAB1LaKboLn7JDO7VtIBScmSHnL3\nJaFWBgAAAKBOnTIcmFm0pEXuPkgSgQAAAAA4R51yWpG7l0g6bGaxdVAPAAAAgHoS0bQiSUclrTaz\nJQquWCRJ7n5vKFUBAAAAqHORhoPXgn8AAAAAzlGRnpD8vJk1lJQUrMpz96LwygIAAABQ1yIKB2aW\nLel5SVslmaROZvY9d18WXmkAAAAA6lKk04omS7rO3fMkycySJM2WlBFWYQAAAADqVqQ3QWtQHgwk\nyd03SmoQTkkAAAAA6kOkRw5yzex/JP0xWP6OpBXhlAQAAACgPkQaDu6W9CNJ96rsnINlkqaGVRQA\nAACAuhdpOIiR9F/u/jvpxF2TG4VWFQAAwDlqxYoVbWJiYv4gKU2RT/EGakOppDXFxcVjMjIydlXV\nINJw8DdJgyQVBstNJC2WlPW1SwQAADiPxMTE/KFt27bd4uLi9kVFRXl914PzR2lpqe3evTtl586d\nf5A0rKo2kabVxu5eHgwUPG5aCzUCAACcb9Li4uIOEAxQ16KiojwuLq5AZUetqm4TYV+HzKx3+YKZ\nZUo68jXrAwAAOB9FEQxQX4LPXrUZINJpRfdJmmdmOyS5pPaSbv365QEAAAA4U0QaDhIl9ZJ0saRv\nSbpcZSEBAAAAX0PCxNdq9aayW3/7zTPmcvN9+vRJnjRp0rarrrrqcCTtS0tLlZWVlbRw4cLNLVu2\nLA27vq/r1VdfbTF58uT4pUuXbg6j/1mzZsWuXbu2yW9+85udlbc1bdq01+HDhz+o7rlZWVlJf/7z\nnz+Ki4srOZ19Rjqt6OfufkDShZKulTRd0u9PZ0cAAAA4exQVFdX5PufOnRubmpp65GwIBnXhO9/5\nTkFVwSASo0aN+mLSpElxp/u8SMNBeeL4pqSn3f3Pkhqe7s4AAABQ/37605+2S0xMTM3Kyupyww03\nJD700EPxUtkv/ePGjetw2WWXJT/yyCPxO3bsiBk8ePAlaWlp3dLS0rotXry4mSQdOHAg6uabb05I\nS0vr1q1bt5SZM2deKEmFhYV2/fXXfyMpKSnlm9/85jeOHj1qkvTEE0+0Hj16dKfy/U+ePLn1mDFj\nOlaua9asWS2/9a1v7ZekvLy8homJiam33npr5y5duqQOGzYscf78+S169+7dtXPnzmlLly5tWlMt\neXl5DTMyMpJTUlK6paSkdFuyZEkzqezX/j59+iQPGTLkG4mJianDhg1LLC2tOYuMGDEi4dvf/vbF\nGRkZyQkJCWmzZ8+Ordxm6dKlTXv16tW1W7duKb169eq6atWqRpKUm5vbuHv37t26du2akpSUlLJ6\n9epGkb62KVOmtPrud797sSRt2LChYc+ePbumpaV1u++++9qX7zc/P79BZmZmcteuXVO6dOmS+vrr\nrzeXpJEjR+5/+eWXW0X4kTgh0mlFn5rZNJVdzvQxM2skrssLAABw1lm2bFnTv/zlLxetXr16XVFR\nkfXs2TOlV69eJ6b97N+/P3r58uV5knTDDTckjh8//vPBgwcXbtq0qeHgwYO7bNmyZe2//uu/ths4\ncOCBefPmbd2zZ090ZmZmt2HDhh343e9+F9ekSZPSjRs3rvvHP/7RpF+/fimSNHr06L2pqakpx44d\n296oUSOfOXNm62nTpuVXrm3FihXN+/Xrd2L9tm3bGs+ZM2dLRkZGfo8ePbrNmjWrVW5u7oY//elP\nF/76179uN3DgwI+qq6V9+/bFb7zxxsamTZv66tWrG40aNeoba9asWS9J69evb7Jy5cotCQkJRRkZ\nGV2XLFnSfPDgwYWV66lo27Ztjd577728devWNRo0aFDy8OHDV1fcnp6efvS9997b0KBBA82fP7/F\nAw880HHRokUfPfnkk3H33HPP53fffffeo0ePWnFxsT799NMGkby2iv3fc889F48ZM2b3uHHjvnj0\n0UdPHBF49tlnW15zzTUFjz322M7i4mIdPHgwSpLi4uJKjh8/bjt37oxu27ZtxFOLIg0Ht0gaImmS\nu+83s3aSfhrpTgAAAHBmyMnJaT506ND9zZs3d0l+7bXX7q+4fdSoUXvLH7/11lsXbNq0qUn5cmFh\nYfS+ffuicnJyLli0aNGFU6ZMaStJx44ds82bNzd88803m9977727JKlv375HkpKSDkvSBRdcUNqv\nX7+Dc+bMie3evfvRoqIi69Onz0lXviwoKIi56KKLTvyM36FDh2Pl7ZKSko5cffXVB6KiotS7d+/D\njzzySPvg9VRZS+fOnYtGjx7ded26dU2ioqKUn59/4ga+3bt3P3TJJZcUSVJqaurhjz766JQzYkaM\nGLE3Ojpa3bt3P9apU6djK1eubFxx+969e6NvvfXWxK1btzY2My8qKjJJuuKKKw5NmjSp3fbt2xuO\nHDlyX/fu3Y9F+toqev/995v/7//+70eS9IMf/OCLX/3qVx0l6fLLLz/0gx/8IKGoqCjqpptu2peV\nlXXifW3VqlXxJ5980rBt27YRX2U0onDg7oclvVxh+TNJn0W6EwAAAJwZ3Gu+pkyLFi1OfDl3d+Xm\n5q4PgsSX+njxxRc3p6enH6v8fDOrst+xY8fu+fWvf902KSnp6G233banqjbR0dFeUlKi6OhoSVLD\nhg1P7DcqKkqNGzf2oJ1KSkqsplrGjx/fvk2bNkUvvfTSx6WlpWrSpMmJE78bNWp0ot/o6GgVFxdX\nXXQNr6vy8oMPPthhwIABB5csWfJRXl5ew6uvvjpZkn74wx/uvfLKKw+98sorsUOHDk2aOnXq1uTk\n5GORvLbKqroE7tChQwuXLVuW99JLL8Xecccdiffee+/n48aN+0IqC0pNmzY9rfM3mBoEAABwHsnO\nzi5ctGhR7OHDh62goCDqr3/964XVte3fv/+Bxx57rE358ttvv91EkgYOHHhg8uTJ8eVz9d96660m\nQfvCmTNntpSk5cuXN964ceOJm+ZeffXVhz777LOGr7zySqvRo0fvVRUSExOPrl+/vlFV26pTXS0F\nBQXR7dq1K4qOjtbUqVNblZScembNj370ow4zZsyo8v14+eWXLyopKdHatWsbbdu2rVF6evrRitsP\nHDgQ3bFjx+OSNG3atNbl69etW9ewW7dux372s5/tuu666/avXLmySeW+I9G7d+/CZ555pqUkPfPM\nMyfOJdi4cWPDDh06FE2YMGHPbbfdtuf9999vKpVd+Wn37t0NkpOTTwpwNYl0WhEAAABCUNeXHh0w\nYMDhIUOGFKSkpKR26NDhWI8ePQ7FxsZW+c15+vTp28aMGXNxUlJSSklJifXt2/dgVlbWJ7/97W93\njB079uKuXbumuLt17Njx2NKlSzf/5Cc/2TVy5MjEpKSklNTU1MPdu3c/VLG/G2+8cd+HH37YtLrL\na1533XUFixcvbpGWlhbxF9rqarn//vt3jRgx4pL58+df1L9//4NNmjQ55S/o69ata1J+QnRll156\n6bE+ffokf/HFFw3+8z//M79p06Zf+hX/wQcf3DlmzJjEKVOmtL3yyisPlK//4x//2HLevHmtYmJi\nPC4urujRRx/dsX///uhIX1+5qVOnfjJy5MhvTJ06NX7YsGH7ytcvWrSoxZQpU9rGxMR406ZNS2bN\nmvWxJL355ptNe/XqdahBgwantR871aGlM01mZqbn5uaGuo+Eia+F2n9dem5IM2VnZ9d3GThHMVaA\nyDBWzl9mtsLdMyuuW7Vq1db09PQqp9XUlYKCgqjY2NjSgwcPRl1xxRXJTz/9dH7//v0juhfB1zFw\n4MBL77///s+HDx9+sKrt+fn5DUaNGpXw9ttvbwq7lqr079+/y5tvvnnSvkeMGJFw/fXXF9x55537\nqnremejOO+/sdOONN+6v6r1etWpV6/T09ISqnse0IgAAgPPMbbfd1rlr164pPXr06HbDDTfsCzsY\n7NmzJzohISGtcePGpdUFA0nq3Llz0V133bVn79699fIdtapgcLZKS0s7UtN7XR2mFQEAAJxn/vKX\nv3xcl/tr3bp1ydatW9dE0nbMmDFn3K/zL7300tb6ruF0TZgw4SsdneLIAQAAAABJhAMAAAAAAcIB\nAAAAAEmEAwAAAAABTkgGAACoT7+MzTh1o9Ppr+CU900ws4zhw4fvnT9//seSVFRUpDZt2qT37Nnz\n0NKlSzdPmTKlVW5ubrMZM2Z8Mn78+PbNmzcvefjhhz+v2Ed0dHRGly5djpSUlNill156ZO7cuVsr\n3l1ZKrsRV1ZWVtLChQs3t2zZsvTmm29O+Nvf/hbbqlWr4k2bNq2tqrZXX321xahRoy7p0KHDcUm6\n/vrr902aNOmzyu1+85vfxD399NPx27Zta7Rjx45V7dq1K5ak2bNnx7733nvNnnjiiR2Rv2kox5ED\nAACA80yTJk1K8/LymhQWFpokvfLKKxfEx8cXnU4fjRo1Kt2wYcO6TZs2rW3QoIFPnjw5rnKbuXPn\nxqamph5p2bJlqSTdddddexYsWHDKy4VmZmYWbtiwYd2GDRvWVRUMJGnAgAGFS5Ys2di+ffvjFdff\neuutBa+//vqFBw8e5HvuV8CbBgAAcB665pprCubNm3ehJM2ePbvliBEj9n7Vvvr371+4efPmRpXX\nz5o1q2XFOw4PHTq0MC4urvir7qeifv36HUlOTj5eeX1UVJSysrIOzpkzJ7Y29nO+IRwAAACch26/\n/fa9c+bMuejw4cO2fv36pldcccWhr9JPUVGRFi1adEH37t2PVN62YsWK5v369Ttlv48//njc448/\nfuLIwwcffNA8OTk55aqrruqSm5vbuHz9gAEDLt26dWuDU/WXmZl56I033mh+Oq8DZUI958DMhkj6\nL0nRkv7g7r+tpt1NkuZJuszdc8OsCQAAAFLfvn2PbN++vdEzzzzTctCgQQWn+/xjx45Fde3aNSXo\n6+B999130k23CgoKYi666KLSk5/9ZQ888MDu8sdZWVmH8vPzP4yNjS2dM2dO7IgRIy7Nz89fI0l/\n//vfN0dSW9u2bYt37tzZMPJXg3KhhQMzi5b0lKRrJW2XtNzMFrj7ukrtWki6V9I/wqoFAAAAJxsy\nZMj+X/ziF50WL16ct2vXrtP6Xlh+zkFNbaKjo72kpETR0dER91t+foJUdv7A+PHjL/7ss89iyk84\njsSRI0escePGpwwlOFmY04r6SNrs7lvc/bikFyQNr6LdryQ9LuloiLUAAACgkrvvvnvPhAkTdvTp\n0+ekKUG1ITEx8ej69etPOhehJp988klMaWnZ9/qlS5c2LS0tVXx8/Gmdp5CXl9c4NTU1lNd0rgtz\nWlEHSdsqLG+X1LdiAzPrJamTu79qZj+priMzGytprCTFx8crJyen9qutYEL3WjlP5oxQWFgY+vuF\n8xdjBYgMYwU1iuDSo2G55JJLin7+85/vOlW7J554ot20adPiy5c///zzDyPp/7rrritYvHhxi7S0\ntGOSdMMNNyS+++67Lfbt2xcTHx/fY+LEiTt+/OMf7yk/3+CBBx7YPXPmzIueffbZNtHR0d64cePS\nGTNmbImKKvs9e8CAAZc+//zz+QkJCUWPPPJImyeffLLtF1980SA9PT1l4MCBBXPmzMmXpGXLlrV4\n7LHHPv0Kb8l5z9w9nI7NbpY02N3HBMu3S+rj7v8cLEdJ+j9Jd7j7VjPLkfSTU51zkJmZ6bm54Z6W\nkDDxtVD7r0vPDWmm7Ozs+i4D5yjGChAZxsr5y8xWuHtmxXWrVq3amp6eftL8/HNRfn5+g1GjRiW8\n/fbbp7x8aW3Ztm1bzC233PKNd955Z2Nd7fNss2rVqtbp6ekJVW0Lc1rRdkmdKix3lFTxZhQtJKVJ\nyjGzrZIul7TAzL40gAAAAHB26ty5c9Fdd921Z+/evXV2hcwtW7Y0nDx58rZTt0RVwpxWtFxSFzNL\nlPSppJGSvl2+0d0LJLUuX470yAEAhOKzldIvqzot6iz0y9O+6AgAhGbMmDH76nJ/AwYMOFyX+zvX\nhJbi3L1Y0jhJiyStlzTX3dea2cNmNiys/QIAAJzhSktLS62+i8D5KfjsVXslp1Dvc+DuCyUtrLTu\noWraZodZCwAAqAUcZasNa3bv3p0SFxdXEBUVFc7Jn0AVSktLbffu3bGS1lTXJtRwAAAAgC8rLi4e\ns3Pnzj/s3LkzTeGe/wlUVippTXFx8ZjqGhAOznX8wgMAwBklIyNjlySmWOOMRFoFAAAAIIlwAAAA\nACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAA\nIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAg\nQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBA\nOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4\nAAAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACAp5HBgZkPMLM/M\nNpvZxCq2jzezdWb2oZn9zcw6h1kPAAAAgOqFFg7MLFrSU5KGSkqRNMrMUio1+0BSprv3kPSipMfD\nqgcAAABAzcI8ctBH0mZ33+LuxyW9IGl4xQbuvtTdDweL70rqGGI9AAAAAGpg7h5Ox2Y3SRri7mOC\n5dsl9XX3cdW0/29JO939kSq2jZU0VpLi4+MzXnjhhVBqLrf604JQ+69LiU2OqfmxHfVdRu1o17O+\nK0AljJUzFGPljMNYOUPVwVgZOHDgCnfPDH1HQC2JCbFvq2JdlUnEzG6TlClpQFXb3X26pOmSlJmZ\n6dnZ2bVUYtXumPhaqP3XpefSP1Z23i/qu4zaMerc+eN6rmCsnKEYK2ccxsoZirECnCTMcLBdUqcK\nyx0lnfRTg5kNkvRvkga4+7EQ6wEAAABQgzDPOVguqYuZJZpZQ0kjJS2o2MDMekmaJmmYu+8KsRYA\nAAAApxBaOHD3YknjJC2StF7SXHdfa2YPm9mwoNl/SGouaZ6ZrTSzBdV0BwAAACBkYU4rkrsvlLSw\n0rqHKjweFOb+AQAAAESOOyQDAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQ\nIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAg\nHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAc\nAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwA\nAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAA\nAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQCDUcmNkQM8szs81mNrGK7Y3MbE6w/R9mlhBmPQAAAACq\nF1o4MLNoSU9JGiopRdIoM0up1Gy0pH3ufqmkJyQ9FlY9AAAAAGoW5pGDPpI2u/sWdz8u6QVJwyu1\nGS7p+eDxi5KuMTMLsSYAAAAA1QgzHHSQtK3C8vZgXZVt3L1YUoGkViHWBAAAAKAaMSH2XdURAP8K\nbWRmYyWNDRYLzSzva9Z23hgotZa0p77rqBX/zkElhIexAkSGsXLaOtfFToDaEmY42C6pU4XljpJ2\nVNNmu5nFSIqVtLdyR+4+XdL0kOo8p5lZrrtn1ncdwJmOsQJEhrECnNvCnFa0XFIXM0s0s4aSRkpa\nUKnNAknfCx7fJOn/3P2kIwcAAAAAwhfakQN3LzazcZIWSYqW9Ky7rzWzhyXluvsCSf8j6Y9mtlll\nRwxGhlUPAAAAgJoZP9Sf28xsbDAtC0ANGCtAZBgrwLmNcAAAAABAUsh3SAYAAABw9iAcAAAAAJBE\nODjvmVmCma0JHmeb2av1XRNO28WKAAAEoklEQVRQH8zs7VNsX2hmF9ZVPcC5wMxyzCwzeLzVzFrX\nd00AahbmfQ4QIjMzlZ0zUlrftQBnGjOLdveS03mOu2edYvs/fb2qgHMPf4uAcw9HDs4iwa/8681s\nqqT3Jd1uZu+Y2ftmNs/MmgftLjOzt81slZm9Z2Ytgue+EbR938xq/CIEnKmCz/IGM3vezD40sxfN\nrGnwq+RDZvampJvN7BIze93MVgSf/a7B8+PN7JVgfKwqHwtmVhj8t52ZLTOzlWa2xsyuDNaf+NXT\nzMYH29aY2f0V6lpvZs+Y2VozW2xmTerlTQJqUPnza2aPmdk9Fbb/0swmBI9/ambLg7H278G6yn+L\nOpnZ780sN/js/3v9vDIAtYFwcPZJljRD0rWSRksa5O69JeVKGh/ccG6OpPvcPV3SIElHJO2SdG3Q\n9lZJU+qjeKCWJEua7u49JB2QVP7F5qi793f3F1R2V/V/dvcMST+RNDVoM0XS34Px0VvS2kp9f1vS\nInfvKSld0sqKG80sQ9KdkvpKulzS982sV7C5i6Sn3D1V0n5JI2rrBQO1oarPr6QXVPZ3odwtkuaZ\n2XUq+0z3kdRTUoaZXRW0SZY0w917uXu+pH8L7prcQ9IAM+tRJy8IQK1jWtHZJ9/d3zWz6yWlSHqr\n7KiuGkp6R2X/w/7M3ZdLkrsfkCQzaybpv82sp6QSSUn1UTxQS7a5+1vB45mS7g0ez5Gk4Chalsq+\n4JQ/p1Hw36slfVeSgqlHBZX6Xi7pWTNrIGm+u6+stL2/pFfc/VCwr5clXamyO75/XKH9CkkJX+M1\nAmGo7vPbxszaS4qTtM/dPzGzeyVdJ+mD4LnNVRYWPlHwt6hCv7eY2ViVfa9op7K/Tx/WxQsCULsI\nB2efQ8F/TdISdx9VcWPwa01VN6/4saTPVfZLaJSko2EWCYSs8me8fLl8fERJ2h/8+n96HbsvC34d\n/abK7uD+H+4+o0ITq+apknSswuMSSUwrwpmmus/vi5JuktRWZUcSyts+6u7TvtSBWYL+/1iTmSWq\n7OjcZe6+z8yek9S4VqsGUGeYVnT2eldSPzO7VJKCOddJkjZIam9mlwXrW5hZjKRYlR1RKJV0u6To\neqobqA0Xm9kVweNRkt6suDE4Yvaxmd0slZ00aWbpwea/Sbo7WB9tZhdUfK6ZdZa0y92fkfQ/Kpt6\nVNEySTcGY66ZpG9JeqP2XhoQquo+vy9IGqmygPBi0HaRpLsqnM/WwczaVNHnBSoLCwVmFi9paMiv\nAUCICAdnKXffLekOSbPN7EOVhYWu7n5cZXNHnzSzVZKWqOwXnKmSvmdm76psStGhKjsGzg7rVfZ5\n/lBSS0m/r6LNdySNDsbBWknDg/X3SRpoZqtVNvUntdLzsiWtNLMPVHbOwH9V3Oju70t6TtJ7kv4h\n6Q/u/oGAs0B1n193XyuphaRP3f2zoO1iSX+S9E4wXl4M2lTuc5XKph6tlfSspLcqtwFw9jD3qmag\nAMCZKZjS8Kq7p9VzKQAAnHM4cgAAAABAEkcOAAAAAAQ4cgAAAABAEuEAAAAAQIBwAAAAAEAS4QAA\nAABAgHAAAAAAQJL0/wBFrKBAHxcihwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f3c93ab4a90>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:5:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>59036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>88349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>7460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>97957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>83174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>3287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0        59036\n",
       "1  plasmid_1        88349\n",
       "2  plasmid_2         7460\n",
       "3  plasmid_3        97957\n",
       "4  plasmid_4        83174\n",
       "5  plasmid_5         3287"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.375719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.389916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.382686</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.375719\n",
       "1  precision  0.389916\n",
       "2   f1_score  0.382686"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP012925.1</th>\n",
       "      <th>CP012926.1</th>\n",
       "      <th>CP012927.1</th>\n",
       "      <th>CP012928.1</th>\n",
       "      <th>CP012929.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.919310</td>\n",
       "      <td>0.086816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002750</td>\n",
       "      <td>0.008725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.088620</td>\n",
       "      <td>0.081745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.948188</td>\n",
       "      <td>0.102830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.027257</td>\n",
       "      <td>0.040992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.072044</td>\n",
       "      <td>0.003957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.058169</td>\n",
       "      <td>0.325066</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP012925.1  CP012926.1  CP012927.1  CP012928.1  CP012929.1\n",
       "0  plasmid_0         0.0         0.0         0.0    0.919310    0.086816\n",
       "1  plasmid_1         0.0         0.0         0.0    0.002750    0.008725\n",
       "2  plasmid_2         0.0         0.0         0.0    0.088620    0.081745\n",
       "3  plasmid_3         0.0         0.0         0.0    0.948188    0.102830\n",
       "4  plasmid_4         0.0         0.0         0.0    0.027257    0.040992\n",
       "5  plasmid_5         0.0         0.0         0.0    0.072044    0.003957\n",
       "6      Total         0.0         0.0         0.0    2.058169    0.325066"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012925.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012926.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012927.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012928.1</td>\n",
       "      <td>0.959008</td>\n",
       "      <td>0.001981</td>\n",
       "      <td>0.690483</td>\n",
       "      <td>0.604612</td>\n",
       "      <td>0.020655</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012929.1</td>\n",
       "      <td>0.131953</td>\n",
       "      <td>0.007504</td>\n",
       "      <td>0.967024</td>\n",
       "      <td>0.099442</td>\n",
       "      <td>0.044954</td>\n",
       "      <td>0.060846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Total</td>\n",
       "      <td>1.090961</td>\n",
       "      <td>0.009485</td>\n",
       "      <td>1.657507</td>\n",
       "      <td>0.704054</td>\n",
       "      <td>0.065609</td>\n",
       "      <td>1.060846</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2  plasmid_3  plasmid_4  \\\n",
       "0  CP012925.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "1  CP012926.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "2  CP012927.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "3  CP012928.1   0.959008   0.001981   0.690483   0.604612   0.020655   \n",
       "4  CP012929.1   0.131953   0.007504   0.967024   0.099442   0.044954   \n",
       "5       Total   1.090961   0.009485   1.657507   0.704054   0.065609   \n",
       "\n",
       "   plasmid_5  \n",
       "0   0.000000  \n",
       "1   0.000000  \n",
       "2   0.000000  \n",
       "3   1.000000  \n",
       "4   0.060846  \n",
       "5   1.060846  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.959059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.007504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.967024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.620293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.044954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.959059\n",
       "1  plasmid_1    0.007504\n",
       "2  plasmid_2    0.967024\n",
       "3  plasmid_3    0.620293\n",
       "4  plasmid_4    0.044954\n",
       "5  plasmid_5    1.000000"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012925.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP012926.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP012927.1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP012928.1</td>\n",
       "      <td>0.995598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP012929.1</td>\n",
       "      <td>0.138481</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP012925.1    0.000000\n",
       "1  CP012926.1    0.000000\n",
       "2  CP012927.1    0.000000\n",
       "3  CP012928.1    0.995598\n",
       "4  CP012929.1    0.138481"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP012928.1</td>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.91931</td>\n",
       "      <td>0.959008</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction   recall  precision\n",
       "0  CP012928.1  plasmid_0  0.91931   0.959008"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 5, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00126169</td>\n",
       "      <td>-0.00463641</td>\n",
       "      <td>0.00809902</td>\n",
       "      <td>-0.00147267</td>\n",
       "      <td>-0.00215429</td>\n",
       "      <td>0.00969438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[27, 46, 26, 36, 13, 48, 30]</td>\n",
       "      <td>[37, 14, 49, 15, 39]</td>\n",
       "      <td>[48, 31, 52, 25, 46, 24]</td>\n",
       "      <td>[46, 24, 49, 84, 48, 30, 52, 23, 36, 13, 71, 5...</td>\n",
       "      <td>[66, 32, 62, 77, 42, 85, 40, 71, 55, 41, 47, 3...</td>\n",
       "      <td>[52, 23]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.755669</td>\n",
       "      <td>-0.73785</td>\n",
       "      <td>-0.734026</td>\n",
       "      <td>-0.466429</td>\n",
       "      <td>-0.421975</td>\n",
       "      <td>-0.634063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.756931</td>\n",
       "      <td>-0.742487</td>\n",
       "      <td>-0.725927</td>\n",
       "      <td>-0.467902</td>\n",
       "      <td>-0.42413</td>\n",
       "      <td>-0.624369</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           plasmid_0             plasmid_1  \\\n",
       "GC_sum                   -0.00126169           -0.00463641   \n",
       "chain   [27, 46, 26, 36, 13, 48, 30]  [37, 14, 49, 15, 39]   \n",
       "gd_sum                     -0.755669              -0.73785   \n",
       "total                      -0.756931             -0.742487   \n",
       "\n",
       "                       plasmid_2  \\\n",
       "GC_sum                0.00809902   \n",
       "chain   [48, 31, 52, 25, 46, 24]   \n",
       "gd_sum                 -0.734026   \n",
       "total                  -0.725927   \n",
       "\n",
       "                                                plasmid_3  \\\n",
       "GC_sum                                        -0.00147267   \n",
       "chain   [46, 24, 49, 84, 48, 30, 52, 23, 36, 13, 71, 5...   \n",
       "gd_sum                                          -0.466429   \n",
       "total                                           -0.467902   \n",
       "\n",
       "                                                plasmid_4   plasmid_5  \n",
       "GC_sum                                        -0.00215429  0.00969438  \n",
       "chain   [66, 32, 62, 77, 42, 85, 40, 71, 55, 41, 47, 3...    [52, 23]  \n",
       "gd_sum                                          -0.421975   -0.634063  \n",
       "total                                            -0.42413   -0.624369  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_11</th>\n",
       "      <th>plasmid_12</th>\n",
       "      <th>plasmid_13</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.00139482</td>\n",
       "      <td>0.00928232</td>\n",
       "      <td>-0.0131559</td>\n",
       "      <td>2.75937e-05</td>\n",
       "      <td>-0.000547016</td>\n",
       "      <td>0.000380912</td>\n",
       "      <td>0.0106062</td>\n",
       "      <td>-0.0180649</td>\n",
       "      <td>-0.00246213</td>\n",
       "      <td>0.00969438</td>\n",
       "      <td>0.00594133</td>\n",
       "      <td>0.0123145</td>\n",
       "      <td>-0.00635206</td>\n",
       "      <td>-0.0102029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[15, 39, 15, 49, 24, 46, 27, 36, 26, 46, 25]</td>\n",
       "      <td>[36, 26, 46, 25, 52, 31, 48, 84, 49, 24, 46, 27]</td>\n",
       "      <td>[39]</td>\n",
       "      <td>[19, 75, 35, 7, 43]</td>\n",
       "      <td>[1, 75, 35, 6, 76]</td>\n",
       "      <td>[4, 75, 35, 3, 58]</td>\n",
       "      <td>[36, 26, 46, 27]</td>\n",
       "      <td>[16]</td>\n",
       "      <td>[13, 48, 30]</td>\n",
       "      <td>[52, 23]</td>\n",
       "      <td>[28]</td>\n",
       "      <td>[48]</td>\n",
       "      <td>[14, 37, 14]</td>\n",
       "      <td>[37]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.759532</td>\n",
       "      <td>-0.752863</td>\n",
       "      <td>-0.766323</td>\n",
       "      <td>-0.00234847</td>\n",
       "      <td>-0.000551762</td>\n",
       "      <td>-0.000425643</td>\n",
       "      <td>-0.777706</td>\n",
       "      <td>-0.684803</td>\n",
       "      <td>-0.75344</td>\n",
       "      <td>-0.634063</td>\n",
       "      <td>-0.393607</td>\n",
       "      <td>-0.833333</td>\n",
       "      <td>-0.717055</td>\n",
       "      <td>-0.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.760927</td>\n",
       "      <td>-0.743581</td>\n",
       "      <td>-0.779479</td>\n",
       "      <td>-0.00232088</td>\n",
       "      <td>-0.00109878</td>\n",
       "      <td>-4.4731e-05</td>\n",
       "      <td>-0.7671</td>\n",
       "      <td>-0.702867</td>\n",
       "      <td>-0.755902</td>\n",
       "      <td>-0.624369</td>\n",
       "      <td>-0.387666</td>\n",
       "      <td>-0.821019</td>\n",
       "      <td>-0.723407</td>\n",
       "      <td>-0.843536</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           plasmid_0  \\\n",
       "GC_sum                                   -0.00139482   \n",
       "chain   [15, 39, 15, 49, 24, 46, 27, 36, 26, 46, 25]   \n",
       "gd_sum                                     -0.759532   \n",
       "total                                      -0.760927   \n",
       "\n",
       "                                               plasmid_1 plasmid_10  \\\n",
       "GC_sum                                        0.00928232 -0.0131559   \n",
       "chain   [36, 26, 46, 25, 52, 31, 48, 84, 49, 24, 46, 27]       [39]   \n",
       "gd_sum                                         -0.752863  -0.766323   \n",
       "total                                          -0.743581  -0.779479   \n",
       "\n",
       "                 plasmid_11          plasmid_12          plasmid_13  \\\n",
       "GC_sum          2.75937e-05        -0.000547016         0.000380912   \n",
       "chain   [19, 75, 35, 7, 43]  [1, 75, 35, 6, 76]  [4, 75, 35, 3, 58]   \n",
       "gd_sum          -0.00234847        -0.000551762        -0.000425643   \n",
       "total           -0.00232088         -0.00109878         -4.4731e-05   \n",
       "\n",
       "               plasmid_2  plasmid_3     plasmid_4   plasmid_5   plasmid_6  \\\n",
       "GC_sum         0.0106062 -0.0180649   -0.00246213  0.00969438  0.00594133   \n",
       "chain   [36, 26, 46, 27]       [16]  [13, 48, 30]    [52, 23]        [28]   \n",
       "gd_sum         -0.777706  -0.684803      -0.75344   -0.634063   -0.393607   \n",
       "total            -0.7671  -0.702867     -0.755902   -0.624369   -0.387666   \n",
       "\n",
       "        plasmid_7     plasmid_8  plasmid_9  \n",
       "GC_sum  0.0123145   -0.00635206 -0.0102029  \n",
       "chain        [48]  [14, 37, 14]       [37]  \n",
       "gd_sum  -0.833333     -0.717055  -0.833333  \n",
       "total   -0.821019     -0.723407  -0.843536  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_2  from MILP completely contained in  plasmid_1  from greedy.\n",
      "\n",
      "plasmid_2  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_4  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_4  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_5  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_5  from MILP completely contained in  plasmid_5  from greedy.\n",
      "\n",
      "plasmid_5  from greedy completely contained in  plasmid_5  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_8  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_9  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_10  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_10  from greedy completely contained in  plasmid_4  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
