{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 103 with $\\alpha_1 = 1$, $\\alpha_2 = 5$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 103\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020824.1</td>\n",
       "      <td>59372</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP020824.1        59372"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xl4VdW9//HPNwkEAhgFQhglUUgg\nAxETsQaUoMhwVdAHB9LaVoVitVy1UJXb2zrVoXihtnilAtbHAaqAA0WlAvVHirMEBWUKIBJAQEAg\nEMYM6/fH2eEeQxIOmp0Twvv1PDycvfc6a39POIucz9lr723OOQEAAABARLgLAAAAAFA/EA4AAAAA\nSCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPBEhbuAk9W6\ndWuXkJAQ7jJOGQcOHFCzZs3CXQZQ7zFWgNAwVk7O0qVLdznn4sJdBxCqUy4cJCQkKD8/P9xlnDLy\n8vKUk5MT7jKAeo+xAoSGsXJyzKww3DUAJ4NpRQAAAAAkEQ4AAAAAeAgHAAAAACQRDgAAAAB4CAcA\nAAAAJBEOAAAAAHgIBwAAAAAkEQ4AAAAAeAgHAAAAACQRDgAAAAB4CAcAAAAAJElR4S6gPkoY91a4\nS6g1zw1qFu4SAAAAcIrgyAEAAAAASYQDAAAAAB7CAQAAAABJhAMAAAAAHsIBAAAAAEmEAwAAAAAe\nLmXa0G1bJj0wNNxV1I4HisJdAQAAQIPGkQMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAA\nAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAA\nkggHAAAAADyEAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAknwOB2Y2yMwKzGy9mY2r\nYvvZZrbIzD4zs8/N7D/8rAcAAABA9XwLB2YWKekpSYMlpUjKNbOUSs1+J2mWc66npOGSJvtVDwAA\nAICa+XnkoJek9c65Dc65o5JeljS0Uhsn6QzvcaykrT7WAwAAAKAGUT723UHS5qDlLZIurNTmAUkL\nzOw/JTWT1N/HegAAAADUwM9wYFWsc5WWcyU955ybaGYXSXrRzNKcc+Xf6chslKRRkhQfH6+8vDw/\n6j1mbHqpr/3XpeLo9spLfjDcZdQOn//dcXorLi72/f8WoCFgrAANm5/hYIukTkHLHXX8tKERkgZJ\nknPuQzNrIqm1pB3BjZxzUyVNlaSsrCyXk5PjU8kBN417y9f+69JzGV8pp+D+cJdRO3KLwl0BGrC8\nvDz5/X8L0BAwVoCGzc9zDpZI6mpmiWbWWIETjudWarNJ0mWSZGbdJTWRtNPHmgAAAABUw7dw4Jwr\nlTRa0nxJqxW4KtFKM3vIzIZ4zcZK+oWZLZf0kqSbnHOVpx4BAAAAqAN+TiuSc26epHmV1t0X9HiV\npN5+1gAAAAAgNNwhGQAAAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAA\nAIAkwgEAAAAAD+EAAAAAgCTCAQAAAAAP4QAAAACAJCkq3AUAANDQJYx7K9wl1JrnBjULdwkAfMSR\nAwAAAACSCAcAAAAAPIQDAAAAAJIIBwAAAAA8hAMAAAAAkggHAAAAADyEAwAAAACSCAcAAAAAPIQD\nAAAAAJIIBwAAAAA8UeEuAAAAnEK2LZMeGBruKmrHA0XhrgCodzhyAAAAAEAS4QAAAACAh3AAAAAA\nQBLhAAAAAICHcAAAAABAEuEAAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAAQBLhAAAAAICHcAAAAABA\nEuEAAAAAgIdwAAAAAEAS4QAAAACAh3AAAAAAQBLhAAAAAICHcAAAAABAEuEAAAAAgIdwAAAAAEAS\n4QAAAACAh3AAAAAAQBLhAAAAAICHcAAAAABAEuEAAAAAgIdwAAAAAECSz+HAzAaZWYGZrTezcdW0\nud7MVpnZSjP7u5/1AAAAAKhelF8dm1mkpKckXS5pi6QlZjbXObcqqE1XSf8lqbdzbo+ZtfGrHgAA\nAAA18/PIQS9J651zG5xzRyW9LGlopTa/kPSUc26PJDnndvhYDwAAAIAa+BkOOkjaHLS8xVsXLElS\nkpm9b2YfmdkgH+sBAAAAUAPfphVJsirWuSr231VSjqSOkt41szTn3N7vdGQ2StIoSYqPj1deXl6t\nFxtsbHqpr/3XpeLo9spLfjDcZdQOn//dcXorLi72/f8WnL74vVJPMeaB4/gZDrZI6hS03FHS1ira\nfOScK5H0lZkVKBAWlgQ3cs5NlTRVkrKyslxOTo5fNUuSbhr3lq/916XnMr5STsH94S6jduQWhbsC\nNGB5eXny+/8WnL74vVJP8XsFOI6f04qWSOpqZolm1ljScElzK7WZI6mfJJlZawWmGW3wsSYAAAAA\n1fAtHDjnSiWNljRf0mpJs5xzK83sITMb4jWbL+lbM1slaZGku51z3/pVEwAAAIDq+TmtSM65eZLm\nVVp3X9BjJ2mM9wcAwmfbMumByhdUO0U9wFQJAMD3wx2SAQAAAEjy+cgBgIYtoUGdZBnuCgAACD+O\nHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEc\nAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwA\nAAAAkHQS4cDM+pjZzd7jODNL9K8sAAAAAHUtpHBgZvdLulfSf3mrGkma7ldRAAAAAOpeqEcOrpE0\nRNIBSXLObZXUwq+iAAAAANS9UMPBUeeck+Qkycya+VcSAAAAgHAINRzMMrMpks40s19I+pekaf6V\nBQAAAKCuRYXSyDk3wcwul7RPUrKk+5xzC32tDAAAAECdOmE4MLNISfOdc/0lEQgAAACABuqE04qc\nc2WSDppZbB3UAwAAACBMQppWJOmwpC/MbKG8KxZJknPuDl+qAgAAAFDnQg0Hb3l/AAAAADRQoZ6Q\n/LyZNZaU5K0qcM6V+FcWAAAAgLoWUjgwsxxJz0vaKMkkdTKznzvnFvtXGgAAAIC6FOq0oomSBjjn\nCiTJzJIkvSQp06/CAAAAANStUG+C1qgiGEiSc26tpEb+lAQAAAAgHEI9cpBvZn+T9KK3/BNJS/0p\nCQAAAEA4hBoObpP0K0l3KHDOwWJJk/0qCgAAAEDdCzUcREn6i3PuT9KxuyZH+1YVAABAA7V06dI2\nUVFRz0hKU+hTvIHaUC5pRWlp6cjMzMwdVTUINRy8I6m/pGJvuamkBZKyf3CJAAAAp5GoqKhn2rZt\n2z0uLm5PRESEC3c9OH2Ul5fbzp07U7Zv3/6MpCFVtQk1rTZxzlUEA3mPY2qhRgAAgNNNWlxc3D6C\nAepaRESEi4uLK1LgqFXVbULs64CZnV+xYGZZkg79wPoAAABORxEEA4SL996rNgOEOq3oTkmzzWyr\nJCepvaQbfnh5AAAAAOqLUMNBoqSeks6WdI2kHykQEgAAAPADJIx7q1ZvKrvxj1fUm8vN9+rVK3nC\nhAmbL7nkkoOhtC8vL1d2dnbSvHnz1rds2bLc7/p+qDfffLPFxIkT4xctWrTej/5nzJgRu3LlyqaP\nPvro9srbYmJieh48ePCz6p6bnZ2d9I9//OPLuLi4spPZZ6jTin7vnNsn6UxJl0uaKumvJ7MjAAAA\nnDpKSkrqfJ+zZs2KTU1NPXQqBIO68JOf/KSoqmAQitzc3G8nTJgQd7LPCzUcVCSOKyQ97Zz7h6TG\nJ7szAAAAhN/dd9/dLjExMTU7O7vrVVddlXjffffFS4Fv+kePHt3hggsuSH744Yfjt27dGjVw4MBz\n09LSuqelpXVfsGBBM0nat29fxHXXXZeQlpbWvXv37inTp08/U5KKi4vtyiuvPCcpKSnliiuuOOfw\n4cMmSU888UTrESNGdKrY/8SJE1uPHDmyY+W6ZsyY0fKaa67ZK0kFBQWNExMTU2+44YbOXbt2TR0y\nZEjinDlzWpx//vndOnfunLZo0aKYmmopKChonJmZmZySktI9JSWl+8KFC5tJgW/7e/XqlTxo0KBz\nEhMTU4cMGZJYXl5zFhk2bFjCj3/847MzMzOTExIS0l566aXYym0WLVoU07Nnz27du3dP6dmzZ7fl\ny5dHS1J+fn6T9PT07t26dUtJSkpK+eKLL6JDfW2TJk1q9bOf/exsSVqzZk3j8847r1taWlr3O++8\ns33FfgsLCxtlZWUld+vWLaVr166pb7/9dnNJGj58+N7XXnutVYhviWNCnVb0tZlNUeBypuPNLFpc\nlxcAAOCUs3jx4pg33njjrC+++GJVSUmJnXfeeSk9e/Y8Nu1n7969kUuWLCmQpKuuuipxzJgx3wwc\nOLB43bp1jQcOHNh1w4YNK3/729+269ev377Zs2dv3LVrV2RWVlb3IUOG7PvTn/4U17Rp0/K1a9eu\n+vjjj5v27t07RZJGjBixOzU1NeXIkSNboqOj3fTp01tPmTKlsHJtS5cubd67d+9j6zdv3txk5syZ\nGzIzMwt79OjRfcaMGa3y8/PX/P3vfz/zkUceadevX78vq6ulffv2pe++++7amJgY98UXX0Tn5uae\ns2LFitWStHr16qbLli3bkJCQUJKZmdlt4cKFzQcOHFhcuZ5gmzdvjv7kk08KVq1aFd2/f//koUOH\nfhG8PSMj4/Ann3yyplGjRpozZ06Le+65p+P8+fO/fPLJJ+Nuv/32b2677bbdhw8fttLSUn399deN\nQnltwf3ffvvtZ48cOXLn6NGjv33ssceOHRF49tlnW1522WVF48eP315aWqr9+/dHSFJcXFzZ0aNH\nbfv27ZFt27YNeWpRqOHgekmDJE1wzu01s3aS7g51JwAAAKgf8vLymg8ePHhv8+bNnSR3+eWX7w3e\nnpubu7vi8fvvv3/GunXrmlYsFxcXR+7ZsyciLy/vjPnz5585adKktpJ05MgRW79+feP33nuv+R13\n3LFDki688MJDSUlJByXpjDPOKO/du/f+mTNnxqanpx8uKSmxXr16HXfly6Kioqizzjrr2Nf4HTp0\nOFLRLikp6dCll166LyIiQueff/7Bhx9+uL33eqqspXPnziUjRozovGrVqqYREREqLCw8dgPf9PT0\nA+eee26JJKWmph788ssvTzgjZtiwYbsjIyOVnp5+pFOnTkeWLVvWJHj77t27I2+44YbEjRs3NjEz\nV1JSYpJ00UUXHZgwYUK7LVu2NB4+fPie9PT0I6G+tmCffvpp83/+859fStKtt9767R/+8IeOkvSj\nH/3owK233ppQUlISce211+7Jzs4+9nNt1apV6aZNmxq3bds25KuMhhQOnHMHJb0WtLxN0rZQdwIA\nAID6wbmarynTokWLYx/OnXPKz89f7QWJ7/TxyiuvrM/IyDhS+flmVmW/o0aN2vXII4+0TUpKOnzj\njTfuqqpNZGSkKysrU2RkpCSpcePGx/YbERGhJk2aOK+dysrKrKZaxowZ075NmzYlr7766lfl5eVq\n2rTpsRO/o6Ojj/UbGRmp0tLSqouu4XVVXr733ns79O3bd//ChQu/LCgoaHzppZcmS9Ivf/nL3Rdf\nfPGB119/PXbw4MFJkydP3picnHwklNdWWVWXwB08eHDx4sWLC1599dXYm266KfGOO+74ZvTo0d9K\ngaAUExNzUudvMDUIAADgNJKTk1M8f/782IMHD1pRUVHEv/71rzOra9unT59948ePb1Ox/MEHHzSV\npH79+u2bOHFifMVc/ffff7+p1754+vTpLSVpyZIlTdauXXvsprmXXnrpgW3btjV+/fXXW40YMWK3\nqpCYmHh49erV0VVtq051tRQVFUW2a9euJDIyUpMnT25VVnbimTW/+tWvOrzwwgtV/jxee+21s8rK\nyrRy5crozZs3R2dkZBwO3r5v377Ijh07HpWkKVOmtK5Yv2rVqsbdu3c/8rvf/W7HgAED9i5btqxp\n5b5Dcf755xdPmzatpSRNmzbt2LkEa9eubdyhQ4eSsWPH7rrxxht3ffrppzFS4MpPO3fubJScnHxc\ngKtJqNOKAAAA4IO6vvRo3759Dw4aNKgoJSUltUOHDkd69OhxIDY2tspPzlOnTt08cuTIs5OSklLK\nysrswgsv3J+dnb3pj3/849ZRo0ad3a1btxTnnHXs2PHIokWL1v/mN7/ZMXz48MSkpKSU1NTUg+np\n6QeC+7v66qv3fP755zHVXV5zwIABRQsWLGiRlpYW8gfa6mq56667dgwbNuzcOXPmnNWnT5/9TZs2\nPeE36KtWrWpacUJ0ZV26dDnSq1ev5G+//bbRn//858KYmJjvfIt/7733bh85cmTipEmT2l588cX7\nKta/+OKLLWfPnt0qKirKxcXFlTz22GNb9+7dGxnq66swefLkTcOHDz9n8uTJ8UOGDNlTsX7+/Pkt\nJk2a1DYqKsrFxMSUzZgx4ytJeu+992J69ux5oFGjRie1HzvRoaUfwswGSfqLpEhJzzjn/lhNu2sl\nzZZ0gXMuv6Y+s7KyXH5+jU1+sIRxb/naf116LmOdcgruD3cZteOBonBXgEoYK/UUY6XeYazUU3Uw\nVsxsqXMuK3jd8uXLN2ZkZFQ5raauFBUVRcTGxpbv378/4qKLLkp++umnC/v06RPSvQh+iH79+nW5\n6667vhk6dOj+qrYXFhY2ys3NTfjggw/W+V1LVfr06dP1vffeO27fw4YNS7jyyiuLbr755j1VPa8+\nuvnmmztdffXVe6v6WS9fvrx1RkZGQlXP821akZlFSnpK0mBJKZJyzSylinYtJN0h6WO/agEAAMD/\nufHGGzt369YtpUePHt2vuuqqPX4Hg127dkUmJCSkNWnSpLy6YCBJnTt3Lrnlllt27d69OyxT36sK\nBqeqtLS0QzX9rKvj57SiXpLWO+c2SJKZvSxpqKRVldr9QdLjkn7jYy0AAADwvPHGG1/V5f5at25d\ntnHjxhWhtB05cmS9+3b+1Vdf3RjuGk7W2LFjv9fRKT9TWQdJm4OWt3jrjjGznpI6Oefe9LEOAAAA\nACHw88hBVZdgOnaCg5lFSHpC0k0n7MhslKRRkhQfH6+8vLzaqbAaY9NLfe2/LhVHt1de8oPhLqN2\n+PzvjpPHWKmnGCv1DmOlnmKsAMfxMxxskdQpaLmjpK1Byy0kpUnK864T21bSXDMbUvmkZOfcVElT\npcAJyTk5OT6WLd3UoE4c+6rhnDiWy0mW9Q1jpZ5irNQ7jJV6irECHMfPaUVLJHU1s0QzayxpuKS5\nFRudc0XOudbOuQTnXIKkjyQdFwwAAAAA1A3fjhw450rNbLSk+QpcyvRZ59xKM3tIUr5zbm7NPQAA\nAJwGHojNPHGjk+mv6IT3TTCzzKFDh+6eM2fOV5JUUlKiNm3aZJx33nkHFi1atH7SpEmt8vPzm73w\nwgubxowZ07558+ZlDz300DfBfURGRmZ27dr1UFlZmXXp0uXQrFmzNgbfXVkK3IgrOzs7ad68eetb\ntmxZft111yW88847sa1atSpdt27dyqpqe/PNN1vk5uae26FDh6OSdOWVV+6ZMGHCtsrtHn300bin\nn346fvPmzdFbt25d3q5du1JJeumll2I/+eSTZk888cTWys/Bifl6mSjn3DznXJJz7lzn3CPeuvuq\nCgbOuRyOGgAAAPivadOm5QUFBU2Li4tNkl5//fUz4uPjS06mj+jo6PI1a9asWrdu3cpGjRq5iRMn\nxlVuM2vWrNjU1NRDLVu2LJekW265ZdfcuXNPeLnQrKys4jVr1qxas2bNqqqCgST17du3eOHChWvb\nt29/NHj9DTfcUPT222+fuX///rBcDvVUxw8NAADgNHTZZZcVzZ49+0xJeumll1oOGzZs9/ftq0+f\nPsXr16+Prrx+xowZLYPvODx48ODiuLi4WjlDv3fv3oeSk5OPVl4fERGh7Ozs/TNnzoytjf2cbggH\nAAAAp6Gf/vSnu2fOnHnWwYMHbfXq1TEXXXTRge/TT0lJiebPn39Genr6ocrbli5d2rx3794n7Pfx\nxx+Pe/zxx48defjss8+aJycnp1xyySVd8/Pzm1Ss79u3b5eNGzc2OlF/WVlZB959993mJ/M6EODn\n1YoAAABQT1144YWHtmzZEj1t2rSW/fv3P+lLNx05ciSiW7duKV5f+++8887jbrpVVFQUddZZZ5Uf\n/+zvuueee3ZWPM7Ozj5QWFj4eWxsbPnMmTNjhw0b1qWwsHCFJP373/9eH0ptbdu2Ld2+fXvj0F8N\nKhAOAAAATlODBg3ae//993dasGBBwY4dO07qc2HFOQc1tYmMjHRlZWWKjIwMud+K8xOkwPkDY8aM\nOXvbtm1RFScch+LQoUPWpEmTE4YSHI9pRQAAAKep2267bdfYsWO39urV67gpQbUhMTHx8OrVq487\nF6EmmzZtiiovD3yuX7RoUUx5ebni4+NP6jyFgoKCJqmpqb68poaOIwcAAADhFMKlR/1y7rnnlvz+\n97/fcaJ2TzzxRLspU6bEVyx/8803n4fS/4ABA4oWLFjQIi0t7YgkXXXVVYkfffRRiz179kTFx8f3\nGDdu3NZf//rXuyrON7jnnnt2Tp8+/axnn322TWRkpGvSpEn5Cy+8sCEiIvB9dt++fbs8//zzhQkJ\nCSUPP/xwmyeffLLtt99+2ygjIyOlX79+RTNnziyUpMWLF7cYP37819/jR3LaM+dcuGs4KVlZWS4/\n398rniY0qDtZrms4d7J8gDtZ1jeMlXqKsVLvMFbqqToYK2a21DmXFbxu+fLlGzMyMo6bn98QFRYW\nNsrNzU344IMPTnj50tqyefPmqOuvv/6cDz/8cG1d7fNUs3z58tYZGRkJVW1jWhEAAAB80blz55Jb\nbrll1+7du+vsM+eGDRsaT5w4cXNd7a+hYVoRAAAAfDNy5Mg9dbm/vn37HqzL/TU0HDkAAACoW+Xl\n5eUW7iJwevLee9VeyYlwAAAAULdW7Ny5M5aAgLpWXl5uO3fujJW0oro2TCsCAACoQ6WlpSO3b9/+\nzPbt29PEF7WoW+WSVpSWlo6srgHhAAAAoA5lZmbukDQk3HUAVSGtAgAAAJBEOAAAAADgIRwAAAAA\nkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAAAACQ\nRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEcAAAAAJBE\nOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4\nAAAAAOAhHAAAAACQRDgAAAAA4CEcAAAAAJBEOAAAAADgIRwAAAAAkEQ4AAAAAOAhHAAAAACQ5HM4\nMLNBZlZgZuvNbFwV28eY2Soz+9zM3jGzzn7WAwAAAKB6voUDM4uU9JSkwZJSJOWaWUqlZp9JynLO\n9ZD0iqTH/aoHAAAAQM38PHLQS9J659wG59xRSS9LGhrcwDm3yDl30Fv8SFJHH+sBAAAAUAM/w0EH\nSZuDlrd466ozQtI/fawHAAAAQA2ifOzbqljnqmxodqOkLEl9q9k+StIoSYqPj1deXl4tlVi1seml\nvvZfl4qj2ysv+cFwl1E7fP53x8ljrNRTjJV6h7FSTzFWgOP4GQ62SOoUtNxR0tbKjcysv6T/ltTX\nOXekqo6cc1MlTZWkrKwsl5OTU+vFBrtp3Fu+9l+Xnsv4SjkF94e7jNqRWxTuClAJY6WeYqzUO4yV\neoqxAhzHz2lFSyR1NbNEM2ssabikucENzKynpCmShjjndvhYCwAAAIAT8C0cOOdKJY2WNF/Sakmz\nnHMrzewhMxviNfsfSc0lzTazZWY2t5ruAAAAAPjMz2lFcs7NkzSv0rr7gh7393P/AAAAAELHHZIB\nAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPAQDgAA\nAABIIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAA\nAEgiHAAAAADwEA4AAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAA\nSCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPAQDgAAAABI\nIhwAAAAA8BAOAAAAAEgiHAAAAADwEA4AAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAAAEgi\nHAAAAADwEA4AAAAASCIcAAAAAPAQDgAAAABIIhwAAAAA8BAOAAAAAEgiHAAAAADw+BoOzGyQmRWY\n2XozG1fF9mgzm+lt/9jMEvysBwAAAED1fAsHZhYp6SlJgyWlSMo1s5RKzUZI2uOc6yLpCUnj/aoH\nAAAAQM38PHLQS9J659wG59xRSS9LGlqpzVBJz3uPX5F0mZmZjzUBAAAAqIaf4aCDpM1By1u8dVW2\ncc6VSiqS1MrHmgAAAABUI8rHvqs6AuC+RxuZ2ShJo7zFYjMr+IG1nTb6Sa0l7Qp3HbXiQQ4qwT+M\nFSA0jJWT1rkudgLUFj/DwRZJnYKWO0raWk2bLWYWJSlW0u7KHTnnpkqa6lOdDZqZ5TvnssJdB1Df\nMVaA0DBWgIbNz2lFSyR1NbNEM2ssabikuZXazJX0c+/xtZL+n3PuuCMHAAAAAPzn25ED51ypmY2W\nNF9SpKRnnXMrzewhSfnOubmS/ibpRTNbr8ARg+F+1QMAAACgZsYX9Q2bmY3ypmUBqAFjBQgNYwVo\n2AgHAAAAACT5fIdkAAAAAKcOwgEAAAAASYSD056ZJZjZCu9xjpm9Ge6agHAwsw9OsH2emZ1ZV/UA\nDYGZ5ZlZlvd4o5m1DndNAGrm530O4CMzMwXOGSkPdy1AfWNmkc65spN5jnMu+wTb/+OHVQU0PPwu\nAhoejhycQrxv+Veb2WRJn0pYeEVLAAAETElEQVT6qZl9aGafmtlsM2vutbvAzD4ws+Vm9omZtfCe\n+67X9lMzq/GDEFBfee/lNWb2vJl9bmavmFmM963kfWb2nqTrzOxcM3vbzJZ67/1u3vPjzex1b3ws\nrxgLZlbs/d3OzBab2TIzW2FmF3vrj33raWZjvG0rzOyuoLpWm9k0M1tpZgvMrGlYfkhADSq/f81s\nvJndHrT9ATMb6z2+28yWeGPtQW9d5d9Fnczsr2aW7733HwzPKwNQGwgHp55kSS9IulzSCEn9nXPn\nS8qXNMa74dxMSXc65zIk9Zd0SNIOSZd7bW+QNCkcxQO1JFnSVOdcD0n7JFV8sDnsnOvjnHtZgbuq\n/6dzLlPSbyRN9tpMkvRvb3ycL2llpb5/LGm+c+48SRmSlgVvNLNMSTdLulDSjyT9wsx6epu7SnrK\nOZcqaa+kYbX1goHaUNX7V9LLCvxeqHC9pNlmNkCB93QvSedJyjSzS7w2yZJecM71dM4VSvpv767J\nPST1NbMedfKCANQ6phWdegqdcx+Z2ZWSUiS9Hziqq8aSPlTgP+xtzrklkuSc2ydJZtZM0v+a2XmS\nyiQlhaN4oJZsds697z2eLukO7/FMSfKOomUr8AGn4jnR3t+XSvqZJHlTj4oq9b1E0rNm1kjSHOfc\nskrb+0h63Tl3wNvXa5IuVuCO718FtV8qKeEHvEbAD9W9f9uYWXtJcZL2OOc2mdkdkgZI+sx7bnMF\nwsImeb+Lgvq93sxGKfC5op0Cv58+r4sXBKB2EQ5OPQe8v03SQudcbvBG79uaqm5e8WtJ3yjwTWiE\npMN+Fgn4rPJ7vGK5YnxESNrrfft/ch07t9j7dvQKBe7g/j/OuReCmlg1T5WkI0GPyyQxrQj1TXXv\n31ckXSuprQJHEiraPuacm/KdDswS9H9jTWaWqMDRuQucc3vM7DlJTWq1agB1hmlFp66PJPU2sy6S\n5M25TpK0RlJ7M7vAW9/CzKIkxSpwRKFc0k8lRYapbqA2nG1mF3mPcyW9F7zRO2L2lZldJwVOmjSz\nDG/zO5Ju89ZHmtkZwc81s86Sdjjnpkn6mwJTj4ItlnS1N+aaSbpG0ru199IAX1X3/n1Z0nAFAsIr\nXtv5km4JOp+tg5m1qaLPMxQIC0VmFi9psM+vAYCPCAenKOfcTkk3SXrJzD5XICx0c84dVWDu6JNm\ntlzSQgW+wZks6edm9pECU4oOVNkxcGpYrcD7+XNJLSX9tYo2P5E0whsHKyUN9dbfKamfmX2hwNSf\n1ErPy5G0zMw+U+Ccgb8Eb3TOfSrpOUmfSPpY0jPOuc8EnAKqe/8651ZKaiHpa+fcNq/tAkl/l/Sh\nN15e8dpU7nO5AlOPVkp6VtL7ldsAOHWYc1XNQAGA+smb0vCmcy4tzKUAANDgcOQAAAAAgCSOHAAA\nAADwcOQAAAAAgCTCAQAAAAAP4QAAAACAJMIBAAAAAA/hAAAAAIAk6f8Db5hjbTOFpqIAAAAASUVO\nRK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f2455080828>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:5:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>37700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>50571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>3166</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0        37700\n",
       "1  plasmid_1        50571\n",
       "2  plasmid_2         3166"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.881796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.583352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.702179</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.881796\n",
       "1  precision  0.583352\n",
       "2   f1_score  0.702179"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP020824.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.851765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.046638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.898403</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP020824.1\n",
       "0  plasmid_0    0.000000\n",
       "1  plasmid_1    0.851765\n",
       "2  plasmid_2    0.046638\n",
       "3      Total    0.898403"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020824.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.874605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.874605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2\n",
       "0  CP020824.1        0.0        1.0   0.874605\n",
       "1       Total        0.0        1.0   0.874605"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.874605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.000000\n",
       "1  plasmid_1    1.000000\n",
       "2  plasmid_2    0.874605"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020824.1</td>\n",
       "      <td>0.881796</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP020824.1    0.881796"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP020824.1</td>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.851765</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction    recall  precision\n",
       "0  CP020824.1  plasmid_1  0.851765        1.0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 5, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0203192</td>\n",
       "      <td>-0.000623571</td>\n",
       "      <td>0.00186416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[43, 26]</td>\n",
       "      <td>[40, 22]</td>\n",
       "      <td>[36, 46, 40]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.636406</td>\n",
       "      <td>-0.693942</td>\n",
       "      <td>-0.570646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.656725</td>\n",
       "      <td>-0.694565</td>\n",
       "      <td>-0.568782</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        plasmid_0    plasmid_1     plasmid_2\n",
       "GC_sum -0.0203192 -0.000623571    0.00186416\n",
       "chain    [43, 26]     [40, 22]  [36, 46, 40]\n",
       "gd_sum  -0.636406    -0.693942     -0.570646\n",
       "total   -0.656725    -0.694565     -0.568782"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0202103</td>\n",
       "      <td>-0.000191801</td>\n",
       "      <td>-0.000738549</td>\n",
       "      <td>0.00172353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[43, 39, 43, 26]</td>\n",
       "      <td>[52, 36, 52, 42, 47, 40, 22]</td>\n",
       "      <td>[46, 40, 4, 57]</td>\n",
       "      <td>[40]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.612305</td>\n",
       "      <td>-0.694834</td>\n",
       "      <td>-0.002027</td>\n",
       "      <td>-0.627958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.632515</td>\n",
       "      <td>-0.695026</td>\n",
       "      <td>-0.00276555</td>\n",
       "      <td>-0.626235</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               plasmid_0                     plasmid_1        plasmid_2  \\\n",
       "GC_sum        -0.0202103                  -0.000191801     -0.000738549   \n",
       "chain   [43, 39, 43, 26]  [52, 36, 52, 42, 47, 40, 22]  [46, 40, 4, 57]   \n",
       "gd_sum         -0.612305                     -0.694834        -0.002027   \n",
       "total          -0.632515                     -0.695026      -0.00276555   \n",
       "\n",
       "         plasmid_3  \n",
       "GC_sum  0.00172353  \n",
       "chain         [40]  \n",
       "gd_sum   -0.627958  \n",
       "total    -0.626235  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_0  from MILP completely contained in  plasmid_0  from greedy.\n",
      "\n",
      "plasmid_1  from MILP completely contained in  plasmid_1  from greedy.\n",
      "\n",
      "plasmid_3  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_3  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
