{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Details about true plasmids for sample 56"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from __future__ import division\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['Density'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\t\tif contigs_dict[sseqid]['Gene_coverage'] > 0:\n",
    "\t\t\tcontigs_dict[sseqid]['Density'] = 1\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sample_id = str(56)\n",
    "\n",
    "sample_dir = '/home/aniket/python_scripts/Plasmids/data/unicycler_pipeline/'\n",
    "\n",
    "assembly_file = sample_dir + 'sample_' + sample_id + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + sample_id + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + sample_id + '/seed_contigs.csv'\n",
    "\n",
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "chain_file = 'contig_chains.csv'\n",
    "\n",
    "contig_chains = {}\n",
    "string_list = read_file(chain_file)\n",
    "for line in string_list:\n",
    "    plasmid_name = line.split(';')[0]\n",
    "    chain = line.split(';')[1]\n",
    "    contig_list = [x[:-1] for x in chain.split(',')]\n",
    "    contig_chains[plasmid_name] = contig_list   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Individual plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of contigs:  227\n",
      "Number of links:  321\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of contigs: \", len(contigs_dict))\n",
    "print(\"Number of links: \", len(links_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "objective_values = [] \n",
    "plasmid_contigs = {}\n",
    "for plasmid in contig_chains:\n",
    "    contig_list = contig_chains[plasmid]\n",
    "    avg_gd, avg_GC, GC_pen, total_len = 0, 0, 0, 0\n",
    "    plasmid_contigs[plasmid] = []\n",
    "    for contig in contig_list:\n",
    "        plasmid_contigs[plasmid].append([contig,contigs_dict[contig]['Gene_coverage'],contigs_dict[contig]['GC_cont'],contigs_dict[contig]['Read_depth'],contigs_dict[contig]['Length']])\n",
    "        \n",
    "        total_len += contigs_dict[contig]['Length']\n",
    "        avg_gd += contigs_dict[contig]['Gene_coverage']*contigs_dict[contig]['Length']\n",
    "        avg_GC += contigs_dict[contig]['GC_cont']*contigs_dict[contig]['Length']\n",
    "    avg_gd = avg_gd/total_len\n",
    "    avg_GC = avg_GC/total_len\n",
    "    for contig in contig_list:\n",
    "        GC_pen += abs(avg_GC - contigs_dict[contig]['GC_cont'])*contigs_dict[contig]['Length']\n",
    "    GC_pen = GC_pen/total_len\n",
    "    \n",
    "    plasmid_contigs[plasmid] = pd.DataFrame(plasmid_contigs[plasmid])\n",
    "    plasmid_contigs[plasmid].rename(columns = {0: 'Contig', 1: 'Gene density', 2: 'GC content', 3: 'Read depth', 4: 'Length'}, inplace = True)  \n",
    "    \n",
    "    objective_values.append([plasmid, avg_gd, GC_pen, total_len])\n",
    "    \n",
    "objective_values = pd.DataFrame(objective_values)\n",
    "objective_values.rename(columns = {0: 'Plasmid', 1: 'Gene density', 2: 'GC penalty', 3: 'Length'}, inplace = True)  \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CP017930.1\n",
      "   Contig  Gene density  GC content  Read depth  Length\n",
      "0      86      0.914264    0.568285    4.020228    1318\n",
      "1      59      0.825267    0.510254    1.621276    3657\n",
      "2      14      0.000000    0.548376    0.946770  166694\n",
      "3      86      0.914264    0.568285    4.020228    1318\n",
      "4      59      0.825267    0.510254    1.621276    3657\n",
      "5      91      1.000000    0.619090    4.618026    1121\n",
      "6     156      0.678571    0.521429    2.661191     140\n",
      "7      60      0.828022    0.516796    1.610469    3483\n",
      "8     106      1.000000    0.524390    5.316353     820\n",
      "9      66      0.871209    0.457831    1.688379    2407\n",
      "10    167      0.000000    0.495652    2.881034     115\n",
      "11    187      0.000000    0.609375    8.791028      64\n",
      "12    151      0.000000    0.554140   11.972298     157\n",
      "13    130      0.000000    0.544910    1.249322     334\n",
      "14    143      0.000000    0.537234   10.103299     188\n",
      "15    155      0.000000    0.492958   12.983792     142\n",
      "16    194      0.000000    0.423077    7.231292      52\n",
      "17     41      0.853910    0.470167    1.630500   20816\n",
      "18    167      0.000000    0.495652    2.881034     115\n",
      "19    187      0.000000    0.609375    8.791028      64\n",
      "20    151      0.000000    0.554140   11.972298     157\n",
      "21    130      0.000000    0.544910    1.249322     334\n",
      "22    143      0.000000    0.537234   10.103299     188\n",
      "23    155      0.000000    0.492958   12.983792     142\n",
      "24    194      0.000000    0.423077    7.231292      52\n",
      "25     58      0.886611    0.462415    1.684712    3951\n",
      "26    167      0.000000    0.495652    2.881034     115\n",
      "27    187      0.000000    0.609375    8.791028      64\n",
      "28    151      0.000000    0.554140   11.972298     157\n",
      "29    130      0.000000    0.544910    1.249322     334\n",
      "..    ...           ...         ...         ...     ...\n",
      "58    125      0.000000    0.551546    3.330511     388\n",
      "59     98      0.333006    0.488212    1.771605    1018\n",
      "60     49      0.923040    0.632069   12.352944    6224\n",
      "61     63      0.726647    0.432645    1.914865    3051\n",
      "62     72      0.907001    0.582550    9.299159    1914\n",
      "63     48      0.897953    0.466891    1.773743    6252\n",
      "64     68      0.858275    0.515618    3.252184    2145\n",
      "65     83      0.639378    0.528417    1.551288    1478\n",
      "66    173      0.000000    0.569767    3.152784      86\n",
      "67    173      0.000000    0.569767    3.152784      86\n",
      "68    173      0.000000    0.569767    3.152784      86\n",
      "69     68      0.858275    0.515618    3.252184    2145\n",
      "70     61      0.796719    0.513181    1.583429    3414\n",
      "71    155      0.000000    0.492958   12.983792     142\n",
      "72    176      0.000000    0.544304    2.677638      79\n",
      "73    159      0.000000    0.488722    2.699550     133\n",
      "74    166      0.000000    0.572650    2.721462     117\n",
      "75    145      0.000000    0.542373    9.978085     177\n",
      "76    151      0.000000    0.554140   11.972298     157\n",
      "77    187      0.000000    0.609375    8.791028      64\n",
      "78    167      0.000000    0.495652    2.881034     115\n",
      "79    104      0.805094    0.489480    1.652539     903\n",
      "80    107      0.927227    0.470514    9.060933     797\n",
      "81    116      0.869245    0.482505    4.833770     543\n",
      "82     67      0.690217    0.519475    1.561003    2208\n",
      "83    106      1.000000    0.524390    5.316353     820\n",
      "84    117      1.000000    0.646840    1.098537     538\n",
      "85     91      1.000000    0.619090    4.618026    1121\n",
      "86    156      0.678571    0.521429    2.661191     140\n",
      "87    101      0.840764    0.523355    1.663910     942\n",
      "\n",
      "[88 rows x 5 columns]\n",
      "\n",
      "\n",
      "\n",
      "CP017932.1\n",
      "  Contig  Gene density  GC content  Read depth  Length\n",
      "0    110      0.349454    0.466459    1.634113     641\n",
      "1    167      0.000000    0.495652    2.881034     115\n",
      "2    187      0.000000    0.609375    8.791028      64\n",
      "3    151      0.000000    0.554140   11.972298     157\n",
      "4    145      0.000000    0.542373    9.978085     177\n",
      "5    150      0.000000    0.541401    5.738444     157\n",
      "6    143      0.000000    0.537234   10.103299     188\n",
      "7    155      0.000000    0.492958   12.983792     142\n",
      "8     28      0.851632    0.521898    2.827378   61489\n",
      "\n",
      "\n",
      "\n",
      "CP017931.1\n",
      "  Contig  Gene density  GC content  Read depth  Length\n",
      "0     16      0.868845    0.488675    1.458063  111395\n",
      "\n",
      "\n",
      "\n",
      "CP017929.1\n",
      "  Contig  Gene density  GC content  Read depth  Length\n",
      "0     19      0.872979    0.537064    2.527550  103794\n",
      "1    107      0.927227    0.470514    9.060933     797\n",
      "2    116      0.869245    0.482505    4.833770     543\n",
      "3     54      0.986312    0.611217    2.132030    5260\n",
      "4     65      0.941438    0.618977    4.788697    2698\n",
      "\n",
      "\n",
      "\n",
      "CP017933.1\n",
      "  Contig  Gene density  GC content  Read depth  Length\n",
      "0     53      0.550425    0.521980   14.284520    5414\n",
      "1     49      0.923040    0.632069   12.352944    6224\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for plasmid in plasmid_contigs:\n",
    "    print(plasmid)\n",
    "    print(plasmid_contigs[plasmid])\n",
    "    print(\"\\n\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Plasmid</th>\n",
       "      <th>Gene density</th>\n",
       "      <th>GC penalty</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP017930.1</td>\n",
       "      <td>0.138497</td>\n",
       "      <td>0.011976</td>\n",
       "      <td>619060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP017932.1</td>\n",
       "      <td>0.833043</td>\n",
       "      <td>0.001342</td>\n",
       "      <td>63130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP017931.1</td>\n",
       "      <td>0.868845</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>111395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP017929.1</td>\n",
       "      <td>0.880248</td>\n",
       "      <td>0.010149</td>\n",
       "      <td>113092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP017933.1</td>\n",
       "      <td>0.749699</td>\n",
       "      <td>0.054778</td>\n",
       "      <td>11638</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Plasmid  Gene density  GC penalty  Length\n",
       "0  CP017930.1      0.138497    0.011976  619060\n",
       "1  CP017932.1      0.833043    0.001342   63130\n",
       "2  CP017931.1      0.868845    0.000000  111395\n",
       "3  CP017929.1      0.880248    0.010149  113092\n",
       "4  CP017933.1      0.749699    0.054778   11638"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "objective_values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Connected components"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Intersects ['167', '145', '155', '151', '187', '110', '143']\n",
      "Intersects ['107', '116']\n",
      "Intersects ['49']\n",
      "[['16'], ['195', '98', '145', '48', '60', '186', '106', '165', '65', '63', '156', '150', '72', '159', '91', '140', '143', '67', '167', '117', '59', '86', '125', '55', '83', '49', '174', '104', '28', '110', '173', '61', '66', '155', '151', '41', '58', '68', '130', '176', '116', '53', '166', '6', '14', '54', '107', '194', '19', '187', '101', '42']]\n"
     ]
    }
   ],
   "source": [
    "conn_comp = []\n",
    "for plasmid in contig_chains:\n",
    "    common = 0\n",
    "    #print(contig_chains[plasmid])\n",
    "    if len(conn_comp) == 0:\n",
    "        conn_comp.append(contig_chains[plasmid])\n",
    "    else:\n",
    "        for comp in conn_comp:\n",
    "            if len(list(set(comp) & set(contig_chains[plasmid]))) != 0:\n",
    "                print(\"Intersects\", list(set(comp) & set(contig_chains[plasmid])))\n",
    "                temp = (list(set().union(comp, contig_chains[plasmid])))\n",
    "                conn_comp.remove(comp)\n",
    "                conn_comp.append(temp)\n",
    "                common = 1\n",
    "                break\n",
    "            else:\n",
    "                common = 0\n",
    "        if common == 0:\n",
    "            conn_comp.append(contig_chains[plasmid])\n",
    "print(conn_comp)                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
