{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 11 with $\\alpha_1 = 1$, $\\alpha_2 = 1$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 11\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP011580.1</td>\n",
       "      <td>90452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP011579.1</td>\n",
       "      <td>33610</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP011580.1        90452\n",
       "1  CP011579.1        33610"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt0VeWd//HPNwmXBDAKhHCVpEqA\nJBAwESpgCYoKUwW78AKtdlQoVsuoxVaZTqvW2lodqS1OqWCnS61ULlYpUxmB9gfFayUoiIABRG4i\nAgKBcAm5fH9/ZIeJIQkHzc5Jwvu1louz937Os7/nrPN48jn72XubuwsAAAAAYqJdAAAAAICGgXAA\nAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAEAgLtoF\nnK727dt7SkpKtMtoNA4fPqxWrVpFuwygwWOsAJFhrJyelStX7nX3pGjXAUSq0YWDlJQU5eXlRbuM\nRmPZsmXKzc2NdhlAg8dYASLDWDk9ZrY12jUAp4NpRQAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAE\nCAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJElx0S6gIUqZ\n8nK0S6gzT49oFe0SAAAA0Ehw5AAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAA\nAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAA\ngADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACA\nAOEAAAAAgKSQw4GZjTCzfDPbZGZTqtl+rpktNbN3zew9M/uXMOsBAAAAULO4sDo2s1hJv5V0maQd\nklaY2QJ3X1ep2Y8lzXX335lZuqSFklLCqgkAgGhImfJytEuoM0+PaBXtEgCEKMwjBwMkbXL3ze5+\nXNJsSaOrtHFJZwWPEyXtDLEeAAAAALUI7ciBpC6Stlda3iFpYJU2D0habGb/JqmVpOHVdWRmEyVN\nlKTk5GQtW7asrmv9nLv7lITaf30qLCwM/f0CmgLGCsLE9wqAxiLMcGDVrPMqy+MkPe3uU83sIkl/\nNLNMdy/73JPcZ0qaKUk5OTmem5sbRr0n3NTEDv+G/X4BTcGyZcsYKwgN3ysAGoswpxXtkNSt0nJX\nnTxtaLykuZLk7m9KaimpfYg1AQAAAKhBmOFghaQeZpZqZs0ljZW0oEqbbZIulSQz663ycLAnxJoA\nAAAA1CC0cODuJZImSVokab3Kr0q01sweNLNRQbO7JX3HzFZLel7STe5edeoRAAAAgHoQ5jkHcveF\nKr88aeV191V6vE7S4DBrAAAAABAZ7pAMAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS\n4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLh\nAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAECSFBftAgA0XilTXo52CXXm6RGtol0CAABRx5EDAAAA\nAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAA\nkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACS\nCAcAAAAAAoQDAAAAAJIIBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJII\nBwAAAAAChAMAAAAAkggHAAAAAAKEAwAAAACSCAcAAAAAAoQDAAAAAJJCDgdmNsLM8s1sk5lNqaHN\ndWa2zszWmtmfwqwHAAAAQM3iwurYzGIl/VbSZZJ2SFphZgvcfV2lNj0k/bukwe6+38w6hFUPAAAA\ngNqFeeRggKRN7r7Z3Y9Lmi1pdJU235H0W3ffL0nuvjvEegAAAADUIrQjB5K6SNpeaXmHpIFV2qRJ\nkpm9LilW0gPu/krVjsxsoqSJkpScnKxly5aFUe8Jd/cpCbX/+lRYWBj6+4UzF2MFiAxjBUBjEWY4\nsGrWeTX77yEpV1JXSa+aWaa7H/jck9xnSpopSTk5OZ6bm1vnxVZ205SXQ+2/Pj09opXCfr9w5mKs\nAJFhrABoLMKcVrRDUrdKy10l7aymzV/cvdjdP5KUr/KwAAAAAKCehRkOVkjqYWapZtZc0lhJC6q0\nmS9pmCSZWXuVTzPaHGJNAAAAAGoQWjhw9xJJkyQtkrRe0lx3X2tmD5rZqKDZIkmfmdk6SUsl/dDd\nPwurJgAAAAA1C/OcA7n7QkkLq6y7r9JjlzQ5+A8AAABAFHGHZAAAAACSCAcAAAAAAoQDAAAAAJJC\nPucADcAnq6QHqt6YupF6oCDaFQAAADRpHDkAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAA\nACCJcAAAAAAgQDgAAAAAIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgQDgAAAAA\nIIlwAAAAACBAOAAAAAAgiXAAAAAAIEA4AAAAACDpNMKBmQ0xs5uDx0lmlhpeWQAAAADqW0ThwMzu\nl3SvpH8PVjWT9FxYRQEAAACof5EeOfiGpFGSDkuSu++U1CasogAAAADUv0jDwXF3d0kuSWbWKryS\nAAAAAERDpOFgrpnNkHS2mX1H0t8kPRVeWQAAAADqW1wkjdz9MTO7TNJBST0l3efuS0KtDAAAAEC9\nOmU4MLNYSYvcfbgkAgEAAADQRJ1yWpG7l0o6YmaJ9VAPAAAAgCiJaFqRpGOS1pjZEgVXLJIkd78j\nlKoAAAAA1LtIw8HLwX8AAAAAmqhIT0h+xsyaS0oLVuW7e3F4ZQEAAACobxGFAzPLlfSMpC2STFI3\nM/tXd18eXmkAAAAA6lOk04qmSrrc3fMlyczSJD0vKTuswgAAAADUr0hvgtasIhhIkrtvkNQsnJIA\nAAAAREOkRw7yzOy/Jf0xWP6WpJXhlAQAAAAgGiINB7dJ+p6kO1R+zsFySdPDKgoAAABA/Ys0HMRJ\n+o27/0o6cdfkFqFVBQAA0EStXLmyQ1xc3O8lZSryKd5AXSiT9H5JScmE7Ozs3dU1iDQc/F3ScEmF\nwXK8pMWSBn3pEgEAAM4gcXFxv+/YsWPvpKSk/TExMR7tenDmKCsrsz179qTv2rXr95JGVdcm0rTa\n0t0rgoGCxwl1UCMAAMCZJjMpKekgwQD1LSYmxpOSkgpUftSq+jYR9nXYzC6oWDCzHElHv2R9AAAA\nZ6IYggGiJfjs1ZgBIp1WdKekeWa2U5JL6izp+i9fHgAAAICGItJwkCqpv6RzJX1D0ldVHhIAAADw\nJaRMeblObyq75ZdfbzCXmx8wYEDPxx57bPvXvva1I5G0Lysr06BBg9IWLly4qW3btmVh1/dl/fWv\nf20zderU5KVLl24Ko/9Zs2Ylrl27Nv4Xv/jFrqrbEhIS+h85cuTdmp47aNCgtL/85S8fJiUllZ7O\nPiOdVvQTdz8o6WxJl0maKel3p7MjAAAANB7FxcX1vs+5c+cmZmRkHG0MwaA+fOtb3yqoLhhEYty4\ncZ899thjSaf7vEjDQUXi+LqkJ939L5Kan+7OAAAAEH0//OEPO6WmpmYMGjSox1VXXZV63333JUvl\nv/RPmjSpy4UXXtjzoYceSt65c2fcFVdccV5mZmbvzMzM3osXL24lSQcPHoy59tprUzIzM3v37t07\n/bnnnjtbkgoLC+3KK6/8SlpaWvrXv/71rxw7dswk6fHHH28/fvz4bhX7nzp1avsJEyZ0rVrXrFmz\n2n7jG984IEn5+fnNU1NTM66//vruPXr0yBg1alTq/Pnz21xwwQW9unfvnrl06dKE2mrJz89vnp2d\n3TM9Pb13enp67yVLlrSSyn/tHzBgQM8RI0Z8JTU1NWPUqFGpZWW1Z5ExY8akfPOb3zw3Ozu7Z0pK\nSubzzz+fWLXN0qVLE/r379+rd+/e6f379++1evXqFpKUl5fXsk+fPr179eqVnpaWlr5mzZoWkb62\nadOmtfv2t799riR98MEHzfv169crMzOz95133tm5Yr9bt25tlpOT07NXr17pPXr0yHjllVdaS9LY\nsWMPvPjii+0i/EicEOm0oo/NbIbKL2f6iJm1ENflBQAAaHSWL1+e8D//8z/nrFmzZl1xcbH169cv\nvX///iem/Rw4cCB2xYoV+ZJ01VVXpU6ePPnTK664onDjxo3Nr7jiih6bN29e+6Mf/ajTsGHDDs6b\nN2/L3r17Y3NycnqPGjXq4K9+9auk+Pj4sg0bNqz75z//GT948OB0SRo/fvy+jIyM9KKioh0tWrTw\n5557rv2MGTO2Vq1t5cqVrQcPHnxi/fbt21vOmTNnc3Z29ta+ffv2njVrVru8vLwP/vSnP53985//\nvNOwYcM+rKmWzp07l7z66qsbEhISfM2aNS3GjRv3lffff3+9JK1fvz5+1apVm1NSUoqzs7N7LVmy\npPUVV1xRWLWeyrZv397i7bffzl+3bl2L4cOH9xw9evSaytuzsrKOvf322x80a9ZM8+fPb3PPPfd0\nXbRo0YdPPPFE0u233/7pbbfdtu/YsWNWUlKijz/+uFkkr61y/7fffvu5EyZM2DNp0qTPHn744RNH\nBP7whz+0vfTSSwseeeSRXSUlJTp06FCMJCUlJZUeP37cdu3aFduxY8eIpxZFGg6ukzRC0mPufsDM\nOkn6YaQ7AQAAQMOwbNmy1iNHjjzQunVrl+SXXXbZgcrbx40bt6/i8euvv37Wxo0b4yuWCwsLY/fv\n3x+zbNmysxYtWnT2tGnTOkpSUVGRbdq0qflrr73W+o477tgtSQMHDjyalpZ2RJLOOuusssGDBx+a\nM2dOYp8+fY4VFxfbgAEDTrryZUFBQdw555xz4mf8Ll26FFW0S0tLO3rJJZccjImJ0QUXXHDkoYce\n6hy8nmpr6d69e/H48eO7r1u3Lj4mJkZbt249cQPfPn36HD7vvPOKJSkjI+PIhx9+eMoZMWPGjNkX\nGxurPn36FHXr1q1o1apVLStv37dvX+z111+fumXLlpZm5sXFxSZJF1100eHHHnus044dO5qPHTt2\nf58+fYoifW2VvfPOO63/93//90NJuvXWWz/72c9+1lWSvvrVrx6+9dZbU4qLi2Ouueaa/YMGDTrx\nvrZr165k27ZtzTt27BjxVUYjCgfufkTSi5WWP5H0SaQ7AQAAQMPgXvs1Zdq0aXPij3N3V15e3vog\nSHyujxdeeGFTVlZWUdXnm1m1/U6cOHHvz3/+845paWnHbrjhhr3VtYmNjfXS0lLFxsZKkpo3b35i\nvzExMWrZsqUH7VRaWmq11TJ58uTOHTp0KP7zn//8UVlZmeLj40+c+N2iRYsT/cbGxqqkpKT6omt5\nXVWX77333i5Dhw49tGTJkg/z8/ObX3LJJT0l6bvf/e6+iy+++PBLL72UOHLkyLTp06dv6dmzZ1Ek\nr62q6i6BO3LkyMLly5fn//nPf0686aabUu+4445PJ02a9JlUHpQSEhJO6/wNpgYBAACcQXJzcwsX\nLVqUeOTIESsoKIj529/+dnZNbYcMGXLwkUce6VCx/MYbb8RL0rBhww5OnTo1uWKu/uuvvx4ftC98\n7rnn2krSihUrWm7YsOHETXMvueSSw5988knzl156qd348eP3qRqpqanH1q9f36K6bTWpqZaCgoLY\nTp06FcfGxmr69OntSktPPbPme9/7Xpdnn3222vfjxRdfPKe0tFRr165tsX379hZZWVnHKm8/ePBg\nbNeuXY9L0owZM9pXrF+3bl3z3r17F/34xz/effnllx9YtWpVfNW+I3HBBRcUPvXUU20l6amnnjpx\nLsGGDRuad+nSpfjuu+/ee8MNN+x95513EqTyKz/t2bOnWc+ePU8KcLWJdFrRF2JmIyT9RlKspN+7\n+y9raHeNpHmSLnT3vDBrAgAAaEjq+9KjQ4cOPTJixIiC9PT0jC5duhT17dv3cGJiYrV/Oc+cOXP7\nhAkTzk1LS0svLS21gQMHHho0aNC2X/7ylzsnTpx4bq9evdLd3bp27Vq0dOnSTT/4wQ92jx07NjUt\nLS09IyPjSJ8+fQ5X7u/qq6/e/9577yXUdHnNyy+/vGDx4sVtMjMzI/6DtqZa7rrrrt1jxow5b/78\n+ecMGTLkUHx8/Cl/QV+3bl18xQnRVZ1//vlFAwYM6PnZZ581+/Wvf701ISHhc7/i33vvvbsmTJiQ\nOm3atI4XX3zxwYr1f/zjH9vOmzevXVxcnCclJRU//PDDOw8cOBAb6eurMH369G1jx479yvTp05NH\njRq1v2L9okWL2kybNq1jXFycJyQklM6aNesjSXrttdcS+vfvf7hZs2antR871aGlL8rMYiVtUPml\nT3dIWiFpnLuvq9KujaSXVX71o0mnCgc5OTmelxdufkiZ8nKo/denp7M2Kjf//miXUTceKIh2Baii\nSY2VEa2Um5sb7TLQRDFWzlxmttLdcyqvW7169ZasrKxqp9XUl4KCgpjExMSyQ4cOxVx00UU9n3zy\nya1DhgyJ6F4EX8awYcPOv+uuuz4dPXr0oeq2b926tdm4ceNS3njjjY1h11KdIUOG9HjttddO2veY\nMWNSrrzyyoKbb755f3XPa4huvvnmbldfffWB6t7r1atXt8/Kykqp7nlhTisaIGmTu2929+OSZksa\nXU27n0l6VNKxarYBAACgjt1www3de/Xqld63b9/eV1111f6wg8HevXtjU1JSMlu2bFlWUzCQpO7d\nuxffcsste/ft2xeVqe/VBYPGKjMz82ht73VNwjxycI2kEe4+IVi+UdJAd59UqU1/ST929zFmtkzS\nD6o7cmBmEyVNlKTk5OTs2bNnh1JzhTUfN51fqFPji9S6aGe0y6gbnfpFuwJU0aTGSmKsWrduHe0y\n0EQxVs5cw4YNa5BHDnBmq+3IQZjnHFR3lvWJJGJmMZIel3TTqTpy95kqvyuzcnJyPOzDmTc1pcO/\nWR81nWlF45rOl2tT0aTGClMlECLGCoDGIsxDNjskdau03FVS5Z+w20jKlLTMzLZI+qqkBWb2uXQN\nAAAAoH6EGQ5WSOphZqlm1lzSWEkLKja6e4G7t3f3FHdPkfSWpFFcrQgAAACIjtDCgbuXSJokaZGk\n9ZLmuvtaM3vQzEaFtV8AAAAAX0yo9zlw94WSFlZZd18NbXPDrAUAANSBT1ZJD1R38cFGqKFcIvuB\nxOxTNzqd/gpOed8EM8sePXr0vvnz538kScXFxerQoUNWv379Di9dunTTtGnT2uXl5bV69tlnt02e\nPLlz69atSx988MFPK/cRGxub3aNHj6OlpaV2/vnnH507d+6WyndXlspvxDVo0KC0hQsXbmrbtm3Z\ntddem/L3v/89sV27diUbN25cW11t7777bsubb745Zd26dQlTpkz5uOp+K/ziF79IevLJJ5O3b9/e\nYufOnas7depUIknPP/984ttvv93q8ccfbyJXZKlf3CEZAADgDBMfH1+Wn58fX1hYaJL00ksvnZWc\nnFx8On20aNGi7IMPPli3cePGtc2aNfOpU6cmVW0zd+7cxIyMjKNt27Ytk6Rbbrll74IFC2q9XGiH\nDh1KfvOb32y79dZbqw0FFYYOHVq4ZMmSDZ07dz5eef31119f8Morr5x96NAh/s79AnjTAAAAzkCX\nXnppwbx5886WpOeff77tmDFj9n3RvoYMGVK4adOmFlXXz5o1q23lOw6PHDmyMCkpqaS2vrp06VIy\ndOjQI82aNav1evuDBw8+2rNnz+NV18fExGjQoEGH5syZk3g6rwHlCAcAAABnoBtvvHHfnDlzzjly\n5IitX78+4aKLLjr8RfopLi7WokWLzurTp8/RqttWrlzZevDgwafs99FHH0169NFHTzryUNXQoUPP\n37JlS7NTtcvJyTn86quvckOOLyDUcw4AAADQMA0cOPDojh07Wjz11FNthw8fftonYBQVFcX06tUr\nPejr0J133nnSjd0KCgrizjnnnLKTn/1599xzz55I9vmPf/xjUyTtOnbsWLJr167mkbTF5xEOAAAA\nzlAjRow4cP/993dbvHhx/u7du0/r78KKcw5qaxMbG+ulpaWKjY39coWepqNHj1rLli1PGUpwMqYV\nAQAAnKFuu+22vXfffffOAQMGnDQlqC6kpqYeW79+/UnnIoQtPz+/ZUZGRiivqanjyAEAAEA0RXDp\n0bCcd955xT/5yU92n6rd448/3mnGjBnJFcuffvrpe5H0f/nllxcsXry4TWZmZpEkXXXVValvvfVW\nm/3798clJyf3nTJlys7vf//7eyvON7jnnnv2bNu2Le7CCy9MP3z4cKyZ+YwZM5LXr1//ftu2bcuG\nDh16/jPPPLM1JSWl+KGHHurwxBNPdPzss8+aZWVlpQ8bNqxgzpw5WyVp+fLlbR555JGPv9i7cmYz\n91pPBG9wcnJyPC8v3Jsop0x5OdT+69PTWRuVm39/tMuoGw3letQ4oUmNlRGtlJubG+0y0EQ1qbHC\n98ppMbOV7p5Ted3q1au3ZGVlnTQ/vynaunVrs3HjxqW88cYbtV6+tC5t37497rrrrvvKm2++uaG+\n9tnYrF69un1WVlZKdduYVgQAAIBQdO/evfiWW27Zu2/fvnr7m3Pz5s3Np06dur2+9tfUMK0IAAAA\noZkwYcL++tzf0KFDj9Tn/poajhwAAADUr7KysjKLdhE4MwWfvRqv5EQ4AAAAqF/v79mzJ5GAgPpW\nVlZme/bsSZT0fk1tmFYEAABQj0pKSibs2rXr97t27coUP9SifpVJer+kpGRCTQ0IBwAAAPUoOzt7\nt6RR0a4DqA5pFQAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAk\nwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgEBctAsAgAbhk1XSA6OjXUXdeKAg2hUA\nABopjhwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAA\nABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAA\nECAcAAAAAJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAEAg1HJjZCDPLN7NNZjal\nmu2TzWydmb1nZn83s+5h1gMAAACgZqGFAzOLlfRbSSMlpUsaZ2bpVZq9KynH3ftKekHSo2HVAwAA\nAKB2YR45GCBpk7tvdvfjkmZLGl25gbsvdfcjweJbkrqGWA8AAACAWpi7h9Ox2TWSRrj7hGD5RkkD\n3X1SDe3/S9Iud3+omm0TJU2UpOTk5OzZs2eHUnOFNR8XhNp/fUqNL1Lrop3RLqNudOoX7QpQBWOl\ngWKsNDiMlQaqHsbKsGHDVrp7Tug7AupIXIh9WzXrqk0iZnaDpBxJQ6vb7u4zJc2UpJycHM/Nza2j\nEqt305SXQ+2/Pj2d9ZFy8++Pdhl1Y1zT+XJtKhgrDRRjpcFhrDRQjBXgJGGGgx2SulVa7irppJ8a\nzGy4pP+QNNTdi0KsBwAAAEAtwjznYIWkHmaWambNJY2VtKByAzPrL2mGpFHuvjvEWgAAAACcQmjh\nwN1LJE2StEjSeklz3X2tmT1oZqOCZv8pqbWkeWa2yswW1NAdAAAAgJCFOa1I7r5Q0sIq6+6r9Hh4\nmPsHAAAAEDnukAwAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABA\nEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS\n4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLh\nAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEA\nAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAA\nAABAgHAAAAAAQBLhAAAAAECAcAAAAABAEuEAAAAAQIBwAAAAAEAS4QAAAABAgHAAAAAAQBLhAAAA\nAEAg1HBgZiPMLN/MNpnZlGq2tzCzOcH2f5pZSpj1AAAAAKhZaOHAzGIl/VbSSEnpksaZWXqVZuMl\n7Xf38yU9LumRsOoBAAAAULswjxwMkLTJ3Te7+3FJsyWNrtJmtKRngscvSLrUzCzEmgAAAADUIMxw\n0EXS9krLO4J11bZx9xJJBZLahVgTAAAAgBrEhdh3dUcA/Au0kZlNlDQxWCw0s/wvWdsZY5jUXtLe\naNdRJ37KQSWEh7ECRIaxctq618dOgLoSZjjYIalbpeWuknbW0GaHmcVJSpS0r2pH7j5T0syQ6mzS\nzCzP3XOiXQfQ0DFWgMgwVoCmLcxpRSsk9TCzVDNrLmmspAVV2iyQ9K/B42sk/T93P+nIAQAAAIDw\nhXbkwN1LzGySpEWSYiX9wd3XmtmDkvLcfYGk/5b0RzPbpPIjBmPDqgcAAABA7Ywf6ps2M5sYTMsC\nUAvGChAZxgrQtBEOAAAAAEgK+Q7JAAAAABoPwgEAAAAASYSDM56ZpZjZ+8HjXDP7a7RrAqLBzN44\nxfaFZnZ2fdUDNAVmtszMcoLHW8ysfbRrAlC7MO9zgBCZman8nJGyaNcCNDRmFuvupafzHHcfdIrt\n//LlqgKaHr6LgKaHIweNSPAr/3ozmy7pHUk3mtmbZvaOmc0zs9ZBuwvN7A0zW21mb5tZm+C5rwZt\n3zGzWv8QAhqq4LP8gZk9Y2bvmdkLZpYQ/Cp5n5m9JulaMzvPzF4xs5XBZ79X8PxkM3spGB+rK8aC\nmRUG/3Yys+VmtsrM3jezi4NHyCaSAAAD3klEQVT1J371NLPJwbb3zeyuSnWtN7OnzGytmS02s/io\nvElALap+fs3sETO7vdL2B8zs7uDxD81sRTDWfhqsq/pd1M3MfmdmecFn/6fReWUA6gLhoPHpKelZ\nSZdJGi9puLtfIClP0uTghnNzJN3p7lmShks6Kmm3pMuCttdLmhaN4oE60lPSTHfvK+mgpIo/bI65\n+xB3n63yu6r/m7tnS/qBpOlBm2mS/hGMjwskra3S9zclLXL3fpKyJK2qvNHMsiXdLGmgpK9K+o6Z\n9Q8295D0W3fPkHRA0pi6esFAXaju8ytptsq/FypcJ2memV2u8s/0AEn9JGWb2deCNj0lPevu/d19\nq6T/CO6a3FfSUDPrWy8vCECdY1pR47PV3d8ysyslpUt6vfyorppLelPl/8P+xN1XSJK7H5QkM2sl\n6b/MrJ+kUklp0SgeqCPb3f314PFzku4IHs+RpOAo2iCV/4FT8ZwWwb+XSPq2JAVTjwqq9L1C0h/M\nrJmk+e6+qsr2IZJecvfDwb5elHSxyu/4/lGl9islpXyJ1wiEoabPbwcz6ywpSdJ+d99mZndIulzS\nu8FzW6s8LGxT8F1Uqd/rzGyiyv+u6KTy76f36uMFAahbhIPG53Dwr0la4u7jKm8Mfq2p7uYV35f0\nqcp/CY2RdCzMIoGQVf2MVyxXjI8YSQeCX/9Pr2P35cGvo19X+R3c/9Pdn63UxGp4qiQVVXpcKolp\nRWhoavr8viDpGkkdVX4koaLtw+4+43MdmKXo/8aazCxV5UfnLnT3/Wb2tKSWdVo1gHrDtKLG6y1J\ng83sfEkK5lynSfpAUmczuzBY38bM4iQlqvyIQpmkGyXFRqluoC6ca2YXBY/HSXqt8sbgiNlHZnat\nVH7SpJllBZv/Lum2YH2smZ1V+blm1l3Sbnd/StJ/q3zqUWXLJV0djLlWkr4h6dW6e2lAqGr6/M6W\nNFblAeGFoO0iSbdUOp+ti5l1qKbPs1QeFgrMLFnSyJBfA4AQEQ4aKXffI+kmSc+b2XsqDwu93P24\nyueOPmFmqyUtUfkvONMl/auZvaXyKUWHq+0YaBzWq/zz/J6ktpJ+V02bb0kaH4yDtZJGB+vvlDTM\nzNaofOpPRpXn5UpaZWbvqvycgd9U3uju70h6WtLbkv4p6ffu/q6ARqCmz6+7r5XURtLH7v5J0Hax\npD9JejMYLy8Ebar2uVrlU4/WSvqDpNertgHQeJh7dTNQAKBhCqY0/NXdM6NcCgAATQ5HDgAAAABI\n4sgBAAAAgABHDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASJL+P2RpS8vfNA2DAAAA\nAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f41222c3a58>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:1:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>101525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>172426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>62845</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0       101525\n",
       "1  plasmid_1       172426\n",
       "2  plasmid_2        62845"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.568562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.288667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.382920</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.568562\n",
       "1  precision  0.288667\n",
       "2   f1_score  0.382920"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP011579.1</th>\n",
       "      <th>CP011580.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.084885</td>\n",
       "      <td>0.256147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.917733</td>\n",
       "      <td>0.220028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.917495</td>\n",
       "      <td>0.110092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Total</td>\n",
       "      <td>1.920113</td>\n",
       "      <td>0.586267</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP011579.1  CP011580.1\n",
       "0  plasmid_0    0.084885    0.256147\n",
       "1  plasmid_1    0.917733    0.220028\n",
       "2  plasmid_2    0.917495    0.110092\n",
       "3      Total    1.920113    0.586267"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP011579.1</td>\n",
       "      <td>0.027727</td>\n",
       "      <td>0.178668</td>\n",
       "      <td>0.490079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP011580.1</td>\n",
       "      <td>0.191894</td>\n",
       "      <td>0.087875</td>\n",
       "      <td>0.083793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.219621</td>\n",
       "      <td>0.266543</td>\n",
       "      <td>0.573872</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2\n",
       "0  CP011579.1   0.027727   0.178668   0.490079\n",
       "1  CP011580.1   0.191894   0.087875   0.083793\n",
       "2       Total   0.219621   0.266543   0.573872"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.191894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.254097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.539852</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.191894\n",
       "1  plasmid_1    0.254097\n",
       "2  plasmid_2    0.539852"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP011579.1</td>\n",
       "      <td>1.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP011580.1</td>\n",
       "      <td>0.40825</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP011579.1     1.00000\n",
       "1  CP011580.1     0.40825"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No high mutual correspondence found.\n"
     ]
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 1, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0126137</td>\n",
       "      <td>-0.0271062</td>\n",
       "      <td>-0.0430691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[25, 39, 41, 35, 60, 54, 51, 64, 28, 80, 46, 7...</td>\n",
       "      <td>[36, 64, 51, 59, 55, 65, 75, 16, 78, 70, 68, 7...</td>\n",
       "      <td>[28, 64, 51, 59, 55, 65, 44, 50, 34, 46, 80, 4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.428825</td>\n",
       "      <td>-0.156986</td>\n",
       "      <td>-0.414631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.441439</td>\n",
       "      <td>-0.184092</td>\n",
       "      <td>-0.4577</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0  \\\n",
       "GC_sum                                         -0.0126137   \n",
       "chain   [25, 39, 41, 35, 60, 54, 51, 64, 28, 80, 46, 7...   \n",
       "gd_sum                                          -0.428825   \n",
       "total                                           -0.441439   \n",
       "\n",
       "                                                plasmid_1  \\\n",
       "GC_sum                                         -0.0271062   \n",
       "chain   [36, 64, 51, 59, 55, 65, 75, 16, 78, 70, 68, 7...   \n",
       "gd_sum                                          -0.156986   \n",
       "total                                           -0.184092   \n",
       "\n",
       "                                                plasmid_2  \n",
       "GC_sum                                         -0.0430691  \n",
       "chain   [28, 64, 51, 59, 55, 65, 44, 50, 34, 46, 80, 4...  \n",
       "gd_sum                                          -0.414631  \n",
       "total                                             -0.4577  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_11</th>\n",
       "      <th>plasmid_12</th>\n",
       "      <th>plasmid_13</th>\n",
       "      <th>plasmid_14</th>\n",
       "      <th>plasmid_15</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.0117709</td>\n",
       "      <td>-0.04946</td>\n",
       "      <td>0.00381224</td>\n",
       "      <td>-0.000874392</td>\n",
       "      <td>-9.22911e-05</td>\n",
       "      <td>-0.00316856</td>\n",
       "      <td>-0.0514102</td>\n",
       "      <td>0.00313343</td>\n",
       "      <td>0.00517664</td>\n",
       "      <td>-0.0426786</td>\n",
       "      <td>0.00656828</td>\n",
       "      <td>-0.0286681</td>\n",
       "      <td>-0.0355077</td>\n",
       "      <td>-0.0109166</td>\n",
       "      <td>-0.000429119</td>\n",
       "      <td>-0.015556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[26, 49, 10, 51, 54, 60, 35, 41, 39, 25, 41, 3...</td>\n",
       "      <td>[60, 65, 36, 64, 51, 54, 60, 35, 41, 39, 25, 41]</td>\n",
       "      <td>[43, 62, 6, 47, 11, 49]</td>\n",
       "      <td>[64, 32, 47, 5, 58, 70, 68, 72, 66]</td>\n",
       "      <td>[77, 23, 47, 9, 48, 2]</td>\n",
       "      <td>[33, 7, 47]</td>\n",
       "      <td>[64, 32, 47]</td>\n",
       "      <td>[53, 18, 48, 1]</td>\n",
       "      <td>[69, 27, 46, 31, 50, 44, 65, 55, 59, 51, 64, 2...</td>\n",
       "      <td>[33, 33, 40, 65, 55, 59, 51, 64, 28]</td>\n",
       "      <td>[43, 13, 46, 79, 37, 46]</td>\n",
       "      <td>[42, 39]</td>\n",
       "      <td>[46, 79]</td>\n",
       "      <td>[75, 81, 55, 65, 75, 16, 78, 70, 68, 82, 74, 66]</td>\n",
       "      <td>[55, 65, 4, 69, 78, 70, 68, 72, 66]</td>\n",
       "      <td>[61, 81, 55]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.113725</td>\n",
       "      <td>-0.398064</td>\n",
       "      <td>-0.0540494</td>\n",
       "      <td>-0.0113242</td>\n",
       "      <td>-0.0012935</td>\n",
       "      <td>-0.00278706</td>\n",
       "      <td>-0.057069</td>\n",
       "      <td>-0.000704025</td>\n",
       "      <td>-0.362927</td>\n",
       "      <td>-0.226833</td>\n",
       "      <td>-0.0248976</td>\n",
       "      <td>-0.379273</td>\n",
       "      <td>-0.447485</td>\n",
       "      <td>-0.00531228</td>\n",
       "      <td>-0.011377</td>\n",
       "      <td>-0.310843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.125496</td>\n",
       "      <td>-0.447524</td>\n",
       "      <td>-0.0502372</td>\n",
       "      <td>-0.0121986</td>\n",
       "      <td>-0.00138579</td>\n",
       "      <td>-0.00595562</td>\n",
       "      <td>-0.108479</td>\n",
       "      <td>0.0024294</td>\n",
       "      <td>-0.357751</td>\n",
       "      <td>-0.269511</td>\n",
       "      <td>-0.0183293</td>\n",
       "      <td>-0.407941</td>\n",
       "      <td>-0.482993</td>\n",
       "      <td>-0.0162289</td>\n",
       "      <td>-0.0118062</td>\n",
       "      <td>-0.326399</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0  \\\n",
       "GC_sum                                         -0.0117709   \n",
       "chain   [26, 49, 10, 51, 54, 60, 35, 41, 39, 25, 41, 3...   \n",
       "gd_sum                                          -0.113725   \n",
       "total                                           -0.125496   \n",
       "\n",
       "                                               plasmid_1  \\\n",
       "GC_sum                                          -0.04946   \n",
       "chain   [60, 65, 36, 64, 51, 54, 60, 35, 41, 39, 25, 41]   \n",
       "gd_sum                                         -0.398064   \n",
       "total                                          -0.447524   \n",
       "\n",
       "                     plasmid_10                           plasmid_11  \\\n",
       "GC_sum               0.00381224                         -0.000874392   \n",
       "chain   [43, 62, 6, 47, 11, 49]  [64, 32, 47, 5, 58, 70, 68, 72, 66]   \n",
       "gd_sum               -0.0540494                           -0.0113242   \n",
       "total                -0.0502372                           -0.0121986   \n",
       "\n",
       "                    plasmid_12   plasmid_13    plasmid_14       plasmid_15  \\\n",
       "GC_sum            -9.22911e-05  -0.00316856    -0.0514102       0.00313343   \n",
       "chain   [77, 23, 47, 9, 48, 2]  [33, 7, 47]  [64, 32, 47]  [53, 18, 48, 1]   \n",
       "gd_sum              -0.0012935  -0.00278706     -0.057069     -0.000704025   \n",
       "total              -0.00138579  -0.00595562     -0.108479        0.0024294   \n",
       "\n",
       "                                                plasmid_2  \\\n",
       "GC_sum                                         0.00517664   \n",
       "chain   [69, 27, 46, 31, 50, 44, 65, 55, 59, 51, 64, 2...   \n",
       "gd_sum                                          -0.362927   \n",
       "total                                           -0.357751   \n",
       "\n",
       "                                   plasmid_3                 plasmid_4  \\\n",
       "GC_sum                            -0.0426786                0.00656828   \n",
       "chain   [33, 33, 40, 65, 55, 59, 51, 64, 28]  [43, 13, 46, 79, 37, 46]   \n",
       "gd_sum                             -0.226833                -0.0248976   \n",
       "total                              -0.269511                -0.0183293   \n",
       "\n",
       "        plasmid_5  plasmid_6  \\\n",
       "GC_sum -0.0286681 -0.0355077   \n",
       "chain    [42, 39]   [46, 79]   \n",
       "gd_sum  -0.379273  -0.447485   \n",
       "total   -0.407941  -0.482993   \n",
       "\n",
       "                                               plasmid_7  \\\n",
       "GC_sum                                        -0.0109166   \n",
       "chain   [75, 81, 55, 65, 75, 16, 78, 70, 68, 82, 74, 66]   \n",
       "gd_sum                                       -0.00531228   \n",
       "total                                         -0.0162289   \n",
       "\n",
       "                                  plasmid_8     plasmid_9  \n",
       "GC_sum                         -0.000429119     -0.015556  \n",
       "chain   [55, 65, 4, 69, 78, 70, 68, 72, 66]  [61, 81, 55]  \n",
       "gd_sum                            -0.011377     -0.310843  \n",
       "total                            -0.0118062     -0.326399  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_5  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n",
      "plasmid_6  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_6  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_6  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
