{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance of MILP on sample 39 with $\\alpha_1 = 1$, $\\alpha_2 = 10$ and $\\alpha_3 = 1$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import linecache\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_structure(filename):\n",
    "    with open(filename) as file:\n",
    "        lines = file.readlines()\n",
    "        for i, line in enumerate(lines):\n",
    "            # number of reference chromosomes:\n",
    "            if line.startswith('number of reference c'):\n",
    "                skip_ref_chromosomes = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of reference plasmids:\n",
    "            if line.startswith('number of reference p'):\n",
    "                skip_ref_plasmids = i + 1\n",
    "                continue\n",
    "                \n",
    "            # number of predicted plasmids:\n",
    "            if line.startswith('number of p'):\n",
    "                skip_pred_plasmids = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > predicted plasmid covers <proportion> of reference plasmid\n",
    "            if line.startswith('> pr'):\n",
    "                skip_ref_by_pred = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > reference plasmid covers <proportion> of predicted plasmid\n",
    "            if line.startswith('> r'):\n",
    "                skip_pred_by_ref = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of predicted plasmid is covered by reference plasmids\n",
    "            if line.startswith('> in total, how much of p'):\n",
    "                skip_pred_by_all = i + 1\n",
    "                continue\n",
    "                \n",
    "            # > in total, how much of reference plasmid is covered by predicted plasmids\n",
    "            if line.startswith('> in total, how much of r'):\n",
    "                skip_ref_by_all = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > pairs of predicted and reference plasmids with coverage >= ... in both directions\n",
    "            if line.startswith('> pa'):\n",
    "                skip_mutual = i + 1\n",
    "                continue\n",
    "            \n",
    "            # > summary scores\n",
    "            if line.startswith('> s'):\n",
    "                skip_summary_scores = i + 1\n",
    "                break\n",
    "        \n",
    "        return skip_ref_chromosomes, skip_ref_plasmids, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, \\\n",
    "                skip_pred_by_all, skip_ref_by_all, skip_mutual, skip_summary_scores\n",
    "\n",
    "def read_table(file, skip_rows, num_rows, h = 0, s = '\\t', col_sums = False, renaming = {}):\n",
    "    df = pd.read_csv(file, skiprows = skip_rows, nrows = num_rows, sep = s, header = h)\n",
    "    if col_sums:\n",
    "        df.loc[len(df.index)] = ['Total'] + list(df.drop(df.columns[0], axis = 1).sum())\n",
    "    df.rename(columns = renaming, inplace = True)\n",
    "    return df\n",
    "\n",
    "def list_plasmids(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'plasmid', 1: 'length (nt)'})\n",
    "    df.plasmid = [v[1:-1] for v in df.plasmid]\n",
    "    return df[['plasmid', 'length (nt)']]\n",
    "\n",
    "def list_scores(file, skip_rows, num_rows):\n",
    "    df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'score', 1: 'value'})\n",
    "    df['score'] = [v[:-1].replace('score_ref_coverage', 'recall').replace('score_pred_coverage', 'precision').replace('overall_score', 'overall') for v in df['score']]\n",
    "    return df\n",
    "\n",
    "def list_correspondences(file, skip_rows, num_rows):\n",
    "    first_line = linecache.getline(file, skip_rows + 1).strip()\n",
    "    if first_line == 'none':\n",
    "        print('No high mutual correspondence found.')\n",
    "        #return pd.DataFrame({'reference': [], 'prediction': [], 'recall': [], 'precision': []})\n",
    "    else:        \n",
    "        df = read_table(file, skip_rows, num_rows, h = None, s = ' ', renaming = {0: 'reference', 2: 'prediction', 3: 'recall', 5: 'precision'})\n",
    "        df['recall'] = [float(v[1:]) for v in df['recall']]\n",
    "        df['precision'] = [float(v[:-1]) for v in df['precision']]\n",
    "        return df[['reference', 'prediction', 'recall', 'precision']]\n",
    "    \n",
    "def plot_scores(information, score_labels):\n",
    "    # information = list of tuples of from (file, label)\n",
    "    pos = list(range(len(score_labels)))\n",
    "    width = 1 / (len(information) + 2)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (10, 5))\n",
    "\n",
    "    offset = -len(information) / 2\n",
    "    for file, lab in information:\n",
    "        plt.bar([p + offset * width for p in pos], list_scores(file, parse_structure(file)[-1], len(score_labels))['value'], width, label = lab)\n",
    "        offset += 1\n",
    "\n",
    "    ax.set_ylabel('score')\n",
    "    ax.set_xticks([p - 0.5 * width for p in pos])\n",
    "    ax.set_xticklabels(score_labels)\n",
    "    ax.legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "    plt.grid()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Precision*: proportion of predicted plasmid covered by reference plasmids (aka score_pred_coverage).\n",
    "\n",
    "*Recall*: proportion of reference plasmid covered by predicted plasmids (aka score_ref_coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "score_labels = ['recall', 'precision', 'overall']\n",
    "\n",
    "# evaluation files\n",
    "sample_id = 39\n",
    "MILP = \"nplasmids_1/eval/MILP/MILP_eval.csv\"\n",
    "greedy_mean = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/%i_greedy_mean.csv\" % (sample_id, sample_id)\n",
    "\n",
    "# reference information\n",
    "_, skip_ref_plasmids, skip_pred_plasmids, _, _, _, _, _, _ = parse_structure(greedy_mean)\n",
    "num_ref_plasmids = skip_pred_plasmids - skip_ref_plasmids - 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP018949.1</td>\n",
       "      <td>16559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP018951.1</td>\n",
       "      <td>43380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP018952.1</td>\n",
       "      <td>132345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP018950.1</td>\n",
       "      <td>35225</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  length (nt)\n",
       "0  CP018949.1        16559\n",
       "1  CP018951.1        43380\n",
       "2  CP018952.1       132345\n",
       "3  CP018950.1        35225"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(greedy_mean, skip_ref_plasmids, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview of the scores on this sample for the different tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwcAAAEyCAYAAACvcKx4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3Xt8FeWdx/HvLwkEAogCIdxJVAKE\nQLgEUMASFCW0AnbxAopdFYqrZb2AVXa3VWu1VldqS1daoOtaKlUBldKKBeqSeqsVUJBrICo3EQGB\nYEAgl9/+kQkbQxIONZME8nm/Xrw8M/OcZ37n9DzN+Z55ZsbcXQAAAAAQVdMFAAAAAKgdCAcAAAAA\nJBEOAAAAAAQIBwAAAAAkEQ4AAAAABAgHAAAAACQRDgAAAAAECAcAAAAAJBEOAAAAAARiarqA09Wi\nRQtPTEys6TLOGIcPH1ajRo1qugyg1mOsAJFhrJyeVatW7XP3+JquA4jUGRcOEhMTtXLlypou44yR\nlZWljIyMmi4DqPUYK0BkGCunx8y21XQNwOlgWhEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIB\nAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAIHQwoGZPW1me8xsXQXbzcymm1mOmX1gZr3DqgUAAADA\nqYV55OAZSZmVbB8uqVPwb6KkX4VYCwAAAIBTCC0cuPvrkvZX0mSUpDle7B1J55pZ67DqAQAAAFC5\nmBrcd1tJO0ot7wzWfVq2oZlNVPHRBSUkJCgrK6s66jsr5OXl8X4hNGs/ya3pEqpMUtNoxgoQAf6u\nAGe3mgwHVs46L6+hu8+SNEuS0tPTPSMjI8Syzi5ZWVni/UJYbpr6Sk2XUGWeyWzEWAEiwN8V4OxW\nk1cr2impfanldpJ21VAtAAAAQJ1Xk+FgkaTvBFctukhSrrufNKUIAAAAQPUIbVqRmT0nKUNSCzPb\nKekBSfUkyd1/LWmxpG9KypF0RNLNYdUCAAAA4NRCCwfuPvYU213S98LaPwAAtUXiWXZ+DoCzF3dI\nBgAAACCJcAAAAAAgQDgAAAAAIIlwAAAAACBQkzdBq7U4cQwAAAB1EUcOAAAAAEgiHAAAAAAIEA4A\nAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAA\nAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAA\nAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAA\nSCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABI\nIhwAAAAACBAOAAAAAEgKORyYWaaZZZtZjplNLWd7BzNbbmbvm9kHZvbNMOsBAAAAULHQwoGZRUt6\nStJwSSmSxppZSplmP5A0z917SRojaUZY9QAAAACoXJhHDvpJynH3j9z9uKTnJY0q08YlnRM8bipp\nV4j1AAAAAKhETIh9t5W0o9TyTkn9y7R5UNJSM/tXSY0kDQ2xHgAAAACVCDMcWDnrvMzyWEnPuPs0\nM7tY0u/MLNXdi77SkdlESRMlKSEhQVlZWWHUe8KU7gWh9l+d8vLyQn+/UHcxVoDIMFYAnCnCDAc7\nJbUvtdxOJ08bGi8pU5Lc/W9m1kBSC0l7Sjdy91mSZklSenq6Z2RkhFRysZumvhJq/9XpmcxGCvv9\nQt3FWAEiw1gBcKYI85yDFZI6mVmSmdVX8QnHi8q02S7pMkkys66SGkjaG2JNAAAAACoQWjhw9wJJ\nkyQtkbRRxVclWm9mD5nZyKDZFEnfNbM1kp6TdJO7l516BAAAAKAahDmtSO6+WNLiMuvuL/V4g6SB\nYdYAAAAAIDLcIRkAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACA\nJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAk\nwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTC\nAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIB\nAAAAgADhAAAAAIAkwgEAAACAAOEAAAAAgCTCAQAAAIAA4QAAAACAJMIBAAAAgADhAAAAAIAkwgEA\nAACAAOEAAAAAgCTCAQAAAIAA4QAAAACApJDDgZllmlm2meWY2dQK2lxrZhvMbL2Z/T7MegAAAABU\nLCasjs0sWtJTki6XtFPSCjNb5O4bSrXpJOnfJA109wNm1jKsegAAAABULswjB/0k5bj7R+5+XNLz\nkkaVafNdSU+5+wFJcvc9IdYDAAAAoBKhHTmQ1FbSjlLLOyX1L9MmWZLM7C1J0ZIedPc/l+3IzCZK\nmihJCQkJysrKCqPeE6Z0Lwi1/+qUl5cX+vuFuouxAkSGsQLgTBFmOLBy1nk5++8kKUNSO0lvmFmq\nux/8ypPcZ0maJUnp6emekZFR5cWWdtPUV0Ltvzo9k9lIYb9fqLsYK0BkGCsAzhRhTivaKal9qeV2\nknaV0+YP7p7v7h9LylZxWAAAAABQzcIMByskdTKzJDOrL2mMpEVl2iyUNESSzKyFiqcZfRRiTQAA\nAAAqEFo4cPcCSZMkLZG0UdI8d19vZg+Z2cig2RJJn5vZBknLJX3f3T8PqyYAAAAAFQvznAO5+2JJ\ni8usu7/UY5c0OfiHMHy6Wnqw7EWizlAP5tZ0BQAAAGc17pAMAAAAQBLhAAAAAECAcAAAAABA0mmE\nAzMbZGY3B4/jzSwpvLIAAAAAVLeIwoGZPSDpPkn/FqyqJ+nZsIoCAAAAUP0iPXLwbUkjJR2WJHff\nJalJWEUBAAAAqH6RhoPjwWVHXZLMrFF4JQEAAACoCZGGg3lmNlPSuWb2XUl/kTQ7vLIAAAAAVLeI\nboLm7k+Y2eWSDknqLOl+d18WamUAAKD24eaawFntlOHAzKIlLXH3oZIIBAAAAMBZ6pTTity9UNIR\nM2taDfUAAAAAqCERTSuSdFTSWjNbpuCKRZLk7neEUhUAAACAahdpOHgl+AcAAADgLBXpCcm/NbP6\nkpKDVdnunh9eWQAAAACqW0ThwMwyJP1W0lZJJqm9mf2zu78eXmkAAAAAqlOk04qmSbrC3bMlycyS\nJT0nqU9YhQEAAACoXpGGg3olwUCS3H2zmdULqSYAqH5cux0AgIjDwUoz+29JvwuWb5C0KpySAAAA\nANSESMPBbZK+J+kOFZ9z8LqkGWEVBQAAAKD6RRoOYiT9wt1/Jp24a3JsaFUBAACcpVatWtUyJibm\nN5JSFcENaYEqVCRpXUFBwYQ+ffrsKa9BpOHgNUlDJeUFyw0lLZU04GuXCAAAUIfExMT8plWrVl3j\n4+MPREVFeU3Xg7qjqKjI9u7dm7J79+7fSBpZXptI02oDdy8JBgoex1VBjQAAAHVNanx8/CGCAapb\nVFSUx8fH56r4qFX5bSLs67CZ9S5ZMLN0SV9+zfoAAADqoiiCAWpK8NmrMANEOq3oTknzzWyXJJfU\nRtJ1X788AAAAALVFpOEgSVIvSR0kfVvSRSoOCQAAAPgaEqe+UqU3ld3602/VmsvN9+vXr/MTTzyx\n4xvf+MaRSNoXFRVpwIAByYsXL85p1qxZUdj1fV1/+tOfmkybNi1h+fLlOWH0P3fu3Kbr169v+JOf\n/GR32W1xcXG9jhw58n5Fzx0wYEDyH/7whw/j4+MLT2efkU4r+qG7H5J0rqTLJc2S9KvT2REAAADO\nHPn5+dW+z3nz5jXt1q3bl2dCMKgON9xwQ255wSASY8eO/fyJJ56IP93nRRoOShLHtyT92t3/IKn+\n6e4MAAAANe/73/9+66SkpG4DBgzoNGLEiKT7778/QSr+pX/SpElt+/bt2/nhhx9O2LVrV8ywYcMu\nSE1N7Zqamtp16dKljSTp0KFDUddcc01iampq165du6Y8++yz50pSXl6eXXnllecnJyenfOtb3zr/\n6NGjJklPPvlki/Hjx7cv2f+0adNaTJgwoV3ZuubOndvs29/+9kFJys7Orp+UlNTtuuuu69ipU6du\nI0eOTFq4cGGT3r17d+nYsWPq8uXL4yqrJTs7u36fPn06p6SkdE1JSem6bNmyRlLxr/39+vXrnJmZ\neX5SUlK3kSNHJhUVVZ5FRo8enXj99dd36NOnT+fExMTU5557rmnZNsuXL4/r1atXl65du6b06tWr\ny5o1a2IlaeXKlQ26d+/etUuXLinJyckpa9eujY30tU2fPr35d77znQ6StGnTpvo9e/bskpqa2vXO\nO+9sU7Lfbdu21UtPT+/cpUuXlE6dOnX785//3FiSxowZc/Cll15qHuFH4oRIpxV9YmYzVXw508fM\nLFZclxcAAOCM8/rrr8f98Y9/PG/t2rUb8vPzrWfPnim9evU6Me3n4MGD0StWrMiWpBEjRiRNnjz5\ns2HDhuVt2bKl/rBhwzp99NFH6//93/+99ZAhQw7Nnz9/6759+6LT09O7jhw58tDPfvaz+IYNGxZt\n3rx5w9///veGAwcOTJGk8ePH7+/WrVvKsWPHdsbGxvqzzz7bYubMmdvK1rZq1arGAwcOPLF+x44d\nDV544YWP+vTps61Hjx5d586d23zlypWbfv/735/7yCOPtB4yZMiHFdXSpk2bgjfeeGNzXFycr127\nNnbs2LHnr1u3bqMkbdy4seHq1as/SkxMzO/Tp0+XZcuWNR42bFhe2XpK27FjR+y7776bvWHDhtih\nQ4d2HjVq1NrS29PS0o6+++67m+rVq6eFCxc2uffee9stWbLkw1/+8pfxt99++2e33Xbb/qNHj1pB\nQYE++eSTepG8ttL933777R0mTJiwd9KkSZ8/+uijJ44IPP30080uu+yy3Mcee2x3QUGBvvjiiyhJ\nio+PLzx+/Ljt3r07ulWrVhFPLYo0HFwrKVPSE+5+0MxaS/p+pDsBAABA7ZCVldV4+PDhBxs3buyS\n/PLLLz9YevvYsWP3lzx+6623ztmyZUvDkuW8vLzoAwcORGVlZZ2zZMmSc6dPn95Kko4dO2Y5OTn1\n33zzzcZ33HHHHknq37//l8nJyUck6ZxzzikaOHDgFy+88ELT7t27H83Pz7d+/fqddOXL3NzcmPPO\nO+/Ez/ht27Y9VtIuOTn5y0svvfRQVFSUevfufeThhx9uE7yecmvp2LFj/vjx4ztu2LChYVRUlLZt\n23biBr7du3c/fMEFF+RLUrdu3Y58+OGHp5wRM3r06P3R0dHq3r37sfbt2x9bvXp1g9Lb9+/fH33d\nddclbd26tYGZeX5+vknSxRdffPiJJ55ovXPnzvpjxow50L1792ORvrbS3nvvvcavvvrqh5J06623\nfv7jH/+4nSRddNFFh2+99dbE/Pz8qKuvvvrAgAEDTryvzZs3L9i+fXv9Vq1aRXyV0YjCgbsfkfRS\nqeVPJX0a6U4AAABQO7hXfk2ZJk2anPhy7u5auXLlxiBIfKWPBQsW5KSlpR0r+3wzK7ffiRMn7nvk\nkUdaJScnHx03bty+8tpER0d7YWGhoqOjJUn169c/sd+oqCg1aNDAg3YqLCy0ymqZPHlym5YtW+a/\n+OKLHxcVFalhw4YnTvyOjY090W90dLQKCgrKL7qS11V2+b777ms7ePDgL5YtW/ZhdnZ2/UsvvbSz\nJP3Lv/zL/ksuueTwyy+/3HT48OHJM2bM2Nq5c+djkby2ssq7BO7w4cPzXn/99ewXX3yx6U033ZR0\nxx13fDZp0qTPpeKgFBcXd1rnbzA1CAAAoA7JyMjIW7JkSdMjR45Ybm5u1F/+8pdzK2o7aNCgQ489\n9ljLkuW33367oSQNGTLk0LRp0xJK5uq/9dZbDYP2ec8++2wzSVqxYkWDzZs3n7hp7qWXXnr4008/\nrf/yyy83Hz9+/H6VIykp6ejGjRtjy9tWkYpqyc3NjW7dunV+dHS0ZsyY0byw8NQza773ve+1nTNn\nTrnvx0svvXReYWGh1q9fH7tjx47YtLS0o6W3Hzp0KLpdu3bHJWnmzJktStZv2LChfteuXY/94Ac/\n2HPFFVccXL16dcOyfUeid+/eebNnz24mSbNnzz5xLsHmzZvrt23bNn/KlCn7xo0bt++9996Lk4qv\n/LR37956nTt3PinAVSbSaUUAAAAIQXVfenTw4MFHMjMzc1NSUrq1bdv2WI8ePQ43bdq03G/Os2bN\n2jFhwoQOycnJKYWFhda/f/8vBgwYsP2nP/3prokTJ3bo0qVLirtbu3btji1fvjznnnvu2TNmzJik\n5OTklG7duh3p3r374dL9XXXVVQc++OCDuIour3nFFVfkLl26tElqamrEX2grquWuu+7aM3r06AsW\nLlx43qBBg75o2LDhKX9B37BhQ8OSE6LLuvDCC4/169ev8+eff17v5z//+ba4uLiv/Ip/33337Z4w\nYULS9OnTW11yySWHStb/7ne/azZ//vzmMTExHh8fn//oo4/uOnjwYHSkr6/EjBkzto8ZM+b8GTNm\nJIwcOfJAyfolS5Y0mT59equYmBiPi4srnDt37seS9Oabb8b16tXrcL169U5rP3aqQ0u1TXp6uq9c\nuTLUfSROfSXU/qvTM2lblJH9QE2XUTUezK3pClAGY6WWYqzUOoyVWqoaxoqZrXL39NLr1qxZszUt\nLa3caTXVJTc3N6pp06ZFX3zxRdTFF1/c+de//vW2QYMGRXQvgq9jyJAhF951112fjRo16ovytm/b\ntq3e2LFjE99+++0tYddSnkGDBnV68803T9r36NGjE6+88srcm2+++UB5z6uNbr755vZXXXXVwfLe\n6zVr1rRIS0tLLO95TCsCAACoY8aNG9exS5cuKT169Og6YsSIA2EHg3379kUnJiamNmjQoKiiYCBJ\nHTt2zL/lllv27d+/v0a+o5YXDM5UqampX1b2XleEaUUAAAB1zB//+MePq3N/LVq0KNy6deu6SNpO\nmDCh1v06/+KLL26t6RpO15QpU/6ho1McOQAAAAAgiXAAAAAAIEA4AAAAACCJcAAAAAAgwAnJAAAA\nNenBpn1O3eh0+ss95X0TzKzPqFGj9i9cuPBjScrPz1fLli3TevbseXj58uU506dPb75y5cpGc+bM\n2T558uQ2jRs3LnzooYc+K91HdHR0n06dOn1ZWFhoF1544Zfz5s3bWvruylLxjbgGDBiQvHjx4pxm\nzZoVXXPNNYmvvfZa0+bNmxds2bJlfXm1vf/++w1uvvnmxA0bNsRNnTr1k9L7XbBgwTn33HNPh6Ki\nIo0bN27fT37yk91ln//qq682njJlSvvNmzfHzZ49+6OSy4/u2rUr5rrrrkt64403zporEoWBIwcA\nAAB1TMOGDYuys7Mb5uXlmSS9/PLL5yQkJOSfTh+xsbFFmzZt2rBly5b19erV82nTpsWXbTNv3rym\n3bp1+7JZs2ZFknTLLbfsW7RoUaVfzlu2bFnwi1/8Yvutt976lTBSUFCgu+++u8PixYs3b968ef2L\nL77YbNWqVQ3KPv/8888//j//8z9bR4wY8Xnp9W3atClISEjIX7p0aaPTeZ11DeEAAACgDrrsssty\n58+ff64kPffcc81Gjx69/x/ta9CgQXk5OTmxZdfPnTu3Wek7Dg8fPjwvPj6+oLK+2rZtWzB48OAj\n9erV+8qderOyshp17NjxWEpKyvEGDRr4P/3TP+1fsGDBuWWf37lz5+P9+/f/Mirq5K+5V1111cE5\nc+Y0P60XV8cQDgAAAOqgG2+8cf8LL7xw3pEjR2zjxo1xF1988eF/pJ/8/HwtWbLknO7du39Zdtuq\nVasaDxw48JT9Pv744/GPP/74SUceStuxY0f9tm3bHi9Zbteu3fFPPvmkviTdddddbebOndv0VPsZ\nOHDg4XfffbfxqdrVZaGec2BmmZJ+ISla0m/c/acVtLta0nxJfd19ZZg1AQAAQOrfv/+XO3fujJ09\ne3azoUOH5p7u848dOxbVpUuXlKCvL+68886TbrqVm5sbc9555xWd/Oyvuvfee/eeqo27n7TOzFyS\nfv7zn++KpOY2bdoU7Nmzp34kbeuq0MKBmUVLekrS5ZJ2SlphZovcfUOZdk0k3SHp72HVAgAAgJNl\nZmYefOCBB9ovXbo0e8+ePaf1vbDknIPK2kRHR3thYaGio6O/XqGSOnTocOJIgSTt3Lmzfps2bU7r\nPIkjR45YbGzsKcNKXRbmtKJ+knLc/SN3Py7peUmjymn3Y0mPSzoaYi0AAAAo47bbbts3ZcqUXf36\n9TtpSlBVSEpKOrpx48aTzkX4RwwePPjw1q1bG2zatKn+0aNH7aWXXmo2evTog6d+5v9bt25dg+Tk\n5FBe69kizGlFbSXtKLW8U1L/0g3MrJek9u7+JzO7p6KOzGyipImSlJCQoKysrKqvtpQp3Ss9T+aM\nkhfbRlmdf1TTZVSNkP93x+ljrNRSjJVah7FSS9WWsRLBpUfDcsEFF+T/8Ic/3HOqdk8++WTrmTNn\nJpQsf/bZZx9E0v8VV1yRu3Tp0iapqanHJGnEiBFJ77zzTpMDBw7EJCQk9Jg6dequu+++e1/J+Qb3\n3nvv3u3bt8f07ds35fDhw9Fm5jNnzkzYuHHjumbNmhVNmzZte2ZmZnJhYaGuv/76fenp6Uel4nMO\n+vbte/iGG27I/etf/xp37bXXXnjo0KHo11577dxHHnmkTU5OznpJWrZsWZPMzMzTnkJVl1h587eq\npGOzayQNc/cJwfKNkvq5+78Gy1GS/lfSTe6+1cyyJN1zqnMO0tPTfeXKcE9LSJz6Sqj9V6dn0rYo\nI/uBmi6jajzIWK5tGCu1FGOl1mGs1FLVMFbMbJW7p5det2bNmq1paWknzc8/G23btq3e2LFjE99+\n++1acW+B9PT0zq+++mpOfHx8YU3XUpPWrFnTIi0tLbG8bWFOK9opqX2p5XaSSp8s0kRSqqQsM9sq\n6SJJi8zsKwMIAAAAZ6aOHTvm33LLLfv2799f41fI3LVrV8ydd975WV0PBqcS5rSiFZI6mVmSpE8k\njZF0fclGd8+V1KJkOdIjBwAAADhzTJgw4UBN1yAVX6noxhtvPK1zFOqi0FKcuxdImiRpiaSNkua5\n+3oze8jMRoa1XwAAgFquqKioyGq6CNRNwWevwis2hXqfA3dfLGlxmXX3V9A2I8xaAAAAaol1e/fu\nTYmPj8+NiooK5+RPoBxFRUW2d+/eppLWVdQm1HAAAACAryooKJiwe/fu3+zevTtV4Z7/CZRVJGld\nQUHBhIoaEA4AAACqUZ8+ffZIYoo1aiXSKgAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAA\nSYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJ\nhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmE\nAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQD\nAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASYQDAAAAAAHCAQAAAABJhAMA\nAAAAAcIBAAAAAEmEAwAAAAABwgEAAAAASSGHAzPLNLNsM8sxs6nlbJ9sZhvM7AMze83MOoZZDwAA\nAICKhRYOzCxa0lOShktKkTTWzFLKNHtfUrq795C0QNLjYdUDAAAAoHJhHjnoJynH3T9y9+OSnpc0\nqnQDd1/u7keCxXcktQuxHgAAAACVMHcPp2OzqyVluvuEYPlGSf3dfVIF7f9L0m53f7icbRMlTZSk\nhISEPs8//3woNZdY+0luqP1Xp6SGx9T42K6aLqNqtO5Z0xWgDMZKLcVYqXUYK7VUNYyVIUOGrHL3\n9NB3BFSRmBD7tnLWlZtEzGycpHRJg8vb7u6zJM2SpPT0dM/IyKiiEst309RXQu2/Oj2T9rEysh+o\n6TKqxtiz54/r2YKxUksxVmodxkotxVgBThJmONgpqX2p5XaSTvqpwcyGSvoPSYPd/ViI9QAAAACo\nRJjnHKyQ1MnMksysvqQxkhaVbmBmvSTNlDTS3feEWAsAAACAUwgtHLh7gaRJkpZI2ihpnruvN7OH\nzGxk0Ow/JTWWNN/MVpvZogq6AwAAABCyMKcVyd0XS1pcZt39pR4PDXP/AAAAACLHHZIBAAAASCIc\nAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwA\nAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAA\nAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAA\nAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAA\nCBAOAAAAAEgiHAAAAAAIEA4AAAAASCIcAAAAAAgQDgAAAABIIhwAAAAACBAOAAAAAEgiHAAAAAAI\nhBoOzCzTzLLNLMfMppazPdbMXgi2/93MEsOsBwAAAEDFQgsHZhYt6SlJwyWlSBprZillmo2XdMDd\nL5T0pKTHwqoHAAAAQOXCPHLQT1KOu3/k7sclPS9pVJk2oyT9Nni8QNJlZmYh1gQAAACgAmGGg7aS\ndpRa3hmsK7eNuxdIypXUPMSaAAAAAFQgJsS+yzsC4P9AG5nZREkTg8U8M8v+mrXVGUOkFpL21XQd\nVeJHHFRCeBgrQGQYK6etY3XsBKgqYYaDnZLal1puJ2lXBW12mlmMpKaS9pftyN1nSZoVUp1nNTNb\n6e7pNV0HUNsxVoDIMFaAs1umpsoGAAAFNUlEQVSY04pWSOpkZklmVl/SGEmLyrRZJOmfg8dXS/pf\ndz/pyAEAAACA8IV25MDdC8xskqQlkqIlPe3u683sIUkr3X2RpP+W9Dszy1HxEYMxYdUDAAAAoHLG\nD/VnNzObGEzLAlAJxgoQGcYKcHYjHAAAAACQFPIdkgEAAACcOQgHAAAAACQRDuo8M0s0s3XB4wwz\n+1NN1wTUBDN7+xTbF5vZudVVD3A2MLMsM0sPHm81sxY1XROAyoV5nwOEyMxMxeeMFNV0LUBtY2bR\n7l54Os9x9wGn2P7Nr1cVcPbhbxFw9uHIwRkk+JV/o5nNkPSepBvN7G9m9p6ZzTezxkG7vmb2tpmt\nMbN3zaxJ8Nw3grbvmVmlX4SA2ir4LG8ys9+a2QdmtsDM4oJfJe83szclXWNmF5jZn81sVfDZ7xI8\nP8HMXg7Gx5qSsWBmecF/W5vZ62a22szWmdklwfoTv3qa2eRg2zozu6tUXRvNbLaZrTezpWbWsEbe\nJKASZT+/ZvaYmd1eavuDZjYlePx9M1sRjLUfBevK/i1qb2a/MrOVwWf/RzXzygBUBcLBmaezpDmS\nLpc0XtJQd+8taaWkycEN516QdKe7p0kaKulLSXskXR60vU7S9JooHqginSXNcvcekg5JKvlic9Td\nB7n78yq+q/q/unsfSfdImhG0mS7pr8H46C1pfZm+r5e0xN17SkqTtLr0RjPrI+lmSf0lXSTpu2bW\nK9jcSdJT7t5N0kFJo6vqBQNVobzPr6TnVfx3ocS1kuab2RUq/kz3k9RTUh8z+0bQprOkOe7ey923\nSfqP4K7JPSQNNrMe1fKCAFQ5phWdeba5+ztmdqWkFElvFR/VVX1Jf1Px/2F/6u4rJMndD0mSmTWS\n9F9m1lNSoaTkmigeqCI73P2t4PGzku4IHr8gScFRtAEq/oJT8pzY4L+XSvqOJAVTj3LL9L1C0tNm\nVk/SQndfXWb7IEkvu/vhYF8vSbpExXd8/7hU+1WSEr/GawTCUNHnt6WZtZEUL+mAu283szskXSHp\n/eC5jVUcFrYr+FtUqt9rzWyiir9XtFbx36cPquMFAahahIMzz+HgvyZpmbuPLb0x+LWmvJtX3C3p\nMxX/Ehol6WiYRQIhK/sZL1kuGR9Rkg4Gv/6fXsfurwe/jn5LxXdw/093n1OqiVXwVEk6VupxoSSm\nFaG2qejzu0DS1ZJaqfhIQknbR9195lc6MEvU/481mVmSio/O9XX3A2b2jKQGVVo1gGrDtKIz1zuS\nBprZhZIUzLlOlrRJUhsz6xusb2JmMZKaqviIQpGkGyVF11DdQFXoYGYXB4/HSnqz9MbgiNnHZnaN\nVHzSpJmlBZtfk3RbsD7azM4p/Vwz6yhpj7vPlvTfKp56VNrrkq4KxlwjSd+W9EbVvTQgVBV9fp+X\nNEbFAWFB0HaJpFtKnc/W1sxaltPnOSoOC7lmliBpeMivAUCICAdnKHffK+kmSc+Z2QcqDgtd3P24\niueO/tLM1khapuJfcGZI+mcze0fFU4oOl9sxcGbYqOLP8weSmkn6VTltbpA0PhgH6yWNCtbfKWmI\nma1V8dSfbmWelyFptZm9r+JzBn5ReqO7vyfpGUnvSvq7pN+4+/sCzgAVfX7dfb2kJpI+cfdPg7ZL\nJf1e0t+C8bIgaFO2zzUqnnq0XtLTkt4q2wbAmcPcy5uBAgC1UzCl4U/unlrDpQAAcNbhyAEAAAAA\nSRw5AAAAABDgyAEAAAAASYQDAAAAAAHCAQAAAABJhAMAAAAAAcIBAAAAAEnS/wE6SdxyxCD/9wAA\nAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fcd4dfe9908>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_scores([(greedy_mean, 'greedy (mean, plasmids)'), (MILP, 'MILP (1:10:1)')], score_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MILP output details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_, _, skip_pred_plasmids, skip_ref_by_pred, skip_pred_by_ref, skip_pred_by_all, skip_ref_by_all, \\\n",
    "    skip_mutual, skip_summary_scores = parse_structure(MILP)\n",
    "num_pred_plasmids = skip_ref_by_pred - skip_pred_plasmids - 2\n",
    "num_mutual = skip_summary_scores - skip_mutual - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>length (nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>158604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>34805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>62315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>17370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>49576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>1582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>4321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>plasmid_7</td>\n",
       "      <td>6558</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  length (nt)\n",
       "0  plasmid_0       158604\n",
       "1  plasmid_1        34805\n",
       "2  plasmid_2        62315\n",
       "3  plasmid_3        17370\n",
       "4  plasmid_4        49576\n",
       "5  plasmid_5         1582\n",
       "6  plasmid_6         4321\n",
       "7  plasmid_7         6558"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_plasmids(MILP, skip_pred_plasmids, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recall</td>\n",
       "      <td>0.580830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>precision</td>\n",
       "      <td>0.517681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f1_score</td>\n",
       "      <td>0.547440</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       score     value\n",
       "0     recall  0.580830\n",
       "1  precision  0.517681\n",
       "2   f1_score  0.547440"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_scores(MILP, skip_summary_scores, len(score_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid (column) covered by individual predicted plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted</th>\n",
       "      <th>CP018949.1</th>\n",
       "      <th>CP018950.1</th>\n",
       "      <th>CP018951.1</th>\n",
       "      <th>CP018952.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.463282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.003382</td>\n",
       "      <td>0.976153</td>\n",
       "      <td>0.018257</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.167418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000853</td>\n",
       "      <td>0.098629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.003382</td>\n",
       "      <td>0.929113</td>\n",
       "      <td>0.018257</td>\n",
       "      <td>0.091345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>plasmid_7</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.019139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.006764</td>\n",
       "      <td>1.905266</td>\n",
       "      <td>0.037367</td>\n",
       "      <td>0.840757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   predicted  CP018949.1  CP018950.1  CP018951.1  CP018952.1\n",
       "0  plasmid_0    0.000000    0.000000    0.000000    0.463282\n",
       "1  plasmid_1    0.003382    0.976153    0.018257    0.000000\n",
       "2  plasmid_2    0.000000    0.000000    0.000000    0.167418\n",
       "3  plasmid_3    0.000000    0.000000    0.000853    0.098629\n",
       "4  plasmid_4    0.003382    0.929113    0.018257    0.091345\n",
       "5  plasmid_5    0.000000    0.000000    0.000000    0.000945\n",
       "6  plasmid_6    0.000000    0.000000    0.000000    0.000000\n",
       "7  plasmid_7    0.000000    0.000000    0.000000    0.019139\n",
       "8      Total    0.006764    1.905266    0.037367    0.840757"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_pred, num_pred_plasmids, renaming = {'Unnamed: 0': 'predicted'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid (column) covered by individual reference plasmid (row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP018949.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001609</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001130</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP018950.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.987933</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.660158</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP018951.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.022755</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002073</td>\n",
       "      <td>0.015975</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP018952.1</td>\n",
       "      <td>0.370413</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.318928</td>\n",
       "      <td>0.749453</td>\n",
       "      <td>0.243646</td>\n",
       "      <td>0.079014</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.386246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Total</td>\n",
       "      <td>0.370413</td>\n",
       "      <td>1.012297</td>\n",
       "      <td>0.318928</td>\n",
       "      <td>0.751526</td>\n",
       "      <td>0.920909</td>\n",
       "      <td>0.079014</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.386246</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference  plasmid_0  plasmid_1  plasmid_2  plasmid_3  plasmid_4  \\\n",
       "0  CP018949.1   0.000000   0.001609   0.000000   0.000000   0.001130   \n",
       "1  CP018950.1   0.000000   0.987933   0.000000   0.000000   0.660158   \n",
       "2  CP018951.1   0.000000   0.022755   0.000000   0.002073   0.015975   \n",
       "3  CP018952.1   0.370413   0.000000   0.318928   0.749453   0.243646   \n",
       "4       Total   0.370413   1.012297   0.318928   0.751526   0.920909   \n",
       "\n",
       "   plasmid_5  plasmid_6  plasmid_7  \n",
       "0   0.000000        0.0   0.000000  \n",
       "1   0.000000        0.0   0.000000  \n",
       "2   0.000000        0.0   0.000000  \n",
       "3   0.079014        0.0   0.386246  \n",
       "4   0.079014        0.0   0.386246  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_ref, num_ref_plasmids, renaming = {'Unnamed: 0': 'reference'}, col_sums = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of predicted plasmid collectively covered by all reference plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasmid_0</td>\n",
       "      <td>0.370413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.987933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plasmid_2</td>\n",
       "      <td>0.318928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plasmid_3</td>\n",
       "      <td>0.749453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmid_4</td>\n",
       "      <td>0.903804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmid_5</td>\n",
       "      <td>0.079014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmid_6</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>plasmid_7</td>\n",
       "      <td>0.386246</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     plasmid  proportion\n",
       "0  plasmid_0    0.370413\n",
       "1  plasmid_1    0.987933\n",
       "2  plasmid_2    0.318928\n",
       "3  plasmid_3    0.749453\n",
       "4  plasmid_4    0.903804\n",
       "5  plasmid_5    0.079014\n",
       "6  plasmid_6    0.000000\n",
       "7  plasmid_7    0.386246"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_pred_by_all, num_pred_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proportion of reference plasmid collectively covered by all predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP018949.1</td>\n",
       "      <td>0.003382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP018950.1</td>\n",
       "      <td>0.976238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP018951.1</td>\n",
       "      <td>0.019110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP018952.1</td>\n",
       "      <td>0.731958</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      plasmid  proportion\n",
       "0  CP018949.1    0.003382\n",
       "1  CP018950.1    0.976238\n",
       "2  CP018951.1    0.019110\n",
       "3  CP018952.1    0.731958"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_table(MILP, skip_ref_by_all, num_ref_plasmids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correspondences between reference and predicted plasmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>prediction</th>\n",
       "      <th>recall</th>\n",
       "      <th>precision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP018950.1</td>\n",
       "      <td>plasmid_1</td>\n",
       "      <td>0.976153</td>\n",
       "      <td>0.987933</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference prediction    recall  precision\n",
       "0  CP018950.1  plasmid_1  0.976153   0.987933"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_correspondences(MILP, skip_mutual, num_mutual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Objective function values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary of the values of components of the objective functions has been provided below. These values have been computed for both the greedy and MILP approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dir = \"../../../../../../Plasmids-Assembly/exp/2018-07-11__plasmid_contigs_from_genes/MOB-suite_samples/unicycler_pipeline/\"\n",
    "assembly_file = sample_dir + 'sample_' + str(sample_id) + '/assembly.gfa'\n",
    "mapping_file = sample_dir + 'sample_' + str(sample_id) + '/filtered_genes_to_contigs.csv'\n",
    "seeds_file = sample_dir + 'sample_' + str(sample_id) + '/seed_contigs.csv'\n",
    "\n",
    "MILP_contig_chains = \"nplasmids_1/MILP/contig_chains.csv\"\n",
    "greedy_contig_chains = \"../../../../../../Plasmids-Assembly/exp/2018-07-20__sample_analyses_greedy_strategy/greedy_v2/L111/sample_%i/greedy_mean/contig_chains.csv\" % (sample_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_file(filename):\n",
    "\tstring = open(filename, \"r\").read()\n",
    "\tstring_list = string.split(\"\\n\")\n",
    "\tstring_list = [line for line in string_list if line and line[0] != '#'] #Read line only if it is nonempty and not a comment.\n",
    "\treturn string_list\n",
    "\n",
    "#Storing contig details\n",
    "#-----------------------------------------------\n",
    "#Stores the id of the contig\n",
    "def get_id(line):\n",
    "\treturn line[1]\n",
    "#Stores the nucleotide sequence of the contig\n",
    "def get_nucleotide_seq(line):\n",
    "\t#print(line[2])\n",
    "\treturn line[2]\t\t\n",
    "#Computes GC ratio: counts no. of 'G'/'C' occurences in the sequence and divide by the sequence length.\n",
    "def compute_GCratio(seq):\n",
    "\tGC = 0\n",
    "\tln_seq = 0\n",
    "\tfor nucl in seq:\n",
    "\t\tif nucl == 'G' or nucl == 'C':\n",
    "\t\t\tGC += 1\n",
    "\t\tln_seq += 1\n",
    "\treturn GC/ln_seq\n",
    "#Stores the length of the sequence\n",
    "def get_length(line):\n",
    "\treturn int(line[3].split(':')[2])\n",
    "#Stores the read depth of the contig\n",
    "def get_read_depth(line):\n",
    "\treturn float(line[4].split(':')[2])\t\t\n",
    "\n",
    "#Takes a contig from the assembly file and initiates an entry in the contigs_dict\n",
    "#Each contig is tagged with the following attributes:\n",
    "#1. Length of the contig (int)\n",
    "#2. Overall read depth of the contig (float)\n",
    "#3. Indication if the contig is a seed (binary)\n",
    "#4. GC content of the contig (float)\n",
    "#5. Gene coverage intervals (list of pairs)\n",
    "#6. Gene coverage (float)\n",
    "def update_contigs_dict(contigs_dict, line):\n",
    "\tc = get_id(line)\n",
    "\tseq = get_nucleotide_seq(line) \n",
    "\tGC_cont = compute_GCratio(seq)\n",
    "\tln = get_length(line)\n",
    "\trd = get_read_depth(line)\n",
    "\n",
    "\tcontigs_dict[c] = {}\n",
    "\tcontigs_dict[c]['Sequence'] = seq\n",
    "\tcontigs_dict[c]['Length'] = ln\n",
    "\tcontigs_dict[c]['Read_depth'] = rd\n",
    "\tcontigs_dict[c]['Seed'] = 0\t\t\t\t\t\t\t#Default\n",
    "\tcontigs_dict[c]['GC_cont'] = GC_cont\n",
    "\tcontigs_dict[c]['Gene_coverage_intervals'] = []\t\t#Default\n",
    "\tcontigs_dict[c]['Gene_coverage'] = 0\t\t\t\t#Default\n",
    "\treturn contigs_dict\n",
    "\n",
    "#A link is of the type: ((l1, e1),(l2, e2)) \n",
    "#where l1, l2 are adjacent links and e1, e2 are the connected link extremities\n",
    "def get_link(line):\n",
    "\tc1, o1, c2, o2 = line[1], line[2], line[3], line[4]\n",
    "\tif o1 == '+':\n",
    "\t\text1 = 'h'\n",
    "\telse:\n",
    "\t\text1 = 't'\n",
    "\tif o2 == '+':\n",
    "\t\text2 = 't'\n",
    "\telse:\n",
    "\t\text2 = 'h'\n",
    "\te = ((c1, ext1),(c2, ext2))\t\n",
    "\treturn e \n",
    "\n",
    "#Reads the assembly file line by line and forwards a line \n",
    "#to update_contigs_dict or get_link depending on the entry\n",
    "def get_data(assembly_file, contigs_dict, links_list):\n",
    "\tstring_list = read_file(assembly_file)\n",
    "\tcount_s = 0\n",
    "\tcount_l = 0\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tif line[0] == 'S':\n",
    "\t\t\tcontigs_dict = update_contigs_dict(contigs_dict, line)\n",
    "\t\telif line[0] == 'L':\n",
    "\t\t\te = get_link(line)\n",
    "\t\t\tlinks_list.append(e)\n",
    "\treturn contigs_dict, links_list\n",
    "\n",
    "#Reads the seed file and makes a list of seeds\n",
    "def get_seeds(seeds_file, seeds_set):\n",
    "\tstring_list = read_file(seeds_file)\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\n",
    "\t\tseeds_set.add(line[0])\n",
    "\treturn seeds_set\n",
    "\n",
    "#Takes the gene covering intervals for a contig and finds their union\n",
    "#The length of the union is used to compute gene coverage\n",
    "def get_union(intervals):\n",
    "\tunion = []\n",
    "\tfor begin, end in sorted(intervals):\n",
    "\t\tif union and union[-1][1] >= begin - 1:\n",
    "\t\t\tunion[-1][1] = max(union[-1][1], end)\n",
    "\t\telse:\n",
    "\t\t\tunion.append([begin, end])\n",
    "\treturn union\t\t\n",
    "\n",
    "#Computes the gene coverage for each contig\n",
    "def get_gene_coverage(mapping_file, contigs_dict):\n",
    "\tstring_list = read_file(mapping_file)\n",
    "\tpossible_seeds = []\n",
    "\tfor line in string_list:\n",
    "\t\tline = line.split(\"\\t\")\t\n",
    "\t\tqseqid, sseqid = line[0], line[1]\n",
    "\t\tsstart, send = line[8], line[9]\n",
    "\t\tif sseqid not in possible_seeds:\n",
    "\t\t\tpossible_seeds.append(sseqid)\n",
    "\t\tif sseqid not in contigs_dict:\n",
    "\t\t\tprint(sseqid, \"not in contigs_dict\")\n",
    "\t\telse:\n",
    "\t\t\tif int(sstart) > int(send):\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(send), int(sstart)))\n",
    "\t\t\telse:\n",
    "\t\t\t\tcontigs_dict[sseqid]['Gene_coverage_intervals'].append((int(sstart), int(send)))\n",
    "\n",
    "\tfor sseqid in contigs_dict:\n",
    "\t\tunion = get_union(contigs_dict[sseqid]['Gene_coverage_intervals'])\n",
    "\t\tln = contigs_dict[sseqid]['Length']\n",
    "\t\tcovered = 0\n",
    "\t\tfor interval in union:\n",
    "\t\t\tcovered += interval[1] - interval[0] + 1\n",
    "\t\tcontigs_dict[sseqid]['Gene_coverage'] = covered/ln\n",
    "\n",
    "\treturn contigs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigs_dict = {}\n",
    "links_list = []\n",
    "seeds_set = set()\n",
    "\n",
    "contigs_dict, links_list = get_data(assembly_file, contigs_dict, links_list)\n",
    "seeds_set = get_seeds(seeds_file, seeds_set)\n",
    "contigs_dict = get_gene_coverage(mapping_file, contigs_dict)\n",
    "\n",
    "GC_total = 0\n",
    "ln_total = 0\n",
    "for c in contigs_dict:\n",
    "\tGC_total += contigs_dict[c]['GC_cont']*contigs_dict[c]['Length']\n",
    "\tln_total += contigs_dict[c]['Length']\n",
    "\t\n",
    "GC_mean = GC_total/ln_total\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_obj_val(chain):\n",
    "    gd_sum, GC_sum = 0, 0\n",
    "    a2, a3 = 10, 1\n",
    "    a_total = a2+a3\n",
    "    p_len = 0\n",
    "    for c in chain:\n",
    "        gd = contigs_dict[c]['Gene_coverage']\n",
    "        ln = contigs_dict[c]['Length']\n",
    "        GC = contigs_dict[c]['GC_cont']\n",
    "        p_len += ln\n",
    "        gd_sum += -(a2/a_total)*gd*ln\n",
    "        GC_sum += -(a3/a_total)*(GC_mean-GC)*ln\n",
    "    gd_sum = gd_sum/p_len\n",
    "    GC_sum = GC_sum/p_len\n",
    "    total = gd_sum + GC_sum\n",
    "    return gd_sum, GC_sum, total\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_plasmids = {}\n",
    "greedy_plasmids = {}\n",
    "\n",
    "plasmids = open(MILP_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        MILP_plasmids[p_id] = {}\n",
    "        MILP_plasmids[p_id]['chain'] = chain\n",
    "        MILP_plasmids[p_id]['gd_sum'], MILP_plasmids[p_id]['GC_sum'], MILP_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)\n",
    "\n",
    "        \n",
    "plasmids = open(greedy_contig_chains, \"r\").read()\n",
    "plasmid_list = plasmids.split(\"\\n\")\n",
    "for line in plasmid_list:\n",
    "    if line:\n",
    "        p_id = line.split(\";\")[0]\n",
    "        #print(p_id)\n",
    "        chain = line.split(\";\")[1].split(\",\")\n",
    "        chain = [x[0:-1] for x in chain]\n",
    "        greedy_plasmids[p_id] = {}\n",
    "        greedy_plasmids[p_id]['chain'] = chain\n",
    "        greedy_plasmids[p_id]['gd_sum'], greedy_plasmids[p_id]['GC_sum'], greedy_plasmids[p_id]['total'] = get_obj_val(chain)\n",
    "        #print(chain)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MILP_obj = pd.DataFrame(MILP_plasmids)\n",
    "MILP_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.000148569</td>\n",
       "      <td>-0.00846045</td>\n",
       "      <td>-0.00200419</td>\n",
       "      <td>-0.00110994</td>\n",
       "      <td>-0.0064141</td>\n",
       "      <td>0.00152381</td>\n",
       "      <td>0.00202632</td>\n",
       "      <td>0.00367495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[113, 83, 92, 89, 30, 56, 46, 53, 76, 36, 125,...</td>\n",
       "      <td>[62, 79, 24]</td>\n",
       "      <td>[28, 88, 124, 112, 82, 73, 42, 55, 52, 77, 51,...</td>\n",
       "      <td>[39, 63, 54, 76, 74, 65, 104]</td>\n",
       "      <td>[63, 54, 76, 74, 65, 104, 43, 60, 79, 24]</td>\n",
       "      <td>[104, 65]</td>\n",
       "      <td>[55, 52, 77, 51, 56, 63]</td>\n",
       "      <td>[51, 56, 92, 83]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.711698</td>\n",
       "      <td>-0.798266</td>\n",
       "      <td>-0.291131</td>\n",
       "      <td>-0.802533</td>\n",
       "      <td>-0.794409</td>\n",
       "      <td>-0.909091</td>\n",
       "      <td>-0.383335</td>\n",
       "      <td>-0.358949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.711847</td>\n",
       "      <td>-0.806726</td>\n",
       "      <td>-0.293135</td>\n",
       "      <td>-0.803643</td>\n",
       "      <td>-0.800823</td>\n",
       "      <td>-0.907567</td>\n",
       "      <td>-0.381309</td>\n",
       "      <td>-0.355274</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0     plasmid_1  \\\n",
       "GC_sum                                       -0.000148569   -0.00846045   \n",
       "chain   [113, 83, 92, 89, 30, 56, 46, 53, 76, 36, 125,...  [62, 79, 24]   \n",
       "gd_sum                                          -0.711698     -0.798266   \n",
       "total                                           -0.711847     -0.806726   \n",
       "\n",
       "                                                plasmid_2  \\\n",
       "GC_sum                                        -0.00200419   \n",
       "chain   [28, 88, 124, 112, 82, 73, 42, 55, 52, 77, 51,...   \n",
       "gd_sum                                          -0.291131   \n",
       "total                                           -0.293135   \n",
       "\n",
       "                            plasmid_3  \\\n",
       "GC_sum                    -0.00110994   \n",
       "chain   [39, 63, 54, 76, 74, 65, 104]   \n",
       "gd_sum                      -0.802533   \n",
       "total                       -0.803643   \n",
       "\n",
       "                                        plasmid_4   plasmid_5  \\\n",
       "GC_sum                                 -0.0064141  0.00152381   \n",
       "chain   [63, 54, 76, 74, 65, 104, 43, 60, 79, 24]   [104, 65]   \n",
       "gd_sum                                  -0.794409   -0.909091   \n",
       "total                                   -0.800823   -0.907567   \n",
       "\n",
       "                       plasmid_6         plasmid_7  \n",
       "GC_sum                0.00202632        0.00367495  \n",
       "chain   [55, 52, 77, 51, 56, 63]  [51, 56, 92, 83]  \n",
       "gd_sum                 -0.383335         -0.358949  \n",
       "total                  -0.381309         -0.355274  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MILP_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "greedy_obj = pd.DataFrame(greedy_plasmids)\n",
    "greedy_obj.rename(columns = {0: 'Plasmid', 1: 'Chain', 2: 'gd', 3: 'GC', 4: 'Total'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plasmid_0</th>\n",
       "      <th>plasmid_1</th>\n",
       "      <th>plasmid_10</th>\n",
       "      <th>plasmid_11</th>\n",
       "      <th>plasmid_12</th>\n",
       "      <th>plasmid_13</th>\n",
       "      <th>plasmid_14</th>\n",
       "      <th>plasmid_15</th>\n",
       "      <th>plasmid_16</th>\n",
       "      <th>plasmid_17</th>\n",
       "      <th>plasmid_18</th>\n",
       "      <th>plasmid_19</th>\n",
       "      <th>plasmid_2</th>\n",
       "      <th>plasmid_20</th>\n",
       "      <th>plasmid_3</th>\n",
       "      <th>plasmid_4</th>\n",
       "      <th>plasmid_5</th>\n",
       "      <th>plasmid_6</th>\n",
       "      <th>plasmid_7</th>\n",
       "      <th>plasmid_8</th>\n",
       "      <th>plasmid_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GC_sum</th>\n",
       "      <td>-0.000751799</td>\n",
       "      <td>-0.00831583</td>\n",
       "      <td>0.00152381</td>\n",
       "      <td>2.11012e-05</td>\n",
       "      <td>0.00206886</td>\n",
       "      <td>-0.000651758</td>\n",
       "      <td>0.00135246</td>\n",
       "      <td>-0.000162189</td>\n",
       "      <td>0.000501196</td>\n",
       "      <td>0.00426066</td>\n",
       "      <td>0.00366916</td>\n",
       "      <td>2.01793e-06</td>\n",
       "      <td>0.00515682</td>\n",
       "      <td>-0.00998782</td>\n",
       "      <td>-0.000937255</td>\n",
       "      <td>-0.00022167</td>\n",
       "      <td>0.00502197</td>\n",
       "      <td>0.00541472</td>\n",
       "      <td>0.00225304</td>\n",
       "      <td>0.000195041</td>\n",
       "      <td>-0.000337058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chain</th>\n",
       "      <td>[99, 87, 81, 101, 49, 60, 43, 104, 111, 65, 74...</td>\n",
       "      <td>[79, 24, 79, 62]</td>\n",
       "      <td>[65, 104]</td>\n",
       "      <td>[27, 56, 2, 72, 41]</td>\n",
       "      <td>[56]</td>\n",
       "      <td>[108, 29, 83, 92, 89, 35, 98]</td>\n",
       "      <td>[83, 92, 89]</td>\n",
       "      <td>[52, 77, 61, 55]</td>\n",
       "      <td>[58, 57, 55, 10]</td>\n",
       "      <td>[55, 34, 103]</td>\n",
       "      <td>[79]</td>\n",
       "      <td>[130, 13, 73, 82, 115]</td>\n",
       "      <td>[31]</td>\n",
       "      <td>[42, 73]</td>\n",
       "      <td>[67, 18, 83, 92, 89, 32, 102, 81, 90, 40, 101,...</td>\n",
       "      <td>[72, 19, 56, 15, 77, 51, 56, 46, 53, 76, 36, 1...</td>\n",
       "      <td>[81, 90]</td>\n",
       "      <td>[90]</td>\n",
       "      <td>[71]</td>\n",
       "      <td>[88, 125, 115, 82, 73, 45, 60, 4, 97]</td>\n",
       "      <td>[108, 14, 55, 33, 65, 104, 1, 68, 11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gd_sum</th>\n",
       "      <td>-0.803333</td>\n",
       "      <td>-0.79809</td>\n",
       "      <td>-0.909091</td>\n",
       "      <td>-0.0127234</td>\n",
       "      <td>-0.882522</td>\n",
       "      <td>-0.0149663</td>\n",
       "      <td>-0.65696</td>\n",
       "      <td>-0.239819</td>\n",
       "      <td>-0.00963433</td>\n",
       "      <td>-0.128135</td>\n",
       "      <td>-0.78355</td>\n",
       "      <td>-0.00192001</td>\n",
       "      <td>-0.800389</td>\n",
       "      <td>-0.0260256</td>\n",
       "      <td>-0.0810319</td>\n",
       "      <td>-0.0311179</td>\n",
       "      <td>-0.909091</td>\n",
       "      <td>-0.909091</td>\n",
       "      <td>-0.909091</td>\n",
       "      <td>-0.00543331</td>\n",
       "      <td>-0.031765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <td>-0.804085</td>\n",
       "      <td>-0.806406</td>\n",
       "      <td>-0.907567</td>\n",
       "      <td>-0.0127023</td>\n",
       "      <td>-0.880453</td>\n",
       "      <td>-0.0156181</td>\n",
       "      <td>-0.655608</td>\n",
       "      <td>-0.239981</td>\n",
       "      <td>-0.00913313</td>\n",
       "      <td>-0.123874</td>\n",
       "      <td>-0.779881</td>\n",
       "      <td>-0.00191799</td>\n",
       "      <td>-0.795232</td>\n",
       "      <td>-0.0360134</td>\n",
       "      <td>-0.0819692</td>\n",
       "      <td>-0.0313395</td>\n",
       "      <td>-0.904069</td>\n",
       "      <td>-0.903676</td>\n",
       "      <td>-0.906838</td>\n",
       "      <td>-0.00523827</td>\n",
       "      <td>-0.032102</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                plasmid_0         plasmid_1  \\\n",
       "GC_sum                                       -0.000751799       -0.00831583   \n",
       "chain   [99, 87, 81, 101, 49, 60, 43, 104, 111, 65, 74...  [79, 24, 79, 62]   \n",
       "gd_sum                                          -0.803333          -0.79809   \n",
       "total                                           -0.804085         -0.806406   \n",
       "\n",
       "        plasmid_10           plasmid_11  plasmid_12  \\\n",
       "GC_sum  0.00152381          2.11012e-05  0.00206886   \n",
       "chain    [65, 104]  [27, 56, 2, 72, 41]        [56]   \n",
       "gd_sum   -0.909091           -0.0127234   -0.882522   \n",
       "total    -0.907567           -0.0127023   -0.880453   \n",
       "\n",
       "                           plasmid_13    plasmid_14        plasmid_15  \\\n",
       "GC_sum                   -0.000651758    0.00135246      -0.000162189   \n",
       "chain   [108, 29, 83, 92, 89, 35, 98]  [83, 92, 89]  [52, 77, 61, 55]   \n",
       "gd_sum                     -0.0149663      -0.65696         -0.239819   \n",
       "total                      -0.0156181     -0.655608         -0.239981   \n",
       "\n",
       "              plasmid_16     plasmid_17  plasmid_18              plasmid_19  \\\n",
       "GC_sum       0.000501196     0.00426066  0.00366916             2.01793e-06   \n",
       "chain   [58, 57, 55, 10]  [55, 34, 103]        [79]  [130, 13, 73, 82, 115]   \n",
       "gd_sum       -0.00963433      -0.128135    -0.78355             -0.00192001   \n",
       "total        -0.00913313      -0.123874   -0.779881             -0.00191799   \n",
       "\n",
       "         plasmid_2  plasmid_20  \\\n",
       "GC_sum  0.00515682 -0.00998782   \n",
       "chain         [31]    [42, 73]   \n",
       "gd_sum   -0.800389  -0.0260256   \n",
       "total    -0.795232  -0.0360134   \n",
       "\n",
       "                                                plasmid_3  \\\n",
       "GC_sum                                       -0.000937255   \n",
       "chain   [67, 18, 83, 92, 89, 32, 102, 81, 90, 40, 101,...   \n",
       "gd_sum                                         -0.0810319   \n",
       "total                                          -0.0819692   \n",
       "\n",
       "                                                plasmid_4   plasmid_5  \\\n",
       "GC_sum                                        -0.00022167  0.00502197   \n",
       "chain   [72, 19, 56, 15, 77, 51, 56, 46, 53, 76, 36, 1...    [81, 90]   \n",
       "gd_sum                                         -0.0311179   -0.909091   \n",
       "total                                          -0.0313395   -0.904069   \n",
       "\n",
       "         plasmid_6   plasmid_7                              plasmid_8  \\\n",
       "GC_sum  0.00541472  0.00225304                            0.000195041   \n",
       "chain         [90]        [71]  [88, 125, 115, 82, 73, 45, 60, 4, 97]   \n",
       "gd_sum   -0.909091   -0.909091                            -0.00543331   \n",
       "total    -0.903676   -0.906838                            -0.00523827   \n",
       "\n",
       "                                    plasmid_9  \n",
       "GC_sum                           -0.000337058  \n",
       "chain   [108, 14, 55, 33, 65, 104, 1, 68, 11]  \n",
       "gd_sum                              -0.031765  \n",
       "total                               -0.032102  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greedy_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding intersections or subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasmid_3  from MILP completely contained in  plasmid_0  from greedy.\n",
      "\n",
      "plasmid_5  from MILP completely contained in  plasmid_0  from greedy.\n",
      "\n",
      "plasmid_1  from MILP completely contained in  plasmid_1  from greedy.\n",
      "\n",
      "plasmid_1  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_5  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_5  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n",
      "plasmid_6  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_6  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n",
      "plasmid_7  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_5  from MILP completely contained in  plasmid_9  from greedy.\n",
      "\n",
      "plasmid_10  from greedy completely contained in  plasmid_3  from MILP.\n",
      "\n",
      "plasmid_10  from greedy completely contained in  plasmid_4  from MILP.\n",
      "\n",
      "plasmid_5  from MILP completely contained in  plasmid_10  from greedy.\n",
      "\n",
      "plasmid_10  from greedy completely contained in  plasmid_5  from MILP.\n",
      "\n",
      "plasmid_12  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_12  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n",
      "plasmid_12  from greedy completely contained in  plasmid_6  from MILP.\n",
      "\n",
      "plasmid_12  from greedy completely contained in  plasmid_7  from MILP.\n",
      "\n",
      "plasmid_14  from greedy completely contained in  plasmid_0  from MILP.\n",
      "\n",
      "plasmid_18  from greedy completely contained in  plasmid_1  from MILP.\n",
      "\n",
      "plasmid_18  from greedy completely contained in  plasmid_4  from MILP.\n",
      "\n",
      "plasmid_20  from greedy completely contained in  plasmid_2  from MILP.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in greedy_plasmids:\n",
    "    x_chain = greedy_plasmids[x]['chain']\n",
    "    for y in MILP_plasmids:\n",
    "        y_chain = MILP_plasmids[y]['chain']\n",
    "        if set(y_chain) <= set(x_chain):\n",
    "            print(y, ' from MILP completely contained in ', x, ' from greedy.\\n')\n",
    "        if set(x_chain) <= set(y_chain):\n",
    "            print(x, ' from greedy completely contained in ', y, ' from MILP.\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
